<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>Class List</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li class="current"><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>Class List</h1>Here are the classes, structs, unions and interfaces with brief descriptions:<table>
  <tr><td class="indexkey"><a class="el" href="classChromosomes.html">Chromosomes</a></td><td class="indexvalue">This class is used to store chromosome DNA sequences </td></tr>
  <tr><td class="indexkey"><a class="el" href="classFileBuffer.html">FileBuffer</a></td><td class="indexvalue">Class for reading lines from file or standard input </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicInterval.html">GenomicInterval</a></td><td class="indexvalue">Set of CIGAR operations that map to the fragment sequence (SAM format) </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicIntervalSetAsArray.html">GenomicIntervalSetAsArray</a></td><td class="indexvalue">[UNDER DEVELOPMENT] This class implements an array of genomic intervals </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegion.html">GenomicRegion</a></td><td class="indexvalue">This class implements the genomic regions (i.e. ordered sets of genomic intervals) </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionBED.html">GenomicRegionBED</a></td><td class="indexvalue">This class processes genomic regions in BED format </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionBEDToREG.html">GenomicRegionBEDToREG</a></td><td class="indexvalue">This class reads BED files but converts them internally to the REG format, so all operations are performed on the REG format </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionGFF.html">GenomicRegionGFF</a></td><td class="indexvalue">This class processes genomic regions in SAM format </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionGFFToREG.html">GenomicRegionGFFToREG</a></td><td class="indexvalue">This class reads GFF files but converts them internally to the REG format, so all operations are performed on the REG format </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionSAM.html">GenomicRegionSAM</a></td><td class="indexvalue">This class processes genomic regions in SAM format </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionSAMToREG.html">GenomicRegionSAMToREG</a></td><td class="indexvalue">This class reads SAM files but converts them internally to the REG format, so all operations are performed on the REG format </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a></td><td class="indexvalue">This class is used to read and manipulate a set of genomic regions from a file or from standard input </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionSetOverlaps.html">GenomicRegionSetOverlaps</a></td><td class="indexvalue">Abstract class for computing overlaps between genomic regions </td></tr>
  <tr><td class="indexkey"><a class="el" href="classGenomicRegionSetScanner.html">GenomicRegionSetScanner</a></td><td class="indexvalue">This class is used to scan a set of regions by sliding windows. See detailed description below for an example </td></tr>
  <tr><td class="indexkey"><a class="el" href="classProgress.html">Progress</a></td><td class="indexvalue">This class is used to report progress of loop computations </td></tr>
  <tr><td class="indexkey"><a class="el" href="classSortedGenomicRegionSetOverlaps.html">SortedGenomicRegionSetOverlaps</a></td><td class="indexvalue">This class is used to find overlaps between two genomic regions sets. See detailed description below for an example </td></tr>
  <tr><td class="indexkey"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html">UnsortedGenomicRegionSetOverlaps</a></td><td class="indexvalue">Class for computing overlaps between unsorted regions </td></tr>
</table>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Sun Sep 25 22:09:23 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
