<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>GenomicRegionSetOverlapScanner Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>GenomicRegionSetOverlapScanner Class Reference</h1><!-- doxytag: class="GenomicRegionSetOverlapScanner" -->This class is used to find overlaps between two genomic regions sets. See detailed description below for an example.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>

<p>
<a href="classGenomicRegionSetOverlapScanner-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#268a52207c12005452c46e86167e4faf">GenomicRegionSetOverlapScanner</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlapScanner.html#f74bf7176b25f99da4c38e84f50efb75">QuerySet</a>, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlapScanner.html#6c9bb71f3f0479b0f773d13315a64162">IndexSet</a>, bool <a class="el" href="classGenomicRegionSetOverlapScanner.html#03774130ef410465b36be25ae28682e7">sorted_by_strand</a>)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor.  <a href="#268a52207c12005452c46e86167e4faf"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5670ab6cb78ef5034d3fd471d000aea0"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::~GenomicRegionSetOverlapScanner" ref="5670ab6cb78ef5034d3fd471d000aea0" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#5670ab6cb78ef5034d3fd471d000aea0">~GenomicRegionSetOverlapScanner</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3fdc80b8ce58503fbf56c3408dfe41a5"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::GetQuery" ref="3fdc80b8ce58503fbf56c3408dfe41a5" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#3fdc80b8ce58503fbf56c3408dfe41a5">GetQuery</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the current query region ('NULL' if no more query regions are available); also, <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> is automatically updated via <a class="el" href="classGenomicRegionSetOverlapScanner.html#2241b4411418e51f21dfbbff56332940">GenomicRegionSetOverlapScanner::LoadIndexBuffer</a>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a237159db6addc8f09c4cc1c078898af"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::NextQuery" ref="a237159db6addc8f09c4cc1c078898af" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#a237159db6addc8f09c4cc1c078898af">NextQuery</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the next query region ('NULL' if no more query regions are available); also, <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> is automatically updated via <a class="el" href="classGenomicRegionSetOverlapScanner.html#2241b4411418e51f21dfbbff56332940">GenomicRegionSetOverlapScanner::LoadIndexBuffer</a>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8fabf222feb13df18c4516a0c11e6c94"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::GetMatch" ref="8fabf222feb13df18c4516a0c11e6c94" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#8fabf222feb13df18c4516a0c11e6c94">GetMatch</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="59683f41c3512969c06ebe12c3c34892"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::NextMatch" ref="59683f41c3512969c06ebe12c3c34892" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#59683f41c3512969c06ebe12c3c34892">NextMatch</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#63d7837172475b7c75d1b5b28a4bca97">GetOverlap</a> (bool match_gaps, bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the current index region that overlaps the current query region.  <a href="#63d7837172475b7c75d1b5b28a4bca97"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#3358b5f338f92581b2e47f8e8ec9b0ef">NextOverlap</a> (bool match_gaps, bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals).  <a href="#3358b5f338f92581b2e47f8e8ec9b0ef"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="78c2866a7a33facc94aaa21513ae31cf"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::Done" ref="78c2866a7a33facc94aaa21513ae31cf" args="()" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#78c2866a7a33facc94aaa21513ae31cf">Done</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns true, if no more query or index regions are available and the <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> buffer is empty. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">unsigned long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#2b51a7827283cf09ce9d0a822714d738">CalcCoverage</a> (bool use_labels_as_values, bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates the total overlap between the current query region and the index region set.  <a href="#2b51a7827283cf09ce9d0a822714d738"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">unsigned long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#7826504ecdb81cf0828b428078ec1844">CountMatches</a> (bool use_labels_as_values, bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates the total number of matches between the current query region and the index region set.  <a href="#7826504ecdb81cf0828b428078ec1844"></a><br></td></tr>
<tr><td colspan="2"><br><h2>Private Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2241b4411418e51f21dfbbff56332940"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::LoadIndexBuffer" ref="2241b4411418e51f21dfbbff56332940" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#2241b4411418e51f21dfbbff56332940">LoadIndexBuffer</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">manages <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4c877e18eeeb397c5c2de2bca6ce0ff0"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::ClearIndexBuffer" ref="4c877e18eeeb397c5c2de2bca6ce0ff0" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#4c877e18eeeb397c5c2de2bca6ce0ff0">ClearIndexBuffer</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">clears <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> <br></td></tr>
<tr><td colspan="2"><br><h2>Private Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f74bf7176b25f99da4c38e84f50efb75"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::QuerySet" ref="f74bf7176b25f99da4c38e84f50efb75" args="" -->
<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#f74bf7176b25f99da4c38e84f50efb75">QuerySet</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">pointer to query region set <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6c9bb71f3f0479b0f773d13315a64162"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::IndexSet" ref="6c9bb71f3f0479b0f773d13315a64162" args="" -->
<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#6c9bb71f3f0479b0f773d13315a64162">IndexSet</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">pointer to index region set <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="03774130ef410465b36be25ae28682e7"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::sorted_by_strand" ref="03774130ef410465b36be25ae28682e7" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#03774130ef410465b36be25ae28682e7">sorted_by_strand</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">if true, query and index regions sorted by strand <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b802e705ad9fcfdb95fa665c26d1c610"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::current_qreg" ref="b802e705ad9fcfdb95fa665c26d1c610" args="" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#b802e705ad9fcfdb95fa665c26d1c610">current_qreg</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">pointer to the current query region <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="44fafd1bccf2b4610bf1979ca0b05c27"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::current_ireg" ref="44fafd1bccf2b4610bf1979ca0b05c27" args="" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#44fafd1bccf2b4610bf1979ca0b05c27">current_ireg</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">pointer to the current index region <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5cef1f4f621533d6ce6a47fc106f35cd"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::IRegBuffer" ref="5cef1f4f621533d6ce6a47fc106f35cd" args="" -->
GenomicRegionList&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">IRegBuffer</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">temporary buffer containing index regions that overlap with current query; the buffer is cleared only if the next query has no overlap <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ab1a082fe808c27618040dfa0537fb5e"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::IRegBufferIterator" ref="ab1a082fe808c27618040dfa0537fb5e" args="" -->
GenomicRegionList::iterator&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#ab1a082fe808c27618040dfa0537fb5e">IRegBufferIterator</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">iterator on <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4dc563efd11721122a7863d574801cc8"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::ireg_buffer_interval" ref="4dc563efd11721122a7863d574801cc8" args="" -->
<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#4dc563efd11721122a7863d574801cc8">ireg_buffer_interval</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">keeps tracks of the (linked) interval stored in <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4c8c433190c5f9a12754ae096ec2e8e5"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::max_ireg_buffer_size" ref="4c8c433190c5f9a12754ae096ec2e8e5" args="" -->
unsigned long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSetOverlapScanner.html#4c8c433190c5f9a12754ae096ec2e8e5">max_ireg_buffer_size</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">keeps track of the maximum size of <a class="el" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">GenomicRegionSetOverlapScanner::IRegBuffer</a> <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class is used to find overlaps between two genomic regions sets. See detailed description below for an example. 
<p>
A simple example for computing RNAseq read densities in known exons (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'density' operation): <div class="fragment"><pre class="fragment">    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *QUERY_REG_FILE = <span class="stringliteral">"rnaseq.reads.reg"</span>;
    <span class="keywordtype">char</span> *INDEX_REG_FILE = <span class="stringliteral">"exons.reg"</span>;
    <span class="keywordtype">double</span> MIN_DENSITY = 0.0;
    <span class="keywordtype">bool</span> SORTED_BY_STRAND = <span class="keyword">false</span>; 
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> QueryRegSet(QUERY_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> IndexRegSet(INDEX_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);

    <a class="code" href="classGenomicRegionSetOverlapScanner.html" title="This class is used to find overlaps between two genomic regions sets. See detailed...">GenomicRegionSetOverlapScanner</a> Overlaps(&amp;QueryRegSet,&amp;IndexRegSet,SORTED_BY_STRAND);
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">"Processing queries..."</span>,1);
    <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *qreg=Overlaps.GetQuery(); qreg!=NULL; qreg=Overlaps.NextQuery()) {
      <span class="keywordtype">double</span> density = (double)Overlaps.CalcCoverage(USE_VALUES,IGNORE_STRAND)/qreg-&gt;GetSize(); 
      <span class="keywordflow">if</span> (density&gt;=MIN_DENSITY) printf(<span class="stringliteral">"%s\t%.4e\n"</span>, qreg-&gt;LABEL, density);
      PRG.Check();
    }
    PRG.Done();
</pre></div> A more complex example for creating ChIPseq read profiles in TSS regions (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'offset' operation): <div class="fragment"><pre class="fragment">    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *QUERY_REG_FILE = <span class="stringliteral">"chipseq.reads.reg"</span>;
    <span class="keywordtype">char</span> *INDEX_REG_FILE = <span class="stringliteral">"TSS.flank10kb.reg"</span>;
    <span class="keywordtype">char</span> *OFFSET_OP = <span class="stringliteral">"5p"</span>;
    <span class="keywordtype">bool</span> SORTED_BY_STRAND = <span class="keyword">false</span>; 
    <span class="keywordtype">bool</span> MATCH_GAPS = <span class="keyword">true</span>;
    <span class="keywordtype">bool</span> IGNORE_STRAND = <span class="keyword">true</span>;
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> QueryRegSet(QUERY_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> IndexRegSet(INDEX_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);

    <a class="code" href="classGenomicRegionSetOverlapScanner.html" title="This class is used to find overlaps between two genomic regions sets. See detailed...">GenomicRegionSetOverlapScanner</a> Overlaps(&amp;QueryRegSet,&amp;IndexRegSet,SORTED_BY_STRAND);
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">"Processing queries..."</span>,1);
    <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *qreg=Overlaps.GetQuery(); Overlaps.Done()==<span class="keyword">false</span>; qreg=Overlaps.NextQuery()) {
      <span class="keywordtype">size_t</span> Qsize = qreg-&gt;GetSize();
      <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *ireg=Overlaps.GetOverlap(MATCH_GAPS,IGNORE_STRAND); ireg!=NULL; ireg=Overlaps.NextOverlap(MATCH_GAPS,IGNORE_STRAND)) {
        cout &lt;&lt; qreg-&gt;LABEL &lt;&lt; <span class="charliteral">'\t'</span>;
        <span class="keywordtype">long</span> <span class="keywordtype">int</span> start_offset, stop_offset;
        ireg-&gt;I.front()-&gt;GetOffsetFrom(qreg-&gt;I.front(),OFFSET_OP,IGNORE_STRAND,&amp;start_offset,&amp;stop_offset);
        printf(<span class="stringliteral">"%f"</span>, ((<span class="keywordtype">float</span>)start_offset/Qsize+(<span class="keywordtype">float</span>)stop_offset/Qsize)/2); 
        cout &lt;&lt; <span class="charliteral">'\n'</span>;
      }
      PRG.Check();
    }
    PRG.Done();
</pre></div> <hr><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" name="268a52207c12005452c46e86167e4faf"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::GenomicRegionSetOverlapScanner" ref="268a52207c12005452c46e86167e4faf" args="(GenomicRegionSet *QuerySet, GenomicRegionSet *IndexSet, bool sorted_by_strand)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">GenomicRegionSetOverlapScanner::GenomicRegionSetOverlapScanner           </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td>
          <td class="paramname"> <em>QuerySet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td>
          <td class="paramname"> <em>IndexSet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>sorted_by_strand</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Class constructor. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>QuerySet</em>&nbsp;</td><td>pointer to query region set </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>IndexSet</em>&nbsp;</td><td>pointer to index region set </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sorted_by_strand</em>&nbsp;</td><td>if true, query and index regions are sorted by strand </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr><h2>Member Function Documentation</h2>
<a class="anchor" name="63d7837172475b7c75d1b5b28a4bca97"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::GetOverlap" ref="63d7837172475b7c75d1b5b28a4bca97" args="(bool match_gaps, bool ignore_strand)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classGenomicRegion.html">GenomicRegion</a> * GenomicRegionSetOverlapScanner::GetOverlap           </td>
          <td>(</td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>match_gaps</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>ignore_strand</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Returns a pointer to the current index region that overlaps the current query region. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>match_gaps</em>&nbsp;</td><td>if 'true', overlaps are defined as in <a class="el" href="classGenomicRegionSetOverlapScanner.html#8fabf222feb13df18c4516a0c11e6c94">GetMatch</a>. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ignore_strand</em>&nbsp;</td><td>if 'true', overlaps are strand-ignorant </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="3358b5f338f92581b2e47f8e8ec9b0ef"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::NextOverlap" ref="3358b5f338f92581b2e47f8e8ec9b0ef" args="(bool match_gaps, bool ignore_strand)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classGenomicRegion.html">GenomicRegion</a> * GenomicRegionSetOverlapScanner::NextOverlap           </td>
          <td>(</td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>match_gaps</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>ignore_strand</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>match_gaps</em>&nbsp;</td><td>if 'true', overlaps are defined as in <a class="el" href="classGenomicRegionSetOverlapScanner.html#8fabf222feb13df18c4516a0c11e6c94">GetMatch</a>. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ignore_strand</em>&nbsp;</td><td>if 'true', overlaps are strand-ignorant </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="2b51a7827283cf09ce9d0a822714d738"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::CalcCoverage" ref="2b51a7827283cf09ce9d0a822714d738" args="(bool use_labels_as_values, bool ignore_strand)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long int GenomicRegionSetOverlapScanner::CalcCoverage           </td>
          <td>(</td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>use_labels_as_values</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>ignore_strand</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Calculates the total overlap between the current query region and the index region set. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>use_labels_as_values</em>&nbsp;</td><td>region labels contain number to be used in calculation </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ignore_strand</em>&nbsp;</td><td>if 'true', overlaps are strand-ignorant </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="7826504ecdb81cf0828b428078ec1844"></a><!-- doxytag: member="GenomicRegionSetOverlapScanner::CountMatches" ref="7826504ecdb81cf0828b428078ec1844" args="(bool use_labels_as_values, bool ignore_strand)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long int GenomicRegionSetOverlapScanner::CountMatches           </td>
          <td>(</td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>use_labels_as_values</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>ignore_strand</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Calculates the total number of matches between the current query region and the index region set. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>use_labels_as_values</em>&nbsp;</td><td>region labels contain number to be used in calculation </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ignore_strand</em>&nbsp;</td><td>if 'true', overlaps are strand-ignorant </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Mon Sep 12 16:49:29 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
