<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>GenomicRegion Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>GenomicRegion Class Reference</h1><!-- doxytag: class="GenomicRegion" -->This class implements the genomic regions (i.e. ordered sets of genomic intervals).  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>
<div class="dynheader">
Inheritance diagram for GenomicRegion:</div>
<div class="dynsection">

<p><center><img src="classGenomicRegion.png" usemap="#GenomicRegion_map" border="0" alt=""></center>
<map name="GenomicRegion_map">
<area href="classGenomicRegionBED.html" alt="GenomicRegionBED" shape="rect" coords="0,56,167,80">
<area href="classGenomicRegionBEDToREG.html" alt="GenomicRegionBEDToREG" shape="rect" coords="177,56,344,80">
<area href="classGenomicRegionGFF.html" alt="GenomicRegionGFF" shape="rect" coords="354,56,521,80">
<area href="classGenomicRegionGFFToREG.html" alt="GenomicRegionGFFToREG" shape="rect" coords="531,56,698,80">
<area href="classGenomicRegionSAM.html" alt="GenomicRegionSAM" shape="rect" coords="708,56,875,80">
<area href="classGenomicRegionSAMToREG.html" alt="GenomicRegionSAMToREG" shape="rect" coords="885,56,1052,80">
</map>
</div>

<p>
<a href="classGenomicRegion-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8ff9fe5618b932a1caccbd9072ddaaef"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="8ff9fe5618b932a1caccbd9072ddaaef" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#8ff9fe5618b932a1caccbd9072ddaaef">GenomicRegion</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Empty class constructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="c45e11c7c0d0477c4b8682ef5266d3d7"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="c45e11c7c0d0477c4b8682ef5266d3d7" args="(FileBuffer *B)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#c45e11c7c0d0477c4b8682ef5266d3d7">GenomicRegion</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor, see example below. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="c22559720502ea450aeb4f65c6c789e1"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="c22559720502ea450aeb4f65c6c789e1" args="(char *inp, long int n_line=0)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#c22559720502ea450aeb4f65c6c789e1">GenomicRegion</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor, see example below. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6de3d0ff71615c36a7311509af56ae79"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="6de3d0ff71615c36a7311509af56ae79" args="(char *label, GenomicInterval *i)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#6de3d0ff71615c36a7311509af56ae79">GenomicRegion</a> (char *label, <a class="el" href="classGenomicInterval.html">GenomicInterval</a> *i)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="37af413ecbed6ee0d36ecbc40d957947"></a><!-- doxytag: member="GenomicRegion::~GenomicRegion" ref="37af413ecbed6ee0d36ecbc40d957947" args="()" -->
virtual&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#37af413ecbed6ee0d36ecbc40d957947">~GenomicRegion</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fd188daab9b4e97ef66883d6971fe103"></a><!-- doxytag: member="GenomicRegion::Read" ref="fd188daab9b4e97ef66883d6971fe103" args="(char *inp, long int n_line=0, char del1=' ', char del2=',')" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#fd188daab9b4e97ef66883d6971fe103">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0, char del1=' ', char del2=',')</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reads label and genomic region intervals from input string <b>inp</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="40397d1fab7b00e5f04b9f7e52f0114e"></a><!-- doxytag: member="GenomicRegion::Print" ref="40397d1fab7b00e5f04b9f7e52f0114e" args="(FILE *file_ptr=stdout)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#40397d1fab7b00e5f04b9f7e52f0114e">Print</a> (FILE *file_ptr=stdout)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b9ea31af7fbb32f206efe0e64533a5be"></a><!-- doxytag: member="GenomicRegion::PrintModified" ref="b9ea31af7fbb32f206efe0e64533a5be" args="(char *label, long int start, long int stop)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#b9ea31af7fbb32f206efe0e64533a5be">PrintModified</a> (char *label, long int start, long int stop)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegion.html#40397d1fab7b00e5f04b9f7e52f0114e" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ba987d40590874eebd1dd8bb0c02bdc8"></a><!-- doxytag: member="GenomicRegion::PrintIntersection" ref="ba987d40590874eebd1dd8bb0c02bdc8" args="(GenomicRegion *r, bool ignore_strand, bool merge_labels)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ba987d40590874eebd1dd8bb0c02bdc8">PrintIntersection</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand, bool merge_labels)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints all pairwise intersection with the intervals in region <b>r</b> (compatible, sorted and non-overlapping). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="cfe5347d9c92da801ce7e97416fc8758"></a><!-- doxytag: member="GenomicRegion::PrintConstrained" ref="cfe5347d9c92da801ce7e97416fc8758" args="(GenomicRegion *r, bool merge_labels=false)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#cfe5347d9c92da801ce7e97416fc8758">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegion.html#40397d1fab7b00e5f04b9f7e52f0114e" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>r-&gt;I.front()-&gt;START</b> and <b>r-&gt;I.back()-&gt;STOP</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9c687565a8ed25da5a5ccf08e5e8461e"></a><!-- doxytag: member="GenomicRegion::PrintREG" ref="9c687565a8ed25da5a5ccf08e5e8461e" args="(bool compact=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#9c687565a8ed25da5a5ccf08e5e8461e">PrintREG</a> (bool compact=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9a00b031d4a119181b8b76ffa34a6f47"></a><!-- doxytag: member="GenomicRegion::PrintQ" ref="9a00b031d4a119181b8b76ffa34a6f47" args="(bool reverse=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#9a00b031d4a119181b8b76ffa34a6f47">PrintQ</a> (bool reverse=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints sequences (SEQ format). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="68ecc1c8c76e9204f6e27801f0ba3d5f"></a><!-- doxytag: member="GenomicRegion::PrintError" ref="68ecc1c8c76e9204f6e27801f0ba3d5f" args="(string error_msg)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#68ecc1c8c76e9204f6e27801f0ba3d5f">PrintError</a> (string error_msg)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints error message and exits. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ad8566a736786134b72e7bf74296e807"></a><!-- doxytag: member="GenomicRegion::DeleteIntervals" ref="ad8566a736786134b72e7bf74296e807" args="(GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ad8566a736786134b72e7bf74296e807">DeleteIntervals</a> (GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Deletes intervals from iterator <b>i</b> to <b>j</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bc547e770917e859fa434d6c54954916"></a><!-- doxytag: member="GenomicRegion::SetLabel" ref="bc547e770917e859fa434d6c54954916" args="(char *label)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#bc547e770917e859fa434d6c54954916">SetLabel</a> (char *label)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Changes the label. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="e795532404bad8b93bec632c20d12b9d"></a><!-- doxytag: member="GenomicRegion::SetStrand" ref="e795532404bad8b93bec632c20d12b9d" args="(char strand)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#e795532404bad8b93bec632c20d12b9d">SetStrand</a> (char strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Interval strands are set to <b>strand</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="815baa00d7d33f179181560a1fe468fc"></a><!-- doxytag: member="GenomicRegion::GetChromosome" ref="815baa00d7d33f179181560a1fe468fc" args="()" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#815baa00d7d33f179181560a1fe468fc">GetChromosome</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns pointer to chromosome name. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5bff7a3e83b0bc9bfb5441270d0ba1ae"></a><!-- doxytag: member="GenomicRegion::GetSize" ref="5bff7a3e83b0bc9bfb5441270d0ba1ae" args="()" -->
size_t&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#5bff7a3e83b0bc9bfb5441270d0ba1ae">GetSize</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns sum of interval's sizes. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="719143f1affa44c38493b2cd43448d3b"></a><!-- doxytag: member="GenomicRegion::GetSeqLength" ref="719143f1affa44c38493b2cd43448d3b" args="(Chromosomes *C)" -->
size_t&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#719143f1affa44c38493b2cd43448d3b">GetSeqLength</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns genomic region sequence length excluding 'N' characters. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fcf7c053989b87f425f084d1f2c078f5"></a><!-- doxytag: member="GenomicRegion::GetSeq" ref="fcf7c053989b87f425f084d1f2c078f5" args="(Chromosomes *C, bool replace=false)" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#fcf7c053989b87f425f084d1f2c078f5">GetSeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool replace=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns genomic region sequence. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="58c2fe162c1e9bbbd7c90d5abd011ff0"></a><!-- doxytag: member="GenomicRegion::IsCompatible" ref="58c2fe162c1e9bbbd7c90d5abd011ff0" args="(bool ignore_strand)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#58c2fe162c1e9bbbd7c90d5abd011ff0">IsCompatible</a> (bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns 'true' if all intervals are in the same chromosome and strand. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2ebd984a2632314a39a6e9ba2c497105"></a><!-- doxytag: member="GenomicRegion::IsCompatibleWith" ref="2ebd984a2632314a39a6e9ba2c497105" args="(GenomicRegion *r, bool ignore_strand)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#2ebd984a2632314a39a6e9ba2c497105">IsCompatibleWith</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns 'true' if all intervals are compatible with the intervals in <b>r</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2e0765dfbe651ad583faa8c8a77e047f"></a><!-- doxytag: member="GenomicRegion::IsCompatibleSorted" ref="2e0765dfbe651ad583faa8c8a77e047f" args="(bool ignore_strand)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#2e0765dfbe651ad583faa8c8a77e047f">IsCompatibleSorted</a> (bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns 'true' if intervals are compatible and sorted by chromosome, strand (if <b>ignore_strand</b> is false) and start position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fb2db74a3cd85a84533375483e12b96c"></a><!-- doxytag: member="GenomicRegion::IsCompatibleSortedAndNonoverlapping" ref="fb2db74a3cd85a84533375483e12b96c" args="()" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#fb2db74a3cd85a84533375483e12b96c">IsCompatibleSortedAndNonoverlapping</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns 'true' if intervals are compatible, sorted and non-overlapping. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9e3617a0046fe5fa8fce9f9cce8c3c37"></a><!-- doxytag: member="GenomicRegion::OverlapsWith" ref="9e3617a0046fe5fa8fce9f9cce8c3c37" args="(GenomicRegion *r, bool ignore_strand)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#9e3617a0046fe5fa8fce9f9cce8c3c37">OverlapsWith</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns 'true' if it overlaps with input region <b>r</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="25634257eb7a2ed379548b17a83414fb"></a><!-- doxytag: member="GenomicRegion::IsBefore" ref="25634257eb7a2ed379548b17a83414fb" args="(GenomicRegion *r, bool sorted_by_strand)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#25634257eb7a2ed379548b17a83414fb">IsBefore</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool sorted_by_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Compares with input region <b>r</b> and returns true if its order is before <b>r</b>. The order is determined first by chromosome, second by strand, then by start position and finally by stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2065cf74a82c2340b1e49e7f7ec0fdd4"></a><!-- doxytag: member="GenomicRegion::IsPosBefore" ref="2065cf74a82c2340b1e49e7f7ec0fdd4" args="(GenomicRegion *r)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#2065cf74a82c2340b1e49e7f7ec0fdd4">IsPosBefore</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Compares with input region <b>r</b> and returns true if its position order is before <b>r</b>. The position order is determined first by start position and then by stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="623e28aab180174d43873c4a85342ad4"></a><!-- doxytag: member="GenomicRegion::CalcOverlap" ref="623e28aab180174d43873c4a85342ad4" args="(GenomicRegion *r, bool ignore_strand)" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#623e28aab180174d43873c4a85342ad4">CalcOverlap</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns the total number of overlapping nucleotides (gaps are not matched) with input region <b>r</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="da548684fc4ae44dd3782aa15eaaf0a8"></a><!-- doxytag: member="GenomicRegion::CalcDirection" ref="da548684fc4ae44dd3782aa15eaaf0a8" args="(GenomicInterval *i, bool sorted_by_strand)" -->
int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#da548684fc4ae44dd3782aa15eaaf0a8">CalcDirection</a> (<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *i, bool sorted_by_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns -1 if before interval <b>i</b>, +1 if after, otherwise returns 0. The order is determined first by chromosome, then (optionally) by strand and finally by start position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b44366e44dc9a83da65b9593ce90dd32"></a><!-- doxytag: member="GenomicRegion::CalcDirection" ref="b44366e44dc9a83da65b9593ce90dd32" args="(GenomicRegion *r, bool sorted_by_strand)" -->
int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#b44366e44dc9a83da65b9593ce90dd32">CalcDirection</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool sorted_by_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns -1 if before region <b>r</b>, +1 if after, otherwise returns 0. The order is determined first by chromosome, then (optionally) by strand and finally by start position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="90b0382a60f0a8532af73fc18f41364b"></a><!-- doxytag: member="GenomicRegion::RunAlign" ref="90b0382a60f0a8532af73fc18f41364b" args="(Chromosomes *C)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#90b0382a60f0a8532af73fc18f41364b">RunAlign</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints alignments of input sequences to reference genome. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="405010d9256a3c08c53adee98133cb62"></a><!-- doxytag: member="GenomicRegion::PrintBEDFormat" ref="405010d9256a3c08c53adee98133cb62" args="(char *color, bool convert_chromosome)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#405010d9256a3c08c53adee98133cb62">PrintBEDFormat</a> (char *color, bool convert_chromosome)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints in BED format. If <b>convert_chromosome</b> is 'true', then convert from ENSEMBL to UCSC names (by adding 'chr' prefix to each chromosome name, and by converting 'MT' to 'chrM'). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="11b494e59775ac255ddb469c98f8a8a8"></a><!-- doxytag: member="GenomicRegion::ApplyBounds" ref="11b494e59775ac255ddb469c98f8a8a8" args="(StringLIntMap *bounds)" -->
virtual bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#11b494e59775ac255ddb469c98f8a8a8">ApplyBounds</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Corrects interval start/stop positions so as to comply with chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1e9a85a086afbe2608411450c9a4c5e6"></a><!-- doxytag: member="GenomicRegion::Center" ref="1e9a85a086afbe2608411450c9a4c5e6" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#1e9a85a086afbe2608411450c9a4c5e6">Center</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace each interval in the region with the corresponding center interval. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="c2c426782bd7c0d302b7a77bb1ff047f"></a><!-- doxytag: member="GenomicRegion::Connect" ref="c2c426782bd7c0d302b7a77bb1ff047f" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#c2c426782bd7c0d302b7a77bb1ff047f">Connect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace region with a single interval from minimum start to maximum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="869de97ecf059355fe5a5b572b5c2575"></a><!-- doxytag: member="GenomicRegion::Diff" ref="869de97ecf059355fe5a5b572b5c2575" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#869de97ecf059355fe5a5b572b5c2575">Diff</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="98695d5d1056d693a1c635b197f4b8bf"></a><!-- doxytag: member="GenomicRegion::RunCalcDistances" ref="98695d5d1056d693a1c635b197f4b8bf" args="(char *op1, char *op2)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#98695d5d1056d693a1c635b197f4b8bf">RunCalcDistances</a> (char *op1, char *op2)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print distance between successive intervals. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0785438db56d7b7cc7539470e85434dd"></a><!-- doxytag: member="GenomicRegion::Divide" ref="0785438db56d7b7cc7539470e85434dd" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#0785438db56d7b7cc7539470e85434dd">Divide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="45fe5d619f2e21644eacb6984aa47eff"></a><!-- doxytag: member="GenomicRegion::RunDivide" ref="45fe5d619f2e21644eacb6984aa47eff" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#45fe5d619f2e21644eacb6984aa47eff">RunDivide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="048d2182b212789b182c2c82a7ac131e"></a><!-- doxytag: member="GenomicRegion::Fix" ref="048d2182b212789b182c2c82a7ac131e" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#048d2182b212789b182c2c82a7ac131e">Fix</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Removes rogue intervals (e.g. start&gt;stop). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8075e77ab6296b59506857cc21ee4c9d"></a><!-- doxytag: member="GenomicRegion::Intersect" ref="8075e77ab6296b59506857cc21ee4c9d" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#8075e77ab6296b59506857cc21ee4c9d">Intersect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="08db190bc63e6f1d6fc90679dc5e6868"></a><!-- doxytag: member="GenomicRegion::RunSize" ref="08db190bc63e6f1d6fc90679dc5e6868" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#08db190bc63e6f1d6fc90679dc5e6868">RunSize</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints region label and sum of intervals' sizes. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#0721b07af0850057e4ab9cd416ecac2f">ModifyPos</a> (char *position_op, long int position_shift)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position.  <a href="#0721b07af0850057e4ab9cd416ecac2f"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="222b3b8f567c306ee5a71ea3a720ef43"></a><!-- doxytag: member="GenomicRegion::Randomize" ref="222b3b8f567c306ee5a71ea3a720ef43" args="(gsl_rng *random_generator, StringLIntMap *bounds)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#222b3b8f567c306ee5a71ea3a720ef43">Randomize</a> (gsl_rng *random_generator, StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Randomizes interval position within chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6120af435fb9ee68cb5a3d5066a4fda4"></a><!-- doxytag: member="GenomicRegion::ReversePos" ref="6120af435fb9ee68cb5a3d5066a4fda4" args="(StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#6120af435fb9ee68cb5a3d5066a4fda4">ReversePos</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reverses interval start/stop positions with respect to chromosomal bounds, as if the reference point were in the end of the chromosome as opposed to the beginning. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="461fbb00db1b45061e641f37614fb146"></a><!-- doxytag: member="GenomicRegion::Select" ref="461fbb00db1b45061e641f37614fb146" args="(bool first, bool last, bool from5p, bool from3p)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#461fbb00db1b45061e641f37614fb146">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#0ee8c165839c79afdc586f8b5e07788e">ShiftPos</a> (long int start_shift, long int stop_shift, bool strand_aware)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately.  <a href="#0ee8c165839c79afdc586f8b5e07788e"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b2c52ae40b34f3607f4558add94c25fd"></a><!-- doxytag: member="GenomicRegion::RunShuffle" ref="b2c52ae40b34f3607f4558add94c25fd" args="(gsl_rng *random_generator, GenomicRegionSet *refReg, StringLIntMap *index, StringVecLIntMap *loc)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#b2c52ae40b34f3607f4558add94c25fd">RunShuffle</a> (gsl_rng *random_generator, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints shuffled region within specified reference regions in <b>refReg</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="03959dc5b5695d817b5a40b47f30a4aa"></a><!-- doxytag: member="GenomicRegion::Sort" ref="03959dc5b5695d817b5a40b47f30a4aa" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#03959dc5b5695d817b5a40b47f30a4aa">Sort</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Sort intervals according to start position (only for compatible intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1686d4f4d2962b22d506a25a5887c31c"></a><!-- doxytag: member="GenomicRegion::RunSplit" ref="1686d4f4d2962b22d506a25a5887c31c" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#1686d4f4d2962b22d506a25a5887c31c">RunSplit</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print region intervals on separate lines. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="e19cd6eaa883e6517ba12567cd756320"></a><!-- doxytag: member="GenomicRegion::ModifyStrand" ref="e19cd6eaa883e6517ba12567cd756320" args="(char *strand_op)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#e19cd6eaa883e6517ba12567cd756320">ModifyStrand</a> (char *strand_op)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Modified strand information: '+'=only forward, '-'=only reverse, 'r'=reverse strands from '+' to '-' and vice versa, 'b'=both strands. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="05cb47d38946b780f7eb585074ccf393"></a><!-- doxytag: member="GenomicRegion::Union" ref="05cb47d38946b780f7eb585074ccf393" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#05cb47d38946b780f7eb585074ccf393">Union</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Compute intervals' union. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3d90c15accabfd5c0279f27f3513ba7c"></a><!-- doxytag: member="GenomicRegion::RunUnion" ref="3d90c15accabfd5c0279f27f3513ba7c" args="()" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#3d90c15accabfd5c0279f27f3513ba7c">RunUnion</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals' union. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3962ea71432f3977b1ac501e4e878480"></a><!-- doxytag: member="GenomicRegion::PrintWindows" ref="3962ea71432f3977b1ac501e4e878480" args="(long int win_step, long int win_size)" -->
virtual void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#3962ea71432f3977b1ac501e4e878480">PrintWindows</a> (long int win_step, long int win_size)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a2a5b753401e9f9d64a30cfbbdf23f60"></a><!-- doxytag: member="GenomicRegion::PrintRemoveN" ref="a2a5b753401e9f9d64a30cfbbdf23f60" args="(Chromosomes *C)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a2a5b753401e9f9d64a30cfbbdf23f60">PrintRemoveN</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Remove sub-intervals that correspond to sequences of 'N' characters. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4b52d78052e0109b1fceafe450cfdedd"></a><!-- doxytag: member="GenomicRegion::PrintSeq" ref="4b52d78052e0109b1fceafe450cfdedd" args="(Chromosomes *C, bool replace=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#4b52d78052e0109b1fceafe450cfdedd">PrintSeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool replace=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints region label, intervals and extracted sequence in SEQ format. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#33d1e5544b3fb81e8c19468b91920b1d">PrintOffsetFormat</a> (char *op, bool fraction)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates the offset distances of interval start and stop position with respect to the first interval in the set.  <a href="#33d1e5544b3fb81e8c19468b91920b1d"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7a8ac4f469447256253bdbdcb1b4a956"></a><!-- doxytag: member="GenomicRegion::PrintSeqLength" ref="7a8ac4f469447256253bdbdcb1b4a956" args="(Chromosomes *C)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#7a8ac4f469447256253bdbdcb1b4a956">PrintSeqLength</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints region label and sequence length (excluding 'N' characters). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="77fa696d55d538b365d15a7a30391132"></a><!-- doxytag: member="GenomicRegion::PrintVerifySeq" ref="77fa696d55d538b365d15a7a30391132" args="(Chromosomes *C, bool ignore)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#77fa696d55d538b365d15a7a30391132">PrintVerifySeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool ignore)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Verifies extracted sequence against region label. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9304039f70c5110b961cc9de77974d09"></a><!-- doxytag: member="GenomicRegion::ReverseStrand" ref="9304039f70c5110b961cc9de77974d09" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#9304039f70c5110b961cc9de77974d09">ReverseStrand</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">If interval strand is '+', it is changed to '-' and vice versa. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="50f5ba99404da1cdae6a8632a19ff837"></a><!-- doxytag: member="GenomicRegion::Constrain" ref="50f5ba99404da1cdae6a8632a19ff837" args="(GenomicRegion *r, char *label=NULL)" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#50f5ba99404da1cdae6a8632a19ff837">Constrain</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, char *label=NULL)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a version of this genomic region constrained by the start/stop coordinates. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="44cf789afda4e77749019045050e5ec7"></a><!-- doxytag: member="GenomicRegion::Diff" ref="44cf789afda4e77749019045050e5ec7" args="(GenomicRegion *r)" -->
virtual <a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#44cf789afda4e77749019045050e5ec7">Diff</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns the difference of this genomic region and region <b>r</b>. <br></td></tr>
<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="efe2255aeed5338060190ded05cb9c0c"></a><!-- doxytag: member="GenomicRegion::n_line" ref="efe2255aeed5338060190ded05cb9c0c" args="" -->
unsigned long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">keeps track of line number <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7eeba95c1e87e100346688681e30ff24"></a><!-- doxytag: member="GenomicRegion::LABEL" ref="7eeba95c1e87e100346688681e30ff24" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#7eeba95c1e87e100346688681e30ff24">LABEL</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">genomic region label <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b10b86b03c258958818b00c21e3672a9"></a><!-- doxytag: member="GenomicRegion::I" ref="b10b86b03c258958818b00c21e3672a9" args="" -->
GenomicIntervalSet&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9">I</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">set of intervals <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class implements the genomic regions (i.e. ordered sets of genomic intervals). 
<p>
Example: <div class="fragment"><pre class="fragment">    GenomicRegionREG x(<span class="stringliteral">"Region#1\tchr1 + 132034 135932 chr1 + 135832 140102"</span>);          <span class="comment">// standard REG format</span>
    GenomicRegionREG y(<span class="stringliteral">"Region#2\tchr1 + 132034,135832 135932,140102"</span>);                         <span class="comment">// compact REG format</span>
    cout &lt;&lt; <span class="stringliteral">"x = \n"</span>; x.Print();
    cout &lt;&lt; <span class="stringliteral">"y = \n"</span>; y.Print(); 
</pre></div> <hr><h2>Member Function Documentation</h2>
<a class="anchor" name="0721b07af0850057e4ab9cd416ecac2f"></a><!-- doxytag: member="GenomicRegion::ModifyPos" ref="0721b07af0850057e4ab9cd416ecac2f" args="(char *position_op, long int position_shift)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegion::ModifyPos           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>position_op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>position_shift</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>position_op</em>&nbsp;</td><td>selects fixed position as follows: '1'=start position, 'c'=center position, '5p'=5-prime, '3p'=3-prime </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>position_shift</em>&nbsp;</td><td>determines the shift of the non-fixed position (in the 'c' case, both positions are shifted symmetrically around the center) </td></tr>
  </table>
</dl>

<p>Reimplemented in <a class="el" href="classGenomicRegionBED.html#c515c70f443db400f911452ed359433b">GenomicRegionBED</a>, and <a class="el" href="classGenomicRegionSAM.html#353207352073db00dee0a9b620dca197">GenomicRegionSAM</a>.</p>

</div>
</div><p>
<a class="anchor" name="0ee8c165839c79afdc586f8b5e07788e"></a><!-- doxytag: member="GenomicRegion::ShiftPos" ref="0ee8c165839c79afdc586f8b5e07788e" args="(long int start_shift, long int stop_shift, bool strand_aware)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegion::ShiftPos           </td>
          <td>(</td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>start_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>stop_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>strand_aware</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>start_shift</em>&nbsp;</td><td>determines the shift of the start position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>stop_shift</em>&nbsp;</td><td>determines the shift of the stop position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>strand_aware</em>&nbsp;</td><td>if 'true', then start=5-prime and stop=3-prime </td></tr>
  </table>
</dl>

<p>Reimplemented in <a class="el" href="classGenomicRegionBED.html#fd2f05cec2af5186794f40f221e040b1">GenomicRegionBED</a>, and <a class="el" href="classGenomicRegionSAM.html#fb2701ba1a521ae2c07ea0ace2f9ee77">GenomicRegionSAM</a>.</p>

</div>
</div><p>
<a class="anchor" name="33d1e5544b3fb81e8c19468b91920b1d"></a><!-- doxytag: member="GenomicRegion::PrintOffsetFormat" ref="33d1e5544b3fb81e8c19468b91920b1d" args="(char *op, bool fraction)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegion::PrintOffsetFormat           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>fraction</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Calculates the offset distances of interval start and stop position with respect to the first interval in the set. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>op</em>&nbsp;</td><td>determines reference point as follows: '1'=start position, '2'=stop position, '5p'=5-prime position, '3p'=3-prime position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>fraction</em>&nbsp;</td><td>if 'true', offsets are reported as a fraction of the total region length </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Sun Sep 25 22:09:25 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
