\section{Class List}
Here are the classes, structs, unions and interfaces with brief descriptions:\begin{CompactList}
\item\contentsline{section}{\hyperlink{classChromosomes}{Chromosomes} (This class is used to store chromosome DNA sequences )}{\pageref{classChromosomes}}{}
\item\contentsline{section}{\hyperlink{classFileBuffer}{FileBuffer} (Class for reading lines from file or standard input )}{\pageref{classFileBuffer}}{}
\item\contentsline{section}{\hyperlink{classGenomicInterval}{GenomicInterval} (Set of CIGAR operations that map to the fragment sequence (SAM format) )}{\pageref{classGenomicInterval}}{}
\item\contentsline{section}{\hyperlink{classGenomicIntervalSetAsArray}{GenomicIntervalSetAsArray} (UNDER DEVELOPMENT: This class implements the an array of genomic intervals )}{\pageref{classGenomicIntervalSetAsArray}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegion}{GenomicRegion} (This class implements the genomic regions (i.e. ordered sets of genomic intervals) )}{\pageref{classGenomicRegion}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionBED}{GenomicRegionBED} (This class processes genomic regions in BED format )}{\pageref{classGenomicRegionBED}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionBEDToREG}{GenomicRegionBEDToREG} (This class reads BED files but converts them internally to the REG format, so all operations are performed on the REG format )}{\pageref{classGenomicRegionBEDToREG}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionGFF}{GenomicRegionGFF} (This class processes genomic regions in SAM format )}{\pageref{classGenomicRegionGFF}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionGFFToREG}{GenomicRegionGFFToREG} (This class reads GFF files but converts them internally to the REG format, so all operations are performed on the REG format )}{\pageref{classGenomicRegionGFFToREG}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionSAM}{GenomicRegionSAM} (This class processes genomic regions in SAM format )}{\pageref{classGenomicRegionSAM}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionSAMToREG}{GenomicRegionSAMToREG} (This class reads SAM files but converts them internally to the REG format, so all operations are performed on the REG format )}{\pageref{classGenomicRegionSAMToREG}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionSet}{GenomicRegionSet} (This class is used to read and manipulate a set of genomic regions from a file or from standard input )}{\pageref{classGenomicRegionSet}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionSetOverlapScanner}{GenomicRegionSetOverlapScanner} (This class is used to find overlaps between two genomic regions sets. See detailed description below for an example )}{\pageref{classGenomicRegionSetOverlapScanner}}{}
\item\contentsline{section}{\hyperlink{classGenomicRegionSetScanner}{GenomicRegionSetScanner} (This class is used to scan a set of regions by sliding windows. See detailed description below for an example )}{\pageref{classGenomicRegionSetScanner}}{}
\item\contentsline{section}{\hyperlink{classIndexedGenomicRegionSet}{IndexedGenomicRegionSet} (\mbox{[}UNDER DEVELOPMENT\mbox{]} An extension of \hyperlink{classGenomicRegionSet}{GenomicRegionSet} has an index for quick searching )}{\pageref{classIndexedGenomicRegionSet}}{}
\item\contentsline{section}{\hyperlink{classProgress}{Progress} (This class is used to report progress of loop computations )}{\pageref{classProgress}}{}
\end{CompactList}
