\hypertarget{classUnsortedGenomicRegionSetOverlaps}{
\section{UnsortedGenomicRegionSetOverlaps Class Reference}
\label{classUnsortedGenomicRegionSetOverlaps}\index{UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}}
}
Class for computing overlaps between unsorted regions.  


{\tt \#include $<$genomic\_\-intervals.h$>$}

Inheritance diagram for UnsortedGenomicRegionSetOverlaps::\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[height=2cm]{classUnsortedGenomicRegionSetOverlaps}
\end{center}
\end{figure}
\subsection*{Public Types}
\begin{CompactItemize}
\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_0f0857dd93ad3c047d1f5056a57edc14}{
typedef pair$<$ long int $\ast$, long int $\ast$$\ast$ $>$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_0f0857dd93ad3c047d1f5056a57edc14}{BinSet}}
\label{classUnsortedGenomicRegionSetOverlaps_0f0857dd93ad3c047d1f5056a57edc14}

\begin{CompactList}\small\item\em a two-dimensional grid of bins \item\end{CompactList}\end{CompactItemize}
\subsection*{Public Member Functions}
\begin{CompactItemize}
\item 
\hyperlink{classUnsortedGenomicRegionSetOverlaps_6a1255dfaac34080fae7880981b47acb}{UnsortedGenomicRegionSetOverlaps} (\hyperlink{classGenomicRegionSet}{GenomicRegionSet} $\ast$\hyperlink{classGenomicRegionSetOverlaps_e513304379055f6d379bc5907733dbe2}{QuerySet}, \hyperlink{classGenomicRegionSet}{GenomicRegionSet} $\ast$\hyperlink{classGenomicRegionSetOverlaps_c587bf854c827381493735b473622e03}{IndexSet})
\begin{CompactList}\small\item\em Class constructor. \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_7682171bb52c651af44e4e8f3bf33f52}{
\hyperlink{classUnsortedGenomicRegionSetOverlaps_7682171bb52c651af44e4e8f3bf33f52}{$\sim$UnsortedGenomicRegionSetOverlaps} ()}
\label{classUnsortedGenomicRegionSetOverlaps_7682171bb52c651af44e4e8f3bf33f52}

\begin{CompactList}\small\item\em Class destructor. \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_bf7e2f8d35f7a9cf8441caa9d765de44}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_bf7e2f8d35f7a9cf8441caa9d765de44}{GetQuery} ()}
\label{classUnsortedGenomicRegionSetOverlaps_bf7e2f8d35f7a9cf8441caa9d765de44}

\begin{CompactList}\small\item\em Returns a pointer to the current query region ('NULL' if no more query regions are available). \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_96e68ef99a958665f5da74572edcc854}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_96e68ef99a958665f5da74572edcc854}{NextQuery} ()}
\label{classUnsortedGenomicRegionSetOverlaps_96e68ef99a958665f5da74572edcc854}

\begin{CompactList}\small\item\em Returns a pointer to the next query region ('NULL' if no more query regions are available). \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_61c2c6e9203acb0e0a9bfa4f0176308f}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_61c2c6e9203acb0e0a9bfa4f0176308f}{GetMatch} ()}
\label{classUnsortedGenomicRegionSetOverlaps_61c2c6e9203acb0e0a9bfa4f0176308f}

\begin{CompactList}\small\item\em Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_0bcd12697611caa5216db7a4c4fca4e2}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_0bcd12697611caa5216db7a4c4fca4e2}{NextMatch} ()}
\label{classUnsortedGenomicRegionSetOverlaps_0bcd12697611caa5216db7a4c4fca4e2}

\begin{CompactList}\small\item\em Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_fc18fc203debad6a2d48200ea8ccb42f}{
bool \hyperlink{classUnsortedGenomicRegionSetOverlaps_fc18fc203debad6a2d48200ea8ccb42f}{Done} ()}
\label{classUnsortedGenomicRegionSetOverlaps_fc18fc203debad6a2d48200ea8ccb42f}

\begin{CompactList}\small\item\em Returns true, if no more query or index regions are available and the \hyperlink{classSortedGenomicRegionSetOverlaps_19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} buffer is empty. \item\end{CompactList}\end{CompactItemize}
\subsection*{Public Attributes}
\begin{CompactItemize}
\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_251e09b7f0a13a56b902c30791228ce4}{
long int $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_251e09b7f0a13a56b902c30791228ce4}{r\_\-next}}
\label{classUnsortedGenomicRegionSetOverlaps_251e09b7f0a13a56b902c30791228ce4}

\begin{CompactList}\small\item\em for each region store, store a pointer to the \char`\"{}next\char`\"{} one in the same bin \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_1cc721de471ed40dcc7d8eae258e5833}{
int \hyperlink{classUnsortedGenomicRegionSetOverlaps_1cc721de471ed40dcc7d8eae258e5833}{n\_\-levels}}
\label{classUnsortedGenomicRegionSetOverlaps_1cc721de471ed40dcc7d8eae258e5833}

\begin{CompactList}\small\item\em number of levels in the two-dimensional grid of bins \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_76ecebfcf839a940bdae26069ceef748}{
int $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_76ecebfcf839a940bdae26069ceef748}{n\_\-bits}}
\label{classUnsortedGenomicRegionSetOverlaps_76ecebfcf839a940bdae26069ceef748}

\begin{CompactList}\small\item\em number of bits (per level) by which the start position is right-shifted to compute its bin \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_beeb94562abdf4aebc36684959a5bc78}{
map$<$ string, \hyperlink{classUnsortedGenomicRegionSetOverlaps_0f0857dd93ad3c047d1f5056a57edc14}{BinSet} $\ast$ $>$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_beeb94562abdf4aebc36684959a5bc78}{index}}
\label{classUnsortedGenomicRegionSetOverlaps_beeb94562abdf4aebc36684959a5bc78}

\begin{CompactList}\small\item\em the two-dimensional grid of bins per chromosome \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_6bbb87f3c66d04527f06a5761df1ab56}{
\hyperlink{classUnsortedGenomicRegionSetOverlaps_0f0857dd93ad3c047d1f5056a57edc14}{BinSet} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_6bbb87f3c66d04527f06a5761df1ab56}{current\_\-binset}}
\label{classUnsortedGenomicRegionSetOverlaps_6bbb87f3c66d04527f06a5761df1ab56}

\begin{CompactList}\small\item\em pointer to the current chromosome set of bins \item\end{CompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_bafc544ce2c677ce071fa4e30ec2ecaf}{
bool \hyperlink{classUnsortedGenomicRegionSetOverlaps_bafc544ce2c677ce071fa4e30ec2ecaf}{new\_\-query}}
\label{classUnsortedGenomicRegionSetOverlaps_bafc544ce2c677ce071fa4e30ec2ecaf}

\begin{CompactList}\small\item\em 'true', if a new query is about to be processed \item\end{CompactList}\end{CompactItemize}


\subsection{Detailed Description}
Class for computing overlaps between unsorted regions. 

A simple example for computing RNAseq read densities in known exons (this is actually implemented in the {\bf genomic\_\-overlaps} command-line tool as 'density' operation): 

\begin{Code}\begin{verbatim}    #include "core.h"
    #include "genomic_intervals.h"
    
    // open region sets
    char *REF_REG_FILE = argv[next_arg];
    char *TEST_REG_FILE = next_arg+1==argc ? NULL : argv[next_arg+1];
    GenomicRegionSet RefRegSet(REF_REG_FILE,BUFFER_SIZE,VERBOSE,true);
    GenomicRegionSet TestRegSet(TEST_REG_FILE,BUFFER_SIZE,VERBOSE,false);

    // process overlaps
    GenomicRegionSetOverlaps *overlaps = new UnsortedGenomicRegionSetOverlaps(&TestRegSet,&RefRegSet);
    unsigned long int *coverage = overlaps->CalcIndexCoverage(MATCH_GAPS,IGNORE_STRAND,USE_VALUES); 
    Progress PRG("Printing densities...",RefRegSet.n_regions);
    for (long int k=0; k<RefRegSet.n_regions; k++) {
      GenomicRegion *qreg = RefRegSet.R[k];
      long int qreg_size = MATCH_GAPS ? (qreg->I.back()->STOP-qreg->I.front()->START+1) : qreg->GetSize();
      double density = (double)coverage[k]/qreg_size;
      if (density>=MIN_DENSITY) printf("%s\t%.4e\n", qreg->LABEL, density);
      PRG.Check();
    }
    PRG.Done();
    delete coverage;
    delete overlaps;
\end{verbatim}
\end{Code}

 

\subsection{Constructor \& Destructor Documentation}
\hypertarget{classUnsortedGenomicRegionSetOverlaps_6a1255dfaac34080fae7880981b47acb}{
\index{UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}!UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}}
\index{UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}!UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}}
\subsubsection[UnsortedGenomicRegionSetOverlaps]{\setlength{\rightskip}{0pt plus 5cm}UnsortedGenomicRegionSetOverlaps::UnsortedGenomicRegionSetOverlaps ({\bf GenomicRegionSet} $\ast$ {\em QuerySet}, \/  {\bf GenomicRegionSet} $\ast$ {\em IndexSet})}}
\label{classUnsortedGenomicRegionSetOverlaps_6a1255dfaac34080fae7880981b47acb}


Class constructor. 

\begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em QuerySet}]pointer to query region set \item[{\em IndexSet}]pointer to index region set \end{description}
\end{Desc}


The documentation for this class was generated from the following files:\begin{CompactItemize}
\item 
genomic\_\-intervals.h\item 
genomic\_\-intervals.cpp\end{CompactItemize}
