<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>GenomicRegionGFF Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>GenomicRegionGFF Class Reference</h1><!-- doxytag: class="GenomicRegionGFF" --><!-- doxytag: inherits="GenomicRegion" -->This class processes genomic regions in SAM format.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>
<div class="dynheader">
Inheritance diagram for GenomicRegionGFF:</div>
<div class="dynsection">

<p><center><img src="classGenomicRegionGFF.png" usemap="#GenomicRegionGFF_map" border="0" alt=""></center>
<map name="GenomicRegionGFF_map">
<area href="classGenomicRegion.html" alt="GenomicRegion" shape="rect" coords="0,0,125,24">
</map>
</div>

<p>
<a href="classGenomicRegionGFF-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9a49f0dc39116964f774e7ce78cfc827"></a><!-- doxytag: member="GenomicRegionGFF::GenomicRegionGFF" ref="9a49f0dc39116964f774e7ce78cfc827" args="(FileBuffer *B)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#9a49f0dc39116964f774e7ce78cfc827">GenomicRegionGFF</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="59dee8d6f9e73b1da46386497030beb2"></a><!-- doxytag: member="GenomicRegionGFF::GenomicRegionGFF" ref="59dee8d6f9e73b1da46386497030beb2" args="(char *inp, long int n_line=0)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#59dee8d6f9e73b1da46386497030beb2">GenomicRegionGFF</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3885d42a347283a5fed52791988076ad"></a><!-- doxytag: member="GenomicRegionGFF::~GenomicRegionGFF" ref="3885d42a347283a5fed52791988076ad" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#3885d42a347283a5fed52791988076ad">~GenomicRegionGFF</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="eb1390d42b388e32d98f58f57a5c169c"></a><!-- doxytag: member="GenomicRegionGFF::Read" ref="eb1390d42b388e32d98f58f57a5c169c" args="(char *inp, long int n_line=0)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#eb1390d42b388e32d98f58f57a5c169c">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reads GFF format from input string <b>inp</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3ffa71d729e4ff060a15a6bab1e924c8"></a><!-- doxytag: member="GenomicRegionGFF::Print" ref="3ffa71d729e4ff060a15a6bab1e924c8" args="(FILE *file_ptr=stdout)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#3ffa71d729e4ff060a15a6bab1e924c8">Print</a> (FILE *file_ptr=stdout)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="96539f4372048c9f0d87826f5856a8ef"></a><!-- doxytag: member="GenomicRegionGFF::PrintConstrained" ref="96539f4372048c9f0d87826f5856a8ef" args="(GenomicRegion *r, bool merge_labels=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#96539f4372048c9f0d87826f5856a8ef">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionGFF.html#3ffa71d729e4ff060a15a6bab1e924c8" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>start</b> and <b>stop</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4b331cd12506aa4793a2cf6ea7620936"></a><!-- doxytag: member="GenomicRegionGFF::PrintModified" ref="4b331cd12506aa4793a2cf6ea7620936" args="(char *label, long int start, long int stop)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#4b331cd12506aa4793a2cf6ea7620936">PrintModified</a> (char *label, long int start, long int stop)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionGFF.html#3ffa71d729e4ff060a15a6bab1e924c8" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ee70fcc81c9a6629bf0aa1ec37cb1d64"></a><!-- doxytag: member="GenomicRegionGFF::Diff" ref="ee70fcc81c9a6629bf0aa1ec37cb1d64" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#ee70fcc81c9a6629bf0aa1ec37cb1d64">Diff</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0ff4768578a03011fabad309072fade4"></a><!-- doxytag: member="GenomicRegionGFF::RunCalcDistances" ref="0ff4768578a03011fabad309072fade4" args="(char *op1, char *op2)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#0ff4768578a03011fabad309072fade4">RunCalcDistances</a> (char *op1, char *op2)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print distance between successive intervals. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="77975c03dd39b404098eb519e81f67d4"></a><!-- doxytag: member="GenomicRegionGFF::Divide" ref="77975c03dd39b404098eb519e81f67d4" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#77975c03dd39b404098eb519e81f67d4">Divide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d4acc1231677efd876084ef36db82de1"></a><!-- doxytag: member="GenomicRegionGFF::RunDivide" ref="d4acc1231677efd876084ef36db82de1" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#d4acc1231677efd876084ef36db82de1">RunDivide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d35d9b6e9c57270c0d5648ae9f702959"></a><!-- doxytag: member="GenomicRegionGFF::Intersect" ref="d35d9b6e9c57270c0d5648ae9f702959" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#d35d9b6e9c57270c0d5648ae9f702959">Intersect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="aa9b1271561e2d62ef1be6464e589cf4"></a><!-- doxytag: member="GenomicRegionGFF::Select" ref="aa9b1271561e2d62ef1be6464e589cf4" args="(bool first, bool last, bool from5p, bool from3p)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#aa9b1271561e2d62ef1be6464e589cf4">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bc08552b0ad4ca512f79064b8cd64e54"></a><!-- doxytag: member="GenomicRegionGFF::RunSplit" ref="bc08552b0ad4ca512f79064b8cd64e54" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#bc08552b0ad4ca512f79064b8cd64e54">RunSplit</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print region intervals on separate lines. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3815f1b27aef231a674c527e11f25880"></a><!-- doxytag: member="GenomicRegionGFF::PrintWindows" ref="3815f1b27aef231a674c527e11f25880" args="(long int win_step, long int win_size)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#3815f1b27aef231a674c527e11f25880">PrintWindows</a> (long int win_step, long int win_size)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions). <br></td></tr>
<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="aec963f555e7e807fb5046566622f5ee"></a><!-- doxytag: member="GenomicRegionGFF::n_tokens" ref="aec963f555e7e807fb5046566622f5ee" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#aec963f555e7e807fb5046566622f5ee">n_tokens</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">number of TAB-separated fields in GFF entry (= 11) <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2f0d4a5f15d8a756233436af03341734"></a><!-- doxytag: member="GenomicRegionGFF::SOURCE" ref="2f0d4a5f15d8a756233436af03341734" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#2f0d4a5f15d8a756233436af03341734">SOURCE</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">feature source <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0de447fbfad3da5b7a9cf07a5a6980f0"></a><!-- doxytag: member="GenomicRegionGFF::FEATURE" ref="0de447fbfad3da5b7a9cf07a5a6980f0" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#0de447fbfad3da5b7a9cf07a5a6980f0">FEATURE</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">feture type <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0e92da1f0fd153504e0630f27229ba4d"></a><!-- doxytag: member="GenomicRegionGFF::SCORE" ref="0e92da1f0fd153504e0630f27229ba4d" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#0e92da1f0fd153504e0630f27229ba4d">SCORE</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">score <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="42122a69eec385e0106e3861cb3b98f3"></a><!-- doxytag: member="GenomicRegionGFF::FRAME" ref="42122a69eec385e0106e3861cb3b98f3" args="" -->
char&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#42122a69eec385e0106e3861cb3b98f3">FRAME</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">coding frame <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="625a023f1bb24fc597e723919a036073"></a><!-- doxytag: member="GenomicRegionGFF::COMMENT" ref="625a023f1bb24fc597e723919a036073" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#625a023f1bb24fc597e723919a036073">COMMENT</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">comment <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class processes genomic regions in SAM format. <hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Tue Sep 20 18:46:13 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
