<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>GenomicRegionSAM Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>GenomicRegionSAM Class Reference</h1><!-- doxytag: class="GenomicRegionSAM" --><!-- doxytag: inherits="GenomicRegion" -->This class processes genomic regions in SAM format.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>
<div class="dynheader">
Inheritance diagram for GenomicRegionSAM:</div>
<div class="dynsection">

<p><center><img src="classGenomicRegionSAM.png" usemap="#GenomicRegionSAM_map" border="0" alt=""></center>
<map name="GenomicRegionSAM_map">
<area href="classGenomicRegion.html" alt="GenomicRegion" shape="rect" coords="0,0,128,24">
</map>
</div>

<p>
<a href="classGenomicRegionSAM-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Types</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="adf8580c5362304ab4866fa76a59df96"></a><!-- doxytag: member="GenomicRegionSAM::CIGARTokens" ref="adf8580c5362304ab4866fa76a59df96" args="" -->
typedef vector&lt; pair&lt; long int, <br>
char &gt; &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#adf8580c5362304ab4866fa76a59df96">CIGARTokens</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">This type is used for storing CIGAR tokens, i.e. (token_type, token_length). <br></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="e6b0828926a573743e7292567346c1a1"></a><!-- doxytag: member="GenomicRegionSAM::GenomicRegionSAM" ref="e6b0828926a573743e7292567346c1a1" args="(FileBuffer *B)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#e6b0828926a573743e7292567346c1a1">GenomicRegionSAM</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="25247d97947a54583fb17d493df4a0cf"></a><!-- doxytag: member="GenomicRegionSAM::GenomicRegionSAM" ref="25247d97947a54583fb17d493df4a0cf" args="(char *inp, long int n_line=0)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#25247d97947a54583fb17d493df4a0cf">GenomicRegionSAM</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1d11cc22134d13dbf67ffe4c7220888d"></a><!-- doxytag: member="GenomicRegionSAM::~GenomicRegionSAM" ref="1d11cc22134d13dbf67ffe4c7220888d" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#1d11cc22134d13dbf67ffe4c7220888d">~GenomicRegionSAM</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="096d90b6743400daf1acb50bcf17086a"></a><!-- doxytag: member="GenomicRegionSAM::Read" ref="096d90b6743400daf1acb50bcf17086a" args="(char *inp, long int n_line=0)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#096d90b6743400daf1acb50bcf17086a">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reads SAM format from input string <b>inp</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4381c6486f767b6e4055cb729beee181"></a><!-- doxytag: member="GenomicRegionSAM::Print" ref="4381c6486f767b6e4055cb729beee181" args="(FILE *file_ptr=stdout)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#4381c6486f767b6e4055cb729beee181">Print</a> (FILE *file_ptr=stdout)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="df8beab78f541121e7b56957afffa4ac"></a><!-- doxytag: member="GenomicRegionSAM::PrintConstrained" ref="df8beab78f541121e7b56957afffa4ac" args="(GenomicRegion *r, bool merge_labels=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#df8beab78f541121e7b56957afffa4ac">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionSAM.html#4381c6486f767b6e4055cb729beee181" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>start</b> and <b>stop</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ecf1d74c572e85bbe9bca0b30c6a6dda"></a><!-- doxytag: member="GenomicRegionSAM::PrintModified" ref="ecf1d74c572e85bbe9bca0b30c6a6dda" args="(char *label, long int start, long int stop)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ecf1d74c572e85bbe9bca0b30c6a6dda">PrintModified</a> (char *label, long int start, long int stop)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionSAM.html#4381c6486f767b6e4055cb729beee181" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6d0949a836520073de57e0f968a6b819"></a><!-- doxytag: member="GenomicRegionSAM::CalcStrandFromFlag" ref="6d0949a836520073de57e0f968a6b819" args="(unsigned long int flag)" -->
char&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#6d0949a836520073de57e0f968a6b819">CalcStrandFromFlag</a> (unsigned long int flag)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Extracts the strand from the FLAG field. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f583a963f7fb0eb071c1b640c660813a"></a><!-- doxytag: member="GenomicRegionSAM::UpdateFlagFromStrand" ref="f583a963f7fb0eb071c1b640c660813a" args="(char strand)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#f583a963f7fb0eb071c1b640c660813a">UpdateFlagFromStrand</a> (char strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Updates the FLAG field given a new strand. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ea563a4a484784ad231399a3b56c7056"></a><!-- doxytag: member="GenomicRegionSAM::PrintGappedSequence" ref="ea563a4a484784ad231399a3b56c7056" args="(char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ea563a4a484784ad231399a3b56c7056">PrintGappedSequence</a> (char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints gapped sequence. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9446540150b384ca0d3dbba630c9c6a6"></a><!-- doxytag: member="GenomicRegionSAM::GetGappedSequence" ref="9446540150b384ca0d3dbba630c9c6a6" args="(char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#9446540150b384ca0d3dbba630c9c6a6">GetGappedSequence</a> (char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns gapped sequence. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="621b3dd6f4d03a5223bd26a57af14f7e"></a><!-- doxytag: member="GenomicRegionSAM::InvalidateSeqData" ref="621b3dd6f4d03a5223bd26a57af14f7e" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#621b3dd6f4d03a5223bd26a57af14f7e">InvalidateSeqData</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Sets CIGAR, SEQ and QUAL fields to "*" and removes OPTIONAL field; also <a class="el" href="classGenomicRegionSAM.html#281e35ef1bd0b7da811d9b8957f3d08f">GenomicRegionSAM::n_tokens</a> is set to 11. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="34b0387b9656e48d0c4046d62fe085e1"></a><!-- doxytag: member="GenomicRegionSAM::TokenizeCIGAR" ref="34b0387b9656e48d0c4046d62fe085e1" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#34b0387b9656e48d0c4046d62fe085e1">TokenizeCIGAR</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Computes and stores CIGAR tokens in <a class="el" href="classGenomicRegionSAM.html#47afdde68b89cf4ea199bc91d4954a92">GenomicRegionSAM::T</a>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="50198fead1c40caadf76f40b61a3c14a"></a><!-- doxytag: member="GenomicRegionSAM::GetNextTokenOfCIGAR" ref="50198fead1c40caadf76f40b61a3c14a" args="(char **cigar, char *type)" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#50198fead1c40caadf76f40b61a3c14a">GetNextTokenOfCIGAR</a> (char **cigar, char *type)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns next token in CIGAR string. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="965bb3cc1217e6c859a07dd23090c2a0"></a><!-- doxytag: member="GenomicRegionSAM::UpdateCigarFromIntervals" ref="965bb3cc1217e6c859a07dd23090c2a0" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#965bb3cc1217e6c859a07dd23090c2a0">UpdateCigarFromIntervals</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Update CIGAR string based only on interval information (chooses simplest possible CIGAR). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5de5d05dfccd33149153a299c1cb2f6b"></a><!-- doxytag: member="GenomicRegionSAM::CalcFragmentLengthFromCIGAR" ref="5de5d05dfccd33149153a299c1cb2f6b" args="()" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#5de5d05dfccd33149153a299c1cb2f6b">CalcFragmentLengthFromCIGAR</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculate interval length from CIGAR string. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="acff4cd74a3b8c450f9f90c6b23412c4"></a><!-- doxytag: member="GenomicRegionSAM::CalcReferenceLengthFromCIGAR" ref="acff4cd74a3b8c450f9f90c6b23412c4" args="()" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#acff4cd74a3b8c450f9f90c6b23412c4">CalcReferenceLengthFromCIGAR</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculate interval length from CIGAR string. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="17e8e7054ffe7205d7c34926b60526d8"></a><!-- doxytag: member="GenomicRegionSAM::TrimCIGAR" ref="17e8e7054ffe7205d7c34926b60526d8" args="(long int d_ref_start, long int d_ref_stop)" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#17e8e7054ffe7205d7c34926b60526d8">TrimCIGAR</a> (long int d_ref_start, long int d_ref_stop)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Trim CIGAR string from left and right by a given distance (measured on the reference sequence). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0ff4472118291a8034efc929c784680c"></a><!-- doxytag: member="GenomicRegionSAM::TrimSequence" ref="0ff4472118291a8034efc929c784680c" args="(char *seq, long int d_ref_start, long int d_ref_stop)" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#0ff4472118291a8034efc929c784680c">TrimSequence</a> (char *seq, long int d_ref_start, long int d_ref_stop)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Trim input <b>seq</b> from left and right (used for trimming both SEQ and QUAL fields). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="97b9fec94e5f21490d839b3f99235a96"></a><!-- doxytag: member="GenomicRegionSAM::RunAlign" ref="97b9fec94e5f21490d839b3f99235a96" args="(Chromosomes *C)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#97b9fec94e5f21490d839b3f99235a96">RunAlign</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints alignments of input sequences to reference genome. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5d82c1d84432274392416f87bfac655a"></a><!-- doxytag: member="GenomicRegionSAM::ApplyBounds" ref="5d82c1d84432274392416f87bfac655a" args="(StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#5d82c1d84432274392416f87bfac655a">ApplyBounds</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Corrects interval start/stop positions so as to comply with chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="52fa8c0420b16fe92f54427365f72899"></a><!-- doxytag: member="GenomicRegionSAM::Center" ref="52fa8c0420b16fe92f54427365f72899" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#52fa8c0420b16fe92f54427365f72899">Center</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace each interval in the region with the corresponding center interval. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d765c84814151be0e710fefb37c7949a"></a><!-- doxytag: member="GenomicRegionSAM::Connect" ref="d765c84814151be0e710fefb37c7949a" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#d765c84814151be0e710fefb37c7949a">Connect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace region with a single interval from minimum start to maximum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ca38b04fa916ae4c9c28c36f537136ba"></a><!-- doxytag: member="GenomicRegionSAM::Diff" ref="ca38b04fa916ae4c9c28c36f537136ba" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ca38b04fa916ae4c9c28c36f537136ba">Diff</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fc9c7ec17be517dd5b1ad7cb62f58826"></a><!-- doxytag: member="GenomicRegionSAM::RunCalcDistances" ref="fc9c7ec17be517dd5b1ad7cb62f58826" args="(char *op1, char *op2)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#fc9c7ec17be517dd5b1ad7cb62f58826">RunCalcDistances</a> (char *op1, char *op2)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print distance between successive intervals. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7c16ba87be562950021d3f19eb14ac5e"></a><!-- doxytag: member="GenomicRegionSAM::Divide" ref="7c16ba87be562950021d3f19eb14ac5e" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#7c16ba87be562950021d3f19eb14ac5e">Divide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="dd2ada73e241f593dcdcdfca97df1eb5"></a><!-- doxytag: member="GenomicRegionSAM::RunDivide" ref="dd2ada73e241f593dcdcdfca97df1eb5" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#dd2ada73e241f593dcdcdfca97df1eb5">RunDivide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d955a6a512678613b4a42d2db0d3f549"></a><!-- doxytag: member="GenomicRegionSAM::Fix" ref="d955a6a512678613b4a42d2db0d3f549" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#d955a6a512678613b4a42d2db0d3f549">Fix</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Removes rogue intervals (e.g. start&gt;stop). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4f26b01325ebd4e20c9bf3247d5fa733"></a><!-- doxytag: member="GenomicRegionSAM::Intersect" ref="4f26b01325ebd4e20c9bf3247d5fa733" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#4f26b01325ebd4e20c9bf3247d5fa733">Intersect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="538b8fc9969f24da6911ec39c8d70a3b"></a><!-- doxytag: member="GenomicRegionSAM::Randomize" ref="538b8fc9969f24da6911ec39c8d70a3b" args="(gsl_rng *random_generator, StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#538b8fc9969f24da6911ec39c8d70a3b">Randomize</a> (gsl_rng *random_generator, StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Randomizes interval position within chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#353207352073db00dee0a9b620dca197">ModifyPos</a> (char *position_op, long int position_shift)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position.  <a href="#353207352073db00dee0a9b620dca197"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4dce925fdb55dcc7ef3dad890bc25408"></a><!-- doxytag: member="GenomicRegionSAM::Select" ref="4dce925fdb55dcc7ef3dad890bc25408" args="(bool first, bool last, bool from5p, bool from3p)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#4dce925fdb55dcc7ef3dad890bc25408">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#c6c559782775bfc65907801148226590">ShiftPos</a> (long int start_shift, long int stop_shift, bool shift_5prime)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately.  <a href="#c6c559782775bfc65907801148226590"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b31fed1760d543fa3a38bb16a064dcf9"></a><!-- doxytag: member="GenomicRegionSAM::RunShuffle" ref="b31fed1760d543fa3a38bb16a064dcf9" args="(gsl_rng *random_generator, GenomicRegionSet *refReg, StringLIntMap *index, StringVecLIntMap *loc)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#b31fed1760d543fa3a38bb16a064dcf9">RunShuffle</a> (gsl_rng *random_generator, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints shuffled region within specified reference regions in <b>refReg</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8a27afdcdf11895d7b3f32bd8eed68ae"></a><!-- doxytag: member="GenomicRegionSAM::Sort" ref="8a27afdcdf11895d7b3f32bd8eed68ae" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#8a27afdcdf11895d7b3f32bd8eed68ae">Sort</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Sort intervals according to start position (only for compatible intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="e195a95ba837bfd9673093f1d451b39c"></a><!-- doxytag: member="GenomicRegionSAM::RunSplit" ref="e195a95ba837bfd9673093f1d451b39c" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#e195a95ba837bfd9673093f1d451b39c">RunSplit</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print region intervals on separate lines. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="beb76df49bb89ecf0f3074805d9fe7b2"></a><!-- doxytag: member="GenomicRegionSAM::ModifyStrand" ref="beb76df49bb89ecf0f3074805d9fe7b2" args="(char *strand_op)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#beb76df49bb89ecf0f3074805d9fe7b2">ModifyStrand</a> (char *strand_op)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Modified strand information: '+'=only forward, '-'=only reverse, 'r'=reverse strands from '+' to '-' and vice versa, 'b'=both strands. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6161ba10d2a4871cc43aa5bc597a787b"></a><!-- doxytag: member="GenomicRegionSAM::Union" ref="6161ba10d2a4871cc43aa5bc597a787b" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#6161ba10d2a4871cc43aa5bc597a787b">Union</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Compute intervals' union. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="27027bc80a3b5a864ebafd415a6ef5a6"></a><!-- doxytag: member="GenomicRegionSAM::PrintWindows" ref="27027bc80a3b5a864ebafd415a6ef5a6" args="(long int win_step, long int win_size)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#27027bc80a3b5a864ebafd415a6ef5a6">PrintWindows</a> (long int win_step, long int win_size)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions). <br></td></tr>
<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="281e35ef1bd0b7da811d9b8957f3d08f"></a><!-- doxytag: member="GenomicRegionSAM::n_tokens" ref="281e35ef1bd0b7da811d9b8957f3d08f" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#281e35ef1bd0b7da811d9b8957f3d08f">n_tokens</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">number of TAB-separated fields in SAM entry (&gt;= 12) <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="200d7f7044dd69311eceb851a3c315e5"></a><!-- doxytag: member="GenomicRegionSAM::FLAG" ref="200d7f7044dd69311eceb851a3c315e5" args="" -->
unsigned long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#200d7f7044dd69311eceb851a3c315e5">FLAG</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">bitwise FLAG <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="edacdd8319f4b614229a41b80d4c6f67"></a><!-- doxytag: member="GenomicRegionSAM::MAPQ" ref="edacdd8319f4b614229a41b80d4c6f67" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#edacdd8319f4b614229a41b80d4c6f67">MAPQ</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">MAPping Quality. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a5acff481ae565db4976e66933212726"></a><!-- doxytag: member="GenomicRegionSAM::CIGAR" ref="a5acff481ae565db4976e66933212726" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a5acff481ae565db4976e66933212726">CIGAR</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">CIGAR string. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f723505bfafff1a1fa616479023d0c3c"></a><!-- doxytag: member="GenomicRegionSAM::RNEXT" ref="f723505bfafff1a1fa616479023d0c3c" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#f723505bfafff1a1fa616479023d0c3c">RNEXT</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reference name of the mate/next fragment. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d6f88b4324ba508b86102641d819661f"></a><!-- doxytag: member="GenomicRegionSAM::PNEXT" ref="d6f88b4324ba508b86102641d819661f" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#d6f88b4324ba508b86102641d819661f">PNEXT</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Position of the mate/next fragment. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f180854a186be37eab49520edaa1262f"></a><!-- doxytag: member="GenomicRegionSAM::TLEN" ref="f180854a186be37eab49520edaa1262f" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#f180854a186be37eab49520edaa1262f">TLEN</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">observed Template LENgth <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6c2887ed1fa3a25d801641ea1b139032"></a><!-- doxytag: member="GenomicRegionSAM::SEQ" ref="6c2887ed1fa3a25d801641ea1b139032" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#6c2887ed1fa3a25d801641ea1b139032">SEQ</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">fragment SEQuence <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b43ca5f2fbfdee520f1424c3a9fd0706"></a><!-- doxytag: member="GenomicRegionSAM::QUAL" ref="b43ca5f2fbfdee520f1424c3a9fd0706" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#b43ca5f2fbfdee520f1424c3a9fd0706">QUAL</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">ASCII of Phred-scaled base QUALity+33. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1622a8804462bad83b1c49d25eeacb49"></a><!-- doxytag: member="GenomicRegionSAM::OPTIONAL" ref="1622a8804462bad83b1c49d25eeacb49" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#1622a8804462bad83b1c49d25eeacb49">OPTIONAL</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">all SAM fields after the 12th mandatory field <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="47afdde68b89cf4ea199bc91d4954a92"></a><!-- doxytag: member="GenomicRegionSAM::T" ref="47afdde68b89cf4ea199bc91d4954a92" args="" -->
<a class="el" href="classGenomicRegionSAM.html#adf8580c5362304ab4866fa76a59df96">CIGARTokens</a>&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#47afdde68b89cf4ea199bc91d4954a92">T</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">CIGAR tokens. <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class processes genomic regions in SAM format. <hr><h2>Member Function Documentation</h2>
<a class="anchor" name="353207352073db00dee0a9b620dca197"></a><!-- doxytag: member="GenomicRegionSAM::ModifyPos" ref="353207352073db00dee0a9b620dca197" args="(char *position_op, long int position_shift)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSAM::ModifyPos           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>position_op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>position_shift</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>position_op</em>&nbsp;</td><td>selects fixed position as follows: '1'=start position, 'c'=center position, '5p'=5-prime, '3p'=3-prime </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>position_shift</em>&nbsp;</td><td>determines the shift of the non-fixed position (in the 'c' case, both positions are shifted symmetrically around the center) </td></tr>
  </table>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#0721b07af0850057e4ab9cd416ecac2f">GenomicRegion</a>.</p>

</div>
</div><p>
<a class="anchor" name="c6c559782775bfc65907801148226590"></a><!-- doxytag: member="GenomicRegionSAM::ShiftPos" ref="c6c559782775bfc65907801148226590" args="(long int start_shift, long int stop_shift, bool shift_5prime)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSAM::ShiftPos           </td>
          <td>(</td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>start_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>stop_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>shift_5prime</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>start_shift</em>&nbsp;</td><td>determines the shift of the start position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>stop_shift</em>&nbsp;</td><td>determines the shift of the stop position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>shift_5prime</em>&nbsp;</td><td>if 'true', then the new start position is the old 5-prime position shifted by <b>start_shift</b> and the new stop position is the old 3-prime position shifted by <b>stop_shift</b> </td></tr>
  </table>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#dc42fea61283fa8134d65467b2292405">GenomicRegion</a>.</p>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Mon Sep 12 16:49:29 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
