<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>SortedGenomicRegionSetOverlaps Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>SortedGenomicRegionSetOverlaps Class Reference</h1><!-- doxytag: class="SortedGenomicRegionSetOverlaps" --><!-- doxytag: inherits="GenomicRegionSetOverlaps" -->This class is used to find overlaps between two genomic regions sets. See detailed description below for an example.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>
<div class="dynheader">
Inheritance diagram for SortedGenomicRegionSetOverlaps:</div>
<div class="dynsection">

<p><center><img src="classSortedGenomicRegionSetOverlaps.png" usemap="#SortedGenomicRegionSetOverlaps_map" border="0" alt=""></center>
<map name="SortedGenomicRegionSetOverlaps_map">
<area href="classGenomicRegionSetOverlaps.html" alt="GenomicRegionSetOverlaps" shape="rect" coords="0,0,207,24">
</map>
</div>

<p>
<a href="classSortedGenomicRegionSetOverlaps-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#0213f6ea908db759f63cea80db4909aa">SortedGenomicRegionSetOverlaps</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#e513304379055f6d379bc5907733dbe2">QuerySet</a>, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#c587bf854c827381493735b473622e03">IndexSet</a>, bool <a class="el" href="classSortedGenomicRegionSetOverlaps.html#f08c449faa6ba7be7c6ed9c07f4226e3">sorted_by_strand</a>)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor.  <a href="#0213f6ea908db759f63cea80db4909aa"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1ac0a70899fce3684da0b7e915a13d1e"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::~SortedGenomicRegionSetOverlaps" ref="1ac0a70899fce3684da0b7e915a13d1e" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#1ac0a70899fce3684da0b7e915a13d1e">~SortedGenomicRegionSetOverlaps</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3b187db0a12f876a9347b4a1641670a0"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::GetQuery" ref="3b187db0a12f876a9347b4a1641670a0" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#3b187db0a12f876a9347b4a1641670a0">GetQuery</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the current query region ('NULL' if no more query regions are available); also, <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> is automatically updated via <a class="el" href="classSortedGenomicRegionSetOverlaps.html#d894d118c61c3cd11be8b37b8f185120">SortedGenomicRegionSetOverlaps::LoadIndexBuffer</a>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="127edf3e9d8fb2ebf189af00e6893e96"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::NextQuery" ref="127edf3e9d8fb2ebf189af00e6893e96" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#127edf3e9d8fb2ebf189af00e6893e96">NextQuery</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the next query region ('NULL' if no more query regions are available); also, <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> is automatically updated via <a class="el" href="classSortedGenomicRegionSetOverlaps.html#d894d118c61c3cd11be8b37b8f185120">SortedGenomicRegionSetOverlaps::LoadIndexBuffer</a>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d0f7b48e1eb685f0aa62aa799088dbc7"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::GetMatch" ref="d0f7b48e1eb685f0aa62aa799088dbc7" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#d0f7b48e1eb685f0aa62aa799088dbc7">GetMatch</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="756bad7ba9a862ae27cd9ad52acbeef2"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::NextMatch" ref="756bad7ba9a862ae27cd9ad52acbeef2" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#756bad7ba9a862ae27cd9ad52acbeef2">NextMatch</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="726105d1ee63db6271a90e8092e243c9"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::Done" ref="726105d1ee63db6271a90e8092e243c9" args="()" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#726105d1ee63db6271a90e8092e243c9">Done</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns true, if no more query or index regions are available and the <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> buffer is empty. <br></td></tr>
<tr><td colspan="2"><br><h2>Private Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d894d118c61c3cd11be8b37b8f185120"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::LoadIndexBuffer" ref="d894d118c61c3cd11be8b37b8f185120" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#d894d118c61c3cd11be8b37b8f185120">LoadIndexBuffer</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">manages <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="c5fe4552253a9d9d802d76b0f4356947"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::ClearIndexBuffer" ref="c5fe4552253a9d9d802d76b0f4356947" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#c5fe4552253a9d9d802d76b0f4356947">ClearIndexBuffer</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">clears <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br></td></tr>
<tr><td colspan="2"><br><h2>Private Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f08c449faa6ba7be7c6ed9c07f4226e3"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::sorted_by_strand" ref="f08c449faa6ba7be7c6ed9c07f4226e3" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#f08c449faa6ba7be7c6ed9c07f4226e3">sorted_by_strand</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">if true, query and index regions sorted by strand <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="19fa18e6abd9f045786698fff48a445f"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::IRegBuffer" ref="19fa18e6abd9f045786698fff48a445f" args="" -->
GenomicRegionList&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">IRegBuffer</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">temporary buffer containing index regions that overlap with current query; the buffer is cleared only if the next query has no overlap <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="71d6eea0c4bf9f19be6cf0c0c562f297"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::IRegBufferIterator" ref="71d6eea0c4bf9f19be6cf0c0c562f297" args="" -->
GenomicRegionList::iterator&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#71d6eea0c4bf9f19be6cf0c0c562f297">IRegBufferIterator</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">iterator on <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="962f8e6ce8c15703cfefa574a07ce42c"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::ireg_buffer_interval" ref="962f8e6ce8c15703cfefa574a07ce42c" args="" -->
<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#962f8e6ce8c15703cfefa574a07ce42c">ireg_buffer_interval</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">keeps tracks of the (linked) interval stored in <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="349d589ff006f742d635b16987fe7b6b"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::max_ireg_buffer_size" ref="349d589ff006f742d635b16987fe7b6b" args="" -->
unsigned long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#349d589ff006f742d635b16987fe7b6b">max_ireg_buffer_size</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">keeps track of the maximum size of <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class is used to find overlaps between two genomic regions sets. See detailed description below for an example. 
<p>
A simple example for computing RNAseq read densities in known exons (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'density' operation): <div class="fragment"><pre class="fragment"><span class="preprocessor">    #include "core.h"</span>
<span class="preprocessor">    #include "genomic_intervals.h"</span>
 
    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *REF_REG_FILE = argv[next_arg];
    <span class="keywordtype">char</span> *TEST_REG_FILE = next_arg+1==argc ? NULL : argv[next_arg+1];
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> RefRegSet(REF_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">true</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> TestRegSet(TEST_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">false</span>);

    <span class="comment">// process overlaps</span>
    <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a> *overlaps = <span class="keyword">new</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#0213f6ea908db759f63cea80db4909aa" title="Class constructor.">SortedGenomicRegionSetOverlaps</a>(&amp;TestRegSet,&amp;RefRegSet,SORTED_BY_STRAND);
    <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *coverage = overlaps-&gt;<a class="code" href="classGenomicRegionSetOverlaps.html#5c31081b154c624ba35487d988082ffa" title="Calculates the total overlap for each index region. The index region set must be...">CalcIndexCoverage</a>(MATCH_GAPS,IGNORE_STRAND,USE_VALUES); 
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">"Printing densities..."</span>,RefRegSet.n_regions);
    <span class="keywordflow">for</span> (<span class="keywordtype">long</span> <span class="keywordtype">int</span> k=0; k&lt;RefRegSet.n_regions; k++) {
      <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *qreg = RefRegSet.R[k];
      <span class="keywordtype">long</span> <span class="keywordtype">int</span> qreg_size = MATCH_GAPS ? (qreg-&gt;<a class="code" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.back()-&gt;STOP-qreg-&gt;<a class="code" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.front()-&gt;START+1) : qreg-&gt;<a class="code" href="classGenomicRegion.html#5bff7a3e83b0bc9bfb5441270d0ba1ae" title="Returns sum of interval&amp;#39;s sizes.">GetSize</a>();
      <span class="keywordtype">double</span> density = (double)coverage[k]/qreg_size;
      <span class="keywordflow">if</span> (density&gt;=MIN_DENSITY) printf(<span class="stringliteral">"%s\t%.4e\n"</span>, qreg-&gt;<a class="code" href="classGenomicRegion.html#7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>, density);
      PRG.Check();
    }
    PRG.Done();
    <span class="keyword">delete</span> coverage;
    <span class="keyword">delete</span> overlaps;
</pre></div> A more complex example for creating ChIPseq read profiles in TSS regions (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'offset' operation): <div class="fragment"><pre class="fragment"><span class="preprocessor">    #include "core.h"</span>
<span class="preprocessor">    #include "genomic_intervals.h"</span>
 
    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *QUERY_REG_FILE = <span class="stringliteral">"chipseq.reads.reg"</span>;
    <span class="keywordtype">char</span> *INDEX_REG_FILE = <span class="stringliteral">"TSS.flank10kb.reg"</span>;
    <span class="keywordtype">char</span> *OFFSET_OP = <span class="stringliteral">"5p"</span>;
    <span class="keywordtype">bool</span> SORTED_BY_STRAND = <span class="keyword">false</span>; 
    <span class="keywordtype">bool</span> MATCH_GAPS = <span class="keyword">true</span>;
    <span class="keywordtype">bool</span> IGNORE_STRAND = <span class="keyword">true</span>;
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> QueryRegSet(QUERY_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> IndexRegSet(INDEX_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);

    <span class="comment">// process overlaps</span>
    <a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed...">SortedGenomicRegionSetOverlaps</a> Overlaps(&amp;QueryRegSet,&amp;IndexRegSet,SORTED_BY_STRAND);
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">"Processing queries..."</span>,1);
    <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *qreg=Overlaps.GetQuery(); Overlaps.Done()==<span class="keyword">false</span>; qreg=Overlaps.NextQuery()) {
      <span class="keywordtype">size_t</span> Qsize = qreg-&gt;<a class="code" href="classGenomicRegion.html#5bff7a3e83b0bc9bfb5441270d0ba1ae" title="Returns sum of interval&amp;#39;s sizes.">GetSize</a>();
      <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *ireg=Overlaps.GetOverlap(MATCH_GAPS,IGNORE_STRAND); ireg!=NULL; ireg=Overlaps.NextOverlap(MATCH_GAPS,IGNORE_STRAND)) {
        cout &lt;&lt; qreg-&gt;<a class="code" href="classGenomicRegion.html#7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a> &lt;&lt; <span class="charliteral">'\t'</span>;
        <span class="keywordtype">long</span> <span class="keywordtype">int</span> start_offset, stop_offset;
        ireg-&gt;I.front()-&gt;GetOffsetFrom(qreg-&gt;<a class="code" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.front(),OFFSET_OP,IGNORE_STRAND,&amp;start_offset,&amp;stop_offset);
        printf(<span class="stringliteral">"%f"</span>, ((<span class="keywordtype">float</span>)start_offset/Qsize+(<span class="keywordtype">float</span>)stop_offset/Qsize)/2); 
        cout &lt;&lt; <span class="charliteral">'\n'</span>;
      }
      PRG.Check();
    }
    PRG.Done();
</pre></div> <hr><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" name="0213f6ea908db759f63cea80db4909aa"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::SortedGenomicRegionSetOverlaps" ref="0213f6ea908db759f63cea80db4909aa" args="(GenomicRegionSet *QuerySet, GenomicRegionSet *IndexSet, bool sorted_by_strand)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">SortedGenomicRegionSetOverlaps::SortedGenomicRegionSetOverlaps           </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td>
          <td class="paramname"> <em>QuerySet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td>
          <td class="paramname"> <em>IndexSet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>sorted_by_strand</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Class constructor. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>QuerySet</em>&nbsp;</td><td>pointer to query region set </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>IndexSet</em>&nbsp;</td><td>pointer to index region set </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sorted_by_strand</em>&nbsp;</td><td>if true, query and index regions are sorted by strand </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Fri Sep 23 12:35:25 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
