<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>GenomicRegionBED Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>GenomicRegionBED Class Reference</h1><!-- doxytag: class="GenomicRegionBED" --><!-- doxytag: inherits="GenomicRegion" -->This class processes genomic regions in BED format.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>
<div class="dynheader">
Inheritance diagram for GenomicRegionBED:</div>
<div class="dynsection">

<p><center><img src="classGenomicRegionBED.png" usemap="#GenomicRegionBED_map" border="0" alt=""></center>
<map name="GenomicRegionBED_map">
<area href="classGenomicRegion.html" alt="GenomicRegion" shape="rect" coords="0,0,125,24">
</map>
</div>

<p>
<a href="classGenomicRegionBED-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="89d54edc5809d113202c9f09f82e1057"></a><!-- doxytag: member="GenomicRegionBED::GenomicRegionBED" ref="89d54edc5809d113202c9f09f82e1057" args="(FileBuffer *B)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#89d54edc5809d113202c9f09f82e1057">GenomicRegionBED</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor, see example below. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4d7c87327c1074258ac1e6b5ac87c9dd"></a><!-- doxytag: member="GenomicRegionBED::GenomicRegionBED" ref="4d7c87327c1074258ac1e6b5ac87c9dd" args="(char *inp, long int n_line=0)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#4d7c87327c1074258ac1e6b5ac87c9dd">GenomicRegionBED</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor, see example below. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a6e68da10b635bcd84bf320b94a4d414"></a><!-- doxytag: member="GenomicRegionBED::GenomicRegionBED" ref="a6e68da10b635bcd84bf320b94a4d414" args="(char *label, GenomicInterval *i)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a6e68da10b635bcd84bf320b94a4d414">GenomicRegionBED</a> (char *label, <a class="el" href="classGenomicInterval.html">GenomicInterval</a> *i)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a27241b7fbdb90e0b4b1e411045d080f"></a><!-- doxytag: member="GenomicRegionBED::~GenomicRegionBED" ref="a27241b7fbdb90e0b4b1e411045d080f" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a27241b7fbdb90e0b4b1e411045d080f">~GenomicRegionBED</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="707a1744e040214d78d37b928f1f0252"></a><!-- doxytag: member="GenomicRegionBED::Read" ref="707a1744e040214d78d37b928f1f0252" args="(char *inp, long int n_line=0)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#707a1744e040214d78d37b928f1f0252">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reads BED format from input string <b>inp</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ffbd8eb279f10f9b45189e2388fe7e89"></a><!-- doxytag: member="GenomicRegionBED::Print" ref="ffbd8eb279f10f9b45189e2388fe7e89" args="(FILE *file_ptr=stdout)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#ffbd8eb279f10f9b45189e2388fe7e89">Print</a> (FILE *file_ptr=stdout)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="e194ce2d67eafe222f10172ee376de94"></a><!-- doxytag: member="GenomicRegionBED::Print" ref="e194ce2d67eafe222f10172ee376de94" args="(FILE *file_ptr, char *color, bool convert_chromosome)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#e194ce2d67eafe222f10172ee376de94">Print</a> (FILE *file_ptr, char *color, bool convert_chromosome)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as generic <a class="el" href="classGenomicRegionBED.html#ffbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a> except that it modifies the color and the chromosome names. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="04eceeb473bee498747bdd68d8ba8806"></a><!-- doxytag: member="GenomicRegionBED::PrintIntersection" ref="04eceeb473bee498747bdd68d8ba8806" args="(GenomicRegion *r, bool ignore_strand, bool merge_labels)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#04eceeb473bee498747bdd68d8ba8806">PrintIntersection</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand, bool merge_labels)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints all pairwise intersection with the intervals in region <b>r</b> (compatible, sorted and non-overlapping). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="25a8584259b6a8d47aa8c46795de1880"></a><!-- doxytag: member="GenomicRegionBED::PrintConstrained" ref="25a8584259b6a8d47aa8c46795de1880" args="(GenomicRegion *r, bool merge_labels=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#25a8584259b6a8d47aa8c46795de1880">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionBED.html#ffbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>start</b> and <b>stop</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="aebba3c31c0fa88e594706f490030549"></a><!-- doxytag: member="GenomicRegionBED::PrintModified" ref="aebba3c31c0fa88e594706f490030549" args="(char *label, long int start, long int stop)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#aebba3c31c0fa88e594706f490030549">PrintModified</a> (char *label, long int start, long int stop)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionBED.html#ffbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="01bd886cfa67a3adae10c67dee941924"></a><!-- doxytag: member="GenomicRegionBED::UpdateThick" ref="01bd886cfa67a3adae10c67dee941924" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#01bd886cfa67a3adae10c67dee941924">UpdateThick</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Auxiliary function: updates thickStart/thickEnd variables to honor changes in coordinates. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="513fb071e2626be58837052daf86bab5"></a><!-- doxytag: member="GenomicRegionBED::PrintBEDFormat" ref="513fb071e2626be58837052daf86bab5" args="(char *color, bool convert_chromosome)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#513fb071e2626be58837052daf86bab5">PrintBEDFormat</a> (char *color, bool convert_chromosome)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints in BED format. If <b>convert_chromosome</b> is 'true', then convert from ENSEMBL to UCSC names (by adding 'chr' prefix to each chromosome name, and by converting 'MT' to 'chrM'). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5056b98f57933e2d24fd6d601f5499ef"></a><!-- doxytag: member="GenomicRegionBED::ApplyBounds" ref="5056b98f57933e2d24fd6d601f5499ef" args="(StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#5056b98f57933e2d24fd6d601f5499ef">ApplyBounds</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Corrects interval start/stop positions so as to comply with chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="35c692c1343327d9cb450cfa1276096f"></a><!-- doxytag: member="GenomicRegionBED::Center" ref="35c692c1343327d9cb450cfa1276096f" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#35c692c1343327d9cb450cfa1276096f">Center</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace each interval in the region with the corresponding center interval. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d55a9ffea8b0924bac3d11a5722cdfd2"></a><!-- doxytag: member="GenomicRegionBED::Connect" ref="d55a9ffea8b0924bac3d11a5722cdfd2" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#d55a9ffea8b0924bac3d11a5722cdfd2">Connect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace region with a single interval from minimum start to maximum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9580179a120834e71ac90e56e5f35aa9"></a><!-- doxytag: member="GenomicRegionBED::Diff" ref="9580179a120834e71ac90e56e5f35aa9" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#9580179a120834e71ac90e56e5f35aa9">Diff</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="833d50369e61f64603c63b6a05f3d80b"></a><!-- doxytag: member="GenomicRegionBED::RunCalcDistances" ref="833d50369e61f64603c63b6a05f3d80b" args="(char *op1, char *op2)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#833d50369e61f64603c63b6a05f3d80b">RunCalcDistances</a> (char *op1, char *op2)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print distance between successive intervals. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="be056ec84d2d6890479d47559839bab4"></a><!-- doxytag: member="GenomicRegionBED::Divide" ref="be056ec84d2d6890479d47559839bab4" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#be056ec84d2d6890479d47559839bab4">Divide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5f456b4af154c98968370568828d6e3d"></a><!-- doxytag: member="GenomicRegionBED::RunDivide" ref="5f456b4af154c98968370568828d6e3d" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#5f456b4af154c98968370568828d6e3d">RunDivide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ac069dcf513265f7756b72c106f2765c"></a><!-- doxytag: member="GenomicRegionBED::Fix" ref="ac069dcf513265f7756b72c106f2765c" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#ac069dcf513265f7756b72c106f2765c">Fix</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Removes rogue intervals (e.g. start&gt;stop). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="cf815a65712e82bf3f8c596c795f55c4"></a><!-- doxytag: member="GenomicRegionBED::Intersect" ref="cf815a65712e82bf3f8c596c795f55c4" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#cf815a65712e82bf3f8c596c795f55c4">Intersect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#c515c70f443db400f911452ed359433b">ModifyPos</a> (char *position_op, long int position_shift)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position.  <a href="#c515c70f443db400f911452ed359433b"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="920d47deb095506c311da2b2324e7a14"></a><!-- doxytag: member="GenomicRegionBED::Randomize" ref="920d47deb095506c311da2b2324e7a14" args="(gsl_rng *random_generator, StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#920d47deb095506c311da2b2324e7a14">Randomize</a> (gsl_rng *random_generator, StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Randomizes interval position within chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="dec7408315765bbdb21465a26400422b"></a><!-- doxytag: member="GenomicRegionBED::Select" ref="dec7408315765bbdb21465a26400422b" args="(bool first, bool last, bool from5p, bool from3p)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#dec7408315765bbdb21465a26400422b">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#97d00cb29eb1f379b9e6ee17197530c7">ShiftPos</a> (long int start_shift, long int stop_shift, bool shift_5prime)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately.  <a href="#97d00cb29eb1f379b9e6ee17197530c7"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f3793387aa335bf68042f9245bc5992d"></a><!-- doxytag: member="GenomicRegionBED::RunShuffle" ref="f3793387aa335bf68042f9245bc5992d" args="(gsl_rng *random_generator, GenomicRegionSet *refReg, StringLIntMap *index, StringVecLIntMap *loc)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#f3793387aa335bf68042f9245bc5992d">RunShuffle</a> (gsl_rng *random_generator, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints shuffled region within specified reference regions in <b>refReg</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fd1a61c04f76782abdc83074ae0f7bd3"></a><!-- doxytag: member="GenomicRegionBED::Sort" ref="fd1a61c04f76782abdc83074ae0f7bd3" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#fd1a61c04f76782abdc83074ae0f7bd3">Sort</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Sort intervals according to start position (only for compatible intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bd3b8e4969c2ff2639eb344de7b224f9"></a><!-- doxytag: member="GenomicRegionBED::RunSplit" ref="bd3b8e4969c2ff2639eb344de7b224f9" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#bd3b8e4969c2ff2639eb344de7b224f9">RunSplit</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print region intervals on separate lines. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="605f5bf86254547d117ed0c3f5c64587"></a><!-- doxytag: member="GenomicRegionBED::Union" ref="605f5bf86254547d117ed0c3f5c64587" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#605f5bf86254547d117ed0c3f5c64587">Union</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Compute intervals' union. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3ebdc24dd97c7b106205d932970de535"></a><!-- doxytag: member="GenomicRegionBED::PrintWindows" ref="3ebdc24dd97c7b106205d932970de535" args="(long int win_step, long int win_size)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#3ebdc24dd97c7b106205d932970de535">PrintWindows</a> (long int win_step, long int win_size)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0c4b2a8a590ce1dfbef142d03d53fe5d"></a><!-- doxytag: member="GenomicRegionBED::Constrain" ref="0c4b2a8a590ce1dfbef142d03d53fe5d" args="(GenomicRegion *r, char *label=NULL)" -->
<a class="el" href="classGenomicRegionBED.html">GenomicRegionBED</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#0c4b2a8a590ce1dfbef142d03d53fe5d">Constrain</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, char *label=NULL)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a version of this genomic region constrained by the start/stop coordinates. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6db4aa9e6d1655bf116095d295b27269"></a><!-- doxytag: member="GenomicRegionBED::Diff" ref="6db4aa9e6d1655bf116095d295b27269" args="(GenomicRegion *r)" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#6db4aa9e6d1655bf116095d295b27269">Diff</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns the difference of this genomic region and region <b>r</b>. <br></td></tr>
<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="60878391f6fb1e0ad0ac9bc88a7625d3"></a><!-- doxytag: member="GenomicRegionBED::n_tokens" ref="60878391f6fb1e0ad0ac9bc88a7625d3" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#60878391f6fb1e0ad0ac9bc88a7625d3">n_tokens</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">number of tokens <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fa7ba810a0d1673f5dbabcf85c7bbb95"></a><!-- doxytag: member="GenomicRegionBED::score" ref="fa7ba810a0d1673f5dbabcf85c7bbb95" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#fa7ba810a0d1673f5dbabcf85c7bbb95">score</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">score field <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="52fc75b347dfe6696e7ad2beeeb623a9"></a><!-- doxytag: member="GenomicRegionBED::itemRgb" ref="52fc75b347dfe6696e7ad2beeeb623a9" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#52fc75b347dfe6696e7ad2beeeb623a9">itemRgb</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">color field <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="019dfaebedc17e8859b36c3359287faa"></a><!-- doxytag: member="GenomicRegionBED::thickStart" ref="019dfaebedc17e8859b36c3359287faa" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#019dfaebedc17e8859b36c3359287faa">thickStart</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">thickStart field <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5333a7fdc1f21db508aa7af4f50553ba"></a><!-- doxytag: member="GenomicRegionBED::thickEnd" ref="5333a7fdc1f21db508aa7af4f50553ba" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#5333a7fdc1f21db508aa7af4f50553ba">thickEnd</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">thickEnd field <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class processes genomic regions in BED format. 
<p>
Example: <div class="fragment"><pre class="fragment">    <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> x(<span class="stringliteral">"chr1 132033 140102 Region#1 1000 + 132033 140102 0 2 3899,4104 0,3965"</span>);
    cout &lt;&lt; <span class="stringliteral">"x = \n"</span>; x.Print();
</pre></div> <hr><h2>Member Function Documentation</h2>
<a class="anchor" name="c515c70f443db400f911452ed359433b"></a><!-- doxytag: member="GenomicRegionBED::ModifyPos" ref="c515c70f443db400f911452ed359433b" args="(char *position_op, long int position_shift)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionBED::ModifyPos           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>position_op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>position_shift</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>position_op</em>&nbsp;</td><td>selects fixed position as follows: '1'=start position, 'c'=center position, '5p'=5-prime, '3p'=3-prime </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>position_shift</em>&nbsp;</td><td>determines the shift of the non-fixed position (in the 'c' case, both positions are shifted symmetrically around the center) </td></tr>
  </table>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#0721b07af0850057e4ab9cd416ecac2f">GenomicRegion</a>.</p>

</div>
</div><p>
<a class="anchor" name="97d00cb29eb1f379b9e6ee17197530c7"></a><!-- doxytag: member="GenomicRegionBED::ShiftPos" ref="97d00cb29eb1f379b9e6ee17197530c7" args="(long int start_shift, long int stop_shift, bool shift_5prime)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionBED::ShiftPos           </td>
          <td>(</td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>start_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>stop_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>shift_5prime</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>start_shift</em>&nbsp;</td><td>determines the shift of the start position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>stop_shift</em>&nbsp;</td><td>determines the shift of the stop position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>shift_5prime</em>&nbsp;</td><td>if 'true', then the new start position is the old 5-prime position shifted by <b>start_shift</b> and the new stop position is the old 3-prime position shifted by <b>stop_shift</b> </td></tr>
  </table>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#dc42fea61283fa8134d65467b2292405">GenomicRegion</a>.</p>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Tue Sep 20 18:46:13 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
