<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>GenomicRegion Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a>  </div>
  <div class="headertitle">
<div class="title">GenomicRegion Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="GenomicRegion" -->
<p>This class implements the genomic regions (i.e. ordered sets of genomic intervals).  
 <a href="classGenomicRegion.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for GenomicRegion:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classGenomicRegion.png" usemap="#GenomicRegion_map" alt=""/>
  <map id="GenomicRegion_map" name="GenomicRegion_map">
<area href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format." alt="GenomicRegionBED" shape="rect" coords="0,56,167,80"/>
<area href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf..." alt="GenomicRegionBEDToREG" shape="rect" coords="177,56,344,80"/>
<area href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format." alt="GenomicRegionGFF" shape="rect" coords="354,56,521,80"/>
<area href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf..." alt="GenomicRegionGFFToREG" shape="rect" coords="531,56,698,80"/>
<area href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format." alt="GenomicRegionSAM" shape="rect" coords="708,56,875,80"/>
<area href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf..." alt="GenomicRegionSAMToREG" shape="rect" coords="885,56,1052,80"/>
<area href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence." alt="GenomicRegionSEQ" shape="rect" coords="1062,56,1229,80"/>
</map>
 </div></div>

<p><a href="classGenomicRegion-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8ff9fe5618b932a1caccbd9072ddaaef"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="a8ff9fe5618b932a1caccbd9072ddaaef" args="()" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a8ff9fe5618b932a1caccbd9072ddaaef">GenomicRegion</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Empty class constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac45e11c7c0d0477c4b8682ef5266d3d7"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="ac45e11c7c0d0477c4b8682ef5266d3d7" args="(FileBuffer *B)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ac45e11c7c0d0477c4b8682ef5266d3d7">GenomicRegion</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor, see example below. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac22559720502ea450aeb4f65c6c789e1"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="ac22559720502ea450aeb4f65c6c789e1" args="(char *inp, long int n_line=0)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ac22559720502ea450aeb4f65c6c789e1">GenomicRegion</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor, see example below. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6de3d0ff71615c36a7311509af56ae79"></a><!-- doxytag: member="GenomicRegion::GenomicRegion" ref="a6de3d0ff71615c36a7311509af56ae79" args="(char *label, GenomicInterval *i)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a6de3d0ff71615c36a7311509af56ae79">GenomicRegion</a> (char *label, <a class="el" href="classGenomicInterval.html">GenomicInterval</a> *i)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a37af413ecbed6ee0d36ecbc40d957947"></a><!-- doxytag: member="GenomicRegion::~GenomicRegion" ref="a37af413ecbed6ee0d36ecbc40d957947" args="()" -->
virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a37af413ecbed6ee0d36ecbc40d957947">~GenomicRegion</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class destructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afd188daab9b4e97ef66883d6971fe103"></a><!-- doxytag: member="GenomicRegion::Read" ref="afd188daab9b4e97ef66883d6971fe103" args="(char *inp, long int n_line=0, char del1=' ', char del2=',')" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#afd188daab9b4e97ef66883d6971fe103">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0, char del1=' ', char del2=',')</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Reads label and genomic region intervals from input string <b>inp</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a40397d1fab7b00e5f04b9f7e52f0114e"></a><!-- doxytag: member="GenomicRegion::Print" ref="a40397d1fab7b00e5f04b9f7e52f0114e" args="(FILE *file_ptr=stdout)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a40397d1fab7b00e5f04b9f7e52f0114e">Print</a> (FILE *file_ptr=stdout)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9846e1c0959ba76c1d552974498a396c"></a><!-- doxytag: member="GenomicRegion::PrintModified" ref="a9846e1c0959ba76c1d552974498a396c" args="(const char *label, long int start, long int stop)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a9846e1c0959ba76c1d552974498a396c">PrintModified</a> (const char *label, long int start, long int stop)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegion.html#a40397d1fab7b00e5f04b9f7e52f0114e" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aba987d40590874eebd1dd8bb0c02bdc8"></a><!-- doxytag: member="GenomicRegion::PrintIntersection" ref="aba987d40590874eebd1dd8bb0c02bdc8" args="(GenomicRegion *r, bool ignore_strand, bool merge_labels)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#aba987d40590874eebd1dd8bb0c02bdc8">PrintIntersection</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand, bool merge_labels)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all pairwise intersection with the intervals in region <b>r</b> (compatible, sorted and non-overlapping). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="acfe5347d9c92da801ce7e97416fc8758"></a><!-- doxytag: member="GenomicRegion::PrintConstrained" ref="acfe5347d9c92da801ce7e97416fc8758" args="(GenomicRegion *r, bool merge_labels=false)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#acfe5347d9c92da801ce7e97416fc8758">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegion.html#a40397d1fab7b00e5f04b9f7e52f0114e" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>r-&gt;I.front()-&gt;START</b> and <b>r-&gt;I.back()-&gt;STOP</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9c687565a8ed25da5a5ccf08e5e8461e"></a><!-- doxytag: member="GenomicRegion::PrintREG" ref="a9c687565a8ed25da5a5ccf08e5e8461e" args="(bool compact=false)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a9c687565a8ed25da5a5ccf08e5e8461e">PrintREG</a> (bool compact=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9a00b031d4a119181b8b76ffa34a6f47"></a><!-- doxytag: member="GenomicRegion::PrintQ" ref="a9a00b031d4a119181b8b76ffa34a6f47" args="(bool reverse=false)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a9a00b031d4a119181b8b76ffa34a6f47">PrintQ</a> (bool reverse=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints sequences (SEQ format) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a68ecc1c8c76e9204f6e27801f0ba3d5f"></a><!-- doxytag: member="GenomicRegion::PrintError" ref="a68ecc1c8c76e9204f6e27801f0ba3d5f" args="(string error_msg)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a68ecc1c8c76e9204f6e27801f0ba3d5f">PrintError</a> (string error_msg)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints error message and exits. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aad8566a736786134b72e7bf74296e807"></a><!-- doxytag: member="GenomicRegion::DeleteIntervals" ref="aad8566a736786134b72e7bf74296e807" args="(GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#aad8566a736786134b72e7bf74296e807">DeleteIntervals</a> (GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes intervals from iterator <b>i</b> to <b>j</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abc547e770917e859fa434d6c54954916"></a><!-- doxytag: member="GenomicRegion::SetLabel" ref="abc547e770917e859fa434d6c54954916" args="(char *label)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#abc547e770917e859fa434d6c54954916">SetLabel</a> (char *label)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Changes the label. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae795532404bad8b93bec632c20d12b9d"></a><!-- doxytag: member="GenomicRegion::SetStrand" ref="ae795532404bad8b93bec632c20d12b9d" args="(char strand)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ae795532404bad8b93bec632c20d12b9d">SetStrand</a> (char strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Interval strands are set to <b>strand</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a815baa00d7d33f179181560a1fe468fc"></a><!-- doxytag: member="GenomicRegion::GetChromosome" ref="a815baa00d7d33f179181560a1fe468fc" args="()" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a815baa00d7d33f179181560a1fe468fc">GetChromosome</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns pointer to chromosome name. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a953613d099a41f0a9e23dcff0bafa85b"></a><!-- doxytag: member="GenomicRegion::GetSize" ref="a953613d099a41f0a9e23dcff0bafa85b" args="(bool skip_gaps)" -->
size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a953613d099a41f0a9e23dcff0bafa85b">GetSize</a> (bool skip_gaps)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns region size. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a719143f1affa44c38493b2cd43448d3b"></a><!-- doxytag: member="GenomicRegion::GetSeqLength" ref="a719143f1affa44c38493b2cd43448d3b" args="(Chromosomes *C)" -->
size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a719143f1affa44c38493b2cd43448d3b">GetSeqLength</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns genomic region sequence length excluding 'N' characters. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afcf7c053989b87f425f084d1f2c078f5"></a><!-- doxytag: member="GenomicRegion::GetSeq" ref="afcf7c053989b87f425f084d1f2c078f5" args="(Chromosomes *C, bool replace=false)" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#afcf7c053989b87f425f084d1f2c078f5">GetSeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool replace=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns genomic region sequence. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a58c2fe162c1e9bbbd7c90d5abd011ff0"></a><!-- doxytag: member="GenomicRegion::IsCompatible" ref="a58c2fe162c1e9bbbd7c90d5abd011ff0" args="(bool ignore_strand)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a58c2fe162c1e9bbbd7c90d5abd011ff0">IsCompatible</a> (bool ignore_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns 'true' if all intervals are in the same chromosome and strand. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2ebd984a2632314a39a6e9ba2c497105"></a><!-- doxytag: member="GenomicRegion::IsCompatibleWith" ref="a2ebd984a2632314a39a6e9ba2c497105" args="(GenomicRegion *r, bool ignore_strand)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a2ebd984a2632314a39a6e9ba2c497105">IsCompatibleWith</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns 'true' if all intervals are compatible with the intervals in <b>r</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2e0765dfbe651ad583faa8c8a77e047f"></a><!-- doxytag: member="GenomicRegion::IsCompatibleSorted" ref="a2e0765dfbe651ad583faa8c8a77e047f" args="(bool ignore_strand)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a2e0765dfbe651ad583faa8c8a77e047f">IsCompatibleSorted</a> (bool ignore_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns 'true' if intervals are compatible and sorted by chromosome, strand (if <b>ignore_strand</b> is false) and start position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afb2db74a3cd85a84533375483e12b96c"></a><!-- doxytag: member="GenomicRegion::IsCompatibleSortedAndNonoverlapping" ref="afb2db74a3cd85a84533375483e12b96c" args="()" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#afb2db74a3cd85a84533375483e12b96c">IsCompatibleSortedAndNonoverlapping</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns 'true' if intervals are compatible, sorted and non-overlapping. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9e3617a0046fe5fa8fce9f9cce8c3c37"></a><!-- doxytag: member="GenomicRegion::OverlapsWith" ref="a9e3617a0046fe5fa8fce9f9cce8c3c37" args="(GenomicRegion *r, bool ignore_strand)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a9e3617a0046fe5fa8fce9f9cce8c3c37">OverlapsWith</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns 'true' if it overlaps with input region <b>r</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a25634257eb7a2ed379548b17a83414fb"></a><!-- doxytag: member="GenomicRegion::IsBefore" ref="a25634257eb7a2ed379548b17a83414fb" args="(GenomicRegion *r, bool sorted_by_strand)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a25634257eb7a2ed379548b17a83414fb">IsBefore</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool sorted_by_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Compares with input region <b>r</b> and returns true if its order is before <b>r</b>. The order is determined first by chromosome, second by strand, then by start position and finally by stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2065cf74a82c2340b1e49e7f7ec0fdd4"></a><!-- doxytag: member="GenomicRegion::IsPosBefore" ref="a2065cf74a82c2340b1e49e7f7ec0fdd4" args="(GenomicRegion *r)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a2065cf74a82c2340b1e49e7f7ec0fdd4">IsPosBefore</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Compares with input region <b>r</b> and returns true if its position order is before <b>r</b>. The position order is determined first by start position and then by stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a623e28aab180174d43873c4a85342ad4"></a><!-- doxytag: member="GenomicRegion::CalcOverlap" ref="a623e28aab180174d43873c4a85342ad4" args="(GenomicRegion *r, bool ignore_strand)" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a623e28aab180174d43873c4a85342ad4">CalcOverlap</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the total number of overlapping nucleotides (gaps are not matched) with input region <b>r</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ada548684fc4ae44dd3782aa15eaaf0a8"></a><!-- doxytag: member="GenomicRegion::CalcDirection" ref="ada548684fc4ae44dd3782aa15eaaf0a8" args="(GenomicInterval *i, bool sorted_by_strand)" -->
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ada548684fc4ae44dd3782aa15eaaf0a8">CalcDirection</a> (<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *i, bool sorted_by_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns -1 if before interval <b>i</b>, +1 if after, otherwise returns 0. The order is determined first by chromosome, then (optionally) by strand and finally by start position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab44366e44dc9a83da65b9593ce90dd32"></a><!-- doxytag: member="GenomicRegion::CalcDirection" ref="ab44366e44dc9a83da65b9593ce90dd32" args="(GenomicRegion *r, bool sorted_by_strand)" -->
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ab44366e44dc9a83da65b9593ce90dd32">CalcDirection</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool sorted_by_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns -1 if before region <b>r</b>, +1 if after, otherwise returns 0. The order is determined first by chromosome, then (optionally) by strand and finally by start position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a90b0382a60f0a8532af73fc18f41364b"></a><!-- doxytag: member="GenomicRegion::RunAlign" ref="a90b0382a60f0a8532af73fc18f41364b" args="(Chromosomes *C)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a90b0382a60f0a8532af73fc18f41364b">RunAlign</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints alignments of input sequences to reference genome. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a405010d9256a3c08c53adee98133cb62"></a><!-- doxytag: member="GenomicRegion::PrintBEDFormat" ref="a405010d9256a3c08c53adee98133cb62" args="(char *color, bool convert_chromosome)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a405010d9256a3c08c53adee98133cb62">PrintBEDFormat</a> (char *color, bool convert_chromosome)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints in BED format. If <b>convert_chromosome</b> is 'true', then convert from ENSEMBL to UCSC names (by adding 'chr' prefix to each chromosome name, and by converting 'MT' to 'chrM'). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a11b494e59775ac255ddb469c98f8a8a8"></a><!-- doxytag: member="GenomicRegion::ApplyBounds" ref="a11b494e59775ac255ddb469c98f8a8a8" args="(StringLIntMap *bounds)" -->
virtual bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a11b494e59775ac255ddb469c98f8a8a8">ApplyBounds</a> (StringLIntMap *bounds)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Corrects interval start/stop positions so as to comply with chromosomal bounds. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1e9a85a086afbe2608411450c9a4c5e6"></a><!-- doxytag: member="GenomicRegion::Center" ref="a1e9a85a086afbe2608411450c9a4c5e6" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a1e9a85a086afbe2608411450c9a4c5e6">Center</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace each interval in the region with the corresponding center interval. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a463b4f14b9a05132441ad91541cd7335"></a><!-- doxytag: member="GenomicRegion::ModifyChromosomeNames" ref="a463b4f14b9a05132441ad91541cd7335" args="(map&lt; string, string &gt; &amp;chromosome_conversion_map)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a463b4f14b9a05132441ad91541cd7335">ModifyChromosomeNames</a> (map&lt; string, string &gt; &amp;chromosome_conversion_map)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Modify chromosome names. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac2c426782bd7c0d302b7a77bb1ff047f"></a><!-- doxytag: member="GenomicRegion::Connect" ref="ac2c426782bd7c0d302b7a77bb1ff047f" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ac2c426782bd7c0d302b7a77bb1ff047f">Connect</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace region with a single interval from minimum start to maximum stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a869de97ecf059355fe5a5b572b5c2575"></a><!-- doxytag: member="GenomicRegion::Diff" ref="a869de97ecf059355fe5a5b572b5c2575" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a869de97ecf059355fe5a5b572b5c2575">Diff</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a98695d5d1056d693a1c635b197f4b8bf"></a><!-- doxytag: member="GenomicRegion::RunCalcDistances" ref="a98695d5d1056d693a1c635b197f4b8bf" args="(char *op1, char *op2)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a98695d5d1056d693a1c635b197f4b8bf">RunCalcDistances</a> (char *op1, char *op2)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print distance between successive intervals. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0785438db56d7b7cc7539470e85434dd"></a><!-- doxytag: member="GenomicRegion::Divide" ref="a0785438db56d7b7cc7539470e85434dd" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a0785438db56d7b7cc7539470e85434dd">Divide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a45fe5d619f2e21644eacb6984aa47eff"></a><!-- doxytag: member="GenomicRegion::RunDivide" ref="a45fe5d619f2e21644eacb6984aa47eff" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a45fe5d619f2e21644eacb6984aa47eff">RunDivide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a048d2182b212789b182c2c82a7ac131e"></a><!-- doxytag: member="GenomicRegion::Fix" ref="a048d2182b212789b182c2c82a7ac131e" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a048d2182b212789b182c2c82a7ac131e">Fix</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes rogue intervals (e.g. start&gt;stop) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8075e77ab6296b59506857cc21ee4c9d"></a><!-- doxytag: member="GenomicRegion::Intersect" ref="a8075e77ab6296b59506857cc21ee4c9d" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a8075e77ab6296b59506857cc21ee4c9d">Intersect</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a08db190bc63e6f1d6fc90679dc5e6868"></a><!-- doxytag: member="GenomicRegion::RunSize" ref="a08db190bc63e6f1d6fc90679dc5e6868" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a08db190bc63e6f1d6fc90679dc5e6868">RunSize</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints region label and sum of intervals' sizes. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a1277874633035d4f6fc9dabfb38998a2">ModifyPos</a> (const char *position_op, long int position_shift)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position.  <a href="#a1277874633035d4f6fc9dabfb38998a2"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a222b3b8f567c306ee5a71ea3a720ef43"></a><!-- doxytag: member="GenomicRegion::Randomize" ref="a222b3b8f567c306ee5a71ea3a720ef43" args="(gsl_rng *random_generator, StringLIntMap *bounds)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a222b3b8f567c306ee5a71ea3a720ef43">Randomize</a> (gsl_rng *random_generator, StringLIntMap *bounds)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Randomizes interval position within chromosomal bounds. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6120af435fb9ee68cb5a3d5066a4fda4"></a><!-- doxytag: member="GenomicRegion::ReversePos" ref="a6120af435fb9ee68cb5a3d5066a4fda4" args="(StringLIntMap *bounds)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a6120af435fb9ee68cb5a3d5066a4fda4">ReversePos</a> (StringLIntMap *bounds)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Reverses interval start/stop positions with respect to chromosomal bounds, as if the reference point were in the end of the chromosome as opposed to the beginning. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a461fbb00db1b45061e641f37614fb146"></a><!-- doxytag: member="GenomicRegion::Select" ref="a461fbb00db1b45061e641f37614fb146" args="(bool first, bool last, bool from5p, bool from3p)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a461fbb00db1b45061e641f37614fb146">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a0ee8c165839c79afdc586f8b5e07788e">ShiftPos</a> (long int start_shift, long int stop_shift, bool strand_aware)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately.  <a href="#a0ee8c165839c79afdc586f8b5e07788e"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab2c52ae40b34f3607f4558add94c25fd"></a><!-- doxytag: member="GenomicRegion::RunShuffle" ref="ab2c52ae40b34f3607f4558add94c25fd" args="(gsl_rng *random_generator, GenomicRegionSet *refReg, StringLIntMap *index, StringVecLIntMap *loc)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ab2c52ae40b34f3607f4558add94c25fd">RunShuffle</a> (gsl_rng *random_generator, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints shuffled region within specified reference regions in <b>refReg</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a03959dc5b5695d817b5a40b47f30a4aa"></a><!-- doxytag: member="GenomicRegion::Sort" ref="a03959dc5b5695d817b5a40b47f30a4aa" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a03959dc5b5695d817b5a40b47f30a4aa">Sort</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Sort intervals according to start position (only for compatible intervals) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a408a727103bce7b33c85601a8221a249"></a><!-- doxytag: member="GenomicRegion::RunSplit" ref="a408a727103bce7b33c85601a8221a249" args="(bool by_chrom_and_strand)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a408a727103bce7b33c85601a8221a249">RunSplit</a> (bool by_chrom_and_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print region intervals on separate lines. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae19cd6eaa883e6517ba12567cd756320"></a><!-- doxytag: member="GenomicRegion::ModifyStrand" ref="ae19cd6eaa883e6517ba12567cd756320" args="(char *strand_op)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ae19cd6eaa883e6517ba12567cd756320">ModifyStrand</a> (char *strand_op)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Modified strand information: '+'=only forward, '-'=only reverse, 'r'=reverse strands from '+' to '-' and vice versa, 'b'=both strands. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a05cb47d38946b780f7eb585074ccf393"></a><!-- doxytag: member="GenomicRegion::Union" ref="a05cb47d38946b780f7eb585074ccf393" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a05cb47d38946b780f7eb585074ccf393">Union</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute intervals' union. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3d90c15accabfd5c0279f27f3513ba7c"></a><!-- doxytag: member="GenomicRegion::RunUnion" ref="a3d90c15accabfd5c0279f27f3513ba7c" args="()" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a3d90c15accabfd5c0279f27f3513ba7c">RunUnion</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intervals' union. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3962ea71432f3977b1ac501e4e878480"></a><!-- doxytag: member="GenomicRegion::PrintWindows" ref="a3962ea71432f3977b1ac501e4e878480" args="(long int win_step, long int win_size)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a3962ea71432f3977b1ac501e4e878480">PrintWindows</a> (long int win_step, long int win_size)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa2a5b753401e9f9d64a30cfbbdf23f60"></a><!-- doxytag: member="GenomicRegion::PrintRemoveN" ref="aa2a5b753401e9f9d64a30cfbbdf23f60" args="(Chromosomes *C)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#aa2a5b753401e9f9d64a30cfbbdf23f60">PrintRemoveN</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove sub-intervals that correspond to sequences of 'N' characters. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7e2ae902016700fdc04c4607ed260edb"></a><!-- doxytag: member="GenomicRegion::PrintSearch" ref="a7e2ae902016700fdc04c4607ed260edb" args="(char *pattern, bool header, bool summary)" -->
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a7e2ae902016700fdc04c4607ed260edb">PrintSearch</a> (char *pattern, bool header, bool summary)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Search sequence for short pattern (requires SEQ input) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4b52d78052e0109b1fceafe450cfdedd"></a><!-- doxytag: member="GenomicRegion::PrintSeq" ref="a4b52d78052e0109b1fceafe450cfdedd" args="(Chromosomes *C, bool replace=false)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a4b52d78052e0109b1fceafe450cfdedd">PrintSeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool replace=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints region label, intervals and extracted sequence in SEQ format. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a33d1e5544b3fb81e8c19468b91920b1d">PrintOffsetFormat</a> (char *op, bool fraction)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the offset distances of interval start and stop position with respect to the first interval in the set.  <a href="#a33d1e5544b3fb81e8c19468b91920b1d"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7a8ac4f469447256253bdbdcb1b4a956"></a><!-- doxytag: member="GenomicRegion::PrintSeqLength" ref="a7a8ac4f469447256253bdbdcb1b4a956" args="(Chromosomes *C)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a7a8ac4f469447256253bdbdcb1b4a956">PrintSeqLength</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints region label and sequence length (excluding 'N' characters) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a77fa696d55d538b365d15a7a30391132"></a><!-- doxytag: member="GenomicRegion::PrintVerifySeq" ref="a77fa696d55d538b365d15a7a30391132" args="(Chromosomes *C, bool ignore)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a77fa696d55d538b365d15a7a30391132">PrintVerifySeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool ignore)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Verifies extracted sequence against region label. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9304039f70c5110b961cc9de77974d09"></a><!-- doxytag: member="GenomicRegion::ReverseStrand" ref="a9304039f70c5110b961cc9de77974d09" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a9304039f70c5110b961cc9de77974d09">ReverseStrand</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">If interval strand is '+', it is changed to '-' and vice versa. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a50f5ba99404da1cdae6a8632a19ff837"></a><!-- doxytag: member="GenomicRegion::Constrain" ref="a50f5ba99404da1cdae6a8632a19ff837" args="(GenomicRegion *r, char *label=NULL)" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a50f5ba99404da1cdae6a8632a19ff837">Constrain</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, char *label=NULL)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a version of this genomic region constrained by the start/stop coordinates. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a44cf789afda4e77749019045050e5ec7"></a><!-- doxytag: member="GenomicRegion::Diff" ref="a44cf789afda4e77749019045050e5ec7" args="(GenomicRegion *r)" -->
virtual <a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a44cf789afda4e77749019045050e5ec7">Diff</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the difference of this genomic region and region <b>r</b>. <br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aefe2255aeed5338060190ded05cb9c0c"></a><!-- doxytag: member="GenomicRegion::n_line" ref="aefe2255aeed5338060190ded05cb9c0c" args="" -->
unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">keeps track of line number <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7eeba95c1e87e100346688681e30ff24"></a><!-- doxytag: member="GenomicRegion::LABEL" ref="a7eeba95c1e87e100346688681e30ff24" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24">LABEL</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">genomic region label <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab10b86b03c258958818b00c21e3672a9"></a><!-- doxytag: member="GenomicRegion::I" ref="ab10b86b03c258958818b00c21e3672a9" args="" -->
GenomicIntervalSet&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9">I</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">set of intervals <br/></td></tr>
</table>
<hr/><a name="details" id="details"></a><h2>Detailed Description</h2>
<div class="textblock"><p>This class implements the genomic regions (i.e. ordered sets of genomic intervals). </p>
<p>Example: </p>
<div class="fragment"><pre class="fragment">    GenomicRegionREG x(<span class="stringliteral">&quot;Region#1\tchr1 + 132034 135932 chr1 + 135832 140102&quot;</span>);          <span class="comment">// standard REG format</span>
    GenomicRegionREG y(<span class="stringliteral">&quot;Region#2\tchr1 + 132034,135832 135932,140102&quot;</span>);                         <span class="comment">// compact REG format</span>
    cout &lt;&lt; <span class="stringliteral">&quot;x = \n&quot;</span>; x.Print();
    cout &lt;&lt; <span class="stringliteral">&quot;y = \n&quot;</span>; y.Print(); 
</pre></div> </div><hr/><h2>Member Function Documentation</h2>
<a class="anchor" id="a1277874633035d4f6fc9dabfb38998a2"></a><!-- doxytag: member="GenomicRegion::ModifyPos" ref="a1277874633035d4f6fc9dabfb38998a2" args="(const char *position_op, long int position_shift)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegion::ModifyPos </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>position_op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>position_shift</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">position_op</td><td>selects fixed position as follows: '1'=start position, 'c'=center position, '5p'=5-prime, '3p'=3-prime </td></tr>
    <tr><td class="paramname">position_shift</td><td>determines the shift of the non-fixed position (in the 'c' case, both positions are shifted symmetrically around the center) </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented in <a class="el" href="classGenomicRegionBED.html#aa001c77413f551274cd2f4c1b7598d3b">GenomicRegionBED</a>, and <a class="el" href="classGenomicRegionSAM.html#a8a5d2a1d1980dcfc238fca5f63765e1f">GenomicRegionSAM</a>.</p>

</div>
</div>
<a class="anchor" id="a33d1e5544b3fb81e8c19468b91920b1d"></a><!-- doxytag: member="GenomicRegion::PrintOffsetFormat" ref="a33d1e5544b3fb81e8c19468b91920b1d" args="(char *op, bool fraction)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegion::PrintOffsetFormat </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>fraction</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculates the offset distances of interval start and stop position with respect to the first interval in the set. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">op</td><td>determines reference point as follows: '1'=start position, '2'=stop position, '5p'=5-prime position, '3p'=3-prime position </td></tr>
    <tr><td class="paramname">fraction</td><td>if 'true', offsets are reported as a fraction of the total region length </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a0ee8c165839c79afdc586f8b5e07788e"></a><!-- doxytag: member="GenomicRegion::ShiftPos" ref="a0ee8c165839c79afdc586f8b5e07788e" args="(long int start_shift, long int stop_shift, bool strand_aware)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegion::ShiftPos </td>
          <td>(</td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>start_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>stop_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>strand_aware</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">start_shift</td><td>determines the shift of the start position </td></tr>
    <tr><td class="paramname">stop_shift</td><td>determines the shift of the stop position </td></tr>
    <tr><td class="paramname">strand_aware</td><td>if 'true', then start=5-prime and stop=3-prime </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented in <a class="el" href="classGenomicRegionBED.html#afd2f05cec2af5186794f40f221e040b1">GenomicRegionBED</a>, and <a class="el" href="classGenomicRegionSAM.html#afb2701ba1a521ae2c07ea0ace2f9ee77">GenomicRegionSAM</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a></li>
<li>genomic_intervals.cpp</li>
</ul>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Mon Jan 7 2013 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
