<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>GenomicRegionGFF Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a>  </div>
  <div class="headertitle">
<div class="title">GenomicRegionGFF Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="GenomicRegionGFF" --><!-- doxytag: inherits="GenomicRegion" -->
<p>This class processes genomic regions in SAM format.  
 <a href="classGenomicRegionGFF.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for GenomicRegionGFF:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classGenomicRegionGFF.png" usemap="#GenomicRegionGFF_map" alt=""/>
  <map id="GenomicRegionGFF_map" name="GenomicRegionGFF_map">
<area href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)." alt="GenomicRegion" shape="rect" coords="0,0,125,24"/>
</map>
 </div></div>

<p><a href="classGenomicRegionGFF-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9a49f0dc39116964f774e7ce78cfc827"></a><!-- doxytag: member="GenomicRegionGFF::GenomicRegionGFF" ref="a9a49f0dc39116964f774e7ce78cfc827" args="(FileBuffer *B)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a9a49f0dc39116964f774e7ce78cfc827">GenomicRegionGFF</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a59dee8d6f9e73b1da46386497030beb2"></a><!-- doxytag: member="GenomicRegionGFF::GenomicRegionGFF" ref="a59dee8d6f9e73b1da46386497030beb2" args="(char *inp, long int n_line=0)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a59dee8d6f9e73b1da46386497030beb2">GenomicRegionGFF</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3885d42a347283a5fed52791988076ad"></a><!-- doxytag: member="GenomicRegionGFF::~GenomicRegionGFF" ref="a3885d42a347283a5fed52791988076ad" args="()" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a3885d42a347283a5fed52791988076ad">~GenomicRegionGFF</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class destructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aeb1390d42b388e32d98f58f57a5c169c"></a><!-- doxytag: member="GenomicRegionGFF::Read" ref="aeb1390d42b388e32d98f58f57a5c169c" args="(char *inp, long int n_line=0)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#aeb1390d42b388e32d98f58f57a5c169c">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Reads GFF format from input string <b>inp</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3ffa71d729e4ff060a15a6bab1e924c8"></a><!-- doxytag: member="GenomicRegionGFF::Print" ref="a3ffa71d729e4ff060a15a6bab1e924c8" args="(FILE *file_ptr=stdout)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a3ffa71d729e4ff060a15a6bab1e924c8">Print</a> (FILE *file_ptr=stdout)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a96539f4372048c9f0d87826f5856a8ef"></a><!-- doxytag: member="GenomicRegionGFF::PrintConstrained" ref="a96539f4372048c9f0d87826f5856a8ef" args="(GenomicRegion *r, bool merge_labels=false)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a96539f4372048c9f0d87826f5856a8ef">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionGFF.html#a3ffa71d729e4ff060a15a6bab1e924c8" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>start</b> and <b>stop</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4b331cd12506aa4793a2cf6ea7620936"></a><!-- doxytag: member="GenomicRegionGFF::PrintModified" ref="a4b331cd12506aa4793a2cf6ea7620936" args="(char *label, long int start, long int stop)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a4b331cd12506aa4793a2cf6ea7620936">PrintModified</a> (char *label, long int start, long int stop)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionGFF.html#a3ffa71d729e4ff060a15a6bab1e924c8" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aee70fcc81c9a6629bf0aa1ec37cb1d64"></a><!-- doxytag: member="GenomicRegionGFF::Diff" ref="aee70fcc81c9a6629bf0aa1ec37cb1d64" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#aee70fcc81c9a6629bf0aa1ec37cb1d64">Diff</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0ff4768578a03011fabad309072fade4"></a><!-- doxytag: member="GenomicRegionGFF::RunCalcDistances" ref="a0ff4768578a03011fabad309072fade4" args="(char *op1, char *op2)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a0ff4768578a03011fabad309072fade4">RunCalcDistances</a> (char *op1, char *op2)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print distance between successive intervals. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a77975c03dd39b404098eb519e81f67d4"></a><!-- doxytag: member="GenomicRegionGFF::Divide" ref="a77975c03dd39b404098eb519e81f67d4" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a77975c03dd39b404098eb519e81f67d4">Divide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad4acc1231677efd876084ef36db82de1"></a><!-- doxytag: member="GenomicRegionGFF::RunDivide" ref="ad4acc1231677efd876084ef36db82de1" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#ad4acc1231677efd876084ef36db82de1">RunDivide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad35d9b6e9c57270c0d5648ae9f702959"></a><!-- doxytag: member="GenomicRegionGFF::Intersect" ref="ad35d9b6e9c57270c0d5648ae9f702959" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#ad35d9b6e9c57270c0d5648ae9f702959">Intersect</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aaa9b1271561e2d62ef1be6464e589cf4"></a><!-- doxytag: member="GenomicRegionGFF::Select" ref="aaa9b1271561e2d62ef1be6464e589cf4" args="(bool first, bool last, bool from5p, bool from3p)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#aaa9b1271561e2d62ef1be6464e589cf4">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abc08552b0ad4ca512f79064b8cd64e54"></a><!-- doxytag: member="GenomicRegionGFF::RunSplit" ref="abc08552b0ad4ca512f79064b8cd64e54" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#abc08552b0ad4ca512f79064b8cd64e54">RunSplit</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print region intervals on separate lines. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3815f1b27aef231a674c527e11f25880"></a><!-- doxytag: member="GenomicRegionGFF::PrintWindows" ref="a3815f1b27aef231a674c527e11f25880" args="(long int win_step, long int win_size)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a3815f1b27aef231a674c527e11f25880">PrintWindows</a> (long int win_step, long int win_size)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions) <br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aaec963f555e7e807fb5046566622f5ee"></a><!-- doxytag: member="GenomicRegionGFF::n_tokens" ref="aaec963f555e7e807fb5046566622f5ee" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#aaec963f555e7e807fb5046566622f5ee">n_tokens</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">number of TAB-separated fields in GFF entry (= 11) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2f0d4a5f15d8a756233436af03341734"></a><!-- doxytag: member="GenomicRegionGFF::SOURCE" ref="a2f0d4a5f15d8a756233436af03341734" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a2f0d4a5f15d8a756233436af03341734">SOURCE</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">feature source <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0de447fbfad3da5b7a9cf07a5a6980f0"></a><!-- doxytag: member="GenomicRegionGFF::FEATURE" ref="a0de447fbfad3da5b7a9cf07a5a6980f0" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a0de447fbfad3da5b7a9cf07a5a6980f0">FEATURE</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">feture type <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0e92da1f0fd153504e0630f27229ba4d"></a><!-- doxytag: member="GenomicRegionGFF::SCORE" ref="a0e92da1f0fd153504e0630f27229ba4d" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a0e92da1f0fd153504e0630f27229ba4d">SCORE</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">score <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a42122a69eec385e0106e3861cb3b98f3"></a><!-- doxytag: member="GenomicRegionGFF::FRAME" ref="a42122a69eec385e0106e3861cb3b98f3" args="" -->
char&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a42122a69eec385e0106e3861cb3b98f3">FRAME</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">coding frame <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a625a023f1bb24fc597e723919a036073"></a><!-- doxytag: member="GenomicRegionGFF::COMMENT" ref="a625a023f1bb24fc597e723919a036073" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionGFF.html#a625a023f1bb24fc597e723919a036073">COMMENT</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">comment <br/></td></tr>
</table>
<hr/><a name="details" id="details"></a><h2>Detailed Description</h2>
<div class="textblock"><p>This class processes genomic regions in SAM format. </p>
</div><hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a></li>
<li>genomic_intervals.cpp</li>
</ul>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Tue Jun 26 2012 15:41:31 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
