<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>genomic_intervals.h Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
<h1>genomic_intervals.h</h1><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">//</span>
<a name="l00002"></a>00002 <span class="comment">// Copyright (c) 2011 IBM Corporation. </span>
<a name="l00003"></a>00003 <span class="comment">// All rights reserved. This program and the accompanying materials are made available under the terms of the GNU General Public License v2.0 </span>
<a name="l00004"></a>00004 <span class="comment">// which accompanies this distribution, and is available at http://www.gnu.org/licenses/old-licenses/gpl-2.0.html</span>
<a name="l00005"></a>00005 <span class="comment">//</span>
<a name="l00006"></a>00006 
<a name="l00007"></a>00007 <span class="preprocessor">#include &lt;stdio.h&gt;</span>
<a name="l00008"></a>00008 <span class="preprocessor">#include &lt;stdlib.h&gt;</span>
<a name="l00009"></a>00009 <span class="preprocessor">#include &lt;string.h&gt;</span>
<a name="l00010"></a>00010 <span class="preprocessor">#include &lt;math.h&gt;</span>
<a name="l00011"></a>00011 <span class="preprocessor">#include &lt;unistd.h&gt;</span>
<a name="l00012"></a>00012 <span class="preprocessor">#include &lt;ctype.h&gt;</span>
<a name="l00013"></a>00013 <span class="preprocessor">#include &lt;sys/types.h&gt;</span>
<a name="l00014"></a>00014 <span class="preprocessor">#include &lt;sys/stat.h&gt;</span>
<a name="l00015"></a>00015 <span class="preprocessor">#include &lt;map&gt;</span>
<a name="l00016"></a>00016 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00017"></a>00017 <span class="preprocessor">#include &lt;list&gt;</span>
<a name="l00018"></a>00018 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00019"></a>00019 <span class="preprocessor">#include &lt;iostream&gt;</span>
<a name="l00020"></a>00020 <span class="preprocessor">#include &lt;algorithm&gt;</span>
<a name="l00021"></a>00021 
<a name="l00022"></a>00022 
<a name="l00023"></a>00023 <span class="keyword">using namespace </span>std;
<a name="l00024"></a>00024 
<a name="l00025"></a>00025 <span class="comment">// forward declarations</span>
<a name="l00026"></a>00026 <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>;
<a name="l00027"></a>00027 <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>;
<a name="l00028"></a>00028 <span class="keyword">typedef</span> vector&lt;GenomicInterval *&gt; GenomicIntervalSetAsVector;
<a name="l00029"></a>00029 <span class="keyword">typedef</span> list&lt;GenomicInterval *&gt; GenomicIntervalSetAsList;
<a name="l00030"></a>00030 <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="UNDER DEVELOPMENT: This class implements the an array of genomic intervals.">GenomicIntervalSetAsArray</a>;
<a name="l00031"></a>00031 <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>;
<a name="l00032"></a>00032 <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a>;
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 <span class="comment">// types</span>
<a name="l00035"></a>00035 <span class="keyword">typedef</span> map&lt;string,string&gt; StringMap;
<a name="l00036"></a>00036 <span class="keyword">typedef</span> map&lt;string,long int&gt; StringLIntMap;
<a name="l00037"></a>00037 <span class="keyword">typedef</span> map&lt;string,vector&lt;long int&gt; &gt; StringVecLIntMap;
<a name="l00038"></a>00038 
<a name="l00039"></a>00039 <span class="comment">// choose container type for interval set (vector, list or array)</span>
<a name="l00040"></a>00040 <span class="keyword">typedef</span> GenomicIntervalSetAsVector GenomicIntervalSet;
<a name="l00041"></a>00041 <span class="comment">//typedef GenomicIntervalSetAsList GenomicIntervalSet;</span>
<a name="l00042"></a>00042 <span class="comment">//typedef GenomicIntervalSetAsArray GenomicIntervalSet;</span>
<a name="l00043"></a>00043 
<a name="l00044"></a>00044 <span class="comment">// for buffer in GenomicRegionSetOverlapScanner</span>
<a name="l00045"></a>00045 <span class="keyword">typedef</span> list&lt;GenomicRegion *&gt; GenomicRegionList;
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 
<a name="l00048"></a>00048 <span class="comment">// constants</span>
<a name="l00049"></a>00049 <span class="preprocessor">#define ALIGNMENT_GAP '*'                                       </span>
<a name="l00050"></a>00050 <span class="preprocessor">const char refCIGARop[] = "MD=X";                       </span>
<a name="l00051"></a>00051 <span class="preprocessor">const char fragCIGARop[] = "MIS=X";                     </span>
<a name="l00052"></a>00052 <span class="preprocessor"></span>
<a name="l00053"></a>00053 <span class="preprocessor"></span>
<a name="l00054"></a>00054 
<a name="l00055"></a>00055   
<a name="l00056"></a>00056   
<a name="l00057"></a>00057 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00058"></a>00058 <span class="comment">// CLASS DECLARATION: GenomicInterval                                                          //</span>
<a name="l00059"></a>00059 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00061"></a>00061 <span class="comment"></span>
<a name="l00075"></a>00075 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00076"></a><a class="code" href="classGenomicInterval.html">00076</a> <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>
<a name="l00077"></a>00077 {
<a name="l00078"></a>00078  <span class="keyword">public</span>:
<a name="l00080"></a>00080   <a class="code" href="classGenomicInterval.html#09dcaee742ca117a8caaa1c656d0e354" title="Class constructor, see example below.">GenomicInterval</a>(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#0dbe6195570468c386eea923e30c762c" title="file line number">n_line</a>=0);
<a name="l00081"></a>00081 
<a name="l00082"></a>00082   
<a name="l00083"></a>00083 
<a name="l00085"></a>00085   <a class="code" href="classGenomicInterval.html#09dcaee742ca117a8caaa1c656d0e354" title="Class constructor, see example below.">GenomicInterval</a>(<span class="keywordtype">string</span> chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#0dbe6195570468c386eea923e30c762c" title="file line number">n_line</a>=0);
<a name="l00086"></a>00086 
<a name="l00087"></a>00087   
<a name="l00088"></a>00088 
<a name="l00090"></a>00090   <a class="code" href="classGenomicInterval.html#09dcaee742ca117a8caaa1c656d0e354" title="Class constructor, see example below.">GenomicInterval</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">char</span> delimiter=<span class="charliteral">' '</span>, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#0dbe6195570468c386eea923e30c762c" title="file line number">n_line</a>=0);
<a name="l00091"></a>00091 
<a name="l00092"></a>00092   
<a name="l00093"></a>00093 
<a name="l00095"></a>00095   <a class="code" href="classGenomicInterval.html#09dcaee742ca117a8caaa1c656d0e354" title="Class constructor, see example below.">GenomicInterval</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00096"></a>00096 
<a name="l00097"></a>00097   
<a name="l00098"></a>00098   
<a name="l00100"></a>00100   <a class="code" href="classGenomicInterval.html#632a255d16cb13b15b2f2d4f826cace7" title="Class Destructor.">~GenomicInterval</a>();
<a name="l00101"></a>00101 
<a name="l00102"></a>00102   
<a name="l00103"></a>00103 
<a name="l00104"></a>00104   <span class="comment">//------------------------------------------------------//</span>
<a name="l00105"></a>00105   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00106"></a>00106   <span class="comment">//------------------------------------------------------//</span>
<a name="l00107"></a>00107 
<a name="l00108"></a>00108 
<a name="l00110"></a>00110   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#feba30dc21d2e5ed7e460eb1b79f021b" title="Print interval (do not print newline character).">PrintInterval</a>();
<a name="l00111"></a>00111 
<a name="l00112"></a>00112   
<a name="l00113"></a>00113 
<a name="l00115"></a>00115   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#feba30dc21d2e5ed7e460eb1b79f021b" title="Print interval (do not print newline character).">PrintInterval</a>(FILE *file_ptr);
<a name="l00116"></a>00116 
<a name="l00117"></a>00117   
<a name="l00118"></a>00118 
<a name="l00120"></a>00120   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#feba30dc21d2e5ed7e460eb1b79f021b" title="Print interval (do not print newline character).">PrintInterval</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00121"></a>00121 
<a name="l00122"></a>00122 
<a name="l00124"></a>00124   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#70f300009a0cc29b024853fb794d0e9a" title="Print interval with a newline character at the end. If point is &amp;#39;true&amp;#39;, then...">Print</a>(<span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00125"></a>00125 
<a name="l00126"></a>00126   
<a name="l00127"></a>00127 
<a name="l00129"></a>00129   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#dbd1571cf6db0ef646f9b1e095bfcb4e" title="Print interval with a label in front and newline character at the end.">PrintWithLabel</a>(<span class="keywordtype">char</span> *label, <span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00130"></a>00130 
<a name="l00131"></a>00131   
<a name="l00132"></a>00132   
<a name="l00133"></a>00133   <span class="comment">//------------------------------------------------------//</span>
<a name="l00134"></a>00134   <span class="comment">//   Check                                              //</span>
<a name="l00135"></a>00135   <span class="comment">//------------------------------------------------------//</span>
<a name="l00136"></a>00136 
<a name="l00137"></a>00137 
<a name="l00139"></a>00139   <span class="keywordtype">bool</span> <a class="code" href="classGenomicInterval.html#79d1e06f62b4c1eddda7b9fc62733c7b" title="Check integrity of interval start/stop positions (start&amp;gt;=1 and stop&amp;gt;=start)...">IsValid</a>();
<a name="l00140"></a>00140 
<a name="l00141"></a>00141   
<a name="l00142"></a>00142 
<a name="l00144"></a>00144   <span class="keywordtype">bool</span> <a class="code" href="classGenomicInterval.html#b9323c2818c13a25fda953ae18fc97aa" title="Same as GenomicInterval::IsValid, but it also prints error messages.">CheckValid</a>(<span class="keywordtype">bool</span> quiet=<span class="keyword">false</span>);
<a name="l00145"></a>00145 
<a name="l00146"></a>00146   
<a name="l00147"></a>00147 
<a name="l00149"></a>00149   <span class="keywordtype">bool</span> <a class="code" href="classGenomicInterval.html#36dc2d4059337cc03a1211b6b8a7471c" title="Compares with input interval I and returns true if its order is before I. The order...">IsBefore</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00150"></a>00150 
<a name="l00151"></a>00151 
<a name="l00152"></a>00152 
<a name="l00154"></a>00154   <span class="keywordtype">bool</span> <a class="code" href="classGenomicInterval.html#8fc5d344a103f145db968e536b30f9cf" title="Compares with input interval I and returns true if its position order is before I...">IsPosBefore</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00155"></a>00155 
<a name="l00156"></a>00156 
<a name="l00157"></a>00157 
<a name="l00159"></a>00159   <span class="keywordtype">bool</span> <a class="code" href="classGenomicInterval.html#b6febc4c04d4919a9485ce8eeec5129d" title="Compares with input interval I and returns true if intervals are in the same chromosome...">IsCompatibleWith</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00160"></a>00160 
<a name="l00161"></a>00161 
<a name="l00162"></a>00162 
<a name="l00164"></a>00164   <span class="keywordtype">bool</span> <a class="code" href="classGenomicInterval.html#a1c591b6d3cc3c576cec24a3ab67abfb" title="Returns &amp;#39;true&amp;#39; if it overlaps with input interval I.">OverlapsWith</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand=<span class="keyword">false</span>);
<a name="l00165"></a>00165 
<a name="l00166"></a>00166 
<a name="l00167"></a>00167 
<a name="l00168"></a>00168   <span class="comment">//------------------------------------------------------//</span>
<a name="l00169"></a>00169   <span class="comment">//   Set &amp; Get                                          //</span>
<a name="l00170"></a>00170   <span class="comment">//------------------------------------------------------//</span>
<a name="l00171"></a>00171 
<a name="l00172"></a>00172 
<a name="l00174"></a>00174   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#c76f5c0d3cae831f6cec9d8a27b1c56d" title="Set start/stop positions to new values.">SetCoordinates</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00175"></a>00175 
<a name="l00176"></a>00176   
<a name="l00177"></a>00177 
<a name="l00179"></a>00179   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#e068e54e894fb239b61ec6d29e5f70bf" title="If: op=&amp;#39;1&amp;#39; returns start position, op=&amp;#39;2&amp;#39; returns stop position,...">GetCoordinate</a>(<span class="keywordtype">char</span> *op);
<a name="l00180"></a>00180   
<a name="l00181"></a>00181 
<a name="l00182"></a>00182 
<a name="l00184"></a>00184   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicInterval.html#7ceb36d952453db9473c7548c885f40a" title="Returns interval size.">GetSize</a>();
<a name="l00185"></a>00185   
<a name="l00186"></a>00186 
<a name="l00187"></a>00187 
<a name="l00189"></a>00189   <span class="keywordtype">char</span> *<a class="code" href="classGenomicInterval.html#ff25b35226220508539b1183d5b9c11d" title="Extracts genomic interval sequence from chromosomal sequences supplied in C. If parameter...">GetSeq</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00190"></a>00190   
<a name="l00191"></a>00191 
<a name="l00192"></a>00192 
<a name="l00194"></a>00194   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicInterval.html#39cd65d08a0c64a8bbddec32c8decbd3" title="Returns sequence length excluding &amp;#39;N&amp;#39; characters.">GetSeqLength</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00195"></a>00195   
<a name="l00196"></a>00196 
<a name="l00197"></a>00197 
<a name="l00198"></a>00198   <span class="comment">//------------------------------------------------------//</span>
<a name="l00199"></a>00199   <span class="comment">//   Operations                                         //</span>
<a name="l00200"></a>00200   <span class="comment">//------------------------------------------------------//</span>
<a name="l00201"></a>00201 
<a name="l00202"></a>00202 
<a name="l00204"></a>00204   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#ea112f495d21eda675de671d9ff2e5ad" title="Calculates the overlap with input interval I.">CalcOverlap</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00205"></a>00205 
<a name="l00206"></a>00206 
<a name="l00207"></a>00207 
<a name="l00209"></a>00209   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#a801336aa07a0029d97d7914a747abd3" title="Calculates the distance from input interval I based on position operators op and...">CalcDistanceFrom</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">char</span> *op, <span class="keywordtype">char</span> *I_op);
<a name="l00210"></a>00210   
<a name="l00211"></a>00211 
<a name="l00212"></a>00212 
<a name="l00214"></a>00214   <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#7856012eda653222a1ac5482510bacc7" title="Returns -1 if before interval i, +1 if after, and 0 if the intervals overlap. The...">CalcDirection</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00215"></a>00215   
<a name="l00216"></a>00216 
<a name="l00217"></a>00217 
<a name="l00219"></a>00219   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#b14f526a4df4f148e1b843ee0dc10d04" title="Prints genomic interval sequence from chromosomal sequences supplied in C. If parameter...">PrintSeq</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00220"></a>00220   
<a name="l00221"></a>00221 
<a name="l00222"></a>00222 
<a name="l00224"></a>00224   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#a410816fa04d76edd7caeb9ae2442ae1" title="If interval strand is &amp;#39;+&amp;#39;, it is changed to &amp;#39;-&amp;#39; and vice versa.">ReverseStrand</a>(); 
<a name="l00225"></a>00225   
<a name="l00226"></a>00226 
<a name="l00227"></a>00227 
<a name="l00229"></a>00229 
<a name="l00234"></a>00234   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#5cc57eb991cca3f1e1b8bd7ca8acaa03" title="Shifts interval start/stop positions. First it chooses reference position, and then...">ShiftPos</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> shift_5prime);
<a name="l00235"></a>00235   
<a name="l00236"></a>00236 
<a name="l00237"></a>00237 
<a name="l00239"></a>00239 
<a name="l00243"></a>00243   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#96b34b76019001c4751d4c4f868ee24d" title="Modifies interval start/stop positions. First it chooses which position to keep fixed...">ModifyPos</a>(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00244"></a>00244 
<a name="l00245"></a>00245 
<a name="l00246"></a>00246 
<a name="l00248"></a>00248   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#bf416450f25b579dfb76b2bd6a81aa6a" title="Reverses interval start/stop positions with respect to chromosomal bounds, as if...">ReversePos</a>(StringLIntMap *bounds);
<a name="l00249"></a>00249 
<a name="l00250"></a>00250 
<a name="l00251"></a>00251 
<a name="l00253"></a>00253   <span class="keywordtype">bool</span> <a class="code" href="classGenomicInterval.html#1834c0f1217585398de74119d680ab7b" title="Checks if interval start/stop positions comply with chromosomal bounds.">CheckBounds</a>(StringLIntMap *bounds, <span class="keywordtype">bool</span> ignore);
<a name="l00254"></a>00254 
<a name="l00255"></a>00255 
<a name="l00256"></a>00256 
<a name="l00258"></a>00258   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#c28099b569624adcf3b327e922e5e4e5" title="Corrects interval start/stop positions so as to comply with chromosomal bounds.">ApplyBounds</a>(StringLIntMap *bounds);
<a name="l00259"></a>00259 
<a name="l00260"></a>00260 
<a name="l00261"></a>00261 
<a name="l00263"></a>00263   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#c06925f3fcf6b47cec5c6f95aaa379c0" title="Prints in BED format. If convert_chromosome is &amp;#39;true&amp;#39;, then convert from...">PrintBEDFormat</a>(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00264"></a>00264 
<a name="l00265"></a>00265 
<a name="l00266"></a>00266 
<a name="l00268"></a>00268   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#75aed02c1df83acb5d7ec89e2a9949bd" title="Randomizes interval position within chromosomal bounds.">Randomize</a>(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00269"></a>00269 
<a name="l00270"></a>00270 
<a name="l00271"></a>00271 
<a name="l00273"></a>00273 
<a name="l00280"></a>00280   <span class="keywordtype">void</span> <a class="code" href="classGenomicInterval.html#ab40ad7dd997948e0c04a13164e955d6" title="Computes start/stop offset distances from ReferenceI.">GetOffsetFrom</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *ReferenceI, <span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *start_offset, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *stop_offset);
<a name="l00281"></a>00281 
<a name="l00282"></a>00282 
<a name="l00283"></a>00283   
<a name="l00284"></a>00284 
<a name="l00285"></a>00285   <span class="comment">//------------------------------------------------------//</span>
<a name="l00286"></a>00286   <span class="comment">//   Data                                               //</span>
<a name="l00287"></a>00287   <span class="comment">//------------------------------------------------------//</span>
<a name="l00288"></a>00288   
<a name="l00289"></a><a class="code" href="classGenomicInterval.html#0dbe6195570468c386eea923e30c762c">00289</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#0dbe6195570468c386eea923e30c762c" title="file line number">n_line</a>;             
<a name="l00290"></a><a class="code" href="classGenomicInterval.html#c489a1195c43cc93d0e192dc91d5a192">00290</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicInterval.html#c489a1195c43cc93d0e192dc91d5a192" title="chromosome name (space and tab characters are not allowed)">CHROMOSOME</a>;                             
<a name="l00291"></a><a class="code" href="classGenomicInterval.html#25f6603b48fcc57f97c89fa768b658f9">00291</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicInterval.html#25f6603b48fcc57f97c89fa768b658f9" title="orientation (&amp;#39;+&amp;#39; or &amp;#39;-&amp;#39;)">STRAND</a>;                                  
<a name="l00292"></a><a class="code" href="classGenomicInterval.html#c0abba93599f5cdf4b5a4e7ec5909bbd">00292</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#c0abba93599f5cdf4b5a4e7ec5909bbd" title="interval start position in the chromosome (&amp;gt;=1)">START</a>;                               
<a name="l00293"></a><a class="code" href="classGenomicInterval.html#c5fac21da3939b7976859a54800b93f0">00293</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#c5fac21da3939b7976859a54800b93f0" title="interval stop position in the chromosome (&amp;lt;=chromosome_size)">STOP</a>;                                
<a name="l00294"></a>00294 };
<a name="l00295"></a>00295 
<a name="l00296"></a>00296 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00297"></a>00297 <span class="comment">// END CLASS DECLARATION: GenomicIntervals                                                     //</span>
<a name="l00298"></a>00298 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00299"></a>00299 
<a name="l00300"></a>00300 
<a name="l00301"></a>00301 
<a name="l00302"></a>00302 
<a name="l00303"></a>00303 
<a name="l00304"></a>00304 
<a name="l00305"></a>00305 
<a name="l00306"></a>00306 
<a name="l00307"></a>00307 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00308"></a>00308 <span class="comment">// CLASS: Chromosomes                                                                          //</span>
<a name="l00309"></a>00309 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00311"></a>00311 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00312"></a><a class="code" href="classChromosomes.html">00312</a> <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>
<a name="l00313"></a>00313 {
<a name="l00314"></a>00314  <span class="keyword">public</span>:
<a name="l00315"></a>00315 
<a name="l00316"></a><a class="code" href="classChromosomes.html#4a233d2b7fabc1a589cbbafd5d0ea243">00316</a>   <span class="keyword">typedef</span> map&lt;string,pair&lt;char*,size_t&gt; &gt; <a class="code" href="classChromosomes.html#4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a>;            
<a name="l00317"></a>00317 
<a name="l00319"></a>00319 
<a name="l00353"></a>00353   <a class="code" href="classChromosomes.html#a596430d4a05ce0dc4d4f43c650db83e" title="Class constructor.">Chromosomes</a>(<span class="keywordtype">char</span> *<a class="code" href="classChromosomes.html#8df459fcb5ad98a2524fd51ac2b5a60c" title="the directory where the chromosome map file is located">map_dir</a>, <span class="keywordtype">char</span> *<a class="code" href="classChromosomes.html#9c690f19e04e80b056dac2072476f24c" title="the name of the chromosome map file">map_name</a>);
<a name="l00354"></a>00354 
<a name="l00355"></a>00355 
<a name="l00357"></a>00357 
<a name="l00373"></a>00373   <a class="code" href="classChromosomes.html#a596430d4a05ce0dc4d4f43c650db83e" title="Class constructor.">Chromosomes</a>(<span class="keywordtype">char</span> *fasta_file, <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#954bae240564f6f589a30d253291eb75" title="if &amp;#39;true&amp;#39;, print messages while loading chromosome sequences">verbose</a>);
<a name="l00374"></a>00374 
<a name="l00375"></a>00375   
<a name="l00377"></a>00377   <a class="code" href="classChromosomes.html#0b04a0781b2ff7088dd8046493385003" title="Class destructor.">~Chromosomes</a>();
<a name="l00378"></a>00378   
<a name="l00379"></a>00379 
<a name="l00381"></a>00381   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#2378b5cab810d50699d97883fe1c04bf" title="Returns &amp;#39;true&amp;#39; if chromosome name is found in the data.">FindChromosome</a>(<span class="keywordtype">char</span> *chromosome_name);
<a name="l00382"></a>00382   
<a name="l00383"></a>00383   
<a name="l00385"></a>00385   <span class="keywordtype">char</span> *<a class="code" href="classChromosomes.html#819d0fa5f41c034a45949156307d02e4" title="Returns interval sequence.">GetSeq</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00386"></a>00386   
<a name="l00387"></a>00387   
<a name="l00389"></a>00389   <span class="keywordtype">void</span> <a class="code" href="classChromosomes.html#c732efe7b6dcc645861507257bf66347" title="Prints interval sequence.">PrintSeq</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00390"></a>00390   
<a name="l00391"></a>00391   
<a name="l00393"></a>00393   <span class="keywordtype">size_t</span> <a class="code" href="classChromosomes.html#044da4339ffcc9a892ea35198bbb9f36" title="Calculates interval sequence length excluding &amp;#39;N&amp;#39; characters.">GetSeqLength</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00394"></a>00394   
<a name="l00395"></a>00395   
<a name="l00397"></a>00397   <span class="keywordtype">void</span> <a class="code" href="classChromosomes.html#7f4e8938bd27e1fdb81f7ab33a44d62a" title="Loads chromosome sequence in memory.">LoadChromosome</a>(<span class="keywordtype">string</span> chromosome_name);
<a name="l00398"></a>00398 
<a name="l00399"></a>00399   
<a name="l00400"></a>00400   <span class="comment">// data</span>
<a name="l00401"></a>00401 
<a name="l00402"></a><a class="code" href="classChromosomes.html#954bae240564f6f589a30d253291eb75">00402</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#954bae240564f6f589a30d253291eb75" title="if &amp;#39;true&amp;#39;, print messages while loading chromosome sequences">verbose</a>;                                                         
<a name="l00403"></a><a class="code" href="classChromosomes.html#408a06b9ae635ceaf0b75452bcb62763">00403</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#408a06b9ae635ceaf0b75452bcb62763" title="&amp;#39;true&amp;#39; if reading from FASTA file">load_in_memory</a>;                                          
<a name="l00404"></a><a class="code" href="classChromosomes.html#8e40e7c99c6c0a40713c7785ba5b5ac4">00404</a>   StringMap <a class="code" href="classChromosomes.html#8e40e7c99c6c0a40713c7785ba5b5ac4" title="maps chromosome names to chromosome sequence file names">map</a>;                                                        
<a name="l00405"></a><a class="code" href="classChromosomes.html#8df459fcb5ad98a2524fd51ac2b5a60c">00405</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#8df459fcb5ad98a2524fd51ac2b5a60c" title="the directory where the chromosome map file is located">map_dir</a>;                                                       
<a name="l00406"></a><a class="code" href="classChromosomes.html#9c690f19e04e80b056dac2072476f24c">00406</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#9c690f19e04e80b056dac2072476f24c" title="the name of the chromosome map file">map_name</a>;                                                      
<a name="l00407"></a><a class="code" href="classChromosomes.html#b3b502eec7bd2028297300aeb47b5844">00407</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#b3b502eec7bd2028297300aeb47b5844" title="name of the currently used chromosome">current_chromosome_name</a>;                       
<a name="l00408"></a><a class="code" href="classChromosomes.html#fb59cb80755187f4546573f10110e669">00408</a>   <span class="keywordtype">size_t</span> <a class="code" href="classChromosomes.html#fb59cb80755187f4546573f10110e669" title="size of the currently used chromosome">current_chromosome_size</a>;                       
<a name="l00409"></a><a class="code" href="classChromosomes.html#f640cf0f09b109b3d46d70aba9207a2e">00409</a>   <span class="keywordtype">char</span> *<a class="code" href="classChromosomes.html#f640cf0f09b109b3d46d70aba9207a2e" title="pointer to the currently used chromosome sequence">current_chromosome_seq</a>;                         
<a name="l00410"></a><a class="code" href="classChromosomes.html#673169d55aa0b94815e17b41566d4d8c">00410</a>   <a class="code" href="classChromosomes.html#4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a> <a class="code" href="classChromosomes.html#673169d55aa0b94815e17b41566d4d8c" title="all chromosomes sequences and their sizes are stored here">chromosome_seq_data</a>;        
<a name="l00411"></a>00411 };
<a name="l00412"></a>00412 
<a name="l00413"></a>00413 
<a name="l00414"></a>00414 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00415"></a>00415 <span class="comment">// END CLASS: Chromosomes                                                                      //</span>
<a name="l00416"></a>00416 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00417"></a>00417 
<a name="l00418"></a>00418 
<a name="l00419"></a>00419 
<a name="l00420"></a>00420 
<a name="l00421"></a>00421 
<a name="l00422"></a>00422 
<a name="l00423"></a>00423 
<a name="l00424"></a>00424 
<a name="l00425"></a>00425 
<a name="l00426"></a>00426 
<a name="l00427"></a>00427 
<a name="l00428"></a>00428 
<a name="l00429"></a>00429 
<a name="l00430"></a>00430 
<a name="l00431"></a>00431 
<a name="l00432"></a>00432 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00433"></a>00433 <span class="comment">// CLASS: GenomicIntervalSetAsArray                                                            //</span>
<a name="l00434"></a>00434 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00436"></a>00436 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00437"></a>00437 
<a name="l00438"></a><a class="code" href="classGenomicIntervalSetAsArray.html">00438</a> <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="UNDER DEVELOPMENT: This class implements the an array of genomic intervals.">GenomicIntervalSetAsArray</a>
<a name="l00439"></a>00439 {
<a name="l00440"></a>00440  <span class="keyword">public</span>:
<a name="l00441"></a><a class="code" href="classGenomicIntervalSetAsArray.html#1a0416c9fac7e9d965ae7fe84f1649c2">00441</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>** <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a>;                           
<a name="l00442"></a><a class="code" href="classGenomicIntervalSetAsArray.html#0310f5d79c529b631a5ef71ef186b41c">00442</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>** <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">reverse_iterator</a>;           
<a name="l00443"></a>00443 
<a name="l00444"></a>00444 
<a name="l00446"></a>00446   <a class="code" href="classGenomicIntervalSetAsArray.html#12686a38ae225b2794feb5a3ef1f4430" title="Empty class constructor.">GenomicIntervalSetAsArray</a>();
<a name="l00447"></a>00447 
<a name="l00448"></a>00448 
<a name="l00450"></a>00450   <a class="code" href="classGenomicIntervalSetAsArray.html#05dafd13fef5a9db6339db5c103fef23" title="Class destructor.">~GenomicIntervalSetAsArray</a>();
<a name="l00451"></a>00451   
<a name="l00452"></a>00452   
<a name="l00454"></a><a class="code" href="classGenomicIntervalSetAsArray.html#9e56381009d164d7625199b183845388">00454</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#9e56381009d164d7625199b183845388" title="Implementation of begin() for this container.">begin</a>() { <span class="keywordflow">return</span> &amp;<a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>[0]; }
<a name="l00455"></a>00455   
<a name="l00456"></a>00456 
<a name="l00458"></a><a class="code" href="classGenomicIntervalSetAsArray.html#5b6145ceff99907d9075047519482584">00458</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#5b6145ceff99907d9075047519482584" title="Implementation of end() for this container.">end</a>() { <span class="keywordflow">return</span> &amp;<a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>[<a class="code" href="classGenomicIntervalSetAsArray.html#9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>]; }
<a name="l00459"></a>00459   
<a name="l00460"></a>00460 
<a name="l00462"></a><a class="code" href="classGenomicIntervalSetAsArray.html#c077589a55a7dfec9b2e6c38392f7444">00462</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#c077589a55a7dfec9b2e6c38392f7444" title="Implementation of rbegin() for this container.">rbegin</a>() { <span class="keywordflow">return</span> &amp;<a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>[<a class="code" href="classGenomicIntervalSetAsArray.html#9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>-1]; }
<a name="l00463"></a>00463   
<a name="l00464"></a>00464 
<a name="l00466"></a><a class="code" href="classGenomicIntervalSetAsArray.html#5b7fc7dc64857dfdadd1198393c0d117">00466</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#5b7fc7dc64857dfdadd1198393c0d117" title="Implementation of rend() for this container.">rend</a>() { <span class="keywordflow">return</span> &amp;<a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>[-1]; }
<a name="l00467"></a>00467   
<a name="l00468"></a>00468 
<a name="l00470"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a74747b36b114c326edfaf1224270d0d">00470</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#a74747b36b114c326edfaf1224270d0d" title="Implementation of front() for this container.">front</a>() { <span class="keywordflow">return</span> <a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>[0]; }
<a name="l00471"></a>00471   
<a name="l00472"></a>00472 
<a name="l00474"></a><a class="code" href="classGenomicIntervalSetAsArray.html#11b483c12111ee96c15fec296c66ee15">00474</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#11b483c12111ee96c15fec296c66ee15" title="Implementation of back() for this container.">back</a>() { <span class="keywordflow">return</span> <a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>[<a class="code" href="classGenomicIntervalSetAsArray.html#9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>-1]; }
<a name="l00475"></a>00475   
<a name="l00476"></a>00476 
<a name="l00478"></a><a class="code" href="classGenomicIntervalSetAsArray.html#d4bfef55744241849841c71d08835403">00478</a>   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicIntervalSetAsArray.html#d4bfef55744241849841c71d08835403" title="Implementation of size() for this container.">size</a>() { <span class="keywordflow">return</span> <a class="code" href="classGenomicIntervalSetAsArray.html#9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>; }
<a name="l00479"></a>00479   
<a name="l00480"></a>00480 
<a name="l00482"></a><a class="code" href="classGenomicIntervalSetAsArray.html#4348e7636cb91405b08374a966e33a7c">00482</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#4348e7636cb91405b08374a966e33a7c" title="Implementation of push_back() for this container.">push_back</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *j) { <a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>[<a class="code" href="classGenomicIntervalSetAsArray.html#9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>++] = j; }
<a name="l00483"></a>00483   
<a name="l00484"></a>00484 
<a name="l00486"></a><a class="code" href="classGenomicIntervalSetAsArray.html#833395eaf9daa5626879eefa861d9ea5">00486</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#833395eaf9daa5626879eefa861d9ea5" title="Implementation of clear() for this container.">clear</a>() {  }
<a name="l00487"></a>00487   
<a name="l00488"></a>00488 
<a name="l00490"></a><a class="code" href="classGenomicIntervalSetAsArray.html#0f8554cfca150d695d15691aa42df7fa">00490</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#0f8554cfca150d695d15691aa42df7fa" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> j) { fprintf(stderr, <span class="stringliteral">"Error: [GenomicIntervalArray]: array entries cannot be erased!"</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00491"></a>00491 
<a name="l00492"></a>00492 
<a name="l00494"></a><a class="code" href="classGenomicIntervalSetAsArray.html#0aae25ac2e9a7cf50a370703aa99df85">00494</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#0f8554cfca150d695d15691aa42df7fa" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i) { fprintf(stderr, <span class="stringliteral">"Error: [GenomicIntervalArray]: array entries cannot be erased!"</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00495"></a>00495   
<a name="l00496"></a>00496 
<a name="l00498"></a><a class="code" href="classGenomicIntervalSetAsArray.html#0cd89244c0e75b88f9b56c9b9478d06b">00498</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#0cd89244c0e75b88f9b56c9b9478d06b" title="Implementation of insert() for this container (exits with error message).">insert</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *u) { fprintf(stderr, <span class="stringliteral">"Error: [GenomicIntervalArray]: array entries cannot be inserted!"</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00499"></a>00499   
<a name="l00500"></a>00500 
<a name="l00501"></a>00501   
<a name="l00502"></a>00502   <span class="comment">//------------------------------------------------------//</span>
<a name="l00503"></a>00503   <span class="comment">//   Data                                               //</span>
<a name="l00504"></a>00504   <span class="comment">//------------------------------------------------------//</span>
<a name="l00505"></a>00505   
<a name="l00506"></a><a class="code" href="classGenomicIntervalSetAsArray.html#9f9558bbaa8d682993a60fa142893998">00506</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>;                                 
<a name="l00507"></a><a class="code" href="classGenomicIntervalSetAsArray.html#6580b875cd9804ba59f88c5000f83f09">00507</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#6580b875cd9804ba59f88c5000f83f09" title="number of maximum allowed intervals (this will be removed)">max_intervals</a>;                               
<a name="l00508"></a><a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2">00508</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> **<a class="code" href="classGenomicIntervalSetAsArray.html#f7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>;                                  
<a name="l00509"></a>00509   
<a name="l00510"></a>00510 };
<a name="l00511"></a>00511 
<a name="l00512"></a>00512 
<a name="l00513"></a>00513 
<a name="l00514"></a>00514 
<a name="l00515"></a>00515 
<a name="l00516"></a>00516 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00517"></a>00517 <span class="comment">// END CLASS: GenomicIntervalArray                                                             //</span>
<a name="l00518"></a>00518 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00519"></a>00519 
<a name="l00520"></a>00520 
<a name="l00521"></a>00521 
<a name="l00522"></a>00522 
<a name="l00523"></a>00523 
<a name="l00524"></a>00524 
<a name="l00525"></a>00525 
<a name="l00526"></a>00526 
<a name="l00527"></a>00527 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00528"></a>00528 <span class="comment">// CLASS: GenomicRegion                                                                        //</span>
<a name="l00529"></a>00529 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00530"></a>00530 <span class="comment">//                                                                                             //</span>
<a name="l00531"></a>00531 <span class="comment">// Input formats:                                                                              //</span>
<a name="l00532"></a>00532 <span class="comment">//  1. LABEL &lt;TAB&gt; CHROMOSOME STRAND START[,START]* STOP[,STOP]*                               //</span>
<a name="l00533"></a>00533 <span class="comment">//  2. LABEL &lt;TAB&gt; [CHROMOSOME STRAND START STOP]+                                             //</span>
<a name="l00534"></a>00534 <span class="comment">//                                                                                             //</span>
<a name="l00535"></a>00535 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00537"></a>00537 <span class="comment"></span>
<a name="l00546"></a>00546 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00547"></a>00547 
<a name="l00548"></a><a class="code" href="classGenomicRegion.html">00548</a> <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>
<a name="l00549"></a>00549 {
<a name="l00550"></a>00550  <span class="keyword">public</span>:
<a name="l00552"></a>00552   <a class="code" href="classGenomicRegion.html#8ff9fe5618b932a1caccbd9072ddaaef" title="Empty class constructor.">GenomicRegion</a>();
<a name="l00553"></a>00553 
<a name="l00554"></a>00554 
<a name="l00555"></a>00555 
<a name="l00557"></a>00557   <a class="code" href="classGenomicRegion.html#8ff9fe5618b932a1caccbd9072ddaaef" title="Empty class constructor.">GenomicRegion</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *B);
<a name="l00558"></a>00558 
<a name="l00559"></a>00559 
<a name="l00560"></a>00560 
<a name="l00562"></a>00562   <a class="code" href="classGenomicRegion.html#8ff9fe5618b932a1caccbd9072ddaaef" title="Empty class constructor.">GenomicRegion</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0); 
<a name="l00563"></a>00563 
<a name="l00564"></a>00564 
<a name="l00565"></a>00565 
<a name="l00567"></a>00567   <a class="code" href="classGenomicRegion.html#8ff9fe5618b932a1caccbd9072ddaaef" title="Empty class constructor.">GenomicRegion</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i);
<a name="l00568"></a>00568 
<a name="l00569"></a>00569 
<a name="l00570"></a>00570 
<a name="l00572"></a>00572   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html#37af413ecbed6ee0d36ecbc40d957947" title="Class destructor.">~GenomicRegion</a>();
<a name="l00573"></a>00573   
<a name="l00574"></a>00574 
<a name="l00575"></a>00575 
<a name="l00576"></a>00576   <span class="comment">//------------------------------------------------------//</span>
<a name="l00577"></a>00577   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00578"></a>00578   <span class="comment">//------------------------------------------------------//</span>
<a name="l00579"></a>00579 
<a name="l00580"></a>00580 
<a name="l00582"></a>00582   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#fd188daab9b4e97ef66883d6971fe103" title="Reads label and genomic region intervals from input string inp.">Read</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0, <span class="keywordtype">char</span> del1=<span class="charliteral">' '</span>, <span class="keywordtype">char</span> del2=<span class="charliteral">','</span>);
<a name="l00583"></a>00583 
<a name="l00584"></a>00584 
<a name="l00585"></a>00585 
<a name="l00587"></a>00587   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#40397d1fab7b00e5f04b9f7e52f0114e" title="Prints label and genomic region intervals with a newline character at the end.">Print</a>(FILE *file_ptr=stdout);
<a name="l00588"></a>00588 
<a name="l00589"></a>00589 
<a name="l00590"></a>00590   
<a name="l00592"></a>00592   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#b9ea31af7fbb32f206efe0e64533a5be" title="Same as Print(), but the label and interval&amp;#39;s coordinates are modified.">PrintModified</a>(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00593"></a>00593 
<a name="l00594"></a>00594 
<a name="l00595"></a>00595 
<a name="l00597"></a>00597   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#ba987d40590874eebd1dd8bb0c02bdc8" title="Prints all pairwise intersection with the intervals in region r (compatible, sorted...">PrintIntersection</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l00598"></a>00598 
<a name="l00599"></a>00599 
<a name="l00600"></a>00600 
<a name="l00602"></a>00602   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#cfe5347d9c92da801ce7e97416fc8758" title="Same as Print(), but the interval is constrained between r-&amp;gt;I.front()-&amp;gt;START...">PrintConstrained</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l00603"></a>00603 
<a name="l00604"></a>00604 
<a name="l00605"></a>00605 
<a name="l00607"></a>00607   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#9c687565a8ed25da5a5ccf08e5e8461e" title="Prints label and genomic region intervals with a newline character at the end.">PrintREG</a>(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l00608"></a>00608 
<a name="l00609"></a>00609 
<a name="l00610"></a>00610 
<a name="l00612"></a>00612   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#9a00b031d4a119181b8b76ffa34a6f47" title="Prints sequences (SEQ format).">PrintQ</a>(<span class="keywordtype">bool</span> reverse=<span class="keyword">false</span>);
<a name="l00613"></a>00613 
<a name="l00614"></a>00614 
<a name="l00615"></a>00615 
<a name="l00617"></a>00617   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#68ecc1c8c76e9204f6e27801f0ba3d5f" title="Prints error message and exits.">PrintError</a>(<span class="keywordtype">string</span> error_msg);
<a name="l00618"></a>00618 
<a name="l00619"></a>00619 
<a name="l00620"></a>00620 
<a name="l00621"></a>00621 
<a name="l00622"></a>00622   <span class="comment">//------------------------------------------------------//</span>
<a name="l00623"></a>00623   <span class="comment">//   Get &amp; Set                                          //</span>
<a name="l00624"></a>00624   <span class="comment">//------------------------------------------------------//</span>
<a name="l00625"></a>00625 
<a name="l00626"></a>00626 
<a name="l00628"></a>00628   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#ad8566a736786134b72e7bf74296e807" title="Deletes intervals from iterator i to j.">DeleteIntervals</a>(GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j); 
<a name="l00629"></a>00629 
<a name="l00630"></a>00630 
<a name="l00631"></a>00631 
<a name="l00633"></a>00633   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#bc547e770917e859fa434d6c54954916" title="Changes the label.">SetLabel</a>(<span class="keywordtype">char</span> *label);
<a name="l00634"></a>00634 
<a name="l00635"></a>00635 
<a name="l00636"></a>00636 
<a name="l00638"></a>00638   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#e795532404bad8b93bec632c20d12b9d" title="Interval strands are set to strand.">SetStrand</a>(<span class="keywordtype">char</span> strand);
<a name="l00639"></a>00639 
<a name="l00640"></a>00640 
<a name="l00641"></a>00641 
<a name="l00643"></a>00643   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegion.html#815baa00d7d33f179181560a1fe468fc" title="Returns pointer to chromosome name.">GetChromosome</a>();
<a name="l00644"></a>00644 
<a name="l00645"></a>00645 
<a name="l00646"></a>00646 
<a name="l00648"></a>00648   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicRegion.html#5bff7a3e83b0bc9bfb5441270d0ba1ae" title="Returns sum of interval&amp;#39;s sizes.">GetSize</a>();
<a name="l00649"></a>00649 
<a name="l00650"></a>00650 
<a name="l00651"></a>00651 
<a name="l00653"></a>00653   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicRegion.html#719143f1affa44c38493b2cd43448d3b" title="Returns genomic region sequence length excluding &amp;#39;N&amp;#39; characters.">GetSeqLength</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00654"></a>00654 
<a name="l00655"></a>00655 
<a name="l00656"></a>00656 
<a name="l00658"></a>00658   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegion.html#fcf7c053989b87f425f084d1f2c078f5" title="Returns genomic region sequence.">GetSeq</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00659"></a>00659 
<a name="l00660"></a>00660 
<a name="l00661"></a>00661 
<a name="l00662"></a>00662   <span class="comment">//------------------------------------------------------//</span>
<a name="l00663"></a>00663   <span class="comment">//   Check                                              //</span>
<a name="l00664"></a>00664   <span class="comment">//------------------------------------------------------//</span>
<a name="l00665"></a>00665 
<a name="l00666"></a>00666 
<a name="l00668"></a>00668   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#58c2fe162c1e9bbbd7c90d5abd011ff0" title="Returns &amp;#39;true&amp;#39; if all intervals are in the same chromosome and strand.">IsCompatible</a>(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00669"></a>00669 
<a name="l00670"></a>00670 
<a name="l00671"></a>00671 
<a name="l00673"></a>00673   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#2ebd984a2632314a39a6e9ba2c497105" title="Returns &amp;#39;true&amp;#39; if all intervals are compatible with the intervals in r.">IsCompatibleWith</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00674"></a>00674 
<a name="l00675"></a>00675 
<a name="l00676"></a>00676 
<a name="l00678"></a>00678   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#2e0765dfbe651ad583faa8c8a77e047f" title="Returns &amp;#39;true&amp;#39; if intervals are compatible and sorted by chromosome, strand...">IsCompatibleSorted</a>(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00679"></a>00679 
<a name="l00680"></a>00680 
<a name="l00681"></a>00681 
<a name="l00683"></a>00683   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#fb2db74a3cd85a84533375483e12b96c" title="Returns &amp;#39;true&amp;#39; if intervals are compatible, sorted and non-overlapping.">IsCompatibleSortedAndNonoverlapping</a>();
<a name="l00684"></a>00684 
<a name="l00685"></a>00685 
<a name="l00686"></a>00686 
<a name="l00688"></a>00688   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#821c468713f787b3d821af7edaab41dc" title="Returns &amp;#39;true&amp;#39; if it overlaps with input region R.">OverlapsWith</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *R, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00689"></a>00689 
<a name="l00690"></a>00690 
<a name="l00691"></a>00691 
<a name="l00693"></a>00693   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#25634257eb7a2ed379548b17a83414fb" title="Compares with input region r and returns true if its order is before r. The order...">IsBefore</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00694"></a>00694 
<a name="l00695"></a>00695 
<a name="l00696"></a>00696 
<a name="l00698"></a>00698   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#2065cf74a82c2340b1e49e7f7ec0fdd4" title="Compares with input region r and returns true if its position order is before r....">IsPosBefore</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r);
<a name="l00699"></a>00699 
<a name="l00700"></a>00700 
<a name="l00701"></a>00701 
<a name="l00703"></a>00703   <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#da548684fc4ae44dd3782aa15eaaf0a8" title="Returns -1 if before interval i, +1 if after, otherwise returns 0. The order is determined...">CalcDirection</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00704"></a>00704 
<a name="l00705"></a>00705   
<a name="l00706"></a>00706 
<a name="l00708"></a>00708   <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#da548684fc4ae44dd3782aa15eaaf0a8" title="Returns -1 if before interval i, +1 if after, otherwise returns 0. The order is determined...">CalcDirection</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00709"></a>00709 
<a name="l00710"></a>00710   
<a name="l00711"></a>00711 
<a name="l00713"></a>00713   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegion.html#2a99aedb0632fd7e1f2f8dc995ba9191" title="Checks if interval start/stop positions comply with chromosomal bounds.">CheckBounds</a>(StringLIntMap *bounds, <span class="keywordtype">bool</span> ignore); 
<a name="l00714"></a>00714 
<a name="l00715"></a>00715 
<a name="l00716"></a>00716 
<a name="l00717"></a>00717   <span class="comment">//------------------------------------------------------//</span>
<a name="l00718"></a>00718   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l00719"></a>00719   <span class="comment">//------------------------------------------------------//</span>
<a name="l00720"></a>00720 
<a name="l00721"></a>00721 
<a name="l00723"></a>00723   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#90b0382a60f0a8532af73fc18f41364b" title="Prints alignments of input sequences to reference genome.">RunAlign</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00724"></a>00724 
<a name="l00725"></a>00725 
<a name="l00726"></a>00726   
<a name="l00728"></a>00728   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#405010d9256a3c08c53adee98133cb62" title="Prints in BED format. If convert_chromosome is &amp;#39;true&amp;#39;, then convert from...">PrintBEDFormat</a>(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00729"></a>00729 
<a name="l00730"></a>00730 
<a name="l00731"></a>00731 
<a name="l00733"></a>00733   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#71367cf4a117918f43a71a5295839733" title="Corrects interval start/stop positions so as to comply with chromosomal bounds.">ApplyBounds</a>(StringLIntMap *bounds);
<a name="l00734"></a>00734 
<a name="l00735"></a>00735 
<a name="l00736"></a>00736 
<a name="l00738"></a>00738   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#1e9a85a086afbe2608411450c9a4c5e6" title="Replace each interval in the region with the corresponding center interval.">Center</a>();
<a name="l00739"></a>00739 
<a name="l00740"></a>00740 
<a name="l00741"></a>00741 
<a name="l00743"></a>00743   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#c2c426782bd7c0d302b7a77bb1ff047f" title="Replace region with a single interval from minimum start to maximum stop position...">Connect</a>();
<a name="l00744"></a>00744 
<a name="l00745"></a>00745 
<a name="l00746"></a>00746 
<a name="l00748"></a>00748   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#869de97ecf059355fe5a5b572b5c2575" title="Replace regions with the corresponding gap intervals between successive intervals...">Diff</a>();
<a name="l00749"></a>00749 
<a name="l00750"></a>00750 
<a name="l00751"></a>00751 
<a name="l00753"></a>00753   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#98695d5d1056d693a1c635b197f4b8bf" title="Print distance between successive intervals.">RunCalcDistances</a>(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l00754"></a>00754 
<a name="l00755"></a>00755 
<a name="l00756"></a>00756 
<a name="l00758"></a>00758   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#0785438db56d7b7cc7539470e85434dd" title="Divide intervals divided in half (i.e. two intervals for each original interval in...">Divide</a>();
<a name="l00759"></a>00759 
<a name="l00760"></a>00760 
<a name="l00761"></a>00761 
<a name="l00763"></a>00763   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#45fe5d619f2e21644eacb6984aa47eff" title="Print intervals divided in half (i.e. two intervals for each original interval in...">RunDivide</a>();
<a name="l00764"></a>00764 
<a name="l00765"></a>00765 
<a name="l00766"></a>00766 
<a name="l00768"></a>00768   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#048d2182b212789b182c2c82a7ac131e" title="Removes rogue intervals (e.g. start&amp;gt;stop).">Fix</a>();
<a name="l00769"></a>00769 
<a name="l00770"></a>00770 
<a name="l00771"></a>00771 
<a name="l00773"></a>00773   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#8075e77ab6296b59506857cc21ee4c9d" title="Print intersection among all intervals, i.e. maximum start to minimum stop position...">Intersect</a>();
<a name="l00774"></a>00774 
<a name="l00775"></a>00775 
<a name="l00776"></a>00776 
<a name="l00778"></a>00778   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#08db190bc63e6f1d6fc90679dc5e6868" title="Prints region label and sum of intervals&amp;#39; sizes.">RunSize</a>();
<a name="l00779"></a>00779 
<a name="l00780"></a>00780 
<a name="l00781"></a>00781 
<a name="l00783"></a>00783 
<a name="l00787"></a>00787   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#0721b07af0850057e4ab9cd416ecac2f" title="Modifies interval start/stop positions. First it chooses which position to keep fixed...">ModifyPos</a>(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00788"></a>00788 
<a name="l00789"></a>00789 
<a name="l00790"></a>00790 
<a name="l00792"></a>00792   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#222b3b8f567c306ee5a71ea3a720ef43" title="Randomizes interval position within chromosomal bounds.">Randomize</a>(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00793"></a>00793 
<a name="l00794"></a>00794 
<a name="l00795"></a>00795 
<a name="l00797"></a>00797   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#6120af435fb9ee68cb5a3d5066a4fda4" title="Reverses interval start/stop positions with respect to chromosomal bounds, as if...">ReversePos</a>(StringLIntMap *bounds);
<a name="l00798"></a>00798 
<a name="l00799"></a>00799 
<a name="l00800"></a>00800   
<a name="l00802"></a>00802   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#461fbb00db1b45061e641f37614fb146" title="Selects a subset of intervals according to their relative positions.">Select</a>(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l00803"></a>00803 
<a name="l00804"></a>00804 
<a name="l00805"></a>00805 
<a name="l00807"></a>00807 
<a name="l00812"></a>00812   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#dc42fea61283fa8134d65467b2292405" title="Shifts interval start/stop positions. First it chooses reference position, and then...">ShiftPos</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> shift_5prime);
<a name="l00813"></a>00813 
<a name="l00814"></a>00814 
<a name="l00815"></a>00815 
<a name="l00817"></a>00817   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#b2c52ae40b34f3607f4558add94c25fd" title="Prints shuffled region within specified reference regions in refReg.">RunShuffle</a>(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l00818"></a>00818 
<a name="l00819"></a>00819 
<a name="l00820"></a>00820 
<a name="l00822"></a>00822   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#03959dc5b5695d817b5a40b47f30a4aa" title="Sort intervals according to start position (only for compatible intervals).">Sort</a>(); 
<a name="l00823"></a>00823 
<a name="l00824"></a>00824 
<a name="l00825"></a>00825   
<a name="l00827"></a>00827   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#1686d4f4d2962b22d506a25a5887c31c" title="Print region intervals on separate lines.">RunSplit</a>(); 
<a name="l00828"></a>00828 
<a name="l00829"></a>00829 
<a name="l00830"></a>00830   
<a name="l00832"></a>00832   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#e19cd6eaa883e6517ba12567cd756320" title="Modified strand information: &amp;#39;+&amp;#39;=only forward, &amp;#39;-&amp;#39;=only reverse,...">ModifyStrand</a>(<span class="keywordtype">char</span> *strand_op);
<a name="l00833"></a>00833 
<a name="l00834"></a>00834 
<a name="l00835"></a>00835   
<a name="l00837"></a>00837   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#05cb47d38946b780f7eb585074ccf393" title="Compute intervals&amp;#39; union.">Union</a>();
<a name="l00838"></a>00838 
<a name="l00839"></a>00839 
<a name="l00840"></a>00840 
<a name="l00842"></a>00842   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#3d90c15accabfd5c0279f27f3513ba7c" title="Print intervals&amp;#39; union.">RunUnion</a>();
<a name="l00843"></a>00843 
<a name="l00844"></a>00844 
<a name="l00845"></a>00845 
<a name="l00847"></a>00847   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#3962ea71432f3977b1ac501e4e878480" title="Print sliding windows (implemented only for single-interval regions).">PrintWindows</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l00848"></a>00848 
<a name="l00849"></a>00849 
<a name="l00850"></a>00850 
<a name="l00852"></a>00852   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#a2a5b753401e9f9d64a30cfbbdf23f60" title="Remove sub-intervals that correspond to sequences of &amp;#39;N&amp;#39; characters.">PrintRemoveN</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00853"></a>00853 
<a name="l00854"></a>00854 
<a name="l00855"></a>00855 
<a name="l00857"></a>00857   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#4b52d78052e0109b1fceafe450cfdedd" title="Prints region label, intervals and extracted sequence in SEQ format.">PrintSeq</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00858"></a>00858 
<a name="l00859"></a>00859 
<a name="l00860"></a>00860 
<a name="l00862"></a>00862 
<a name="l00866"></a>00866   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#33d1e5544b3fb81e8c19468b91920b1d" title="Calculates the offset distances of interval start and stop position with respect...">PrintOffsetFormat</a>(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l00867"></a>00867 
<a name="l00868"></a>00868 
<a name="l00869"></a>00869 
<a name="l00871"></a>00871   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#7a8ac4f469447256253bdbdcb1b4a956" title="Prints region label and sequence length (excluding &amp;#39;N&amp;#39; characters).">PrintSeqLength</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00872"></a>00872 
<a name="l00873"></a>00873 
<a name="l00874"></a>00874 
<a name="l00876"></a>00876   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#77fa696d55d538b365d15a7a30391132" title="Verifies extracted sequence against region label.">PrintVerifySeq</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l00877"></a>00877 
<a name="l00878"></a>00878 
<a name="l00879"></a>00879 
<a name="l00881"></a>00881   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegion.html#9304039f70c5110b961cc9de77974d09" title="If interval strand is &amp;#39;+&amp;#39;, it is changed to &amp;#39;-&amp;#39; and vice versa.">ReverseStrand</a>();
<a name="l00882"></a>00882 
<a name="l00883"></a>00883 
<a name="l00884"></a>00884 
<a name="l00885"></a>00885   <span class="comment">//------------------------------------------------------//</span>
<a name="l00886"></a>00886   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l00887"></a>00887   <span class="comment">//------------------------------------------------------//</span>
<a name="l00888"></a>00888 
<a name="l00889"></a>00889   
<a name="l00891"></a>00891   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegion.html#50f5ba99404da1cdae6a8632a19ff837" title="Returns a version of this genomic region constrained by the start/stop coordinates...">Constrain</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l00892"></a>00892 
<a name="l00893"></a>00893   
<a name="l00894"></a>00894 
<a name="l00896"></a>00896   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegion.html#869de97ecf059355fe5a5b572b5c2575" title="Replace regions with the corresponding gap intervals between successive intervals...">Diff</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r);
<a name="l00897"></a>00897 
<a name="l00898"></a>00898   
<a name="l00899"></a>00899 
<a name="l00900"></a>00900   <span class="comment">//------------------------------------------------------//</span>
<a name="l00901"></a>00901   <span class="comment">//   Data                                               //</span>
<a name="l00902"></a>00902   <span class="comment">//------------------------------------------------------//</span>
<a name="l00903"></a>00903 
<a name="l00904"></a><a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c">00904</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>;                             
<a name="l00905"></a><a class="code" href="classGenomicRegion.html#7eeba95c1e87e100346688681e30ff24">00905</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegion.html#7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>;                                                  
<a name="l00906"></a><a class="code" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9">00906</a>   GenomicIntervalSet <a class="code" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>;                                 
<a name="l00907"></a>00907   
<a name="l00908"></a>00908   
<a name="l00909"></a>00909 };
<a name="l00910"></a>00910 
<a name="l00911"></a>00911 
<a name="l00912"></a>00912 
<a name="l00913"></a>00913 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00914"></a>00914 <span class="comment">// END CLASS: GenomicRegion                                                                    //</span>
<a name="l00915"></a>00915 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00916"></a>00916 
<a name="l00917"></a>00917 
<a name="l00918"></a>00918 
<a name="l00919"></a>00919 
<a name="l00920"></a>00920 
<a name="l00921"></a>00921 
<a name="l00922"></a>00922 
<a name="l00923"></a>00923 
<a name="l00924"></a>00924 
<a name="l00925"></a>00925 
<a name="l00926"></a>00926 
<a name="l00927"></a>00927 
<a name="l00928"></a>00928 
<a name="l00929"></a>00929 <span class="comment">/*</span>
<a name="l00930"></a>00930 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00931"></a>00931 <span class="comment">// CLASS: GenomicRegionSEQ                                                                     //</span>
<a name="l00932"></a>00932 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00934"></a>00934 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00935"></a>00935 <span class="comment"></span>
<a name="l00936"></a>00936 <span class="comment">class GenomicRegionSEQ : public GenomicRegion</span>
<a name="l00937"></a>00937 <span class="comment">{</span>
<a name="l00938"></a>00938 <span class="comment"></span>
<a name="l00939"></a>00939 <span class="comment"></span>
<a name="l00940"></a>00940 <span class="comment">  //------------------------------------------------------//</span>
<a name="l00941"></a>00941 <span class="comment">  //   Data                                               //</span>
<a name="l00942"></a>00942 <span class="comment">  //------------------------------------------------------//</span>
<a name="l00943"></a>00943 <span class="comment"></span>
<a name="l00944"></a>00944 <span class="comment">  char *SEQ;                                                    //!&lt; sequence</span>
<a name="l00945"></a>00945 <span class="comment">  </span>
<a name="l00946"></a>00946 <span class="comment">  </span>
<a name="l00947"></a>00947 <span class="comment">  // NOTE: utilize all data &amp; code related to the SEQ format in a new class: GenomicRegionSEQ.</span>
<a name="l00949"></a>00949 <span class="comment">  void PrintSearch(char *pattern, bool header, bool summary);</span>
<a name="l00950"></a>00950 <span class="comment"></span>
<a name="l00952"></a>00952 <span class="comment">  void PrintNoGaps();                                                                                   // alignments</span>
<a name="l00953"></a>00953 <span class="comment"></span>
<a name="l00955"></a>00955 <span class="comment">  void PrintModifyLabel();</span>
<a name="l00956"></a>00956 <span class="comment">  </span>
<a name="l00957"></a>00957 <span class="comment">  vector&lt;string&gt; Q;                                             //!&lt; sequences (e.g. from a multiple sequence alignment)</span>
<a name="l00958"></a>00958 <span class="comment">  vector&lt;string&gt; L;                                             //!&lt; sequence labels (e.g. names of species in multiple sequence alignment)</span>
<a name="l00959"></a>00959 <span class="comment">};</span>
<a name="l00960"></a>00960 <span class="comment"></span>
<a name="l00961"></a>00961 <span class="comment"></span>
<a name="l00962"></a>00962 <span class="comment"></span>
<a name="l00963"></a>00963 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00964"></a>00964 <span class="comment">// END CLASS: GenomicRegionSEQ                                                                 //</span>
<a name="l00965"></a>00965 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00966"></a>00966 <span class="comment">*/</span>
<a name="l00967"></a>00967 
<a name="l00968"></a>00968 
<a name="l00969"></a>00969 
<a name="l00970"></a>00970 
<a name="l00971"></a>00971 
<a name="l00972"></a>00972 
<a name="l00973"></a>00973 
<a name="l00974"></a>00974 
<a name="l00975"></a>00975 
<a name="l00976"></a>00976 
<a name="l00977"></a>00977 
<a name="l00978"></a>00978 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00979"></a>00979 <span class="comment">// CLASS: GenomicRegionBEDToREG                                                                //</span>
<a name="l00980"></a>00980 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00982"></a>00982 <span class="comment"></span>
<a name="l00989"></a>00989 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00990"></a><a class="code" href="classGenomicRegionBEDToREG.html">00990</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all...">GenomicRegionBEDToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>
<a name="l00991"></a>00991 {
<a name="l00992"></a>00992  <span class="keyword">public</span>:
<a name="l00994"></a>00994   <a class="code" href="classGenomicRegionBEDToREG.html#964b37e08c0c513d974e14cadaba5d20" title="Class constructor, see example below.">GenomicRegionBEDToREG</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *B);
<a name="l00995"></a>00995 
<a name="l00996"></a>00996 
<a name="l00997"></a>00997 
<a name="l00999"></a>00999   <a class="code" href="classGenomicRegionBEDToREG.html#964b37e08c0c513d974e14cadaba5d20" title="Class constructor, see example below.">GenomicRegionBEDToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01000"></a>01000 
<a name="l01001"></a>01001 
<a name="l01002"></a>01002 
<a name="l01004"></a>01004   <a class="code" href="classGenomicRegionBEDToREG.html#59650c4d898e5835271627bf8c45c609" title="Class destructor.">~GenomicRegionBEDToREG</a>();
<a name="l01005"></a>01005   
<a name="l01006"></a>01006 
<a name="l01007"></a>01007 
<a name="l01009"></a>01009   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBEDToREG.html#3f18f7f3946f2258b30e9796fafee093" title="Reads BED format from input string inp.">Read</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01010"></a>01010   
<a name="l01011"></a>01011 
<a name="l01012"></a>01012 
<a name="l01013"></a>01013   <span class="comment">// data</span>
<a name="l01014"></a>01014 
<a name="l01015"></a>01015 
<a name="l01016"></a>01016 };
<a name="l01017"></a>01017 
<a name="l01018"></a>01018 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01019"></a>01019 <span class="comment">// END CLASS: GenomicRegionBEDToREG                                                            //</span>
<a name="l01020"></a>01020 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01021"></a>01021 
<a name="l01022"></a>01022 
<a name="l01023"></a>01023 
<a name="l01024"></a>01024 
<a name="l01025"></a>01025 
<a name="l01026"></a>01026 
<a name="l01027"></a>01027 
<a name="l01028"></a>01028 
<a name="l01029"></a>01029 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01030"></a>01030 <span class="comment">// CLASS: GenomicRegionBED                                                                     //</span>
<a name="l01031"></a>01031 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01033"></a>01033 <span class="comment"></span>
<a name="l01040"></a>01040 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01041"></a><a class="code" href="classGenomicRegionBED.html">01041</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>
<a name="l01042"></a>01042 {
<a name="l01043"></a>01043  <span class="keyword">public</span>:
<a name="l01045"></a>01045   <a class="code" href="classGenomicRegionBED.html#89d54edc5809d113202c9f09f82e1057" title="Class constructor, see example below.">GenomicRegionBED</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *B);
<a name="l01046"></a>01046 
<a name="l01047"></a>01047 
<a name="l01048"></a>01048 
<a name="l01050"></a>01050   <a class="code" href="classGenomicRegionBED.html#89d54edc5809d113202c9f09f82e1057" title="Class constructor, see example below.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01051"></a>01051 
<a name="l01052"></a>01052 
<a name="l01053"></a>01053 
<a name="l01055"></a>01055   <a class="code" href="classGenomicRegionBED.html#89d54edc5809d113202c9f09f82e1057" title="Class constructor, see example below.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i);
<a name="l01056"></a>01056 
<a name="l01057"></a>01057 
<a name="l01058"></a>01058 
<a name="l01060"></a>01060   <a class="code" href="classGenomicRegionBED.html#a27241b7fbdb90e0b4b1e411045d080f" title="Class destructor.">~GenomicRegionBED</a>();
<a name="l01061"></a>01061   
<a name="l01062"></a>01062 
<a name="l01063"></a>01063 
<a name="l01064"></a>01064   <span class="comment">//------------------------------------------------------//</span>
<a name="l01065"></a>01065   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01066"></a>01066   <span class="comment">//------------------------------------------------------//</span>
<a name="l01067"></a>01067 
<a name="l01068"></a>01068 
<a name="l01070"></a>01070   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#707a1744e040214d78d37b928f1f0252" title="Reads BED format from input string inp.">Read</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01071"></a>01071   
<a name="l01072"></a>01072 
<a name="l01073"></a>01073 
<a name="l01075"></a>01075   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#ffbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print</a>(FILE *file_ptr=stdout);
<a name="l01076"></a>01076 
<a name="l01077"></a>01077 
<a name="l01078"></a>01078 
<a name="l01080"></a>01080   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#ffbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print</a>(FILE *file_ptr, <span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01081"></a>01081 
<a name="l01082"></a>01082 
<a name="l01083"></a>01083 
<a name="l01085"></a>01085   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#04eceeb473bee498747bdd68d8ba8806" title="Prints all pairwise intersection with the intervals in region r (compatible, sorted...">PrintIntersection</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l01086"></a>01086 
<a name="l01087"></a>01087 
<a name="l01088"></a>01088 
<a name="l01090"></a>01090   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#25a8584259b6a8d47aa8c46795de1880" title="Same as Print(), but the interval is constrained between start and stop.">PrintConstrained</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01091"></a>01091 
<a name="l01092"></a>01092 
<a name="l01093"></a>01093 
<a name="l01095"></a>01095   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#aebba3c31c0fa88e594706f490030549" title="Same as Print(), but the label and interval&amp;#39;s coordinates are modified.">PrintModified</a>(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01096"></a>01096 
<a name="l01097"></a>01097 
<a name="l01098"></a>01098 
<a name="l01099"></a>01099   <span class="comment">//------------------------------------------------------//</span>
<a name="l01100"></a>01100   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01101"></a>01101   <span class="comment">//------------------------------------------------------//</span>
<a name="l01102"></a>01102 
<a name="l01103"></a>01103 
<a name="l01105"></a>01105   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#01bd886cfa67a3adae10c67dee941924" title="Auxiliary function: updates thickStart/thickEnd variables to honor changes in coordinates...">UpdateThick</a>();
<a name="l01106"></a>01106   
<a name="l01107"></a>01107   
<a name="l01108"></a>01108   
<a name="l01110"></a>01110   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#513fb071e2626be58837052daf86bab5" title="Prints in BED format. If convert_chromosome is &amp;#39;true&amp;#39;, then convert from...">PrintBEDFormat</a>(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01111"></a>01111 
<a name="l01112"></a>01112 
<a name="l01113"></a>01113 
<a name="l01115"></a>01115   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#5056b98f57933e2d24fd6d601f5499ef" title="Corrects interval start/stop positions so as to comply with chromosomal bounds.">ApplyBounds</a>(StringLIntMap *bounds);
<a name="l01116"></a>01116 
<a name="l01117"></a>01117 
<a name="l01118"></a>01118 
<a name="l01120"></a>01120   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#35c692c1343327d9cb450cfa1276096f" title="Replace each interval in the region with the corresponding center interval.">Center</a>();
<a name="l01121"></a>01121 
<a name="l01122"></a>01122 
<a name="l01123"></a>01123 
<a name="l01125"></a>01125   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#d55a9ffea8b0924bac3d11a5722cdfd2" title="Replace region with a single interval from minimum start to maximum stop position...">Connect</a>();
<a name="l01126"></a>01126 
<a name="l01127"></a>01127 
<a name="l01128"></a>01128 
<a name="l01130"></a>01130   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#9580179a120834e71ac90e56e5f35aa9" title="Replace regions with the corresponding gap intervals between successive intervals...">Diff</a>();
<a name="l01131"></a>01131 
<a name="l01132"></a>01132 
<a name="l01133"></a>01133 
<a name="l01135"></a>01135   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#833d50369e61f64603c63b6a05f3d80b" title="Print distance between successive intervals.">RunCalcDistances</a>(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01136"></a>01136 
<a name="l01137"></a>01137 
<a name="l01138"></a>01138 
<a name="l01140"></a>01140   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#be056ec84d2d6890479d47559839bab4" title="Divide intervals divided in half (i.e. two intervals for each original interval in...">Divide</a>();
<a name="l01141"></a>01141 
<a name="l01142"></a>01142 
<a name="l01143"></a>01143 
<a name="l01145"></a>01145   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#5f456b4af154c98968370568828d6e3d" title="Print intervals divided in half (i.e. two intervals for each original interval in...">RunDivide</a>();
<a name="l01146"></a>01146 
<a name="l01147"></a>01147 
<a name="l01148"></a>01148 
<a name="l01150"></a>01150   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#ac069dcf513265f7756b72c106f2765c" title="Removes rogue intervals (e.g. start&amp;gt;stop).">Fix</a>();
<a name="l01151"></a>01151 
<a name="l01152"></a>01152 
<a name="l01153"></a>01153 
<a name="l01155"></a>01155   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#cf815a65712e82bf3f8c596c795f55c4" title="Print intersection among all intervals, i.e. maximum start to minimum stop position...">Intersect</a>();
<a name="l01156"></a>01156 
<a name="l01157"></a>01157 
<a name="l01158"></a>01158 
<a name="l01160"></a>01160 
<a name="l01164"></a>01164   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#c515c70f443db400f911452ed359433b" title="Modifies interval start/stop positions. First it chooses which position to keep fixed...">ModifyPos</a>(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01165"></a>01165 
<a name="l01166"></a>01166 
<a name="l01167"></a>01167 
<a name="l01169"></a>01169   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#920d47deb095506c311da2b2324e7a14" title="Randomizes interval position within chromosomal bounds.">Randomize</a>(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01170"></a>01170 
<a name="l01171"></a>01171 
<a name="l01172"></a>01172 
<a name="l01174"></a>01174   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#dec7408315765bbdb21465a26400422b" title="Selects a subset of intervals according to their relative positions.">Select</a>(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01175"></a>01175 
<a name="l01176"></a>01176 
<a name="l01177"></a>01177 
<a name="l01179"></a>01179 
<a name="l01184"></a>01184   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#97d00cb29eb1f379b9e6ee17197530c7" title="Shifts interval start/stop positions. First it chooses reference position, and then...">ShiftPos</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> shift_5prime);
<a name="l01185"></a>01185 
<a name="l01186"></a>01186 
<a name="l01187"></a>01187 
<a name="l01189"></a>01189   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#f3793387aa335bf68042f9245bc5992d" title="Prints shuffled region within specified reference regions in refReg.">RunShuffle</a>(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01190"></a>01190 
<a name="l01191"></a>01191 
<a name="l01192"></a>01192 
<a name="l01194"></a>01194   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#fd1a61c04f76782abdc83074ae0f7bd3" title="Sort intervals according to start position (only for compatible intervals).">Sort</a>(); 
<a name="l01195"></a>01195 
<a name="l01196"></a>01196 
<a name="l01197"></a>01197   
<a name="l01199"></a>01199   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#bd3b8e4969c2ff2639eb344de7b224f9" title="Print region intervals on separate lines.">RunSplit</a>(); 
<a name="l01200"></a>01200 
<a name="l01201"></a>01201 
<a name="l01202"></a>01202   
<a name="l01204"></a>01204   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#605f5bf86254547d117ed0c3f5c64587" title="Compute intervals&amp;#39; union.">Union</a>();
<a name="l01205"></a>01205 
<a name="l01206"></a>01206 
<a name="l01207"></a>01207 
<a name="l01209"></a>01209   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionBED.html#3ebdc24dd97c7b106205d932970de535" title="Print sliding windows (implemented only for single-interval regions).">PrintWindows</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01210"></a>01210 
<a name="l01211"></a>01211 
<a name="l01212"></a>01212 
<a name="l01213"></a>01213   <span class="comment">//------------------------------------------------------//</span>
<a name="l01214"></a>01214   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l01215"></a>01215   <span class="comment">//------------------------------------------------------//</span>
<a name="l01216"></a>01216 
<a name="l01217"></a>01217   
<a name="l01219"></a>01219   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> *<a class="code" href="classGenomicRegionBED.html#0c4b2a8a590ce1dfbef142d03d53fe5d" title="Returns a version of this genomic region constrained by the start/stop coordinates...">Constrain</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l01220"></a>01220 
<a name="l01221"></a>01221   
<a name="l01222"></a>01222 
<a name="l01224"></a>01224   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionBED.html#9580179a120834e71ac90e56e5f35aa9" title="Replace regions with the corresponding gap intervals between successive intervals...">Diff</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r);
<a name="l01225"></a>01225 
<a name="l01226"></a>01226   
<a name="l01227"></a>01227 
<a name="l01228"></a>01228   <span class="comment">//------------------------------------------------------//</span>
<a name="l01229"></a>01229   <span class="comment">//   Data                                               //</span>
<a name="l01230"></a>01230   <span class="comment">//------------------------------------------------------//</span>
<a name="l01231"></a>01231 
<a name="l01232"></a><a class="code" href="classGenomicRegionBED.html#60878391f6fb1e0ad0ac9bc88a7625d3">01232</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#60878391f6fb1e0ad0ac9bc88a7625d3" title="number of tokens">n_tokens</a>;                    
<a name="l01233"></a><a class="code" href="classGenomicRegionBED.html#fa7ba810a0d1673f5dbabcf85c7bbb95">01233</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#fa7ba810a0d1673f5dbabcf85c7bbb95" title="score field">score</a>;                               
<a name="l01234"></a><a class="code" href="classGenomicRegionBED.html#52fc75b347dfe6696e7ad2beeeb623a9">01234</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionBED.html#52fc75b347dfe6696e7ad2beeeb623a9" title="color field">itemRgb</a>;                                
<a name="l01235"></a><a class="code" href="classGenomicRegionBED.html#019dfaebedc17e8859b36c3359287faa">01235</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#019dfaebedc17e8859b36c3359287faa" title="thickStart field">thickStart</a>;                  
<a name="l01236"></a><a class="code" href="classGenomicRegionBED.html#5333a7fdc1f21db508aa7af4f50553ba">01236</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#5333a7fdc1f21db508aa7af4f50553ba" title="thickEnd field">thickEnd</a>;                    
<a name="l01237"></a>01237 
<a name="l01238"></a>01238 };
<a name="l01239"></a>01239 
<a name="l01240"></a>01240 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01241"></a>01241 <span class="comment">// END CLASS: GenomicRegionBED                                                                 //</span>
<a name="l01242"></a>01242 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01243"></a>01243 
<a name="l01244"></a>01244 
<a name="l01245"></a>01245 
<a name="l01246"></a>01246 
<a name="l01247"></a>01247 
<a name="l01248"></a>01248 
<a name="l01249"></a>01249 
<a name="l01250"></a>01250 
<a name="l01251"></a>01251 
<a name="l01252"></a>01252 
<a name="l01253"></a>01253 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01254"></a>01254 <span class="comment">// CLASS: GenomicRegionSAMToREG                                                                //</span>
<a name="l01255"></a>01255 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01257"></a>01257 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01258"></a><a class="code" href="classGenomicRegionSAMToREG.html">01258</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all...">GenomicRegionSAMToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>
<a name="l01259"></a>01259 {
<a name="l01260"></a>01260  <span class="keyword">public</span>:
<a name="l01261"></a>01261   
<a name="l01263"></a>01263   <a class="code" href="classGenomicRegionSAMToREG.html#67555a42a7bba86d7e0389a09d1e3407" title="Class constructor, see example below.">GenomicRegionSAMToREG</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *B);
<a name="l01264"></a>01264 
<a name="l01265"></a>01265 
<a name="l01266"></a>01266 
<a name="l01268"></a>01268   <a class="code" href="classGenomicRegionSAMToREG.html#67555a42a7bba86d7e0389a09d1e3407" title="Class constructor, see example below.">GenomicRegionSAMToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01269"></a>01269 
<a name="l01270"></a>01270 
<a name="l01271"></a>01271 
<a name="l01273"></a>01273   <a class="code" href="classGenomicRegionSAMToREG.html#19dea1653efdc013f4425a43470eab02" title="Class destructor.">~GenomicRegionSAMToREG</a>();
<a name="l01274"></a>01274   
<a name="l01275"></a>01275 
<a name="l01276"></a>01276 
<a name="l01278"></a>01278   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAMToREG.html#b4a3b874cafac1537a7d7287cae39e76" title="Reads SAM format from input string inp.">Read</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01279"></a>01279   
<a name="l01280"></a>01280 
<a name="l01281"></a>01281 
<a name="l01283"></a>01283   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSAMToREG.html#cc2044e03dc7e75bdcada1eb1e77a5ea" title="Extracts the strand from the FLAG field.">CalcStrandFromFlag</a>(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01284"></a>01284 
<a name="l01285"></a>01285 
<a name="l01286"></a>01286   
<a name="l01288"></a>01288   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAMToREG.html#dbc035818bcc64cc715b16c34b314799" title="Returns next token in CIGAR string.">GetNextTokenOfCIGAR</a>(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01289"></a>01289 
<a name="l01290"></a>01290   
<a name="l01291"></a>01291   
<a name="l01292"></a>01292   <span class="comment">// data</span>
<a name="l01293"></a>01293 
<a name="l01294"></a>01294 
<a name="l01295"></a>01295 };
<a name="l01296"></a>01296 
<a name="l01297"></a>01297 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01298"></a>01298 <span class="comment">// END CLASS: GenomicRegionSAMToREG                                                            //</span>
<a name="l01299"></a>01299 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01300"></a>01300 
<a name="l01301"></a>01301 
<a name="l01302"></a>01302 
<a name="l01303"></a>01303 
<a name="l01304"></a>01304 
<a name="l01305"></a>01305 
<a name="l01306"></a>01306 
<a name="l01307"></a>01307 
<a name="l01308"></a>01308 
<a name="l01309"></a>01309 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01310"></a>01310 <span class="comment">// CLASS: GenomicRegionSAM                                                                     //</span>
<a name="l01311"></a>01311 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01313"></a>01313 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01314"></a><a class="code" href="classGenomicRegionSAM.html">01314</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>
<a name="l01315"></a>01315 {
<a name="l01316"></a>01316  <span class="keyword">public</span>:
<a name="l01317"></a><a class="code" href="classGenomicRegionSAM.html#adf8580c5362304ab4866fa76a59df96">01317</a>   <span class="keyword">typedef</span> vector&lt;pair&lt;long int,char&gt; &gt; <a class="code" href="classGenomicRegionSAM.html#adf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length).">CIGARTokens</a>;                     
<a name="l01318"></a>01318   
<a name="l01320"></a>01320   <a class="code" href="classGenomicRegionSAM.html#e6b0828926a573743e7292567346c1a1" title="Class constructor.">GenomicRegionSAM</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *B);
<a name="l01321"></a>01321 
<a name="l01322"></a>01322 
<a name="l01323"></a>01323 
<a name="l01325"></a>01325   <a class="code" href="classGenomicRegionSAM.html#e6b0828926a573743e7292567346c1a1" title="Class constructor.">GenomicRegionSAM</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01326"></a>01326 
<a name="l01327"></a>01327 
<a name="l01328"></a>01328 
<a name="l01330"></a>01330   <a class="code" href="classGenomicRegionSAM.html#1d11cc22134d13dbf67ffe4c7220888d" title="Class destructor.">~GenomicRegionSAM</a>();
<a name="l01331"></a>01331   
<a name="l01332"></a>01332 
<a name="l01333"></a>01333 
<a name="l01334"></a>01334   <span class="comment">//------------------------------------------------------//</span>
<a name="l01335"></a>01335   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01336"></a>01336   <span class="comment">//------------------------------------------------------//</span>
<a name="l01337"></a>01337 
<a name="l01338"></a>01338 
<a name="l01340"></a>01340   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#096d90b6743400daf1acb50bcf17086a" title="Reads SAM format from input string inp.">Read</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01341"></a>01341 
<a name="l01342"></a>01342 
<a name="l01343"></a>01343 
<a name="l01345"></a>01345   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#4381c6486f767b6e4055cb729beee181" title="Prints label and genomic region intervals with a newline character at the end.">Print</a>(FILE *file_ptr=stdout);
<a name="l01346"></a>01346 
<a name="l01347"></a>01347 
<a name="l01348"></a>01348 
<a name="l01350"></a>01350   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#df8beab78f541121e7b56957afffa4ac" title="Same as Print(), but the interval is constrained between start and stop.">PrintConstrained</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01351"></a>01351 
<a name="l01352"></a>01352 
<a name="l01353"></a>01353 
<a name="l01355"></a>01355   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#ecf1d74c572e85bbe9bca0b30c6a6dda" title="Same as Print(), but the label and interval&amp;#39;s coordinates are modified.">PrintModified</a>(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01356"></a>01356 
<a name="l01357"></a>01357 
<a name="l01358"></a>01358   
<a name="l01359"></a>01359   <span class="comment">//------------------------------------------------------//</span>
<a name="l01360"></a>01360   <span class="comment">//   Auxiliary functions                                //</span>
<a name="l01361"></a>01361   <span class="comment">//------------------------------------------------------//</span>
<a name="l01362"></a>01362 
<a name="l01364"></a>01364   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSAM.html#6d0949a836520073de57e0f968a6b819" title="Extracts the strand from the FLAG field.">CalcStrandFromFlag</a>(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01365"></a>01365 
<a name="l01366"></a>01366 
<a name="l01367"></a>01367   
<a name="l01369"></a>01369   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#f583a963f7fb0eb071c1b640c660813a" title="Updates the FLAG field given a new strand.">UpdateFlagFromStrand</a>(<span class="keywordtype">char</span> strand);
<a name="l01370"></a>01370   
<a name="l01371"></a>01371   
<a name="l01372"></a>01372   
<a name="l01374"></a>01374   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#ea563a4a484784ad231399a3b56c7056" title="Prints gapped sequence.">PrintGappedSequence</a>(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01375"></a>01375 
<a name="l01376"></a>01376   
<a name="l01377"></a>01377   
<a name="l01379"></a>01379   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#9446540150b384ca0d3dbba630c9c6a6" title="Returns gapped sequence.">GetGappedSequence</a>(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01380"></a>01380 
<a name="l01381"></a>01381   
<a name="l01382"></a>01382   
<a name="l01384"></a>01384   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#621b3dd6f4d03a5223bd26a57af14f7e" title="Sets CIGAR, SEQ and QUAL fields to &amp;quot;*&amp;quot; and removes OPTIONAL field; also...">InvalidateSeqData</a>();
<a name="l01385"></a>01385 
<a name="l01386"></a>01386 
<a name="l01387"></a>01387   
<a name="l01389"></a>01389   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#34b0387b9656e48d0c4046d62fe085e1" title="Computes and stores CIGAR tokens in GenomicRegionSAM::T.">TokenizeCIGAR</a>();
<a name="l01390"></a>01390 
<a name="l01391"></a>01391   
<a name="l01392"></a>01392   
<a name="l01394"></a>01394   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#50198fead1c40caadf76f40b61a3c14a" title="Returns next token in CIGAR string.">GetNextTokenOfCIGAR</a>(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01395"></a>01395 
<a name="l01396"></a>01396   
<a name="l01397"></a>01397   
<a name="l01399"></a>01399   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#965bb3cc1217e6c859a07dd23090c2a0" title="Update CIGAR string based only on interval information (chooses simplest possible...">UpdateCigarFromIntervals</a>();
<a name="l01400"></a>01400   
<a name="l01401"></a>01401   
<a name="l01402"></a>01402  
<a name="l01404"></a>01404   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#5de5d05dfccd33149153a299c1cb2f6b" title="Calculate interval length from CIGAR string.">CalcFragmentLengthFromCIGAR</a>();
<a name="l01405"></a>01405   
<a name="l01406"></a>01406   
<a name="l01407"></a>01407   
<a name="l01409"></a>01409   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#acff4cd74a3b8c450f9f90c6b23412c4" title="Calculate interval length from CIGAR string.">CalcReferenceLengthFromCIGAR</a>();
<a name="l01410"></a>01410   
<a name="l01411"></a>01411   
<a name="l01412"></a>01412   
<a name="l01414"></a>01414   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#17e8e7054ffe7205d7c34926b60526d8" title="Trim CIGAR string from left and right by a given distance (measured on the reference...">TrimCIGAR</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01415"></a>01415  
<a name="l01416"></a>01416  
<a name="l01417"></a>01417  
<a name="l01419"></a>01419   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#0ff4472118291a8034efc929c784680c" title="Trim input seq from left and right (used for trimming both SEQ and QUAL fields).">TrimSequence</a>(<span class="keywordtype">char</span> *seq, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01420"></a>01420  
<a name="l01421"></a>01421  
<a name="l01422"></a>01422  
<a name="l01423"></a>01423   <span class="comment">//------------------------------------------------------//</span>
<a name="l01424"></a>01424   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01425"></a>01425   <span class="comment">//------------------------------------------------------//</span>
<a name="l01426"></a>01426 
<a name="l01427"></a>01427 
<a name="l01429"></a>01429   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#97b9fec94e5f21490d839b3f99235a96" title="Prints alignments of input sequences to reference genome.">RunAlign</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01430"></a>01430 
<a name="l01431"></a>01431 
<a name="l01432"></a>01432   
<a name="l01434"></a>01434   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#5d82c1d84432274392416f87bfac655a" title="Corrects interval start/stop positions so as to comply with chromosomal bounds.">ApplyBounds</a>(StringLIntMap *bounds);
<a name="l01435"></a>01435 
<a name="l01436"></a>01436 
<a name="l01437"></a>01437 
<a name="l01439"></a>01439   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#52fa8c0420b16fe92f54427365f72899" title="Replace each interval in the region with the corresponding center interval.">Center</a>();
<a name="l01440"></a>01440 
<a name="l01441"></a>01441 
<a name="l01442"></a>01442 
<a name="l01444"></a>01444   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#d765c84814151be0e710fefb37c7949a" title="Replace region with a single interval from minimum start to maximum stop position...">Connect</a>();
<a name="l01445"></a>01445 
<a name="l01446"></a>01446 
<a name="l01447"></a>01447 
<a name="l01449"></a>01449   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#ca38b04fa916ae4c9c28c36f537136ba" title="Replace regions with the corresponding gap intervals between successive intervals...">Diff</a>();
<a name="l01450"></a>01450 
<a name="l01451"></a>01451 
<a name="l01452"></a>01452 
<a name="l01454"></a>01454   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#fc9c7ec17be517dd5b1ad7cb62f58826" title="Print distance between successive intervals.">RunCalcDistances</a>(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01455"></a>01455 
<a name="l01456"></a>01456 
<a name="l01457"></a>01457 
<a name="l01459"></a>01459   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#7c16ba87be562950021d3f19eb14ac5e" title="Divide intervals divided in half (i.e. two intervals for each original interval in...">Divide</a>();
<a name="l01460"></a>01460 
<a name="l01461"></a>01461 
<a name="l01462"></a>01462 
<a name="l01464"></a>01464   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#dd2ada73e241f593dcdcdfca97df1eb5" title="Print intervals divided in half (i.e. two intervals for each original interval in...">RunDivide</a>();
<a name="l01465"></a>01465 
<a name="l01466"></a>01466 
<a name="l01467"></a>01467 
<a name="l01469"></a>01469   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#d955a6a512678613b4a42d2db0d3f549" title="Removes rogue intervals (e.g. start&amp;gt;stop).">Fix</a>();
<a name="l01470"></a>01470 
<a name="l01471"></a>01471 
<a name="l01472"></a>01472 
<a name="l01474"></a>01474   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#4f26b01325ebd4e20c9bf3247d5fa733" title="Print intersection among all intervals, i.e. maximum start to minimum stop position...">Intersect</a>();
<a name="l01475"></a>01475 
<a name="l01476"></a>01476 
<a name="l01477"></a>01477 
<a name="l01479"></a>01479   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#538b8fc9969f24da6911ec39c8d70a3b" title="Randomizes interval position within chromosomal bounds.">Randomize</a>(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01480"></a>01480 
<a name="l01481"></a>01481 
<a name="l01482"></a>01482 
<a name="l01484"></a>01484 
<a name="l01488"></a>01488   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#353207352073db00dee0a9b620dca197" title="Modifies interval start/stop positions. First it chooses which position to keep fixed...">ModifyPos</a>(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01489"></a>01489 
<a name="l01490"></a>01490 
<a name="l01491"></a>01491 
<a name="l01493"></a>01493   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#4dce925fdb55dcc7ef3dad890bc25408" title="Selects a subset of intervals according to their relative positions.">Select</a>(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01494"></a>01494 
<a name="l01495"></a>01495 
<a name="l01496"></a>01496 
<a name="l01498"></a>01498 
<a name="l01503"></a>01503   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#c6c559782775bfc65907801148226590" title="Shifts interval start/stop positions. First it chooses reference position, and then...">ShiftPos</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> shift_5prime);
<a name="l01504"></a>01504 
<a name="l01505"></a>01505 
<a name="l01506"></a>01506 
<a name="l01508"></a>01508   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#b31fed1760d543fa3a38bb16a064dcf9" title="Prints shuffled region within specified reference regions in refReg.">RunShuffle</a>(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01509"></a>01509 
<a name="l01510"></a>01510 
<a name="l01511"></a>01511 
<a name="l01513"></a>01513   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#8a27afdcdf11895d7b3f32bd8eed68ae" title="Sort intervals according to start position (only for compatible intervals).">Sort</a>(); 
<a name="l01514"></a>01514 
<a name="l01515"></a>01515 
<a name="l01516"></a>01516   
<a name="l01518"></a>01518   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#e195a95ba837bfd9673093f1d451b39c" title="Print region intervals on separate lines.">RunSplit</a>(); 
<a name="l01519"></a>01519 
<a name="l01520"></a>01520 
<a name="l01521"></a>01521   
<a name="l01523"></a>01523   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#beb76df49bb89ecf0f3074805d9fe7b2" title="Modified strand information: &amp;#39;+&amp;#39;=only forward, &amp;#39;-&amp;#39;=only reverse,...">ModifyStrand</a>(<span class="keywordtype">char</span> *strand_op);
<a name="l01524"></a>01524 
<a name="l01525"></a>01525 
<a name="l01526"></a>01526   
<a name="l01528"></a>01528   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#6161ba10d2a4871cc43aa5bc597a787b" title="Compute intervals&amp;#39; union.">Union</a>();
<a name="l01529"></a>01529 
<a name="l01530"></a>01530 
<a name="l01531"></a>01531 
<a name="l01533"></a>01533   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSAM.html#27027bc80a3b5a864ebafd415a6ef5a6" title="Print sliding windows (implemented only for single-interval regions).">PrintWindows</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01534"></a>01534 
<a name="l01535"></a>01535 
<a name="l01536"></a>01536 
<a name="l01537"></a>01537   <span class="comment">//------------------------------------------------------//</span>
<a name="l01538"></a>01538   <span class="comment">//   Data                                               //</span>
<a name="l01539"></a>01539   <span class="comment">//------------------------------------------------------//</span>
<a name="l01540"></a>01540 
<a name="l01541"></a><a class="code" href="classGenomicRegionSAM.html#281e35ef1bd0b7da811d9b8957f3d08f">01541</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#281e35ef1bd0b7da811d9b8957f3d08f" title="number of TAB-separated fields in SAM entry (&amp;gt;= 12)">n_tokens</a>;                                    
<a name="l01542"></a>01542   <span class="comment">//char *QNAME;                                                //!&lt; Query template NAME: mapped to GenomicRegion::LABEL</span>
<a name="l01543"></a><a class="code" href="classGenomicRegionSAM.html#200d7f7044dd69311eceb851a3c315e5">01543</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#200d7f7044dd69311eceb851a3c315e5" title="bitwise FLAG">FLAG</a>;                               
<a name="l01544"></a>01544   <span class="comment">//char *RNAME;                                                //!&lt; Reference sequence NAME: mapped to GenomicRegion::I.front()-&gt;CHROMOSOME</span>
<a name="l01545"></a>01545   <span class="comment">//long int POS;                                               //!&lt; 1-based leftmost mapping POSition: mapped to GenomicRegion::I.front()-&gt;START</span>
<a name="l01546"></a><a class="code" href="classGenomicRegionSAM.html#edacdd8319f4b614229a41b80d4c6f67">01546</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#edacdd8319f4b614229a41b80d4c6f67" title="MAPping Quality.">MAPQ</a>;                                                
<a name="l01547"></a><a class="code" href="classGenomicRegionSAM.html#a5acff481ae565db4976e66933212726">01547</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#a5acff481ae565db4976e66933212726" title="CIGAR string.">CIGAR</a>;                                                  
<a name="l01548"></a><a class="code" href="classGenomicRegionSAM.html#f723505bfafff1a1fa616479023d0c3c">01548</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#f723505bfafff1a1fa616479023d0c3c" title="Reference name of the mate/next fragment.">RNEXT</a>;                                                  
<a name="l01549"></a><a class="code" href="classGenomicRegionSAM.html#d6f88b4324ba508b86102641d819661f">01549</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#d6f88b4324ba508b86102641d819661f" title="Position of the mate/next fragment.">PNEXT</a>;                                               
<a name="l01550"></a><a class="code" href="classGenomicRegionSAM.html#f180854a186be37eab49520edaa1262f">01550</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#f180854a186be37eab49520edaa1262f" title="observed Template LENgth">TLEN</a>;                                                
<a name="l01551"></a><a class="code" href="classGenomicRegionSAM.html#6c2887ed1fa3a25d801641ea1b139032">01551</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#6c2887ed1fa3a25d801641ea1b139032" title="fragment SEQuence">SEQ</a>;                                                    
<a name="l01552"></a><a class="code" href="classGenomicRegionSAM.html#b43ca5f2fbfdee520f1424c3a9fd0706">01552</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#b43ca5f2fbfdee520f1424c3a9fd0706" title="ASCII of Phred-scaled base QUALity+33.">QUAL</a>;                                                   
<a name="l01553"></a><a class="code" href="classGenomicRegionSAM.html#1622a8804462bad83b1c49d25eeacb49">01553</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#1622a8804462bad83b1c49d25eeacb49" title="all SAM fields after the 12th mandatory field">OPTIONAL</a>;                                               
<a name="l01554"></a><a class="code" href="classGenomicRegionSAM.html#47afdde68b89cf4ea199bc91d4954a92">01554</a>   <a class="code" href="classGenomicRegionSAM.html#adf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length).">CIGARTokens</a> <a class="code" href="classGenomicRegionSAM.html#47afdde68b89cf4ea199bc91d4954a92" title="CIGAR tokens.">T</a>;                                                
<a name="l01555"></a>01555   
<a name="l01556"></a>01556 };
<a name="l01557"></a>01557 
<a name="l01558"></a>01558 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01559"></a>01559 <span class="comment">// END CLASS: GenomicRegionSAM                                                                 //</span>
<a name="l01560"></a>01560 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01561"></a>01561 
<a name="l01562"></a>01562 
<a name="l01563"></a>01563 
<a name="l01564"></a>01564 
<a name="l01565"></a>01565 
<a name="l01566"></a>01566 
<a name="l01567"></a>01567 
<a name="l01568"></a>01568 
<a name="l01569"></a>01569 
<a name="l01570"></a>01570 
<a name="l01571"></a>01571 
<a name="l01572"></a>01572 
<a name="l01573"></a>01573 
<a name="l01574"></a>01574 
<a name="l01575"></a>01575 
<a name="l01576"></a>01576 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01577"></a>01577 <span class="comment">// CLASS: GenomicRegionGFFToREG                                                                //</span>
<a name="l01578"></a>01578 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01580"></a>01580 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01581"></a><a class="code" href="classGenomicRegionGFFToREG.html">01581</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all...">GenomicRegionGFFToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>
<a name="l01582"></a>01582 {
<a name="l01583"></a>01583  <span class="keyword">public</span>:
<a name="l01585"></a>01585   <a class="code" href="classGenomicRegionGFFToREG.html#8ad38850fd8e7e19aabca0bf2ba62d55" title="Class constructor, see example below.">GenomicRegionGFFToREG</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *B);
<a name="l01586"></a>01586 
<a name="l01587"></a>01587 
<a name="l01588"></a>01588 
<a name="l01590"></a>01590   <a class="code" href="classGenomicRegionGFFToREG.html#8ad38850fd8e7e19aabca0bf2ba62d55" title="Class constructor, see example below.">GenomicRegionGFFToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01591"></a>01591 
<a name="l01592"></a>01592 
<a name="l01593"></a>01593 
<a name="l01595"></a>01595   <a class="code" href="classGenomicRegionGFFToREG.html#5a41c3de64cd8ad69952b77805816c5e" title="Class destructor.">~GenomicRegionGFFToREG</a>();
<a name="l01596"></a>01596   
<a name="l01597"></a>01597 
<a name="l01598"></a>01598 
<a name="l01600"></a>01600   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFFToREG.html#bc519dd06e8d03711b90679ec279be3b" title="Reads GFF format from input string inp.">Read</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01601"></a>01601   
<a name="l01602"></a>01602 
<a name="l01603"></a>01603 
<a name="l01604"></a>01604   <span class="comment">// data</span>
<a name="l01605"></a>01605 
<a name="l01606"></a>01606 
<a name="l01607"></a>01607 };
<a name="l01608"></a>01608 
<a name="l01609"></a>01609 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01610"></a>01610 <span class="comment">// END CLASS: GenomicRegionGFFToREG                                                            //</span>
<a name="l01611"></a>01611 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01612"></a>01612 
<a name="l01613"></a>01613 
<a name="l01614"></a>01614 
<a name="l01615"></a>01615 
<a name="l01616"></a>01616 
<a name="l01617"></a>01617 
<a name="l01618"></a>01618 
<a name="l01619"></a>01619 
<a name="l01620"></a>01620 
<a name="l01621"></a>01621 
<a name="l01622"></a>01622 
<a name="l01623"></a>01623 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01624"></a>01624 <span class="comment">// CLASS: GenomicRegionGFF                                                                     //</span>
<a name="l01625"></a>01625 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01627"></a>01627 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01628"></a><a class="code" href="classGenomicRegionGFF.html">01628</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a>
<a name="l01629"></a>01629 {
<a name="l01630"></a>01630  <span class="keyword">public</span>:
<a name="l01632"></a>01632   <a class="code" href="classGenomicRegionGFF.html#9a49f0dc39116964f774e7ce78cfc827" title="Class constructor.">GenomicRegionGFF</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *B);
<a name="l01633"></a>01633 
<a name="l01634"></a>01634 
<a name="l01635"></a>01635 
<a name="l01637"></a>01637   <a class="code" href="classGenomicRegionGFF.html#9a49f0dc39116964f774e7ce78cfc827" title="Class constructor.">GenomicRegionGFF</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01638"></a>01638 
<a name="l01639"></a>01639 
<a name="l01640"></a>01640 
<a name="l01642"></a>01642   <a class="code" href="classGenomicRegionGFF.html#3885d42a347283a5fed52791988076ad" title="Class destructor.">~GenomicRegionGFF</a>();
<a name="l01643"></a>01643   
<a name="l01644"></a>01644 
<a name="l01645"></a>01645 
<a name="l01646"></a>01646   <span class="comment">//------------------------------------------------------//</span>
<a name="l01647"></a>01647   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01648"></a>01648   <span class="comment">//------------------------------------------------------//</span>
<a name="l01649"></a>01649 
<a name="l01650"></a>01650 
<a name="l01652"></a>01652   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#eb1390d42b388e32d98f58f57a5c169c" title="Reads GFF format from input string inp.">Read</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#efe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>=0);
<a name="l01653"></a>01653 
<a name="l01654"></a>01654 
<a name="l01655"></a>01655 
<a name="l01657"></a>01657   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#3ffa71d729e4ff060a15a6bab1e924c8" title="Prints label and genomic region intervals with a newline character at the end.">Print</a>(FILE *file_ptr=stdout);
<a name="l01658"></a>01658 
<a name="l01659"></a>01659 
<a name="l01660"></a>01660 
<a name="l01662"></a>01662   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#96539f4372048c9f0d87826f5856a8ef" title="Same as Print(), but the interval is constrained between start and stop.">PrintConstrained</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01663"></a>01663 
<a name="l01664"></a>01664 
<a name="l01665"></a>01665 
<a name="l01667"></a>01667   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#4b331cd12506aa4793a2cf6ea7620936" title="Same as Print(), but the label and interval&amp;#39;s coordinates are modified.">PrintModified</a>(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01668"></a>01668 
<a name="l01669"></a>01669 
<a name="l01670"></a>01670 
<a name="l01672"></a>01672   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#ee70fcc81c9a6629bf0aa1ec37cb1d64" title="Replace regions with the corresponding gap intervals between successive intervals...">Diff</a>();
<a name="l01673"></a>01673 
<a name="l01674"></a>01674 
<a name="l01675"></a>01675 
<a name="l01677"></a>01677   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#0ff4768578a03011fabad309072fade4" title="Print distance between successive intervals.">RunCalcDistances</a>(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01678"></a>01678 
<a name="l01679"></a>01679 
<a name="l01680"></a>01680 
<a name="l01682"></a>01682   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#77975c03dd39b404098eb519e81f67d4" title="Divide intervals divided in half (i.e. two intervals for each original interval in...">Divide</a>();
<a name="l01683"></a>01683 
<a name="l01684"></a>01684 
<a name="l01685"></a>01685 
<a name="l01687"></a>01687   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#d4acc1231677efd876084ef36db82de1" title="Print intervals divided in half (i.e. two intervals for each original interval in...">RunDivide</a>();
<a name="l01688"></a>01688 
<a name="l01689"></a>01689 
<a name="l01690"></a>01690 
<a name="l01692"></a>01692   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#d35d9b6e9c57270c0d5648ae9f702959" title="Print intersection among all intervals, i.e. maximum start to minimum stop position...">Intersect</a>();
<a name="l01693"></a>01693 
<a name="l01694"></a>01694 
<a name="l01695"></a>01695 
<a name="l01697"></a>01697   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#aa9b1271561e2d62ef1be6464e589cf4" title="Selects a subset of intervals according to their relative positions.">Select</a>(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01698"></a>01698 
<a name="l01699"></a>01699 
<a name="l01700"></a>01700 
<a name="l01702"></a>01702   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#bc08552b0ad4ca512f79064b8cd64e54" title="Print region intervals on separate lines.">RunSplit</a>(); 
<a name="l01703"></a>01703 
<a name="l01704"></a>01704 
<a name="l01705"></a>01705   
<a name="l01707"></a>01707   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionGFF.html#3815f1b27aef231a674c527e11f25880" title="Print sliding windows (implemented only for single-interval regions).">PrintWindows</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01708"></a>01708 
<a name="l01709"></a>01709 
<a name="l01710"></a>01710 
<a name="l01711"></a>01711   <span class="comment">//---------------------------------------//</span>
<a name="l01712"></a>01712   <span class="comment">//  Data                                 //</span>
<a name="l01713"></a>01713   <span class="comment">//---------------------------------------//</span>
<a name="l01714"></a>01714   
<a name="l01715"></a><a class="code" href="classGenomicRegionGFF.html#aec963f555e7e807fb5046566622f5ee">01715</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionGFF.html#aec963f555e7e807fb5046566622f5ee" title="number of TAB-separated fields in GFF entry (= 11)">n_tokens</a>;                    
<a name="l01716"></a>01716   <span class="comment">//char *SEQNAME;                              //!&lt; sequence name: mapped to GenomicRegion::I-&gt;front()-&gt;CHROMOSOME</span>
<a name="l01717"></a><a class="code" href="classGenomicRegionGFF.html#2f0d4a5f15d8a756233436af03341734">01717</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#2f0d4a5f15d8a756233436af03341734" title="feature source">SOURCE</a>;                                 
<a name="l01718"></a><a class="code" href="classGenomicRegionGFF.html#0de447fbfad3da5b7a9cf07a5a6980f0">01718</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#0de447fbfad3da5b7a9cf07a5a6980f0" title="feture type">FEATURE</a>;                                
<a name="l01719"></a>01719   <span class="comment">//long int START;                             //!&lt; 1-based start position: mapped to GenomicRegion::I-&gt;front()-&gt;START</span>
<a name="l01720"></a>01720   <span class="comment">//long int END;                               //!&lt; 1-based end position: mapped to GenomicRegion::I-&gt;front()-&gt;STOP</span>
<a name="l01721"></a><a class="code" href="classGenomicRegionGFF.html#0e92da1f0fd153504e0630f27229ba4d">01721</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#0e92da1f0fd153504e0630f27229ba4d" title="score">SCORE</a>;                                  
<a name="l01722"></a>01722   <span class="comment">//char STRAND;                                //!&lt; strand: mapped to GenomicRegion::I-&gt;front()-&gt;STRAND</span>
<a name="l01723"></a><a class="code" href="classGenomicRegionGFF.html#42122a69eec385e0106e3861cb3b98f3">01723</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionGFF.html#42122a69eec385e0106e3861cb3b98f3" title="coding frame">FRAME</a>;                                   
<a name="l01724"></a>01724   <span class="comment">//char *ATTRIBUTE;                    //!&lt; attribute: mapped to GenomicRegion::LABEL if non-empty</span>
<a name="l01725"></a><a class="code" href="classGenomicRegionGFF.html#625a023f1bb24fc597e723919a036073">01725</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#625a023f1bb24fc597e723919a036073" title="comment">COMMENT</a>;                                
<a name="l01726"></a>01726 
<a name="l01727"></a>01727 };
<a name="l01728"></a>01728 
<a name="l01729"></a>01729 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01730"></a>01730 <span class="comment">// END CLASS: GenomicRegionGFF                                                                 //</span>
<a name="l01731"></a>01731 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01732"></a>01732 
<a name="l01733"></a>01733 
<a name="l01734"></a>01734 
<a name="l01735"></a>01735 
<a name="l01736"></a>01736 
<a name="l01737"></a>01737 
<a name="l01738"></a>01738 
<a name="l01739"></a>01739 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01740"></a>01740 <span class="comment">// CLASS: GenomicRegionSet                                                                     //</span>
<a name="l01741"></a>01741 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01743"></a>01743 <span class="comment"></span>
<a name="l01756"></a>01756 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01757"></a><a class="code" href="classGenomicRegionSet.html">01757</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a>
<a name="l01758"></a>01758 {
<a name="l01759"></a>01759  <span class="keyword">public</span>:
<a name="l01761"></a>01761   <a class="code" href="classGenomicRegionSet.html#afc417b2a9a27f3ad833828be5720e92" title="Class constructor, see example below.">GenomicRegionSet</a>(<span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSet.html#7c593d1beab775590d5b810097644583" title="file name (&amp;#39;NULL&amp;#39; if reading from standard input)">file</a>, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#4d60b5403d2d6aaeda541ed45625b245" title="line buffer size (automatically adjusted during execution to ensure the entire line...">buffer_size</a>, <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#2646e9e6d5da1693b4bea36a5acc082e" title="verbose mode">verbose</a>, <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687" title="&amp;#39;true&amp;#39; if the entire region set is loaded in memory">load_in_memory</a>=<span class="keyword">true</span>); 
<a name="l01762"></a>01762 
<a name="l01763"></a>01763 
<a name="l01765"></a>01765   <a class="code" href="classGenomicRegionSet.html#afc417b2a9a27f3ad833828be5720e92" title="Class constructor, see example below.">GenomicRegionSet</a>(FILE *<a class="code" href="classGenomicRegionSet.html#4ca6c6268e39fdea2add6e6381c66962" title="file pointer (&amp;#39;NULL&amp;#39; if reading from standard input)">file_ptr</a>, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size, <span class="keywordtype">bool</span> verbose, <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687" title="&amp;#39;true&amp;#39; if the entire region set is loaded in memory">load_in_memory</a>=<span class="keyword">true</span>);
<a name="l01766"></a>01766 
<a name="l01767"></a>01767 
<a name="l01769"></a>01769   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegionSet.html#3fc7c6dc31cffcedf97a7b69a7f3c2c8" title="Class destructor.">~GenomicRegionSet</a>();
<a name="l01770"></a>01770 
<a name="l01771"></a>01771 
<a name="l01772"></a>01772   <span class="comment">//-------------------------------------//</span>
<a name="l01773"></a>01773   <span class="comment">//  Read &amp; Print                       //</span>
<a name="l01774"></a>01774   <span class="comment">//-------------------------------------//</span>
<a name="l01775"></a>01775 
<a name="l01777"></a>01777   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#6992ccc0c4dd8be0ad4be9b991ab9f36" title="Processes file header.">ProcessFileHeader</a>(<span class="keywordtype">bool</span> hide);
<a name="l01778"></a>01778 
<a name="l01779"></a>01779   
<a name="l01781"></a>01781   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#5721a1d9ca4ca18ddbcad1e563d42019" title="Detects input format.">DetectFileFormat</a>();
<a name="l01782"></a>01782 
<a name="l01783"></a>01783   
<a name="l01785"></a>01785   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#4e8e101aebcedd6a5892d463fe830450" title="Creates new genomic region based on input format.">CreateGenomicRegion</a>(<a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *<a class="code" href="classGenomicRegionSet.html#df3a5598e31ddbb5b381125925362085" title="pointer to the buffer class used to read the region sets">buffer</a>);
<a name="l01786"></a>01786 
<a name="l01787"></a>01787 
<a name="l01789"></a>01789   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#94b1955e7a4127efaceba25429af7568" title="Initializes data structures.">Init</a>();
<a name="l01790"></a>01790 
<a name="l01791"></a>01791 
<a name="l01793"></a>01793   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#ae7295913424148013df6a154c5458aa" title="Resets file; note, that this is not possible if the regions are being read from the...">Reset</a>();
<a name="l01794"></a>01794 
<a name="l01795"></a>01795 
<a name="l01797"></a>01797   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#06e2e8f91623c01e6e00500888b9ba3d" title="Returns pointer to the first region in the set; note, that this is not possible if...">Begin</a>(); 
<a name="l01798"></a>01798 
<a name="l01799"></a>01799 
<a name="l01801"></a>01801   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#77031066d2648c8935c291ef227b37be" title="Returns pointer to current region in the set.">Get</a>();
<a name="l01802"></a>01802 
<a name="l01803"></a>01803 
<a name="l01805"></a>01805   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#4f7501a27d13cae65247686182b78d51" title="Moves pointer to next region in the set; if GenomicRegionSet::load_in_memory is &amp;#39;false&amp;#39;...">Next</a>();
<a name="l01806"></a>01806 
<a name="l01807"></a>01807 
<a name="l01809"></a>01809 
<a name="l01813"></a>01813   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#4f7501a27d13cae65247686182b78d51" title="Moves pointer to next region in the set; if GenomicRegionSet::load_in_memory is &amp;#39;false&amp;#39;...">Next</a>(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">bool</span> retain_current=<span class="keyword">false</span>);
<a name="l01814"></a>01814 
<a name="l01815"></a>01815 
<a name="l01817"></a>01817   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#8993c741ef2513ba5c5b88eb390f1f99" title="Deletes genomic region from memory.">Release</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r);
<a name="l01818"></a>01818 
<a name="l01819"></a>01819 
<a name="l01821"></a>01821   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#d7dfea15eb8968d376771d74e042f3a1" title="[DEPRECATED] Returns current region, and protects its from deletion by GenomicRegionSet::Next()...">GetRetain</a>();
<a name="l01822"></a>01822 
<a name="l01823"></a>01823 
<a name="l01825"></a>01825   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#39c746333affb32acffbf0f31b7f6fcf" title="Prints error message.">PrintError</a>(<span class="keywordtype">string</span> error_msg);
<a name="l01826"></a>01826   
<a name="l01827"></a>01827 
<a name="l01828"></a>01828 
<a name="l01829"></a>01829   <span class="comment">//-------------------------------------//</span>
<a name="l01830"></a>01830   <span class="comment">//  File-based (vertical) operations   //</span>
<a name="l01831"></a>01831   <span class="comment">//-------------------------------------//</span>
<a name="l01832"></a>01832 
<a name="l01833"></a>01833 
<a name="l01835"></a>01835   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#260acb32612d31325099c9e53366c9d2" title="Counts the number of regions in the set. If use_label_counts, the region labels are...">CountRegions</a>(<span class="keywordtype">bool</span> use_label_counts);
<a name="l01836"></a>01836 
<a name="l01837"></a>01837 
<a name="l01839"></a>01839   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#78c6f4879ac131e6db2094f21b466a98" title="Annotate input region neighborhoods as upstream, downstream, etc.">RunGlobalAnnotate</a>(StringLIntMap *bounds);
<a name="l01840"></a>01840 
<a name="l01841"></a>01841 
<a name="l01843"></a>01843   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#619b51265f7bf5d6c92144de689ca8a2" title="Cluster input regions based on pair-wise overlaps (note that this is computationally...">RunGlobalCluster</a>(<span class="keywordtype">bool</span> merge=<span class="keyword">false</span>);
<a name="l01844"></a>01844 
<a name="l01845"></a>01845 
<a name="l01847"></a>01847   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#3d550572975732566378797ac1d8119b" title="Print distance between successive regions.">RunGlobalCalcDistances</a>(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01848"></a>01848 
<a name="l01849"></a>01849 
<a name="l01851"></a>01851   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#2d6dc99608a1d4938b421fa5a85f5477" title="Prints sorted region set.">RunGlobalSort</a>();
<a name="l01852"></a>01852 
<a name="l01853"></a>01853 
<a name="l01855"></a>01855   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#1056659ccda2e39febb6b87d858a88c4" title="Prints the difference between chromosome intervals and input region set (input region...">RunGlobalInvert</a>(StringLIntMap *bounds);
<a name="l01856"></a>01856 
<a name="l01857"></a>01857 
<a name="l01859"></a>01859   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#b9783fb4fead289c7fe107f9bec1865b" title="Links successive regions if they overlap (input region set must be sorted).">RunGlobalLink</a>(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> max_difference);
<a name="l01860"></a>01860 
<a name="l01861"></a>01861 
<a name="l01863"></a>01863   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#40dd79548f9d0333f1169bd17a6e9465" title="Partitions overlapping regions into a non-overlapping region set (input region set...">RunGlobalPartition</a>();
<a name="l01864"></a>01864 
<a name="l01865"></a>01865 
<a name="l01867"></a>01867   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#83853c6af4b779a37de284751d09348d" title="Tests if input region set is sorted.">RunGlobalTest</a>(<span class="keywordtype">bool</span> sorted_by_strand);
<a name="l01868"></a>01868 
<a name="l01869"></a>01869 
<a name="l01871"></a>01871   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#53885e10c5dc9aa51ebece77dae63313" title="Merges successive regions if they are separated by a distance of merge_win or less...">RunGlobalMerge</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> merge_win);
<a name="l01872"></a>01872 
<a name="l01873"></a>01873 
<a name="l01875"></a>01875   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#962bb1b26543df14b324d7ab4dd8e404" title="Reverses the order of reverse-strand regions (for this operation, the entire region...">RunGlobalReverseOrder</a>();
<a name="l01876"></a>01876 
<a name="l01877"></a>01877 
<a name="l01879"></a>01879   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#b8abdbc3c1f4530dc1992fb0766ba5a4" title="Computes densities of input regions in index regions in Index. This is OBSOLETE,...">PrintDensities</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> &amp;Index, <span class="keywordtype">bool</span> report_exact_boundaries);
<a name="l01880"></a>01880 
<a name="l01881"></a>01881 
<a name="l01883"></a>01883   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#14610521cc47631b1459c2e3488b0eec" title="Scans input regions in sliding windows and reports count distribution (input regions...">RunGlobalScan</a>(StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01884"></a>01884 
<a name="l01885"></a>01885 
<a name="l01887"></a>01887 
<a name="l01897"></a>01897   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#45cac26bedec4d983127b93d2096f735" title="Scans input regions in sliding windows (input regions must be sorted). See also class...">RunGlobalScanCount</a>(StringLIntMap *bounds, <span class="keywordtype">char</span> *ref_reg_file, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> ignore_reverse_strand, <span class="keywordtype">char</span> preprocess, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">long</span> <span class="keywordtype">int</span> min_reads);
<a name="l01898"></a>01898 
<a name="l01899"></a>01899 
<a name="l01901"></a>01901   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#6377c10b1e9903ff5ae3b0ff9fe20e4d" title="Prints in Wiggle format. If convert_chromosome is &amp;#39;true&amp;#39;, then convert from...">RunConvertToWIG</a>(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">char</span> *options, <span class="keywordtype">long</span> <span class="keywordtype">int</span> span, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01902"></a>01902 
<a name="l01903"></a>01903 
<a name="l01904"></a>01904 
<a name="l01905"></a>01905 
<a name="l01906"></a>01906   <span class="comment">//---------------------------------------//</span>
<a name="l01907"></a>01907   <span class="comment">//  Line-based (horizontal) operations   //</span>
<a name="l01908"></a>01908   <span class="comment">//---------------------------------------//</span>
<a name="l01909"></a>01909 
<a name="l01911"></a>01911   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#da3da7ef918c09867341cda32881d63e" title="Prints alignments of input sequences to reference genome.">RunAlign</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01912"></a>01912 
<a name="l01913"></a>01913 
<a name="l01915"></a>01915   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#18669fb4dfdf1287aaf9cfceb8329cb5" title="Converts to BED format. If convert_chromosome is &amp;#39;true&amp;#39;, then convert from...">RunConvertToBED</a>(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01916"></a>01916 
<a name="l01917"></a>01917 
<a name="l01919"></a>01919   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#d7ad039dd702837f37c2cdfebd9b2845" title="Corrects interval start/stop positions so as to comply with chromosomal bounds.">RunBounds</a>(StringLIntMap *bounds);
<a name="l01920"></a>01920 
<a name="l01921"></a>01921 
<a name="l01923"></a>01923   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#1aa8f41f0978e9b66c6f11c0ccb95a0b" title="Print intervals&amp;#39; center.">RunCenter</a>();
<a name="l01924"></a>01924 
<a name="l01925"></a>01925 
<a name="l01927"></a>01927   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#752b7d4d2ac3568e7bdb13edbfc00af3" title="Print connected intervals as a single interval from minimum start to maximum stop...">RunConnect</a>();
<a name="l01928"></a>01928 
<a name="l01929"></a>01929 
<a name="l01931"></a>01931   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#90256738bd85501460874ab8e69b894b" title="Print gap intervals between successive intervals (e.g. for computing intron boundaries)...">RunDiff</a>();
<a name="l01932"></a>01932 
<a name="l01933"></a>01933 
<a name="l01935"></a>01935   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#22f3fce9421a0f80c96ec0a9346fec19" title="Print distances between successive intervals of each region (i.e. local operation)...">RunCalcDistances</a>(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01936"></a>01936 
<a name="l01937"></a>01937 
<a name="l01939"></a>01939   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#9acb3424739ed62c0efddad01536ad18" title="Print intervals divided in half (i.e. two intervals for each original interval in...">RunDivide</a>();
<a name="l01940"></a>01940 
<a name="l01941"></a>01941 
<a name="l01943"></a>01943   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#efa09dc62c768a06e4ca1f56fd683be5" title="Removes rogue intervals (e.g. start&amp;gt;stop).">RunFix</a>();
<a name="l01944"></a>01944 
<a name="l01945"></a>01945 
<a name="l01946"></a>01946 
<a name="l01948"></a>01948   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#e598cd9c5f94555a04e43f9fe6334aec" title="Print intersection among all intervals, i.e. maximum start to minimum stop position...">RunIntersection</a>();
<a name="l01949"></a>01949 
<a name="l01950"></a>01950 
<a name="l01951"></a>01951 
<a name="l01953"></a>01953   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#b1e96bd7781dac63a1745a8cf5ac5538" title="Prints region label and sum of intervals&amp;#39; sizes.">RunSize</a>();
<a name="l01954"></a>01954 
<a name="l01955"></a>01955 
<a name="l01956"></a>01956 
<a name="l01958"></a>01958 
<a name="l01962"></a>01962   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#11f1e58c92e9b0b2e3feb5821cff1a55" title="Modifies interval start/stop positions. First it chooses which position to keep fixed...">RunModifyPos</a>(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01963"></a>01963 
<a name="l01964"></a>01964 
<a name="l01965"></a>01965 
<a name="l01967"></a>01967   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#8f71dd5fa48a0878751cd78a3b485daf" title="Prints label and genomic region intervals with a newline character at the end for...">RunConvertToREG</a>(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l01968"></a>01968 
<a name="l01969"></a>01969 
<a name="l01970"></a>01970 
<a name="l01972"></a>01972   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#1b2b9edad211ca961f319921d38bfbf6" title="Randomizes interval position within chromosomal bounds.">RunRandomize</a>(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01973"></a>01973 
<a name="l01974"></a>01974 
<a name="l01975"></a>01975 
<a name="l01977"></a>01977   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#7b522944fe4f7f489c99fb76ce7f576c" title="Selects a subset of intervals according to their relative positions.">RunSelect</a>(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01978"></a>01978 
<a name="l01979"></a>01979 
<a name="l01980"></a>01980 
<a name="l01982"></a>01982 
<a name="l01987"></a>01987   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#d4b42becffb21de5e89129476c79c851" title="Shifts interval start/stop positions. First it chooses reference position, and then...">RunShiftPos</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> shift_5prime);
<a name="l01988"></a>01988 
<a name="l01989"></a>01989 
<a name="l01991"></a>01991   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#53143cb3271c65fb289faece622cd8bd" title="Print sorted intervals according to start position (only for compatible intervals)...">RunSort</a>(); 
<a name="l01992"></a>01992 
<a name="l01993"></a>01993   
<a name="l01995"></a>01995   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#570e56362af1b30e7f5cb149ea42e54e" title="Split regions into intervals which are printed on separate lines.">RunSplit</a>(); 
<a name="l01996"></a>01996 
<a name="l01997"></a>01997   
<a name="l01999"></a>01999   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#f1146729843b82d7637e0adc1b07e51a" title="Print region with modified strands: &amp;#39;+&amp;#39;=only forward, &amp;#39;-&amp;#39;=only reverse...">RunModifyStrand</a>(<span class="keywordtype">char</span> *strand_op, <span class="keywordtype">bool</span> sorted);
<a name="l02000"></a>02000 
<a name="l02001"></a>02001 
<a name="l02003"></a>02003   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#7c0a328896b88a0c837f60e89fd85584" title="Same as GenomicRegionSet::RunModifyStrand but returns a sorted region set (note:...">RunModifyStrandSorted</a>(<span class="keywordtype">char</span> *strand_op);
<a name="l02004"></a>02004 
<a name="l02005"></a>02005 
<a name="l02007"></a>02007   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#015049fd0bd08587cbdbb49332ee352d" title="Shuffles input regions within specified reference regions loaded from file ref_reg_file...">RunShuffle</a>(gsl_rng *random_generator, <span class="keywordtype">char</span> *ref_reg_file);
<a name="l02008"></a>02008 
<a name="l02009"></a>02009 
<a name="l02011"></a>02011   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#b281c8ab57c4181a06161182031c9df4" title="Print intervals&amp;#39; union.">RunUnion</a>();
<a name="l02012"></a>02012 
<a name="l02013"></a>02013 
<a name="l02015"></a>02015   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#a326d5adcc97f47ed2de35353b60b393" title="Print sliding windows (implemented only for single-interval regions).">RunSlidingWindows</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l02016"></a>02016 
<a name="l02017"></a>02017 
<a name="l02019"></a>02019   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#b219b834c232c091982b066a2163f76a" title="Prints region label, intervals and extracted sequence in SEQ format.">RunExtractSeq</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l02020"></a>02020 
<a name="l02021"></a>02021 
<a name="l02022"></a>02022 
<a name="l02023"></a>02023   <span class="comment">//---------------------------------------//</span>
<a name="l02024"></a>02024   <span class="comment">//  UNDER DEVELOPMENT/TESTING            //</span>
<a name="l02025"></a>02025   <span class="comment">//---------------------------------------//</span>
<a name="l02026"></a>02026 
<a name="l02027"></a>02027 
<a name="l02029"></a>02029   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#de00553eb76b71f68e540739bdc776f4" title="Prints in BEDGraph format. If convert_chromosome is &amp;#39;true&amp;#39;, then convert...">PrintBEDGraphFormat</a>(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l02030"></a>02030 
<a name="l02031"></a>02031 
<a name="l02033"></a>02033   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#3da972129b9b69577bbe0bb9837bbea3" title="Prints region label and sequence length excluding &amp;#39;N&amp;#39; characters.">PrintSeqLength</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02034"></a>02034 
<a name="l02035"></a>02035 
<a name="l02037"></a>02037   <span class="comment">//void PrintNoGaps();</span>
<a name="l02038"></a>02038 
<a name="l02039"></a>02039 
<a name="l02041"></a>02041   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#6dc7150acdd7a0614f8a8497da687d0e" title="Print sequence with no alignment gaps (requires SEQ input).">PrintRemoveN</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02042"></a>02042 
<a name="l02043"></a>02043 
<a name="l02045"></a>02045 
<a name="l02049"></a>02049   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#e27a5c9f17f08afae277603e33d4fb91" title="Calculates the offset distances of interval start and stop position with respect...">PrintOffsetFormat</a>(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l02050"></a>02050 
<a name="l02051"></a>02051 
<a name="l02053"></a>02053   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#c8de83151ba0cc30ca1b72e83ecea82f" title="Reverses interval start/stop positions with respect to chromosomal bounds, as if...">PrintReversePos</a>(StringLIntMap *bounds);
<a name="l02054"></a>02054 
<a name="l02055"></a>02055 
<a name="l02057"></a>02057   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#dfbad7f0acf42c2179bd33103e5d183c" title="Search sequence for short pattern (requires SEQ input).">PrintSearch</a>(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l02058"></a>02058 
<a name="l02059"></a>02059 
<a name="l02061"></a>02061   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSet.html#4c197a09ffcace2ec2736aa089eec448" title="Verifies extracted sequence against region label.">PrintVerifySeq</a>(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l02062"></a>02062 
<a name="l02063"></a>02063 
<a name="l02064"></a>02064 
<a name="l02065"></a>02065  <span class="keyword">private</span>:
<a name="l02067"></a>02067   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSet.html#cd823956ec9f510ee3d52b08eec1b2f6" title="Creates a temporary file from this instance of GenomicRegionSet which contains the...">StoreInTempFile</a>();
<a name="l02068"></a>02068 
<a name="l02069"></a>02069 
<a name="l02071"></a>02071   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#b118bb8a6b63ae47a37ef89bff25ab72" title="Uses mergesort between tempSet and this instance of GenomicRegionSet (used in GenomicRegionSet::RunM...">PrintMergeSort</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *rtempSet, <span class="keywordtype">char</span> strand);
<a name="l02072"></a>02072 
<a name="l02073"></a>02073 
<a name="l02075"></a>02075   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSet.html#e601878a7754a7e5bc8b4ed340d012f4" title="Reverses the order between tempSet and this instance of GenomicRegionSet (used in...">PrintReverse</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *rtempSet);
<a name="l02076"></a>02076 
<a name="l02077"></a>02077 
<a name="l02078"></a>02078 
<a name="l02079"></a>02079   <span class="comment">//---------------------------------------//</span>
<a name="l02080"></a>02080   <span class="comment">//  Data                                 //</span>
<a name="l02081"></a>02081   <span class="comment">//---------------------------------------//</span>
<a name="l02082"></a>02082 
<a name="l02083"></a>02083  <span class="keyword">public</span>:
<a name="l02084"></a><a class="code" href="classGenomicRegionSet.html#7c593d1beab775590d5b810097644583">02084</a>   <span class="keywordtype">char</span> *file;                                           
<a name="l02085"></a><a class="code" href="classGenomicRegionSet.html#4ca6c6268e39fdea2add6e6381c66962">02085</a>   FILE *file_ptr;                                       
<a name="l02086"></a><a class="code" href="classGenomicRegionSet.html#4d60b5403d2d6aaeda541ed45625b245">02086</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size;        
<a name="l02087"></a><a class="code" href="classGenomicRegionSet.html#2646e9e6d5da1693b4bea36a5acc082e">02087</a>   <span class="keywordtype">bool</span> verbose;                                         
<a name="l02088"></a><a class="code" href="classGenomicRegionSet.html#bd0fc2542e717e5855e03d3b3892f027">02088</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#bd0fc2542e717e5855e03d3b3892f027" title="&amp;#39;true&amp;#39; if reading from standard input">from_stdin</a>;                                      
<a name="l02089"></a><a class="code" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687">02089</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687" title="&amp;#39;true&amp;#39; if the entire region set is loaded in memory">load_in_memory</a>;                          
<a name="l02090"></a><a class="code" href="classGenomicRegionSet.html#ea8b5c9bae036da50d4e8720d2e13231">02090</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#ea8b5c9bae036da50d4e8720d2e13231" title="if &amp;#39;true&amp;#39; use an array for storing genomic regions&amp;#39; intervals (for time...">use_interval_array</a>;                      
<a name="l02091"></a><a class="code" href="classGenomicRegionSet.html#a4494ed2ae3f7594c29b18c39fed6748">02091</a>   <span class="keywordtype">string</span> <a class="code" href="classGenomicRegionSet.html#a4494ed2ae3f7594c29b18c39fed6748" title="input format">format</a>;                                        
<a name="l02092"></a><a class="code" href="classGenomicRegionSet.html#08d9a35a7688ac2c193b53c9a8693c6e">02092</a>   <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> <a class="code" href="classGenomicRegionSet.html#08d9a35a7688ac2c193b53c9a8693c6e" title="keeps track of progress of computation">progress</a>;                            
<a name="l02093"></a><a class="code" href="classGenomicRegionSet.html#df3a5598e31ddbb5b381125925362085">02093</a>   <a class="code" href="classFileBuffer.html" title="Class for reading lines from file or standard input.">FileBuffer</a> *buffer;                           
<a name="l02094"></a><a class="code" href="classGenomicRegionSet.html#5293292428aa06145b38be4b96ab05e4">02094</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#5293292428aa06145b38be4b96ab05e4" title="number of regions (&amp;#39;1&amp;#39; if the regions are not loaded in memory)">n_regions</a>;                           
<a name="l02095"></a><a class="code" href="classGenomicRegionSet.html#206d91b443832800afc04f786d15971d">02095</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> **<a class="code" href="classGenomicRegionSet.html#206d91b443832800afc04f786d15971d" title="array where the regions are stored; if the region set is not entirely loaded in memory...">R</a>;                            
<a name="l02096"></a><a class="code" href="classGenomicRegionSet.html#d89185b88d5b37161cc65bc84b421c0e">02096</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#d89185b88d5b37161cc65bc84b421c0e" title="index pointing to the current region in R">r_index</a>;                                     
<a name="l02097"></a>02097 };
<a name="l02098"></a>02098 
<a name="l02099"></a>02099 
<a name="l02100"></a>02100 
<a name="l02101"></a>02101 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02102"></a>02102 <span class="comment">// END CLASS: GenomicRegionSet                                                                 //</span>
<a name="l02103"></a>02103 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02104"></a>02104 
<a name="l02105"></a>02105 
<a name="l02106"></a>02106 
<a name="l02107"></a>02107 
<a name="l02108"></a>02108 
<a name="l02109"></a>02109 
<a name="l02110"></a>02110 
<a name="l02111"></a>02111 
<a name="l02112"></a>02112 
<a name="l02113"></a>02113 
<a name="l02114"></a>02114 
<a name="l02115"></a>02115 
<a name="l02116"></a>02116 
<a name="l02117"></a>02117 
<a name="l02118"></a>02118 
<a name="l02119"></a>02119 
<a name="l02120"></a>02120 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02121"></a>02121 <span class="comment">// CLASS: IndexedGenomicRegionSet                                                              //</span>
<a name="l02122"></a>02122 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02124"></a>02124 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02125"></a><a class="code" href="classIndexedGenomicRegionSet.html">02125</a> <span class="keyword">class </span><a class="code" href="classIndexedGenomicRegionSet.html" title="[UNDER DEVELOPMENT] An extension of GenomicRegionSet has an index for quick searching...">IndexedGenomicRegionSet</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a>
<a name="l02126"></a>02126 {
<a name="l02127"></a>02127  <span class="keyword">public</span>:
<a name="l02128"></a><a class="code" href="classIndexedGenomicRegionSet.html#870fa974397cb4487b5e2fe3fa6f8ec5">02128</a>   <span class="keyword">typedef</span> list&lt;GenomicInterval*&gt; <a class="code" href="classIndexedGenomicRegionSet.html#870fa974397cb4487b5e2fe3fa6f8ec5" title="list of genomic intervals">GenomicIntervalList</a>;                   
<a name="l02129"></a><a class="code" href="classIndexedGenomicRegionSet.html#c1e7aa6c1a55659c5756a70d0289c997">02129</a>   <span class="keyword">typedef</span> map&lt;string,GenomicIntervalList**&gt; <a class="code" href="classIndexedGenomicRegionSet.html#c1e7aa6c1a55659c5756a70d0289c997" title="map of string to genomic interval list">RegIndex</a>;                   
<a name="l02130"></a>02130 
<a name="l02132"></a>02132   <a class="code" href="classIndexedGenomicRegionSet.html#f4e3d07d6b9c1dfbc795c636edbcb939" title="Class constructor, see example below.">IndexedGenomicRegionSet</a>(<span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSet.html#7c593d1beab775590d5b810097644583" title="file name (&amp;#39;NULL&amp;#39; if reading from standard input)">file</a>, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#4d60b5403d2d6aaeda541ed45625b245" title="line buffer size (automatically adjusted during execution to ensure the entire line...">buffer_size</a>, <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#2646e9e6d5da1693b4bea36a5acc082e" title="verbose mode">verbose</a>); 
<a name="l02133"></a>02133 
<a name="l02134"></a>02134 
<a name="l02136"></a>02136   <a class="code" href="classIndexedGenomicRegionSet.html#9b29696100e6af8b184d13d135ecf37b" title="Class destructor.">~IndexedGenomicRegionSet</a>();
<a name="l02137"></a>02137 
<a name="l02138"></a>02138 
<a name="l02140"></a>02140   <span class="keywordtype">void</span> <a class="code" href="classIndexedGenomicRegionSet.html#d05963371eebfcbd246d736290951888" title="Finds a match with genomic interval i.">Find</a>(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *i);
<a name="l02141"></a>02141 
<a name="l02142"></a>02142   
<a name="l02144"></a>02144   <span class="keywordtype">int</span> <a class="code" href="classIndexedGenomicRegionSet.html#e302fefddacaa1b77e69e0f1657a88d7" title="Given start,end in chromosome coordinates assign it a bin (Copyright 2002 Jim Kent)...">CalcBinFromRange</a>(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> end);
<a name="l02145"></a>02145 
<a name="l02146"></a>02146   
<a name="l02147"></a>02147   <span class="comment">// data</span>
<a name="l02148"></a>02148   
<a name="l02149"></a><a class="code" href="classIndexedGenomicRegionSet.html#594ca1013c6d77da12dfb9b626781f71">02149</a>   <a class="code" href="classIndexedGenomicRegionSet.html#c1e7aa6c1a55659c5756a70d0289c997" title="map of string to genomic interval list">RegIndex</a> <a class="code" href="classIndexedGenomicRegionSet.html#594ca1013c6d77da12dfb9b626781f71" title="index">regIndex</a>;                                            
<a name="l02150"></a>02150   
<a name="l02151"></a>02151 };
<a name="l02152"></a>02152 
<a name="l02153"></a>02153 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02154"></a>02154 <span class="comment">// END CLASS: IndexedGenomicRegionSet                                                          //</span>
<a name="l02155"></a>02155 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02156"></a>02156 
<a name="l02157"></a>02157 
<a name="l02158"></a>02158 
<a name="l02159"></a>02159 
<a name="l02160"></a>02160 
<a name="l02161"></a>02161 
<a name="l02162"></a>02162 
<a name="l02163"></a>02163 
<a name="l02164"></a>02164 
<a name="l02165"></a>02165 
<a name="l02166"></a>02166 
<a name="l02167"></a>02167 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02168"></a>02168 <span class="comment">// CLASS: GenomicRegionSetScanner                                                              //</span>
<a name="l02169"></a>02169 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02171"></a>02171 <span class="comment"></span>
<a name="l02195"></a>02195 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02196"></a><a class="code" href="classGenomicRegionSetScanner.html">02196</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description...">GenomicRegionSetScanner</a>
<a name="l02197"></a>02197 {
<a name="l02198"></a>02198  <span class="keyword">public</span>:
<a name="l02200"></a>02200 
<a name="l02209"></a>02209   <a class="code" href="classGenomicRegionSetScanner.html#2703f6892137282f00618895e02f628b" title="Class constructor.">GenomicRegionSetScanner</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetScanner.html#f76bf4fef482886f23f58cfa12c3eea1" title="pointer to the GenomicRegionSet which will be scanned by sliding windows">R</a>, StringLIntMap *<a class="code" href="classGenomicRegionSetScanner.html#7f64551c26331cd4ead1ab01405f1b72" title="chromosome sizes">bounds</a>, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#b278dfa27c3589865b9243156c1c727f" title="sliding window step">win_step</a>, <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#eed625b2a12aa2f7900997c4e20cd9b5" title="sliding window size (must be a multiple of window step)">win_size</a>, <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#d47603d08614ab993fe9036662c421e7" title="if true, genomic region labels are assumed to be integers and are included in the...">use_labels_as_values</a>, <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#10c1d22ae74e6f295cd17259f70eff8a" title="if true, no sliding windows on the negative strand are reported">ignore_reverse_strand</a>, <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#f2a536e23023f96c57f51e2276b4be0e" title="if &amp;#39;1&amp;#39;, only start position is counted; if &amp;#39;p&amp;#39;, all positions are...">preprocess</a>=<span class="charliteral">'1'</span>);
<a name="l02210"></a>02210   ~<a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description...">GenomicRegionSetScanner</a>();
<a name="l02211"></a>02211   
<a name="l02212"></a>02212   <span class="comment">// operations</span>
<a name="l02213"></a>02213   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#d0e45066dc2f22d3c211d1e7d885bd87" title="computes value in the next window">Next</a>();                                                      
<a name="l02214"></a>02214   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#d0e45066dc2f22d3c211d1e7d885bd87" title="computes value in the next window">Next</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *Ref);         
<a name="l02215"></a>02215   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSetScanner.html#e494d96f93477902ca339516906bb6da" title="prints current window&amp;#39;s interval">PrintInterval</a>();                                         
<a name="l02216"></a>02216   
<a name="l02217"></a>02217   <span class="comment">// data</span>
<a name="l02218"></a>02218  <span class="keyword">private</span>: 
<a name="l02219"></a>02219   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#2687f9ec4d1fac4856f6fd8bf823b65d" title="tests whether current input region should be skipped because it does not match any...">Test</a>();                          
<a name="l02220"></a>02220 
<a name="l02221"></a><a class="code" href="classGenomicRegionSetScanner.html#f76bf4fef482886f23f58cfa12c3eea1">02221</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *R;                          
<a name="l02222"></a><a class="code" href="classGenomicRegionSetScanner.html#7f64551c26331cd4ead1ab01405f1b72">02222</a>   StringLIntMap *bounds;                        
<a name="l02223"></a><a class="code" href="classGenomicRegionSetScanner.html#b278dfa27c3589865b9243156c1c727f">02223</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step;                            
<a name="l02224"></a><a class="code" href="classGenomicRegionSetScanner.html#eed625b2a12aa2f7900997c4e20cd9b5">02224</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size;                            
<a name="l02225"></a><a class="code" href="classGenomicRegionSetScanner.html#d47603d08614ab993fe9036662c421e7">02225</a>   <span class="keywordtype">bool</span> use_labels_as_values;            
<a name="l02226"></a><a class="code" href="classGenomicRegionSetScanner.html#10c1d22ae74e6f295cd17259f70eff8a">02226</a>   <span class="keywordtype">bool</span> ignore_reverse_strand;           
<a name="l02227"></a><a class="code" href="classGenomicRegionSetScanner.html#f2a536e23023f96c57f51e2276b4be0e">02227</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#f2a536e23023f96c57f51e2276b4be0e" title="if &amp;#39;1&amp;#39;, only start position is counted; if &amp;#39;p&amp;#39;, all positions are...">preprocess</a>;                                      
<a name="l02228"></a><a class="code" href="classGenomicRegionSetScanner.html#bab8a5c92ec438168e22eb6e33098aba">02228</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#bab8a5c92ec438168e22eb6e33098aba" title="win_size divided by win_step (note: each window is comprised of non-overlapping sub-windows...">n_win_combine</a>;                       
<a name="l02229"></a><a class="code" href="classGenomicRegionSetScanner.html#60dff3d69ed195c0492711ee295720c2">02229</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetScanner.html#60dff3d69ed195c0492711ee295720c2" title="pointer to the most recently scanned region">r</a>;                                     
<a name="l02230"></a><a class="code" href="classGenomicRegionSetScanner.html#3f0ba2ab42e8029297f63ba5cf2f9b95">02230</a>   StringLIntMap::iterator <a class="code" href="classGenomicRegionSetScanner.html#3f0ba2ab42e8029297f63ba5cf2f9b95" title="keeps track of current window&amp;#39;s chromosome information">chr</a>;          
<a name="l02231"></a><a class="code" href="classGenomicRegionSetScanner.html#0b9c57901528e75d9d32f4f98a708f4b">02231</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#0b9c57901528e75d9d32f4f98a708f4b" title="keeps track of current window&amp;#39;s strand information">strand</a>;                                          
<a name="l02232"></a><a class="code" href="classGenomicRegionSetScanner.html#305a53be4686bb873996e83b3c393701">02232</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#305a53be4686bb873996e83b3c393701" title="keeps track of current window&amp;#39;s start information">start</a>;                                       
<a name="l02233"></a><a class="code" href="classGenomicRegionSetScanner.html#f39ff3f8e4f1a1ab55c004cd06c2378a">02233</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#f39ff3f8e4f1a1ab55c004cd06c2378a" title="keeps track of current window&amp;#39;s stop information">stop</a>;                                        
<a name="l02234"></a><a class="code" href="classGenomicRegionSetScanner.html#19e26d0661ce5f380a72f3eb43b0b07f">02234</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classGenomicRegionSetScanner.html#19e26d0661ce5f380a72f3eb43b0b07f" title="a vector that keeps track of values in the sub-windows (of size win_step) that comprise...">v</a>;                                          
<a name="l02235"></a><a class="code" href="classGenomicRegionSetScanner.html#fa341bbfce6cffdc2880c8f1c80abeb1">02235</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#fa341bbfce6cffdc2880c8f1c80abeb1" title="the k-th element of v to be updated next (in a round-robin manner)">k</a>;                                           
<a name="l02236"></a><a class="code" href="classGenomicRegionSetScanner.html#80be27fad3a3456e7aecd71fb7439c5a">02236</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#80be27fad3a3456e7aecd71fb7439c5a" title="keeps track of current window&amp;#39;s value information (this is the sum of elements...">v_sum</a>;                                       
<a name="l02237"></a>02237 };
<a name="l02238"></a>02238 
<a name="l02239"></a>02239 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02240"></a>02240 <span class="comment">// END CLASS: GenomicRegionSetScanner                                                          //</span>
<a name="l02241"></a>02241 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02242"></a>02242 
<a name="l02243"></a>02243 
<a name="l02244"></a>02244 
<a name="l02245"></a>02245 
<a name="l02246"></a>02246 
<a name="l02247"></a>02247 
<a name="l02248"></a>02248 
<a name="l02249"></a>02249 
<a name="l02250"></a>02250 
<a name="l02251"></a>02251 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02252"></a>02252 <span class="comment">// CLASS: GenomicRegionSetOverlapScanner                                                       //</span>
<a name="l02253"></a>02253 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02255"></a>02255 <span class="comment"></span>
<a name="l02303"></a>02303 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02304"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html">02304</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetOverlapScanner.html" title="This class is used to find overlaps between two genomic regions sets. See detailed...">GenomicRegionSetOverlapScanner</a>
<a name="l02305"></a>02305 {
<a name="l02306"></a>02306  <span class="keyword">public</span>:
<a name="l02308"></a>02308 
<a name="l02313"></a>02313   <a class="code" href="classGenomicRegionSetOverlapScanner.html#268a52207c12005452c46e86167e4faf" title="Class constructor.">GenomicRegionSetOverlapScanner</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#f74bf7176b25f99da4c38e84f50efb75" title="pointer to query region set">QuerySet</a>, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#6c9bb71f3f0479b0f773d13315a64162" title="pointer to index region set">IndexSet</a>, <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetOverlapScanner.html#03774130ef410465b36be25ae28682e7" title="if true, query and index regions sorted by strand">sorted_by_strand</a>);
<a name="l02314"></a>02314 
<a name="l02315"></a>02315 
<a name="l02317"></a>02317   <a class="code" href="classGenomicRegionSetOverlapScanner.html#5670ab6cb78ef5034d3fd471d000aea0" title="Class destructor.">~GenomicRegionSetOverlapScanner</a>();
<a name="l02318"></a>02318 
<a name="l02319"></a>02319   
<a name="l02321"></a>02321   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#3fdc80b8ce58503fbf56c3408dfe41a5" title="Returns a pointer to the current query region (&amp;#39;NULL&amp;#39; if no more query regions...">GetQuery</a>();
<a name="l02322"></a>02322 
<a name="l02323"></a>02323 
<a name="l02325"></a>02325   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#a237159db6addc8f09c4cc1c078898af" title="Returns a pointer to the next query region (&amp;#39;NULL&amp;#39; if no more query regions...">NextQuery</a>();
<a name="l02326"></a>02326 
<a name="l02327"></a>02327 
<a name="l02329"></a>02329   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#8fabf222feb13df18c4516a0c11e6c94" title="Returns a pointer to the current index region that overlaps the current query region...">GetMatch</a>();
<a name="l02330"></a>02330 
<a name="l02331"></a>02331  
<a name="l02333"></a>02333   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#59683f41c3512969c06ebe12c3c34892" title="Returns a pointer to the next index region that overlaps the current query region...">NextMatch</a>();
<a name="l02334"></a>02334 
<a name="l02335"></a>02335  
<a name="l02337"></a>02337 
<a name="l02341"></a>02341   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#63d7837172475b7c75d1b5b28a4bca97" title="Returns a pointer to the current index region that overlaps the current query region...">GetOverlap</a>(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02342"></a>02342 
<a name="l02343"></a>02343 
<a name="l02345"></a>02345 
<a name="l02349"></a>02349   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#3358b5f338f92581b2e47f8e8ec9b0ef" title="Returns a pointer to the next index region that overlaps the current query region...">NextOverlap</a>(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02350"></a>02350 
<a name="l02351"></a>02351   
<a name="l02353"></a>02353   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetOverlapScanner.html#78c2866a7a33facc94aaa21513ae31cf" title="Returns true, if no more query or index regions are available and the GenomicRegionSetOverlapScanner...">Done</a>();
<a name="l02354"></a>02354  
<a name="l02355"></a>02355 
<a name="l02357"></a>02357 
<a name="l02361"></a>02361   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetOverlapScanner.html#2b51a7827283cf09ce9d0a822714d738" title="Calculates the total overlap between the current query region and the index region...">CalcCoverage</a>(<span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02362"></a>02362 
<a name="l02363"></a>02363 
<a name="l02365"></a>02365 
<a name="l02369"></a>02369   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetOverlapScanner.html#7826504ecdb81cf0828b428078ec1844" title="Calculates the total number of matches between the current query region and the index...">CountMatches</a>(<span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02370"></a>02370 
<a name="l02371"></a>02371  <span class="keyword">private</span>:
<a name="l02372"></a>02372   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSetOverlapScanner.html#2241b4411418e51f21dfbbff56332940" title="manages GenomicRegionSetOverlapScanner::IRegBuffer">LoadIndexBuffer</a>();                               
<a name="l02373"></a>02373   <span class="keywordtype">void</span> <a class="code" href="classGenomicRegionSetOverlapScanner.html#4c877e18eeeb397c5c2de2bca6ce0ff0" title="clears GenomicRegionSetOverlapScanner::IRegBuffer">ClearIndexBuffer</a>();                              
<a name="l02374"></a>02374 
<a name="l02375"></a>02375   <span class="comment">// data</span>
<a name="l02376"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#f74bf7176b25f99da4c38e84f50efb75">02376</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *QuerySet;                           
<a name="l02377"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#6c9bb71f3f0479b0f773d13315a64162">02377</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> *IndexSet;                           
<a name="l02378"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#03774130ef410465b36be25ae28682e7">02378</a>   <span class="keywordtype">bool</span> sorted_by_strand;                                
<a name="l02379"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#b802e705ad9fcfdb95fa665c26d1c610">02379</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#b802e705ad9fcfdb95fa665c26d1c610" title="pointer to the current query region">current_qreg</a>;                          
<a name="l02380"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#44fafd1bccf2b4610bf1979ca0b05c27">02380</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#44fafd1bccf2b4610bf1979ca0b05c27" title="pointer to the current index region">current_ireg</a>;                          
<a name="l02381"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd">02381</a>   GenomicRegionList <a class="code" href="classGenomicRegionSetOverlapScanner.html#5cef1f4f621533d6ce6a47fc106f35cd" title="temporary buffer containing index regions that overlap with current query; the buffer...">IRegBuffer</a>;                         
<a name="l02382"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#ab1a082fe808c27618040dfa0537fb5e">02382</a>   GenomicRegionList::iterator <a class="code" href="classGenomicRegionSetOverlapScanner.html#ab1a082fe808c27618040dfa0537fb5e" title="iterator on GenomicRegionSetOverlapScanner::IRegBuffer">IRegBufferIterator</a>;       
<a name="l02383"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#4dc563efd11721122a7863d574801cc8">02383</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicRegionSetOverlapScanner.html#4dc563efd11721122a7863d574801cc8" title="keeps tracks of the (linked) interval stored in GenomicRegionSetOverlapScanner::IRegBuffer...">ireg_buffer_interval</a>;                
<a name="l02384"></a><a class="code" href="classGenomicRegionSetOverlapScanner.html#4c8c433190c5f9a12754ae096ec2e8e5">02384</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetOverlapScanner.html#4c8c433190c5f9a12754ae096ec2e8e5" title="keeps track of the maximum size of GenomicRegionSetOverlapScanner::IRegBuffer">max_ireg_buffer_size</a>;               
<a name="l02385"></a>02385 };
<a name="l02386"></a>02386 
<a name="l02387"></a>02387 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02388"></a>02388 <span class="comment">// END CLASS: GenomicRegionSetOverlapScanner                                                   //</span>
<a name="l02389"></a>02389 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02390"></a>02390 
<a name="l02391"></a>02391 
<a name="l02392"></a>02392 
<a name="l02393"></a>02393 
<a name="l02394"></a>02394 <span class="keywordtype">char</span> ProcessStrand(<span class="keywordtype">char</span> *strand);
<a name="l02395"></a>02395 <span class="keywordtype">void</span> PrintChromosome(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">bool</span> convert);
<a name="l02396"></a>02396 <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *RegCenter(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r);
<a name="l02397"></a>02397 
<a name="l02398"></a>02398 StringLIntMap *ReadBounds(<span class="keywordtype">char</span> *genome_reg_file, <span class="keywordtype">bool</span> verbose=<span class="keyword">false</span>);           
<a name="l02399"></a>02399 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcBoundSize(StringLIntMap *bounds);
<a name="l02400"></a>02400 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcRegSize(<span class="keywordtype">char</span> *reg_file);
<a name="l02401"></a>02401 <span class="keywordtype">bool</span> CompareGenomicIntervals(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *J);
<a name="l02402"></a>02402 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsList *L);
<a name="l02403"></a>02403 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsVector *V);
<a name="l02404"></a>02404 <span class="keywordtype">void</span> SortGenomicIntervals(<a class="code" href="classGenomicIntervalSetAsArray.html" title="UNDER DEVELOPMENT: This class implements the an array of genomic intervals.">GenomicIntervalSetAsArray</a> *A);
<a name="l02405"></a>02405 
<a name="l02406"></a>02406 
<a name="l02407"></a>02407 
<a name="l02408"></a>02408 
<a name="l02409"></a>02409 
</pre></div></div>
<hr size="1"><address style="text-align: right;"><small>Generated on Mon Sep 12 16:49:28 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
