<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>genomic_intervals.h Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
<div class="header">
  <div class="headertitle">
<div class="title">genomic_intervals.h</div>  </div>
</div>
<div class="contents">
<div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">//</span>
<a name="l00002"></a>00002 <span class="comment">// Copyright (c) 2011 IBM Corporation. </span>
<a name="l00003"></a>00003 <span class="comment">// All rights reserved. This program and the accompanying materials are made available under the terms of the GNU General Public License v2.0 </span>
<a name="l00004"></a>00004 <span class="comment">// which accompanies this distribution, and is available at http://www.gnu.org/licenses/old-licenses/gpl-2.0.html</span>
<a name="l00005"></a>00005 <span class="comment">//</span>
<a name="l00006"></a>00006 
<a name="l00007"></a>00007 <span class="keyword">const</span> <span class="keywordtype">string</span> VERSION = <span class="stringliteral">&quot;genomic_tools 2.7.0&quot;</span>;
<a name="l00008"></a>00008 
<a name="l00009"></a>00009 
<a name="l00010"></a>00010 <span class="preprocessor">#include &lt;stdio.h&gt;</span>
<a name="l00011"></a>00011 <span class="preprocessor">#include &lt;stdlib.h&gt;</span>
<a name="l00012"></a>00012 <span class="preprocessor">#include &lt;string.h&gt;</span>
<a name="l00013"></a>00013 <span class="preprocessor">#include &lt;math.h&gt;</span>
<a name="l00014"></a>00014 <span class="preprocessor">#include &lt;unistd.h&gt;</span>
<a name="l00015"></a>00015 <span class="preprocessor">#include &lt;ctype.h&gt;</span>
<a name="l00016"></a>00016 <span class="preprocessor">#include &lt;sys/types.h&gt;</span>
<a name="l00017"></a>00017 <span class="preprocessor">#include &lt;sys/stat.h&gt;</span>
<a name="l00018"></a>00018 <span class="preprocessor">#include &lt;map&gt;</span>
<a name="l00019"></a>00019 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00020"></a>00020 <span class="preprocessor">#include &lt;list&gt;</span>
<a name="l00021"></a>00021 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00022"></a>00022 <span class="preprocessor">#include &lt;iostream&gt;</span>
<a name="l00023"></a>00023 <span class="preprocessor">#include &lt;algorithm&gt;</span>
<a name="l00024"></a>00024 
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 <span class="keyword">using namespace </span>std;
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <span class="comment">// forward declarations</span>
<a name="l00029"></a>00029 <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>;
<a name="l00030"></a>00030 <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>;
<a name="l00031"></a>00031 <span class="keyword">typedef</span> vector&lt;GenomicInterval *&gt; GenomicIntervalSetAsVector;
<a name="l00032"></a>00032 <span class="keyword">typedef</span> list&lt;GenomicInterval *&gt; GenomicIntervalSetAsList;
<a name="l00033"></a>00033 <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>;
<a name="l00034"></a>00034 <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>;
<a name="l00035"></a>00035 <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>;
<a name="l00036"></a>00036 <span class="keyword">class </span><a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a>;
<a name="l00037"></a>00037 
<a name="l00038"></a>00038 
<a name="l00039"></a>00039 <span class="comment">// types</span>
<a name="l00040"></a>00040 <span class="keyword">typedef</span> map&lt;string,string&gt; StringMap;
<a name="l00041"></a>00041 <span class="keyword">typedef</span> map&lt;string,long int&gt; StringLIntMap;
<a name="l00042"></a>00042 <span class="keyword">typedef</span> map&lt;string,vector&lt;long int&gt; &gt; StringVecLIntMap;
<a name="l00043"></a>00043 <span class="keyword">typedef</span> list&lt; pair&lt;long int,long int&gt; &gt; OffsetList;
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 <span class="comment">// choose container type for interval set (vector, list or array)</span>
<a name="l00046"></a>00046 <span class="keyword">typedef</span> GenomicIntervalSetAsVector GenomicIntervalSet;
<a name="l00047"></a>00047 <span class="comment">//typedef GenomicIntervalSetAsList GenomicIntervalSet;</span>
<a name="l00048"></a>00048 <span class="comment">//typedef GenomicIntervalSetAsArray GenomicIntervalSet;</span>
<a name="l00049"></a>00049 
<a name="l00050"></a>00050 <span class="keyword">typedef</span> list&lt;GenomicRegion*&gt; GenomicRegionList;
<a name="l00051"></a>00051 
<a name="l00052"></a>00052 <span class="keyword">struct </span>ChromosomeBins {
<a name="l00053"></a>00053         ChromosomeBins(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start) : min_start(start),max_start(start),n_bins(0),b_fwd(NULL),b_rev(NULL) { }
<a name="l00054"></a>00054         ~ChromosomeBins() { <span class="keywordflow">if</span> (b_fwd!=NULL) <span class="keyword">delete</span> b_fwd; <span class="keywordflow">if</span> (b_rev!=NULL) <span class="keyword">delete</span> b_rev; }
<a name="l00055"></a>00055         <span class="keywordtype">long</span> <span class="keywordtype">int</span> min_start, max_start;
<a name="l00056"></a>00056         <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_bins;
<a name="l00057"></a>00057         GenomicRegionList *b_fwd, *b_rev;
<a name="l00058"></a>00058 };
<a name="l00059"></a>00059 <span class="keyword">typedef</span> map&lt;string,ChromosomeBins*&gt; GenomicRegionSetBins;
<a name="l00060"></a>00060 
<a name="l00061"></a>00061 
<a name="l00062"></a>00062 
<a name="l00063"></a>00063 <span class="comment">// constants</span>
<a name="l00064"></a>00064 <span class="preprocessor">#define ALIGNMENT_GAP &#39;*&#39;                                       //! character for alignment gaps </span>
<a name="l00065"></a>00065 <span class="preprocessor"></span><span class="keyword">const</span> <span class="keywordtype">char</span> refCIGARop[] = <span class="stringliteral">&quot;MD=X&quot;</span>;                       
<a name="l00066"></a>00066 <span class="keyword">const</span> <span class="keywordtype">char</span> fragCIGARop[] = <span class="stringliteral">&quot;MIS=X&quot;</span>;                     
<a name="l00067"></a>00067 <span class="keyword">const</span> <span class="keywordtype">char</span> BED_SEPARATOR = <span class="charliteral">&#39;\t&#39;</span>;
<a name="l00068"></a>00068 
<a name="l00069"></a>00069 
<a name="l00070"></a>00070   
<a name="l00071"></a>00071   
<a name="l00072"></a>00072 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00073"></a>00073 <span class="comment">// CLASS DECLARATION: GenomicInterval                                                          //</span>
<a name="l00074"></a>00074 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00076"></a>00076 <span class="comment"></span>
<a name="l00090"></a>00090 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00091"></a><a class="code" href="classGenomicInterval.html">00091</a> <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>
<a name="l00092"></a>00092 {
<a name="l00093"></a>00093  <span class="keyword">public</span>:
<a name="l00095"></a>00095   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00096"></a>00096 
<a name="l00097"></a>00097   
<a name="l00098"></a>00098 
<a name="l00100"></a>00100   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>(<span class="keywordtype">string</span> chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00101"></a>00101 
<a name="l00102"></a>00102   
<a name="l00103"></a>00103 
<a name="l00105"></a>00105   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">char</span> delimiter=<span class="charliteral">&#39; &#39;</span>, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00106"></a>00106 
<a name="l00107"></a>00107   
<a name="l00108"></a>00108 
<a name="l00110"></a>00110   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I);
<a name="l00111"></a>00111 
<a name="l00112"></a>00112   
<a name="l00113"></a>00113   
<a name="l00115"></a>00115   ~<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>();
<a name="l00116"></a>00116 
<a name="l00117"></a>00117   
<a name="l00118"></a>00118 
<a name="l00119"></a>00119   <span class="comment">//------------------------------------------------------//</span>
<a name="l00120"></a>00120   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00121"></a>00121   <span class="comment">//------------------------------------------------------//</span>
<a name="l00122"></a>00122 
<a name="l00123"></a>00123 
<a name="l00125"></a>00125   <span class="keywordtype">void</span> PrintInterval();
<a name="l00126"></a>00126 
<a name="l00127"></a>00127   
<a name="l00128"></a>00128 
<a name="l00130"></a>00130   <span class="keywordtype">void</span> PrintInterval(FILE *file_ptr);
<a name="l00131"></a>00131 
<a name="l00132"></a>00132   
<a name="l00133"></a>00133 
<a name="l00135"></a>00135   <span class="keywordtype">void</span> PrintInterval(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00136"></a>00136 
<a name="l00137"></a>00137 
<a name="l00139"></a>00139   <span class="keywordtype">void</span> Print(<span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00140"></a>00140 
<a name="l00141"></a>00141   
<a name="l00142"></a>00142 
<a name="l00144"></a>00144   <span class="keywordtype">void</span> PrintWithLabel(<span class="keywordtype">char</span> *label, <span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00145"></a>00145 
<a name="l00146"></a>00146   
<a name="l00147"></a>00147   
<a name="l00148"></a>00148   <span class="comment">//------------------------------------------------------//</span>
<a name="l00149"></a>00149   <span class="comment">//   Check                                              //</span>
<a name="l00150"></a>00150   <span class="comment">//------------------------------------------------------//</span>
<a name="l00151"></a>00151 
<a name="l00152"></a>00152 
<a name="l00154"></a>00154   <span class="keywordtype">bool</span> IsValid();
<a name="l00155"></a>00155 
<a name="l00156"></a>00156   
<a name="l00157"></a>00157 
<a name="l00159"></a>00159   <span class="keywordtype">bool</span> CheckValid(<span class="keywordtype">bool</span> quiet=<span class="keyword">false</span>);
<a name="l00160"></a>00160 
<a name="l00161"></a>00161   
<a name="l00162"></a>00162 
<a name="l00164"></a>00164   <span class="keywordtype">bool</span> IsBefore(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00165"></a>00165 
<a name="l00166"></a>00166 
<a name="l00167"></a>00167 
<a name="l00169"></a>00169   <span class="keywordtype">bool</span> IsPosBefore(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I);
<a name="l00170"></a>00170 
<a name="l00171"></a>00171 
<a name="l00172"></a>00172 
<a name="l00174"></a>00174   <span class="keywordtype">bool</span> IsCompatibleWith(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00175"></a>00175 
<a name="l00176"></a>00176 
<a name="l00177"></a>00177 
<a name="l00179"></a>00179   <span class="keywordtype">bool</span> IsContained(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand=<span class="keyword">false</span>);
<a name="l00180"></a>00180 
<a name="l00181"></a>00181 
<a name="l00183"></a>00183   <span class="keywordtype">bool</span> OverlapsWith(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand=<span class="keyword">false</span>);
<a name="l00184"></a>00184 
<a name="l00185"></a>00185 
<a name="l00187"></a>00187   <span class="keywordtype">bool</span> OverlapsWith(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand=<span class="keyword">false</span>);
<a name="l00188"></a>00188 
<a name="l00189"></a>00189 
<a name="l00190"></a>00190 
<a name="l00191"></a>00191   <span class="comment">//------------------------------------------------------//</span>
<a name="l00192"></a>00192   <span class="comment">//   Set &amp; Get                                          //</span>
<a name="l00193"></a>00193   <span class="comment">//------------------------------------------------------//</span>
<a name="l00194"></a>00194 
<a name="l00195"></a>00195 
<a name="l00197"></a>00197   <span class="keywordtype">void</span> SetCoordinates(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00198"></a>00198 
<a name="l00199"></a>00199   
<a name="l00200"></a>00200 
<a name="l00202"></a>00202   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetCoordinate(<span class="keyword">const</span> <span class="keywordtype">char</span> *op);
<a name="l00203"></a>00203   
<a name="l00204"></a>00204 
<a name="l00205"></a>00205 
<a name="l00207"></a>00207   <span class="keywordtype">size_t</span> GetSize();
<a name="l00208"></a>00208   
<a name="l00209"></a>00209 
<a name="l00210"></a>00210 
<a name="l00212"></a>00212   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00213"></a>00213   
<a name="l00214"></a>00214 
<a name="l00215"></a>00215 
<a name="l00217"></a>00217   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00218"></a>00218   
<a name="l00219"></a>00219 
<a name="l00220"></a>00220 
<a name="l00221"></a>00221   <span class="comment">//------------------------------------------------------//</span>
<a name="l00222"></a>00222   <span class="comment">//   Operations                                         //</span>
<a name="l00223"></a>00223   <span class="comment">//------------------------------------------------------//</span>
<a name="l00224"></a>00224 
<a name="l00225"></a>00225 
<a name="l00227"></a>00227   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcOverlap(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00228"></a>00228 
<a name="l00229"></a>00229 
<a name="l00230"></a>00230 
<a name="l00232"></a>00232   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcDistanceFrom(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keyword">const</span> <span class="keywordtype">char</span> *op, <span class="keyword">const</span> <span class="keywordtype">char</span> *I_op);
<a name="l00233"></a>00233   
<a name="l00234"></a>00234 
<a name="l00235"></a>00235 
<a name="l00237"></a>00237   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00238"></a>00238   
<a name="l00239"></a>00239 
<a name="l00240"></a>00240 
<a name="l00242"></a>00242   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00243"></a>00243   
<a name="l00244"></a>00244 
<a name="l00245"></a>00245 
<a name="l00247"></a>00247   <span class="keywordtype">void</span> ReverseStrand(); 
<a name="l00248"></a>00248   
<a name="l00249"></a>00249 
<a name="l00250"></a>00250 
<a name="l00252"></a>00252 
<a name="l00257"></a>00257   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l00258"></a>00258   
<a name="l00259"></a>00259 
<a name="l00260"></a>00260 
<a name="l00262"></a>00262 
<a name="l00266"></a>00266   <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00267"></a>00267 
<a name="l00268"></a>00268 
<a name="l00269"></a>00269 
<a name="l00271"></a>00271   <span class="keywordtype">void</span> ReversePos(StringLIntMap *bounds);
<a name="l00272"></a>00272 
<a name="l00273"></a>00273 
<a name="l00274"></a>00274 
<a name="l00276"></a>00276   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l00277"></a>00277 
<a name="l00278"></a>00278 
<a name="l00279"></a>00279 
<a name="l00281"></a>00281   <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00282"></a>00282 
<a name="l00283"></a>00283 
<a name="l00284"></a>00284 
<a name="l00286"></a>00286   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00287"></a>00287 
<a name="l00288"></a>00288 
<a name="l00289"></a>00289 
<a name="l00291"></a>00291 
<a name="l00298"></a>00298   <span class="keywordtype">void</span> GetOffsetFrom(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *ref_int, <span class="keyword">const</span> <span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *start_offset, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *stop_offset);
<a name="l00299"></a>00299 
<a name="l00300"></a>00300 
<a name="l00301"></a>00301   
<a name="l00302"></a>00302 
<a name="l00304"></a>00304 
<a name="l00311"></a>00311   <span class="keywordtype">void</span> GetOffsetFrom(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *ref_reg, <span class="keyword">const</span> <span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *start_offset, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *stop_offset);
<a name="l00312"></a>00312 
<a name="l00313"></a>00313 
<a name="l00314"></a>00314 
<a name="l00315"></a>00315 
<a name="l00316"></a>00316   <span class="comment">//------------------------------------------------------//</span>
<a name="l00317"></a>00317   <span class="comment">//   Create                                             //</span>
<a name="l00318"></a>00318   <span class="comment">//------------------------------------------------------//</span>
<a name="l00319"></a>00319 
<a name="l00321"></a>00321 
<a name="l00324"></a>00324   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *CreateCenter(<span class="keywordtype">bool</span> stop_equals_start=<span class="keyword">false</span>);
<a name="l00325"></a>00325 
<a name="l00326"></a>00326 
<a name="l00327"></a>00327   
<a name="l00328"></a>00328 
<a name="l00329"></a>00329   <span class="comment">//------------------------------------------------------//</span>
<a name="l00330"></a>00330   <span class="comment">//   Data                                               //</span>
<a name="l00331"></a>00331   <span class="comment">//------------------------------------------------------//</span>
<a name="l00332"></a>00332   
<a name="l00333"></a><a class="code" href="classGenomicInterval.html#a0dbe6195570468c386eea923e30c762c">00333</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#a0dbe6195570468c386eea923e30c762c" title="file line number">n_line</a>;             
<a name="l00334"></a><a class="code" href="classGenomicInterval.html#ac489a1195c43cc93d0e192dc91d5a192">00334</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicInterval.html#ac489a1195c43cc93d0e192dc91d5a192" title="chromosome name (space and tab characters are not allowed)">CHROMOSOME</a>;                             
<a name="l00335"></a><a class="code" href="classGenomicInterval.html#a25f6603b48fcc57f97c89fa768b658f9">00335</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicInterval.html#a25f6603b48fcc57f97c89fa768b658f9" title="orientation (&#39;+&#39; or &#39;-&#39;)">STRAND</a>;                                  
<a name="l00336"></a><a class="code" href="classGenomicInterval.html#ac0abba93599f5cdf4b5a4e7ec5909bbd">00336</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#ac0abba93599f5cdf4b5a4e7ec5909bbd" title="interval start position in the chromosome (&gt;=1)">START</a>;                               
<a name="l00337"></a><a class="code" href="classGenomicInterval.html#ac5fac21da3939b7976859a54800b93f0">00337</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#ac5fac21da3939b7976859a54800b93f0" title="interval stop position in the chromosome (&lt;=chromosome_size)">STOP</a>;                                
<a name="l00338"></a>00338 };
<a name="l00339"></a>00339 
<a name="l00340"></a>00340 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00341"></a>00341 <span class="comment">// END CLASS DECLARATION: GenomicIntervals                                                     //</span>
<a name="l00342"></a>00342 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00343"></a>00343 
<a name="l00344"></a>00344 
<a name="l00345"></a>00345 
<a name="l00346"></a>00346 
<a name="l00347"></a>00347 
<a name="l00348"></a>00348 
<a name="l00349"></a>00349 
<a name="l00350"></a>00350 
<a name="l00351"></a>00351 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00352"></a>00352 <span class="comment">// CLASS: Chromosomes                                                                          //</span>
<a name="l00353"></a>00353 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00355"></a>00355 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00356"></a><a class="code" href="classChromosomes.html">00356</a> <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>
<a name="l00357"></a>00357 {
<a name="l00358"></a>00358  <span class="keyword">public</span>:
<a name="l00359"></a>00359 
<a name="l00360"></a><a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243">00360</a>   <span class="keyword">typedef</span> map&lt;string,pair&lt;char*,size_t&gt; &gt; <a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a>;            
<a name="l00361"></a>00361 
<a name="l00363"></a>00363 
<a name="l00396"></a>00396   <a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>(<span class="keywordtype">char</span> *chrom_map_dir, <span class="keywordtype">char</span> *chrom_map_name);
<a name="l00397"></a>00397 
<a name="l00398"></a>00398 
<a name="l00400"></a>00400 
<a name="l00416"></a>00416   <a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>(<span class="keywordtype">char</span> *fasta_file, <span class="keywordtype">bool</span> verbose);
<a name="l00417"></a>00417 
<a name="l00418"></a>00418   
<a name="l00420"></a>00420   ~<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>();
<a name="l00421"></a>00421   
<a name="l00422"></a>00422 
<a name="l00424"></a>00424   <span class="keywordtype">bool</span> FindChromosome(<span class="keywordtype">char</span> *chromosome_name);
<a name="l00425"></a>00425   
<a name="l00426"></a>00426   
<a name="l00428"></a>00428   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00429"></a>00429   
<a name="l00430"></a>00430   
<a name="l00432"></a>00432   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00433"></a>00433   
<a name="l00434"></a>00434   
<a name="l00436"></a>00436   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I);
<a name="l00437"></a>00437   
<a name="l00438"></a>00438   
<a name="l00440"></a>00440   <span class="keywordtype">void</span> LoadChromosome(<span class="keywordtype">string</span> chromosome_name);
<a name="l00441"></a>00441 
<a name="l00442"></a>00442   
<a name="l00443"></a>00443   <span class="comment">// data</span>
<a name="l00444"></a>00444 
<a name="l00445"></a><a class="code" href="classChromosomes.html#a954bae240564f6f589a30d253291eb75">00445</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#a954bae240564f6f589a30d253291eb75" title="if &#39;true&#39;, print messages while loading chromosome sequences">verbose</a>;                                                         
<a name="l00446"></a><a class="code" href="classChromosomes.html#a408a06b9ae635ceaf0b75452bcb62763">00446</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#a408a06b9ae635ceaf0b75452bcb62763" title="&#39;true&#39; if reading from FASTA file">load_in_memory</a>;                                          
<a name="l00447"></a><a class="code" href="classChromosomes.html#ad72dcc9b9d3b0b041b737d895a20eef6">00447</a>   StringMap <a class="code" href="classChromosomes.html#ad72dcc9b9d3b0b041b737d895a20eef6" title="maps chromosome names to chromosome sequence file names">chrom_map</a>;                                          
<a name="l00448"></a><a class="code" href="classChromosomes.html#a5fb6a314ac56a7ace50070fd3e5ae708">00448</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#a5fb6a314ac56a7ace50070fd3e5ae708" title="the directory where the chromosome map file is located">chrom_map_dir</a>;                                         
<a name="l00449"></a><a class="code" href="classChromosomes.html#a71b69dff9fc8c4cae1671df17e11a016">00449</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#a71b69dff9fc8c4cae1671df17e11a016" title="the name of the chromosome map file">chrom_map_name</a>;                                        
<a name="l00450"></a><a class="code" href="classChromosomes.html#ab3b502eec7bd2028297300aeb47b5844">00450</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#ab3b502eec7bd2028297300aeb47b5844" title="name of the currently used chromosome">current_chromosome_name</a>;                       
<a name="l00451"></a><a class="code" href="classChromosomes.html#afb59cb80755187f4546573f10110e669">00451</a>   <span class="keywordtype">size_t</span> <a class="code" href="classChromosomes.html#afb59cb80755187f4546573f10110e669" title="size of the currently used chromosome">current_chromosome_size</a>;                       
<a name="l00452"></a><a class="code" href="classChromosomes.html#af640cf0f09b109b3d46d70aba9207a2e">00452</a>   <span class="keywordtype">char</span> *<a class="code" href="classChromosomes.html#af640cf0f09b109b3d46d70aba9207a2e" title="pointer to the currently used chromosome sequence">current_chromosome_seq</a>;                         
<a name="l00453"></a><a class="code" href="classChromosomes.html#a673169d55aa0b94815e17b41566d4d8c">00453</a>   <a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a> <a class="code" href="classChromosomes.html#a673169d55aa0b94815e17b41566d4d8c" title="all chromosomes sequences and their sizes are stored here">chromosome_seq_data</a>;        
<a name="l00454"></a>00454 };
<a name="l00455"></a>00455 
<a name="l00456"></a>00456 
<a name="l00457"></a>00457 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00458"></a>00458 <span class="comment">// END CLASS: Chromosomes                                                                      //</span>
<a name="l00459"></a>00459 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00460"></a>00460 
<a name="l00461"></a>00461 
<a name="l00462"></a>00462 
<a name="l00463"></a>00463 
<a name="l00464"></a>00464 
<a name="l00465"></a>00465 
<a name="l00466"></a>00466 
<a name="l00467"></a>00467 
<a name="l00468"></a>00468 
<a name="l00469"></a>00469 
<a name="l00470"></a>00470 
<a name="l00471"></a>00471 
<a name="l00472"></a>00472 
<a name="l00473"></a>00473 
<a name="l00474"></a>00474 
<a name="l00475"></a>00475 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00476"></a>00476 <span class="comment">// CLASS: GenomicIntervalSetAsArray                                                            //</span>
<a name="l00477"></a>00477 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00479"></a>00479 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00480"></a>00480 
<a name="l00481"></a><a class="code" href="classGenomicIntervalSetAsArray.html">00481</a> <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>
<a name="l00482"></a>00482 {
<a name="l00483"></a>00483  <span class="keyword">public</span>:
<a name="l00484"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a1a0416c9fac7e9d965ae7fe84f1649c2">00484</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>** <a class="code" href="classGenomicIntervalSetAsArray.html#a1a0416c9fac7e9d965ae7fe84f1649c2" title="iterator defined as a pointer in the array of GenomicInterval">iterator</a>;                           
<a name="l00485"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0310f5d79c529b631a5ef71ef186b41c">00485</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a>** <a class="code" href="classGenomicIntervalSetAsArray.html#a0310f5d79c529b631a5ef71ef186b41c" title="reverse_iterator defined as a pointer in the array of GenomicInterval">reverse_iterator</a>;           
<a name="l00486"></a>00486 
<a name="l00487"></a>00487 
<a name="l00489"></a>00489   <a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>();
<a name="l00490"></a>00490 
<a name="l00491"></a>00491 
<a name="l00493"></a>00493   ~<a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>();
<a name="l00494"></a>00494   
<a name="l00495"></a>00495   
<a name="l00497"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a9e56381009d164d7625199b183845388">00497</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a9e56381009d164d7625199b183845388" title="Implementation of begin() for this container.">begin</a>() { <span class="keywordflow">return</span> &amp;I[0]; }
<a name="l00498"></a>00498   
<a name="l00499"></a>00499 
<a name="l00501"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a5b6145ceff99907d9075047519482584">00501</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a5b6145ceff99907d9075047519482584" title="Implementation of end() for this container.">end</a>() { <span class="keywordflow">return</span> &amp;I[n_intervals]; }
<a name="l00502"></a>00502   
<a name="l00503"></a>00503 
<a name="l00505"></a><a class="code" href="classGenomicIntervalSetAsArray.html#ac077589a55a7dfec9b2e6c38392f7444">00505</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#ac077589a55a7dfec9b2e6c38392f7444" title="Implementation of rbegin() for this container.">rbegin</a>() { <span class="keywordflow">return</span> &amp;I[n_intervals-1]; }
<a name="l00506"></a>00506   
<a name="l00507"></a>00507 
<a name="l00509"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a5b7fc7dc64857dfdadd1198393c0d117">00509</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a5b7fc7dc64857dfdadd1198393c0d117" title="Implementation of rend() for this container.">rend</a>() { <span class="keywordflow">return</span> &amp;I[-1]; }
<a name="l00510"></a>00510   
<a name="l00511"></a>00511 
<a name="l00513"></a><a class="code" href="classGenomicIntervalSetAsArray.html#aa74747b36b114c326edfaf1224270d0d">00513</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#aa74747b36b114c326edfaf1224270d0d" title="Implementation of front() for this container.">front</a>() { <span class="keywordflow">return</span> I[0]; }
<a name="l00514"></a>00514   
<a name="l00515"></a>00515 
<a name="l00517"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a11b483c12111ee96c15fec296c66ee15">00517</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#a11b483c12111ee96c15fec296c66ee15" title="Implementation of back() for this container.">back</a>() { <span class="keywordflow">return</span> I[n_intervals-1]; }
<a name="l00518"></a>00518   
<a name="l00519"></a>00519 
<a name="l00521"></a><a class="code" href="classGenomicIntervalSetAsArray.html#ad4bfef55744241849841c71d08835403">00521</a>   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicIntervalSetAsArray.html#ad4bfef55744241849841c71d08835403" title="Implementation of size() for this container.">size</a>() { <span class="keywordflow">return</span> n_intervals; }
<a name="l00522"></a>00522   
<a name="l00523"></a>00523 
<a name="l00525"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a4348e7636cb91405b08374a966e33a7c">00525</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a4348e7636cb91405b08374a966e33a7c" title="Implementation of push_back() for this container.">push_back</a>(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *j) { I[n_intervals++] = j; }
<a name="l00526"></a>00526   
<a name="l00527"></a>00527 
<a name="l00529"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a833395eaf9daa5626879eefa861d9ea5">00529</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a833395eaf9daa5626879eefa861d9ea5" title="Implementation of clear() for this container.">clear</a>() {  }
<a name="l00530"></a>00530   
<a name="l00531"></a>00531 
<a name="l00533"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0f8554cfca150d695d15691aa42df7fa">00533</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0f8554cfca150d695d15691aa42df7fa" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> j) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be erased!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00534"></a>00534 
<a name="l00535"></a>00535 
<a name="l00537"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0aae25ac2e9a7cf50a370703aa99df85">00537</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0aae25ac2e9a7cf50a370703aa99df85" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> i) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be erased!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00538"></a>00538   
<a name="l00539"></a>00539 
<a name="l00541"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0cd89244c0e75b88f9b56c9b9478d06b">00541</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0cd89244c0e75b88f9b56c9b9478d06b" title="Implementation of insert() for this container (exits with error message).">insert</a>(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *u) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be inserted!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00542"></a>00542   
<a name="l00543"></a>00543 
<a name="l00544"></a>00544   
<a name="l00545"></a>00545   <span class="comment">//------------------------------------------------------//</span>
<a name="l00546"></a>00546   <span class="comment">//   Data                                               //</span>
<a name="l00547"></a>00547   <span class="comment">//------------------------------------------------------//</span>
<a name="l00548"></a>00548   
<a name="l00549"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a9f9558bbaa8d682993a60fa142893998">00549</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>;                                 
<a name="l00550"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a6580b875cd9804ba59f88c5000f83f09">00550</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a6580b875cd9804ba59f88c5000f83f09" title="number of maximum allowed intervals (this will be removed)">max_intervals</a>;                               
<a name="l00551"></a><a class="code" href="classGenomicIntervalSetAsArray.html#af7ec8bbfea7748695aa35e6ec467a4b2">00551</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> **<a class="code" href="classGenomicIntervalSetAsArray.html#af7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>;                                  
<a name="l00552"></a>00552   
<a name="l00553"></a>00553 };
<a name="l00554"></a>00554 
<a name="l00555"></a>00555 
<a name="l00556"></a>00556 
<a name="l00557"></a>00557 
<a name="l00558"></a>00558 
<a name="l00559"></a>00559 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00560"></a>00560 <span class="comment">// END CLASS: GenomicIntervalArray                                                             //</span>
<a name="l00561"></a>00561 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00562"></a>00562 
<a name="l00563"></a>00563 
<a name="l00564"></a>00564 
<a name="l00565"></a>00565 
<a name="l00566"></a>00566 
<a name="l00567"></a>00567 
<a name="l00568"></a>00568 
<a name="l00569"></a>00569 
<a name="l00570"></a>00570 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00571"></a>00571 <span class="comment">// CLASS: GenomicRegion                                                                        //</span>
<a name="l00572"></a>00572 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00573"></a>00573 <span class="comment">//                                                                                             //</span>
<a name="l00574"></a>00574 <span class="comment">// Input formats:                                                                              //</span>
<a name="l00575"></a>00575 <span class="comment">//  1. LABEL &lt;TAB&gt; CHROMOSOME STRAND START[,START]* STOP[,STOP]*                               //</span>
<a name="l00576"></a>00576 <span class="comment">//  2. LABEL &lt;TAB&gt; [CHROMOSOME STRAND START STOP]+                                             //</span>
<a name="l00577"></a>00577 <span class="comment">//                                                                                             //</span>
<a name="l00578"></a>00578 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00580"></a>00580 <span class="comment"></span>
<a name="l00589"></a>00589 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00590"></a>00590 
<a name="l00591"></a><a class="code" href="classGenomicRegion.html">00591</a> <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l00592"></a>00592 {
<a name="l00593"></a>00593  <span class="keyword">public</span>:
<a name="l00595"></a>00595   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>();
<a name="l00596"></a>00596 
<a name="l00597"></a>00597 
<a name="l00598"></a>00598 
<a name="l00600"></a>00600   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l00601"></a>00601 
<a name="l00602"></a>00602 
<a name="l00603"></a>00603 
<a name="l00605"></a>00605   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0); 
<a name="l00606"></a>00606 
<a name="l00607"></a>00607 
<a name="l00608"></a>00608 
<a name="l00610"></a>00610   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *i);
<a name="l00611"></a>00611 
<a name="l00612"></a>00612 
<a name="l00613"></a>00613 
<a name="l00615"></a>00615   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>();
<a name="l00616"></a>00616   
<a name="l00617"></a>00617 
<a name="l00618"></a>00618 
<a name="l00619"></a>00619   <span class="comment">//------------------------------------------------------//</span>
<a name="l00620"></a>00620   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00621"></a>00621   <span class="comment">//------------------------------------------------------//</span>
<a name="l00622"></a>00622 
<a name="l00623"></a>00623 
<a name="l00625"></a>00625   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0, <span class="keywordtype">char</span> del1=<span class="charliteral">&#39; &#39;</span>, <span class="keywordtype">char</span> del2=<span class="charliteral">&#39;,&#39;</span>);
<a name="l00626"></a>00626 
<a name="l00627"></a>00627 
<a name="l00628"></a>00628 
<a name="l00630"></a>00630   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l00631"></a>00631 
<a name="l00632"></a>00632 
<a name="l00633"></a>00633   
<a name="l00635"></a>00635   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintModified(<span class="keyword">const</span> <span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00636"></a>00636 
<a name="l00637"></a>00637 
<a name="l00638"></a>00638 
<a name="l00640"></a>00640   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintIntersection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l00641"></a>00641 
<a name="l00642"></a>00642 
<a name="l00643"></a>00643 
<a name="l00645"></a>00645   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l00646"></a>00646 
<a name="l00647"></a>00647 
<a name="l00648"></a>00648 
<a name="l00650"></a>00650   <span class="keywordtype">void</span> PrintREG(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l00651"></a>00651 
<a name="l00652"></a>00652 
<a name="l00653"></a>00653 
<a name="l00655"></a>00655   <span class="keywordtype">void</span> PrintQ(<span class="keywordtype">bool</span> reverse=<span class="keyword">false</span>);
<a name="l00656"></a>00656 
<a name="l00657"></a>00657 
<a name="l00658"></a>00658 
<a name="l00660"></a>00660   <span class="keywordtype">void</span> PrintError(<span class="keywordtype">string</span> error_msg);
<a name="l00661"></a>00661 
<a name="l00662"></a>00662 
<a name="l00663"></a>00663 
<a name="l00664"></a>00664 
<a name="l00665"></a>00665   <span class="comment">//------------------------------------------------------//</span>
<a name="l00666"></a>00666   <span class="comment">//   Get &amp; Set                                          //</span>
<a name="l00667"></a>00667   <span class="comment">//------------------------------------------------------//</span>
<a name="l00668"></a>00668 
<a name="l00669"></a>00669 
<a name="l00671"></a>00671   <span class="keywordtype">void</span> DeleteIntervals(GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j); 
<a name="l00672"></a>00672 
<a name="l00673"></a>00673 
<a name="l00674"></a>00674 
<a name="l00676"></a>00676   <span class="keywordtype">void</span> SetLabel(<span class="keywordtype">char</span> *label);
<a name="l00677"></a>00677 
<a name="l00678"></a>00678 
<a name="l00679"></a>00679 
<a name="l00681"></a>00681   <span class="keywordtype">void</span> SetStrand(<span class="keywordtype">char</span> strand);
<a name="l00682"></a>00682 
<a name="l00683"></a>00683 
<a name="l00684"></a>00684 
<a name="l00686"></a>00686   <span class="keywordtype">char</span> *GetChromosome();
<a name="l00687"></a>00687 
<a name="l00688"></a>00688 
<a name="l00689"></a>00689 
<a name="l00691"></a>00691   <span class="keywordtype">size_t</span> GetSize(<span class="keywordtype">bool</span> skip_gaps);
<a name="l00692"></a>00692 
<a name="l00693"></a>00693 
<a name="l00694"></a>00694 
<a name="l00696"></a>00696   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00697"></a>00697 
<a name="l00698"></a>00698 
<a name="l00699"></a>00699 
<a name="l00701"></a>00701   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00702"></a>00702 
<a name="l00703"></a>00703 
<a name="l00704"></a>00704 
<a name="l00705"></a>00705   <span class="comment">//------------------------------------------------------//</span>
<a name="l00706"></a>00706   <span class="comment">//   Check                                              //</span>
<a name="l00707"></a>00707   <span class="comment">//------------------------------------------------------//</span>
<a name="l00708"></a>00708 
<a name="l00709"></a>00709 
<a name="l00711"></a>00711   <span class="keywordtype">bool</span> IsCompatible(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00712"></a>00712 
<a name="l00713"></a>00713 
<a name="l00714"></a>00714 
<a name="l00716"></a>00716   <span class="keywordtype">bool</span> IsCompatibleWith(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00717"></a>00717 
<a name="l00718"></a>00718 
<a name="l00719"></a>00719 
<a name="l00721"></a>00721   <span class="keywordtype">bool</span> IsCompatibleSorted(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00722"></a>00722 
<a name="l00723"></a>00723 
<a name="l00724"></a>00724 
<a name="l00726"></a>00726   <span class="keywordtype">bool</span> IsCompatibleSortedAndNonoverlapping();
<a name="l00727"></a>00727 
<a name="l00728"></a>00728 
<a name="l00729"></a>00729 
<a name="l00731"></a>00731   <span class="keywordtype">bool</span> OverlapsWith(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00732"></a>00732 
<a name="l00733"></a>00733 
<a name="l00734"></a>00734 
<a name="l00736"></a>00736   <span class="keywordtype">bool</span> IsBefore(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00737"></a>00737 
<a name="l00738"></a>00738 
<a name="l00739"></a>00739 
<a name="l00741"></a>00741   <span class="keywordtype">bool</span> IsPosBefore(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l00742"></a>00742 
<a name="l00743"></a>00743 
<a name="l00744"></a>00744 
<a name="l00746"></a>00746   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcOverlap(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00747"></a>00747 
<a name="l00748"></a>00748 
<a name="l00749"></a>00749 
<a name="l00751"></a>00751   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00752"></a>00752 
<a name="l00753"></a>00753   
<a name="l00754"></a>00754 
<a name="l00756"></a>00756   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00757"></a>00757 
<a name="l00758"></a>00758   
<a name="l00759"></a>00759 
<a name="l00760"></a>00760   <span class="comment">//------------------------------------------------------//</span>
<a name="l00761"></a>00761   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l00762"></a>00762   <span class="comment">//------------------------------------------------------//</span>
<a name="l00763"></a>00763 
<a name="l00764"></a>00764 
<a name="l00766"></a>00766   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00767"></a>00767 
<a name="l00768"></a>00768 
<a name="l00769"></a>00769   
<a name="l00771"></a>00771   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00772"></a>00772 
<a name="l00773"></a>00773 
<a name="l00774"></a>00774 
<a name="l00776"></a>00776   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l00777"></a>00777 
<a name="l00778"></a>00778 
<a name="l00779"></a>00779 
<a name="l00781"></a>00781   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Center();
<a name="l00782"></a>00782 
<a name="l00783"></a>00783 
<a name="l00784"></a>00784 
<a name="l00786"></a>00786   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyChromosomeNames(map&lt;string,string&gt; &amp;chromosome_conversion_map);
<a name="l00787"></a>00787 
<a name="l00788"></a>00788 
<a name="l00789"></a>00789 
<a name="l00791"></a>00791   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Connect();
<a name="l00792"></a>00792 
<a name="l00793"></a>00793 
<a name="l00794"></a>00794 
<a name="l00796"></a>00796   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Diff();
<a name="l00797"></a>00797 
<a name="l00798"></a>00798 
<a name="l00799"></a>00799 
<a name="l00801"></a>00801   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l00802"></a>00802 
<a name="l00803"></a>00803 
<a name="l00804"></a>00804 
<a name="l00806"></a>00806   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Divide();
<a name="l00807"></a>00807 
<a name="l00808"></a>00808 
<a name="l00809"></a>00809 
<a name="l00811"></a>00811   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunDivide();
<a name="l00812"></a>00812 
<a name="l00813"></a>00813 
<a name="l00814"></a>00814 
<a name="l00816"></a>00816   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Fix();
<a name="l00817"></a>00817 
<a name="l00818"></a>00818 
<a name="l00819"></a>00819 
<a name="l00821"></a>00821   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Intersect();
<a name="l00822"></a>00822 
<a name="l00823"></a>00823 
<a name="l00824"></a>00824 
<a name="l00826"></a>00826   <span class="keywordtype">void</span> RunSize();
<a name="l00827"></a>00827 
<a name="l00828"></a>00828 
<a name="l00829"></a>00829 
<a name="l00831"></a>00831 
<a name="l00835"></a>00835   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00836"></a>00836 
<a name="l00837"></a>00837 
<a name="l00838"></a>00838 
<a name="l00840"></a>00840   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00841"></a>00841 
<a name="l00842"></a>00842 
<a name="l00843"></a>00843 
<a name="l00845"></a>00845   <span class="keywordtype">void</span> ReversePos(StringLIntMap *bounds);
<a name="l00846"></a>00846 
<a name="l00847"></a>00847 
<a name="l00848"></a>00848   
<a name="l00850"></a>00850   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l00851"></a>00851 
<a name="l00852"></a>00852 
<a name="l00853"></a>00853 
<a name="l00855"></a>00855 
<a name="l00860"></a>00860   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l00861"></a>00861 
<a name="l00862"></a>00862 
<a name="l00863"></a>00863 
<a name="l00865"></a>00865   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l00866"></a>00866 
<a name="l00867"></a>00867 
<a name="l00868"></a>00868 
<a name="l00870"></a>00870   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Sort(); 
<a name="l00871"></a>00871 
<a name="l00872"></a>00872 
<a name="l00873"></a>00873   
<a name="l00875"></a>00875   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunSplit(<span class="keywordtype">bool</span> by_chrom_and_strand); 
<a name="l00876"></a>00876 
<a name="l00877"></a>00877 
<a name="l00878"></a>00878   
<a name="l00880"></a>00880   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyStrand(<span class="keywordtype">char</span> *strand_op);
<a name="l00881"></a>00881 
<a name="l00882"></a>00882 
<a name="l00883"></a>00883   
<a name="l00885"></a>00885   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Union();
<a name="l00886"></a>00886 
<a name="l00887"></a>00887 
<a name="l00888"></a>00888 
<a name="l00890"></a>00890   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunUnion();
<a name="l00891"></a>00891 
<a name="l00892"></a>00892 
<a name="l00893"></a>00893 
<a name="l00895"></a>00895   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l00896"></a>00896 
<a name="l00897"></a>00897 
<a name="l00898"></a>00898 
<a name="l00900"></a>00900   <span class="keywordtype">void</span> PrintRemoveN(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00901"></a>00901 
<a name="l00902"></a>00902 
<a name="l00903"></a>00903 
<a name="l00905"></a>00905   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l00906"></a>00906 
<a name="l00907"></a>00907   
<a name="l00909"></a>00909   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00910"></a>00910 
<a name="l00911"></a>00911 
<a name="l00912"></a>00912 
<a name="l00914"></a>00914 
<a name="l00918"></a>00918   <span class="keywordtype">void</span> PrintOffsetFormat(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l00919"></a>00919 
<a name="l00920"></a>00920 
<a name="l00921"></a>00921 
<a name="l00923"></a>00923   <span class="keywordtype">void</span> PrintSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00924"></a>00924 
<a name="l00925"></a>00925 
<a name="l00926"></a>00926 
<a name="l00928"></a>00928   <span class="keywordtype">void</span> PrintVerifySeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l00929"></a>00929 
<a name="l00930"></a>00930 
<a name="l00931"></a>00931 
<a name="l00933"></a>00933   <span class="keywordtype">void</span> ReverseStrand();
<a name="l00934"></a>00934 
<a name="l00935"></a>00935 
<a name="l00936"></a>00936 
<a name="l00937"></a>00937   <span class="comment">//------------------------------------------------------//</span>
<a name="l00938"></a>00938   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l00939"></a>00939   <span class="comment">//------------------------------------------------------//</span>
<a name="l00940"></a>00940 
<a name="l00941"></a>00941   
<a name="l00943"></a>00943   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Constrain(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l00944"></a>00944 
<a name="l00945"></a>00945   
<a name="l00946"></a>00946 
<a name="l00948"></a>00948   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Diff(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l00949"></a>00949 
<a name="l00950"></a>00950 
<a name="l00951"></a>00951   
<a name="l00952"></a>00952 
<a name="l00953"></a>00953   <span class="comment">//------------------------------------------------------//</span>
<a name="l00954"></a>00954   <span class="comment">//   Data                                               //</span>
<a name="l00955"></a>00955   <span class="comment">//------------------------------------------------------//</span>
<a name="l00956"></a>00956 
<a name="l00957"></a><a class="code" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">00957</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>;                             
<a name="l00958"></a><a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24">00958</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>;                                                  
<a name="l00959"></a><a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9">00959</a>   GenomicIntervalSet <a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>;                                 
<a name="l00960"></a>00960   
<a name="l00961"></a>00961   
<a name="l00962"></a>00962 };
<a name="l00963"></a>00963 
<a name="l00964"></a>00964 
<a name="l00965"></a>00965 
<a name="l00966"></a>00966 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00967"></a>00967 <span class="comment">// END CLASS: GenomicRegion                                                                    //</span>
<a name="l00968"></a>00968 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00969"></a>00969 
<a name="l00970"></a>00970 
<a name="l00971"></a>00971 
<a name="l00972"></a>00972 
<a name="l00973"></a>00973 
<a name="l00974"></a>00974 
<a name="l00975"></a>00975 
<a name="l00976"></a>00976 
<a name="l00977"></a>00977 
<a name="l00978"></a>00978 
<a name="l00979"></a>00979 
<a name="l00980"></a>00980 
<a name="l00981"></a>00981 
<a name="l00982"></a>00982 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00983"></a>00983 <span class="comment">// CLASS: GenomicRegionSEQ                                                                     //</span>
<a name="l00984"></a>00984 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00986"></a>00986 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00987"></a>00987 
<a name="l00988"></a><a class="code" href="classGenomicRegionSEQ.html">00988</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l00989"></a>00989 {
<a name="l00990"></a>00990  <span class="keyword">public</span>:
<a name="l00992"></a>00992   <a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l00993"></a>00993 
<a name="l00995"></a>00995   ~<a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a>();
<a name="l00996"></a>00996   
<a name="l00997"></a>00997 
<a name="l00998"></a>00998 
<a name="l00999"></a>00999   <span class="comment">//------------------------------------------------------//</span>
<a name="l01000"></a>01000   <span class="comment">//   Operations                                         //</span>
<a name="l01001"></a>01001   <span class="comment">//------------------------------------------------------//</span>
<a name="l01002"></a>01002   
<a name="l01004"></a>01004   <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l01005"></a>01005 
<a name="l01006"></a>01006   
<a name="l01007"></a>01007   <span class="comment">//------------------------------------------------------//</span>
<a name="l01008"></a>01008   <span class="comment">//   Data                                               //</span>
<a name="l01009"></a>01009   <span class="comment">//------------------------------------------------------//</span>
<a name="l01010"></a>01010 
<a name="l01011"></a><a class="code" href="classGenomicRegionSEQ.html#a37f8aa6c3c6d8d0059aa56944efa3a52">01011</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSEQ.html#a37f8aa6c3c6d8d0059aa56944efa3a52" title="sequence">SEQ</a>;                                                    
<a name="l01012"></a>01012 
<a name="l01013"></a>01013   
<a name="l01014"></a>01014 <span class="comment">/*</span>
<a name="l01015"></a>01015 <span class="comment">  // NOTE: utilize all data &amp; code related to the SEQ format in a new class: GenomicRegionSEQ.</span>
<a name="l01017"></a>01017 <span class="comment">  void PrintNoGaps();                                                                                   // alignments</span>
<a name="l01018"></a>01018 <span class="comment"></span>
<a name="l01020"></a>01020 <span class="comment">  void PrintModifyLabel();</span>
<a name="l01021"></a>01021 <span class="comment">  </span>
<a name="l01022"></a>01022 <span class="comment">  vector&lt;string&gt; Q;                                             //!&lt; sequences (e.g. from a multiple sequence alignment)</span>
<a name="l01023"></a>01023 <span class="comment">  vector&lt;string&gt; L;                                             //!&lt; sequence labels (e.g. names of species in multiple sequence alignment)</span>
<a name="l01024"></a>01024 <span class="comment">*/</span>
<a name="l01025"></a>01025 
<a name="l01026"></a>01026 };
<a name="l01027"></a>01027 
<a name="l01028"></a>01028 
<a name="l01029"></a>01029 
<a name="l01030"></a>01030 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01031"></a>01031 <span class="comment">// END CLASS: GenomicRegionSEQ                                                                 //</span>
<a name="l01032"></a>01032 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01033"></a>01033 
<a name="l01034"></a>01034 
<a name="l01035"></a>01035 
<a name="l01036"></a>01036 
<a name="l01037"></a>01037 
<a name="l01038"></a>01038 
<a name="l01039"></a>01039 
<a name="l01040"></a>01040 
<a name="l01041"></a>01041 
<a name="l01042"></a>01042 
<a name="l01043"></a>01043 
<a name="l01044"></a>01044 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01045"></a>01045 <span class="comment">// CLASS: GenomicRegionBEDToREG                                                                //</span>
<a name="l01046"></a>01046 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01048"></a>01048 <span class="comment"></span>
<a name="l01055"></a>01055 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01056"></a><a class="code" href="classGenomicRegionBEDToREG.html">01056</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01057"></a>01057 {
<a name="l01058"></a>01058  <span class="keyword">public</span>:
<a name="l01060"></a>01060   <a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01061"></a>01061 
<a name="l01062"></a>01062 
<a name="l01063"></a>01063 
<a name="l01065"></a>01065   <a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01066"></a>01066 
<a name="l01067"></a>01067 
<a name="l01068"></a>01068 
<a name="l01070"></a>01070   ~<a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>();
<a name="l01071"></a>01071   
<a name="l01072"></a>01072 
<a name="l01073"></a>01073 
<a name="l01075"></a>01075   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01076"></a>01076   
<a name="l01077"></a>01077 
<a name="l01078"></a>01078 
<a name="l01079"></a>01079   <span class="comment">// data</span>
<a name="l01080"></a>01080 
<a name="l01081"></a>01081 
<a name="l01082"></a>01082 };
<a name="l01083"></a>01083 
<a name="l01084"></a>01084 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01085"></a>01085 <span class="comment">// END CLASS: GenomicRegionBEDToREG                                                            //</span>
<a name="l01086"></a>01086 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01087"></a>01087 
<a name="l01088"></a>01088 
<a name="l01089"></a>01089 
<a name="l01090"></a>01090 
<a name="l01091"></a>01091 
<a name="l01092"></a>01092 
<a name="l01093"></a>01093 
<a name="l01094"></a>01094 
<a name="l01095"></a>01095 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01096"></a>01096 <span class="comment">// CLASS: GenomicRegionBED                                                                     //</span>
<a name="l01097"></a>01097 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01099"></a>01099 <span class="comment"></span>
<a name="l01106"></a>01106 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01107"></a><a class="code" href="classGenomicRegionBED.html">01107</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01108"></a>01108 {
<a name="l01109"></a>01109  <span class="keyword">public</span>:
<a name="l01111"></a>01111   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01112"></a>01112 
<a name="l01113"></a>01113 
<a name="l01114"></a>01114 
<a name="l01116"></a>01116   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01117"></a>01117 
<a name="l01118"></a>01118 
<a name="l01119"></a>01119 
<a name="l01121"></a>01121   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *i);
<a name="l01122"></a>01122 
<a name="l01123"></a>01123 
<a name="l01124"></a>01124 
<a name="l01126"></a>01126   ~<a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>();
<a name="l01127"></a>01127   
<a name="l01128"></a>01128 
<a name="l01129"></a>01129 
<a name="l01130"></a>01130   <span class="comment">//------------------------------------------------------//</span>
<a name="l01131"></a>01131   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01132"></a>01132   <span class="comment">//------------------------------------------------------//</span>
<a name="l01133"></a>01133 
<a name="l01134"></a>01134 
<a name="l01136"></a>01136   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01137"></a>01137   
<a name="l01138"></a>01138 
<a name="l01139"></a>01139 
<a name="l01141"></a>01141   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01142"></a>01142 
<a name="l01143"></a>01143 
<a name="l01144"></a>01144 
<a name="l01146"></a>01146   <span class="keywordtype">void</span> Print(FILE *file_ptr, <span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01147"></a>01147 
<a name="l01148"></a>01148 
<a name="l01149"></a>01149 
<a name="l01151"></a>01151   <span class="keywordtype">void</span> PrintIntersection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l01152"></a>01152 
<a name="l01153"></a>01153 
<a name="l01154"></a>01154 
<a name="l01156"></a>01156   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01157"></a>01157 
<a name="l01158"></a>01158 
<a name="l01159"></a>01159 
<a name="l01161"></a>01161   <span class="keywordtype">void</span> PrintModified(<span class="keyword">const</span> <span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01162"></a>01162 
<a name="l01163"></a>01163 
<a name="l01164"></a>01164 
<a name="l01165"></a>01165   <span class="comment">//------------------------------------------------------//</span>
<a name="l01166"></a>01166   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01167"></a>01167   <span class="comment">//------------------------------------------------------//</span>
<a name="l01168"></a>01168 
<a name="l01169"></a>01169 
<a name="l01171"></a>01171   <span class="keywordtype">void</span> UpdateThick();
<a name="l01172"></a>01172   
<a name="l01173"></a>01173   
<a name="l01174"></a>01174   
<a name="l01176"></a>01176   <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01177"></a>01177 
<a name="l01178"></a>01178 
<a name="l01179"></a>01179 
<a name="l01181"></a>01181   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l01182"></a>01182 
<a name="l01183"></a>01183 
<a name="l01184"></a>01184 
<a name="l01186"></a>01186   <span class="keywordtype">void</span> Center();
<a name="l01187"></a>01187 
<a name="l01188"></a>01188 
<a name="l01189"></a>01189 
<a name="l01191"></a>01191   <span class="keywordtype">void</span> Connect();
<a name="l01192"></a>01192 
<a name="l01193"></a>01193 
<a name="l01194"></a>01194 
<a name="l01196"></a>01196   <span class="keywordtype">void</span> Diff();
<a name="l01197"></a>01197 
<a name="l01198"></a>01198 
<a name="l01199"></a>01199 
<a name="l01201"></a>01201   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01202"></a>01202 
<a name="l01203"></a>01203 
<a name="l01204"></a>01204 
<a name="l01206"></a>01206   <span class="keywordtype">void</span> Divide();
<a name="l01207"></a>01207 
<a name="l01208"></a>01208 
<a name="l01209"></a>01209 
<a name="l01211"></a>01211   <span class="keywordtype">void</span> RunDivide();
<a name="l01212"></a>01212 
<a name="l01213"></a>01213 
<a name="l01214"></a>01214 
<a name="l01216"></a>01216   <span class="keywordtype">void</span> Fix();
<a name="l01217"></a>01217 
<a name="l01218"></a>01218 
<a name="l01219"></a>01219 
<a name="l01221"></a>01221   <span class="keywordtype">void</span> Intersect();
<a name="l01222"></a>01222 
<a name="l01223"></a>01223 
<a name="l01224"></a>01224 
<a name="l01226"></a>01226 
<a name="l01230"></a>01230   <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01231"></a>01231 
<a name="l01232"></a>01232 
<a name="l01233"></a>01233 
<a name="l01235"></a>01235   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01236"></a>01236 
<a name="l01237"></a>01237 
<a name="l01238"></a>01238 
<a name="l01240"></a>01240   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01241"></a>01241 
<a name="l01242"></a>01242 
<a name="l01243"></a>01243 
<a name="l01245"></a>01245 
<a name="l01250"></a>01250   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l01251"></a>01251 
<a name="l01252"></a>01252 
<a name="l01253"></a>01253 
<a name="l01255"></a>01255   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01256"></a>01256 
<a name="l01257"></a>01257 
<a name="l01258"></a>01258 
<a name="l01260"></a>01260   <span class="keywordtype">void</span> Sort(); 
<a name="l01261"></a>01261 
<a name="l01262"></a>01262 
<a name="l01263"></a>01263   
<a name="l01265"></a>01265   <span class="keywordtype">void</span> RunSplit(<span class="keywordtype">bool</span> by_chrom_and_strand); 
<a name="l01266"></a>01266 
<a name="l01267"></a>01267 
<a name="l01268"></a>01268   
<a name="l01270"></a>01270   <span class="keywordtype">void</span> Union();
<a name="l01271"></a>01271 
<a name="l01272"></a>01272 
<a name="l01273"></a>01273 
<a name="l01275"></a>01275   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01276"></a>01276 
<a name="l01277"></a>01277 
<a name="l01278"></a>01278 
<a name="l01279"></a>01279   <span class="comment">//------------------------------------------------------//</span>
<a name="l01280"></a>01280   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l01281"></a>01281   <span class="comment">//------------------------------------------------------//</span>
<a name="l01282"></a>01282 
<a name="l01283"></a>01283   
<a name="l01285"></a>01285   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> *Constrain(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l01286"></a>01286 
<a name="l01287"></a>01287   
<a name="l01288"></a>01288 
<a name="l01290"></a>01290   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Diff(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l01291"></a>01291 
<a name="l01292"></a>01292   
<a name="l01293"></a>01293 
<a name="l01294"></a>01294   <span class="comment">//------------------------------------------------------//</span>
<a name="l01295"></a>01295   <span class="comment">//   Data                                               //</span>
<a name="l01296"></a>01296   <span class="comment">//------------------------------------------------------//</span>
<a name="l01297"></a>01297 
<a name="l01298"></a><a class="code" href="classGenomicRegionBED.html#a60878391f6fb1e0ad0ac9bc88a7625d3">01298</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a60878391f6fb1e0ad0ac9bc88a7625d3" title="number of tokens">n_tokens</a>;                    
<a name="l01299"></a><a class="code" href="classGenomicRegionBED.html#afa7ba810a0d1673f5dbabcf85c7bbb95">01299</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#afa7ba810a0d1673f5dbabcf85c7bbb95" title="score field">score</a>;                               
<a name="l01300"></a><a class="code" href="classGenomicRegionBED.html#a52fc75b347dfe6696e7ad2beeeb623a9">01300</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionBED.html#a52fc75b347dfe6696e7ad2beeeb623a9" title="color field">itemRgb</a>;                                
<a name="l01301"></a><a class="code" href="classGenomicRegionBED.html#a019dfaebedc17e8859b36c3359287faa">01301</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a019dfaebedc17e8859b36c3359287faa" title="thickStart field">thickStart</a>;                  
<a name="l01302"></a><a class="code" href="classGenomicRegionBED.html#a5333a7fdc1f21db508aa7af4f50553ba">01302</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a5333a7fdc1f21db508aa7af4f50553ba" title="thickEnd field">thickEnd</a>;                    
<a name="l01303"></a>01303 
<a name="l01304"></a>01304 };
<a name="l01305"></a>01305 
<a name="l01306"></a>01306 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01307"></a>01307 <span class="comment">// END CLASS: GenomicRegionBED                                                                 //</span>
<a name="l01308"></a>01308 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01309"></a>01309 
<a name="l01310"></a>01310 
<a name="l01311"></a>01311 
<a name="l01312"></a>01312 
<a name="l01313"></a>01313 
<a name="l01314"></a>01314 
<a name="l01315"></a>01315 
<a name="l01316"></a>01316 
<a name="l01317"></a>01317 
<a name="l01318"></a>01318 
<a name="l01319"></a>01319 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01320"></a>01320 <span class="comment">// CLASS: GenomicRegionSAMToREG                                                                //</span>
<a name="l01321"></a>01321 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01323"></a>01323 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01324"></a><a class="code" href="classGenomicRegionSAMToREG.html">01324</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01325"></a>01325 {
<a name="l01326"></a>01326  <span class="keyword">public</span>:
<a name="l01327"></a>01327   
<a name="l01329"></a>01329   <a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01330"></a>01330 
<a name="l01331"></a>01331 
<a name="l01332"></a>01332 
<a name="l01334"></a>01334   <a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01335"></a>01335 
<a name="l01336"></a>01336 
<a name="l01337"></a>01337 
<a name="l01339"></a>01339   ~<a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>();
<a name="l01340"></a>01340   
<a name="l01341"></a>01341 
<a name="l01342"></a>01342 
<a name="l01344"></a>01344   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01345"></a>01345   
<a name="l01346"></a>01346 
<a name="l01347"></a>01347 
<a name="l01349"></a>01349   <span class="keywordtype">char</span> CalcStrandFromFlag(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01350"></a>01350 
<a name="l01351"></a>01351 
<a name="l01352"></a>01352   
<a name="l01354"></a>01354   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetNextTokenOfCIGAR(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01355"></a>01355 
<a name="l01356"></a>01356   
<a name="l01357"></a>01357   
<a name="l01358"></a>01358   <span class="comment">// data</span>
<a name="l01359"></a>01359 
<a name="l01360"></a>01360 
<a name="l01361"></a>01361 };
<a name="l01362"></a>01362 
<a name="l01363"></a>01363 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01364"></a>01364 <span class="comment">// END CLASS: GenomicRegionSAMToREG                                                            //</span>
<a name="l01365"></a>01365 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01366"></a>01366 
<a name="l01367"></a>01367 
<a name="l01368"></a>01368 
<a name="l01369"></a>01369 
<a name="l01370"></a>01370 
<a name="l01371"></a>01371 
<a name="l01372"></a>01372 
<a name="l01373"></a>01373 
<a name="l01374"></a>01374 
<a name="l01375"></a>01375 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01376"></a>01376 <span class="comment">// CLASS: GenomicRegionSAM                                                                     //</span>
<a name="l01377"></a>01377 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01379"></a>01379 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01380"></a><a class="code" href="classGenomicRegionSAM.html">01380</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01381"></a>01381 {
<a name="l01382"></a>01382  <span class="keyword">public</span>:
<a name="l01383"></a><a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96">01383</a>   <span class="keyword">typedef</span> vector&lt;pair&lt;long int,char&gt; &gt; <a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length)">CIGARTokens</a>;                     
<a name="l01384"></a>01384   
<a name="l01386"></a>01386   <a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01387"></a>01387 
<a name="l01388"></a>01388 
<a name="l01389"></a>01389 
<a name="l01391"></a>01391   <a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01392"></a>01392 
<a name="l01393"></a>01393 
<a name="l01394"></a>01394 
<a name="l01396"></a>01396   ~<a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>();
<a name="l01397"></a>01397   
<a name="l01398"></a>01398 
<a name="l01399"></a>01399 
<a name="l01400"></a>01400   <span class="comment">//------------------------------------------------------//</span>
<a name="l01401"></a>01401   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01402"></a>01402   <span class="comment">//------------------------------------------------------//</span>
<a name="l01403"></a>01403 
<a name="l01404"></a>01404 
<a name="l01406"></a>01406   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01407"></a>01407 
<a name="l01408"></a>01408 
<a name="l01409"></a>01409 
<a name="l01411"></a>01411   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01412"></a>01412 
<a name="l01413"></a>01413 
<a name="l01414"></a>01414 
<a name="l01416"></a>01416   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01417"></a>01417 
<a name="l01418"></a>01418 
<a name="l01419"></a>01419 
<a name="l01421"></a>01421   <span class="keywordtype">void</span> PrintModified(<span class="keyword">const</span> <span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01422"></a>01422 
<a name="l01423"></a>01423 
<a name="l01424"></a>01424   
<a name="l01425"></a>01425   <span class="comment">//------------------------------------------------------//</span>
<a name="l01426"></a>01426   <span class="comment">//   Auxiliary functions                                //</span>
<a name="l01427"></a>01427   <span class="comment">//------------------------------------------------------//</span>
<a name="l01428"></a>01428 
<a name="l01430"></a>01430   <span class="keywordtype">char</span> CalcStrandFromFlag(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01431"></a>01431 
<a name="l01432"></a>01432 
<a name="l01433"></a>01433   
<a name="l01435"></a>01435   <span class="keywordtype">void</span> UpdateFlagFromStrand(<span class="keywordtype">char</span> strand);
<a name="l01436"></a>01436   
<a name="l01437"></a>01437   
<a name="l01438"></a>01438   
<a name="l01440"></a>01440   <span class="keywordtype">void</span> PrintGappedSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01441"></a>01441 
<a name="l01442"></a>01442   
<a name="l01443"></a>01443   
<a name="l01445"></a>01445   <span class="keywordtype">char</span> *GetGappedSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01446"></a>01446 
<a name="l01447"></a>01447   
<a name="l01448"></a>01448   
<a name="l01450"></a>01450   <span class="keywordtype">void</span> InvalidateSeqData();
<a name="l01451"></a>01451 
<a name="l01452"></a>01452 
<a name="l01453"></a>01453   
<a name="l01455"></a>01455   <span class="keywordtype">void</span> TokenizeCIGAR();
<a name="l01456"></a>01456 
<a name="l01457"></a>01457   
<a name="l01458"></a>01458   
<a name="l01460"></a>01460   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetNextTokenOfCIGAR(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01461"></a>01461 
<a name="l01462"></a>01462   
<a name="l01463"></a>01463   
<a name="l01465"></a>01465   <span class="keywordtype">void</span> UpdateCigarFromIntervals();
<a name="l01466"></a>01466   
<a name="l01467"></a>01467   
<a name="l01468"></a>01468  
<a name="l01470"></a>01470   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcFragmentLengthFromCIGAR();
<a name="l01471"></a>01471   
<a name="l01472"></a>01472   
<a name="l01473"></a>01473   
<a name="l01475"></a>01475   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcReferenceLengthFromCIGAR();
<a name="l01476"></a>01476   
<a name="l01477"></a>01477   
<a name="l01478"></a>01478   
<a name="l01480"></a>01480   <span class="keywordtype">char</span> *TrimCIGAR(<span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01481"></a>01481  
<a name="l01482"></a>01482  
<a name="l01483"></a>01483  
<a name="l01485"></a>01485   <span class="keywordtype">char</span> *TrimSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01486"></a>01486  
<a name="l01487"></a>01487  
<a name="l01488"></a>01488  
<a name="l01489"></a>01489   <span class="comment">//------------------------------------------------------//</span>
<a name="l01490"></a>01490   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01491"></a>01491   <span class="comment">//------------------------------------------------------//</span>
<a name="l01492"></a>01492 
<a name="l01493"></a>01493 
<a name="l01495"></a>01495   <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01496"></a>01496 
<a name="l01497"></a>01497 
<a name="l01498"></a>01498   
<a name="l01500"></a>01500   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l01501"></a>01501 
<a name="l01502"></a>01502 
<a name="l01503"></a>01503 
<a name="l01505"></a>01505   <span class="keywordtype">void</span> Center();
<a name="l01506"></a>01506 
<a name="l01507"></a>01507 
<a name="l01508"></a>01508 
<a name="l01510"></a>01510   <span class="keywordtype">void</span> Connect();
<a name="l01511"></a>01511 
<a name="l01512"></a>01512 
<a name="l01513"></a>01513 
<a name="l01515"></a>01515   <span class="keywordtype">void</span> Diff();
<a name="l01516"></a>01516 
<a name="l01517"></a>01517 
<a name="l01518"></a>01518 
<a name="l01520"></a>01520   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01521"></a>01521 
<a name="l01522"></a>01522 
<a name="l01523"></a>01523 
<a name="l01525"></a>01525   <span class="keywordtype">void</span> Divide();
<a name="l01526"></a>01526 
<a name="l01527"></a>01527 
<a name="l01528"></a>01528 
<a name="l01530"></a>01530   <span class="keywordtype">void</span> RunDivide();
<a name="l01531"></a>01531 
<a name="l01532"></a>01532 
<a name="l01533"></a>01533 
<a name="l01535"></a>01535   <span class="keywordtype">void</span> Fix();
<a name="l01536"></a>01536 
<a name="l01537"></a>01537 
<a name="l01538"></a>01538 
<a name="l01540"></a>01540   <span class="keywordtype">void</span> Intersect();
<a name="l01541"></a>01541 
<a name="l01542"></a>01542 
<a name="l01543"></a>01543 
<a name="l01545"></a>01545   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01546"></a>01546 
<a name="l01547"></a>01547 
<a name="l01548"></a>01548 
<a name="l01550"></a>01550 
<a name="l01554"></a>01554   <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01555"></a>01555 
<a name="l01556"></a>01556 
<a name="l01557"></a>01557 
<a name="l01559"></a>01559   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01560"></a>01560 
<a name="l01561"></a>01561 
<a name="l01562"></a>01562 
<a name="l01564"></a>01564 
<a name="l01569"></a>01569   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l01570"></a>01570 
<a name="l01571"></a>01571 
<a name="l01572"></a>01572 
<a name="l01574"></a>01574   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01575"></a>01575 
<a name="l01576"></a>01576 
<a name="l01577"></a>01577 
<a name="l01579"></a>01579   <span class="keywordtype">void</span> Sort(); 
<a name="l01580"></a>01580 
<a name="l01581"></a>01581 
<a name="l01582"></a>01582   
<a name="l01584"></a>01584   <span class="keywordtype">void</span> RunSplit(<span class="keywordtype">bool</span> by_chrom_and_strand); 
<a name="l01585"></a>01585 
<a name="l01586"></a>01586 
<a name="l01587"></a>01587   
<a name="l01589"></a>01589   <span class="keywordtype">void</span> ModifyStrand(<span class="keywordtype">char</span> *strand_op);
<a name="l01590"></a>01590 
<a name="l01591"></a>01591 
<a name="l01592"></a>01592   
<a name="l01594"></a>01594   <span class="keywordtype">void</span> Union();
<a name="l01595"></a>01595 
<a name="l01596"></a>01596 
<a name="l01597"></a>01597 
<a name="l01599"></a>01599   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01600"></a>01600 
<a name="l01601"></a>01601 
<a name="l01602"></a>01602 
<a name="l01603"></a>01603   <span class="comment">//------------------------------------------------------//</span>
<a name="l01604"></a>01604   <span class="comment">//   Data                                               //</span>
<a name="l01605"></a>01605   <span class="comment">//------------------------------------------------------//</span>
<a name="l01606"></a>01606 
<a name="l01607"></a><a class="code" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f">01607</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f" title="number of TAB-separated fields in SAM entry (&gt;= 12)">n_tokens</a>;                                    
<a name="l01608"></a>01608   <span class="comment">//char *QNAME;                                                //!&lt; Query template NAME: mapped to GenomicRegion::LABEL</span>
<a name="l01609"></a><a class="code" href="classGenomicRegionSAM.html#a200d7f7044dd69311eceb851a3c315e5">01609</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#a200d7f7044dd69311eceb851a3c315e5" title="bitwise FLAG">FLAG</a>;                               
<a name="l01610"></a>01610   <span class="comment">//char *RNAME;                                                //!&lt; Reference sequence NAME: mapped to GenomicRegion::I.front()-&gt;CHROMOSOME</span>
<a name="l01611"></a>01611   <span class="comment">//long int POS;                                               //!&lt; 1-based leftmost mapping POSition: mapped to GenomicRegion::I.front()-&gt;START</span>
<a name="l01612"></a><a class="code" href="classGenomicRegionSAM.html#aedacdd8319f4b614229a41b80d4c6f67">01612</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#aedacdd8319f4b614229a41b80d4c6f67" title="MAPping Quality.">MAPQ</a>;                                                
<a name="l01613"></a><a class="code" href="classGenomicRegionSAM.html#aa5acff481ae565db4976e66933212726">01613</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#aa5acff481ae565db4976e66933212726" title="CIGAR string.">CIGAR</a>;                                                  
<a name="l01614"></a><a class="code" href="classGenomicRegionSAM.html#af723505bfafff1a1fa616479023d0c3c">01614</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#af723505bfafff1a1fa616479023d0c3c" title="Reference name of the mate/next fragment.">RNEXT</a>;                                                  
<a name="l01615"></a><a class="code" href="classGenomicRegionSAM.html#ad6f88b4324ba508b86102641d819661f">01615</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#ad6f88b4324ba508b86102641d819661f" title="Position of the mate/next fragment.">PNEXT</a>;                                               
<a name="l01616"></a><a class="code" href="classGenomicRegionSAM.html#af180854a186be37eab49520edaa1262f">01616</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#af180854a186be37eab49520edaa1262f" title="observed Template LENgth">TLEN</a>;                                                
<a name="l01617"></a><a class="code" href="classGenomicRegionSAM.html#a6c2887ed1fa3a25d801641ea1b139032">01617</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#a6c2887ed1fa3a25d801641ea1b139032" title="fragment SEQuence">SEQ</a>;                                                    
<a name="l01618"></a><a class="code" href="classGenomicRegionSAM.html#ab43ca5f2fbfdee520f1424c3a9fd0706">01618</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#ab43ca5f2fbfdee520f1424c3a9fd0706" title="ASCII of Phred-scaled base QUALity+33.">QUAL</a>;                                                   
<a name="l01619"></a><a class="code" href="classGenomicRegionSAM.html#a1622a8804462bad83b1c49d25eeacb49">01619</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#a1622a8804462bad83b1c49d25eeacb49" title="all SAM fields after the 12th mandatory field">OPTIONAL</a>;                                               
<a name="l01620"></a><a class="code" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92">01620</a>   <a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length)">CIGARTokens</a> <a class="code" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92" title="CIGAR tokens.">T</a>;                                                
<a name="l01621"></a>01621   
<a name="l01622"></a>01622 };
<a name="l01623"></a>01623 
<a name="l01624"></a>01624 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01625"></a>01625 <span class="comment">// END CLASS: GenomicRegionSAM                                                                 //</span>
<a name="l01626"></a>01626 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01627"></a>01627 
<a name="l01628"></a>01628 
<a name="l01629"></a>01629 
<a name="l01630"></a>01630 
<a name="l01631"></a>01631 
<a name="l01632"></a>01632 
<a name="l01633"></a>01633 
<a name="l01634"></a>01634 
<a name="l01635"></a>01635 
<a name="l01636"></a>01636 
<a name="l01637"></a>01637 
<a name="l01638"></a>01638 
<a name="l01639"></a>01639 
<a name="l01640"></a>01640 
<a name="l01641"></a>01641 
<a name="l01642"></a>01642 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01643"></a>01643 <span class="comment">// CLASS: GenomicRegionGFFToREG                                                                //</span>
<a name="l01644"></a>01644 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01646"></a>01646 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01647"></a><a class="code" href="classGenomicRegionGFFToREG.html">01647</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01648"></a>01648 {
<a name="l01649"></a>01649  <span class="keyword">public</span>:
<a name="l01651"></a>01651   <a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01652"></a>01652 
<a name="l01653"></a>01653 
<a name="l01654"></a>01654 
<a name="l01656"></a>01656   <a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01657"></a>01657 
<a name="l01658"></a>01658 
<a name="l01659"></a>01659 
<a name="l01661"></a>01661   ~<a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>();
<a name="l01662"></a>01662   
<a name="l01663"></a>01663 
<a name="l01664"></a>01664 
<a name="l01666"></a>01666   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01667"></a>01667   
<a name="l01668"></a>01668 
<a name="l01669"></a>01669 
<a name="l01670"></a>01670   <span class="comment">// data</span>
<a name="l01671"></a>01671 
<a name="l01672"></a>01672 
<a name="l01673"></a>01673 };
<a name="l01674"></a>01674 
<a name="l01675"></a>01675 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01676"></a>01676 <span class="comment">// END CLASS: GenomicRegionGFFToREG                                                            //</span>
<a name="l01677"></a>01677 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01678"></a>01678 
<a name="l01679"></a>01679 
<a name="l01680"></a>01680 
<a name="l01681"></a>01681 
<a name="l01682"></a>01682 
<a name="l01683"></a>01683 
<a name="l01684"></a>01684 
<a name="l01685"></a>01685 
<a name="l01686"></a>01686 
<a name="l01687"></a>01687 
<a name="l01688"></a>01688 
<a name="l01689"></a>01689 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01690"></a>01690 <span class="comment">// CLASS: GenomicRegionGFF                                                                     //</span>
<a name="l01691"></a>01691 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01693"></a>01693 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01694"></a><a class="code" href="classGenomicRegionGFF.html">01694</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01695"></a>01695 {
<a name="l01696"></a>01696  <span class="keyword">public</span>:
<a name="l01698"></a>01698   <a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01699"></a>01699 
<a name="l01700"></a>01700 
<a name="l01701"></a>01701 
<a name="l01703"></a>01703   <a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01704"></a>01704 
<a name="l01705"></a>01705 
<a name="l01706"></a>01706 
<a name="l01708"></a>01708   ~<a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>();
<a name="l01709"></a>01709   
<a name="l01710"></a>01710 
<a name="l01711"></a>01711 
<a name="l01712"></a>01712   <span class="comment">//------------------------------------------------------//</span>
<a name="l01713"></a>01713   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01714"></a>01714   <span class="comment">//------------------------------------------------------//</span>
<a name="l01715"></a>01715 
<a name="l01716"></a>01716 
<a name="l01718"></a>01718   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01719"></a>01719 
<a name="l01720"></a>01720 
<a name="l01721"></a>01721 
<a name="l01723"></a>01723   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01724"></a>01724 
<a name="l01725"></a>01725 
<a name="l01726"></a>01726 
<a name="l01728"></a>01728   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01729"></a>01729 
<a name="l01730"></a>01730 
<a name="l01731"></a>01731 
<a name="l01733"></a>01733   <span class="keywordtype">void</span> PrintModified(<span class="keyword">const</span> <span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01734"></a>01734 
<a name="l01735"></a>01735 
<a name="l01736"></a>01736 
<a name="l01738"></a>01738   <span class="keywordtype">void</span> Diff();
<a name="l01739"></a>01739 
<a name="l01740"></a>01740 
<a name="l01741"></a>01741 
<a name="l01743"></a>01743   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01744"></a>01744 
<a name="l01745"></a>01745 
<a name="l01746"></a>01746 
<a name="l01748"></a>01748   <span class="keywordtype">void</span> Divide();
<a name="l01749"></a>01749 
<a name="l01750"></a>01750 
<a name="l01751"></a>01751 
<a name="l01753"></a>01753   <span class="keywordtype">void</span> RunDivide();
<a name="l01754"></a>01754 
<a name="l01755"></a>01755 
<a name="l01756"></a>01756 
<a name="l01758"></a>01758   <span class="keywordtype">void</span> Intersect();
<a name="l01759"></a>01759 
<a name="l01760"></a>01760 
<a name="l01761"></a>01761 
<a name="l01763"></a>01763   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01764"></a>01764 
<a name="l01765"></a>01765 
<a name="l01766"></a>01766 
<a name="l01768"></a>01768   <span class="keywordtype">void</span> RunSplit(<span class="keywordtype">bool</span> by_chrom_and_strand); 
<a name="l01769"></a>01769 
<a name="l01770"></a>01770 
<a name="l01771"></a>01771   
<a name="l01773"></a>01773   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01774"></a>01774 
<a name="l01775"></a>01775 
<a name="l01776"></a>01776 
<a name="l01777"></a>01777   <span class="comment">//---------------------------------------//</span>
<a name="l01778"></a>01778   <span class="comment">//  Data                                 //</span>
<a name="l01779"></a>01779   <span class="comment">//---------------------------------------//</span>
<a name="l01780"></a>01780   
<a name="l01781"></a><a class="code" href="classGenomicRegionGFF.html#aaec963f555e7e807fb5046566622f5ee">01781</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionGFF.html#aaec963f555e7e807fb5046566622f5ee" title="number of TAB-separated fields in GFF entry (= 11)">n_tokens</a>;                    
<a name="l01782"></a>01782   <span class="comment">//char *SEQNAME;                              //!&lt; sequence name: mapped to GenomicRegion::I-&gt;front()-&gt;CHROMOSOME</span>
<a name="l01783"></a><a class="code" href="classGenomicRegionGFF.html#a2f0d4a5f15d8a756233436af03341734">01783</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a2f0d4a5f15d8a756233436af03341734" title="feature source">SOURCE</a>;                                 
<a name="l01784"></a><a class="code" href="classGenomicRegionGFF.html#a0de447fbfad3da5b7a9cf07a5a6980f0">01784</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a0de447fbfad3da5b7a9cf07a5a6980f0" title="feture type">FEATURE</a>;                                
<a name="l01785"></a>01785   <span class="comment">//long int START;                             //!&lt; 1-based start position: mapped to GenomicRegion::I-&gt;front()-&gt;START</span>
<a name="l01786"></a>01786   <span class="comment">//long int END;                               //!&lt; 1-based end position: mapped to GenomicRegion::I-&gt;front()-&gt;STOP</span>
<a name="l01787"></a><a class="code" href="classGenomicRegionGFF.html#a0e92da1f0fd153504e0630f27229ba4d">01787</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a0e92da1f0fd153504e0630f27229ba4d" title="score">SCORE</a>;                                  
<a name="l01788"></a>01788   <span class="comment">//char STRAND;                                //!&lt; strand: mapped to GenomicRegion::I-&gt;front()-&gt;STRAND</span>
<a name="l01789"></a><a class="code" href="classGenomicRegionGFF.html#a42122a69eec385e0106e3861cb3b98f3">01789</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionGFF.html#a42122a69eec385e0106e3861cb3b98f3" title="coding frame">FRAME</a>;                                   
<a name="l01790"></a>01790   <span class="comment">//char *ATTRIBUTE;                    //!&lt; attribute: mapped to GenomicRegion::LABEL if non-empty</span>
<a name="l01791"></a><a class="code" href="classGenomicRegionGFF.html#a625a023f1bb24fc597e723919a036073">01791</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a625a023f1bb24fc597e723919a036073" title="comment">COMMENT</a>;                                
<a name="l01792"></a>01792 
<a name="l01793"></a>01793 };
<a name="l01794"></a>01794 
<a name="l01795"></a>01795 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01796"></a>01796 <span class="comment">// END CLASS: GenomicRegionGFF                                                                 //</span>
<a name="l01797"></a>01797 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01798"></a>01798 
<a name="l01799"></a>01799 
<a name="l01800"></a>01800 
<a name="l01801"></a>01801 
<a name="l01802"></a>01802 
<a name="l01803"></a>01803 
<a name="l01804"></a>01804 
<a name="l01805"></a>01805 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01806"></a>01806 <span class="comment">// CLASS: GenomicRegionSet                                                                     //</span>
<a name="l01807"></a>01807 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01809"></a>01809 <span class="comment"></span>
<a name="l01822"></a>01822 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01823"></a><a class="code" href="classGenomicRegionSet.html">01823</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>
<a name="l01824"></a>01824 {
<a name="l01825"></a>01825  <span class="keyword">public</span>:
<a name="l01827"></a>01827   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>(<span class="keywordtype">char</span> *file, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size, <span class="keywordtype">bool</span> verbose, <span class="keywordtype">bool</span> load_in_memory, <span class="keywordtype">bool</span> hide_header); 
<a name="l01828"></a>01828 
<a name="l01829"></a>01829 
<a name="l01831"></a>01831   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>(FILE *file_ptr, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size, <span class="keywordtype">bool</span> verbose, <span class="keywordtype">bool</span> load_in_memory, <span class="keywordtype">bool</span> hide_header);
<a name="l01832"></a>01832 
<a name="l01833"></a>01833 
<a name="l01835"></a>01835   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>();
<a name="l01836"></a>01836 
<a name="l01837"></a>01837 
<a name="l01838"></a>01838   <span class="comment">//-------------------------------------//</span>
<a name="l01839"></a>01839   <span class="comment">//  Read &amp; Print                       //</span>
<a name="l01840"></a>01840   <span class="comment">//-------------------------------------//</span>
<a name="l01841"></a>01841 
<a name="l01843"></a>01843   <span class="keywordtype">void</span> ProcessFileHeader(<span class="keywordtype">bool</span> hide);
<a name="l01844"></a>01844 
<a name="l01845"></a>01845   
<a name="l01847"></a>01847   <span class="keywordtype">void</span> DetectFileFormat();
<a name="l01848"></a>01848 
<a name="l01849"></a>01849   
<a name="l01851"></a>01851   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *CreateGenomicRegion(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *buffer);
<a name="l01852"></a>01852 
<a name="l01853"></a>01853 
<a name="l01855"></a>01855   <span class="keywordtype">void</span> Init();
<a name="l01856"></a>01856 
<a name="l01857"></a>01857 
<a name="l01859"></a>01859   <span class="keywordtype">void</span> Reset();
<a name="l01860"></a>01860 
<a name="l01861"></a>01861 
<a name="l01863"></a>01863   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Get();
<a name="l01864"></a>01864 
<a name="l01865"></a>01865 
<a name="l01867"></a>01867 
<a name="l01870"></a>01870   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Next(<span class="keywordtype">bool</span> retain_current=<span class="keyword">false</span>);
<a name="l01871"></a>01871 
<a name="l01872"></a>01872 
<a name="l01874"></a>01874 
<a name="l01878"></a>01878   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Next(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">bool</span> retain_current);
<a name="l01879"></a>01879 
<a name="l01880"></a>01880 
<a name="l01882"></a>01882   <span class="keywordtype">void</span> Release(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l01883"></a>01883 
<a name="l01884"></a>01884 
<a name="l01886"></a>01886   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetRetain();
<a name="l01887"></a>01887 
<a name="l01888"></a>01888 
<a name="l01890"></a>01890   <span class="keywordtype">void</span> PrintError(<span class="keywordtype">string</span> error_msg);
<a name="l01891"></a>01891   
<a name="l01892"></a>01892 
<a name="l01893"></a>01893 
<a name="l01894"></a>01894   <span class="comment">//-------------------------------------//</span>
<a name="l01895"></a>01895   <span class="comment">//  File-based (vertical) operations   //</span>
<a name="l01896"></a>01896   <span class="comment">//-------------------------------------//</span>
<a name="l01897"></a>01897 
<a name="l01898"></a>01898 
<a name="l01900"></a>01900   <span class="keywordtype">void</span> RunGlobalCreateAnnotator(StringLIntMap *bounds, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> upstream_max_distance, <span class="keywordtype">long</span> <span class="keywordtype">int</span> upstream_min_distance);
<a name="l01901"></a>01901 
<a name="l01902"></a>01902 
<a name="l01904"></a>01904   <span class="keywordtype">void</span> RunGlobalCluster(<span class="keywordtype">bool</span> merge=<span class="keyword">false</span>);
<a name="l01905"></a>01905 
<a name="l01906"></a>01906 
<a name="l01908"></a>01908   <span class="keywordtype">void</span> RunGlobalCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01909"></a>01909 
<a name="l01910"></a>01910 
<a name="l01912"></a>01912   <span class="keywordtype">void</span> RunGlobalSort(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">int</span> bin_bits);
<a name="l01913"></a>01913 
<a name="l01914"></a>01914 
<a name="l01916"></a>01916   <span class="keywordtype">void</span> RunGlobalInvert(StringLIntMap *bounds);
<a name="l01917"></a>01917 
<a name="l01918"></a>01918 
<a name="l01920"></a>01920   <span class="keywordtype">void</span> RunGlobalLink(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> max_difference, <span class="keywordtype">char</span> *label_func);
<a name="l01921"></a>01921 
<a name="l01922"></a>01922 
<a name="l01924"></a>01924   <span class="keywordtype">void</span> RunGlobalPartition();
<a name="l01925"></a>01925 
<a name="l01926"></a>01926 
<a name="l01928"></a>01928   <span class="keywordtype">void</span> RunGlobalTest(<span class="keywordtype">bool</span> sorted_by_strand);
<a name="l01929"></a>01929 
<a name="l01930"></a>01930 
<a name="l01932"></a>01932   <span class="keywordtype">void</span> RunGlobalMerge(<span class="keywordtype">long</span> <span class="keywordtype">int</span> merge_win);
<a name="l01933"></a>01933 
<a name="l01934"></a>01934 
<a name="l01936"></a>01936   <span class="keywordtype">void</span> RunGlobalReverseOrder();
<a name="l01937"></a>01937 
<a name="l01938"></a>01938 
<a name="l01940"></a>01940   <span class="keywordtype">void</span> PrintDensities(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> &amp;Index, <span class="keywordtype">bool</span> report_exact_boundaries);
<a name="l01941"></a>01941 
<a name="l01942"></a>01942 
<a name="l01944"></a>01944   <span class="keywordtype">void</span> RunGlobalScan(StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01945"></a>01945 
<a name="l01946"></a>01946 
<a name="l01948"></a>01948 
<a name="l01958"></a>01958   <span class="keywordtype">void</span> RunGlobalScanCount(StringLIntMap *bounds, <span class="keywordtype">char</span> *ref_reg_file, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">char</span> preprocess, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">long</span> <span class="keywordtype">int</span> min_reads);
<a name="l01959"></a>01959 
<a name="l01960"></a>01960 
<a name="l01962"></a>01962   <span class="keywordtype">void</span> RunConvertToWIG(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">char</span> *options, <span class="keywordtype">long</span> <span class="keywordtype">int</span> span, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01963"></a>01963 
<a name="l01964"></a>01964 
<a name="l01965"></a>01965 
<a name="l01966"></a>01966 
<a name="l01967"></a>01967   <span class="comment">//---------------------------------------//</span>
<a name="l01968"></a>01968   <span class="comment">//  Line-based (horizontal) operations   //</span>
<a name="l01969"></a>01969   <span class="comment">//---------------------------------------//</span>
<a name="l01970"></a>01970 
<a name="l01972"></a>01972   <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01973"></a>01973 
<a name="l01974"></a>01974 
<a name="l01976"></a>01976   <span class="keywordtype">void</span> RunConvertToBED(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01977"></a>01977 
<a name="l01978"></a>01978 
<a name="l01980"></a>01980   <span class="keywordtype">void</span> RunBounds(StringLIntMap *bounds);
<a name="l01981"></a>01981 
<a name="l01982"></a>01982 
<a name="l01984"></a>01984   <span class="keywordtype">void</span> RunCenter();
<a name="l01985"></a>01985 
<a name="l01986"></a>01986 
<a name="l01988"></a>01988   <span class="keywordtype">void</span> RunModifyChromosomeNames(<span class="keywordtype">char</span> *chromosome_conversion_file);
<a name="l01989"></a>01989 
<a name="l01990"></a>01990 
<a name="l01992"></a>01992   <span class="keywordtype">void</span> RunConnect();
<a name="l01993"></a>01993 
<a name="l01994"></a>01994 
<a name="l01996"></a>01996   <span class="keywordtype">void</span> RunDiff();
<a name="l01997"></a>01997 
<a name="l01998"></a>01998 
<a name="l02000"></a>02000   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l02001"></a>02001 
<a name="l02002"></a>02002 
<a name="l02004"></a>02004   <span class="keywordtype">void</span> RunDivide();
<a name="l02005"></a>02005 
<a name="l02006"></a>02006 
<a name="l02008"></a>02008   <span class="keywordtype">void</span> RunFix();
<a name="l02009"></a>02009 
<a name="l02010"></a>02010 
<a name="l02011"></a>02011 
<a name="l02013"></a>02013   <span class="keywordtype">void</span> RunIntersection();
<a name="l02014"></a>02014 
<a name="l02015"></a>02015 
<a name="l02016"></a>02016 
<a name="l02018"></a>02018   <span class="keywordtype">void</span> RunSize();
<a name="l02019"></a>02019 
<a name="l02020"></a>02020 
<a name="l02021"></a>02021 
<a name="l02023"></a>02023 
<a name="l02027"></a>02027   <span class="keywordtype">void</span> RunModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l02028"></a>02028 
<a name="l02029"></a>02029 
<a name="l02030"></a>02030 
<a name="l02032"></a>02032   <span class="keywordtype">void</span> RunConvertToREG(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l02033"></a>02033 
<a name="l02034"></a>02034 
<a name="l02035"></a>02035 
<a name="l02037"></a>02037   <span class="keywordtype">void</span> RunRandomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l02038"></a>02038 
<a name="l02039"></a>02039 
<a name="l02040"></a>02040 
<a name="l02042"></a>02042   <span class="keywordtype">void</span> RunSelect(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l02043"></a>02043 
<a name="l02044"></a>02044 
<a name="l02045"></a>02045 
<a name="l02047"></a>02047 
<a name="l02052"></a>02052   <span class="keywordtype">void</span> RunShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l02053"></a>02053 
<a name="l02054"></a>02054 
<a name="l02056"></a>02056   <span class="keywordtype">void</span> RunSort(); 
<a name="l02057"></a>02057 
<a name="l02058"></a>02058   
<a name="l02060"></a>02060   <span class="keywordtype">void</span> RunSplit(<span class="keywordtype">bool</span> by_chrom_and_strand); 
<a name="l02061"></a>02061 
<a name="l02062"></a>02062   
<a name="l02064"></a>02064   <span class="keywordtype">void</span> RunModifyStrand(<span class="keywordtype">char</span> *strand_op, <span class="keywordtype">bool</span> sorted);
<a name="l02065"></a>02065 
<a name="l02066"></a>02066 
<a name="l02068"></a>02068   <span class="keywordtype">void</span> RunModifyStrandSorted(<span class="keywordtype">char</span> *strand_op);
<a name="l02069"></a>02069 
<a name="l02070"></a>02070 
<a name="l02072"></a>02072   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <span class="keywordtype">char</span> *ref_reg_file);
<a name="l02073"></a>02073 
<a name="l02074"></a>02074 
<a name="l02076"></a>02076   <span class="keywordtype">void</span> RunUnion();
<a name="l02077"></a>02077 
<a name="l02078"></a>02078 
<a name="l02080"></a>02080   <span class="keywordtype">void</span> RunSlidingWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l02081"></a>02081 
<a name="l02082"></a>02082 
<a name="l02084"></a>02084   <span class="keywordtype">void</span> RunExtractSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l02085"></a>02085 
<a name="l02086"></a>02086 
<a name="l02087"></a>02087 
<a name="l02088"></a>02088   <span class="comment">//---------------------------------------//</span>
<a name="l02089"></a>02089   <span class="comment">//  UNDER DEVELOPMENT/TESTING            //</span>
<a name="l02090"></a>02090   <span class="comment">//---------------------------------------//</span>
<a name="l02091"></a>02091 
<a name="l02092"></a>02092 
<a name="l02094"></a>02094   <span class="keywordtype">void</span> PrintBEDGraphFormat(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l02095"></a>02095 
<a name="l02096"></a>02096 
<a name="l02098"></a>02098   <span class="keywordtype">void</span> PrintSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02099"></a>02099 
<a name="l02100"></a>02100 
<a name="l02102"></a>02102   <span class="comment">//void PrintNoGaps();</span>
<a name="l02103"></a>02103 
<a name="l02104"></a>02104 
<a name="l02106"></a>02106   <span class="keywordtype">void</span> PrintRemoveN(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02107"></a>02107 
<a name="l02108"></a>02108 
<a name="l02110"></a>02110 
<a name="l02114"></a>02114   <span class="keywordtype">void</span> PrintOffsetFormat(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l02115"></a>02115 
<a name="l02116"></a>02116 
<a name="l02118"></a>02118   <span class="keywordtype">void</span> PrintReversePos(StringLIntMap *bounds);
<a name="l02119"></a>02119 
<a name="l02120"></a>02120 
<a name="l02122"></a>02122   <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l02123"></a>02123 
<a name="l02124"></a>02124 
<a name="l02126"></a>02126   <span class="keywordtype">void</span> PrintVerifySeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l02127"></a>02127 
<a name="l02128"></a>02128 
<a name="l02129"></a>02129 
<a name="l02130"></a>02130  <span class="keyword">private</span>:
<a name="l02132"></a>02132   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *StoreInTempFile();
<a name="l02133"></a>02133 
<a name="l02134"></a>02134 
<a name="l02136"></a>02136   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *PrintMergeSort(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *rtempSet, <span class="keywordtype">char</span> strand);
<a name="l02137"></a>02137 
<a name="l02138"></a>02138 
<a name="l02140"></a>02140   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *PrintReverse(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *rtempSet);
<a name="l02141"></a>02141 
<a name="l02142"></a>02142 
<a name="l02143"></a>02143 
<a name="l02144"></a>02144   <span class="comment">//---------------------------------------//</span>
<a name="l02145"></a>02145   <span class="comment">//  Data                                 //</span>
<a name="l02146"></a>02146   <span class="comment">//---------------------------------------//</span>
<a name="l02147"></a>02147 
<a name="l02148"></a>02148  <span class="keyword">public</span>:
<a name="l02149"></a><a class="code" href="classGenomicRegionSet.html#a7c593d1beab775590d5b810097644583">02149</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSet.html#a7c593d1beab775590d5b810097644583" title="file name (&#39;NULL&#39; if reading from standard input)">file</a>;                                           
<a name="l02150"></a><a class="code" href="classGenomicRegionSet.html#a4ca6c6268e39fdea2add6e6381c66962">02150</a>   FILE *<a class="code" href="classGenomicRegionSet.html#a4ca6c6268e39fdea2add6e6381c66962" title="file pointer (&#39;NULL&#39; if reading from standard input)">file_ptr</a>;                                       
<a name="l02151"></a><a class="code" href="classGenomicRegionSet.html#a4d60b5403d2d6aaeda541ed45625b245">02151</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#a4d60b5403d2d6aaeda541ed45625b245" title="line buffer size (automatically adjusted during execution to ensure the entire line is read)...">buffer_size</a>;        
<a name="l02152"></a><a class="code" href="classGenomicRegionSet.html#a2646e9e6d5da1693b4bea36a5acc082e">02152</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#a2646e9e6d5da1693b4bea36a5acc082e" title="verbose mode">verbose</a>;                                         
<a name="l02153"></a><a class="code" href="classGenomicRegionSet.html#abd0fc2542e717e5855e03d3b3892f027">02153</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#abd0fc2542e717e5855e03d3b3892f027" title="&#39;true&#39; if reading from standard input">from_stdin</a>;                                      
<a name="l02154"></a><a class="code" href="classGenomicRegionSet.html#ade43c6cd72b95da75395129b00a52687">02154</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#ade43c6cd72b95da75395129b00a52687" title="&#39;true&#39; if the entire region set is loaded in memory">load_in_memory</a>;                          
<a name="l02155"></a><a class="code" href="classGenomicRegionSet.html#a84dd2f8055524a933e963175781f7fd2">02155</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#a84dd2f8055524a933e963175781f7fd2" title="if &#39;true&#39;, then do not print file header">hide_header</a>;                                     
<a name="l02156"></a><a class="code" href="classGenomicRegionSet.html#aea8b5c9bae036da50d4e8720d2e13231">02156</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#aea8b5c9bae036da50d4e8720d2e13231" title="if &#39;true&#39;, use an array for storing genomic regions&#39; intervals (for time efficiency)">use_interval_array</a>;                      
<a name="l02157"></a><a class="code" href="classGenomicRegionSet.html#aa4494ed2ae3f7594c29b18c39fed6748">02157</a>   <span class="keywordtype">string</span> <a class="code" href="classGenomicRegionSet.html#aa4494ed2ae3f7594c29b18c39fed6748" title="input format">format</a>;                                        
<a name="l02158"></a><a class="code" href="classGenomicRegionSet.html#a08d9a35a7688ac2c193b53c9a8693c6e">02158</a>   <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> <a class="code" href="classGenomicRegionSet.html#a08d9a35a7688ac2c193b53c9a8693c6e" title="keeps track of progress of computation">progress</a>;                            
<a name="l02159"></a><a class="code" href="classGenomicRegionSet.html#adf3a5598e31ddbb5b381125925362085">02159</a>   <a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *<a class="code" href="classGenomicRegionSet.html#adf3a5598e31ddbb5b381125925362085" title="pointer to the buffer class used to read the region sets">buffer</a>;                           
<a name="l02160"></a><a class="code" href="classGenomicRegionSet.html#a5293292428aa06145b38be4b96ab05e4">02160</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#a5293292428aa06145b38be4b96ab05e4" title="number of regions (&#39;1&#39; if the regions are not loaded in memory)">n_regions</a>;                           
<a name="l02161"></a><a class="code" href="classGenomicRegionSet.html#a206d91b443832800afc04f786d15971d">02161</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> **<a class="code" href="classGenomicRegionSet.html#a206d91b443832800afc04f786d15971d" title="array where the regions are stored; if the region set is not entirely loaded in memory, only R[0] is used">R</a>;                            
<a name="l02162"></a><a class="code" href="classGenomicRegionSet.html#ad89185b88d5b37161cc65bc84b421c0e">02162</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#ad89185b88d5b37161cc65bc84b421c0e" title="index pointing to the current region in R">r_index</a>;                                     
<a name="l02163"></a>02163 };
<a name="l02164"></a>02164 
<a name="l02165"></a>02165 
<a name="l02166"></a>02166 
<a name="l02167"></a>02167 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02168"></a>02168 <span class="comment">// END CLASS: GenomicRegionSet                                                                 //</span>
<a name="l02169"></a>02169 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02170"></a>02170 
<a name="l02171"></a>02171 
<a name="l02172"></a>02172 
<a name="l02173"></a>02173 
<a name="l02174"></a>02174 
<a name="l02175"></a>02175 
<a name="l02176"></a>02176 
<a name="l02177"></a>02177 
<a name="l02178"></a>02178 
<a name="l02179"></a>02179 
<a name="l02180"></a>02180 
<a name="l02181"></a>02181 
<a name="l02182"></a>02182 
<a name="l02183"></a>02183 
<a name="l02184"></a>02184 
<a name="l02185"></a>02185 
<a name="l02186"></a>02186 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02187"></a>02187 <span class="comment">// CLASS: GenomicRegionSetScanner                                                              //</span>
<a name="l02188"></a>02188 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02190"></a>02190 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02191"></a><a class="code" href="classGenomicRegionSetScanner.html">02191</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetScanner.html" title="Abstract class for scanning a set of genomic regions by sliding windows.">GenomicRegionSetScanner</a>
<a name="l02192"></a>02192 {
<a name="l02193"></a>02193  <span class="keyword">public</span>:
<a name="l02195"></a>02195 
<a name="l02204"></a>02204   <a class="code" href="classGenomicRegionSetScanner.html" title="Abstract class for scanning a set of genomic regions by sliding windows.">GenomicRegionSetScanner</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *R, StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">char</span> preprocess);
<a name="l02205"></a>02205   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegionSetScanner.html" title="Abstract class for scanning a set of genomic regions by sliding windows.">GenomicRegionSetScanner</a>();
<a name="l02206"></a>02206 
<a name="l02207"></a>02207   <span class="comment">// operations</span>
<a name="l02208"></a>02208   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintInterval(FILE *out_file=stdout) = 0;        
<a name="l02209"></a>02209   <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next() = 0;                                                          
<a name="l02210"></a>02210   <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *Ref) = 0;                     
<a name="l02211"></a>02211   <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a> *index) = 0;      
<a name="l02212"></a>02212 
<a name="l02213"></a>02213   <span class="comment">// data</span>
<a name="l02214"></a>02214  <span class="keyword">public</span>:
<a name="l02215"></a><a class="code" href="classGenomicRegionSetScanner.html#af76bf4fef482886f23f58cfa12c3eea1">02215</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetScanner.html#af76bf4fef482886f23f58cfa12c3eea1" title="pointer to the GenomicRegionSet which will be scanned by sliding windows">R</a>;                          
<a name="l02216"></a><a class="code" href="classGenomicRegionSetScanner.html#a7f64551c26331cd4ead1ab01405f1b72">02216</a>   StringLIntMap *<a class="code" href="classGenomicRegionSetScanner.html#a7f64551c26331cd4ead1ab01405f1b72" title="chromosome sizes">bounds</a>;                        
<a name="l02217"></a><a class="code" href="classGenomicRegionSetScanner.html#ab278dfa27c3589865b9243156c1c727f">02217</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#ab278dfa27c3589865b9243156c1c727f" title="sliding window step">win_step</a>;                            
<a name="l02218"></a><a class="code" href="classGenomicRegionSetScanner.html#aeed625b2a12aa2f7900997c4e20cd9b5">02218</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#aeed625b2a12aa2f7900997c4e20cd9b5" title="sliding window size (must be a multiple of window step)">win_size</a>;                            
<a name="l02219"></a><a class="code" href="classGenomicRegionSetScanner.html#ad47603d08614ab993fe9036662c421e7">02219</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#ad47603d08614ab993fe9036662c421e7" title="if true, genomic region labels are assumed to be integers and are included in the counting...">use_labels_as_values</a>;            
<a name="l02220"></a><a class="code" href="classGenomicRegionSetScanner.html#a6e5e0f274eae714f0076404a20c4c736">02220</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#a6e5e0f274eae714f0076404a20c4c736" title="if true, strand information is ignored and no sliding windows on the negative strand are reported...">ignore_strand</a>;                           
<a name="l02221"></a><a class="code" href="classGenomicRegionSetScanner.html#af2a536e23023f96c57f51e2276b4be0e">02221</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#af2a536e23023f96c57f51e2276b4be0e" title="if &#39;1&#39;, only start position is counted; if &#39;p&#39;, all positions are counted; if &#39;c&#39;, center of interval is counted">preprocess</a>;                                      
<a name="l02222"></a><a class="code" href="classGenomicRegionSetScanner.html#abab8a5c92ec438168e22eb6e33098aba">02222</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#abab8a5c92ec438168e22eb6e33098aba" title="win_size divided by win_step (note: each window is comprised of non-overlapping sub-windows of size w...">n_win_combine</a>;                       
<a name="l02223"></a>02223 };
<a name="l02224"></a>02224 
<a name="l02225"></a>02225 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02226"></a>02226 <span class="comment">// END CLASS: GenomicRegionSetScanner                                                          //</span>
<a name="l02227"></a>02227 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02228"></a>02228 
<a name="l02229"></a>02229 
<a name="l02230"></a>02230 
<a name="l02231"></a>02231 
<a name="l02232"></a>02232 
<a name="l02233"></a>02233 
<a name="l02234"></a>02234 
<a name="l02235"></a>02235 
<a name="l02236"></a>02236 
<a name="l02237"></a>02237 
<a name="l02238"></a>02238 
<a name="l02239"></a>02239 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02240"></a>02240 <span class="comment">// CLASS: SortedGenomicRegionSetScanner                                                        //</span>
<a name="l02241"></a>02241 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02243"></a>02243 <span class="comment"></span>
<a name="l02267"></a>02267 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02268"></a><a class="code" href="classSortedGenomicRegionSetScanner.html">02268</a> <span class="keyword">class </span><a class="code" href="classSortedGenomicRegionSetScanner.html" title="This class is used to scan a set of sorted regions by sliding windows. See detailed description below...">SortedGenomicRegionSetScanner</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetScanner.html" title="Abstract class for scanning a set of genomic regions by sliding windows.">GenomicRegionSetScanner</a>
<a name="l02269"></a>02269 {
<a name="l02270"></a>02270  <span class="keyword">public</span>:
<a name="l02272"></a>02272 
<a name="l02281"></a>02281   <a class="code" href="classSortedGenomicRegionSetScanner.html" title="This class is used to scan a set of sorted regions by sliding windows. See detailed description below...">SortedGenomicRegionSetScanner</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *R, StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">char</span> preprocess);
<a name="l02282"></a>02282   <span class="keyword">virtual</span> ~<a class="code" href="classSortedGenomicRegionSetScanner.html" title="This class is used to scan a set of sorted regions by sliding windows. See detailed description below...">SortedGenomicRegionSetScanner</a>();
<a name="l02283"></a>02283   
<a name="l02284"></a>02284   <span class="comment">// operations</span>
<a name="l02285"></a>02285   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintInterval(FILE *out_file=stdout);            
<a name="l02286"></a>02286   <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next();                                                                      
<a name="l02287"></a>02287   <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *Ref);                         
<a name="l02288"></a>02288   <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a> *index);          
<a name="l02289"></a>02289   
<a name="l02290"></a>02290   <span class="comment">// data</span>
<a name="l02291"></a>02291  <span class="keyword">private</span>: 
<a name="l02292"></a>02292   <span class="keywordtype">bool</span> Test();                                  
<a name="l02293"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#a1bc1e61a86173ef65cb2cdbb92500e5f">02293</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classSortedGenomicRegionSetScanner.html#a1bc1e61a86173ef65cb2cdbb92500e5f" title="pointer to the most recently scanned region">r</a>;                             
<a name="l02294"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#a2d03ec2a63f2efde03045243fd1b845a">02294</a>   StringLIntMap::iterator <a class="code" href="classSortedGenomicRegionSetScanner.html#a2d03ec2a63f2efde03045243fd1b845a" title="keeps track of current window&#39;s chromosome information">chr</a>;  
<a name="l02295"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#a4dabc5bab629e73c7ad67e8b721c3fe2">02295</a>   <span class="keywordtype">char</span> <a class="code" href="classSortedGenomicRegionSetScanner.html#a4dabc5bab629e73c7ad67e8b721c3fe2" title="keeps track of current window&#39;s strand information">strand</a>;                                  
<a name="l02296"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#adea7c2663132d48d57d69a0678240fa3">02296</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classSortedGenomicRegionSetScanner.html#adea7c2663132d48d57d69a0678240fa3" title="keeps track of current window&#39;s start information">start</a>;                               
<a name="l02297"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#a45e693afdaec2e13d2d180bcea6bec95">02297</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classSortedGenomicRegionSetScanner.html#a45e693afdaec2e13d2d180bcea6bec95" title="keeps track of current window&#39;s stop information">stop</a>;                                
<a name="l02298"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#a7abc4c7fa14c402b04a249046f4ce7da">02298</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classSortedGenomicRegionSetScanner.html#a7abc4c7fa14c402b04a249046f4ce7da" title="a vector that keeps track of values in the sub-windows (of size win_step) that comprise the window (o...">v</a>;                                  
<a name="l02299"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#ab41bd3644ee03a42781be3a0d6185911">02299</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classSortedGenomicRegionSetScanner.html#ab41bd3644ee03a42781be3a0d6185911" title="the k-th element of v to be updated next (in a round-robin manner)">k</a>;                                   
<a name="l02300"></a><a class="code" href="classSortedGenomicRegionSetScanner.html#af6c4fe6e94f26074c8fa919cebe3f652">02300</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classSortedGenomicRegionSetScanner.html#af6c4fe6e94f26074c8fa919cebe3f652" title="keeps track of current window&#39;s value information (this is the sum of elements in vector v); it is se...">v_sum</a>;                               
<a name="l02301"></a>02301 };
<a name="l02302"></a>02302 
<a name="l02303"></a>02303 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02304"></a>02304 <span class="comment">// END CLASS: SortedGenomicRegionSetScanner                                                    //</span>
<a name="l02305"></a>02305 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02306"></a>02306 
<a name="l02307"></a>02307 
<a name="l02308"></a>02308 
<a name="l02309"></a>02309 
<a name="l02310"></a>02310 
<a name="l02311"></a>02311 
<a name="l02312"></a>02312 
<a name="l02313"></a>02313 
<a name="l02314"></a>02314 
<a name="l02315"></a>02315 
<a name="l02316"></a>02316 
<a name="l02317"></a>02317 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02318"></a>02318 <span class="comment">// CLASS: UnsortedGenomicRegionSetScanner                                                      //</span>
<a name="l02319"></a>02319 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02321"></a>02321 <span class="comment"></span>
<a name="l02322"></a>02322 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02323"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html">02323</a> <span class="keyword">class </span><a class="code" href="classUnsortedGenomicRegionSetScanner.html" title="This class is used to scan a set of unsorted regions by sliding windows.">UnsortedGenomicRegionSetScanner</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetScanner.html" title="Abstract class for scanning a set of genomic regions by sliding windows.">GenomicRegionSetScanner</a>
<a name="l02324"></a>02324 {
<a name="l02325"></a>02325  <span class="keyword">public</span>:
<a name="l02326"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html#aae780b6a128d3c6a634273c65478565f">02326</a>         <span class="keyword">typedef</span> map&lt;string,unsigned long int**&gt; <a class="code" href="classUnsortedGenomicRegionSetScanner.html#aae780b6a128d3c6a634273c65478565f" title="a map from chromosome name to sliding window counts">CountMap</a>;               
<a name="l02327"></a>02327 
<a name="l02329"></a>02329 
<a name="l02338"></a>02338         <a class="code" href="classUnsortedGenomicRegionSetScanner.html" title="This class is used to scan a set of unsorted regions by sliding windows.">UnsortedGenomicRegionSetScanner</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *R, StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">char</span> preprocess);
<a name="l02339"></a>02339         <span class="keyword">virtual</span> ~<a class="code" href="classUnsortedGenomicRegionSetScanner.html" title="This class is used to scan a set of unsorted regions by sliding windows.">UnsortedGenomicRegionSetScanner</a>();
<a name="l02340"></a>02340 
<a name="l02341"></a>02341         <span class="comment">// operations</span>
<a name="l02342"></a>02342         <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintInterval(FILE *out_file=stdout);                      
<a name="l02343"></a>02343         <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next();                                                                        
<a name="l02344"></a>02344         <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *Ref);                           
<a name="l02345"></a>02345         <span class="keyword">virtual</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a> *index);            
<a name="l02346"></a>02346 
<a name="l02347"></a>02347  <span class="keyword">private</span>:
<a name="l02348"></a>02348         <span class="keywordtype">void</span> Init();                                                    
<a name="l02349"></a>02349 
<a name="l02350"></a>02350         <span class="comment">// data</span>
<a name="l02351"></a>02351  <span class="keyword">public</span>:
<a name="l02352"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html#aa3f47941bb72c74bffe78e18a09f522f">02352</a>         <a class="code" href="classUnsortedGenomicRegionSetScanner.html#aae780b6a128d3c6a634273c65478565f" title="a map from chromosome name to sliding window counts">CountMap</a> <a class="code" href="classUnsortedGenomicRegionSetScanner.html#aa3f47941bb72c74bffe78e18a09f522f" title="keeps track of read counts per chromosome per sliding window (forward &amp; reverse strand)">count</a>;                                                 
<a name="l02353"></a>02353  <span class="keyword">private</span>:
<a name="l02354"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html#aae982064a41baacc9869ed5cce63015e">02354</a>         CountMap::iterator <a class="code" href="classUnsortedGenomicRegionSetScanner.html#aae982064a41baacc9869ed5cce63015e" title="current window&#39;s chromosome information">current_chr</a>;                 
<a name="l02355"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html#a5543f3adff3f4eefc4897545fec17db6">02355</a>         <span class="keywordtype">char</span> <a class="code" href="classUnsortedGenomicRegionSetScanner.html#a5543f3adff3f4eefc4897545fec17db6" title="current window&#39;s strand information">current_strand</a>;                                    
<a name="l02356"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html#a47b0d1dc4cea90fb85a86f7193f01d96">02356</a>         <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classUnsortedGenomicRegionSetScanner.html#a47b0d1dc4cea90fb85a86f7193f01d96" title="current chromosome count vector (for forward or reverse strand)">current_v</a>;                   
<a name="l02357"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html#a958adb3fe36090683ec20d7491b50b88">02357</a>         <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classUnsortedGenomicRegionSetScanner.html#a958adb3fe36090683ec20d7491b50b88" title="number of windows in current window&#39;s chromosome count vector">current_n</a>;                    
<a name="l02358"></a><a class="code" href="classUnsortedGenomicRegionSetScanner.html#adaca86c33f9a8f75e62b59378ccdd163">02358</a>         <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classUnsortedGenomicRegionSetScanner.html#adaca86c33f9a8f75e62b59378ccdd163" title="current window&#39;s start information">current_win</a>;                  
<a name="l02359"></a>02359 
<a name="l02360"></a>02360 };
<a name="l02361"></a>02361 
<a name="l02362"></a>02362 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02363"></a>02363 <span class="comment">// END CLASS: UnsortedGenomicRegionSetScanner                                                  //</span>
<a name="l02364"></a>02364 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02365"></a>02365 
<a name="l02366"></a>02366 
<a name="l02367"></a>02367 
<a name="l02368"></a>02368 
<a name="l02369"></a>02369 
<a name="l02370"></a>02370 
<a name="l02371"></a>02371 
<a name="l02372"></a>02372 
<a name="l02373"></a>02373 
<a name="l02374"></a>02374 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02375"></a>02375 <span class="comment">// CLASS: GenomicRegionSetOverlaps                                                             //</span>
<a name="l02376"></a>02376 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02378"></a>02378 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02379"></a><a class="code" href="classGenomicRegionSetOverlaps.html">02379</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02380"></a>02380 {
<a name="l02381"></a>02381  <span class="keyword">public</span>:
<a name="l02383"></a>02383 
<a name="l02387"></a>02387   <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet);
<a name="l02388"></a>02388 
<a name="l02389"></a>02389 
<a name="l02391"></a>02391   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>();
<a name="l02392"></a>02392 
<a name="l02393"></a>02393   
<a name="l02395"></a>02395   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery() = 0;
<a name="l02396"></a>02396 
<a name="l02397"></a>02397 
<a name="l02399"></a>02399   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery() = 0;
<a name="l02400"></a>02400 
<a name="l02401"></a>02401 
<a name="l02403"></a>02403   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch() = 0;
<a name="l02404"></a>02404 
<a name="l02405"></a>02405  
<a name="l02407"></a>02407   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch() = 0;
<a name="l02408"></a>02408 
<a name="l02409"></a>02409  
<a name="l02411"></a>02411   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> Done() = 0;
<a name="l02412"></a>02412  
<a name="l02413"></a>02413 
<a name="l02415"></a>02415 
<a name="l02419"></a>02419   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetOverlap(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02420"></a>02420 
<a name="l02421"></a>02421 
<a name="l02423"></a>02423 
<a name="l02427"></a>02427   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextOverlap(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02428"></a>02428 
<a name="l02429"></a>02429   
<a name="l02431"></a>02431 
<a name="l02436"></a>02436   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcQueryCoverage(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02437"></a>02437 
<a name="l02438"></a>02438 
<a name="l02440"></a>02440 
<a name="l02445"></a>02445   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *CalcIndexCoverage(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02446"></a>02446 
<a name="l02447"></a>02447 
<a name="l02449"></a>02449 
<a name="l02454"></a>02454   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CountQueryOverlaps(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02455"></a>02455 
<a name="l02456"></a>02456 
<a name="l02458"></a>02458 
<a name="l02463"></a>02463   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *CountIndexOverlaps(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02464"></a>02464 
<a name="l02465"></a>02465  
<a name="l02466"></a>02466   
<a name="l02467"></a>02467   <span class="comment">// data</span>
<a name="l02468"></a>02468  <span class="keyword">public</span>: 
<a name="l02469"></a><a class="code" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2">02469</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2" title="pointer to query region set">QuerySet</a>;                           
<a name="l02470"></a><a class="code" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03">02470</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03" title="pointer to index region set">IndexSet</a>;                           
<a name="l02471"></a><a class="code" href="classGenomicRegionSetOverlaps.html#a31d1dbc3407c616a0bda0f42af7acbae">02471</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#a31d1dbc3407c616a0bda0f42af7acbae" title="pointer to the current query region">current_qreg</a>;                          
<a name="l02472"></a><a class="code" href="classGenomicRegionSetOverlaps.html#a537336b7c40b20e6f5113395c11614b6">02472</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#a537336b7c40b20e6f5113395c11614b6" title="pointer to the current index region">current_ireg</a>;                          
<a name="l02473"></a>02473    
<a name="l02474"></a>02474 };
<a name="l02475"></a>02475 
<a name="l02476"></a>02476 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02477"></a>02477 <span class="comment">// END CLASS: GenomicRegionSetOverlaps                                                         //</span>
<a name="l02478"></a>02478 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02479"></a>02479 
<a name="l02480"></a>02480 
<a name="l02481"></a>02481 
<a name="l02482"></a>02482 
<a name="l02483"></a>02483 
<a name="l02484"></a>02484 
<a name="l02485"></a>02485 
<a name="l02486"></a>02486 
<a name="l02487"></a>02487 
<a name="l02488"></a>02488 
<a name="l02489"></a>02489 
<a name="l02490"></a>02490 <span class="comment">//------------------------------------------------------------------------------------------------------------//</span>
<a name="l02491"></a>02491 <span class="comment">// CLASS: GenomicRegionSetIndex                                                                               //</span>
<a name="l02492"></a>02492 <span class="comment">//------------------------------------------------------------------------------------------------------------//</span>
<a name="l02494"></a>02494 <span class="comment"></span><span class="comment">//------------------------------------------------------------------------------------------------------------//</span>
<a name="l02495"></a><a class="code" href="classGenomicRegionSetIndex.html">02495</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a> 
<a name="l02496"></a>02496 {
<a name="l02497"></a>02497  <span class="keyword">public</span>:
<a name="l02499"></a>02499 
<a name="l02503"></a>02503   <a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *regSet, <span class="keyword">const</span> <span class="keywordtype">char</span> *bin_bits=NULL);
<a name="l02504"></a>02504 
<a name="l02505"></a>02505 
<a name="l02507"></a>02507   ~<a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a>();
<a name="l02508"></a>02508 
<a name="l02509"></a>02509   
<a name="l02511"></a>02511 
<a name="l02514"></a>02514   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *i);
<a name="l02515"></a>02515 
<a name="l02516"></a>02516  
<a name="l02518"></a>02518   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02519"></a>02519 
<a name="l02520"></a>02520  
<a name="l02522"></a>02522 
<a name="l02527"></a>02527   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetOverlap(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *i, <span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02528"></a>02528 
<a name="l02529"></a>02529 
<a name="l02531"></a>02531 
<a name="l02535"></a>02535   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextOverlap(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02536"></a>02536 
<a name="l02537"></a>02537 
<a name="l02538"></a>02538   <span class="comment">// data</span>
<a name="l02539"></a>02539  <span class="keyword">public</span>: 
<a name="l02540"></a><a class="code" href="classGenomicRegionSetIndex.html#a69385a469c77555e783aa2801c7608cf">02540</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetIndex.html#a69385a469c77555e783aa2801c7608cf" title="pointer to index region set">regSet</a>;                                                     
<a name="l02541"></a><a class="code" href="classGenomicRegionSetIndex.html#ae9d898b579f6b1b37a466f13ce9b1524">02541</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetIndex.html#ae9d898b579f6b1b37a466f13ce9b1524" title="pointer to the current index region">current_ireg</a>;                                          
<a name="l02542"></a><a class="code" href="classGenomicRegionSetIndex.html#a563202d9d5ee7e8b70fc510cdf02a35b">02542</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *<a class="code" href="classGenomicRegionSetIndex.html#a563202d9d5ee7e8b70fc510cdf02a35b" title="pointer to the current query interval">current_qint</a>;                                        
<a name="l02543"></a><a class="code" href="classGenomicRegionSetIndex.html#a6f2134bcaf8718d2fb19444cd9c2dfa8">02543</a>   <span class="keyword">typedef</span> pair&lt;long int*,long int**&gt; <a class="code" href="classGenomicRegionSetIndex.html#a6f2134bcaf8718d2fb19444cd9c2dfa8" title="a two-dimensional grid of bins">BinSet</a>;            
<a name="l02544"></a><a class="code" href="classGenomicRegionSetIndex.html#a5c147e94bc5208615ceaee6bf65ddd99">02544</a>   <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetIndex.html#a5c147e94bc5208615ceaee6bf65ddd99" title="number of levels in the two-dimensional grid of bins">n_levels</a>;                                                                         
<a name="l02545"></a><a class="code" href="classGenomicRegionSetIndex.html#ab458e36338671d25a4add20f3265473e">02545</a>   <span class="keywordtype">int</span> *<a class="code" href="classGenomicRegionSetIndex.html#ab458e36338671d25a4add20f3265473e" title="number of bits (per level) by which the start position is right-shifted to compute its bin...">n_bits</a>;                                                                          
<a name="l02546"></a><a class="code" href="classGenomicRegionSetIndex.html#a5876efc8c7ca70685413752a9aabd651">02546</a>   map&lt;string,BinSet*&gt; <a class="code" href="classGenomicRegionSetIndex.html#a5876efc8c7ca70685413752a9aabd651" title="the two-dimensional grid of bins per chromosome">index</a>;                                            
<a name="l02547"></a><a class="code" href="classGenomicRegionSetIndex.html#a70be632459700206f8d5ea74fe913d87">02547</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classGenomicRegionSetIndex.html#a70be632459700206f8d5ea74fe913d87" title="for each region, store a pointer to the &quot;next&quot; one in the same bin">r_next</a>;                                                                     
<a name="l02548"></a><a class="code" href="classGenomicRegionSetIndex.html#a55bdee4bb523a127ef744c9fc9eb53f3">02548</a>   <a class="code" href="classGenomicRegionSetIndex.html#a6f2134bcaf8718d2fb19444cd9c2dfa8" title="a two-dimensional grid of bins">BinSet</a> *<a class="code" href="classGenomicRegionSetIndex.html#a55bdee4bb523a127ef744c9fc9eb53f3" title="pointer to the current chromosome set of bins">current_binset</a>;                                                       
<a name="l02549"></a><a class="code" href="classGenomicRegionSetIndex.html#adc1e41cc8005f46f07db2bdbf86b86e9">02549</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetIndex.html#adc1e41cc8005f46f07db2bdbf86b86e9" title="&#39;true&#39;, if a new query is about to be processed">new_query</a>;                                                                       
<a name="l02550"></a>02550 };
<a name="l02551"></a>02551 
<a name="l02552"></a>02552 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02553"></a>02553 <span class="comment">// END CLASS: GenomicRegionSetIndex                                                            //</span>
<a name="l02554"></a>02554 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02555"></a>02555 
<a name="l02556"></a>02556 
<a name="l02557"></a>02557 
<a name="l02558"></a>02558 
<a name="l02559"></a>02559 
<a name="l02560"></a>02560 
<a name="l02561"></a>02561 
<a name="l02562"></a>02562 
<a name="l02563"></a>02563 
<a name="l02564"></a>02564 
<a name="l02565"></a>02565 
<a name="l02566"></a>02566 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02567"></a>02567 <span class="comment">// CLASS: UnsortedGenomicRegionSetOverlaps                                                     //</span>
<a name="l02568"></a>02568 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02570"></a>02570 <span class="comment"></span>
<a name="l02598"></a>02598 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02599"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html">02599</a> <span class="keyword">class </span><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02600"></a>02600 {
<a name="l02601"></a>02601  <span class="keyword">public</span>:
<a name="l02603"></a>02603 
<a name="l02608"></a>02608   <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet, <span class="keyword">const</span> <span class="keywordtype">char</span> *bin_bits=NULL);
<a name="l02609"></a>02609 
<a name="l02610"></a>02610 
<a name="l02612"></a>02612   ~<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a>();
<a name="l02613"></a>02613 
<a name="l02614"></a>02614   
<a name="l02616"></a>02616   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery();
<a name="l02617"></a>02617 
<a name="l02618"></a>02618 
<a name="l02620"></a>02620   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery();
<a name="l02621"></a>02621 
<a name="l02622"></a>02622 
<a name="l02624"></a>02624   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch();
<a name="l02625"></a>02625 
<a name="l02626"></a>02626  
<a name="l02628"></a>02628   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02629"></a>02629 
<a name="l02630"></a>02630  
<a name="l02632"></a>02632   <span class="keywordtype">bool</span> Done();
<a name="l02633"></a>02633  
<a name="l02634"></a>02634 
<a name="l02635"></a>02635   
<a name="l02636"></a>02636   <span class="comment">// data</span>
<a name="l02637"></a>02637  <span class="keyword">public</span>: 
<a name="l02638"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14">02638</a>   <span class="keyword">typedef</span> pair&lt;long int*,long int**&gt; <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14" title="a two-dimensional grid of bins">BinSet</a>;            
<a name="l02639"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a251e09b7f0a13a56b902c30791228ce4">02639</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a251e09b7f0a13a56b902c30791228ce4" title="for each region, store a pointer to the &quot;next&quot; one in the same bin">r_next</a>;                                                                     
<a name="l02640"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a1cc721de471ed40dcc7d8eae258e5833">02640</a>   <span class="keywordtype">int</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a1cc721de471ed40dcc7d8eae258e5833" title="number of levels in the two-dimensional grid of bins">n_levels</a>;                                                                         
<a name="l02641"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a76ecebfcf839a940bdae26069ceef748">02641</a>   <span class="keywordtype">int</span> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a76ecebfcf839a940bdae26069ceef748" title="number of bits (per level) by which the start position is right-shifted to compute its bin...">n_bits</a>;                                                                          
<a name="l02642"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abeeb94562abdf4aebc36684959a5bc78">02642</a>   map&lt;string,BinSet*&gt; <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abeeb94562abdf4aebc36684959a5bc78" title="the two-dimensional grid of bins per chromosome">index</a>;                                            
<a name="l02643"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a6bbb87f3c66d04527f06a5761df1ab56">02643</a>   <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14" title="a two-dimensional grid of bins">BinSet</a> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a6bbb87f3c66d04527f06a5761df1ab56" title="pointer to the current chromosome set of bins">current_binset</a>;                                                       
<a name="l02644"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abafc544ce2c677ce071fa4e30ec2ecaf">02644</a>   <span class="keywordtype">bool</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abafc544ce2c677ce071fa4e30ec2ecaf" title="&#39;true&#39;, if a new query is about to be processed">new_query</a>;                                                                       
<a name="l02645"></a>02645 };
<a name="l02646"></a>02646 
<a name="l02647"></a>02647 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02648"></a>02648 <span class="comment">// END CLASS: UnsortedGenomicRegionSetOverlaps                                                 //</span>
<a name="l02649"></a>02649 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02650"></a>02650 
<a name="l02651"></a>02651 
<a name="l02652"></a>02652 
<a name="l02653"></a>02653 
<a name="l02654"></a>02654 
<a name="l02655"></a>02655 
<a name="l02656"></a>02656 
<a name="l02657"></a>02657 
<a name="l02658"></a>02658 
<a name="l02659"></a>02659 
<a name="l02660"></a>02660 
<a name="l02661"></a>02661 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02662"></a>02662 <span class="comment">// CLASS: SortedGenomicRegionSetOverlaps                                                       //</span>
<a name="l02663"></a>02663 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02665"></a>02665 <span class="comment"></span>
<a name="l02724"></a>02724 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02725"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html">02725</a> <span class="keyword">class </span><a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02726"></a>02726 {
<a name="l02727"></a>02727  <span class="keyword">public</span>:
<a name="l02729"></a>02729 
<a name="l02734"></a>02734   <a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l02735"></a>02735 
<a name="l02736"></a>02736 
<a name="l02738"></a>02738   ~<a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a>();
<a name="l02739"></a>02739 
<a name="l02740"></a>02740   
<a name="l02742"></a>02742   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery();
<a name="l02743"></a>02743 
<a name="l02744"></a>02744 
<a name="l02746"></a>02746   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery();
<a name="l02747"></a>02747 
<a name="l02748"></a>02748 
<a name="l02750"></a>02750   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch();
<a name="l02751"></a>02751 
<a name="l02752"></a>02752  
<a name="l02754"></a>02754   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02755"></a>02755 
<a name="l02756"></a>02756  
<a name="l02758"></a>02758   <span class="keywordtype">bool</span> Done();
<a name="l02759"></a>02759  
<a name="l02760"></a>02760 
<a name="l02761"></a>02761  <span class="keyword">private</span>:
<a name="l02762"></a>02762   <span class="keywordtype">void</span> LoadIndexBuffer();                               
<a name="l02763"></a>02763   <span class="keywordtype">void</span> ClearIndexBuffer();                              
<a name="l02764"></a>02764 
<a name="l02765"></a>02765   <span class="comment">// data</span>
<a name="l02766"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3">02766</a>   <span class="keywordtype">bool</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3" title="if true, query and index regions sorted by strand">sorted_by_strand</a>;                                
<a name="l02767"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">02767</a>   GenomicRegionList <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f" title="temporary buffer containing index regions that overlap with current query; the buffer is cleared only...">IRegBuffer</a>;                         
<a name="l02768"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a71d6eea0c4bf9f19be6cf0c0c562f297">02768</a>   GenomicRegionList::iterator <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a71d6eea0c4bf9f19be6cf0c0c562f297" title="iterator on SortedGenomicRegionSetOverlaps::IRegBuffer">IRegBufferIterator</a>;       
<a name="l02769"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a962f8e6ce8c15703cfefa574a07ce42c">02769</a>   <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *<a class="code" href="classSortedGenomicRegionSetOverlaps.html#a962f8e6ce8c15703cfefa574a07ce42c" title="keeps tracks of the (linked) interval stored in SortedGenomicRegionSetOverlaps::IRegBuffer">ireg_buffer_interval</a>;                
<a name="l02770"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a349d589ff006f742d635b16987fe7b6b">02770</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a349d589ff006f742d635b16987fe7b6b" title="keeps track of the maximum size of SortedGenomicRegionSetOverlaps::IRegBuffer">max_ireg_buffer_size</a>;               
<a name="l02771"></a>02771 };
<a name="l02772"></a>02772 
<a name="l02773"></a>02773 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02774"></a>02774 <span class="comment">// END CLASS: SortedGenomicRegionSetOverlaps                                                   //</span>
<a name="l02775"></a>02775 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02776"></a>02776 
<a name="l02777"></a>02777 
<a name="l02778"></a>02778 
<a name="l02779"></a>02779 
<a name="l02780"></a>02780 
<a name="l02781"></a>02781 
<a name="l02782"></a>02782 
<a name="l02783"></a>02783 
<a name="l02784"></a>02784 
<a name="l02785"></a>02785 
<a name="l02786"></a>02786 <span class="keywordtype">char</span> ProcessStrand(<span class="keywordtype">char</span> *strand);
<a name="l02787"></a>02787 <span class="keywordtype">void</span> PrintChromosome(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">bool</span> convert);
<a name="l02788"></a>02788 <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *RegCenter(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l02789"></a>02789 <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetStart(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> preprocess);
<a name="l02790"></a>02790 
<a name="l02791"></a>02791 StringLIntMap *ReadBounds(<span class="keywordtype">char</span> *genome_reg_file, <span class="keywordtype">bool</span> verbose=<span class="keyword">false</span>);           
<a name="l02792"></a>02792 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcBoundSize(StringLIntMap *bounds);
<a name="l02793"></a>02793 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcRegSize(<span class="keywordtype">char</span> *reg_file);
<a name="l02794"></a>02794 <span class="keywordtype">bool</span> CompareBinnedGenomicRegions(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *q);
<a name="l02795"></a>02795 <span class="keywordtype">bool</span> CompareGenomicIntervals(<a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *I, <a class="code" href="classGenomicInterval.html" title="This class implements the genomic interval.">GenomicInterval</a> *J);
<a name="l02796"></a>02796 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsList *L);
<a name="l02797"></a>02797 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsVector *V);
<a name="l02798"></a>02798 <span class="keywordtype">void</span> SortGenomicIntervals(<a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a> *A);
<a name="l02799"></a>02799 
<a name="l02800"></a>02800 GenomicRegionSetBins *BinGenomicRegions(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *rset, <span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">int</span> bin_bits);
<a name="l02801"></a>02801 
<a name="l02802"></a>02802 <span class="keywordtype">long</span> <span class="keywordtype">int</span> *GetGapSizes(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> *offset_op);
<a name="l02803"></a>02803 OffsetList *CalcOffsetsWithoutGaps(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *query_reg, <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *ref_reg, <span class="keywordtype">char</span> *offset_op, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02804"></a>02804 
<a name="l02805"></a>02805 <span class="keywordtype">long</span> <span class="keywordtype">int</span> CountGenomicRegions(<span class="keywordtype">char</span> *reg_file, <span class="keywordtype">bool</span> use_label_counts);
<a name="l02806"></a>02806 
<a name="l02807"></a>02807 <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *CreateGenomicRegionSetAnnotator(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *RefRegSet, StringLIntMap *bounds, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> upstream_max_distance, <span class="keywordtype">long</span> <span class="keywordtype">int</span> upstream_min_distance, <span class="keywordtype">char</span> *bin_bits);
<a name="l02808"></a>02808 
</pre></div></div>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Tue Oct 9 2012 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
