<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>genomic_intervals.h Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
<div class="header">
  <div class="headertitle">
<div class="title">genomic_intervals.h</div>  </div>
</div>
<div class="contents">
<div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">//</span>
<a name="l00002"></a>00002 <span class="comment">// Copyright (c) 2011 IBM Corporation. </span>
<a name="l00003"></a>00003 <span class="comment">// All rights reserved. This program and the accompanying materials are made available under the terms of the GNU General Public License v2.0 </span>
<a name="l00004"></a>00004 <span class="comment">// which accompanies this distribution, and is available at http://www.gnu.org/licenses/old-licenses/gpl-2.0.html</span>
<a name="l00005"></a>00005 <span class="comment">//</span>
<a name="l00006"></a>00006 
<a name="l00007"></a>00007 <span class="keyword">const</span> <span class="keywordtype">string</span> VERSION = <span class="stringliteral">&quot;genomic_tools 2.3.0b&quot;</span>;
<a name="l00008"></a>00008 
<a name="l00009"></a>00009 
<a name="l00010"></a>00010 <span class="preprocessor">#include &lt;stdio.h&gt;</span>
<a name="l00011"></a>00011 <span class="preprocessor">#include &lt;stdlib.h&gt;</span>
<a name="l00012"></a>00012 <span class="preprocessor">#include &lt;string.h&gt;</span>
<a name="l00013"></a>00013 <span class="preprocessor">#include &lt;math.h&gt;</span>
<a name="l00014"></a>00014 <span class="preprocessor">#include &lt;unistd.h&gt;</span>
<a name="l00015"></a>00015 <span class="preprocessor">#include &lt;ctype.h&gt;</span>
<a name="l00016"></a>00016 <span class="preprocessor">#include &lt;sys/types.h&gt;</span>
<a name="l00017"></a>00017 <span class="preprocessor">#include &lt;sys/stat.h&gt;</span>
<a name="l00018"></a>00018 <span class="preprocessor">#include &lt;map&gt;</span>
<a name="l00019"></a>00019 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00020"></a>00020 <span class="preprocessor">#include &lt;list&gt;</span>
<a name="l00021"></a>00021 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00022"></a>00022 <span class="preprocessor">#include &lt;iostream&gt;</span>
<a name="l00023"></a>00023 <span class="preprocessor">#include &lt;algorithm&gt;</span>
<a name="l00024"></a>00024 
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 <span class="keyword">using namespace </span>std;
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <span class="comment">// forward declarations</span>
<a name="l00029"></a>00029 <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>;
<a name="l00030"></a>00030 <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>;
<a name="l00031"></a>00031 <span class="keyword">typedef</span> vector&lt;GenomicInterval *&gt; GenomicIntervalSetAsVector;
<a name="l00032"></a>00032 <span class="keyword">typedef</span> list&lt;GenomicInterval *&gt; GenomicIntervalSetAsList;
<a name="l00033"></a>00033 <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>;
<a name="l00034"></a>00034 <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>;
<a name="l00035"></a>00035 <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>;
<a name="l00036"></a>00036 
<a name="l00037"></a>00037 <span class="comment">// types</span>
<a name="l00038"></a>00038 <span class="keyword">typedef</span> map&lt;string,string&gt; StringMap;
<a name="l00039"></a>00039 <span class="keyword">typedef</span> map&lt;string,long int&gt; StringLIntMap;
<a name="l00040"></a>00040 <span class="keyword">typedef</span> map&lt;string,vector&lt;long int&gt; &gt; StringVecLIntMap;
<a name="l00041"></a>00041 
<a name="l00042"></a>00042 <span class="comment">// choose container type for interval set (vector, list or array)</span>
<a name="l00043"></a>00043 <span class="keyword">typedef</span> GenomicIntervalSetAsVector GenomicIntervalSet;
<a name="l00044"></a>00044 <span class="comment">//typedef GenomicIntervalSetAsList GenomicIntervalSet;</span>
<a name="l00045"></a>00045 <span class="comment">//typedef GenomicIntervalSetAsArray GenomicIntervalSet;</span>
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 <span class="comment">// for buffer in SortedGenomicRegionSetOverlaps</span>
<a name="l00048"></a>00048 <span class="keyword">typedef</span> list&lt;GenomicRegion *&gt; GenomicRegionList;
<a name="l00049"></a>00049 
<a name="l00050"></a>00050 
<a name="l00051"></a>00051 <span class="comment">// constants</span>
<a name="l00052"></a>00052 <span class="preprocessor">#define ALIGNMENT_GAP &#39;*&#39;                                       //! character for alignment gaps </span>
<a name="l00053"></a>00053 <span class="preprocessor"></span><span class="keyword">const</span> <span class="keywordtype">char</span> refCIGARop[] = <span class="stringliteral">&quot;MD=X&quot;</span>;                       
<a name="l00054"></a>00054 <span class="keyword">const</span> <span class="keywordtype">char</span> fragCIGARop[] = <span class="stringliteral">&quot;MIS=X&quot;</span>;                     
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 
<a name="l00057"></a>00057 
<a name="l00058"></a>00058   
<a name="l00059"></a>00059   
<a name="l00060"></a>00060 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00061"></a>00061 <span class="comment">// CLASS DECLARATION: GenomicInterval                                                          //</span>
<a name="l00062"></a>00062 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00064"></a>00064 <span class="comment"></span>
<a name="l00078"></a>00078 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00079"></a><a class="code" href="classGenomicInterval.html">00079</a> <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>
<a name="l00080"></a>00080 {
<a name="l00081"></a>00081  <span class="keyword">public</span>:
<a name="l00083"></a>00083   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00084"></a>00084 
<a name="l00085"></a>00085   
<a name="l00086"></a>00086 
<a name="l00088"></a>00088   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<span class="keywordtype">string</span> chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00089"></a>00089 
<a name="l00090"></a>00090   
<a name="l00091"></a>00091 
<a name="l00093"></a>00093   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">char</span> delimiter=<span class="charliteral">&#39; &#39;</span>, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00094"></a>00094 
<a name="l00095"></a>00095   
<a name="l00096"></a>00096 
<a name="l00098"></a>00098   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00099"></a>00099 
<a name="l00100"></a>00100   
<a name="l00101"></a>00101   
<a name="l00103"></a>00103   ~<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>();
<a name="l00104"></a>00104 
<a name="l00105"></a>00105   
<a name="l00106"></a>00106 
<a name="l00107"></a>00107   <span class="comment">//------------------------------------------------------//</span>
<a name="l00108"></a>00108   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00109"></a>00109   <span class="comment">//------------------------------------------------------//</span>
<a name="l00110"></a>00110 
<a name="l00111"></a>00111 
<a name="l00113"></a>00113   <span class="keywordtype">void</span> PrintInterval();
<a name="l00114"></a>00114 
<a name="l00115"></a>00115   
<a name="l00116"></a>00116 
<a name="l00118"></a>00118   <span class="keywordtype">void</span> PrintInterval(FILE *file_ptr);
<a name="l00119"></a>00119 
<a name="l00120"></a>00120   
<a name="l00121"></a>00121 
<a name="l00123"></a>00123   <span class="keywordtype">void</span> PrintInterval(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00124"></a>00124 
<a name="l00125"></a>00125 
<a name="l00127"></a>00127   <span class="keywordtype">void</span> Print(<span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00128"></a>00128 
<a name="l00129"></a>00129   
<a name="l00130"></a>00130 
<a name="l00132"></a>00132   <span class="keywordtype">void</span> PrintWithLabel(<span class="keywordtype">char</span> *label, <span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00133"></a>00133 
<a name="l00134"></a>00134   
<a name="l00135"></a>00135   
<a name="l00136"></a>00136   <span class="comment">//------------------------------------------------------//</span>
<a name="l00137"></a>00137   <span class="comment">//   Check                                              //</span>
<a name="l00138"></a>00138   <span class="comment">//------------------------------------------------------//</span>
<a name="l00139"></a>00139 
<a name="l00140"></a>00140 
<a name="l00142"></a>00142   <span class="keywordtype">bool</span> IsValid();
<a name="l00143"></a>00143 
<a name="l00144"></a>00144   
<a name="l00145"></a>00145 
<a name="l00147"></a>00147   <span class="keywordtype">bool</span> CheckValid(<span class="keywordtype">bool</span> quiet=<span class="keyword">false</span>);
<a name="l00148"></a>00148 
<a name="l00149"></a>00149   
<a name="l00150"></a>00150 
<a name="l00152"></a>00152   <span class="keywordtype">bool</span> IsBefore(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00153"></a>00153 
<a name="l00154"></a>00154 
<a name="l00155"></a>00155 
<a name="l00157"></a>00157   <span class="keywordtype">bool</span> IsPosBefore(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00158"></a>00158 
<a name="l00159"></a>00159 
<a name="l00160"></a>00160 
<a name="l00162"></a>00162   <span class="keywordtype">bool</span> IsCompatibleWith(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00163"></a>00163 
<a name="l00164"></a>00164 
<a name="l00165"></a>00165 
<a name="l00167"></a>00167   <span class="keywordtype">bool</span> OverlapsWith(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand=<span class="keyword">false</span>);
<a name="l00168"></a>00168 
<a name="l00169"></a>00169 
<a name="l00170"></a>00170 
<a name="l00171"></a>00171   <span class="comment">//------------------------------------------------------//</span>
<a name="l00172"></a>00172   <span class="comment">//   Set &amp; Get                                          //</span>
<a name="l00173"></a>00173   <span class="comment">//------------------------------------------------------//</span>
<a name="l00174"></a>00174 
<a name="l00175"></a>00175 
<a name="l00177"></a>00177   <span class="keywordtype">void</span> SetCoordinates(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00178"></a>00178 
<a name="l00179"></a>00179   
<a name="l00180"></a>00180 
<a name="l00182"></a>00182   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetCoordinate(<span class="keywordtype">char</span> *op);
<a name="l00183"></a>00183   
<a name="l00184"></a>00184 
<a name="l00185"></a>00185 
<a name="l00187"></a>00187   <span class="keywordtype">size_t</span> GetSize();
<a name="l00188"></a>00188   
<a name="l00189"></a>00189 
<a name="l00190"></a>00190 
<a name="l00192"></a>00192   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00193"></a>00193   
<a name="l00194"></a>00194 
<a name="l00195"></a>00195 
<a name="l00197"></a>00197   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00198"></a>00198   
<a name="l00199"></a>00199 
<a name="l00200"></a>00200 
<a name="l00201"></a>00201   <span class="comment">//------------------------------------------------------//</span>
<a name="l00202"></a>00202   <span class="comment">//   Operations                                         //</span>
<a name="l00203"></a>00203   <span class="comment">//------------------------------------------------------//</span>
<a name="l00204"></a>00204 
<a name="l00205"></a>00205 
<a name="l00207"></a>00207   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcOverlap(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00208"></a>00208 
<a name="l00209"></a>00209 
<a name="l00210"></a>00210 
<a name="l00212"></a>00212   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcDistanceFrom(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">char</span> *op, <span class="keywordtype">char</span> *I_op);
<a name="l00213"></a>00213   
<a name="l00214"></a>00214 
<a name="l00215"></a>00215 
<a name="l00217"></a>00217   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00218"></a>00218   
<a name="l00219"></a>00219 
<a name="l00220"></a>00220 
<a name="l00222"></a>00222   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00223"></a>00223   
<a name="l00224"></a>00224 
<a name="l00225"></a>00225 
<a name="l00227"></a>00227   <span class="keywordtype">void</span> ReverseStrand(); 
<a name="l00228"></a>00228   
<a name="l00229"></a>00229 
<a name="l00230"></a>00230 
<a name="l00232"></a>00232 
<a name="l00237"></a>00237   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l00238"></a>00238   
<a name="l00239"></a>00239 
<a name="l00240"></a>00240 
<a name="l00242"></a>00242 
<a name="l00246"></a>00246   <span class="keywordtype">void</span> ModifyPos(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00247"></a>00247 
<a name="l00248"></a>00248 
<a name="l00249"></a>00249 
<a name="l00251"></a>00251   <span class="keywordtype">void</span> ReversePos(StringLIntMap *bounds);
<a name="l00252"></a>00252 
<a name="l00253"></a>00253 
<a name="l00254"></a>00254 
<a name="l00256"></a>00256   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l00257"></a>00257 
<a name="l00258"></a>00258 
<a name="l00259"></a>00259 
<a name="l00261"></a>00261   <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00262"></a>00262 
<a name="l00263"></a>00263 
<a name="l00264"></a>00264 
<a name="l00266"></a>00266   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00267"></a>00267 
<a name="l00268"></a>00268 
<a name="l00269"></a>00269 
<a name="l00271"></a>00271 
<a name="l00278"></a>00278   <span class="keywordtype">void</span> GetOffsetFrom(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *ReferenceI, <span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *start_offset, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *stop_offset);
<a name="l00279"></a>00279 
<a name="l00280"></a>00280 
<a name="l00281"></a>00281   
<a name="l00282"></a>00282 
<a name="l00283"></a>00283   <span class="comment">//------------------------------------------------------//</span>
<a name="l00284"></a>00284   <span class="comment">//   Data                                               //</span>
<a name="l00285"></a>00285   <span class="comment">//------------------------------------------------------//</span>
<a name="l00286"></a>00286   
<a name="l00287"></a><a class="code" href="classGenomicInterval.html#a0dbe6195570468c386eea923e30c762c">00287</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#a0dbe6195570468c386eea923e30c762c" title="file line number">n_line</a>;             
<a name="l00288"></a><a class="code" href="classGenomicInterval.html#ac489a1195c43cc93d0e192dc91d5a192">00288</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicInterval.html#ac489a1195c43cc93d0e192dc91d5a192" title="chromosome name (space and tab characters are not allowed)">CHROMOSOME</a>;                             
<a name="l00289"></a><a class="code" href="classGenomicInterval.html#a25f6603b48fcc57f97c89fa768b658f9">00289</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicInterval.html#a25f6603b48fcc57f97c89fa768b658f9" title="orientation (&#39;+&#39; or &#39;-&#39;)">STRAND</a>;                                  
<a name="l00290"></a><a class="code" href="classGenomicInterval.html#ac0abba93599f5cdf4b5a4e7ec5909bbd">00290</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#ac0abba93599f5cdf4b5a4e7ec5909bbd" title="interval start position in the chromosome (&gt;=1)">START</a>;                               
<a name="l00291"></a><a class="code" href="classGenomicInterval.html#ac5fac21da3939b7976859a54800b93f0">00291</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#ac5fac21da3939b7976859a54800b93f0" title="interval stop position in the chromosome (&lt;=chromosome_size)">STOP</a>;                                
<a name="l00292"></a>00292 };
<a name="l00293"></a>00293 
<a name="l00294"></a>00294 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00295"></a>00295 <span class="comment">// END CLASS DECLARATION: GenomicIntervals                                                     //</span>
<a name="l00296"></a>00296 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00297"></a>00297 
<a name="l00298"></a>00298 
<a name="l00299"></a>00299 
<a name="l00300"></a>00300 
<a name="l00301"></a>00301 
<a name="l00302"></a>00302 
<a name="l00303"></a>00303 
<a name="l00304"></a>00304 
<a name="l00305"></a>00305 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00306"></a>00306 <span class="comment">// CLASS: Chromosomes                                                                          //</span>
<a name="l00307"></a>00307 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00309"></a>00309 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00310"></a><a class="code" href="classChromosomes.html">00310</a> <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>
<a name="l00311"></a>00311 {
<a name="l00312"></a>00312  <span class="keyword">public</span>:
<a name="l00313"></a>00313 
<a name="l00314"></a><a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243">00314</a>   <span class="keyword">typedef</span> map&lt;string,pair&lt;char*,size_t&gt; &gt; <a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a>;            
<a name="l00315"></a>00315 
<a name="l00317"></a>00317 
<a name="l00351"></a>00351   <a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>(<span class="keywordtype">char</span> *chrom_map_dir, <span class="keywordtype">char</span> *chrom_map_name);
<a name="l00352"></a>00352 
<a name="l00353"></a>00353 
<a name="l00355"></a>00355 
<a name="l00371"></a>00371   <a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>(<span class="keywordtype">char</span> *fasta_file, <span class="keywordtype">bool</span> verbose);
<a name="l00372"></a>00372 
<a name="l00373"></a>00373   
<a name="l00375"></a>00375   ~<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>();
<a name="l00376"></a>00376   
<a name="l00377"></a>00377 
<a name="l00379"></a>00379   <span class="keywordtype">bool</span> FindChromosome(<span class="keywordtype">char</span> *chromosome_name);
<a name="l00380"></a>00380   
<a name="l00381"></a>00381   
<a name="l00383"></a>00383   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00384"></a>00384   
<a name="l00385"></a>00385   
<a name="l00387"></a>00387   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00388"></a>00388   
<a name="l00389"></a>00389   
<a name="l00391"></a>00391   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00392"></a>00392   
<a name="l00393"></a>00393   
<a name="l00395"></a>00395   <span class="keywordtype">void</span> LoadChromosome(<span class="keywordtype">string</span> chromosome_name);
<a name="l00396"></a>00396 
<a name="l00397"></a>00397   
<a name="l00398"></a>00398   <span class="comment">// data</span>
<a name="l00399"></a>00399 
<a name="l00400"></a><a class="code" href="classChromosomes.html#a954bae240564f6f589a30d253291eb75">00400</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#a954bae240564f6f589a30d253291eb75" title="if &#39;true&#39;, print messages while loading chromosome sequences">verbose</a>;                                                         
<a name="l00401"></a><a class="code" href="classChromosomes.html#a408a06b9ae635ceaf0b75452bcb62763">00401</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#a408a06b9ae635ceaf0b75452bcb62763" title="&#39;true&#39; if reading from FASTA file">load_in_memory</a>;                                          
<a name="l00402"></a><a class="code" href="classChromosomes.html#ad72dcc9b9d3b0b041b737d895a20eef6">00402</a>   StringMap <a class="code" href="classChromosomes.html#ad72dcc9b9d3b0b041b737d895a20eef6" title="maps chromosome names to chromosome sequence file names">chrom_map</a>;                                          
<a name="l00403"></a><a class="code" href="classChromosomes.html#a5fb6a314ac56a7ace50070fd3e5ae708">00403</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#a5fb6a314ac56a7ace50070fd3e5ae708" title="the directory where the chromosome map file is located">chrom_map_dir</a>;                                         
<a name="l00404"></a><a class="code" href="classChromosomes.html#a71b69dff9fc8c4cae1671df17e11a016">00404</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#a71b69dff9fc8c4cae1671df17e11a016" title="the name of the chromosome map file">chrom_map_name</a>;                                        
<a name="l00405"></a><a class="code" href="classChromosomes.html#ab3b502eec7bd2028297300aeb47b5844">00405</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#ab3b502eec7bd2028297300aeb47b5844" title="name of the currently used chromosome">current_chromosome_name</a>;                       
<a name="l00406"></a><a class="code" href="classChromosomes.html#afb59cb80755187f4546573f10110e669">00406</a>   <span class="keywordtype">size_t</span> <a class="code" href="classChromosomes.html#afb59cb80755187f4546573f10110e669" title="size of the currently used chromosome">current_chromosome_size</a>;                       
<a name="l00407"></a><a class="code" href="classChromosomes.html#af640cf0f09b109b3d46d70aba9207a2e">00407</a>   <span class="keywordtype">char</span> *<a class="code" href="classChromosomes.html#af640cf0f09b109b3d46d70aba9207a2e" title="pointer to the currently used chromosome sequence">current_chromosome_seq</a>;                         
<a name="l00408"></a><a class="code" href="classChromosomes.html#a673169d55aa0b94815e17b41566d4d8c">00408</a>   <a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a> <a class="code" href="classChromosomes.html#a673169d55aa0b94815e17b41566d4d8c" title="all chromosomes sequences and their sizes are stored here">chromosome_seq_data</a>;        
<a name="l00409"></a>00409 };
<a name="l00410"></a>00410 
<a name="l00411"></a>00411 
<a name="l00412"></a>00412 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00413"></a>00413 <span class="comment">// END CLASS: Chromosomes                                                                      //</span>
<a name="l00414"></a>00414 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00415"></a>00415 
<a name="l00416"></a>00416 
<a name="l00417"></a>00417 
<a name="l00418"></a>00418 
<a name="l00419"></a>00419 
<a name="l00420"></a>00420 
<a name="l00421"></a>00421 
<a name="l00422"></a>00422 
<a name="l00423"></a>00423 
<a name="l00424"></a>00424 
<a name="l00425"></a>00425 
<a name="l00426"></a>00426 
<a name="l00427"></a>00427 
<a name="l00428"></a>00428 
<a name="l00429"></a>00429 
<a name="l00430"></a>00430 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00431"></a>00431 <span class="comment">// CLASS: GenomicIntervalSetAsArray                                                            //</span>
<a name="l00432"></a>00432 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00434"></a>00434 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00435"></a>00435 
<a name="l00436"></a><a class="code" href="classGenomicIntervalSetAsArray.html">00436</a> <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>
<a name="l00437"></a>00437 {
<a name="l00438"></a>00438  <span class="keyword">public</span>:
<a name="l00439"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a1a0416c9fac7e9d965ae7fe84f1649c2">00439</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>** <a class="code" href="classGenomicIntervalSetAsArray.html#a1a0416c9fac7e9d965ae7fe84f1649c2" title="iterator defined as a pointer in the array of GenomicInterval">iterator</a>;                           
<a name="l00440"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0310f5d79c529b631a5ef71ef186b41c">00440</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>** <a class="code" href="classGenomicIntervalSetAsArray.html#a0310f5d79c529b631a5ef71ef186b41c" title="reverse_iterator defined as a pointer in the array of GenomicInterval">reverse_iterator</a>;           
<a name="l00441"></a>00441 
<a name="l00442"></a>00442 
<a name="l00444"></a>00444   <a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>();
<a name="l00445"></a>00445 
<a name="l00446"></a>00446 
<a name="l00448"></a>00448   ~<a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>();
<a name="l00449"></a>00449   
<a name="l00450"></a>00450   
<a name="l00452"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a9e56381009d164d7625199b183845388">00452</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a9e56381009d164d7625199b183845388" title="Implementation of begin() for this container.">begin</a>() { <span class="keywordflow">return</span> &amp;I[0]; }
<a name="l00453"></a>00453   
<a name="l00454"></a>00454 
<a name="l00456"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a5b6145ceff99907d9075047519482584">00456</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a5b6145ceff99907d9075047519482584" title="Implementation of end() for this container.">end</a>() { <span class="keywordflow">return</span> &amp;I[n_intervals]; }
<a name="l00457"></a>00457   
<a name="l00458"></a>00458 
<a name="l00460"></a><a class="code" href="classGenomicIntervalSetAsArray.html#ac077589a55a7dfec9b2e6c38392f7444">00460</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#ac077589a55a7dfec9b2e6c38392f7444" title="Implementation of rbegin() for this container.">rbegin</a>() { <span class="keywordflow">return</span> &amp;I[n_intervals-1]; }
<a name="l00461"></a>00461   
<a name="l00462"></a>00462 
<a name="l00464"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a5b7fc7dc64857dfdadd1198393c0d117">00464</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a5b7fc7dc64857dfdadd1198393c0d117" title="Implementation of rend() for this container.">rend</a>() { <span class="keywordflow">return</span> &amp;I[-1]; }
<a name="l00465"></a>00465   
<a name="l00466"></a>00466 
<a name="l00468"></a><a class="code" href="classGenomicIntervalSetAsArray.html#aa74747b36b114c326edfaf1224270d0d">00468</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#aa74747b36b114c326edfaf1224270d0d" title="Implementation of front() for this container.">front</a>() { <span class="keywordflow">return</span> I[0]; }
<a name="l00469"></a>00469   
<a name="l00470"></a>00470 
<a name="l00472"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a11b483c12111ee96c15fec296c66ee15">00472</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#a11b483c12111ee96c15fec296c66ee15" title="Implementation of back() for this container.">back</a>() { <span class="keywordflow">return</span> I[n_intervals-1]; }
<a name="l00473"></a>00473   
<a name="l00474"></a>00474 
<a name="l00476"></a><a class="code" href="classGenomicIntervalSetAsArray.html#ad4bfef55744241849841c71d08835403">00476</a>   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicIntervalSetAsArray.html#ad4bfef55744241849841c71d08835403" title="Implementation of size() for this container.">size</a>() { <span class="keywordflow">return</span> n_intervals; }
<a name="l00477"></a>00477   
<a name="l00478"></a>00478 
<a name="l00480"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a4348e7636cb91405b08374a966e33a7c">00480</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a4348e7636cb91405b08374a966e33a7c" title="Implementation of push_back() for this container.">push_back</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *j) { I[n_intervals++] = j; }
<a name="l00481"></a>00481   
<a name="l00482"></a>00482 
<a name="l00484"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a833395eaf9daa5626879eefa861d9ea5">00484</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a833395eaf9daa5626879eefa861d9ea5" title="Implementation of clear() for this container.">clear</a>() {  }
<a name="l00485"></a>00485   
<a name="l00486"></a>00486 
<a name="l00488"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0f8554cfca150d695d15691aa42df7fa">00488</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0f8554cfca150d695d15691aa42df7fa" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> j) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be erased!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00489"></a>00489 
<a name="l00490"></a>00490 
<a name="l00492"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0aae25ac2e9a7cf50a370703aa99df85">00492</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0aae25ac2e9a7cf50a370703aa99df85" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be erased!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00493"></a>00493   
<a name="l00494"></a>00494 
<a name="l00496"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0cd89244c0e75b88f9b56c9b9478d06b">00496</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0cd89244c0e75b88f9b56c9b9478d06b" title="Implementation of insert() for this container (exits with error message).">insert</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *u) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be inserted!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00497"></a>00497   
<a name="l00498"></a>00498 
<a name="l00499"></a>00499   
<a name="l00500"></a>00500   <span class="comment">//------------------------------------------------------//</span>
<a name="l00501"></a>00501   <span class="comment">//   Data                                               //</span>
<a name="l00502"></a>00502   <span class="comment">//------------------------------------------------------//</span>
<a name="l00503"></a>00503   
<a name="l00504"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a9f9558bbaa8d682993a60fa142893998">00504</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>;                                 
<a name="l00505"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a6580b875cd9804ba59f88c5000f83f09">00505</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a6580b875cd9804ba59f88c5000f83f09" title="number of maximum allowed intervals (this will be removed)">max_intervals</a>;                               
<a name="l00506"></a><a class="code" href="classGenomicIntervalSetAsArray.html#af7ec8bbfea7748695aa35e6ec467a4b2">00506</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> **<a class="code" href="classGenomicIntervalSetAsArray.html#af7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>;                                  
<a name="l00507"></a>00507   
<a name="l00508"></a>00508 };
<a name="l00509"></a>00509 
<a name="l00510"></a>00510 
<a name="l00511"></a>00511 
<a name="l00512"></a>00512 
<a name="l00513"></a>00513 
<a name="l00514"></a>00514 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00515"></a>00515 <span class="comment">// END CLASS: GenomicIntervalArray                                                             //</span>
<a name="l00516"></a>00516 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00517"></a>00517 
<a name="l00518"></a>00518 
<a name="l00519"></a>00519 
<a name="l00520"></a>00520 
<a name="l00521"></a>00521 
<a name="l00522"></a>00522 
<a name="l00523"></a>00523 
<a name="l00524"></a>00524 
<a name="l00525"></a>00525 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00526"></a>00526 <span class="comment">// CLASS: GenomicRegion                                                                        //</span>
<a name="l00527"></a>00527 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00528"></a>00528 <span class="comment">//                                                                                             //</span>
<a name="l00529"></a>00529 <span class="comment">// Input formats:                                                                              //</span>
<a name="l00530"></a>00530 <span class="comment">//  1. LABEL &lt;TAB&gt; CHROMOSOME STRAND START[,START]* STOP[,STOP]*                               //</span>
<a name="l00531"></a>00531 <span class="comment">//  2. LABEL &lt;TAB&gt; [CHROMOSOME STRAND START STOP]+                                             //</span>
<a name="l00532"></a>00532 <span class="comment">//                                                                                             //</span>
<a name="l00533"></a>00533 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00535"></a>00535 <span class="comment"></span>
<a name="l00544"></a>00544 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00545"></a>00545 
<a name="l00546"></a><a class="code" href="classGenomicRegion.html">00546</a> <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l00547"></a>00547 {
<a name="l00548"></a>00548  <span class="keyword">public</span>:
<a name="l00550"></a>00550   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>();
<a name="l00551"></a>00551 
<a name="l00552"></a>00552 
<a name="l00553"></a>00553 
<a name="l00555"></a>00555   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l00556"></a>00556 
<a name="l00557"></a>00557 
<a name="l00558"></a>00558 
<a name="l00560"></a>00560   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0); 
<a name="l00561"></a>00561 
<a name="l00562"></a>00562 
<a name="l00563"></a>00563 
<a name="l00565"></a>00565   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i);
<a name="l00566"></a>00566 
<a name="l00567"></a>00567 
<a name="l00568"></a>00568 
<a name="l00570"></a>00570   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>();
<a name="l00571"></a>00571   
<a name="l00572"></a>00572 
<a name="l00573"></a>00573 
<a name="l00574"></a>00574   <span class="comment">//------------------------------------------------------//</span>
<a name="l00575"></a>00575   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00576"></a>00576   <span class="comment">//------------------------------------------------------//</span>
<a name="l00577"></a>00577 
<a name="l00578"></a>00578 
<a name="l00580"></a>00580   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0, <span class="keywordtype">char</span> del1=<span class="charliteral">&#39; &#39;</span>, <span class="keywordtype">char</span> del2=<span class="charliteral">&#39;,&#39;</span>);
<a name="l00581"></a>00581 
<a name="l00582"></a>00582 
<a name="l00583"></a>00583 
<a name="l00585"></a>00585   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l00586"></a>00586 
<a name="l00587"></a>00587 
<a name="l00588"></a>00588   
<a name="l00590"></a>00590   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00591"></a>00591 
<a name="l00592"></a>00592 
<a name="l00593"></a>00593 
<a name="l00595"></a>00595   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintIntersection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l00596"></a>00596 
<a name="l00597"></a>00597 
<a name="l00598"></a>00598 
<a name="l00600"></a>00600   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l00601"></a>00601 
<a name="l00602"></a>00602 
<a name="l00603"></a>00603 
<a name="l00605"></a>00605   <span class="keywordtype">void</span> PrintREG(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l00606"></a>00606 
<a name="l00607"></a>00607 
<a name="l00608"></a>00608 
<a name="l00610"></a>00610   <span class="keywordtype">void</span> PrintQ(<span class="keywordtype">bool</span> reverse=<span class="keyword">false</span>);
<a name="l00611"></a>00611 
<a name="l00612"></a>00612 
<a name="l00613"></a>00613 
<a name="l00615"></a>00615   <span class="keywordtype">void</span> PrintError(<span class="keywordtype">string</span> error_msg);
<a name="l00616"></a>00616 
<a name="l00617"></a>00617 
<a name="l00618"></a>00618 
<a name="l00619"></a>00619 
<a name="l00620"></a>00620   <span class="comment">//------------------------------------------------------//</span>
<a name="l00621"></a>00621   <span class="comment">//   Get &amp; Set                                          //</span>
<a name="l00622"></a>00622   <span class="comment">//------------------------------------------------------//</span>
<a name="l00623"></a>00623 
<a name="l00624"></a>00624 
<a name="l00626"></a>00626   <span class="keywordtype">void</span> DeleteIntervals(GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j); 
<a name="l00627"></a>00627 
<a name="l00628"></a>00628 
<a name="l00629"></a>00629 
<a name="l00631"></a>00631   <span class="keywordtype">void</span> SetLabel(<span class="keywordtype">char</span> *label);
<a name="l00632"></a>00632 
<a name="l00633"></a>00633 
<a name="l00634"></a>00634 
<a name="l00636"></a>00636   <span class="keywordtype">void</span> SetStrand(<span class="keywordtype">char</span> strand);
<a name="l00637"></a>00637 
<a name="l00638"></a>00638 
<a name="l00639"></a>00639 
<a name="l00641"></a>00641   <span class="keywordtype">char</span> *GetChromosome();
<a name="l00642"></a>00642 
<a name="l00643"></a>00643 
<a name="l00644"></a>00644 
<a name="l00646"></a>00646   <span class="keywordtype">size_t</span> GetSize();
<a name="l00647"></a>00647 
<a name="l00648"></a>00648 
<a name="l00649"></a>00649 
<a name="l00651"></a>00651   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00652"></a>00652 
<a name="l00653"></a>00653 
<a name="l00654"></a>00654 
<a name="l00656"></a>00656   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00657"></a>00657 
<a name="l00658"></a>00658 
<a name="l00659"></a>00659 
<a name="l00660"></a>00660   <span class="comment">//------------------------------------------------------//</span>
<a name="l00661"></a>00661   <span class="comment">//   Check                                              //</span>
<a name="l00662"></a>00662   <span class="comment">//------------------------------------------------------//</span>
<a name="l00663"></a>00663 
<a name="l00664"></a>00664 
<a name="l00666"></a>00666   <span class="keywordtype">bool</span> IsCompatible(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00667"></a>00667 
<a name="l00668"></a>00668 
<a name="l00669"></a>00669 
<a name="l00671"></a>00671   <span class="keywordtype">bool</span> IsCompatibleWith(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00672"></a>00672 
<a name="l00673"></a>00673 
<a name="l00674"></a>00674 
<a name="l00676"></a>00676   <span class="keywordtype">bool</span> IsCompatibleSorted(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00677"></a>00677 
<a name="l00678"></a>00678 
<a name="l00679"></a>00679 
<a name="l00681"></a>00681   <span class="keywordtype">bool</span> IsCompatibleSortedAndNonoverlapping();
<a name="l00682"></a>00682 
<a name="l00683"></a>00683 
<a name="l00684"></a>00684 
<a name="l00686"></a>00686   <span class="keywordtype">bool</span> OverlapsWith(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00687"></a>00687 
<a name="l00688"></a>00688 
<a name="l00689"></a>00689 
<a name="l00691"></a>00691   <span class="keywordtype">bool</span> IsBefore(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00692"></a>00692 
<a name="l00693"></a>00693 
<a name="l00694"></a>00694 
<a name="l00696"></a>00696   <span class="keywordtype">bool</span> IsPosBefore(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l00697"></a>00697 
<a name="l00698"></a>00698 
<a name="l00699"></a>00699 
<a name="l00701"></a>00701   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcOverlap(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00702"></a>00702 
<a name="l00703"></a>00703 
<a name="l00704"></a>00704 
<a name="l00706"></a>00706   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00707"></a>00707 
<a name="l00708"></a>00708   
<a name="l00709"></a>00709 
<a name="l00711"></a>00711   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00712"></a>00712 
<a name="l00713"></a>00713   
<a name="l00714"></a>00714 
<a name="l00715"></a>00715   <span class="comment">//------------------------------------------------------//</span>
<a name="l00716"></a>00716   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l00717"></a>00717   <span class="comment">//------------------------------------------------------//</span>
<a name="l00718"></a>00718 
<a name="l00719"></a>00719 
<a name="l00721"></a>00721   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00722"></a>00722 
<a name="l00723"></a>00723 
<a name="l00724"></a>00724   
<a name="l00726"></a>00726   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00727"></a>00727 
<a name="l00728"></a>00728 
<a name="l00729"></a>00729 
<a name="l00731"></a>00731   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l00732"></a>00732 
<a name="l00733"></a>00733 
<a name="l00734"></a>00734 
<a name="l00736"></a>00736   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Center();
<a name="l00737"></a>00737 
<a name="l00738"></a>00738 
<a name="l00739"></a>00739 
<a name="l00741"></a>00741   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Connect();
<a name="l00742"></a>00742 
<a name="l00743"></a>00743 
<a name="l00744"></a>00744 
<a name="l00746"></a>00746   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Diff();
<a name="l00747"></a>00747 
<a name="l00748"></a>00748 
<a name="l00749"></a>00749 
<a name="l00751"></a>00751   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l00752"></a>00752 
<a name="l00753"></a>00753 
<a name="l00754"></a>00754 
<a name="l00756"></a>00756   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Divide();
<a name="l00757"></a>00757 
<a name="l00758"></a>00758 
<a name="l00759"></a>00759 
<a name="l00761"></a>00761   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunDivide();
<a name="l00762"></a>00762 
<a name="l00763"></a>00763 
<a name="l00764"></a>00764 
<a name="l00766"></a>00766   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Fix();
<a name="l00767"></a>00767 
<a name="l00768"></a>00768 
<a name="l00769"></a>00769 
<a name="l00771"></a>00771   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Intersect();
<a name="l00772"></a>00772 
<a name="l00773"></a>00773 
<a name="l00774"></a>00774 
<a name="l00776"></a>00776   <span class="keywordtype">void</span> RunSize();
<a name="l00777"></a>00777 
<a name="l00778"></a>00778 
<a name="l00779"></a>00779 
<a name="l00781"></a>00781 
<a name="l00785"></a>00785   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyPos(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00786"></a>00786 
<a name="l00787"></a>00787 
<a name="l00788"></a>00788 
<a name="l00790"></a>00790   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00791"></a>00791 
<a name="l00792"></a>00792 
<a name="l00793"></a>00793 
<a name="l00795"></a>00795   <span class="keywordtype">void</span> ReversePos(StringLIntMap *bounds);
<a name="l00796"></a>00796 
<a name="l00797"></a>00797 
<a name="l00798"></a>00798   
<a name="l00800"></a>00800   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l00801"></a>00801 
<a name="l00802"></a>00802 
<a name="l00803"></a>00803 
<a name="l00805"></a>00805 
<a name="l00810"></a>00810   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l00811"></a>00811 
<a name="l00812"></a>00812 
<a name="l00813"></a>00813 
<a name="l00815"></a>00815   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l00816"></a>00816 
<a name="l00817"></a>00817 
<a name="l00818"></a>00818 
<a name="l00820"></a>00820   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Sort(); 
<a name="l00821"></a>00821 
<a name="l00822"></a>00822 
<a name="l00823"></a>00823   
<a name="l00825"></a>00825   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunSplit(); 
<a name="l00826"></a>00826 
<a name="l00827"></a>00827 
<a name="l00828"></a>00828   
<a name="l00830"></a>00830   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyStrand(<span class="keywordtype">char</span> *strand_op);
<a name="l00831"></a>00831 
<a name="l00832"></a>00832 
<a name="l00833"></a>00833   
<a name="l00835"></a>00835   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Union();
<a name="l00836"></a>00836 
<a name="l00837"></a>00837 
<a name="l00838"></a>00838 
<a name="l00840"></a>00840   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunUnion();
<a name="l00841"></a>00841 
<a name="l00842"></a>00842 
<a name="l00843"></a>00843 
<a name="l00845"></a>00845   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l00846"></a>00846 
<a name="l00847"></a>00847 
<a name="l00848"></a>00848 
<a name="l00850"></a>00850   <span class="keywordtype">void</span> PrintRemoveN(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00851"></a>00851 
<a name="l00852"></a>00852 
<a name="l00853"></a>00853 
<a name="l00855"></a>00855   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l00856"></a>00856 
<a name="l00857"></a>00857   
<a name="l00859"></a>00859   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00860"></a>00860 
<a name="l00861"></a>00861 
<a name="l00862"></a>00862 
<a name="l00864"></a>00864 
<a name="l00868"></a>00868   <span class="keywordtype">void</span> PrintOffsetFormat(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l00869"></a>00869 
<a name="l00870"></a>00870 
<a name="l00871"></a>00871 
<a name="l00873"></a>00873   <span class="keywordtype">void</span> PrintSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00874"></a>00874 
<a name="l00875"></a>00875 
<a name="l00876"></a>00876 
<a name="l00878"></a>00878   <span class="keywordtype">void</span> PrintVerifySeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l00879"></a>00879 
<a name="l00880"></a>00880 
<a name="l00881"></a>00881 
<a name="l00883"></a>00883   <span class="keywordtype">void</span> ReverseStrand();
<a name="l00884"></a>00884 
<a name="l00885"></a>00885 
<a name="l00886"></a>00886 
<a name="l00887"></a>00887   <span class="comment">//------------------------------------------------------//</span>
<a name="l00888"></a>00888   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l00889"></a>00889   <span class="comment">//------------------------------------------------------//</span>
<a name="l00890"></a>00890 
<a name="l00891"></a>00891   
<a name="l00893"></a>00893   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Constrain(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l00894"></a>00894 
<a name="l00895"></a>00895   
<a name="l00896"></a>00896 
<a name="l00898"></a>00898   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Diff(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l00899"></a>00899 
<a name="l00900"></a>00900   
<a name="l00901"></a>00901 
<a name="l00902"></a>00902   <span class="comment">//------------------------------------------------------//</span>
<a name="l00903"></a>00903   <span class="comment">//   Data                                               //</span>
<a name="l00904"></a>00904   <span class="comment">//------------------------------------------------------//</span>
<a name="l00905"></a>00905 
<a name="l00906"></a><a class="code" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">00906</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>;                             
<a name="l00907"></a><a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24">00907</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>;                                                  
<a name="l00908"></a><a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9">00908</a>   GenomicIntervalSet <a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>;                                 
<a name="l00909"></a>00909   
<a name="l00910"></a>00910   
<a name="l00911"></a>00911 };
<a name="l00912"></a>00912 
<a name="l00913"></a>00913 
<a name="l00914"></a>00914 
<a name="l00915"></a>00915 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00916"></a>00916 <span class="comment">// END CLASS: GenomicRegion                                                                    //</span>
<a name="l00917"></a>00917 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00918"></a>00918 
<a name="l00919"></a>00919 
<a name="l00920"></a>00920 
<a name="l00921"></a>00921 
<a name="l00922"></a>00922 
<a name="l00923"></a>00923 
<a name="l00924"></a>00924 
<a name="l00925"></a>00925 
<a name="l00926"></a>00926 
<a name="l00927"></a>00927 
<a name="l00928"></a>00928 
<a name="l00929"></a>00929 
<a name="l00930"></a>00930 
<a name="l00931"></a>00931 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00932"></a>00932 <span class="comment">// CLASS: GenomicRegionSEQ                                                                     //</span>
<a name="l00933"></a>00933 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00935"></a>00935 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00936"></a>00936 
<a name="l00937"></a><a class="code" href="classGenomicRegionSEQ.html">00937</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l00938"></a>00938 {
<a name="l00939"></a>00939  <span class="keyword">public</span>:
<a name="l00941"></a>00941   <a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l00942"></a>00942 
<a name="l00944"></a>00944   ~<a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a>();
<a name="l00945"></a>00945   
<a name="l00946"></a>00946 
<a name="l00947"></a>00947 
<a name="l00948"></a>00948   <span class="comment">//------------------------------------------------------//</span>
<a name="l00949"></a>00949   <span class="comment">//   Operations                                         //</span>
<a name="l00950"></a>00950   <span class="comment">//------------------------------------------------------//</span>
<a name="l00951"></a>00951   
<a name="l00953"></a>00953   <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l00954"></a>00954 
<a name="l00955"></a>00955   
<a name="l00956"></a>00956   <span class="comment">//------------------------------------------------------//</span>
<a name="l00957"></a>00957   <span class="comment">//   Data                                               //</span>
<a name="l00958"></a>00958   <span class="comment">//------------------------------------------------------//</span>
<a name="l00959"></a>00959 
<a name="l00960"></a><a class="code" href="classGenomicRegionSEQ.html#a37f8aa6c3c6d8d0059aa56944efa3a52">00960</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSEQ.html#a37f8aa6c3c6d8d0059aa56944efa3a52" title="sequence">SEQ</a>;                                                    
<a name="l00961"></a>00961 
<a name="l00962"></a>00962   
<a name="l00963"></a>00963 <span class="comment">/*</span>
<a name="l00964"></a>00964 <span class="comment">  // NOTE: utilize all data &amp; code related to the SEQ format in a new class: GenomicRegionSEQ.</span>
<a name="l00966"></a>00966 <span class="comment">  void PrintNoGaps();                                                                                   // alignments</span>
<a name="l00967"></a>00967 <span class="comment"></span>
<a name="l00969"></a>00969 <span class="comment">  void PrintModifyLabel();</span>
<a name="l00970"></a>00970 <span class="comment">  </span>
<a name="l00971"></a>00971 <span class="comment">  vector&lt;string&gt; Q;                                             //!&lt; sequences (e.g. from a multiple sequence alignment)</span>
<a name="l00972"></a>00972 <span class="comment">  vector&lt;string&gt; L;                                             //!&lt; sequence labels (e.g. names of species in multiple sequence alignment)</span>
<a name="l00973"></a>00973 <span class="comment">*/</span>
<a name="l00974"></a>00974 
<a name="l00975"></a>00975 };
<a name="l00976"></a>00976 
<a name="l00977"></a>00977 
<a name="l00978"></a>00978 
<a name="l00979"></a>00979 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00980"></a>00980 <span class="comment">// END CLASS: GenomicRegionSEQ                                                                 //</span>
<a name="l00981"></a>00981 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00982"></a>00982 
<a name="l00983"></a>00983 
<a name="l00984"></a>00984 
<a name="l00985"></a>00985 
<a name="l00986"></a>00986 
<a name="l00987"></a>00987 
<a name="l00988"></a>00988 
<a name="l00989"></a>00989 
<a name="l00990"></a>00990 
<a name="l00991"></a>00991 
<a name="l00992"></a>00992 
<a name="l00993"></a>00993 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00994"></a>00994 <span class="comment">// CLASS: GenomicRegionBEDToREG                                                                //</span>
<a name="l00995"></a>00995 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00997"></a>00997 <span class="comment"></span>
<a name="l01004"></a>01004 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01005"></a><a class="code" href="classGenomicRegionBEDToREG.html">01005</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01006"></a>01006 {
<a name="l01007"></a>01007  <span class="keyword">public</span>:
<a name="l01009"></a>01009   <a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01010"></a>01010 
<a name="l01011"></a>01011 
<a name="l01012"></a>01012 
<a name="l01014"></a>01014   <a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01015"></a>01015 
<a name="l01016"></a>01016 
<a name="l01017"></a>01017 
<a name="l01019"></a>01019   ~<a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>();
<a name="l01020"></a>01020   
<a name="l01021"></a>01021 
<a name="l01022"></a>01022 
<a name="l01024"></a>01024   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01025"></a>01025   
<a name="l01026"></a>01026 
<a name="l01027"></a>01027 
<a name="l01028"></a>01028   <span class="comment">// data</span>
<a name="l01029"></a>01029 
<a name="l01030"></a>01030 
<a name="l01031"></a>01031 };
<a name="l01032"></a>01032 
<a name="l01033"></a>01033 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01034"></a>01034 <span class="comment">// END CLASS: GenomicRegionBEDToREG                                                            //</span>
<a name="l01035"></a>01035 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01036"></a>01036 
<a name="l01037"></a>01037 
<a name="l01038"></a>01038 
<a name="l01039"></a>01039 
<a name="l01040"></a>01040 
<a name="l01041"></a>01041 
<a name="l01042"></a>01042 
<a name="l01043"></a>01043 
<a name="l01044"></a>01044 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01045"></a>01045 <span class="comment">// CLASS: GenomicRegionBED                                                                     //</span>
<a name="l01046"></a>01046 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01048"></a>01048 <span class="comment"></span>
<a name="l01055"></a>01055 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01056"></a><a class="code" href="classGenomicRegionBED.html">01056</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01057"></a>01057 {
<a name="l01058"></a>01058  <span class="keyword">public</span>:
<a name="l01060"></a>01060   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01061"></a>01061 
<a name="l01062"></a>01062 
<a name="l01063"></a>01063 
<a name="l01065"></a>01065   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01066"></a>01066 
<a name="l01067"></a>01067 
<a name="l01068"></a>01068 
<a name="l01070"></a>01070   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i);
<a name="l01071"></a>01071 
<a name="l01072"></a>01072 
<a name="l01073"></a>01073 
<a name="l01075"></a>01075   ~<a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>();
<a name="l01076"></a>01076   
<a name="l01077"></a>01077 
<a name="l01078"></a>01078 
<a name="l01079"></a>01079   <span class="comment">//------------------------------------------------------//</span>
<a name="l01080"></a>01080   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01081"></a>01081   <span class="comment">//------------------------------------------------------//</span>
<a name="l01082"></a>01082 
<a name="l01083"></a>01083 
<a name="l01085"></a>01085   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01086"></a>01086   
<a name="l01087"></a>01087 
<a name="l01088"></a>01088 
<a name="l01090"></a>01090   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01091"></a>01091 
<a name="l01092"></a>01092 
<a name="l01093"></a>01093 
<a name="l01095"></a>01095   <span class="keywordtype">void</span> Print(FILE *file_ptr, <span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01096"></a>01096 
<a name="l01097"></a>01097 
<a name="l01098"></a>01098 
<a name="l01100"></a>01100   <span class="keywordtype">void</span> PrintIntersection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l01101"></a>01101 
<a name="l01102"></a>01102 
<a name="l01103"></a>01103 
<a name="l01105"></a>01105   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01106"></a>01106 
<a name="l01107"></a>01107 
<a name="l01108"></a>01108 
<a name="l01110"></a>01110   <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01111"></a>01111 
<a name="l01112"></a>01112 
<a name="l01113"></a>01113 
<a name="l01114"></a>01114   <span class="comment">//------------------------------------------------------//</span>
<a name="l01115"></a>01115   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01116"></a>01116   <span class="comment">//------------------------------------------------------//</span>
<a name="l01117"></a>01117 
<a name="l01118"></a>01118 
<a name="l01120"></a>01120   <span class="keywordtype">void</span> UpdateThick();
<a name="l01121"></a>01121   
<a name="l01122"></a>01122   
<a name="l01123"></a>01123   
<a name="l01125"></a>01125   <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01126"></a>01126 
<a name="l01127"></a>01127 
<a name="l01128"></a>01128 
<a name="l01130"></a>01130   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l01131"></a>01131 
<a name="l01132"></a>01132 
<a name="l01133"></a>01133 
<a name="l01135"></a>01135   <span class="keywordtype">void</span> Center();
<a name="l01136"></a>01136 
<a name="l01137"></a>01137 
<a name="l01138"></a>01138 
<a name="l01140"></a>01140   <span class="keywordtype">void</span> Connect();
<a name="l01141"></a>01141 
<a name="l01142"></a>01142 
<a name="l01143"></a>01143 
<a name="l01145"></a>01145   <span class="keywordtype">void</span> Diff();
<a name="l01146"></a>01146 
<a name="l01147"></a>01147 
<a name="l01148"></a>01148 
<a name="l01150"></a>01150   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01151"></a>01151 
<a name="l01152"></a>01152 
<a name="l01153"></a>01153 
<a name="l01155"></a>01155   <span class="keywordtype">void</span> Divide();
<a name="l01156"></a>01156 
<a name="l01157"></a>01157 
<a name="l01158"></a>01158 
<a name="l01160"></a>01160   <span class="keywordtype">void</span> RunDivide();
<a name="l01161"></a>01161 
<a name="l01162"></a>01162 
<a name="l01163"></a>01163 
<a name="l01165"></a>01165   <span class="keywordtype">void</span> Fix();
<a name="l01166"></a>01166 
<a name="l01167"></a>01167 
<a name="l01168"></a>01168 
<a name="l01170"></a>01170   <span class="keywordtype">void</span> Intersect();
<a name="l01171"></a>01171 
<a name="l01172"></a>01172 
<a name="l01173"></a>01173 
<a name="l01175"></a>01175 
<a name="l01179"></a>01179   <span class="keywordtype">void</span> ModifyPos(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01180"></a>01180 
<a name="l01181"></a>01181 
<a name="l01182"></a>01182 
<a name="l01184"></a>01184   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01185"></a>01185 
<a name="l01186"></a>01186 
<a name="l01187"></a>01187 
<a name="l01189"></a>01189   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01190"></a>01190 
<a name="l01191"></a>01191 
<a name="l01192"></a>01192 
<a name="l01194"></a>01194 
<a name="l01199"></a>01199   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l01200"></a>01200 
<a name="l01201"></a>01201 
<a name="l01202"></a>01202 
<a name="l01204"></a>01204   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01205"></a>01205 
<a name="l01206"></a>01206 
<a name="l01207"></a>01207 
<a name="l01209"></a>01209   <span class="keywordtype">void</span> Sort(); 
<a name="l01210"></a>01210 
<a name="l01211"></a>01211 
<a name="l01212"></a>01212   
<a name="l01214"></a>01214   <span class="keywordtype">void</span> RunSplit(); 
<a name="l01215"></a>01215 
<a name="l01216"></a>01216 
<a name="l01217"></a>01217   
<a name="l01219"></a>01219   <span class="keywordtype">void</span> Union();
<a name="l01220"></a>01220 
<a name="l01221"></a>01221 
<a name="l01222"></a>01222 
<a name="l01224"></a>01224   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01225"></a>01225 
<a name="l01226"></a>01226 
<a name="l01227"></a>01227 
<a name="l01228"></a>01228   <span class="comment">//------------------------------------------------------//</span>
<a name="l01229"></a>01229   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l01230"></a>01230   <span class="comment">//------------------------------------------------------//</span>
<a name="l01231"></a>01231 
<a name="l01232"></a>01232   
<a name="l01234"></a>01234   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> *Constrain(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l01235"></a>01235 
<a name="l01236"></a>01236   
<a name="l01237"></a>01237 
<a name="l01239"></a>01239   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Diff(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l01240"></a>01240 
<a name="l01241"></a>01241   
<a name="l01242"></a>01242 
<a name="l01243"></a>01243   <span class="comment">//------------------------------------------------------//</span>
<a name="l01244"></a>01244   <span class="comment">//   Data                                               //</span>
<a name="l01245"></a>01245   <span class="comment">//------------------------------------------------------//</span>
<a name="l01246"></a>01246 
<a name="l01247"></a><a class="code" href="classGenomicRegionBED.html#a60878391f6fb1e0ad0ac9bc88a7625d3">01247</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a60878391f6fb1e0ad0ac9bc88a7625d3" title="number of tokens">n_tokens</a>;                    
<a name="l01248"></a><a class="code" href="classGenomicRegionBED.html#afa7ba810a0d1673f5dbabcf85c7bbb95">01248</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#afa7ba810a0d1673f5dbabcf85c7bbb95" title="score field">score</a>;                               
<a name="l01249"></a><a class="code" href="classGenomicRegionBED.html#a52fc75b347dfe6696e7ad2beeeb623a9">01249</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionBED.html#a52fc75b347dfe6696e7ad2beeeb623a9" title="color field">itemRgb</a>;                                
<a name="l01250"></a><a class="code" href="classGenomicRegionBED.html#a019dfaebedc17e8859b36c3359287faa">01250</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a019dfaebedc17e8859b36c3359287faa" title="thickStart field">thickStart</a>;                  
<a name="l01251"></a><a class="code" href="classGenomicRegionBED.html#a5333a7fdc1f21db508aa7af4f50553ba">01251</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a5333a7fdc1f21db508aa7af4f50553ba" title="thickEnd field">thickEnd</a>;                    
<a name="l01252"></a>01252 
<a name="l01253"></a>01253 };
<a name="l01254"></a>01254 
<a name="l01255"></a>01255 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01256"></a>01256 <span class="comment">// END CLASS: GenomicRegionBED                                                                 //</span>
<a name="l01257"></a>01257 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01258"></a>01258 
<a name="l01259"></a>01259 
<a name="l01260"></a>01260 
<a name="l01261"></a>01261 
<a name="l01262"></a>01262 
<a name="l01263"></a>01263 
<a name="l01264"></a>01264 
<a name="l01265"></a>01265 
<a name="l01266"></a>01266 
<a name="l01267"></a>01267 
<a name="l01268"></a>01268 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01269"></a>01269 <span class="comment">// CLASS: GenomicRegionSAMToREG                                                                //</span>
<a name="l01270"></a>01270 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01272"></a>01272 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01273"></a><a class="code" href="classGenomicRegionSAMToREG.html">01273</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01274"></a>01274 {
<a name="l01275"></a>01275  <span class="keyword">public</span>:
<a name="l01276"></a>01276   
<a name="l01278"></a>01278   <a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01279"></a>01279 
<a name="l01280"></a>01280 
<a name="l01281"></a>01281 
<a name="l01283"></a>01283   <a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01284"></a>01284 
<a name="l01285"></a>01285 
<a name="l01286"></a>01286 
<a name="l01288"></a>01288   ~<a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>();
<a name="l01289"></a>01289   
<a name="l01290"></a>01290 
<a name="l01291"></a>01291 
<a name="l01293"></a>01293   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01294"></a>01294   
<a name="l01295"></a>01295 
<a name="l01296"></a>01296 
<a name="l01298"></a>01298   <span class="keywordtype">char</span> CalcStrandFromFlag(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01299"></a>01299 
<a name="l01300"></a>01300 
<a name="l01301"></a>01301   
<a name="l01303"></a>01303   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetNextTokenOfCIGAR(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01304"></a>01304 
<a name="l01305"></a>01305   
<a name="l01306"></a>01306   
<a name="l01307"></a>01307   <span class="comment">// data</span>
<a name="l01308"></a>01308 
<a name="l01309"></a>01309 
<a name="l01310"></a>01310 };
<a name="l01311"></a>01311 
<a name="l01312"></a>01312 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01313"></a>01313 <span class="comment">// END CLASS: GenomicRegionSAMToREG                                                            //</span>
<a name="l01314"></a>01314 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01315"></a>01315 
<a name="l01316"></a>01316 
<a name="l01317"></a>01317 
<a name="l01318"></a>01318 
<a name="l01319"></a>01319 
<a name="l01320"></a>01320 
<a name="l01321"></a>01321 
<a name="l01322"></a>01322 
<a name="l01323"></a>01323 
<a name="l01324"></a>01324 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01325"></a>01325 <span class="comment">// CLASS: GenomicRegionSAM                                                                     //</span>
<a name="l01326"></a>01326 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01328"></a>01328 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01329"></a><a class="code" href="classGenomicRegionSAM.html">01329</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01330"></a>01330 {
<a name="l01331"></a>01331  <span class="keyword">public</span>:
<a name="l01332"></a><a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96">01332</a>   <span class="keyword">typedef</span> vector&lt;pair&lt;long int,char&gt; &gt; <a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length)">CIGARTokens</a>;                     
<a name="l01333"></a>01333   
<a name="l01335"></a>01335   <a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01336"></a>01336 
<a name="l01337"></a>01337 
<a name="l01338"></a>01338 
<a name="l01340"></a>01340   <a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01341"></a>01341 
<a name="l01342"></a>01342 
<a name="l01343"></a>01343 
<a name="l01345"></a>01345   ~<a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>();
<a name="l01346"></a>01346   
<a name="l01347"></a>01347 
<a name="l01348"></a>01348 
<a name="l01349"></a>01349   <span class="comment">//------------------------------------------------------//</span>
<a name="l01350"></a>01350   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01351"></a>01351   <span class="comment">//------------------------------------------------------//</span>
<a name="l01352"></a>01352 
<a name="l01353"></a>01353 
<a name="l01355"></a>01355   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01356"></a>01356 
<a name="l01357"></a>01357 
<a name="l01358"></a>01358 
<a name="l01360"></a>01360   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01361"></a>01361 
<a name="l01362"></a>01362 
<a name="l01363"></a>01363 
<a name="l01365"></a>01365   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01366"></a>01366 
<a name="l01367"></a>01367 
<a name="l01368"></a>01368 
<a name="l01370"></a>01370   <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01371"></a>01371 
<a name="l01372"></a>01372 
<a name="l01373"></a>01373   
<a name="l01374"></a>01374   <span class="comment">//------------------------------------------------------//</span>
<a name="l01375"></a>01375   <span class="comment">//   Auxiliary functions                                //</span>
<a name="l01376"></a>01376   <span class="comment">//------------------------------------------------------//</span>
<a name="l01377"></a>01377 
<a name="l01379"></a>01379   <span class="keywordtype">char</span> CalcStrandFromFlag(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01380"></a>01380 
<a name="l01381"></a>01381 
<a name="l01382"></a>01382   
<a name="l01384"></a>01384   <span class="keywordtype">void</span> UpdateFlagFromStrand(<span class="keywordtype">char</span> strand);
<a name="l01385"></a>01385   
<a name="l01386"></a>01386   
<a name="l01387"></a>01387   
<a name="l01389"></a>01389   <span class="keywordtype">void</span> PrintGappedSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01390"></a>01390 
<a name="l01391"></a>01391   
<a name="l01392"></a>01392   
<a name="l01394"></a>01394   <span class="keywordtype">char</span> *GetGappedSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01395"></a>01395 
<a name="l01396"></a>01396   
<a name="l01397"></a>01397   
<a name="l01399"></a>01399   <span class="keywordtype">void</span> InvalidateSeqData();
<a name="l01400"></a>01400 
<a name="l01401"></a>01401 
<a name="l01402"></a>01402   
<a name="l01404"></a>01404   <span class="keywordtype">void</span> TokenizeCIGAR();
<a name="l01405"></a>01405 
<a name="l01406"></a>01406   
<a name="l01407"></a>01407   
<a name="l01409"></a>01409   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetNextTokenOfCIGAR(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01410"></a>01410 
<a name="l01411"></a>01411   
<a name="l01412"></a>01412   
<a name="l01414"></a>01414   <span class="keywordtype">void</span> UpdateCigarFromIntervals();
<a name="l01415"></a>01415   
<a name="l01416"></a>01416   
<a name="l01417"></a>01417  
<a name="l01419"></a>01419   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcFragmentLengthFromCIGAR();
<a name="l01420"></a>01420   
<a name="l01421"></a>01421   
<a name="l01422"></a>01422   
<a name="l01424"></a>01424   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcReferenceLengthFromCIGAR();
<a name="l01425"></a>01425   
<a name="l01426"></a>01426   
<a name="l01427"></a>01427   
<a name="l01429"></a>01429   <span class="keywordtype">char</span> *TrimCIGAR(<span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01430"></a>01430  
<a name="l01431"></a>01431  
<a name="l01432"></a>01432  
<a name="l01434"></a>01434   <span class="keywordtype">char</span> *TrimSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01435"></a>01435  
<a name="l01436"></a>01436  
<a name="l01437"></a>01437  
<a name="l01438"></a>01438   <span class="comment">//------------------------------------------------------//</span>
<a name="l01439"></a>01439   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01440"></a>01440   <span class="comment">//------------------------------------------------------//</span>
<a name="l01441"></a>01441 
<a name="l01442"></a>01442 
<a name="l01444"></a>01444   <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01445"></a>01445 
<a name="l01446"></a>01446 
<a name="l01447"></a>01447   
<a name="l01449"></a>01449   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l01450"></a>01450 
<a name="l01451"></a>01451 
<a name="l01452"></a>01452 
<a name="l01454"></a>01454   <span class="keywordtype">void</span> Center();
<a name="l01455"></a>01455 
<a name="l01456"></a>01456 
<a name="l01457"></a>01457 
<a name="l01459"></a>01459   <span class="keywordtype">void</span> Connect();
<a name="l01460"></a>01460 
<a name="l01461"></a>01461 
<a name="l01462"></a>01462 
<a name="l01464"></a>01464   <span class="keywordtype">void</span> Diff();
<a name="l01465"></a>01465 
<a name="l01466"></a>01466 
<a name="l01467"></a>01467 
<a name="l01469"></a>01469   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01470"></a>01470 
<a name="l01471"></a>01471 
<a name="l01472"></a>01472 
<a name="l01474"></a>01474   <span class="keywordtype">void</span> Divide();
<a name="l01475"></a>01475 
<a name="l01476"></a>01476 
<a name="l01477"></a>01477 
<a name="l01479"></a>01479   <span class="keywordtype">void</span> RunDivide();
<a name="l01480"></a>01480 
<a name="l01481"></a>01481 
<a name="l01482"></a>01482 
<a name="l01484"></a>01484   <span class="keywordtype">void</span> Fix();
<a name="l01485"></a>01485 
<a name="l01486"></a>01486 
<a name="l01487"></a>01487 
<a name="l01489"></a>01489   <span class="keywordtype">void</span> Intersect();
<a name="l01490"></a>01490 
<a name="l01491"></a>01491 
<a name="l01492"></a>01492 
<a name="l01494"></a>01494   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01495"></a>01495 
<a name="l01496"></a>01496 
<a name="l01497"></a>01497 
<a name="l01499"></a>01499 
<a name="l01503"></a>01503   <span class="keywordtype">void</span> ModifyPos(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01504"></a>01504 
<a name="l01505"></a>01505 
<a name="l01506"></a>01506 
<a name="l01508"></a>01508   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01509"></a>01509 
<a name="l01510"></a>01510 
<a name="l01511"></a>01511 
<a name="l01513"></a>01513 
<a name="l01518"></a>01518   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l01519"></a>01519 
<a name="l01520"></a>01520 
<a name="l01521"></a>01521 
<a name="l01523"></a>01523   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01524"></a>01524 
<a name="l01525"></a>01525 
<a name="l01526"></a>01526 
<a name="l01528"></a>01528   <span class="keywordtype">void</span> Sort(); 
<a name="l01529"></a>01529 
<a name="l01530"></a>01530 
<a name="l01531"></a>01531   
<a name="l01533"></a>01533   <span class="keywordtype">void</span> RunSplit(); 
<a name="l01534"></a>01534 
<a name="l01535"></a>01535 
<a name="l01536"></a>01536   
<a name="l01538"></a>01538   <span class="keywordtype">void</span> ModifyStrand(<span class="keywordtype">char</span> *strand_op);
<a name="l01539"></a>01539 
<a name="l01540"></a>01540 
<a name="l01541"></a>01541   
<a name="l01543"></a>01543   <span class="keywordtype">void</span> Union();
<a name="l01544"></a>01544 
<a name="l01545"></a>01545 
<a name="l01546"></a>01546 
<a name="l01548"></a>01548   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01549"></a>01549 
<a name="l01550"></a>01550 
<a name="l01551"></a>01551 
<a name="l01552"></a>01552   <span class="comment">//------------------------------------------------------//</span>
<a name="l01553"></a>01553   <span class="comment">//   Data                                               //</span>
<a name="l01554"></a>01554   <span class="comment">//------------------------------------------------------//</span>
<a name="l01555"></a>01555 
<a name="l01556"></a><a class="code" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f">01556</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f" title="number of TAB-separated fields in SAM entry (&gt;= 12)">n_tokens</a>;                                    
<a name="l01557"></a>01557   <span class="comment">//char *QNAME;                                                //!&lt; Query template NAME: mapped to GenomicRegion::LABEL</span>
<a name="l01558"></a><a class="code" href="classGenomicRegionSAM.html#a200d7f7044dd69311eceb851a3c315e5">01558</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#a200d7f7044dd69311eceb851a3c315e5" title="bitwise FLAG">FLAG</a>;                               
<a name="l01559"></a>01559   <span class="comment">//char *RNAME;                                                //!&lt; Reference sequence NAME: mapped to GenomicRegion::I.front()-&gt;CHROMOSOME</span>
<a name="l01560"></a>01560   <span class="comment">//long int POS;                                               //!&lt; 1-based leftmost mapping POSition: mapped to GenomicRegion::I.front()-&gt;START</span>
<a name="l01561"></a><a class="code" href="classGenomicRegionSAM.html#aedacdd8319f4b614229a41b80d4c6f67">01561</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#aedacdd8319f4b614229a41b80d4c6f67" title="MAPping Quality.">MAPQ</a>;                                                
<a name="l01562"></a><a class="code" href="classGenomicRegionSAM.html#aa5acff481ae565db4976e66933212726">01562</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#aa5acff481ae565db4976e66933212726" title="CIGAR string.">CIGAR</a>;                                                  
<a name="l01563"></a><a class="code" href="classGenomicRegionSAM.html#af723505bfafff1a1fa616479023d0c3c">01563</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#af723505bfafff1a1fa616479023d0c3c" title="Reference name of the mate/next fragment.">RNEXT</a>;                                                  
<a name="l01564"></a><a class="code" href="classGenomicRegionSAM.html#ad6f88b4324ba508b86102641d819661f">01564</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#ad6f88b4324ba508b86102641d819661f" title="Position of the mate/next fragment.">PNEXT</a>;                                               
<a name="l01565"></a><a class="code" href="classGenomicRegionSAM.html#af180854a186be37eab49520edaa1262f">01565</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#af180854a186be37eab49520edaa1262f" title="observed Template LENgth">TLEN</a>;                                                
<a name="l01566"></a><a class="code" href="classGenomicRegionSAM.html#a6c2887ed1fa3a25d801641ea1b139032">01566</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#a6c2887ed1fa3a25d801641ea1b139032" title="fragment SEQuence">SEQ</a>;                                                    
<a name="l01567"></a><a class="code" href="classGenomicRegionSAM.html#ab43ca5f2fbfdee520f1424c3a9fd0706">01567</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#ab43ca5f2fbfdee520f1424c3a9fd0706" title="ASCII of Phred-scaled base QUALity+33.">QUAL</a>;                                                   
<a name="l01568"></a><a class="code" href="classGenomicRegionSAM.html#a1622a8804462bad83b1c49d25eeacb49">01568</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#a1622a8804462bad83b1c49d25eeacb49" title="all SAM fields after the 12th mandatory field">OPTIONAL</a>;                                               
<a name="l01569"></a><a class="code" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92">01569</a>   <a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length)">CIGARTokens</a> <a class="code" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92" title="CIGAR tokens.">T</a>;                                                
<a name="l01570"></a>01570   
<a name="l01571"></a>01571 };
<a name="l01572"></a>01572 
<a name="l01573"></a>01573 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01574"></a>01574 <span class="comment">// END CLASS: GenomicRegionSAM                                                                 //</span>
<a name="l01575"></a>01575 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01576"></a>01576 
<a name="l01577"></a>01577 
<a name="l01578"></a>01578 
<a name="l01579"></a>01579 
<a name="l01580"></a>01580 
<a name="l01581"></a>01581 
<a name="l01582"></a>01582 
<a name="l01583"></a>01583 
<a name="l01584"></a>01584 
<a name="l01585"></a>01585 
<a name="l01586"></a>01586 
<a name="l01587"></a>01587 
<a name="l01588"></a>01588 
<a name="l01589"></a>01589 
<a name="l01590"></a>01590 
<a name="l01591"></a>01591 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01592"></a>01592 <span class="comment">// CLASS: GenomicRegionGFFToREG                                                                //</span>
<a name="l01593"></a>01593 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01595"></a>01595 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01596"></a><a class="code" href="classGenomicRegionGFFToREG.html">01596</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01597"></a>01597 {
<a name="l01598"></a>01598  <span class="keyword">public</span>:
<a name="l01600"></a>01600   <a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01601"></a>01601 
<a name="l01602"></a>01602 
<a name="l01603"></a>01603 
<a name="l01605"></a>01605   <a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01606"></a>01606 
<a name="l01607"></a>01607 
<a name="l01608"></a>01608 
<a name="l01610"></a>01610   ~<a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>();
<a name="l01611"></a>01611   
<a name="l01612"></a>01612 
<a name="l01613"></a>01613 
<a name="l01615"></a>01615   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01616"></a>01616   
<a name="l01617"></a>01617 
<a name="l01618"></a>01618 
<a name="l01619"></a>01619   <span class="comment">// data</span>
<a name="l01620"></a>01620 
<a name="l01621"></a>01621 
<a name="l01622"></a>01622 };
<a name="l01623"></a>01623 
<a name="l01624"></a>01624 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01625"></a>01625 <span class="comment">// END CLASS: GenomicRegionGFFToREG                                                            //</span>
<a name="l01626"></a>01626 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01627"></a>01627 
<a name="l01628"></a>01628 
<a name="l01629"></a>01629 
<a name="l01630"></a>01630 
<a name="l01631"></a>01631 
<a name="l01632"></a>01632 
<a name="l01633"></a>01633 
<a name="l01634"></a>01634 
<a name="l01635"></a>01635 
<a name="l01636"></a>01636 
<a name="l01637"></a>01637 
<a name="l01638"></a>01638 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01639"></a>01639 <span class="comment">// CLASS: GenomicRegionGFF                                                                     //</span>
<a name="l01640"></a>01640 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01642"></a>01642 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01643"></a><a class="code" href="classGenomicRegionGFF.html">01643</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01644"></a>01644 {
<a name="l01645"></a>01645  <span class="keyword">public</span>:
<a name="l01647"></a>01647   <a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01648"></a>01648 
<a name="l01649"></a>01649 
<a name="l01650"></a>01650 
<a name="l01652"></a>01652   <a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01653"></a>01653 
<a name="l01654"></a>01654 
<a name="l01655"></a>01655 
<a name="l01657"></a>01657   ~<a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>();
<a name="l01658"></a>01658   
<a name="l01659"></a>01659 
<a name="l01660"></a>01660 
<a name="l01661"></a>01661   <span class="comment">//------------------------------------------------------//</span>
<a name="l01662"></a>01662   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01663"></a>01663   <span class="comment">//------------------------------------------------------//</span>
<a name="l01664"></a>01664 
<a name="l01665"></a>01665 
<a name="l01667"></a>01667   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01668"></a>01668 
<a name="l01669"></a>01669 
<a name="l01670"></a>01670 
<a name="l01672"></a>01672   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01673"></a>01673 
<a name="l01674"></a>01674 
<a name="l01675"></a>01675 
<a name="l01677"></a>01677   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01678"></a>01678 
<a name="l01679"></a>01679 
<a name="l01680"></a>01680 
<a name="l01682"></a>01682   <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01683"></a>01683 
<a name="l01684"></a>01684 
<a name="l01685"></a>01685 
<a name="l01687"></a>01687   <span class="keywordtype">void</span> Diff();
<a name="l01688"></a>01688 
<a name="l01689"></a>01689 
<a name="l01690"></a>01690 
<a name="l01692"></a>01692   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01693"></a>01693 
<a name="l01694"></a>01694 
<a name="l01695"></a>01695 
<a name="l01697"></a>01697   <span class="keywordtype">void</span> Divide();
<a name="l01698"></a>01698 
<a name="l01699"></a>01699 
<a name="l01700"></a>01700 
<a name="l01702"></a>01702   <span class="keywordtype">void</span> RunDivide();
<a name="l01703"></a>01703 
<a name="l01704"></a>01704 
<a name="l01705"></a>01705 
<a name="l01707"></a>01707   <span class="keywordtype">void</span> Intersect();
<a name="l01708"></a>01708 
<a name="l01709"></a>01709 
<a name="l01710"></a>01710 
<a name="l01712"></a>01712   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01713"></a>01713 
<a name="l01714"></a>01714 
<a name="l01715"></a>01715 
<a name="l01717"></a>01717   <span class="keywordtype">void</span> RunSplit(); 
<a name="l01718"></a>01718 
<a name="l01719"></a>01719 
<a name="l01720"></a>01720   
<a name="l01722"></a>01722   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01723"></a>01723 
<a name="l01724"></a>01724 
<a name="l01725"></a>01725 
<a name="l01726"></a>01726   <span class="comment">//---------------------------------------//</span>
<a name="l01727"></a>01727   <span class="comment">//  Data                                 //</span>
<a name="l01728"></a>01728   <span class="comment">//---------------------------------------//</span>
<a name="l01729"></a>01729   
<a name="l01730"></a><a class="code" href="classGenomicRegionGFF.html#aaec963f555e7e807fb5046566622f5ee">01730</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionGFF.html#aaec963f555e7e807fb5046566622f5ee" title="number of TAB-separated fields in GFF entry (= 11)">n_tokens</a>;                    
<a name="l01731"></a>01731   <span class="comment">//char *SEQNAME;                              //!&lt; sequence name: mapped to GenomicRegion::I-&gt;front()-&gt;CHROMOSOME</span>
<a name="l01732"></a><a class="code" href="classGenomicRegionGFF.html#a2f0d4a5f15d8a756233436af03341734">01732</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a2f0d4a5f15d8a756233436af03341734" title="feature source">SOURCE</a>;                                 
<a name="l01733"></a><a class="code" href="classGenomicRegionGFF.html#a0de447fbfad3da5b7a9cf07a5a6980f0">01733</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a0de447fbfad3da5b7a9cf07a5a6980f0" title="feture type">FEATURE</a>;                                
<a name="l01734"></a>01734   <span class="comment">//long int START;                             //!&lt; 1-based start position: mapped to GenomicRegion::I-&gt;front()-&gt;START</span>
<a name="l01735"></a>01735   <span class="comment">//long int END;                               //!&lt; 1-based end position: mapped to GenomicRegion::I-&gt;front()-&gt;STOP</span>
<a name="l01736"></a><a class="code" href="classGenomicRegionGFF.html#a0e92da1f0fd153504e0630f27229ba4d">01736</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a0e92da1f0fd153504e0630f27229ba4d" title="score">SCORE</a>;                                  
<a name="l01737"></a>01737   <span class="comment">//char STRAND;                                //!&lt; strand: mapped to GenomicRegion::I-&gt;front()-&gt;STRAND</span>
<a name="l01738"></a><a class="code" href="classGenomicRegionGFF.html#a42122a69eec385e0106e3861cb3b98f3">01738</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionGFF.html#a42122a69eec385e0106e3861cb3b98f3" title="coding frame">FRAME</a>;                                   
<a name="l01739"></a>01739   <span class="comment">//char *ATTRIBUTE;                    //!&lt; attribute: mapped to GenomicRegion::LABEL if non-empty</span>
<a name="l01740"></a><a class="code" href="classGenomicRegionGFF.html#a625a023f1bb24fc597e723919a036073">01740</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a625a023f1bb24fc597e723919a036073" title="comment">COMMENT</a>;                                
<a name="l01741"></a>01741 
<a name="l01742"></a>01742 };
<a name="l01743"></a>01743 
<a name="l01744"></a>01744 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01745"></a>01745 <span class="comment">// END CLASS: GenomicRegionGFF                                                                 //</span>
<a name="l01746"></a>01746 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01747"></a>01747 
<a name="l01748"></a>01748 
<a name="l01749"></a>01749 
<a name="l01750"></a>01750 
<a name="l01751"></a>01751 
<a name="l01752"></a>01752 
<a name="l01753"></a>01753 
<a name="l01754"></a>01754 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01755"></a>01755 <span class="comment">// CLASS: GenomicRegionSet                                                                     //</span>
<a name="l01756"></a>01756 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01758"></a>01758 <span class="comment"></span>
<a name="l01771"></a>01771 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01772"></a><a class="code" href="classGenomicRegionSet.html">01772</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>
<a name="l01773"></a>01773 {
<a name="l01774"></a>01774  <span class="keyword">public</span>:
<a name="l01776"></a>01776   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>(<span class="keywordtype">char</span> *file, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size, <span class="keywordtype">bool</span> verbose, <span class="keywordtype">bool</span> load_in_memory, <span class="keywordtype">bool</span> hide_header); 
<a name="l01777"></a>01777 
<a name="l01778"></a>01778 
<a name="l01780"></a>01780   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>(FILE *file_ptr, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size, <span class="keywordtype">bool</span> verbose, <span class="keywordtype">bool</span> load_in_memory, <span class="keywordtype">bool</span> hide_header);
<a name="l01781"></a>01781 
<a name="l01782"></a>01782 
<a name="l01784"></a>01784   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>();
<a name="l01785"></a>01785 
<a name="l01786"></a>01786 
<a name="l01787"></a>01787   <span class="comment">//-------------------------------------//</span>
<a name="l01788"></a>01788   <span class="comment">//  Read &amp; Print                       //</span>
<a name="l01789"></a>01789   <span class="comment">//-------------------------------------//</span>
<a name="l01790"></a>01790 
<a name="l01792"></a>01792   <span class="keywordtype">void</span> ProcessFileHeader(<span class="keywordtype">bool</span> hide);
<a name="l01793"></a>01793 
<a name="l01794"></a>01794   
<a name="l01796"></a>01796   <span class="keywordtype">void</span> DetectFileFormat();
<a name="l01797"></a>01797 
<a name="l01798"></a>01798   
<a name="l01800"></a>01800   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *CreateGenomicRegion(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *buffer);
<a name="l01801"></a>01801 
<a name="l01802"></a>01802 
<a name="l01804"></a>01804   <span class="keywordtype">void</span> Init();
<a name="l01805"></a>01805 
<a name="l01806"></a>01806 
<a name="l01808"></a>01808   <span class="keywordtype">void</span> Reset();
<a name="l01809"></a>01809 
<a name="l01810"></a>01810 
<a name="l01812"></a>01812   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Begin(); 
<a name="l01813"></a>01813 
<a name="l01814"></a>01814 
<a name="l01816"></a>01816   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Get();
<a name="l01817"></a>01817 
<a name="l01818"></a>01818 
<a name="l01820"></a>01820 
<a name="l01823"></a>01823   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Next(<span class="keywordtype">bool</span> retain_current=<span class="keyword">false</span>);
<a name="l01824"></a>01824 
<a name="l01825"></a>01825 
<a name="l01827"></a>01827 
<a name="l01831"></a>01831   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Next(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">bool</span> retain_current);
<a name="l01832"></a>01832 
<a name="l01833"></a>01833 
<a name="l01835"></a>01835   <span class="keywordtype">void</span> Release(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l01836"></a>01836 
<a name="l01837"></a>01837 
<a name="l01839"></a>01839   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetRetain();
<a name="l01840"></a>01840 
<a name="l01841"></a>01841 
<a name="l01843"></a>01843   <span class="keywordtype">void</span> PrintError(<span class="keywordtype">string</span> error_msg);
<a name="l01844"></a>01844   
<a name="l01845"></a>01845 
<a name="l01846"></a>01846 
<a name="l01847"></a>01847   <span class="comment">//-------------------------------------//</span>
<a name="l01848"></a>01848   <span class="comment">//  File-based (vertical) operations   //</span>
<a name="l01849"></a>01849   <span class="comment">//-------------------------------------//</span>
<a name="l01850"></a>01850 
<a name="l01851"></a>01851 
<a name="l01853"></a>01853   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CountRegions(<span class="keywordtype">bool</span> use_label_counts);
<a name="l01854"></a>01854 
<a name="l01855"></a>01855 
<a name="l01857"></a>01857   <span class="keywordtype">void</span> RunGlobalAnnotate(StringLIntMap *bounds);
<a name="l01858"></a>01858 
<a name="l01859"></a>01859 
<a name="l01861"></a>01861   <span class="keywordtype">void</span> RunGlobalCluster(<span class="keywordtype">bool</span> merge=<span class="keyword">false</span>);
<a name="l01862"></a>01862 
<a name="l01863"></a>01863 
<a name="l01865"></a>01865   <span class="keywordtype">void</span> RunGlobalCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01866"></a>01866 
<a name="l01867"></a>01867 
<a name="l01869"></a>01869   <span class="keywordtype">void</span> RunGlobalSort();
<a name="l01870"></a>01870 
<a name="l01871"></a>01871 
<a name="l01873"></a>01873   <span class="keywordtype">void</span> RunGlobalInvert(StringLIntMap *bounds);
<a name="l01874"></a>01874 
<a name="l01875"></a>01875 
<a name="l01877"></a>01877   <span class="keywordtype">void</span> RunGlobalLink(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> max_difference);
<a name="l01878"></a>01878 
<a name="l01879"></a>01879 
<a name="l01881"></a>01881   <span class="keywordtype">void</span> RunGlobalPartition();
<a name="l01882"></a>01882 
<a name="l01883"></a>01883 
<a name="l01885"></a>01885   <span class="keywordtype">void</span> RunGlobalTest(<span class="keywordtype">bool</span> sorted_by_strand);
<a name="l01886"></a>01886 
<a name="l01887"></a>01887 
<a name="l01889"></a>01889   <span class="keywordtype">void</span> RunGlobalMerge(<span class="keywordtype">long</span> <span class="keywordtype">int</span> merge_win);
<a name="l01890"></a>01890 
<a name="l01891"></a>01891 
<a name="l01893"></a>01893   <span class="keywordtype">void</span> RunGlobalReverseOrder();
<a name="l01894"></a>01894 
<a name="l01895"></a>01895 
<a name="l01897"></a>01897   <span class="keywordtype">void</span> PrintDensities(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> &amp;Index, <span class="keywordtype">bool</span> report_exact_boundaries);
<a name="l01898"></a>01898 
<a name="l01899"></a>01899 
<a name="l01901"></a>01901   <span class="keywordtype">void</span> RunGlobalScan(StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01902"></a>01902 
<a name="l01903"></a>01903 
<a name="l01905"></a>01905 
<a name="l01915"></a>01915   <span class="keywordtype">void</span> RunGlobalScanCount(StringLIntMap *bounds, <span class="keywordtype">char</span> *ref_reg_file, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> ignore_reverse_strand, <span class="keywordtype">char</span> preprocess, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">long</span> <span class="keywordtype">int</span> min_reads);
<a name="l01916"></a>01916 
<a name="l01917"></a>01917 
<a name="l01919"></a>01919   <span class="keywordtype">void</span> RunConvertToWIG(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">char</span> *options, <span class="keywordtype">long</span> <span class="keywordtype">int</span> span, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01920"></a>01920 
<a name="l01921"></a>01921 
<a name="l01922"></a>01922 
<a name="l01923"></a>01923 
<a name="l01924"></a>01924   <span class="comment">//---------------------------------------//</span>
<a name="l01925"></a>01925   <span class="comment">//  Line-based (horizontal) operations   //</span>
<a name="l01926"></a>01926   <span class="comment">//---------------------------------------//</span>
<a name="l01927"></a>01927 
<a name="l01929"></a>01929   <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01930"></a>01930 
<a name="l01931"></a>01931 
<a name="l01933"></a>01933   <span class="keywordtype">void</span> RunConvertToBED(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01934"></a>01934 
<a name="l01935"></a>01935 
<a name="l01937"></a>01937   <span class="keywordtype">void</span> RunBounds(StringLIntMap *bounds);
<a name="l01938"></a>01938 
<a name="l01939"></a>01939 
<a name="l01941"></a>01941   <span class="keywordtype">void</span> RunCenter();
<a name="l01942"></a>01942 
<a name="l01943"></a>01943 
<a name="l01945"></a>01945   <span class="keywordtype">void</span> RunConnect();
<a name="l01946"></a>01946 
<a name="l01947"></a>01947 
<a name="l01949"></a>01949   <span class="keywordtype">void</span> RunDiff();
<a name="l01950"></a>01950 
<a name="l01951"></a>01951 
<a name="l01953"></a>01953   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01954"></a>01954 
<a name="l01955"></a>01955 
<a name="l01957"></a>01957   <span class="keywordtype">void</span> RunDivide();
<a name="l01958"></a>01958 
<a name="l01959"></a>01959 
<a name="l01961"></a>01961   <span class="keywordtype">void</span> RunFix();
<a name="l01962"></a>01962 
<a name="l01963"></a>01963 
<a name="l01964"></a>01964 
<a name="l01966"></a>01966   <span class="keywordtype">void</span> RunIntersection();
<a name="l01967"></a>01967 
<a name="l01968"></a>01968 
<a name="l01969"></a>01969 
<a name="l01971"></a>01971   <span class="keywordtype">void</span> RunSize();
<a name="l01972"></a>01972 
<a name="l01973"></a>01973 
<a name="l01974"></a>01974 
<a name="l01976"></a>01976 
<a name="l01980"></a>01980   <span class="keywordtype">void</span> RunModifyPos(<span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01981"></a>01981 
<a name="l01982"></a>01982 
<a name="l01983"></a>01983 
<a name="l01985"></a>01985   <span class="keywordtype">void</span> RunConvertToREG(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l01986"></a>01986 
<a name="l01987"></a>01987 
<a name="l01988"></a>01988 
<a name="l01990"></a>01990   <span class="keywordtype">void</span> RunRandomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01991"></a>01991 
<a name="l01992"></a>01992 
<a name="l01993"></a>01993 
<a name="l01995"></a>01995   <span class="keywordtype">void</span> RunSelect(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01996"></a>01996 
<a name="l01997"></a>01997 
<a name="l01998"></a>01998 
<a name="l02000"></a>02000 
<a name="l02005"></a>02005   <span class="keywordtype">void</span> RunShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l02006"></a>02006 
<a name="l02007"></a>02007 
<a name="l02009"></a>02009   <span class="keywordtype">void</span> RunSort(); 
<a name="l02010"></a>02010 
<a name="l02011"></a>02011   
<a name="l02013"></a>02013   <span class="keywordtype">void</span> RunSplit(); 
<a name="l02014"></a>02014 
<a name="l02015"></a>02015   
<a name="l02017"></a>02017   <span class="keywordtype">void</span> RunModifyStrand(<span class="keywordtype">char</span> *strand_op, <span class="keywordtype">bool</span> sorted);
<a name="l02018"></a>02018 
<a name="l02019"></a>02019 
<a name="l02021"></a>02021   <span class="keywordtype">void</span> RunModifyStrandSorted(<span class="keywordtype">char</span> *strand_op);
<a name="l02022"></a>02022 
<a name="l02023"></a>02023 
<a name="l02025"></a>02025   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <span class="keywordtype">char</span> *ref_reg_file);
<a name="l02026"></a>02026 
<a name="l02027"></a>02027 
<a name="l02029"></a>02029   <span class="keywordtype">void</span> RunUnion();
<a name="l02030"></a>02030 
<a name="l02031"></a>02031 
<a name="l02033"></a>02033   <span class="keywordtype">void</span> RunSlidingWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l02034"></a>02034 
<a name="l02035"></a>02035 
<a name="l02037"></a>02037   <span class="keywordtype">void</span> RunExtractSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l02038"></a>02038 
<a name="l02039"></a>02039 
<a name="l02040"></a>02040 
<a name="l02041"></a>02041   <span class="comment">//---------------------------------------//</span>
<a name="l02042"></a>02042   <span class="comment">//  UNDER DEVELOPMENT/TESTING            //</span>
<a name="l02043"></a>02043   <span class="comment">//---------------------------------------//</span>
<a name="l02044"></a>02044 
<a name="l02045"></a>02045 
<a name="l02047"></a>02047   <span class="keywordtype">void</span> PrintBEDGraphFormat(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l02048"></a>02048 
<a name="l02049"></a>02049 
<a name="l02051"></a>02051   <span class="keywordtype">void</span> PrintSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02052"></a>02052 
<a name="l02053"></a>02053 
<a name="l02055"></a>02055   <span class="comment">//void PrintNoGaps();</span>
<a name="l02056"></a>02056 
<a name="l02057"></a>02057 
<a name="l02059"></a>02059   <span class="keywordtype">void</span> PrintRemoveN(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02060"></a>02060 
<a name="l02061"></a>02061 
<a name="l02063"></a>02063 
<a name="l02067"></a>02067   <span class="keywordtype">void</span> PrintOffsetFormat(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l02068"></a>02068 
<a name="l02069"></a>02069 
<a name="l02071"></a>02071   <span class="keywordtype">void</span> PrintReversePos(StringLIntMap *bounds);
<a name="l02072"></a>02072 
<a name="l02073"></a>02073 
<a name="l02075"></a>02075   <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l02076"></a>02076 
<a name="l02077"></a>02077 
<a name="l02079"></a>02079   <span class="keywordtype">void</span> PrintVerifySeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l02080"></a>02080 
<a name="l02081"></a>02081 
<a name="l02082"></a>02082 
<a name="l02083"></a>02083  <span class="keyword">private</span>:
<a name="l02085"></a>02085   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *StoreInTempFile();
<a name="l02086"></a>02086 
<a name="l02087"></a>02087 
<a name="l02089"></a>02089   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *PrintMergeSort(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *rtempSet, <span class="keywordtype">char</span> strand);
<a name="l02090"></a>02090 
<a name="l02091"></a>02091 
<a name="l02093"></a>02093   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *PrintReverse(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *rtempSet);
<a name="l02094"></a>02094 
<a name="l02095"></a>02095 
<a name="l02096"></a>02096 
<a name="l02097"></a>02097   <span class="comment">//---------------------------------------//</span>
<a name="l02098"></a>02098   <span class="comment">//  Data                                 //</span>
<a name="l02099"></a>02099   <span class="comment">//---------------------------------------//</span>
<a name="l02100"></a>02100 
<a name="l02101"></a>02101  <span class="keyword">public</span>:
<a name="l02102"></a><a class="code" href="classGenomicRegionSet.html#a7c593d1beab775590d5b810097644583">02102</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSet.html#a7c593d1beab775590d5b810097644583" title="file name (&#39;NULL&#39; if reading from standard input)">file</a>;                                           
<a name="l02103"></a><a class="code" href="classGenomicRegionSet.html#a4ca6c6268e39fdea2add6e6381c66962">02103</a>   FILE *<a class="code" href="classGenomicRegionSet.html#a4ca6c6268e39fdea2add6e6381c66962" title="file pointer (&#39;NULL&#39; if reading from standard input)">file_ptr</a>;                                       
<a name="l02104"></a><a class="code" href="classGenomicRegionSet.html#a4d60b5403d2d6aaeda541ed45625b245">02104</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#a4d60b5403d2d6aaeda541ed45625b245" title="line buffer size (automatically adjusted during execution to ensure the entire line is read)...">buffer_size</a>;        
<a name="l02105"></a><a class="code" href="classGenomicRegionSet.html#a2646e9e6d5da1693b4bea36a5acc082e">02105</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#a2646e9e6d5da1693b4bea36a5acc082e" title="verbose mode">verbose</a>;                                         
<a name="l02106"></a><a class="code" href="classGenomicRegionSet.html#abd0fc2542e717e5855e03d3b3892f027">02106</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#abd0fc2542e717e5855e03d3b3892f027" title="&#39;true&#39; if reading from standard input">from_stdin</a>;                                      
<a name="l02107"></a><a class="code" href="classGenomicRegionSet.html#ade43c6cd72b95da75395129b00a52687">02107</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#ade43c6cd72b95da75395129b00a52687" title="&#39;true&#39; if the entire region set is loaded in memory">load_in_memory</a>;                          
<a name="l02108"></a><a class="code" href="classGenomicRegionSet.html#a84dd2f8055524a933e963175781f7fd2">02108</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#a84dd2f8055524a933e963175781f7fd2" title="if &#39;true&#39;, then do not print file header">hide_header</a>;                                     
<a name="l02109"></a><a class="code" href="classGenomicRegionSet.html#aea8b5c9bae036da50d4e8720d2e13231">02109</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#aea8b5c9bae036da50d4e8720d2e13231" title="if &#39;true&#39;, use an array for storing genomic regions&#39; intervals (for time efficiency)">use_interval_array</a>;                      
<a name="l02110"></a><a class="code" href="classGenomicRegionSet.html#aa4494ed2ae3f7594c29b18c39fed6748">02110</a>   <span class="keywordtype">string</span> <a class="code" href="classGenomicRegionSet.html#aa4494ed2ae3f7594c29b18c39fed6748" title="input format">format</a>;                                        
<a name="l02111"></a><a class="code" href="classGenomicRegionSet.html#a08d9a35a7688ac2c193b53c9a8693c6e">02111</a>   <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> <a class="code" href="classGenomicRegionSet.html#a08d9a35a7688ac2c193b53c9a8693c6e" title="keeps track of progress of computation">progress</a>;                            
<a name="l02112"></a><a class="code" href="classGenomicRegionSet.html#adf3a5598e31ddbb5b381125925362085">02112</a>   <a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *<a class="code" href="classGenomicRegionSet.html#adf3a5598e31ddbb5b381125925362085" title="pointer to the buffer class used to read the region sets">buffer</a>;                           
<a name="l02113"></a><a class="code" href="classGenomicRegionSet.html#a5293292428aa06145b38be4b96ab05e4">02113</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#a5293292428aa06145b38be4b96ab05e4" title="number of regions (&#39;1&#39; if the regions are not loaded in memory)">n_regions</a>;                           
<a name="l02114"></a><a class="code" href="classGenomicRegionSet.html#a206d91b443832800afc04f786d15971d">02114</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> **<a class="code" href="classGenomicRegionSet.html#a206d91b443832800afc04f786d15971d" title="array where the regions are stored; if the region set is not entirely loaded in memory, only R[0] is used">R</a>;                            
<a name="l02115"></a><a class="code" href="classGenomicRegionSet.html#ad89185b88d5b37161cc65bc84b421c0e">02115</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#ad89185b88d5b37161cc65bc84b421c0e" title="index pointing to the current region in R">r_index</a>;                                     
<a name="l02116"></a>02116 };
<a name="l02117"></a>02117 
<a name="l02118"></a>02118 
<a name="l02119"></a>02119 
<a name="l02120"></a>02120 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02121"></a>02121 <span class="comment">// END CLASS: GenomicRegionSet                                                                 //</span>
<a name="l02122"></a>02122 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02123"></a>02123 
<a name="l02124"></a>02124 
<a name="l02125"></a>02125 
<a name="l02126"></a>02126 
<a name="l02127"></a>02127 
<a name="l02128"></a>02128 
<a name="l02129"></a>02129 
<a name="l02130"></a>02130 
<a name="l02131"></a>02131 
<a name="l02132"></a>02132 
<a name="l02133"></a>02133 
<a name="l02134"></a>02134 
<a name="l02135"></a>02135 
<a name="l02136"></a>02136 
<a name="l02137"></a>02137 
<a name="l02138"></a>02138 
<a name="l02139"></a>02139 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02140"></a>02140 <span class="comment">// CLASS: GenomicRegionSetScanner                                                              //</span>
<a name="l02141"></a>02141 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02143"></a>02143 <span class="comment"></span>
<a name="l02167"></a>02167 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02168"></a><a class="code" href="classGenomicRegionSetScanner.html">02168</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description below for an...">GenomicRegionSetScanner</a>
<a name="l02169"></a>02169 {
<a name="l02170"></a>02170  <span class="keyword">public</span>:
<a name="l02172"></a>02172 
<a name="l02181"></a>02181   <a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description below for an...">GenomicRegionSetScanner</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *R, StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">bool</span> ignore_reverse_strand, <span class="keywordtype">char</span> preprocess);
<a name="l02182"></a>02182   ~<a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description below for an...">GenomicRegionSetScanner</a>();
<a name="l02183"></a>02183   
<a name="l02184"></a>02184   <span class="comment">// operations</span>
<a name="l02185"></a>02185   <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next();                                                              
<a name="l02186"></a>02186   <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *Ref);                 
<a name="l02187"></a>02187   <span class="keywordtype">void</span> PrintInterval(FILE *out_file=stdout);    
<a name="l02188"></a>02188   
<a name="l02189"></a>02189   <span class="comment">// data</span>
<a name="l02190"></a>02190  <span class="keyword">private</span>: 
<a name="l02191"></a>02191   <span class="keywordtype">bool</span> Test();                                  
<a name="l02192"></a>02192 
<a name="l02193"></a><a class="code" href="classGenomicRegionSetScanner.html#af76bf4fef482886f23f58cfa12c3eea1">02193</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetScanner.html#af76bf4fef482886f23f58cfa12c3eea1" title="pointer to the GenomicRegionSet which will be scanned by sliding windows">R</a>;                          
<a name="l02194"></a><a class="code" href="classGenomicRegionSetScanner.html#a7f64551c26331cd4ead1ab01405f1b72">02194</a>   StringLIntMap *<a class="code" href="classGenomicRegionSetScanner.html#a7f64551c26331cd4ead1ab01405f1b72" title="chromosome sizes">bounds</a>;                        
<a name="l02195"></a><a class="code" href="classGenomicRegionSetScanner.html#ab278dfa27c3589865b9243156c1c727f">02195</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#ab278dfa27c3589865b9243156c1c727f" title="sliding window step">win_step</a>;                            
<a name="l02196"></a><a class="code" href="classGenomicRegionSetScanner.html#aeed625b2a12aa2f7900997c4e20cd9b5">02196</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#aeed625b2a12aa2f7900997c4e20cd9b5" title="sliding window size (must be a multiple of window step)">win_size</a>;                            
<a name="l02197"></a><a class="code" href="classGenomicRegionSetScanner.html#ad47603d08614ab993fe9036662c421e7">02197</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#ad47603d08614ab993fe9036662c421e7" title="if true, genomic region labels are assumed to be integers and are included in the counting...">use_labels_as_values</a>;                    
<a name="l02198"></a><a class="code" href="classGenomicRegionSetScanner.html#a10c1d22ae74e6f295cd17259f70eff8a">02198</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#a10c1d22ae74e6f295cd17259f70eff8a" title="if true, no sliding windows on the negative strand are reported">ignore_reverse_strand</a>;                   
<a name="l02199"></a><a class="code" href="classGenomicRegionSetScanner.html#af2a536e23023f96c57f51e2276b4be0e">02199</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#af2a536e23023f96c57f51e2276b4be0e" title="if &#39;1&#39;, only start position is counted; if &#39;p&#39;, all positions are counted; if &#39;c&#39;, center of interval is counted">preprocess</a>;                              
<a name="l02200"></a><a class="code" href="classGenomicRegionSetScanner.html#abab8a5c92ec438168e22eb6e33098aba">02200</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#abab8a5c92ec438168e22eb6e33098aba" title="win_size divided by win_step (note: each window is comprised of non-overlapping sub-windows of size w...">n_win_combine</a>;                       
<a name="l02201"></a><a class="code" href="classGenomicRegionSetScanner.html#a60dff3d69ed195c0492711ee295720c2">02201</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetScanner.html#a60dff3d69ed195c0492711ee295720c2" title="pointer to the most recently scanned region">r</a>;                             
<a name="l02202"></a><a class="code" href="classGenomicRegionSetScanner.html#a3f0ba2ab42e8029297f63ba5cf2f9b95">02202</a>   StringLIntMap::iterator <a class="code" href="classGenomicRegionSetScanner.html#a3f0ba2ab42e8029297f63ba5cf2f9b95" title="keeps track of current window&#39;s chromosome information">chr</a>;                  
<a name="l02203"></a><a class="code" href="classGenomicRegionSetScanner.html#a0b9c57901528e75d9d32f4f98a708f4b">02203</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#a0b9c57901528e75d9d32f4f98a708f4b" title="keeps track of current window&#39;s strand information">strand</a>;                                  
<a name="l02204"></a><a class="code" href="classGenomicRegionSetScanner.html#a305a53be4686bb873996e83b3c393701">02204</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#a305a53be4686bb873996e83b3c393701" title="keeps track of current window&#39;s start information">start</a>;                               
<a name="l02205"></a><a class="code" href="classGenomicRegionSetScanner.html#af39ff3f8e4f1a1ab55c004cd06c2378a">02205</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#af39ff3f8e4f1a1ab55c004cd06c2378a" title="keeps track of current window&#39;s stop information">stop</a>;                                
<a name="l02206"></a><a class="code" href="classGenomicRegionSetScanner.html#a19e26d0661ce5f380a72f3eb43b0b07f">02206</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classGenomicRegionSetScanner.html#a19e26d0661ce5f380a72f3eb43b0b07f" title="a vector that keeps track of values in the sub-windows (of size win_step) that comprise the window (o...">v</a>;                                  
<a name="l02207"></a><a class="code" href="classGenomicRegionSetScanner.html#afa341bbfce6cffdc2880c8f1c80abeb1">02207</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#afa341bbfce6cffdc2880c8f1c80abeb1" title="the k-th element of v to be updated next (in a round-robin manner)">k</a>;                                   
<a name="l02208"></a><a class="code" href="classGenomicRegionSetScanner.html#a80be27fad3a3456e7aecd71fb7439c5a">02208</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#a80be27fad3a3456e7aecd71fb7439c5a" title="keeps track of current window&#39;s value information (this is the sum of elements in vector v); it is se...">v_sum</a>;                               
<a name="l02209"></a>02209 };
<a name="l02210"></a>02210 
<a name="l02211"></a>02211 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02212"></a>02212 <span class="comment">// END CLASS: GenomicRegionSetScanner                                                          //</span>
<a name="l02213"></a>02213 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02214"></a>02214 
<a name="l02215"></a>02215 
<a name="l02216"></a>02216 
<a name="l02217"></a>02217 
<a name="l02218"></a>02218 
<a name="l02219"></a>02219 
<a name="l02220"></a>02220 
<a name="l02221"></a>02221 
<a name="l02222"></a>02222 
<a name="l02223"></a>02223 
<a name="l02224"></a>02224 
<a name="l02225"></a>02225 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02226"></a>02226 <span class="comment">// CLASS: GenomicRegionSetOverlaps                                                             //</span>
<a name="l02227"></a>02227 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02229"></a>02229 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02230"></a><a class="code" href="classGenomicRegionSetOverlaps.html">02230</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02231"></a>02231 {
<a name="l02232"></a>02232  <span class="keyword">public</span>:
<a name="l02234"></a>02234 
<a name="l02238"></a>02238   <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet);
<a name="l02239"></a>02239 
<a name="l02240"></a>02240 
<a name="l02242"></a>02242   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>();
<a name="l02243"></a>02243 
<a name="l02244"></a>02244   
<a name="l02246"></a>02246   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery() = 0;
<a name="l02247"></a>02247 
<a name="l02248"></a>02248 
<a name="l02250"></a>02250   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery() = 0;
<a name="l02251"></a>02251 
<a name="l02252"></a>02252 
<a name="l02254"></a>02254   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch() = 0;
<a name="l02255"></a>02255 
<a name="l02256"></a>02256  
<a name="l02258"></a>02258   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch() = 0;
<a name="l02259"></a>02259 
<a name="l02260"></a>02260  
<a name="l02262"></a>02262   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> Done() = 0;
<a name="l02263"></a>02263  
<a name="l02264"></a>02264 
<a name="l02266"></a>02266 
<a name="l02270"></a>02270   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetOverlap(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02271"></a>02271 
<a name="l02272"></a>02272 
<a name="l02274"></a>02274 
<a name="l02278"></a>02278   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextOverlap(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02279"></a>02279 
<a name="l02280"></a>02280   
<a name="l02282"></a>02282 
<a name="l02287"></a>02287   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcQueryCoverage(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02288"></a>02288 
<a name="l02289"></a>02289 
<a name="l02291"></a>02291 
<a name="l02296"></a>02296   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *CalcIndexCoverage(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02297"></a>02297 
<a name="l02298"></a>02298 
<a name="l02300"></a>02300 
<a name="l02305"></a>02305   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CountQueryOverlaps(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02306"></a>02306 
<a name="l02307"></a>02307 
<a name="l02309"></a>02309 
<a name="l02314"></a>02314   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *CountIndexOverlaps(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02315"></a>02315 
<a name="l02316"></a>02316  
<a name="l02317"></a>02317   
<a name="l02318"></a>02318   <span class="comment">// data</span>
<a name="l02319"></a>02319  <span class="keyword">public</span>: 
<a name="l02320"></a><a class="code" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2">02320</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2" title="pointer to query region set">QuerySet</a>;                           
<a name="l02321"></a><a class="code" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03">02321</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03" title="pointer to index region set">IndexSet</a>;                           
<a name="l02322"></a><a class="code" href="classGenomicRegionSetOverlaps.html#a31d1dbc3407c616a0bda0f42af7acbae">02322</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#a31d1dbc3407c616a0bda0f42af7acbae" title="pointer to the current query region">current_qreg</a>;                          
<a name="l02323"></a><a class="code" href="classGenomicRegionSetOverlaps.html#a537336b7c40b20e6f5113395c11614b6">02323</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#a537336b7c40b20e6f5113395c11614b6" title="pointer to the current index region">current_ireg</a>;                          
<a name="l02324"></a>02324    
<a name="l02325"></a>02325 };
<a name="l02326"></a>02326 
<a name="l02327"></a>02327 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02328"></a>02328 <span class="comment">// END CLASS: GenomicRegionSetOverlaps                                                         //</span>
<a name="l02329"></a>02329 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02330"></a>02330 
<a name="l02331"></a>02331 
<a name="l02332"></a>02332 
<a name="l02333"></a>02333 
<a name="l02334"></a>02334 
<a name="l02335"></a>02335 
<a name="l02336"></a>02336 
<a name="l02337"></a>02337 
<a name="l02338"></a>02338 
<a name="l02339"></a>02339 
<a name="l02340"></a>02340 
<a name="l02341"></a>02341 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02342"></a>02342 <span class="comment">// CLASS: UnsortedGenomicRegionSetOverlaps                                                     //</span>
<a name="l02343"></a>02343 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02345"></a>02345 <span class="comment"></span>
<a name="l02373"></a>02373 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02374"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html">02374</a> <span class="keyword">class </span><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02375"></a>02375 {
<a name="l02376"></a>02376  <span class="keyword">public</span>:
<a name="l02378"></a>02378 
<a name="l02383"></a>02383   <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet, <span class="keywordtype">char</span> *bin_bits=NULL);
<a name="l02384"></a>02384 
<a name="l02385"></a>02385 
<a name="l02387"></a>02387   ~<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a>();
<a name="l02388"></a>02388 
<a name="l02389"></a>02389   
<a name="l02391"></a>02391   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery();
<a name="l02392"></a>02392 
<a name="l02393"></a>02393 
<a name="l02395"></a>02395   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery();
<a name="l02396"></a>02396 
<a name="l02397"></a>02397 
<a name="l02399"></a>02399   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch();
<a name="l02400"></a>02400 
<a name="l02401"></a>02401  
<a name="l02403"></a>02403   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02404"></a>02404 
<a name="l02405"></a>02405  
<a name="l02407"></a>02407   <span class="keywordtype">bool</span> Done();
<a name="l02408"></a>02408  
<a name="l02409"></a>02409 
<a name="l02410"></a>02410   
<a name="l02411"></a>02411   <span class="comment">// data</span>
<a name="l02412"></a>02412  <span class="keyword">public</span>: 
<a name="l02413"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14">02413</a>   <span class="keyword">typedef</span> pair&lt;long int*,long int**&gt; <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14" title="a two-dimensional grid of bins">BinSet</a>;            
<a name="l02414"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a251e09b7f0a13a56b902c30791228ce4">02414</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a251e09b7f0a13a56b902c30791228ce4" title="for each region store, store a pointer to the &quot;next&quot; one in the same bin">r_next</a>;                                     
<a name="l02415"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a1cc721de471ed40dcc7d8eae258e5833">02415</a>   <span class="keywordtype">int</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a1cc721de471ed40dcc7d8eae258e5833" title="number of levels in the two-dimensional grid of bins">n_levels</a>;                                         
<a name="l02416"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a76ecebfcf839a940bdae26069ceef748">02416</a>   <span class="keywordtype">int</span> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a76ecebfcf839a940bdae26069ceef748" title="number of bits (per level) by which the start position is right-shifted to compute its bin...">n_bits</a>;                                          
<a name="l02417"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abeeb94562abdf4aebc36684959a5bc78">02417</a>   map&lt;string,BinSet*&gt; <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abeeb94562abdf4aebc36684959a5bc78" title="the two-dimensional grid of bins per chromosome">index</a>;                            
<a name="l02418"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a6bbb87f3c66d04527f06a5761df1ab56">02418</a>   <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14" title="a two-dimensional grid of bins">BinSet</a> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a6bbb87f3c66d04527f06a5761df1ab56" title="pointer to the current chromosome set of bins">current_binset</a>;                               
<a name="l02419"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abafc544ce2c677ce071fa4e30ec2ecaf">02419</a>   <span class="keywordtype">bool</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abafc544ce2c677ce071fa4e30ec2ecaf" title="&#39;true&#39;, if a new query is about to be processed">new_query</a>;                                       
<a name="l02420"></a>02420 };
<a name="l02421"></a>02421 
<a name="l02422"></a>02422 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02423"></a>02423 <span class="comment">// END CLASS: UnsortedGenomicRegionSetOverlaps                                                 //</span>
<a name="l02424"></a>02424 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02425"></a>02425 
<a name="l02426"></a>02426 
<a name="l02427"></a>02427 
<a name="l02428"></a>02428 
<a name="l02429"></a>02429 
<a name="l02430"></a>02430 
<a name="l02431"></a>02431 
<a name="l02432"></a>02432 
<a name="l02433"></a>02433 
<a name="l02434"></a>02434 
<a name="l02435"></a>02435 
<a name="l02436"></a>02436 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02437"></a>02437 <span class="comment">// CLASS: SortedGenomicRegionSetOverlaps                                                       //</span>
<a name="l02438"></a>02438 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02440"></a>02440 <span class="comment"></span>
<a name="l02499"></a>02499 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02500"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html">02500</a> <span class="keyword">class </span><a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02501"></a>02501 {
<a name="l02502"></a>02502  <span class="keyword">public</span>:
<a name="l02504"></a>02504 
<a name="l02509"></a>02509   <a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l02510"></a>02510 
<a name="l02511"></a>02511 
<a name="l02513"></a>02513   ~<a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a>();
<a name="l02514"></a>02514 
<a name="l02515"></a>02515   
<a name="l02517"></a>02517   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery();
<a name="l02518"></a>02518 
<a name="l02519"></a>02519 
<a name="l02521"></a>02521   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery();
<a name="l02522"></a>02522 
<a name="l02523"></a>02523 
<a name="l02525"></a>02525   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch();
<a name="l02526"></a>02526 
<a name="l02527"></a>02527  
<a name="l02529"></a>02529   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02530"></a>02530 
<a name="l02531"></a>02531  
<a name="l02533"></a>02533   <span class="keywordtype">bool</span> Done();
<a name="l02534"></a>02534  
<a name="l02535"></a>02535 
<a name="l02536"></a>02536  <span class="keyword">private</span>:
<a name="l02537"></a>02537   <span class="keywordtype">void</span> LoadIndexBuffer();                               
<a name="l02538"></a>02538   <span class="keywordtype">void</span> ClearIndexBuffer();                              
<a name="l02539"></a>02539 
<a name="l02540"></a>02540   <span class="comment">// data</span>
<a name="l02541"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3">02541</a>   <span class="keywordtype">bool</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3" title="if true, query and index regions sorted by strand">sorted_by_strand</a>;                                
<a name="l02542"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">02542</a>   GenomicRegionList <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f" title="temporary buffer containing index regions that overlap with current query; the buffer is cleared only...">IRegBuffer</a>;                         
<a name="l02543"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a71d6eea0c4bf9f19be6cf0c0c562f297">02543</a>   GenomicRegionList::iterator <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a71d6eea0c4bf9f19be6cf0c0c562f297" title="iterator on SortedGenomicRegionSetOverlaps::IRegBuffer">IRegBufferIterator</a>;       
<a name="l02544"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a962f8e6ce8c15703cfefa574a07ce42c">02544</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classSortedGenomicRegionSetOverlaps.html#a962f8e6ce8c15703cfefa574a07ce42c" title="keeps tracks of the (linked) interval stored in SortedGenomicRegionSetOverlaps::IRegBuffer">ireg_buffer_interval</a>;                
<a name="l02545"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a349d589ff006f742d635b16987fe7b6b">02545</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a349d589ff006f742d635b16987fe7b6b" title="keeps track of the maximum size of SortedGenomicRegionSetOverlaps::IRegBuffer">max_ireg_buffer_size</a>;               
<a name="l02546"></a>02546 };
<a name="l02547"></a>02547 
<a name="l02548"></a>02548 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02549"></a>02549 <span class="comment">// END CLASS: SortedGenomicRegionSetOverlaps                                                   //</span>
<a name="l02550"></a>02550 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02551"></a>02551 
<a name="l02552"></a>02552 
<a name="l02553"></a>02553 
<a name="l02554"></a>02554 
<a name="l02555"></a>02555 
<a name="l02556"></a>02556 
<a name="l02557"></a>02557 
<a name="l02558"></a>02558 
<a name="l02559"></a>02559 
<a name="l02560"></a>02560 
<a name="l02561"></a>02561 <span class="keywordtype">char</span> ProcessStrand(<span class="keywordtype">char</span> *strand);
<a name="l02562"></a>02562 <span class="keywordtype">void</span> PrintChromosome(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">bool</span> convert);
<a name="l02563"></a>02563 <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *RegCenter(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l02564"></a>02564 <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetStart(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> preprocess);
<a name="l02565"></a>02565 
<a name="l02566"></a>02566 StringLIntMap *ReadBounds(<span class="keywordtype">char</span> *genome_reg_file, <span class="keywordtype">bool</span> verbose=<span class="keyword">false</span>);           
<a name="l02567"></a>02567 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcBoundSize(StringLIntMap *bounds);
<a name="l02568"></a>02568 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcRegSize(<span class="keywordtype">char</span> *reg_file);
<a name="l02569"></a>02569 <span class="keywordtype">bool</span> CompareGenomicIntervals(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *J);
<a name="l02570"></a>02570 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsList *L);
<a name="l02571"></a>02571 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsVector *V);
<a name="l02572"></a>02572 <span class="keywordtype">void</span> SortGenomicIntervals(<a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a> *A);
<a name="l02573"></a>02573 
<a name="l02574"></a>02574 
<a name="l02575"></a>02575 
<a name="l02576"></a>02576 
<a name="l02577"></a>02577 
</pre></div></div>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Thu Apr 5 2012 09:34:48 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
