<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>genomic_intervals.h Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
<div class="header">
  <div class="headertitle">
<div class="title">genomic_intervals.h</div>  </div>
</div>
<div class="contents">
<div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">//</span>
<a name="l00002"></a>00002 <span class="comment">// Copyright (c) 2011 IBM Corporation. </span>
<a name="l00003"></a>00003 <span class="comment">// All rights reserved. This program and the accompanying materials are made available under the terms of the GNU General Public License v2.0 </span>
<a name="l00004"></a>00004 <span class="comment">// which accompanies this distribution, and is available at http://www.gnu.org/licenses/old-licenses/gpl-2.0.html</span>
<a name="l00005"></a>00005 <span class="comment">//</span>
<a name="l00006"></a>00006 
<a name="l00007"></a>00007 <span class="keyword">const</span> <span class="keywordtype">string</span> VERSION = <span class="stringliteral">&quot;genomic_tools 2.3.2&quot;</span>;
<a name="l00008"></a>00008 
<a name="l00009"></a>00009 
<a name="l00010"></a>00010 <span class="preprocessor">#include &lt;stdio.h&gt;</span>
<a name="l00011"></a>00011 <span class="preprocessor">#include &lt;stdlib.h&gt;</span>
<a name="l00012"></a>00012 <span class="preprocessor">#include &lt;string.h&gt;</span>
<a name="l00013"></a>00013 <span class="preprocessor">#include &lt;math.h&gt;</span>
<a name="l00014"></a>00014 <span class="preprocessor">#include &lt;unistd.h&gt;</span>
<a name="l00015"></a>00015 <span class="preprocessor">#include &lt;ctype.h&gt;</span>
<a name="l00016"></a>00016 <span class="preprocessor">#include &lt;sys/types.h&gt;</span>
<a name="l00017"></a>00017 <span class="preprocessor">#include &lt;sys/stat.h&gt;</span>
<a name="l00018"></a>00018 <span class="preprocessor">#include &lt;map&gt;</span>
<a name="l00019"></a>00019 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00020"></a>00020 <span class="preprocessor">#include &lt;list&gt;</span>
<a name="l00021"></a>00021 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00022"></a>00022 <span class="preprocessor">#include &lt;iostream&gt;</span>
<a name="l00023"></a>00023 <span class="preprocessor">#include &lt;algorithm&gt;</span>
<a name="l00024"></a>00024 
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 <span class="keyword">using namespace </span>std;
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <span class="comment">// forward declarations</span>
<a name="l00029"></a>00029 <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>;
<a name="l00030"></a>00030 <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>;
<a name="l00031"></a>00031 <span class="keyword">typedef</span> vector&lt;GenomicInterval *&gt; GenomicIntervalSetAsVector;
<a name="l00032"></a>00032 <span class="keyword">typedef</span> list&lt;GenomicInterval *&gt; GenomicIntervalSetAsList;
<a name="l00033"></a>00033 <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>;
<a name="l00034"></a>00034 <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>;
<a name="l00035"></a>00035 <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>;
<a name="l00036"></a>00036 
<a name="l00037"></a>00037 <span class="comment">// types</span>
<a name="l00038"></a>00038 <span class="keyword">typedef</span> map&lt;string,string&gt; StringMap;
<a name="l00039"></a>00039 <span class="keyword">typedef</span> map&lt;string,long int&gt; StringLIntMap;
<a name="l00040"></a>00040 <span class="keyword">typedef</span> map&lt;string,vector&lt;long int&gt; &gt; StringVecLIntMap;
<a name="l00041"></a>00041 
<a name="l00042"></a>00042 <span class="comment">// choose container type for interval set (vector, list or array)</span>
<a name="l00043"></a>00043 <span class="keyword">typedef</span> GenomicIntervalSetAsVector GenomicIntervalSet;
<a name="l00044"></a>00044 <span class="comment">//typedef GenomicIntervalSetAsList GenomicIntervalSet;</span>
<a name="l00045"></a>00045 <span class="comment">//typedef GenomicIntervalSetAsArray GenomicIntervalSet;</span>
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 <span class="comment">// for buffer in SortedGenomicRegionSetOverlaps</span>
<a name="l00048"></a>00048 <span class="keyword">typedef</span> list&lt;GenomicRegion *&gt; GenomicRegionList;
<a name="l00049"></a>00049 
<a name="l00050"></a>00050 
<a name="l00051"></a>00051 <span class="comment">// constants</span>
<a name="l00052"></a>00052 <span class="preprocessor">#define ALIGNMENT_GAP &#39;*&#39;                                       //! character for alignment gaps </span>
<a name="l00053"></a>00053 <span class="preprocessor"></span><span class="keyword">const</span> <span class="keywordtype">char</span> refCIGARop[] = <span class="stringliteral">&quot;MD=X&quot;</span>;                       
<a name="l00054"></a>00054 <span class="keyword">const</span> <span class="keywordtype">char</span> fragCIGARop[] = <span class="stringliteral">&quot;MIS=X&quot;</span>;                     
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 
<a name="l00057"></a>00057 
<a name="l00058"></a>00058   
<a name="l00059"></a>00059   
<a name="l00060"></a>00060 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00061"></a>00061 <span class="comment">// CLASS DECLARATION: GenomicInterval                                                          //</span>
<a name="l00062"></a>00062 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00064"></a>00064 <span class="comment"></span>
<a name="l00078"></a>00078 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00079"></a><a class="code" href="classGenomicInterval.html">00079</a> <span class="keyword">class </span><a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>
<a name="l00080"></a>00080 {
<a name="l00081"></a>00081  <span class="keyword">public</span>:
<a name="l00083"></a>00083   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00084"></a>00084 
<a name="l00085"></a>00085   
<a name="l00086"></a>00086 
<a name="l00088"></a>00088   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<span class="keywordtype">string</span> chromosome, <span class="keywordtype">char</span> strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00089"></a>00089 
<a name="l00090"></a>00090   
<a name="l00091"></a>00091 
<a name="l00093"></a>00093   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">char</span> delimiter=<span class="charliteral">&#39; &#39;</span>, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l00094"></a>00094 
<a name="l00095"></a>00095   
<a name="l00096"></a>00096 
<a name="l00098"></a>00098   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00099"></a>00099 
<a name="l00100"></a>00100   
<a name="l00101"></a>00101   
<a name="l00103"></a>00103   ~<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>();
<a name="l00104"></a>00104 
<a name="l00105"></a>00105   
<a name="l00106"></a>00106 
<a name="l00107"></a>00107   <span class="comment">//------------------------------------------------------//</span>
<a name="l00108"></a>00108   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00109"></a>00109   <span class="comment">//------------------------------------------------------//</span>
<a name="l00110"></a>00110 
<a name="l00111"></a>00111 
<a name="l00113"></a>00113   <span class="keywordtype">void</span> PrintInterval();
<a name="l00114"></a>00114 
<a name="l00115"></a>00115   
<a name="l00116"></a>00116 
<a name="l00118"></a>00118   <span class="keywordtype">void</span> PrintInterval(FILE *file_ptr);
<a name="l00119"></a>00119 
<a name="l00120"></a>00120   
<a name="l00121"></a>00121 
<a name="l00123"></a>00123   <span class="keywordtype">void</span> PrintInterval(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00124"></a>00124 
<a name="l00125"></a>00125 
<a name="l00127"></a>00127   <span class="keywordtype">void</span> Print(<span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00128"></a>00128 
<a name="l00129"></a>00129   
<a name="l00130"></a>00130 
<a name="l00132"></a>00132   <span class="keywordtype">void</span> PrintWithLabel(<span class="keywordtype">char</span> *label, <span class="keywordtype">bool</span> point=<span class="keyword">false</span>);
<a name="l00133"></a>00133 
<a name="l00134"></a>00134   
<a name="l00135"></a>00135   
<a name="l00136"></a>00136   <span class="comment">//------------------------------------------------------//</span>
<a name="l00137"></a>00137   <span class="comment">//   Check                                              //</span>
<a name="l00138"></a>00138   <span class="comment">//------------------------------------------------------//</span>
<a name="l00139"></a>00139 
<a name="l00140"></a>00140 
<a name="l00142"></a>00142   <span class="keywordtype">bool</span> IsValid();
<a name="l00143"></a>00143 
<a name="l00144"></a>00144   
<a name="l00145"></a>00145 
<a name="l00147"></a>00147   <span class="keywordtype">bool</span> CheckValid(<span class="keywordtype">bool</span> quiet=<span class="keyword">false</span>);
<a name="l00148"></a>00148 
<a name="l00149"></a>00149   
<a name="l00150"></a>00150 
<a name="l00152"></a>00152   <span class="keywordtype">bool</span> IsBefore(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00153"></a>00153 
<a name="l00154"></a>00154 
<a name="l00155"></a>00155 
<a name="l00157"></a>00157   <span class="keywordtype">bool</span> IsPosBefore(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00158"></a>00158 
<a name="l00159"></a>00159 
<a name="l00160"></a>00160 
<a name="l00162"></a>00162   <span class="keywordtype">bool</span> IsCompatibleWith(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00163"></a>00163 
<a name="l00164"></a>00164 
<a name="l00165"></a>00165 
<a name="l00167"></a>00167   <span class="keywordtype">bool</span> OverlapsWith(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand=<span class="keyword">false</span>);
<a name="l00168"></a>00168 
<a name="l00169"></a>00169 
<a name="l00170"></a>00170 
<a name="l00171"></a>00171   <span class="comment">//------------------------------------------------------//</span>
<a name="l00172"></a>00172   <span class="comment">//   Set &amp; Get                                          //</span>
<a name="l00173"></a>00173   <span class="comment">//------------------------------------------------------//</span>
<a name="l00174"></a>00174 
<a name="l00175"></a>00175 
<a name="l00177"></a>00177   <span class="keywordtype">void</span> SetCoordinates(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00178"></a>00178 
<a name="l00179"></a>00179   
<a name="l00180"></a>00180 
<a name="l00182"></a>00182   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetCoordinate(<span class="keyword">const</span> <span class="keywordtype">char</span> *op);
<a name="l00183"></a>00183   
<a name="l00184"></a>00184 
<a name="l00185"></a>00185 
<a name="l00187"></a>00187   <span class="keywordtype">size_t</span> GetSize();
<a name="l00188"></a>00188   
<a name="l00189"></a>00189 
<a name="l00190"></a>00190 
<a name="l00192"></a>00192   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00193"></a>00193   
<a name="l00194"></a>00194 
<a name="l00195"></a>00195 
<a name="l00197"></a>00197   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00198"></a>00198   
<a name="l00199"></a>00199 
<a name="l00200"></a>00200 
<a name="l00201"></a>00201   <span class="comment">//------------------------------------------------------//</span>
<a name="l00202"></a>00202   <span class="comment">//   Operations                                         //</span>
<a name="l00203"></a>00203   <span class="comment">//------------------------------------------------------//</span>
<a name="l00204"></a>00204 
<a name="l00205"></a>00205 
<a name="l00207"></a>00207   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcOverlap(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00208"></a>00208 
<a name="l00209"></a>00209 
<a name="l00210"></a>00210 
<a name="l00212"></a>00212   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcDistanceFrom(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keyword">const</span> <span class="keywordtype">char</span> *op, <span class="keyword">const</span> <span class="keywordtype">char</span> *I_op);
<a name="l00213"></a>00213   
<a name="l00214"></a>00214 
<a name="l00215"></a>00215 
<a name="l00217"></a>00217   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00218"></a>00218   
<a name="l00219"></a>00219 
<a name="l00220"></a>00220 
<a name="l00222"></a>00222   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00223"></a>00223   
<a name="l00224"></a>00224 
<a name="l00225"></a>00225 
<a name="l00227"></a>00227   <span class="keywordtype">void</span> ReverseStrand(); 
<a name="l00228"></a>00228   
<a name="l00229"></a>00229 
<a name="l00230"></a>00230 
<a name="l00232"></a>00232 
<a name="l00237"></a>00237   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l00238"></a>00238   
<a name="l00239"></a>00239 
<a name="l00240"></a>00240 
<a name="l00242"></a>00242 
<a name="l00246"></a>00246   <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00247"></a>00247 
<a name="l00248"></a>00248 
<a name="l00249"></a>00249 
<a name="l00251"></a>00251   <span class="keywordtype">void</span> ReversePos(StringLIntMap *bounds);
<a name="l00252"></a>00252 
<a name="l00253"></a>00253 
<a name="l00254"></a>00254 
<a name="l00256"></a>00256   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l00257"></a>00257 
<a name="l00258"></a>00258 
<a name="l00259"></a>00259 
<a name="l00261"></a>00261   <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00262"></a>00262 
<a name="l00263"></a>00263 
<a name="l00264"></a>00264 
<a name="l00266"></a>00266   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00267"></a>00267 
<a name="l00268"></a>00268 
<a name="l00269"></a>00269 
<a name="l00271"></a>00271 
<a name="l00278"></a>00278   <span class="keywordtype">void</span> GetOffsetFrom(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *ReferenceI, <span class="keyword">const</span> <span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *start_offset, <span class="keywordtype">long</span> <span class="keywordtype">int</span> *stop_offset);
<a name="l00279"></a>00279 
<a name="l00280"></a>00280 
<a name="l00281"></a>00281   
<a name="l00282"></a>00282 
<a name="l00283"></a>00283   <span class="comment">//------------------------------------------------------//</span>
<a name="l00284"></a>00284   <span class="comment">//   Create                                             //</span>
<a name="l00285"></a>00285   <span class="comment">//------------------------------------------------------//</span>
<a name="l00286"></a>00286 
<a name="l00288"></a>00288 
<a name="l00291"></a>00291   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *CreateCenter(<span class="keywordtype">bool</span> stop_equals_start=<span class="keyword">false</span>);
<a name="l00292"></a>00292 
<a name="l00293"></a>00293 
<a name="l00294"></a>00294   
<a name="l00295"></a>00295 
<a name="l00296"></a>00296   <span class="comment">//------------------------------------------------------//</span>
<a name="l00297"></a>00297   <span class="comment">//   Data                                               //</span>
<a name="l00298"></a>00298   <span class="comment">//------------------------------------------------------//</span>
<a name="l00299"></a>00299   
<a name="l00300"></a><a class="code" href="classGenomicInterval.html#a0dbe6195570468c386eea923e30c762c">00300</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#a0dbe6195570468c386eea923e30c762c" title="file line number">n_line</a>;             
<a name="l00301"></a><a class="code" href="classGenomicInterval.html#ac489a1195c43cc93d0e192dc91d5a192">00301</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicInterval.html#ac489a1195c43cc93d0e192dc91d5a192" title="chromosome name (space and tab characters are not allowed)">CHROMOSOME</a>;                             
<a name="l00302"></a><a class="code" href="classGenomicInterval.html#a25f6603b48fcc57f97c89fa768b658f9">00302</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicInterval.html#a25f6603b48fcc57f97c89fa768b658f9" title="orientation (&#39;+&#39; or &#39;-&#39;)">STRAND</a>;                                  
<a name="l00303"></a><a class="code" href="classGenomicInterval.html#ac0abba93599f5cdf4b5a4e7ec5909bbd">00303</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#ac0abba93599f5cdf4b5a4e7ec5909bbd" title="interval start position in the chromosome (&gt;=1)">START</a>;                               
<a name="l00304"></a><a class="code" href="classGenomicInterval.html#ac5fac21da3939b7976859a54800b93f0">00304</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicInterval.html#ac5fac21da3939b7976859a54800b93f0" title="interval stop position in the chromosome (&lt;=chromosome_size)">STOP</a>;                                
<a name="l00305"></a>00305 };
<a name="l00306"></a>00306 
<a name="l00307"></a>00307 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00308"></a>00308 <span class="comment">// END CLASS DECLARATION: GenomicIntervals                                                     //</span>
<a name="l00309"></a>00309 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00310"></a>00310 
<a name="l00311"></a>00311 
<a name="l00312"></a>00312 
<a name="l00313"></a>00313 
<a name="l00314"></a>00314 
<a name="l00315"></a>00315 
<a name="l00316"></a>00316 
<a name="l00317"></a>00317 
<a name="l00318"></a>00318 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00319"></a>00319 <span class="comment">// CLASS: Chromosomes                                                                          //</span>
<a name="l00320"></a>00320 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00322"></a>00322 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00323"></a><a class="code" href="classChromosomes.html">00323</a> <span class="keyword">class </span><a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>
<a name="l00324"></a>00324 {
<a name="l00325"></a>00325  <span class="keyword">public</span>:
<a name="l00326"></a>00326 
<a name="l00327"></a><a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243">00327</a>   <span class="keyword">typedef</span> map&lt;string,pair&lt;char*,size_t&gt; &gt; <a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a>;            
<a name="l00328"></a>00328 
<a name="l00330"></a>00330 
<a name="l00363"></a>00363   <a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>(<span class="keywordtype">char</span> *chrom_map_dir, <span class="keywordtype">char</span> *chrom_map_name);
<a name="l00364"></a>00364 
<a name="l00365"></a>00365 
<a name="l00367"></a>00367 
<a name="l00383"></a>00383   <a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>(<span class="keywordtype">char</span> *fasta_file, <span class="keywordtype">bool</span> verbose);
<a name="l00384"></a>00384 
<a name="l00385"></a>00385   
<a name="l00387"></a>00387   ~<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a>();
<a name="l00388"></a>00388   
<a name="l00389"></a>00389 
<a name="l00391"></a>00391   <span class="keywordtype">bool</span> FindChromosome(<span class="keywordtype">char</span> *chromosome_name);
<a name="l00392"></a>00392   
<a name="l00393"></a>00393   
<a name="l00395"></a>00395   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00396"></a>00396   
<a name="l00397"></a>00397   
<a name="l00399"></a>00399   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00400"></a>00400   
<a name="l00401"></a>00401   
<a name="l00403"></a>00403   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I);
<a name="l00404"></a>00404   
<a name="l00405"></a>00405   
<a name="l00407"></a>00407   <span class="keywordtype">void</span> LoadChromosome(<span class="keywordtype">string</span> chromosome_name);
<a name="l00408"></a>00408 
<a name="l00409"></a>00409   
<a name="l00410"></a>00410   <span class="comment">// data</span>
<a name="l00411"></a>00411 
<a name="l00412"></a><a class="code" href="classChromosomes.html#a954bae240564f6f589a30d253291eb75">00412</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#a954bae240564f6f589a30d253291eb75" title="if &#39;true&#39;, print messages while loading chromosome sequences">verbose</a>;                                                         
<a name="l00413"></a><a class="code" href="classChromosomes.html#a408a06b9ae635ceaf0b75452bcb62763">00413</a>   <span class="keywordtype">bool</span> <a class="code" href="classChromosomes.html#a408a06b9ae635ceaf0b75452bcb62763" title="&#39;true&#39; if reading from FASTA file">load_in_memory</a>;                                          
<a name="l00414"></a><a class="code" href="classChromosomes.html#ad72dcc9b9d3b0b041b737d895a20eef6">00414</a>   StringMap <a class="code" href="classChromosomes.html#ad72dcc9b9d3b0b041b737d895a20eef6" title="maps chromosome names to chromosome sequence file names">chrom_map</a>;                                          
<a name="l00415"></a><a class="code" href="classChromosomes.html#a5fb6a314ac56a7ace50070fd3e5ae708">00415</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#a5fb6a314ac56a7ace50070fd3e5ae708" title="the directory where the chromosome map file is located">chrom_map_dir</a>;                                         
<a name="l00416"></a><a class="code" href="classChromosomes.html#a71b69dff9fc8c4cae1671df17e11a016">00416</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#a71b69dff9fc8c4cae1671df17e11a016" title="the name of the chromosome map file">chrom_map_name</a>;                                        
<a name="l00417"></a><a class="code" href="classChromosomes.html#ab3b502eec7bd2028297300aeb47b5844">00417</a>   <span class="keywordtype">string</span> <a class="code" href="classChromosomes.html#ab3b502eec7bd2028297300aeb47b5844" title="name of the currently used chromosome">current_chromosome_name</a>;                       
<a name="l00418"></a><a class="code" href="classChromosomes.html#afb59cb80755187f4546573f10110e669">00418</a>   <span class="keywordtype">size_t</span> <a class="code" href="classChromosomes.html#afb59cb80755187f4546573f10110e669" title="size of the currently used chromosome">current_chromosome_size</a>;                       
<a name="l00419"></a><a class="code" href="classChromosomes.html#af640cf0f09b109b3d46d70aba9207a2e">00419</a>   <span class="keywordtype">char</span> *<a class="code" href="classChromosomes.html#af640cf0f09b109b3d46d70aba9207a2e" title="pointer to the currently used chromosome sequence">current_chromosome_seq</a>;                         
<a name="l00420"></a><a class="code" href="classChromosomes.html#a673169d55aa0b94815e17b41566d4d8c">00420</a>   <a class="code" href="classChromosomes.html#a4a233d2b7fabc1a589cbbafd5d0ea243" title="for every chromosome, store a pointer to its sequence, and its size.">ChromosomeSeqData</a> <a class="code" href="classChromosomes.html#a673169d55aa0b94815e17b41566d4d8c" title="all chromosomes sequences and their sizes are stored here">chromosome_seq_data</a>;        
<a name="l00421"></a>00421 };
<a name="l00422"></a>00422 
<a name="l00423"></a>00423 
<a name="l00424"></a>00424 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00425"></a>00425 <span class="comment">// END CLASS: Chromosomes                                                                      //</span>
<a name="l00426"></a>00426 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00427"></a>00427 
<a name="l00428"></a>00428 
<a name="l00429"></a>00429 
<a name="l00430"></a>00430 
<a name="l00431"></a>00431 
<a name="l00432"></a>00432 
<a name="l00433"></a>00433 
<a name="l00434"></a>00434 
<a name="l00435"></a>00435 
<a name="l00436"></a>00436 
<a name="l00437"></a>00437 
<a name="l00438"></a>00438 
<a name="l00439"></a>00439 
<a name="l00440"></a>00440 
<a name="l00441"></a>00441 
<a name="l00442"></a>00442 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00443"></a>00443 <span class="comment">// CLASS: GenomicIntervalSetAsArray                                                            //</span>
<a name="l00444"></a>00444 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00446"></a>00446 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00447"></a>00447 
<a name="l00448"></a><a class="code" href="classGenomicIntervalSetAsArray.html">00448</a> <span class="keyword">class </span><a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>
<a name="l00449"></a>00449 {
<a name="l00450"></a>00450  <span class="keyword">public</span>:
<a name="l00451"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a1a0416c9fac7e9d965ae7fe84f1649c2">00451</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>** <a class="code" href="classGenomicIntervalSetAsArray.html#a1a0416c9fac7e9d965ae7fe84f1649c2" title="iterator defined as a pointer in the array of GenomicInterval">iterator</a>;                           
<a name="l00452"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0310f5d79c529b631a5ef71ef186b41c">00452</a>   <span class="keyword">typedef</span> <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a>** <a class="code" href="classGenomicIntervalSetAsArray.html#a0310f5d79c529b631a5ef71ef186b41c" title="reverse_iterator defined as a pointer in the array of GenomicInterval">reverse_iterator</a>;           
<a name="l00453"></a>00453 
<a name="l00454"></a>00454 
<a name="l00456"></a>00456   <a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>();
<a name="l00457"></a>00457 
<a name="l00458"></a>00458 
<a name="l00460"></a>00460   ~<a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a>();
<a name="l00461"></a>00461   
<a name="l00462"></a>00462   
<a name="l00464"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a9e56381009d164d7625199b183845388">00464</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a9e56381009d164d7625199b183845388" title="Implementation of begin() for this container.">begin</a>() { <span class="keywordflow">return</span> &amp;I[0]; }
<a name="l00465"></a>00465   
<a name="l00466"></a>00466 
<a name="l00468"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a5b6145ceff99907d9075047519482584">00468</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a5b6145ceff99907d9075047519482584" title="Implementation of end() for this container.">end</a>() { <span class="keywordflow">return</span> &amp;I[n_intervals]; }
<a name="l00469"></a>00469   
<a name="l00470"></a>00470 
<a name="l00472"></a><a class="code" href="classGenomicIntervalSetAsArray.html#ac077589a55a7dfec9b2e6c38392f7444">00472</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#ac077589a55a7dfec9b2e6c38392f7444" title="Implementation of rbegin() for this container.">rbegin</a>() { <span class="keywordflow">return</span> &amp;I[n_intervals-1]; }
<a name="l00473"></a>00473   
<a name="l00474"></a>00474 
<a name="l00476"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a5b7fc7dc64857dfdadd1198393c0d117">00476</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a5b7fc7dc64857dfdadd1198393c0d117" title="Implementation of rend() for this container.">rend</a>() { <span class="keywordflow">return</span> &amp;I[-1]; }
<a name="l00477"></a>00477   
<a name="l00478"></a>00478 
<a name="l00480"></a><a class="code" href="classGenomicIntervalSetAsArray.html#aa74747b36b114c326edfaf1224270d0d">00480</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#aa74747b36b114c326edfaf1224270d0d" title="Implementation of front() for this container.">front</a>() { <span class="keywordflow">return</span> I[0]; }
<a name="l00481"></a>00481   
<a name="l00482"></a>00482 
<a name="l00484"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a11b483c12111ee96c15fec296c66ee15">00484</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicIntervalSetAsArray.html#a11b483c12111ee96c15fec296c66ee15" title="Implementation of back() for this container.">back</a>() { <span class="keywordflow">return</span> I[n_intervals-1]; }
<a name="l00485"></a>00485   
<a name="l00486"></a>00486 
<a name="l00488"></a><a class="code" href="classGenomicIntervalSetAsArray.html#ad4bfef55744241849841c71d08835403">00488</a>   <span class="keywordtype">size_t</span> <a class="code" href="classGenomicIntervalSetAsArray.html#ad4bfef55744241849841c71d08835403" title="Implementation of size() for this container.">size</a>() { <span class="keywordflow">return</span> n_intervals; }
<a name="l00489"></a>00489   
<a name="l00490"></a>00490 
<a name="l00492"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a4348e7636cb91405b08374a966e33a7c">00492</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a4348e7636cb91405b08374a966e33a7c" title="Implementation of push_back() for this container.">push_back</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *j) { I[n_intervals++] = j; }
<a name="l00493"></a>00493   
<a name="l00494"></a>00494 
<a name="l00496"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a833395eaf9daa5626879eefa861d9ea5">00496</a>   <span class="keywordtype">void</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a833395eaf9daa5626879eefa861d9ea5" title="Implementation of clear() for this container.">clear</a>() {  }
<a name="l00497"></a>00497   
<a name="l00498"></a>00498 
<a name="l00500"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0f8554cfca150d695d15691aa42df7fa">00500</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0f8554cfca150d695d15691aa42df7fa" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> j) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be erased!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00501"></a>00501 
<a name="l00502"></a>00502 
<a name="l00504"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0aae25ac2e9a7cf50a370703aa99df85">00504</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0aae25ac2e9a7cf50a370703aa99df85" title="Implementation of erase() for this container (exits with error message).">erase</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be erased!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00505"></a>00505   
<a name="l00506"></a>00506 
<a name="l00508"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a0cd89244c0e75b88f9b56c9b9478d06b">00508</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> <a class="code" href="classGenomicIntervalSetAsArray.html#a0cd89244c0e75b88f9b56c9b9478d06b" title="Implementation of insert() for this container (exits with error message).">insert</a>(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">iterator</a> i, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *u) { fprintf(stderr, <span class="stringliteral">&quot;Error: [GenomicIntervalArray]: array entries cannot be inserted!&quot;</span>); exit(1); <span class="keywordflow">return</span> i; }
<a name="l00509"></a>00509   
<a name="l00510"></a>00510 
<a name="l00511"></a>00511   
<a name="l00512"></a>00512   <span class="comment">//------------------------------------------------------//</span>
<a name="l00513"></a>00513   <span class="comment">//   Data                                               //</span>
<a name="l00514"></a>00514   <span class="comment">//------------------------------------------------------//</span>
<a name="l00515"></a>00515   
<a name="l00516"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a9f9558bbaa8d682993a60fa142893998">00516</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a9f9558bbaa8d682993a60fa142893998" title="number of intervals">n_intervals</a>;                                 
<a name="l00517"></a><a class="code" href="classGenomicIntervalSetAsArray.html#a6580b875cd9804ba59f88c5000f83f09">00517</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicIntervalSetAsArray.html#a6580b875cd9804ba59f88c5000f83f09" title="number of maximum allowed intervals (this will be removed)">max_intervals</a>;                               
<a name="l00518"></a><a class="code" href="classGenomicIntervalSetAsArray.html#af7ec8bbfea7748695aa35e6ec467a4b2">00518</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> **<a class="code" href="classGenomicIntervalSetAsArray.html#af7ec8bbfea7748695aa35e6ec467a4b2" title="array of GenomicInterval">I</a>;                                  
<a name="l00519"></a>00519   
<a name="l00520"></a>00520 };
<a name="l00521"></a>00521 
<a name="l00522"></a>00522 
<a name="l00523"></a>00523 
<a name="l00524"></a>00524 
<a name="l00525"></a>00525 
<a name="l00526"></a>00526 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00527"></a>00527 <span class="comment">// END CLASS: GenomicIntervalArray                                                             //</span>
<a name="l00528"></a>00528 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00529"></a>00529 
<a name="l00530"></a>00530 
<a name="l00531"></a>00531 
<a name="l00532"></a>00532 
<a name="l00533"></a>00533 
<a name="l00534"></a>00534 
<a name="l00535"></a>00535 
<a name="l00536"></a>00536 
<a name="l00537"></a>00537 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00538"></a>00538 <span class="comment">// CLASS: GenomicRegion                                                                        //</span>
<a name="l00539"></a>00539 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00540"></a>00540 <span class="comment">//                                                                                             //</span>
<a name="l00541"></a>00541 <span class="comment">// Input formats:                                                                              //</span>
<a name="l00542"></a>00542 <span class="comment">//  1. LABEL &lt;TAB&gt; CHROMOSOME STRAND START[,START]* STOP[,STOP]*                               //</span>
<a name="l00543"></a>00543 <span class="comment">//  2. LABEL &lt;TAB&gt; [CHROMOSOME STRAND START STOP]+                                             //</span>
<a name="l00544"></a>00544 <span class="comment">//                                                                                             //</span>
<a name="l00545"></a>00545 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00547"></a>00547 <span class="comment"></span>
<a name="l00556"></a>00556 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00557"></a>00557 
<a name="l00558"></a><a class="code" href="classGenomicRegion.html">00558</a> <span class="keyword">class </span><a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l00559"></a>00559 {
<a name="l00560"></a>00560  <span class="keyword">public</span>:
<a name="l00562"></a>00562   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>();
<a name="l00563"></a>00563 
<a name="l00564"></a>00564 
<a name="l00565"></a>00565 
<a name="l00567"></a>00567   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l00568"></a>00568 
<a name="l00569"></a>00569 
<a name="l00570"></a>00570 
<a name="l00572"></a>00572   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0); 
<a name="l00573"></a>00573 
<a name="l00574"></a>00574 
<a name="l00575"></a>00575 
<a name="l00577"></a>00577   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i);
<a name="l00578"></a>00578 
<a name="l00579"></a>00579 
<a name="l00580"></a>00580 
<a name="l00582"></a>00582   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>();
<a name="l00583"></a>00583   
<a name="l00584"></a>00584 
<a name="l00585"></a>00585 
<a name="l00586"></a>00586   <span class="comment">//------------------------------------------------------//</span>
<a name="l00587"></a>00587   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l00588"></a>00588   <span class="comment">//------------------------------------------------------//</span>
<a name="l00589"></a>00589 
<a name="l00590"></a>00590 
<a name="l00592"></a>00592   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0, <span class="keywordtype">char</span> del1=<span class="charliteral">&#39; &#39;</span>, <span class="keywordtype">char</span> del2=<span class="charliteral">&#39;,&#39;</span>);
<a name="l00593"></a>00593 
<a name="l00594"></a>00594 
<a name="l00595"></a>00595 
<a name="l00597"></a>00597   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l00598"></a>00598 
<a name="l00599"></a>00599 
<a name="l00600"></a>00600   
<a name="l00602"></a>00602   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l00603"></a>00603 
<a name="l00604"></a>00604 
<a name="l00605"></a>00605 
<a name="l00607"></a>00607   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintIntersection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l00608"></a>00608 
<a name="l00609"></a>00609 
<a name="l00610"></a>00610 
<a name="l00612"></a>00612   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l00613"></a>00613 
<a name="l00614"></a>00614 
<a name="l00615"></a>00615 
<a name="l00617"></a>00617   <span class="keywordtype">void</span> PrintREG(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l00618"></a>00618 
<a name="l00619"></a>00619 
<a name="l00620"></a>00620 
<a name="l00622"></a>00622   <span class="keywordtype">void</span> PrintQ(<span class="keywordtype">bool</span> reverse=<span class="keyword">false</span>);
<a name="l00623"></a>00623 
<a name="l00624"></a>00624 
<a name="l00625"></a>00625 
<a name="l00627"></a>00627   <span class="keywordtype">void</span> PrintError(<span class="keywordtype">string</span> error_msg);
<a name="l00628"></a>00628 
<a name="l00629"></a>00629 
<a name="l00630"></a>00630 
<a name="l00631"></a>00631 
<a name="l00632"></a>00632   <span class="comment">//------------------------------------------------------//</span>
<a name="l00633"></a>00633   <span class="comment">//   Get &amp; Set                                          //</span>
<a name="l00634"></a>00634   <span class="comment">//------------------------------------------------------//</span>
<a name="l00635"></a>00635 
<a name="l00636"></a>00636 
<a name="l00638"></a>00638   <span class="keywordtype">void</span> DeleteIntervals(GenomicIntervalSet::iterator i, GenomicIntervalSet::iterator j); 
<a name="l00639"></a>00639 
<a name="l00640"></a>00640 
<a name="l00641"></a>00641 
<a name="l00643"></a>00643   <span class="keywordtype">void</span> SetLabel(<span class="keywordtype">char</span> *label);
<a name="l00644"></a>00644 
<a name="l00645"></a>00645 
<a name="l00646"></a>00646 
<a name="l00648"></a>00648   <span class="keywordtype">void</span> SetStrand(<span class="keywordtype">char</span> strand);
<a name="l00649"></a>00649 
<a name="l00650"></a>00650 
<a name="l00651"></a>00651 
<a name="l00653"></a>00653   <span class="keywordtype">char</span> *GetChromosome();
<a name="l00654"></a>00654 
<a name="l00655"></a>00655 
<a name="l00656"></a>00656 
<a name="l00658"></a>00658   <span class="keywordtype">size_t</span> GetSize();
<a name="l00659"></a>00659 
<a name="l00660"></a>00660 
<a name="l00661"></a>00661 
<a name="l00663"></a>00663   <span class="keywordtype">size_t</span> GetSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00664"></a>00664 
<a name="l00665"></a>00665 
<a name="l00666"></a>00666 
<a name="l00668"></a>00668   <span class="keywordtype">char</span> *GetSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00669"></a>00669 
<a name="l00670"></a>00670 
<a name="l00671"></a>00671 
<a name="l00672"></a>00672   <span class="comment">//------------------------------------------------------//</span>
<a name="l00673"></a>00673   <span class="comment">//   Check                                              //</span>
<a name="l00674"></a>00674   <span class="comment">//------------------------------------------------------//</span>
<a name="l00675"></a>00675 
<a name="l00676"></a>00676 
<a name="l00678"></a>00678   <span class="keywordtype">bool</span> IsCompatible(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00679"></a>00679 
<a name="l00680"></a>00680 
<a name="l00681"></a>00681 
<a name="l00683"></a>00683   <span class="keywordtype">bool</span> IsCompatibleWith(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00684"></a>00684 
<a name="l00685"></a>00685 
<a name="l00686"></a>00686 
<a name="l00688"></a>00688   <span class="keywordtype">bool</span> IsCompatibleSorted(<span class="keywordtype">bool</span> ignore_strand);
<a name="l00689"></a>00689 
<a name="l00690"></a>00690 
<a name="l00691"></a>00691 
<a name="l00693"></a>00693   <span class="keywordtype">bool</span> IsCompatibleSortedAndNonoverlapping();
<a name="l00694"></a>00694 
<a name="l00695"></a>00695 
<a name="l00696"></a>00696 
<a name="l00698"></a>00698   <span class="keywordtype">bool</span> OverlapsWith(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00699"></a>00699 
<a name="l00700"></a>00700 
<a name="l00701"></a>00701 
<a name="l00703"></a>00703   <span class="keywordtype">bool</span> IsBefore(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00704"></a>00704 
<a name="l00705"></a>00705 
<a name="l00706"></a>00706 
<a name="l00708"></a>00708   <span class="keywordtype">bool</span> IsPosBefore(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l00709"></a>00709 
<a name="l00710"></a>00710 
<a name="l00711"></a>00711 
<a name="l00713"></a>00713   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcOverlap(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand);
<a name="l00714"></a>00714 
<a name="l00715"></a>00715 
<a name="l00716"></a>00716 
<a name="l00718"></a>00718   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00719"></a>00719 
<a name="l00720"></a>00720   
<a name="l00721"></a>00721 
<a name="l00723"></a>00723   <span class="keywordtype">int</span> CalcDirection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l00724"></a>00724 
<a name="l00725"></a>00725   
<a name="l00726"></a>00726 
<a name="l00727"></a>00727   <span class="comment">//------------------------------------------------------//</span>
<a name="l00728"></a>00728   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l00729"></a>00729   <span class="comment">//------------------------------------------------------//</span>
<a name="l00730"></a>00730 
<a name="l00731"></a>00731 
<a name="l00733"></a>00733   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00734"></a>00734 
<a name="l00735"></a>00735 
<a name="l00736"></a>00736   
<a name="l00738"></a>00738   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l00739"></a>00739 
<a name="l00740"></a>00740 
<a name="l00741"></a>00741 
<a name="l00743"></a>00743   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l00744"></a>00744 
<a name="l00745"></a>00745 
<a name="l00746"></a>00746 
<a name="l00748"></a>00748   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Center();
<a name="l00749"></a>00749 
<a name="l00750"></a>00750 
<a name="l00751"></a>00751 
<a name="l00753"></a>00753   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyChromosomeNames(map&lt;string,string&gt; &amp;chromosome_conversion_map);
<a name="l00754"></a>00754 
<a name="l00755"></a>00755 
<a name="l00756"></a>00756 
<a name="l00758"></a>00758   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Connect();
<a name="l00759"></a>00759 
<a name="l00760"></a>00760 
<a name="l00761"></a>00761 
<a name="l00763"></a>00763   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Diff();
<a name="l00764"></a>00764 
<a name="l00765"></a>00765 
<a name="l00766"></a>00766 
<a name="l00768"></a>00768   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l00769"></a>00769 
<a name="l00770"></a>00770 
<a name="l00771"></a>00771 
<a name="l00773"></a>00773   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Divide();
<a name="l00774"></a>00774 
<a name="l00775"></a>00775 
<a name="l00776"></a>00776 
<a name="l00778"></a>00778   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunDivide();
<a name="l00779"></a>00779 
<a name="l00780"></a>00780 
<a name="l00781"></a>00781 
<a name="l00783"></a>00783   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Fix();
<a name="l00784"></a>00784 
<a name="l00785"></a>00785 
<a name="l00786"></a>00786 
<a name="l00788"></a>00788   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Intersect();
<a name="l00789"></a>00789 
<a name="l00790"></a>00790 
<a name="l00791"></a>00791 
<a name="l00793"></a>00793   <span class="keywordtype">void</span> RunSize();
<a name="l00794"></a>00794 
<a name="l00795"></a>00795 
<a name="l00796"></a>00796 
<a name="l00798"></a>00798 
<a name="l00802"></a>00802   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l00803"></a>00803 
<a name="l00804"></a>00804 
<a name="l00805"></a>00805 
<a name="l00807"></a>00807   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l00808"></a>00808 
<a name="l00809"></a>00809 
<a name="l00810"></a>00810 
<a name="l00812"></a>00812   <span class="keywordtype">void</span> ReversePos(StringLIntMap *bounds);
<a name="l00813"></a>00813 
<a name="l00814"></a>00814 
<a name="l00815"></a>00815   
<a name="l00817"></a>00817   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l00818"></a>00818 
<a name="l00819"></a>00819 
<a name="l00820"></a>00820 
<a name="l00822"></a>00822 
<a name="l00827"></a>00827   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l00828"></a>00828 
<a name="l00829"></a>00829 
<a name="l00830"></a>00830 
<a name="l00832"></a>00832   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l00833"></a>00833 
<a name="l00834"></a>00834 
<a name="l00835"></a>00835 
<a name="l00837"></a>00837   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Sort(); 
<a name="l00838"></a>00838 
<a name="l00839"></a>00839 
<a name="l00840"></a>00840   
<a name="l00842"></a>00842   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunSplit(); 
<a name="l00843"></a>00843 
<a name="l00844"></a>00844 
<a name="l00845"></a>00845   
<a name="l00847"></a>00847   <span class="keyword">virtual</span> <span class="keywordtype">void</span> ModifyStrand(<span class="keywordtype">char</span> *strand_op);
<a name="l00848"></a>00848 
<a name="l00849"></a>00849 
<a name="l00850"></a>00850   
<a name="l00852"></a>00852   <span class="keyword">virtual</span> <span class="keywordtype">void</span> Union();
<a name="l00853"></a>00853 
<a name="l00854"></a>00854 
<a name="l00855"></a>00855 
<a name="l00857"></a>00857   <span class="keyword">virtual</span> <span class="keywordtype">void</span> RunUnion();
<a name="l00858"></a>00858 
<a name="l00859"></a>00859 
<a name="l00860"></a>00860 
<a name="l00862"></a>00862   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l00863"></a>00863 
<a name="l00864"></a>00864 
<a name="l00865"></a>00865 
<a name="l00867"></a>00867   <span class="keywordtype">void</span> PrintRemoveN(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00868"></a>00868 
<a name="l00869"></a>00869 
<a name="l00870"></a>00870 
<a name="l00872"></a>00872   <span class="keyword">virtual</span> <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l00873"></a>00873 
<a name="l00874"></a>00874   
<a name="l00876"></a>00876   <span class="keywordtype">void</span> PrintSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l00877"></a>00877 
<a name="l00878"></a>00878 
<a name="l00879"></a>00879 
<a name="l00881"></a>00881 
<a name="l00885"></a>00885   <span class="keywordtype">void</span> PrintOffsetFormat(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l00886"></a>00886 
<a name="l00887"></a>00887 
<a name="l00888"></a>00888 
<a name="l00890"></a>00890   <span class="keywordtype">void</span> PrintSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l00891"></a>00891 
<a name="l00892"></a>00892 
<a name="l00893"></a>00893 
<a name="l00895"></a>00895   <span class="keywordtype">void</span> PrintVerifySeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l00896"></a>00896 
<a name="l00897"></a>00897 
<a name="l00898"></a>00898 
<a name="l00900"></a>00900   <span class="keywordtype">void</span> ReverseStrand();
<a name="l00901"></a>00901 
<a name="l00902"></a>00902 
<a name="l00903"></a>00903 
<a name="l00904"></a>00904   <span class="comment">//------------------------------------------------------//</span>
<a name="l00905"></a>00905   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l00906"></a>00906   <span class="comment">//------------------------------------------------------//</span>
<a name="l00907"></a>00907 
<a name="l00908"></a>00908   
<a name="l00910"></a>00910   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Constrain(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l00911"></a>00911 
<a name="l00912"></a>00912   
<a name="l00913"></a>00913 
<a name="l00915"></a>00915   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Diff(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l00916"></a>00916 
<a name="l00917"></a>00917 
<a name="l00918"></a>00918   
<a name="l00919"></a>00919 
<a name="l00920"></a>00920   <span class="comment">//------------------------------------------------------//</span>
<a name="l00921"></a>00921   <span class="comment">//   Data                                               //</span>
<a name="l00922"></a>00922   <span class="comment">//------------------------------------------------------//</span>
<a name="l00923"></a>00923 
<a name="l00924"></a><a class="code" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">00924</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c" title="keeps track of line number">n_line</a>;                             
<a name="l00925"></a><a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24">00925</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>;                                                  
<a name="l00926"></a><a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9">00926</a>   GenomicIntervalSet <a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>;                                 
<a name="l00927"></a>00927   
<a name="l00928"></a>00928   
<a name="l00929"></a>00929 };
<a name="l00930"></a>00930 
<a name="l00931"></a>00931 
<a name="l00932"></a>00932 
<a name="l00933"></a>00933 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00934"></a>00934 <span class="comment">// END CLASS: GenomicRegion                                                                    //</span>
<a name="l00935"></a>00935 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00936"></a>00936 
<a name="l00937"></a>00937 
<a name="l00938"></a>00938 
<a name="l00939"></a>00939 
<a name="l00940"></a>00940 
<a name="l00941"></a>00941 
<a name="l00942"></a>00942 
<a name="l00943"></a>00943 
<a name="l00944"></a>00944 
<a name="l00945"></a>00945 
<a name="l00946"></a>00946 
<a name="l00947"></a>00947 
<a name="l00948"></a>00948 
<a name="l00949"></a>00949 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00950"></a>00950 <span class="comment">// CLASS: GenomicRegionSEQ                                                                     //</span>
<a name="l00951"></a>00951 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00953"></a>00953 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00954"></a>00954 
<a name="l00955"></a><a class="code" href="classGenomicRegionSEQ.html">00955</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l00956"></a>00956 {
<a name="l00957"></a>00957  <span class="keyword">public</span>:
<a name="l00959"></a>00959   <a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l00960"></a>00960 
<a name="l00962"></a>00962   ~<a class="code" href="classGenomicRegionSEQ.html" title="This class implements the genomic regions REG format with attached sequence.">GenomicRegionSEQ</a>();
<a name="l00963"></a>00963   
<a name="l00964"></a>00964 
<a name="l00965"></a>00965 
<a name="l00966"></a>00966   <span class="comment">//------------------------------------------------------//</span>
<a name="l00967"></a>00967   <span class="comment">//   Operations                                         //</span>
<a name="l00968"></a>00968   <span class="comment">//------------------------------------------------------//</span>
<a name="l00969"></a>00969   
<a name="l00971"></a>00971   <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l00972"></a>00972 
<a name="l00973"></a>00973   
<a name="l00974"></a>00974   <span class="comment">//------------------------------------------------------//</span>
<a name="l00975"></a>00975   <span class="comment">//   Data                                               //</span>
<a name="l00976"></a>00976   <span class="comment">//------------------------------------------------------//</span>
<a name="l00977"></a>00977 
<a name="l00978"></a><a class="code" href="classGenomicRegionSEQ.html#a37f8aa6c3c6d8d0059aa56944efa3a52">00978</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSEQ.html#a37f8aa6c3c6d8d0059aa56944efa3a52" title="sequence">SEQ</a>;                                                    
<a name="l00979"></a>00979 
<a name="l00980"></a>00980   
<a name="l00981"></a>00981 <span class="comment">/*</span>
<a name="l00982"></a>00982 <span class="comment">  // NOTE: utilize all data &amp; code related to the SEQ format in a new class: GenomicRegionSEQ.</span>
<a name="l00984"></a>00984 <span class="comment">  void PrintNoGaps();                                                                                   // alignments</span>
<a name="l00985"></a>00985 <span class="comment"></span>
<a name="l00987"></a>00987 <span class="comment">  void PrintModifyLabel();</span>
<a name="l00988"></a>00988 <span class="comment">  </span>
<a name="l00989"></a>00989 <span class="comment">  vector&lt;string&gt; Q;                                             //!&lt; sequences (e.g. from a multiple sequence alignment)</span>
<a name="l00990"></a>00990 <span class="comment">  vector&lt;string&gt; L;                                             //!&lt; sequence labels (e.g. names of species in multiple sequence alignment)</span>
<a name="l00991"></a>00991 <span class="comment">*/</span>
<a name="l00992"></a>00992 
<a name="l00993"></a>00993 };
<a name="l00994"></a>00994 
<a name="l00995"></a>00995 
<a name="l00996"></a>00996 
<a name="l00997"></a>00997 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l00998"></a>00998 <span class="comment">// END CLASS: GenomicRegionSEQ                                                                 //</span>
<a name="l00999"></a>00999 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01000"></a>01000 
<a name="l01001"></a>01001 
<a name="l01002"></a>01002 
<a name="l01003"></a>01003 
<a name="l01004"></a>01004 
<a name="l01005"></a>01005 
<a name="l01006"></a>01006 
<a name="l01007"></a>01007 
<a name="l01008"></a>01008 
<a name="l01009"></a>01009 
<a name="l01010"></a>01010 
<a name="l01011"></a>01011 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01012"></a>01012 <span class="comment">// CLASS: GenomicRegionBEDToREG                                                                //</span>
<a name="l01013"></a>01013 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01015"></a>01015 <span class="comment"></span>
<a name="l01022"></a>01022 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01023"></a><a class="code" href="classGenomicRegionBEDToREG.html">01023</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01024"></a>01024 {
<a name="l01025"></a>01025  <span class="keyword">public</span>:
<a name="l01027"></a>01027   <a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01028"></a>01028 
<a name="l01029"></a>01029 
<a name="l01030"></a>01030 
<a name="l01032"></a>01032   <a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01033"></a>01033 
<a name="l01034"></a>01034 
<a name="l01035"></a>01035 
<a name="l01037"></a>01037   ~<a class="code" href="classGenomicRegionBEDToREG.html" title="This class reads BED files but converts them internally to the REG format, so all operations are perf...">GenomicRegionBEDToREG</a>();
<a name="l01038"></a>01038   
<a name="l01039"></a>01039 
<a name="l01040"></a>01040 
<a name="l01042"></a>01042   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01043"></a>01043   
<a name="l01044"></a>01044 
<a name="l01045"></a>01045 
<a name="l01046"></a>01046   <span class="comment">// data</span>
<a name="l01047"></a>01047 
<a name="l01048"></a>01048 
<a name="l01049"></a>01049 };
<a name="l01050"></a>01050 
<a name="l01051"></a>01051 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01052"></a>01052 <span class="comment">// END CLASS: GenomicRegionBEDToREG                                                            //</span>
<a name="l01053"></a>01053 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01054"></a>01054 
<a name="l01055"></a>01055 
<a name="l01056"></a>01056 
<a name="l01057"></a>01057 
<a name="l01058"></a>01058 
<a name="l01059"></a>01059 
<a name="l01060"></a>01060 
<a name="l01061"></a>01061 
<a name="l01062"></a>01062 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01063"></a>01063 <span class="comment">// CLASS: GenomicRegionBED                                                                     //</span>
<a name="l01064"></a>01064 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01066"></a>01066 <span class="comment"></span>
<a name="l01073"></a>01073 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01074"></a><a class="code" href="classGenomicRegionBED.html">01074</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01075"></a>01075 {
<a name="l01076"></a>01076  <span class="keyword">public</span>:
<a name="l01078"></a>01078   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01079"></a>01079 
<a name="l01080"></a>01080 
<a name="l01081"></a>01081 
<a name="l01083"></a>01083   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01084"></a>01084 
<a name="l01085"></a>01085 
<a name="l01086"></a>01086 
<a name="l01088"></a>01088   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>(<span class="keywordtype">char</span> *label, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i);
<a name="l01089"></a>01089 
<a name="l01090"></a>01090 
<a name="l01091"></a>01091 
<a name="l01093"></a>01093   ~<a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a>();
<a name="l01094"></a>01094   
<a name="l01095"></a>01095 
<a name="l01096"></a>01096 
<a name="l01097"></a>01097   <span class="comment">//------------------------------------------------------//</span>
<a name="l01098"></a>01098   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01099"></a>01099   <span class="comment">//------------------------------------------------------//</span>
<a name="l01100"></a>01100 
<a name="l01101"></a>01101 
<a name="l01103"></a>01103   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01104"></a>01104   
<a name="l01105"></a>01105 
<a name="l01106"></a>01106 
<a name="l01108"></a>01108   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01109"></a>01109 
<a name="l01110"></a>01110 
<a name="l01111"></a>01111 
<a name="l01113"></a>01113   <span class="keywordtype">void</span> Print(FILE *file_ptr, <span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01114"></a>01114 
<a name="l01115"></a>01115 
<a name="l01116"></a>01116 
<a name="l01118"></a>01118   <span class="keywordtype">void</span> PrintIntersection(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> merge_labels);
<a name="l01119"></a>01119 
<a name="l01120"></a>01120 
<a name="l01121"></a>01121 
<a name="l01123"></a>01123   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01124"></a>01124 
<a name="l01125"></a>01125 
<a name="l01126"></a>01126 
<a name="l01128"></a>01128   <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01129"></a>01129 
<a name="l01130"></a>01130 
<a name="l01131"></a>01131 
<a name="l01132"></a>01132   <span class="comment">//------------------------------------------------------//</span>
<a name="l01133"></a>01133   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01134"></a>01134   <span class="comment">//------------------------------------------------------//</span>
<a name="l01135"></a>01135 
<a name="l01136"></a>01136 
<a name="l01138"></a>01138   <span class="keywordtype">void</span> UpdateThick();
<a name="l01139"></a>01139   
<a name="l01140"></a>01140   
<a name="l01141"></a>01141   
<a name="l01143"></a>01143   <span class="keywordtype">void</span> PrintBEDFormat(<span class="keywordtype">char</span> *color, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01144"></a>01144 
<a name="l01145"></a>01145 
<a name="l01146"></a>01146 
<a name="l01148"></a>01148   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l01149"></a>01149 
<a name="l01150"></a>01150 
<a name="l01151"></a>01151 
<a name="l01153"></a>01153   <span class="keywordtype">void</span> Center();
<a name="l01154"></a>01154 
<a name="l01155"></a>01155 
<a name="l01156"></a>01156 
<a name="l01158"></a>01158   <span class="keywordtype">void</span> Connect();
<a name="l01159"></a>01159 
<a name="l01160"></a>01160 
<a name="l01161"></a>01161 
<a name="l01163"></a>01163   <span class="keywordtype">void</span> Diff();
<a name="l01164"></a>01164 
<a name="l01165"></a>01165 
<a name="l01166"></a>01166 
<a name="l01168"></a>01168   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01169"></a>01169 
<a name="l01170"></a>01170 
<a name="l01171"></a>01171 
<a name="l01173"></a>01173   <span class="keywordtype">void</span> Divide();
<a name="l01174"></a>01174 
<a name="l01175"></a>01175 
<a name="l01176"></a>01176 
<a name="l01178"></a>01178   <span class="keywordtype">void</span> RunDivide();
<a name="l01179"></a>01179 
<a name="l01180"></a>01180 
<a name="l01181"></a>01181 
<a name="l01183"></a>01183   <span class="keywordtype">void</span> Fix();
<a name="l01184"></a>01184 
<a name="l01185"></a>01185 
<a name="l01186"></a>01186 
<a name="l01188"></a>01188   <span class="keywordtype">void</span> Intersect();
<a name="l01189"></a>01189 
<a name="l01190"></a>01190 
<a name="l01191"></a>01191 
<a name="l01193"></a>01193 
<a name="l01197"></a>01197   <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01198"></a>01198 
<a name="l01199"></a>01199 
<a name="l01200"></a>01200 
<a name="l01202"></a>01202   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01203"></a>01203 
<a name="l01204"></a>01204 
<a name="l01205"></a>01205 
<a name="l01207"></a>01207   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01208"></a>01208 
<a name="l01209"></a>01209 
<a name="l01210"></a>01210 
<a name="l01212"></a>01212 
<a name="l01217"></a>01217   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l01218"></a>01218 
<a name="l01219"></a>01219 
<a name="l01220"></a>01220 
<a name="l01222"></a>01222   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01223"></a>01223 
<a name="l01224"></a>01224 
<a name="l01225"></a>01225 
<a name="l01227"></a>01227   <span class="keywordtype">void</span> Sort(); 
<a name="l01228"></a>01228 
<a name="l01229"></a>01229 
<a name="l01230"></a>01230   
<a name="l01232"></a>01232   <span class="keywordtype">void</span> RunSplit(); 
<a name="l01233"></a>01233 
<a name="l01234"></a>01234 
<a name="l01235"></a>01235   
<a name="l01237"></a>01237   <span class="keywordtype">void</span> Union();
<a name="l01238"></a>01238 
<a name="l01239"></a>01239 
<a name="l01240"></a>01240 
<a name="l01242"></a>01242   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01243"></a>01243 
<a name="l01244"></a>01244 
<a name="l01245"></a>01245 
<a name="l01246"></a>01246   <span class="comment">//------------------------------------------------------//</span>
<a name="l01247"></a>01247   <span class="comment">//   Operations returning new regions                   //</span>
<a name="l01248"></a>01248   <span class="comment">//------------------------------------------------------//</span>
<a name="l01249"></a>01249 
<a name="l01250"></a>01250   
<a name="l01252"></a>01252   <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> *Constrain(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> *label=NULL);
<a name="l01253"></a>01253 
<a name="l01254"></a>01254   
<a name="l01255"></a>01255 
<a name="l01257"></a>01257   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Diff(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l01258"></a>01258 
<a name="l01259"></a>01259   
<a name="l01260"></a>01260 
<a name="l01261"></a>01261   <span class="comment">//------------------------------------------------------//</span>
<a name="l01262"></a>01262   <span class="comment">//   Data                                               //</span>
<a name="l01263"></a>01263   <span class="comment">//------------------------------------------------------//</span>
<a name="l01264"></a>01264 
<a name="l01265"></a><a class="code" href="classGenomicRegionBED.html#a60878391f6fb1e0ad0ac9bc88a7625d3">01265</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a60878391f6fb1e0ad0ac9bc88a7625d3" title="number of tokens">n_tokens</a>;                    
<a name="l01266"></a><a class="code" href="classGenomicRegionBED.html#afa7ba810a0d1673f5dbabcf85c7bbb95">01266</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#afa7ba810a0d1673f5dbabcf85c7bbb95" title="score field">score</a>;                               
<a name="l01267"></a><a class="code" href="classGenomicRegionBED.html#a52fc75b347dfe6696e7ad2beeeb623a9">01267</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionBED.html#a52fc75b347dfe6696e7ad2beeeb623a9" title="color field">itemRgb</a>;                                
<a name="l01268"></a><a class="code" href="classGenomicRegionBED.html#a019dfaebedc17e8859b36c3359287faa">01268</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a019dfaebedc17e8859b36c3359287faa" title="thickStart field">thickStart</a>;                  
<a name="l01269"></a><a class="code" href="classGenomicRegionBED.html#a5333a7fdc1f21db508aa7af4f50553ba">01269</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionBED.html#a5333a7fdc1f21db508aa7af4f50553ba" title="thickEnd field">thickEnd</a>;                    
<a name="l01270"></a>01270 
<a name="l01271"></a>01271 };
<a name="l01272"></a>01272 
<a name="l01273"></a>01273 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01274"></a>01274 <span class="comment">// END CLASS: GenomicRegionBED                                                                 //</span>
<a name="l01275"></a>01275 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01276"></a>01276 
<a name="l01277"></a>01277 
<a name="l01278"></a>01278 
<a name="l01279"></a>01279 
<a name="l01280"></a>01280 
<a name="l01281"></a>01281 
<a name="l01282"></a>01282 
<a name="l01283"></a>01283 
<a name="l01284"></a>01284 
<a name="l01285"></a>01285 
<a name="l01286"></a>01286 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01287"></a>01287 <span class="comment">// CLASS: GenomicRegionSAMToREG                                                                //</span>
<a name="l01288"></a>01288 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01290"></a>01290 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01291"></a><a class="code" href="classGenomicRegionSAMToREG.html">01291</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01292"></a>01292 {
<a name="l01293"></a>01293  <span class="keyword">public</span>:
<a name="l01294"></a>01294   
<a name="l01296"></a>01296   <a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01297"></a>01297 
<a name="l01298"></a>01298 
<a name="l01299"></a>01299 
<a name="l01301"></a>01301   <a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01302"></a>01302 
<a name="l01303"></a>01303 
<a name="l01304"></a>01304 
<a name="l01306"></a>01306   ~<a class="code" href="classGenomicRegionSAMToREG.html" title="This class reads SAM files but converts them internally to the REG format, so all operations are perf...">GenomicRegionSAMToREG</a>();
<a name="l01307"></a>01307   
<a name="l01308"></a>01308 
<a name="l01309"></a>01309 
<a name="l01311"></a>01311   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01312"></a>01312   
<a name="l01313"></a>01313 
<a name="l01314"></a>01314 
<a name="l01316"></a>01316   <span class="keywordtype">char</span> CalcStrandFromFlag(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01317"></a>01317 
<a name="l01318"></a>01318 
<a name="l01319"></a>01319   
<a name="l01321"></a>01321   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetNextTokenOfCIGAR(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01322"></a>01322 
<a name="l01323"></a>01323   
<a name="l01324"></a>01324   
<a name="l01325"></a>01325   <span class="comment">// data</span>
<a name="l01326"></a>01326 
<a name="l01327"></a>01327 
<a name="l01328"></a>01328 };
<a name="l01329"></a>01329 
<a name="l01330"></a>01330 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01331"></a>01331 <span class="comment">// END CLASS: GenomicRegionSAMToREG                                                            //</span>
<a name="l01332"></a>01332 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01333"></a>01333 
<a name="l01334"></a>01334 
<a name="l01335"></a>01335 
<a name="l01336"></a>01336 
<a name="l01337"></a>01337 
<a name="l01338"></a>01338 
<a name="l01339"></a>01339 
<a name="l01340"></a>01340 
<a name="l01341"></a>01341 
<a name="l01342"></a>01342 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01343"></a>01343 <span class="comment">// CLASS: GenomicRegionSAM                                                                     //</span>
<a name="l01344"></a>01344 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01346"></a>01346 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01347"></a><a class="code" href="classGenomicRegionSAM.html">01347</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01348"></a>01348 {
<a name="l01349"></a>01349  <span class="keyword">public</span>:
<a name="l01350"></a><a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96">01350</a>   <span class="keyword">typedef</span> vector&lt;pair&lt;long int,char&gt; &gt; <a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length)">CIGARTokens</a>;                     
<a name="l01351"></a>01351   
<a name="l01353"></a>01353   <a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01354"></a>01354 
<a name="l01355"></a>01355 
<a name="l01356"></a>01356 
<a name="l01358"></a>01358   <a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01359"></a>01359 
<a name="l01360"></a>01360 
<a name="l01361"></a>01361 
<a name="l01363"></a>01363   ~<a class="code" href="classGenomicRegionSAM.html" title="This class processes genomic regions in SAM format.">GenomicRegionSAM</a>();
<a name="l01364"></a>01364   
<a name="l01365"></a>01365 
<a name="l01366"></a>01366 
<a name="l01367"></a>01367   <span class="comment">//------------------------------------------------------//</span>
<a name="l01368"></a>01368   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01369"></a>01369   <span class="comment">//------------------------------------------------------//</span>
<a name="l01370"></a>01370 
<a name="l01371"></a>01371 
<a name="l01373"></a>01373   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01374"></a>01374 
<a name="l01375"></a>01375 
<a name="l01376"></a>01376 
<a name="l01378"></a>01378   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01379"></a>01379 
<a name="l01380"></a>01380 
<a name="l01381"></a>01381 
<a name="l01383"></a>01383   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01384"></a>01384 
<a name="l01385"></a>01385 
<a name="l01386"></a>01386 
<a name="l01388"></a>01388   <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01389"></a>01389 
<a name="l01390"></a>01390 
<a name="l01391"></a>01391   
<a name="l01392"></a>01392   <span class="comment">//------------------------------------------------------//</span>
<a name="l01393"></a>01393   <span class="comment">//   Auxiliary functions                                //</span>
<a name="l01394"></a>01394   <span class="comment">//------------------------------------------------------//</span>
<a name="l01395"></a>01395 
<a name="l01397"></a>01397   <span class="keywordtype">char</span> CalcStrandFromFlag(<span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> flag);
<a name="l01398"></a>01398 
<a name="l01399"></a>01399 
<a name="l01400"></a>01400   
<a name="l01402"></a>01402   <span class="keywordtype">void</span> UpdateFlagFromStrand(<span class="keywordtype">char</span> strand);
<a name="l01403"></a>01403   
<a name="l01404"></a>01404   
<a name="l01405"></a>01405   
<a name="l01407"></a>01407   <span class="keywordtype">void</span> PrintGappedSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01408"></a>01408 
<a name="l01409"></a>01409   
<a name="l01410"></a>01410   
<a name="l01412"></a>01412   <span class="keywordtype">char</span> *GetGappedSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">char</span> *gap_token_types, <span class="keywordtype">bool</span> lowercase_for_sort_clipping=<span class="keyword">true</span>);
<a name="l01413"></a>01413 
<a name="l01414"></a>01414   
<a name="l01415"></a>01415   
<a name="l01417"></a>01417   <span class="keywordtype">void</span> InvalidateSeqData();
<a name="l01418"></a>01418 
<a name="l01419"></a>01419 
<a name="l01420"></a>01420   
<a name="l01422"></a>01422   <span class="keywordtype">void</span> TokenizeCIGAR();
<a name="l01423"></a>01423 
<a name="l01424"></a>01424   
<a name="l01425"></a>01425   
<a name="l01427"></a>01427   <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetNextTokenOfCIGAR(<span class="keywordtype">char</span> **cigar, <span class="keywordtype">char</span> *type);
<a name="l01428"></a>01428 
<a name="l01429"></a>01429   
<a name="l01430"></a>01430   
<a name="l01432"></a>01432   <span class="keywordtype">void</span> UpdateCigarFromIntervals();
<a name="l01433"></a>01433   
<a name="l01434"></a>01434   
<a name="l01435"></a>01435  
<a name="l01437"></a>01437   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcFragmentLengthFromCIGAR();
<a name="l01438"></a>01438   
<a name="l01439"></a>01439   
<a name="l01440"></a>01440   
<a name="l01442"></a>01442   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcReferenceLengthFromCIGAR();
<a name="l01443"></a>01443   
<a name="l01444"></a>01444   
<a name="l01445"></a>01445   
<a name="l01447"></a>01447   <span class="keywordtype">char</span> *TrimCIGAR(<span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01448"></a>01448  
<a name="l01449"></a>01449  
<a name="l01450"></a>01450  
<a name="l01452"></a>01452   <span class="keywordtype">char</span> *TrimSequence(<span class="keywordtype">char</span> *seq, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> d_ref_stop); 
<a name="l01453"></a>01453  
<a name="l01454"></a>01454  
<a name="l01455"></a>01455  
<a name="l01456"></a>01456   <span class="comment">//------------------------------------------------------//</span>
<a name="l01457"></a>01457   <span class="comment">//   Line-based (horizontal) operations                 //</span>
<a name="l01458"></a>01458   <span class="comment">//------------------------------------------------------//</span>
<a name="l01459"></a>01459 
<a name="l01460"></a>01460 
<a name="l01462"></a>01462   <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01463"></a>01463 
<a name="l01464"></a>01464 
<a name="l01465"></a>01465   
<a name="l01467"></a>01467   <span class="keywordtype">bool</span> ApplyBounds(StringLIntMap *bounds);
<a name="l01468"></a>01468 
<a name="l01469"></a>01469 
<a name="l01470"></a>01470 
<a name="l01472"></a>01472   <span class="keywordtype">void</span> Center();
<a name="l01473"></a>01473 
<a name="l01474"></a>01474 
<a name="l01475"></a>01475 
<a name="l01477"></a>01477   <span class="keywordtype">void</span> Connect();
<a name="l01478"></a>01478 
<a name="l01479"></a>01479 
<a name="l01480"></a>01480 
<a name="l01482"></a>01482   <span class="keywordtype">void</span> Diff();
<a name="l01483"></a>01483 
<a name="l01484"></a>01484 
<a name="l01485"></a>01485 
<a name="l01487"></a>01487   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01488"></a>01488 
<a name="l01489"></a>01489 
<a name="l01490"></a>01490 
<a name="l01492"></a>01492   <span class="keywordtype">void</span> Divide();
<a name="l01493"></a>01493 
<a name="l01494"></a>01494 
<a name="l01495"></a>01495 
<a name="l01497"></a>01497   <span class="keywordtype">void</span> RunDivide();
<a name="l01498"></a>01498 
<a name="l01499"></a>01499 
<a name="l01500"></a>01500 
<a name="l01502"></a>01502   <span class="keywordtype">void</span> Fix();
<a name="l01503"></a>01503 
<a name="l01504"></a>01504 
<a name="l01505"></a>01505 
<a name="l01507"></a>01507   <span class="keywordtype">void</span> Intersect();
<a name="l01508"></a>01508 
<a name="l01509"></a>01509 
<a name="l01510"></a>01510 
<a name="l01512"></a>01512   <span class="keywordtype">void</span> Randomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l01513"></a>01513 
<a name="l01514"></a>01514 
<a name="l01515"></a>01515 
<a name="l01517"></a>01517 
<a name="l01521"></a>01521   <span class="keywordtype">void</span> ModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l01522"></a>01522 
<a name="l01523"></a>01523 
<a name="l01524"></a>01524 
<a name="l01526"></a>01526   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01527"></a>01527 
<a name="l01528"></a>01528 
<a name="l01529"></a>01529 
<a name="l01531"></a>01531 
<a name="l01536"></a>01536   <span class="keywordtype">void</span> ShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l01537"></a>01537 
<a name="l01538"></a>01538 
<a name="l01539"></a>01539 
<a name="l01541"></a>01541   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc);
<a name="l01542"></a>01542 
<a name="l01543"></a>01543 
<a name="l01544"></a>01544 
<a name="l01546"></a>01546   <span class="keywordtype">void</span> Sort(); 
<a name="l01547"></a>01547 
<a name="l01548"></a>01548 
<a name="l01549"></a>01549   
<a name="l01551"></a>01551   <span class="keywordtype">void</span> RunSplit(); 
<a name="l01552"></a>01552 
<a name="l01553"></a>01553 
<a name="l01554"></a>01554   
<a name="l01556"></a>01556   <span class="keywordtype">void</span> ModifyStrand(<span class="keywordtype">char</span> *strand_op);
<a name="l01557"></a>01557 
<a name="l01558"></a>01558 
<a name="l01559"></a>01559   
<a name="l01561"></a>01561   <span class="keywordtype">void</span> Union();
<a name="l01562"></a>01562 
<a name="l01563"></a>01563 
<a name="l01564"></a>01564 
<a name="l01566"></a>01566   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01567"></a>01567 
<a name="l01568"></a>01568 
<a name="l01569"></a>01569 
<a name="l01570"></a>01570   <span class="comment">//------------------------------------------------------//</span>
<a name="l01571"></a>01571   <span class="comment">//   Data                                               //</span>
<a name="l01572"></a>01572   <span class="comment">//------------------------------------------------------//</span>
<a name="l01573"></a>01573 
<a name="l01574"></a><a class="code" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f">01574</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f" title="number of TAB-separated fields in SAM entry (&gt;= 12)">n_tokens</a>;                                    
<a name="l01575"></a>01575   <span class="comment">//char *QNAME;                                                //!&lt; Query template NAME: mapped to GenomicRegion::LABEL</span>
<a name="l01576"></a><a class="code" href="classGenomicRegionSAM.html#a200d7f7044dd69311eceb851a3c315e5">01576</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#a200d7f7044dd69311eceb851a3c315e5" title="bitwise FLAG">FLAG</a>;                               
<a name="l01577"></a>01577   <span class="comment">//char *RNAME;                                                //!&lt; Reference sequence NAME: mapped to GenomicRegion::I.front()-&gt;CHROMOSOME</span>
<a name="l01578"></a>01578   <span class="comment">//long int POS;                                               //!&lt; 1-based leftmost mapping POSition: mapped to GenomicRegion::I.front()-&gt;START</span>
<a name="l01579"></a><a class="code" href="classGenomicRegionSAM.html#aedacdd8319f4b614229a41b80d4c6f67">01579</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#aedacdd8319f4b614229a41b80d4c6f67" title="MAPping Quality.">MAPQ</a>;                                                
<a name="l01580"></a><a class="code" href="classGenomicRegionSAM.html#aa5acff481ae565db4976e66933212726">01580</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#aa5acff481ae565db4976e66933212726" title="CIGAR string.">CIGAR</a>;                                                  
<a name="l01581"></a><a class="code" href="classGenomicRegionSAM.html#af723505bfafff1a1fa616479023d0c3c">01581</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#af723505bfafff1a1fa616479023d0c3c" title="Reference name of the mate/next fragment.">RNEXT</a>;                                                  
<a name="l01582"></a><a class="code" href="classGenomicRegionSAM.html#ad6f88b4324ba508b86102641d819661f">01582</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#ad6f88b4324ba508b86102641d819661f" title="Position of the mate/next fragment.">PNEXT</a>;                                               
<a name="l01583"></a><a class="code" href="classGenomicRegionSAM.html#af180854a186be37eab49520edaa1262f">01583</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSAM.html#af180854a186be37eab49520edaa1262f" title="observed Template LENgth">TLEN</a>;                                                
<a name="l01584"></a><a class="code" href="classGenomicRegionSAM.html#a6c2887ed1fa3a25d801641ea1b139032">01584</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#a6c2887ed1fa3a25d801641ea1b139032" title="fragment SEQuence">SEQ</a>;                                                    
<a name="l01585"></a><a class="code" href="classGenomicRegionSAM.html#ab43ca5f2fbfdee520f1424c3a9fd0706">01585</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#ab43ca5f2fbfdee520f1424c3a9fd0706" title="ASCII of Phred-scaled base QUALity+33.">QUAL</a>;                                                   
<a name="l01586"></a><a class="code" href="classGenomicRegionSAM.html#a1622a8804462bad83b1c49d25eeacb49">01586</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSAM.html#a1622a8804462bad83b1c49d25eeacb49" title="all SAM fields after the 12th mandatory field">OPTIONAL</a>;                                               
<a name="l01587"></a><a class="code" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92">01587</a>   <a class="code" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96" title="This type is used for storing CIGAR tokens, i.e. (token_type, token_length)">CIGARTokens</a> <a class="code" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92" title="CIGAR tokens.">T</a>;                                                
<a name="l01588"></a>01588   
<a name="l01589"></a>01589 };
<a name="l01590"></a>01590 
<a name="l01591"></a>01591 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01592"></a>01592 <span class="comment">// END CLASS: GenomicRegionSAM                                                                 //</span>
<a name="l01593"></a>01593 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01594"></a>01594 
<a name="l01595"></a>01595 
<a name="l01596"></a>01596 
<a name="l01597"></a>01597 
<a name="l01598"></a>01598 
<a name="l01599"></a>01599 
<a name="l01600"></a>01600 
<a name="l01601"></a>01601 
<a name="l01602"></a>01602 
<a name="l01603"></a>01603 
<a name="l01604"></a>01604 
<a name="l01605"></a>01605 
<a name="l01606"></a>01606 
<a name="l01607"></a>01607 
<a name="l01608"></a>01608 
<a name="l01609"></a>01609 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01610"></a>01610 <span class="comment">// CLASS: GenomicRegionGFFToREG                                                                //</span>
<a name="l01611"></a>01611 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01613"></a>01613 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01614"></a><a class="code" href="classGenomicRegionGFFToREG.html">01614</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01615"></a>01615 {
<a name="l01616"></a>01616  <span class="keyword">public</span>:
<a name="l01618"></a>01618   <a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01619"></a>01619 
<a name="l01620"></a>01620 
<a name="l01621"></a>01621 
<a name="l01623"></a>01623   <a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01624"></a>01624 
<a name="l01625"></a>01625 
<a name="l01626"></a>01626 
<a name="l01628"></a>01628   ~<a class="code" href="classGenomicRegionGFFToREG.html" title="This class reads GFF files but converts them internally to the REG format, so all operations are perf...">GenomicRegionGFFToREG</a>();
<a name="l01629"></a>01629   
<a name="l01630"></a>01630 
<a name="l01631"></a>01631 
<a name="l01633"></a>01633   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01634"></a>01634   
<a name="l01635"></a>01635 
<a name="l01636"></a>01636 
<a name="l01637"></a>01637   <span class="comment">// data</span>
<a name="l01638"></a>01638 
<a name="l01639"></a>01639 
<a name="l01640"></a>01640 };
<a name="l01641"></a>01641 
<a name="l01642"></a>01642 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01643"></a>01643 <span class="comment">// END CLASS: GenomicRegionGFFToREG                                                            //</span>
<a name="l01644"></a>01644 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01645"></a>01645 
<a name="l01646"></a>01646 
<a name="l01647"></a>01647 
<a name="l01648"></a>01648 
<a name="l01649"></a>01649 
<a name="l01650"></a>01650 
<a name="l01651"></a>01651 
<a name="l01652"></a>01652 
<a name="l01653"></a>01653 
<a name="l01654"></a>01654 
<a name="l01655"></a>01655 
<a name="l01656"></a>01656 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01657"></a>01657 <span class="comment">// CLASS: GenomicRegionGFF                                                                     //</span>
<a name="l01658"></a>01658 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01660"></a>01660 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01661"></a><a class="code" href="classGenomicRegionGFF.html">01661</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a>
<a name="l01662"></a>01662 {
<a name="l01663"></a>01663  <span class="keyword">public</span>:
<a name="l01665"></a>01665   <a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *B);
<a name="l01666"></a>01666 
<a name="l01667"></a>01667 
<a name="l01668"></a>01668 
<a name="l01670"></a>01670   <a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01671"></a>01671 
<a name="l01672"></a>01672 
<a name="l01673"></a>01673 
<a name="l01675"></a>01675   ~<a class="code" href="classGenomicRegionGFF.html" title="This class processes genomic regions in SAM format.">GenomicRegionGFF</a>();
<a name="l01676"></a>01676   
<a name="l01677"></a>01677 
<a name="l01678"></a>01678 
<a name="l01679"></a>01679   <span class="comment">//------------------------------------------------------//</span>
<a name="l01680"></a>01680   <span class="comment">//   Read &amp; Print                                       //</span>
<a name="l01681"></a>01681   <span class="comment">//------------------------------------------------------//</span>
<a name="l01682"></a>01682 
<a name="l01683"></a>01683 
<a name="l01685"></a>01685   <span class="keywordtype">void</span> Read(<span class="keywordtype">char</span> *inp, <span class="keywordtype">long</span> <span class="keywordtype">int</span> n_line=0);
<a name="l01686"></a>01686 
<a name="l01687"></a>01687 
<a name="l01688"></a>01688 
<a name="l01690"></a>01690   <span class="keywordtype">void</span> Print(FILE *file_ptr=stdout);
<a name="l01691"></a>01691 
<a name="l01692"></a>01692 
<a name="l01693"></a>01693 
<a name="l01695"></a>01695   <span class="keywordtype">void</span> PrintConstrained(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">bool</span> merge_labels=<span class="keyword">false</span>);
<a name="l01696"></a>01696 
<a name="l01697"></a>01697 
<a name="l01698"></a>01698 
<a name="l01700"></a>01700   <span class="keywordtype">void</span> PrintModified(<span class="keywordtype">char</span> *label, <span class="keywordtype">long</span> <span class="keywordtype">int</span> start, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop);
<a name="l01701"></a>01701 
<a name="l01702"></a>01702 
<a name="l01703"></a>01703 
<a name="l01705"></a>01705   <span class="keywordtype">void</span> Diff();
<a name="l01706"></a>01706 
<a name="l01707"></a>01707 
<a name="l01708"></a>01708 
<a name="l01710"></a>01710   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01711"></a>01711 
<a name="l01712"></a>01712 
<a name="l01713"></a>01713 
<a name="l01715"></a>01715   <span class="keywordtype">void</span> Divide();
<a name="l01716"></a>01716 
<a name="l01717"></a>01717 
<a name="l01718"></a>01718 
<a name="l01720"></a>01720   <span class="keywordtype">void</span> RunDivide();
<a name="l01721"></a>01721 
<a name="l01722"></a>01722 
<a name="l01723"></a>01723 
<a name="l01725"></a>01725   <span class="keywordtype">void</span> Intersect();
<a name="l01726"></a>01726 
<a name="l01727"></a>01727 
<a name="l01728"></a>01728 
<a name="l01730"></a>01730   <span class="keywordtype">void</span> Select(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l01731"></a>01731 
<a name="l01732"></a>01732 
<a name="l01733"></a>01733 
<a name="l01735"></a>01735   <span class="keywordtype">void</span> RunSplit(); 
<a name="l01736"></a>01736 
<a name="l01737"></a>01737 
<a name="l01738"></a>01738   
<a name="l01740"></a>01740   <span class="keywordtype">void</span> PrintWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01741"></a>01741 
<a name="l01742"></a>01742 
<a name="l01743"></a>01743 
<a name="l01744"></a>01744   <span class="comment">//---------------------------------------//</span>
<a name="l01745"></a>01745   <span class="comment">//  Data                                 //</span>
<a name="l01746"></a>01746   <span class="comment">//---------------------------------------//</span>
<a name="l01747"></a>01747   
<a name="l01748"></a><a class="code" href="classGenomicRegionGFF.html#aaec963f555e7e807fb5046566622f5ee">01748</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionGFF.html#aaec963f555e7e807fb5046566622f5ee" title="number of TAB-separated fields in GFF entry (= 11)">n_tokens</a>;                    
<a name="l01749"></a>01749   <span class="comment">//char *SEQNAME;                              //!&lt; sequence name: mapped to GenomicRegion::I-&gt;front()-&gt;CHROMOSOME</span>
<a name="l01750"></a><a class="code" href="classGenomicRegionGFF.html#a2f0d4a5f15d8a756233436af03341734">01750</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a2f0d4a5f15d8a756233436af03341734" title="feature source">SOURCE</a>;                                 
<a name="l01751"></a><a class="code" href="classGenomicRegionGFF.html#a0de447fbfad3da5b7a9cf07a5a6980f0">01751</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a0de447fbfad3da5b7a9cf07a5a6980f0" title="feture type">FEATURE</a>;                                
<a name="l01752"></a>01752   <span class="comment">//long int START;                             //!&lt; 1-based start position: mapped to GenomicRegion::I-&gt;front()-&gt;START</span>
<a name="l01753"></a>01753   <span class="comment">//long int END;                               //!&lt; 1-based end position: mapped to GenomicRegion::I-&gt;front()-&gt;STOP</span>
<a name="l01754"></a><a class="code" href="classGenomicRegionGFF.html#a0e92da1f0fd153504e0630f27229ba4d">01754</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a0e92da1f0fd153504e0630f27229ba4d" title="score">SCORE</a>;                                  
<a name="l01755"></a>01755   <span class="comment">//char STRAND;                                //!&lt; strand: mapped to GenomicRegion::I-&gt;front()-&gt;STRAND</span>
<a name="l01756"></a><a class="code" href="classGenomicRegionGFF.html#a42122a69eec385e0106e3861cb3b98f3">01756</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionGFF.html#a42122a69eec385e0106e3861cb3b98f3" title="coding frame">FRAME</a>;                                   
<a name="l01757"></a>01757   <span class="comment">//char *ATTRIBUTE;                    //!&lt; attribute: mapped to GenomicRegion::LABEL if non-empty</span>
<a name="l01758"></a><a class="code" href="classGenomicRegionGFF.html#a625a023f1bb24fc597e723919a036073">01758</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionGFF.html#a625a023f1bb24fc597e723919a036073" title="comment">COMMENT</a>;                                
<a name="l01759"></a>01759 
<a name="l01760"></a>01760 };
<a name="l01761"></a>01761 
<a name="l01762"></a>01762 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01763"></a>01763 <span class="comment">// END CLASS: GenomicRegionGFF                                                                 //</span>
<a name="l01764"></a>01764 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01765"></a>01765 
<a name="l01766"></a>01766 
<a name="l01767"></a>01767 
<a name="l01768"></a>01768 
<a name="l01769"></a>01769 
<a name="l01770"></a>01770 
<a name="l01771"></a>01771 
<a name="l01772"></a>01772 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01773"></a>01773 <span class="comment">// CLASS: GenomicRegionSet                                                                     //</span>
<a name="l01774"></a>01774 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01776"></a>01776 <span class="comment"></span>
<a name="l01789"></a>01789 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l01790"></a><a class="code" href="classGenomicRegionSet.html">01790</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>
<a name="l01791"></a>01791 {
<a name="l01792"></a>01792  <span class="keyword">public</span>:
<a name="l01794"></a>01794   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>(<span class="keywordtype">char</span> *file, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size, <span class="keywordtype">bool</span> verbose, <span class="keywordtype">bool</span> load_in_memory, <span class="keywordtype">bool</span> hide_header); 
<a name="l01795"></a>01795 
<a name="l01796"></a>01796 
<a name="l01798"></a>01798   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>(FILE *file_ptr, <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> buffer_size, <span class="keywordtype">bool</span> verbose, <span class="keywordtype">bool</span> load_in_memory, <span class="keywordtype">bool</span> hide_header);
<a name="l01799"></a>01799 
<a name="l01800"></a>01800 
<a name="l01802"></a>01802   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a>();
<a name="l01803"></a>01803 
<a name="l01804"></a>01804 
<a name="l01805"></a>01805   <span class="comment">//-------------------------------------//</span>
<a name="l01806"></a>01806   <span class="comment">//  Read &amp; Print                       //</span>
<a name="l01807"></a>01807   <span class="comment">//-------------------------------------//</span>
<a name="l01808"></a>01808 
<a name="l01810"></a>01810   <span class="keywordtype">void</span> ProcessFileHeader(<span class="keywordtype">bool</span> hide);
<a name="l01811"></a>01811 
<a name="l01812"></a>01812   
<a name="l01814"></a>01814   <span class="keywordtype">void</span> DetectFileFormat();
<a name="l01815"></a>01815 
<a name="l01816"></a>01816   
<a name="l01818"></a>01818   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *CreateGenomicRegion(<a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *buffer);
<a name="l01819"></a>01819 
<a name="l01820"></a>01820 
<a name="l01822"></a>01822   <span class="keywordtype">void</span> Init();
<a name="l01823"></a>01823 
<a name="l01824"></a>01824 
<a name="l01826"></a>01826   <span class="keywordtype">void</span> Reset();
<a name="l01827"></a>01827 
<a name="l01828"></a>01828 
<a name="l01830"></a>01830   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Begin(); 
<a name="l01831"></a>01831 
<a name="l01832"></a>01832 
<a name="l01834"></a>01834   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Get();
<a name="l01835"></a>01835 
<a name="l01836"></a>01836 
<a name="l01838"></a>01838 
<a name="l01841"></a>01841   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Next(<span class="keywordtype">bool</span> retain_current=<span class="keyword">false</span>);
<a name="l01842"></a>01842 
<a name="l01843"></a>01843 
<a name="l01845"></a>01845 
<a name="l01849"></a>01849   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *Next(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">bool</span> retain_current);
<a name="l01850"></a>01850 
<a name="l01851"></a>01851 
<a name="l01853"></a>01853   <span class="keywordtype">void</span> Release(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l01854"></a>01854 
<a name="l01855"></a>01855 
<a name="l01857"></a>01857   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetRetain();
<a name="l01858"></a>01858 
<a name="l01859"></a>01859 
<a name="l01861"></a>01861   <span class="keywordtype">void</span> PrintError(<span class="keywordtype">string</span> error_msg);
<a name="l01862"></a>01862   
<a name="l01863"></a>01863 
<a name="l01864"></a>01864 
<a name="l01865"></a>01865   <span class="comment">//-------------------------------------//</span>
<a name="l01866"></a>01866   <span class="comment">//  File-based (vertical) operations   //</span>
<a name="l01867"></a>01867   <span class="comment">//-------------------------------------//</span>
<a name="l01868"></a>01868 
<a name="l01869"></a>01869 
<a name="l01871"></a>01871   <span class="keywordtype">long</span> <span class="keywordtype">int</span> CountRegions(<span class="keywordtype">bool</span> use_label_counts);
<a name="l01872"></a>01872 
<a name="l01873"></a>01873 
<a name="l01875"></a>01875   <span class="keywordtype">void</span> RunGlobalAnnotate(StringLIntMap *bounds);
<a name="l01876"></a>01876 
<a name="l01877"></a>01877 
<a name="l01879"></a>01879   <span class="keywordtype">void</span> RunGlobalCluster(<span class="keywordtype">bool</span> merge=<span class="keyword">false</span>);
<a name="l01880"></a>01880 
<a name="l01881"></a>01881 
<a name="l01883"></a>01883   <span class="keywordtype">void</span> RunGlobalCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01884"></a>01884 
<a name="l01885"></a>01885 
<a name="l01887"></a>01887   <span class="keywordtype">void</span> RunGlobalSort();
<a name="l01888"></a>01888 
<a name="l01889"></a>01889 
<a name="l01891"></a>01891   <span class="keywordtype">void</span> RunGlobalInvert(StringLIntMap *bounds);
<a name="l01892"></a>01892 
<a name="l01893"></a>01893 
<a name="l01895"></a>01895   <span class="keywordtype">void</span> RunGlobalLink(<span class="keywordtype">bool</span> sorted_by_strand, <span class="keywordtype">long</span> <span class="keywordtype">int</span> max_difference);
<a name="l01896"></a>01896 
<a name="l01897"></a>01897 
<a name="l01899"></a>01899   <span class="keywordtype">void</span> RunGlobalPartition();
<a name="l01900"></a>01900 
<a name="l01901"></a>01901 
<a name="l01903"></a>01903   <span class="keywordtype">void</span> RunGlobalTest(<span class="keywordtype">bool</span> sorted_by_strand);
<a name="l01904"></a>01904 
<a name="l01905"></a>01905 
<a name="l01907"></a>01907   <span class="keywordtype">void</span> RunGlobalMerge(<span class="keywordtype">long</span> <span class="keywordtype">int</span> merge_win);
<a name="l01908"></a>01908 
<a name="l01909"></a>01909 
<a name="l01911"></a>01911   <span class="keywordtype">void</span> RunGlobalReverseOrder();
<a name="l01912"></a>01912 
<a name="l01913"></a>01913 
<a name="l01915"></a>01915   <span class="keywordtype">void</span> PrintDensities(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> &amp;Index, <span class="keywordtype">bool</span> report_exact_boundaries);
<a name="l01916"></a>01916 
<a name="l01917"></a>01917 
<a name="l01919"></a>01919   <span class="keywordtype">void</span> RunGlobalScan(StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l01920"></a>01920 
<a name="l01921"></a>01921 
<a name="l01923"></a>01923 
<a name="l01933"></a>01933   <span class="keywordtype">void</span> RunGlobalScanCount(StringLIntMap *bounds, <span class="keywordtype">char</span> *ref_reg_file, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> ignore_reverse_strand, <span class="keywordtype">char</span> preprocess, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">long</span> <span class="keywordtype">int</span> min_reads);
<a name="l01934"></a>01934 
<a name="l01935"></a>01935 
<a name="l01937"></a>01937   <span class="keywordtype">void</span> RunConvertToWIG(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">char</span> *options, <span class="keywordtype">long</span> <span class="keywordtype">int</span> span, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01938"></a>01938 
<a name="l01939"></a>01939 
<a name="l01940"></a>01940 
<a name="l01941"></a>01941 
<a name="l01942"></a>01942   <span class="comment">//---------------------------------------//</span>
<a name="l01943"></a>01943   <span class="comment">//  Line-based (horizontal) operations   //</span>
<a name="l01944"></a>01944   <span class="comment">//---------------------------------------//</span>
<a name="l01945"></a>01945 
<a name="l01947"></a>01947   <span class="keywordtype">void</span> RunAlign(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l01948"></a>01948 
<a name="l01949"></a>01949 
<a name="l01951"></a>01951   <span class="keywordtype">void</span> RunConvertToBED(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l01952"></a>01952 
<a name="l01953"></a>01953 
<a name="l01955"></a>01955   <span class="keywordtype">void</span> RunBounds(StringLIntMap *bounds);
<a name="l01956"></a>01956 
<a name="l01957"></a>01957 
<a name="l01959"></a>01959   <span class="keywordtype">void</span> RunCenter();
<a name="l01960"></a>01960 
<a name="l01961"></a>01961 
<a name="l01963"></a>01963   <span class="keywordtype">void</span> RunModifyChromosomeNames(<span class="keywordtype">char</span> *chromosome_conversion_file);
<a name="l01964"></a>01964 
<a name="l01965"></a>01965 
<a name="l01967"></a>01967   <span class="keywordtype">void</span> RunConnect();
<a name="l01968"></a>01968 
<a name="l01969"></a>01969 
<a name="l01971"></a>01971   <span class="keywordtype">void</span> RunDiff();
<a name="l01972"></a>01972 
<a name="l01973"></a>01973 
<a name="l01975"></a>01975   <span class="keywordtype">void</span> RunCalcDistances(<span class="keywordtype">char</span> *op1, <span class="keywordtype">char</span> *op2);
<a name="l01976"></a>01976 
<a name="l01977"></a>01977 
<a name="l01979"></a>01979   <span class="keywordtype">void</span> RunDivide();
<a name="l01980"></a>01980 
<a name="l01981"></a>01981 
<a name="l01983"></a>01983   <span class="keywordtype">void</span> RunFix();
<a name="l01984"></a>01984 
<a name="l01985"></a>01985 
<a name="l01986"></a>01986 
<a name="l01988"></a>01988   <span class="keywordtype">void</span> RunIntersection();
<a name="l01989"></a>01989 
<a name="l01990"></a>01990 
<a name="l01991"></a>01991 
<a name="l01993"></a>01993   <span class="keywordtype">void</span> RunSize();
<a name="l01994"></a>01994 
<a name="l01995"></a>01995 
<a name="l01996"></a>01996 
<a name="l01998"></a>01998 
<a name="l02002"></a>02002   <span class="keywordtype">void</span> RunModifyPos(<span class="keyword">const</span> <span class="keywordtype">char</span> *position_op, <span class="keywordtype">long</span> <span class="keywordtype">int</span> position_shift);           
<a name="l02003"></a>02003 
<a name="l02004"></a>02004 
<a name="l02005"></a>02005 
<a name="l02007"></a>02007   <span class="keywordtype">void</span> RunConvertToREG(<span class="keywordtype">bool</span> compact=<span class="keyword">false</span>);
<a name="l02008"></a>02008 
<a name="l02009"></a>02009 
<a name="l02010"></a>02010 
<a name="l02012"></a>02012   <span class="keywordtype">void</span> RunRandomize(gsl_rng *random_generator, StringLIntMap *bounds);
<a name="l02013"></a>02013 
<a name="l02014"></a>02014 
<a name="l02015"></a>02015 
<a name="l02017"></a>02017   <span class="keywordtype">void</span> RunSelect(<span class="keywordtype">bool</span> first, <span class="keywordtype">bool</span> last, <span class="keywordtype">bool</span> from5p, <span class="keywordtype">bool</span> from3p);
<a name="l02018"></a>02018 
<a name="l02019"></a>02019 
<a name="l02020"></a>02020 
<a name="l02022"></a>02022 
<a name="l02027"></a>02027   <span class="keywordtype">void</span> RunShiftPos(<span class="keywordtype">long</span> <span class="keywordtype">int</span> start_shift, <span class="keywordtype">long</span> <span class="keywordtype">int</span> stop_shift, <span class="keywordtype">bool</span> strand_aware);
<a name="l02028"></a>02028 
<a name="l02029"></a>02029 
<a name="l02031"></a>02031   <span class="keywordtype">void</span> RunSort(); 
<a name="l02032"></a>02032 
<a name="l02033"></a>02033   
<a name="l02035"></a>02035   <span class="keywordtype">void</span> RunSplit(); 
<a name="l02036"></a>02036 
<a name="l02037"></a>02037   
<a name="l02039"></a>02039   <span class="keywordtype">void</span> RunModifyStrand(<span class="keywordtype">char</span> *strand_op, <span class="keywordtype">bool</span> sorted);
<a name="l02040"></a>02040 
<a name="l02041"></a>02041 
<a name="l02043"></a>02043   <span class="keywordtype">void</span> RunModifyStrandSorted(<span class="keywordtype">char</span> *strand_op);
<a name="l02044"></a>02044 
<a name="l02045"></a>02045 
<a name="l02047"></a>02047   <span class="keywordtype">void</span> RunShuffle(gsl_rng *random_generator, <span class="keywordtype">char</span> *ref_reg_file);
<a name="l02048"></a>02048 
<a name="l02049"></a>02049 
<a name="l02051"></a>02051   <span class="keywordtype">void</span> RunUnion();
<a name="l02052"></a>02052 
<a name="l02053"></a>02053 
<a name="l02055"></a>02055   <span class="keywordtype">void</span> RunSlidingWindows(<span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size);
<a name="l02056"></a>02056 
<a name="l02057"></a>02057 
<a name="l02059"></a>02059   <span class="keywordtype">void</span> RunExtractSeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> replace=<span class="keyword">false</span>);
<a name="l02060"></a>02060 
<a name="l02061"></a>02061 
<a name="l02062"></a>02062 
<a name="l02063"></a>02063   <span class="comment">//---------------------------------------//</span>
<a name="l02064"></a>02064   <span class="comment">//  UNDER DEVELOPMENT/TESTING            //</span>
<a name="l02065"></a>02065   <span class="comment">//---------------------------------------//</span>
<a name="l02066"></a>02066 
<a name="l02067"></a>02067 
<a name="l02069"></a>02069   <span class="keywordtype">void</span> PrintBEDGraphFormat(<span class="keywordtype">char</span> *title, <span class="keywordtype">char</span> *color, <span class="keywordtype">char</span> *position, <span class="keywordtype">bool</span> convert_chromosome);
<a name="l02070"></a>02070 
<a name="l02071"></a>02071 
<a name="l02073"></a>02073   <span class="keywordtype">void</span> PrintSeqLength(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02074"></a>02074 
<a name="l02075"></a>02075 
<a name="l02077"></a>02077   <span class="comment">//void PrintNoGaps();</span>
<a name="l02078"></a>02078 
<a name="l02079"></a>02079 
<a name="l02081"></a>02081   <span class="keywordtype">void</span> PrintRemoveN(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C);
<a name="l02082"></a>02082 
<a name="l02083"></a>02083 
<a name="l02085"></a>02085 
<a name="l02089"></a>02089   <span class="keywordtype">void</span> PrintOffsetFormat(<span class="keywordtype">char</span> *op, <span class="keywordtype">bool</span> fraction);
<a name="l02090"></a>02090 
<a name="l02091"></a>02091 
<a name="l02093"></a>02093   <span class="keywordtype">void</span> PrintReversePos(StringLIntMap *bounds);
<a name="l02094"></a>02094 
<a name="l02095"></a>02095 
<a name="l02097"></a>02097   <span class="keywordtype">void</span> PrintSearch(<span class="keywordtype">char</span> *pattern, <span class="keywordtype">bool</span> header, <span class="keywordtype">bool</span> summary);
<a name="l02098"></a>02098 
<a name="l02099"></a>02099 
<a name="l02101"></a>02101   <span class="keywordtype">void</span> PrintVerifySeq(<a class="code" href="classChromosomes.html" title="This class is used to store chromosome DNA sequences.">Chromosomes</a> *C, <span class="keywordtype">bool</span> ignore);
<a name="l02102"></a>02102 
<a name="l02103"></a>02103 
<a name="l02104"></a>02104 
<a name="l02105"></a>02105  <span class="keyword">private</span>:
<a name="l02107"></a>02107   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *StoreInTempFile();
<a name="l02108"></a>02108 
<a name="l02109"></a>02109 
<a name="l02111"></a>02111   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *PrintMergeSort(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *rtempSet, <span class="keywordtype">char</span> strand);
<a name="l02112"></a>02112 
<a name="l02113"></a>02113 
<a name="l02115"></a>02115   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *PrintReverse(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *rtempSet);
<a name="l02116"></a>02116 
<a name="l02117"></a>02117 
<a name="l02118"></a>02118 
<a name="l02119"></a>02119   <span class="comment">//---------------------------------------//</span>
<a name="l02120"></a>02120   <span class="comment">//  Data                                 //</span>
<a name="l02121"></a>02121   <span class="comment">//---------------------------------------//</span>
<a name="l02122"></a>02122 
<a name="l02123"></a>02123  <span class="keyword">public</span>:
<a name="l02124"></a><a class="code" href="classGenomicRegionSet.html#a7c593d1beab775590d5b810097644583">02124</a>   <span class="keywordtype">char</span> *<a class="code" href="classGenomicRegionSet.html#a7c593d1beab775590d5b810097644583" title="file name (&#39;NULL&#39; if reading from standard input)">file</a>;                                           
<a name="l02125"></a><a class="code" href="classGenomicRegionSet.html#a4ca6c6268e39fdea2add6e6381c66962">02125</a>   FILE *<a class="code" href="classGenomicRegionSet.html#a4ca6c6268e39fdea2add6e6381c66962" title="file pointer (&#39;NULL&#39; if reading from standard input)">file_ptr</a>;                                       
<a name="l02126"></a><a class="code" href="classGenomicRegionSet.html#a4d60b5403d2d6aaeda541ed45625b245">02126</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#a4d60b5403d2d6aaeda541ed45625b245" title="line buffer size (automatically adjusted during execution to ensure the entire line is read)...">buffer_size</a>;        
<a name="l02127"></a><a class="code" href="classGenomicRegionSet.html#a2646e9e6d5da1693b4bea36a5acc082e">02127</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#a2646e9e6d5da1693b4bea36a5acc082e" title="verbose mode">verbose</a>;                                         
<a name="l02128"></a><a class="code" href="classGenomicRegionSet.html#abd0fc2542e717e5855e03d3b3892f027">02128</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#abd0fc2542e717e5855e03d3b3892f027" title="&#39;true&#39; if reading from standard input">from_stdin</a>;                                      
<a name="l02129"></a><a class="code" href="classGenomicRegionSet.html#ade43c6cd72b95da75395129b00a52687">02129</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#ade43c6cd72b95da75395129b00a52687" title="&#39;true&#39; if the entire region set is loaded in memory">load_in_memory</a>;                          
<a name="l02130"></a><a class="code" href="classGenomicRegionSet.html#a84dd2f8055524a933e963175781f7fd2">02130</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#a84dd2f8055524a933e963175781f7fd2" title="if &#39;true&#39;, then do not print file header">hide_header</a>;                                     
<a name="l02131"></a><a class="code" href="classGenomicRegionSet.html#aea8b5c9bae036da50d4e8720d2e13231">02131</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSet.html#aea8b5c9bae036da50d4e8720d2e13231" title="if &#39;true&#39;, use an array for storing genomic regions&#39; intervals (for time efficiency)">use_interval_array</a>;                      
<a name="l02132"></a><a class="code" href="classGenomicRegionSet.html#aa4494ed2ae3f7594c29b18c39fed6748">02132</a>   <span class="keywordtype">string</span> <a class="code" href="classGenomicRegionSet.html#aa4494ed2ae3f7594c29b18c39fed6748" title="input format">format</a>;                                        
<a name="l02133"></a><a class="code" href="classGenomicRegionSet.html#a08d9a35a7688ac2c193b53c9a8693c6e">02133</a>   <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> <a class="code" href="classGenomicRegionSet.html#a08d9a35a7688ac2c193b53c9a8693c6e" title="keeps track of progress of computation">progress</a>;                            
<a name="l02134"></a><a class="code" href="classGenomicRegionSet.html#adf3a5598e31ddbb5b381125925362085">02134</a>   <a class="code" href="classFileBuffer.html" title="Abstract class for reading lines from file.">FileBuffer</a> *<a class="code" href="classGenomicRegionSet.html#adf3a5598e31ddbb5b381125925362085" title="pointer to the buffer class used to read the region sets">buffer</a>;                           
<a name="l02135"></a><a class="code" href="classGenomicRegionSet.html#a5293292428aa06145b38be4b96ab05e4">02135</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#a5293292428aa06145b38be4b96ab05e4" title="number of regions (&#39;1&#39; if the regions are not loaded in memory)">n_regions</a>;                           
<a name="l02136"></a><a class="code" href="classGenomicRegionSet.html#a206d91b443832800afc04f786d15971d">02136</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> **<a class="code" href="classGenomicRegionSet.html#a206d91b443832800afc04f786d15971d" title="array where the regions are stored; if the region set is not entirely loaded in memory, only R[0] is used">R</a>;                            
<a name="l02137"></a><a class="code" href="classGenomicRegionSet.html#ad89185b88d5b37161cc65bc84b421c0e">02137</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSet.html#ad89185b88d5b37161cc65bc84b421c0e" title="index pointing to the current region in R">r_index</a>;                                     
<a name="l02138"></a>02138 };
<a name="l02139"></a>02139 
<a name="l02140"></a>02140 
<a name="l02141"></a>02141 
<a name="l02142"></a>02142 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02143"></a>02143 <span class="comment">// END CLASS: GenomicRegionSet                                                                 //</span>
<a name="l02144"></a>02144 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02145"></a>02145 
<a name="l02146"></a>02146 
<a name="l02147"></a>02147 
<a name="l02148"></a>02148 
<a name="l02149"></a>02149 
<a name="l02150"></a>02150 
<a name="l02151"></a>02151 
<a name="l02152"></a>02152 
<a name="l02153"></a>02153 
<a name="l02154"></a>02154 
<a name="l02155"></a>02155 
<a name="l02156"></a>02156 
<a name="l02157"></a>02157 
<a name="l02158"></a>02158 
<a name="l02159"></a>02159 
<a name="l02160"></a>02160 
<a name="l02161"></a>02161 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02162"></a>02162 <span class="comment">// CLASS: GenomicRegionSetScanner                                                              //</span>
<a name="l02163"></a>02163 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02165"></a>02165 <span class="comment"></span>
<a name="l02189"></a>02189 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02190"></a><a class="code" href="classGenomicRegionSetScanner.html">02190</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description below for an...">GenomicRegionSetScanner</a>
<a name="l02191"></a>02191 {
<a name="l02192"></a>02192  <span class="keyword">public</span>:
<a name="l02194"></a>02194 
<a name="l02203"></a>02203   <a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description below for an...">GenomicRegionSetScanner</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *R, StringLIntMap *bounds, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_step, <span class="keywordtype">long</span> <span class="keywordtype">int</span> win_size, <span class="keywordtype">bool</span> use_labels_as_values, <span class="keywordtype">bool</span> ignore_reverse_strand, <span class="keywordtype">char</span> preprocess);
<a name="l02204"></a>02204   ~<a class="code" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description below for an...">GenomicRegionSetScanner</a>();
<a name="l02205"></a>02205   
<a name="l02206"></a>02206   <span class="comment">// operations</span>
<a name="l02207"></a>02207   <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next();                                                              
<a name="l02208"></a>02208   <span class="keywordtype">long</span> <span class="keywordtype">int</span> Next(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *Ref);                 
<a name="l02209"></a>02209   <span class="keywordtype">void</span> PrintInterval(FILE *out_file=stdout);    
<a name="l02210"></a>02210   
<a name="l02211"></a>02211   <span class="comment">// data</span>
<a name="l02212"></a>02212  <span class="keyword">private</span>: 
<a name="l02213"></a>02213   <span class="keywordtype">bool</span> Test();                                  
<a name="l02214"></a>02214 
<a name="l02215"></a><a class="code" href="classGenomicRegionSetScanner.html#af76bf4fef482886f23f58cfa12c3eea1">02215</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetScanner.html#af76bf4fef482886f23f58cfa12c3eea1" title="pointer to the GenomicRegionSet which will be scanned by sliding windows">R</a>;                          
<a name="l02216"></a><a class="code" href="classGenomicRegionSetScanner.html#a7f64551c26331cd4ead1ab01405f1b72">02216</a>   StringLIntMap *<a class="code" href="classGenomicRegionSetScanner.html#a7f64551c26331cd4ead1ab01405f1b72" title="chromosome sizes">bounds</a>;                        
<a name="l02217"></a><a class="code" href="classGenomicRegionSetScanner.html#ab278dfa27c3589865b9243156c1c727f">02217</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#ab278dfa27c3589865b9243156c1c727f" title="sliding window step">win_step</a>;                            
<a name="l02218"></a><a class="code" href="classGenomicRegionSetScanner.html#aeed625b2a12aa2f7900997c4e20cd9b5">02218</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#aeed625b2a12aa2f7900997c4e20cd9b5" title="sliding window size (must be a multiple of window step)">win_size</a>;                            
<a name="l02219"></a><a class="code" href="classGenomicRegionSetScanner.html#ad47603d08614ab993fe9036662c421e7">02219</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#ad47603d08614ab993fe9036662c421e7" title="if true, genomic region labels are assumed to be integers and are included in the counting...">use_labels_as_values</a>;                    
<a name="l02220"></a><a class="code" href="classGenomicRegionSetScanner.html#a10c1d22ae74e6f295cd17259f70eff8a">02220</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetScanner.html#a10c1d22ae74e6f295cd17259f70eff8a" title="if true, no sliding windows on the negative strand are reported">ignore_reverse_strand</a>;                   
<a name="l02221"></a><a class="code" href="classGenomicRegionSetScanner.html#af2a536e23023f96c57f51e2276b4be0e">02221</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#af2a536e23023f96c57f51e2276b4be0e" title="if &#39;1&#39;, only start position is counted; if &#39;p&#39;, all positions are counted; if &#39;c&#39;, center of interval is counted">preprocess</a>;                              
<a name="l02222"></a><a class="code" href="classGenomicRegionSetScanner.html#abab8a5c92ec438168e22eb6e33098aba">02222</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#abab8a5c92ec438168e22eb6e33098aba" title="win_size divided by win_step (note: each window is comprised of non-overlapping sub-windows of size w...">n_win_combine</a>;                       
<a name="l02223"></a><a class="code" href="classGenomicRegionSetScanner.html#a60dff3d69ed195c0492711ee295720c2">02223</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetScanner.html#a60dff3d69ed195c0492711ee295720c2" title="pointer to the most recently scanned region">r</a>;                             
<a name="l02224"></a><a class="code" href="classGenomicRegionSetScanner.html#a3f0ba2ab42e8029297f63ba5cf2f9b95">02224</a>   StringLIntMap::iterator <a class="code" href="classGenomicRegionSetScanner.html#a3f0ba2ab42e8029297f63ba5cf2f9b95" title="keeps track of current window&#39;s chromosome information">chr</a>;                  
<a name="l02225"></a><a class="code" href="classGenomicRegionSetScanner.html#a0b9c57901528e75d9d32f4f98a708f4b">02225</a>   <span class="keywordtype">char</span> <a class="code" href="classGenomicRegionSetScanner.html#a0b9c57901528e75d9d32f4f98a708f4b" title="keeps track of current window&#39;s strand information">strand</a>;                                  
<a name="l02226"></a><a class="code" href="classGenomicRegionSetScanner.html#a305a53be4686bb873996e83b3c393701">02226</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#a305a53be4686bb873996e83b3c393701" title="keeps track of current window&#39;s start information">start</a>;                               
<a name="l02227"></a><a class="code" href="classGenomicRegionSetScanner.html#af39ff3f8e4f1a1ab55c004cd06c2378a">02227</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#af39ff3f8e4f1a1ab55c004cd06c2378a" title="keeps track of current window&#39;s stop information">stop</a>;                                
<a name="l02228"></a><a class="code" href="classGenomicRegionSetScanner.html#a19e26d0661ce5f380a72f3eb43b0b07f">02228</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classGenomicRegionSetScanner.html#a19e26d0661ce5f380a72f3eb43b0b07f" title="a vector that keeps track of values in the sub-windows (of size win_step) that comprise the window (o...">v</a>;                                  
<a name="l02229"></a><a class="code" href="classGenomicRegionSetScanner.html#afa341bbfce6cffdc2880c8f1c80abeb1">02229</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#afa341bbfce6cffdc2880c8f1c80abeb1" title="the k-th element of v to be updated next (in a round-robin manner)">k</a>;                                   
<a name="l02230"></a><a class="code" href="classGenomicRegionSetScanner.html#a80be27fad3a3456e7aecd71fb7439c5a">02230</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetScanner.html#a80be27fad3a3456e7aecd71fb7439c5a" title="keeps track of current window&#39;s value information (this is the sum of elements in vector v); it is se...">v_sum</a>;                               
<a name="l02231"></a>02231 };
<a name="l02232"></a>02232 
<a name="l02233"></a>02233 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02234"></a>02234 <span class="comment">// END CLASS: GenomicRegionSetScanner                                                          //</span>
<a name="l02235"></a>02235 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02236"></a>02236 
<a name="l02237"></a>02237 
<a name="l02238"></a>02238 
<a name="l02239"></a>02239 
<a name="l02240"></a>02240 
<a name="l02241"></a>02241 
<a name="l02242"></a>02242 
<a name="l02243"></a>02243 
<a name="l02244"></a>02244 
<a name="l02245"></a>02245 
<a name="l02246"></a>02246 
<a name="l02247"></a>02247 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02248"></a>02248 <span class="comment">// CLASS: GenomicRegionSetOverlaps                                                             //</span>
<a name="l02249"></a>02249 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02251"></a>02251 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02252"></a><a class="code" href="classGenomicRegionSetOverlaps.html">02252</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02253"></a>02253 {
<a name="l02254"></a>02254  <span class="keyword">public</span>:
<a name="l02256"></a>02256 
<a name="l02260"></a>02260   <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet);
<a name="l02261"></a>02261 
<a name="l02262"></a>02262 
<a name="l02264"></a>02264   <span class="keyword">virtual</span> ~<a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>();
<a name="l02265"></a>02265 
<a name="l02266"></a>02266   
<a name="l02268"></a>02268   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery() = 0;
<a name="l02269"></a>02269 
<a name="l02270"></a>02270 
<a name="l02272"></a>02272   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery() = 0;
<a name="l02273"></a>02273 
<a name="l02274"></a>02274 
<a name="l02276"></a>02276   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch() = 0;
<a name="l02277"></a>02277 
<a name="l02278"></a>02278  
<a name="l02280"></a>02280   <span class="keyword">virtual</span> <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch() = 0;
<a name="l02281"></a>02281 
<a name="l02282"></a>02282  
<a name="l02284"></a>02284   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> Done() = 0;
<a name="l02285"></a>02285  
<a name="l02286"></a>02286 
<a name="l02288"></a>02288 
<a name="l02292"></a>02292   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetOverlap(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02293"></a>02293 
<a name="l02294"></a>02294 
<a name="l02296"></a>02296 
<a name="l02300"></a>02300   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextOverlap(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand);
<a name="l02301"></a>02301 
<a name="l02302"></a>02302   
<a name="l02304"></a>02304 
<a name="l02309"></a>02309   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcQueryCoverage(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02310"></a>02310 
<a name="l02311"></a>02311 
<a name="l02313"></a>02313 
<a name="l02318"></a>02318   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *CalcIndexCoverage(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02319"></a>02319 
<a name="l02320"></a>02320 
<a name="l02322"></a>02322 
<a name="l02327"></a>02327   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CountQueryOverlaps(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02328"></a>02328 
<a name="l02329"></a>02329 
<a name="l02331"></a>02331 
<a name="l02336"></a>02336   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *CountIndexOverlaps(<span class="keywordtype">bool</span> match_gaps, <span class="keywordtype">bool</span> ignore_strand, <span class="keywordtype">bool</span> use_labels_as_values);
<a name="l02337"></a>02337 
<a name="l02338"></a>02338  
<a name="l02339"></a>02339   
<a name="l02340"></a>02340   <span class="comment">// data</span>
<a name="l02341"></a>02341  <span class="keyword">public</span>: 
<a name="l02342"></a><a class="code" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2">02342</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2" title="pointer to query region set">QuerySet</a>;                           
<a name="l02343"></a><a class="code" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03">02343</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03" title="pointer to index region set">IndexSet</a>;                           
<a name="l02344"></a><a class="code" href="classGenomicRegionSetOverlaps.html#a31d1dbc3407c616a0bda0f42af7acbae">02344</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#a31d1dbc3407c616a0bda0f42af7acbae" title="pointer to the current query region">current_qreg</a>;                          
<a name="l02345"></a><a class="code" href="classGenomicRegionSetOverlaps.html#a537336b7c40b20e6f5113395c11614b6">02345</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetOverlaps.html#a537336b7c40b20e6f5113395c11614b6" title="pointer to the current index region">current_ireg</a>;                          
<a name="l02346"></a>02346    
<a name="l02347"></a>02347 };
<a name="l02348"></a>02348 
<a name="l02349"></a>02349 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02350"></a>02350 <span class="comment">// END CLASS: GenomicRegionSetOverlaps                                                         //</span>
<a name="l02351"></a>02351 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02352"></a>02352 
<a name="l02353"></a>02353 
<a name="l02354"></a>02354 
<a name="l02355"></a>02355 
<a name="l02356"></a>02356 
<a name="l02357"></a>02357 
<a name="l02358"></a>02358 
<a name="l02359"></a>02359 
<a name="l02360"></a>02360 
<a name="l02361"></a>02361 
<a name="l02362"></a>02362 
<a name="l02363"></a>02363 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02364"></a>02364 <span class="comment">// CLASS: GenomicRegionSetIndex                                                                //</span>
<a name="l02365"></a>02365 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02367"></a>02367 <span class="comment"></span><span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02368"></a><a class="code" href="classGenomicRegionSetIndex.html">02368</a> <span class="keyword">class </span><a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a> 
<a name="l02369"></a>02369 {
<a name="l02370"></a>02370  <span class="keyword">public</span>:
<a name="l02372"></a>02372 
<a name="l02376"></a>02376   <a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *regSet, <span class="keywordtype">char</span> *bin_bits=NULL);
<a name="l02377"></a>02377 
<a name="l02378"></a>02378 
<a name="l02380"></a>02380   ~<a class="code" href="classGenomicRegionSetIndex.html" title="Class for creating an index for genomic regions based on the &quot;genome binning&quot; approach by Kent et al...">GenomicRegionSetIndex</a>();
<a name="l02381"></a>02381 
<a name="l02382"></a>02382   
<a name="l02384"></a>02384   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *i);
<a name="l02385"></a>02385 
<a name="l02386"></a>02386  
<a name="l02388"></a>02388   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02389"></a>02389 
<a name="l02390"></a>02390  
<a name="l02391"></a>02391   <span class="comment">// data</span>
<a name="l02392"></a>02392  <span class="keyword">public</span>: 
<a name="l02393"></a><a class="code" href="classGenomicRegionSetIndex.html#a69385a469c77555e783aa2801c7608cf">02393</a>   <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *<a class="code" href="classGenomicRegionSetIndex.html#a69385a469c77555e783aa2801c7608cf" title="pointer to index region set">regSet</a>;                                                     
<a name="l02394"></a><a class="code" href="classGenomicRegionSetIndex.html#ae9d898b579f6b1b37a466f13ce9b1524">02394</a>   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *<a class="code" href="classGenomicRegionSetIndex.html#ae9d898b579f6b1b37a466f13ce9b1524" title="pointer to the current index region">current_ireg</a>;                                          
<a name="l02395"></a><a class="code" href="classGenomicRegionSetIndex.html#a563202d9d5ee7e8b70fc510cdf02a35b">02395</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classGenomicRegionSetIndex.html#a563202d9d5ee7e8b70fc510cdf02a35b" title="pointer to the current query interval">current_qint</a>;                                        
<a name="l02396"></a><a class="code" href="classGenomicRegionSetIndex.html#a6f2134bcaf8718d2fb19444cd9c2dfa8">02396</a>   <span class="keyword">typedef</span> pair&lt;long int*,long int**&gt; <a class="code" href="classGenomicRegionSetIndex.html#a6f2134bcaf8718d2fb19444cd9c2dfa8" title="a two-dimensional grid of bins">BinSet</a>;            
<a name="l02397"></a><a class="code" href="classGenomicRegionSetIndex.html#a5c147e94bc5208615ceaee6bf65ddd99">02397</a>   <span class="keywordtype">int</span> <a class="code" href="classGenomicRegionSetIndex.html#a5c147e94bc5208615ceaee6bf65ddd99" title="number of levels in the two-dimensional grid of bins">n_levels</a>;                                                                         
<a name="l02398"></a><a class="code" href="classGenomicRegionSetIndex.html#ab458e36338671d25a4add20f3265473e">02398</a>   <span class="keywordtype">int</span> *<a class="code" href="classGenomicRegionSetIndex.html#ab458e36338671d25a4add20f3265473e" title="number of bits (per level) by which the start position is right-shifted to compute its bin...">n_bits</a>;                                                                          
<a name="l02399"></a><a class="code" href="classGenomicRegionSetIndex.html#a5876efc8c7ca70685413752a9aabd651">02399</a>   map&lt;string,BinSet*&gt; <a class="code" href="classGenomicRegionSetIndex.html#a5876efc8c7ca70685413752a9aabd651" title="the two-dimensional grid of bins per chromosome">index</a>;                                            
<a name="l02400"></a><a class="code" href="classGenomicRegionSetIndex.html#a70be632459700206f8d5ea74fe913d87">02400</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classGenomicRegionSetIndex.html#a70be632459700206f8d5ea74fe913d87" title="for each region, store a pointer to the &quot;next&quot; one in the same bin">r_next</a>;                                                                     
<a name="l02401"></a><a class="code" href="classGenomicRegionSetIndex.html#a55bdee4bb523a127ef744c9fc9eb53f3">02401</a>   <a class="code" href="classGenomicRegionSetIndex.html#a6f2134bcaf8718d2fb19444cd9c2dfa8" title="a two-dimensional grid of bins">BinSet</a> *<a class="code" href="classGenomicRegionSetIndex.html#a55bdee4bb523a127ef744c9fc9eb53f3" title="pointer to the current chromosome set of bins">current_binset</a>;                                                       
<a name="l02402"></a><a class="code" href="classGenomicRegionSetIndex.html#adc1e41cc8005f46f07db2bdbf86b86e9">02402</a>   <span class="keywordtype">bool</span> <a class="code" href="classGenomicRegionSetIndex.html#adc1e41cc8005f46f07db2bdbf86b86e9" title="&#39;true&#39;, if a new query is about to be processed">new_query</a>;                                                                       
<a name="l02403"></a>02403 };
<a name="l02404"></a>02404 
<a name="l02405"></a>02405 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02406"></a>02406 <span class="comment">// END CLASS: GenomicRegionSetIndex                                                            //</span>
<a name="l02407"></a>02407 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02408"></a>02408 
<a name="l02409"></a>02409 
<a name="l02410"></a>02410 
<a name="l02411"></a>02411 
<a name="l02412"></a>02412 
<a name="l02413"></a>02413 
<a name="l02414"></a>02414 
<a name="l02415"></a>02415 
<a name="l02416"></a>02416 
<a name="l02417"></a>02417 
<a name="l02418"></a>02418 
<a name="l02419"></a>02419 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02420"></a>02420 <span class="comment">// CLASS: UnsortedGenomicRegionSetOverlaps                                                     //</span>
<a name="l02421"></a>02421 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02423"></a>02423 <span class="comment"></span>
<a name="l02451"></a>02451 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02452"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html">02452</a> <span class="keyword">class </span><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02453"></a>02453 {
<a name="l02454"></a>02454  <span class="keyword">public</span>:
<a name="l02456"></a>02456 
<a name="l02461"></a>02461   <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet, <span class="keywordtype">char</span> *bin_bits=NULL);
<a name="l02462"></a>02462 
<a name="l02463"></a>02463 
<a name="l02465"></a>02465   ~<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html" title="Class for computing overlaps between unsorted regions.">UnsortedGenomicRegionSetOverlaps</a>();
<a name="l02466"></a>02466 
<a name="l02467"></a>02467   
<a name="l02469"></a>02469   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery();
<a name="l02470"></a>02470 
<a name="l02471"></a>02471 
<a name="l02473"></a>02473   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery();
<a name="l02474"></a>02474 
<a name="l02475"></a>02475 
<a name="l02477"></a>02477   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch();
<a name="l02478"></a>02478 
<a name="l02479"></a>02479  
<a name="l02481"></a>02481   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02482"></a>02482 
<a name="l02483"></a>02483  
<a name="l02485"></a>02485   <span class="keywordtype">bool</span> Done();
<a name="l02486"></a>02486  
<a name="l02487"></a>02487 
<a name="l02488"></a>02488   
<a name="l02489"></a>02489   <span class="comment">// data</span>
<a name="l02490"></a>02490  <span class="keyword">public</span>: 
<a name="l02491"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14">02491</a>   <span class="keyword">typedef</span> pair&lt;long int*,long int**&gt; <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14" title="a two-dimensional grid of bins">BinSet</a>;            
<a name="l02492"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a251e09b7f0a13a56b902c30791228ce4">02492</a>   <span class="keywordtype">long</span> <span class="keywordtype">int</span> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a251e09b7f0a13a56b902c30791228ce4" title="for each region, store a pointer to the &quot;next&quot; one in the same bin">r_next</a>;                                                                     
<a name="l02493"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a1cc721de471ed40dcc7d8eae258e5833">02493</a>   <span class="keywordtype">int</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a1cc721de471ed40dcc7d8eae258e5833" title="number of levels in the two-dimensional grid of bins">n_levels</a>;                                                                         
<a name="l02494"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a76ecebfcf839a940bdae26069ceef748">02494</a>   <span class="keywordtype">int</span> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a76ecebfcf839a940bdae26069ceef748" title="number of bits (per level) by which the start position is right-shifted to compute its bin...">n_bits</a>;                                                                          
<a name="l02495"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abeeb94562abdf4aebc36684959a5bc78">02495</a>   map&lt;string,BinSet*&gt; <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abeeb94562abdf4aebc36684959a5bc78" title="the two-dimensional grid of bins per chromosome">index</a>;                                            
<a name="l02496"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a6bbb87f3c66d04527f06a5761df1ab56">02496</a>   <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14" title="a two-dimensional grid of bins">BinSet</a> *<a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a6bbb87f3c66d04527f06a5761df1ab56" title="pointer to the current chromosome set of bins">current_binset</a>;                                                       
<a name="l02497"></a><a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abafc544ce2c677ce071fa4e30ec2ecaf">02497</a>   <span class="keywordtype">bool</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#abafc544ce2c677ce071fa4e30ec2ecaf" title="&#39;true&#39;, if a new query is about to be processed">new_query</a>;                                                                       
<a name="l02498"></a>02498 };
<a name="l02499"></a>02499 
<a name="l02500"></a>02500 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02501"></a>02501 <span class="comment">// END CLASS: UnsortedGenomicRegionSetOverlaps                                                 //</span>
<a name="l02502"></a>02502 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02503"></a>02503 
<a name="l02504"></a>02504 
<a name="l02505"></a>02505 
<a name="l02506"></a>02506 
<a name="l02507"></a>02507 
<a name="l02508"></a>02508 
<a name="l02509"></a>02509 
<a name="l02510"></a>02510 
<a name="l02511"></a>02511 
<a name="l02512"></a>02512 
<a name="l02513"></a>02513 
<a name="l02514"></a>02514 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02515"></a>02515 <span class="comment">// CLASS: SortedGenomicRegionSetOverlaps                                                       //</span>
<a name="l02516"></a>02516 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02518"></a>02518 <span class="comment"></span>
<a name="l02577"></a>02577 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02578"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html">02578</a> <span class="keyword">class </span><a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a> : <span class="keyword">public</span> <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a>
<a name="l02579"></a>02579 {
<a name="l02580"></a>02580  <span class="keyword">public</span>:
<a name="l02582"></a>02582 
<a name="l02587"></a>02587   <a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a>(<a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *QuerySet, <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> *IndexSet, <span class="keywordtype">bool</span> sorted_by_strand);
<a name="l02588"></a>02588 
<a name="l02589"></a>02589 
<a name="l02591"></a>02591   ~<a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a>();
<a name="l02592"></a>02592 
<a name="l02593"></a>02593   
<a name="l02595"></a>02595   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetQuery();
<a name="l02596"></a>02596 
<a name="l02597"></a>02597 
<a name="l02599"></a>02599   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextQuery();
<a name="l02600"></a>02600 
<a name="l02601"></a>02601 
<a name="l02603"></a>02603   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *GetMatch();
<a name="l02604"></a>02604 
<a name="l02605"></a>02605  
<a name="l02607"></a>02607   <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *NextMatch();
<a name="l02608"></a>02608 
<a name="l02609"></a>02609  
<a name="l02611"></a>02611   <span class="keywordtype">bool</span> Done();
<a name="l02612"></a>02612  
<a name="l02613"></a>02613 
<a name="l02614"></a>02614  <span class="keyword">private</span>:
<a name="l02615"></a>02615   <span class="keywordtype">void</span> LoadIndexBuffer();                               
<a name="l02616"></a>02616   <span class="keywordtype">void</span> ClearIndexBuffer();                              
<a name="l02617"></a>02617 
<a name="l02618"></a>02618   <span class="comment">// data</span>
<a name="l02619"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3">02619</a>   <span class="keywordtype">bool</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3" title="if true, query and index regions sorted by strand">sorted_by_strand</a>;                                
<a name="l02620"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">02620</a>   GenomicRegionList <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f" title="temporary buffer containing index regions that overlap with current query; the buffer is cleared only...">IRegBuffer</a>;                         
<a name="l02621"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a71d6eea0c4bf9f19be6cf0c0c562f297">02621</a>   GenomicRegionList::iterator <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a71d6eea0c4bf9f19be6cf0c0c562f297" title="iterator on SortedGenomicRegionSetOverlaps::IRegBuffer">IRegBufferIterator</a>;       
<a name="l02622"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a962f8e6ce8c15703cfefa574a07ce42c">02622</a>   <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *<a class="code" href="classSortedGenomicRegionSetOverlaps.html#a962f8e6ce8c15703cfefa574a07ce42c" title="keeps tracks of the (linked) interval stored in SortedGenomicRegionSetOverlaps::IRegBuffer">ireg_buffer_interval</a>;                
<a name="l02623"></a><a class="code" href="classSortedGenomicRegionSetOverlaps.html#a349d589ff006f742d635b16987fe7b6b">02623</a>   <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a349d589ff006f742d635b16987fe7b6b" title="keeps track of the maximum size of SortedGenomicRegionSetOverlaps::IRegBuffer">max_ireg_buffer_size</a>;               
<a name="l02624"></a>02624 };
<a name="l02625"></a>02625 
<a name="l02626"></a>02626 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02627"></a>02627 <span class="comment">// END CLASS: SortedGenomicRegionSetOverlaps                                                   //</span>
<a name="l02628"></a>02628 <span class="comment">//---------------------------------------------------------------------------------------------//</span>
<a name="l02629"></a>02629 
<a name="l02630"></a>02630 
<a name="l02631"></a>02631 
<a name="l02632"></a>02632 
<a name="l02633"></a>02633 
<a name="l02634"></a>02634 
<a name="l02635"></a>02635 
<a name="l02636"></a>02636 
<a name="l02637"></a>02637 
<a name="l02638"></a>02638 
<a name="l02639"></a>02639 <span class="keywordtype">char</span> ProcessStrand(<span class="keywordtype">char</span> *strand);
<a name="l02640"></a>02640 <span class="keywordtype">void</span> PrintChromosome(<span class="keywordtype">char</span> *chromosome, <span class="keywordtype">bool</span> convert);
<a name="l02641"></a>02641 <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *RegCenter(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r);
<a name="l02642"></a>02642 <span class="keywordtype">long</span> <span class="keywordtype">int</span> GetStart(<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *r, <span class="keywordtype">char</span> preprocess);
<a name="l02643"></a>02643 
<a name="l02644"></a>02644 StringLIntMap *ReadBounds(<span class="keywordtype">char</span> *genome_reg_file, <span class="keywordtype">bool</span> verbose=<span class="keyword">false</span>);           
<a name="l02645"></a>02645 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcBoundSize(StringLIntMap *bounds);
<a name="l02646"></a>02646 <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> CalcRegSize(<span class="keywordtype">char</span> *reg_file);
<a name="l02647"></a>02647 <span class="keywordtype">bool</span> CompareGenomicIntervals(<a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *I, <a class="code" href="classGenomicInterval.html" title="set of CIGAR operations that map to the fragment sequence (SAM format)">GenomicInterval</a> *J);
<a name="l02648"></a>02648 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsList *L);
<a name="l02649"></a>02649 <span class="keywordtype">void</span> SortGenomicIntervals(GenomicIntervalSetAsVector *V);
<a name="l02650"></a>02650 <span class="keywordtype">void</span> SortGenomicIntervals(<a class="code" href="classGenomicIntervalSetAsArray.html" title="[UNDER DEVELOPMENT] This class implements an array of genomic intervals.">GenomicIntervalSetAsArray</a> *A);
<a name="l02651"></a>02651 
<a name="l02652"></a>02652 
<a name="l02653"></a>02653 
<a name="l02654"></a>02654 
<a name="l02655"></a>02655 
</pre></div></div>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Tue Jun 5 2012 12:08:22 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
