\hypertarget{classChromosomes}{
\section{Chromosomes Class Reference}
\label{classChromosomes}\index{Chromosomes@{Chromosomes}}
}
This class is used to store chromosome DNA sequences.  


{\tt \#include $<$genomic\_\-intervals.h$>$}

\subsection*{Public Types}
\begin{CompactItemize}
\item 
\hypertarget{classChromosomes_4a233d2b7fabc1a589cbbafd5d0ea243}{
typedef \hyperlink{classChromosomes_8e40e7c99c6c0a40713c7785ba5b5ac4}{map}$<$ string, pair$<$ char $\ast$, size\_\-t $>$ $>$ \hyperlink{classChromosomes_4a233d2b7fabc1a589cbbafd5d0ea243}{ChromosomeSeqData}}
\label{classChromosomes_4a233d2b7fabc1a589cbbafd5d0ea243}

\begin{CompactList}\small\item\em for every chromosome, store a pointer to its sequence, and its size. \item\end{CompactList}\end{CompactItemize}
\subsection*{Public Member Functions}
\begin{CompactItemize}
\item 
\hyperlink{classChromosomes_a596430d4a05ce0dc4d4f43c650db83e}{Chromosomes} (char $\ast$\hyperlink{classChromosomes_8df459fcb5ad98a2524fd51ac2b5a60c}{map\_\-dir}, char $\ast$\hyperlink{classChromosomes_9c690f19e04e80b056dac2072476f24c}{map\_\-name})
\begin{CompactList}\small\item\em Class constructor. \item\end{CompactList}\item 
\hyperlink{classChromosomes_bb3466c6c94f42f22a872b043c415b67}{Chromosomes} (char $\ast$fasta\_\-file, bool \hyperlink{classChromosomes_954bae240564f6f589a30d253291eb75}{verbose})
\begin{CompactList}\small\item\em Class constructor. \item\end{CompactList}\item 
\hypertarget{classChromosomes_0b04a0781b2ff7088dd8046493385003}{
\hyperlink{classChromosomes_0b04a0781b2ff7088dd8046493385003}{$\sim$Chromosomes} ()}
\label{classChromosomes_0b04a0781b2ff7088dd8046493385003}

\begin{CompactList}\small\item\em Class destructor. \item\end{CompactList}\item 
\hypertarget{classChromosomes_2378b5cab810d50699d97883fe1c04bf}{
bool \hyperlink{classChromosomes_2378b5cab810d50699d97883fe1c04bf}{FindChromosome} (char $\ast$chromosome\_\-name)}
\label{classChromosomes_2378b5cab810d50699d97883fe1c04bf}

\begin{CompactList}\small\item\em Returns 'true' if chromosome name is found in the data. \item\end{CompactList}\item 
\hypertarget{classChromosomes_819d0fa5f41c034a45949156307d02e4}{
char $\ast$ \hyperlink{classChromosomes_819d0fa5f41c034a45949156307d02e4}{GetSeq} (\hyperlink{classGenomicInterval}{GenomicInterval} $\ast$I, bool replace=false)}
\label{classChromosomes_819d0fa5f41c034a45949156307d02e4}

\begin{CompactList}\small\item\em Returns interval sequence. \item\end{CompactList}\item 
\hypertarget{classChromosomes_c732efe7b6dcc645861507257bf66347}{
void \hyperlink{classChromosomes_c732efe7b6dcc645861507257bf66347}{PrintSeq} (\hyperlink{classGenomicInterval}{GenomicInterval} $\ast$I, bool replace=false)}
\label{classChromosomes_c732efe7b6dcc645861507257bf66347}

\begin{CompactList}\small\item\em Prints interval sequence. \item\end{CompactList}\item 
\hypertarget{classChromosomes_044da4339ffcc9a892ea35198bbb9f36}{
size\_\-t \hyperlink{classChromosomes_044da4339ffcc9a892ea35198bbb9f36}{GetSeqLength} (\hyperlink{classGenomicInterval}{GenomicInterval} $\ast$I)}
\label{classChromosomes_044da4339ffcc9a892ea35198bbb9f36}

\begin{CompactList}\small\item\em Calculates interval sequence length excluding 'N' characters. \item\end{CompactList}\item 
\hypertarget{classChromosomes_7f4e8938bd27e1fdb81f7ab33a44d62a}{
void \hyperlink{classChromosomes_7f4e8938bd27e1fdb81f7ab33a44d62a}{LoadChromosome} (string chromosome\_\-name)}
\label{classChromosomes_7f4e8938bd27e1fdb81f7ab33a44d62a}

\begin{CompactList}\small\item\em Loads chromosome sequence in memory. \item\end{CompactList}\end{CompactItemize}
\subsection*{Public Attributes}
\begin{CompactItemize}
\item 
\hypertarget{classChromosomes_954bae240564f6f589a30d253291eb75}{
bool \hyperlink{classChromosomes_954bae240564f6f589a30d253291eb75}{verbose}}
\label{classChromosomes_954bae240564f6f589a30d253291eb75}

\begin{CompactList}\small\item\em if 'true', print messages while loading chromosome sequences \item\end{CompactList}\item 
\hypertarget{classChromosomes_408a06b9ae635ceaf0b75452bcb62763}{
bool \hyperlink{classChromosomes_408a06b9ae635ceaf0b75452bcb62763}{load\_\-in\_\-memory}}
\label{classChromosomes_408a06b9ae635ceaf0b75452bcb62763}

\begin{CompactList}\small\item\em 'true' if reading from FASTA file \item\end{CompactList}\item 
\hypertarget{classChromosomes_8e40e7c99c6c0a40713c7785ba5b5ac4}{
StringMap \hyperlink{classChromosomes_8e40e7c99c6c0a40713c7785ba5b5ac4}{map}}
\label{classChromosomes_8e40e7c99c6c0a40713c7785ba5b5ac4}

\begin{CompactList}\small\item\em maps chromosome names to chromosome sequence file names \item\end{CompactList}\item 
\hypertarget{classChromosomes_8df459fcb5ad98a2524fd51ac2b5a60c}{
string \hyperlink{classChromosomes_8df459fcb5ad98a2524fd51ac2b5a60c}{map\_\-dir}}
\label{classChromosomes_8df459fcb5ad98a2524fd51ac2b5a60c}

\begin{CompactList}\small\item\em the directory where the chromosome map file is located \item\end{CompactList}\item 
\hypertarget{classChromosomes_9c690f19e04e80b056dac2072476f24c}{
string \hyperlink{classChromosomes_9c690f19e04e80b056dac2072476f24c}{map\_\-name}}
\label{classChromosomes_9c690f19e04e80b056dac2072476f24c}

\begin{CompactList}\small\item\em the name of the chromosome map file \item\end{CompactList}\item 
\hypertarget{classChromosomes_b3b502eec7bd2028297300aeb47b5844}{
string \hyperlink{classChromosomes_b3b502eec7bd2028297300aeb47b5844}{current\_\-chromosome\_\-name}}
\label{classChromosomes_b3b502eec7bd2028297300aeb47b5844}

\begin{CompactList}\small\item\em name of the currently used chromosome \item\end{CompactList}\item 
\hypertarget{classChromosomes_fb59cb80755187f4546573f10110e669}{
size\_\-t \hyperlink{classChromosomes_fb59cb80755187f4546573f10110e669}{current\_\-chromosome\_\-size}}
\label{classChromosomes_fb59cb80755187f4546573f10110e669}

\begin{CompactList}\small\item\em size of the currently used chromosome \item\end{CompactList}\item 
\hypertarget{classChromosomes_f640cf0f09b109b3d46d70aba9207a2e}{
char $\ast$ \hyperlink{classChromosomes_f640cf0f09b109b3d46d70aba9207a2e}{current\_\-chromosome\_\-seq}}
\label{classChromosomes_f640cf0f09b109b3d46d70aba9207a2e}

\begin{CompactList}\small\item\em pointer to the currently used chromosome sequence \item\end{CompactList}\item 
\hypertarget{classChromosomes_673169d55aa0b94815e17b41566d4d8c}{
\hyperlink{classChromosomes_4a233d2b7fabc1a589cbbafd5d0ea243}{ChromosomeSeqData} \hyperlink{classChromosomes_673169d55aa0b94815e17b41566d4d8c}{chromosome\_\-seq\_\-data}}
\label{classChromosomes_673169d55aa0b94815e17b41566d4d8c}

\begin{CompactList}\small\item\em all chromosomes sequences and their sizes are stored here \item\end{CompactList}\end{CompactItemize}


\subsection{Detailed Description}
This class is used to store chromosome DNA sequences. 

\subsection{Constructor \& Destructor Documentation}
\hypertarget{classChromosomes_a596430d4a05ce0dc4d4f43c650db83e}{
\index{Chromosomes@{Chromosomes}!Chromosomes@{Chromosomes}}
\index{Chromosomes@{Chromosomes}!Chromosomes@{Chromosomes}}
\subsubsection[Chromosomes]{\setlength{\rightskip}{0pt plus 5cm}Chromosomes::Chromosomes (char $\ast$ {\em map\_\-dir}, \/  char $\ast$ {\em map\_\-name})}}
\label{classChromosomes_a596430d4a05ce0dc4d4f43c650db83e}


Class constructor. 

\begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em map\_\-dir}]the directory where the chromosome map file is located \item[{\em map\_\-name}]the name of the chromosome map file\end{description}
\end{Desc}
Chromosome sequences are loaded on the fly from the $\ast$.dna files. For efficient sequence extraction, genomic intervals must be grouped by chromosome. If not, use the alternative constructor described below.

Example of map file (stored in \hyperlink{classChromosomes_9c690f19e04e80b056dac2072476f24c}{map\_\-name}):

1 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.1.fa.dna \par
 10 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.10.fa.dna \par
 11 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.11.fa.dna \par
 12 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.12.fa.dna \par
 13 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.13.fa.dna \par
 14 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.14.fa.dna \par
 15 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.15.fa.dna \par
 16 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.16.fa.dna \par
 17 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.17.fa.dna \par
 18 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.18.fa.dna \par
 19 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.19.fa.dna \par
 2 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.2.fa.dna \par
 3 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.3.fa.dna \par
 4 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.4.fa.dna \par
 5 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.5.fa.dna \par
 6 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.6.fa.dna \par
 7 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.7.fa.dna \par
 8 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.8.fa.dna \par
 9 {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.9.fa.dna \par
 MT {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.MT.fa.dna \par
 X {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.X.fa.dna \par
 Y {\em $<$TAB$>$\/} Mus\_\-musculus.NCBIM37.52.dna.chromosome.Y.fa.dna \par


Each $\ast$.dna file contains the chromosome sequence in one line. \hypertarget{classChromosomes_bb3466c6c94f42f22a872b043c415b67}{
\index{Chromosomes@{Chromosomes}!Chromosomes@{Chromosomes}}
\index{Chromosomes@{Chromosomes}!Chromosomes@{Chromosomes}}
\subsubsection[Chromosomes]{\setlength{\rightskip}{0pt plus 5cm}Chromosomes::Chromosomes (char $\ast$ {\em fasta\_\-file}, \/  bool {\em verbose})}}
\label{classChromosomes_bb3466c6c94f42f22a872b043c415b67}


Class constructor. 

\begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em fasta\_\-file}]FASTA file containing chromosome sequences \item[{\em verbose}]if 'true', print messages while loading chromosome sequences\end{description}
\end{Desc}
Chromosome sequences are loaded in memory from the FASTA file. Example FASTA file:

$>$chromosome\_\-1 \par
 TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTA \par
 $>$chromosome\_\-2 \par
 CGTATCCCACACACCACACCCACACACCACACCCACACACACCCACACCC \par
 $>$chromosome\_\-3 \par
 TCCTCCCAGAATCTGGAGAGGTCAACCTGTTCTTCAAAGCAGTGGTGGAT \par
 $>$chromosome\_\-4 \par
 ACCCTAACCCTAACCCTAACCCTAACCCTAACCCTACCCTAACCCTAACC \par
 

The documentation for this class was generated from the following files:\begin{CompactItemize}
\item 
genomic\_\-intervals.h\item 
genomic\_\-intervals.cpp\end{CompactItemize}
