\hypertarget{classSortedGenomicRegionSetOverlaps}{
\section{SortedGenomicRegionSetOverlaps Class Reference}
\label{classSortedGenomicRegionSetOverlaps}\index{SortedGenomicRegionSetOverlaps@{SortedGenomicRegionSetOverlaps}}
}


This class is used to find overlaps between two genomic regions sets. See detailed description below for an example.  




{\ttfamily \#include $<$genomic\_\-intervals.h$>$}

Inheritance diagram for SortedGenomicRegionSetOverlaps:\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[height=2.000000cm]{classSortedGenomicRegionSetOverlaps}
\end{center}
\end{figure}
\subsection*{Public Member Functions}
\begin{DoxyCompactItemize}
\item 
\hyperlink{classSortedGenomicRegionSetOverlaps_a0213f6ea908db759f63cea80db4909aa}{SortedGenomicRegionSetOverlaps} (\hyperlink{classGenomicRegionSet}{GenomicRegionSet} $\ast$\hyperlink{classGenomicRegionSetOverlaps_ae513304379055f6d379bc5907733dbe2}{QuerySet}, \hyperlink{classGenomicRegionSet}{GenomicRegionSet} $\ast$\hyperlink{classGenomicRegionSetOverlaps_ac587bf854c827381493735b473622e03}{IndexSet}, bool \hyperlink{classSortedGenomicRegionSetOverlaps_af08c449faa6ba7be7c6ed9c07f4226e3}{sorted\_\-by\_\-strand})
\begin{DoxyCompactList}\small\item\em Class constructor. \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a1ac0a70899fce3684da0b7e915a13d1e}{
\hyperlink{classSortedGenomicRegionSetOverlaps_a1ac0a70899fce3684da0b7e915a13d1e}{$\sim$SortedGenomicRegionSetOverlaps} ()}
\label{classSortedGenomicRegionSetOverlaps_a1ac0a70899fce3684da0b7e915a13d1e}

\begin{DoxyCompactList}\small\item\em Class destructor. \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a3b187db0a12f876a9347b4a1641670a0}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classSortedGenomicRegionSetOverlaps_a3b187db0a12f876a9347b4a1641670a0}{GetQuery} ()}
\label{classSortedGenomicRegionSetOverlaps_a3b187db0a12f876a9347b4a1641670a0}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the current query region ('NULL' if no more query regions are available); also, \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} is automatically updated via \hyperlink{classSortedGenomicRegionSetOverlaps_ad894d118c61c3cd11be8b37b8f185120}{SortedGenomicRegionSetOverlaps::LoadIndexBuffer}. \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a127edf3e9d8fb2ebf189af00e6893e96}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classSortedGenomicRegionSetOverlaps_a127edf3e9d8fb2ebf189af00e6893e96}{NextQuery} ()}
\label{classSortedGenomicRegionSetOverlaps_a127edf3e9d8fb2ebf189af00e6893e96}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the next query region ('NULL' if no more query regions are available); also, \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} is automatically updated via \hyperlink{classSortedGenomicRegionSetOverlaps_ad894d118c61c3cd11be8b37b8f185120}{SortedGenomicRegionSetOverlaps::LoadIndexBuffer}. \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_ad0f7b48e1eb685f0aa62aa799088dbc7}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classSortedGenomicRegionSetOverlaps_ad0f7b48e1eb685f0aa62aa799088dbc7}{GetMatch} ()}
\label{classSortedGenomicRegionSetOverlaps_ad0f7b48e1eb685f0aa62aa799088dbc7}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a756bad7ba9a862ae27cd9ad52acbeef2}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classSortedGenomicRegionSetOverlaps_a756bad7ba9a862ae27cd9ad52acbeef2}{NextMatch} ()}
\label{classSortedGenomicRegionSetOverlaps_a756bad7ba9a862ae27cd9ad52acbeef2}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a726105d1ee63db6271a90e8092e243c9}{
bool \hyperlink{classSortedGenomicRegionSetOverlaps_a726105d1ee63db6271a90e8092e243c9}{Done} ()}
\label{classSortedGenomicRegionSetOverlaps_a726105d1ee63db6271a90e8092e243c9}

\begin{DoxyCompactList}\small\item\em Returns true, if no more query or index regions are available and the \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} buffer is empty. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Private Member Functions}
\begin{DoxyCompactItemize}
\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_ad894d118c61c3cd11be8b37b8f185120}{
void \hyperlink{classSortedGenomicRegionSetOverlaps_ad894d118c61c3cd11be8b37b8f185120}{LoadIndexBuffer} ()}
\label{classSortedGenomicRegionSetOverlaps_ad894d118c61c3cd11be8b37b8f185120}

\begin{DoxyCompactList}\small\item\em manages \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_ac5fe4552253a9d9d802d76b0f4356947}{
void \hyperlink{classSortedGenomicRegionSetOverlaps_ac5fe4552253a9d9d802d76b0f4356947}{ClearIndexBuffer} ()}
\label{classSortedGenomicRegionSetOverlaps_ac5fe4552253a9d9d802d76b0f4356947}

\begin{DoxyCompactList}\small\item\em clears \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Private Attributes}
\begin{DoxyCompactItemize}
\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_af08c449faa6ba7be7c6ed9c07f4226e3}{
bool \hyperlink{classSortedGenomicRegionSetOverlaps_af08c449faa6ba7be7c6ed9c07f4226e3}{sorted\_\-by\_\-strand}}
\label{classSortedGenomicRegionSetOverlaps_af08c449faa6ba7be7c6ed9c07f4226e3}

\begin{DoxyCompactList}\small\item\em if true, query and index regions sorted by strand \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{
GenomicRegionList \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{IRegBuffer}}
\label{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}

\begin{DoxyCompactList}\small\item\em temporary buffer containing index regions that overlap with current query; the buffer is cleared only if the next query has no overlap \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a71d6eea0c4bf9f19be6cf0c0c562f297}{
GenomicRegionList::iterator \hyperlink{classSortedGenomicRegionSetOverlaps_a71d6eea0c4bf9f19be6cf0c0c562f297}{IRegBufferIterator}}
\label{classSortedGenomicRegionSetOverlaps_a71d6eea0c4bf9f19be6cf0c0c562f297}

\begin{DoxyCompactList}\small\item\em iterator on \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a962f8e6ce8c15703cfefa574a07ce42c}{
\hyperlink{classGenomicInterval}{GenomicInterval} $\ast$ \hyperlink{classSortedGenomicRegionSetOverlaps_a962f8e6ce8c15703cfefa574a07ce42c}{ireg\_\-buffer\_\-interval}}
\label{classSortedGenomicRegionSetOverlaps_a962f8e6ce8c15703cfefa574a07ce42c}

\begin{DoxyCompactList}\small\item\em keeps tracks of the (linked) interval stored in \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} \end{DoxyCompactList}\item 
\hypertarget{classSortedGenomicRegionSetOverlaps_a349d589ff006f742d635b16987fe7b6b}{
unsigned long int \hyperlink{classSortedGenomicRegionSetOverlaps_a349d589ff006f742d635b16987fe7b6b}{max\_\-ireg\_\-buffer\_\-size}}
\label{classSortedGenomicRegionSetOverlaps_a349d589ff006f742d635b16987fe7b6b}

\begin{DoxyCompactList}\small\item\em keeps track of the maximum size of \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} \end{DoxyCompactList}\end{DoxyCompactItemize}


\subsection{Detailed Description}
This class is used to find overlaps between two genomic regions sets. See detailed description below for an example. 

A simple example for computing RNAseq read densities in known exons (this is actually implemented in the {\bfseries genomic\_\-overlaps} command-\/line tool as 'density' operation): 
\begin{DoxyCode}
    #include "core.h"
    #include "genomic_intervals.h"
 
    // open region sets
    char *REF_REG_FILE = argv[next_arg];
    char *TEST_REG_FILE = next_arg+1==argc ? NULL : argv[next_arg+1];
    GenomicRegionSet RefRegSet(REF_REG_FILE,BUFFER_SIZE,VERBOSE,true);
    GenomicRegionSet TestRegSet(TEST_REG_FILE,BUFFER_SIZE,VERBOSE,false);

    // process overlaps
    GenomicRegionSetOverlaps *overlaps = new SortedGenomicRegionSetOverlaps(&Test
      RegSet,&RefRegSet,SORTED_BY_STRAND);
    unsigned long int *coverage = overlaps->CalcIndexCoverage(MATCH_GAPS,IGNORE_S
      TRAND,USE_VALUES); 
    Progress PRG("Printing densities...",RefRegSet.n_regions);
    for (long int k=0; k<RefRegSet.n_regions; k++) {
      GenomicRegion *qreg = RefRegSet.R[k];
      long int qreg_size = MATCH_GAPS ? (qreg->I.back()->STOP-qreg->I.front()->ST
      ART+1) : qreg->GetSize();
      double density = (double)coverage[k]/qreg_size;
      if (density>=MIN_DENSITY) printf("%s\t%.4e\n", qreg->LABEL, density);
      PRG.Check();
    }
    PRG.Done();
    delete coverage;
    delete overlaps;
\end{DoxyCode}
 A more complex example for creating ChIPseq read profiles in TSS regions (this is actually implemented in the {\bfseries genomic\_\-overlaps} command-\/line tool as 'offset' operation): 
\begin{DoxyCode}
    #include "core.h"
    #include "genomic_intervals.h"
 
    // open region sets
    char *QUERY_REG_FILE = "chipseq.reads.reg";
    char *INDEX_REG_FILE = "TSS.flank10kb.reg";
    char *OFFSET_OP = "5p";
    bool SORTED_BY_STRAND = false; 
    bool MATCH_GAPS = true;
    bool IGNORE_STRAND = true;
    GenomicRegionSet QueryRegSet(QUERY_REG_FILE,10000,true,false);
    GenomicRegionSet IndexRegSet(INDEX_REG_FILE,10000,true,false);

    // process overlaps
    SortedGenomicRegionSetOverlaps Overlaps(&QueryRegSet,&IndexRegSet,SORTED_BY_S
      TRAND);
    Progress PRG("Processing queries...",1);
    for (GenomicRegion *qreg=Overlaps.GetQuery(); Overlaps.Done()==false; qreg=Ov
      erlaps.NextQuery()) {
      size_t Qsize = qreg->GetSize();
      for (GenomicRegion *ireg=Overlaps.GetOverlap(MATCH_GAPS,IGNORE_STRAND); ire
      g!=NULL; ireg=Overlaps.NextOverlap(MATCH_GAPS,IGNORE_STRAND)) {
        cout << qreg->LABEL << '\t';
        long int start_offset, stop_offset;
        ireg->I.front()->GetOffsetFrom(qreg->I.front(),OFFSET_OP,IGNORE_STRAND,&s
      tart_offset,&stop_offset);
        printf("%f", ((float)start_offset/Qsize+(float)stop_offset/Qsize)/2); 
        cout << '\n';
      }
      PRG.Check();
    }
    PRG.Done();
\end{DoxyCode}
 

\subsection{Constructor \& Destructor Documentation}
\hypertarget{classSortedGenomicRegionSetOverlaps_a0213f6ea908db759f63cea80db4909aa}{
\index{SortedGenomicRegionSetOverlaps@{SortedGenomicRegionSetOverlaps}!SortedGenomicRegionSetOverlaps@{SortedGenomicRegionSetOverlaps}}
\index{SortedGenomicRegionSetOverlaps@{SortedGenomicRegionSetOverlaps}!SortedGenomicRegionSetOverlaps@{SortedGenomicRegionSetOverlaps}}
\subsubsection[{SortedGenomicRegionSetOverlaps}]{\setlength{\rightskip}{0pt plus 5cm}SortedGenomicRegionSetOverlaps::SortedGenomicRegionSetOverlaps (
\begin{DoxyParamCaption}
\item[{{\bf GenomicRegionSet} $\ast$}]{QuerySet, }
\item[{{\bf GenomicRegionSet} $\ast$}]{IndexSet, }
\item[{bool}]{sorted\_\-by\_\-strand}
\end{DoxyParamCaption}
)}}
\label{classSortedGenomicRegionSetOverlaps_a0213f6ea908db759f63cea80db4909aa}


Class constructor. 


\begin{DoxyParams}{Parameters}
{\em QuerySet} & pointer to query region set \\
\hline
{\em IndexSet} & pointer to index region set \\
\hline
{\em sorted\_\-by\_\-strand} & if true, query and index regions are sorted by strand \\
\hline
\end{DoxyParams}


The documentation for this class was generated from the following files:\begin{DoxyCompactItemize}
\item 
genomic\_\-intervals.h\item 
genomic\_\-intervals.cpp\end{DoxyCompactItemize}
