\hypertarget{classGenomicRegionGFF}{
\section{GenomicRegionGFF Class Reference}
\label{classGenomicRegionGFF}\index{GenomicRegionGFF@{GenomicRegionGFF}}
}
This class processes genomic regions in SAM format.  


{\tt \#include $<$genomic\_\-intervals.h$>$}

Inheritance diagram for GenomicRegionGFF::\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[height=2cm]{classGenomicRegionGFF}
\end{center}
\end{figure}
\subsection*{Public Member Functions}
\begin{CompactItemize}
\item 
\hypertarget{classGenomicRegionGFF_9a49f0dc39116964f774e7ce78cfc827}{
\hyperlink{classGenomicRegionGFF_9a49f0dc39116964f774e7ce78cfc827}{GenomicRegionGFF} (\hyperlink{classFileBuffer}{FileBuffer} $\ast$B)}
\label{classGenomicRegionGFF_9a49f0dc39116964f774e7ce78cfc827}

\begin{CompactList}\small\item\em Class constructor. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_59dee8d6f9e73b1da46386497030beb2}{
\hyperlink{classGenomicRegionGFF_59dee8d6f9e73b1da46386497030beb2}{GenomicRegionGFF} (char $\ast$inp, long int \hyperlink{classGenomicRegion_efe2255aeed5338060190ded05cb9c0c}{n\_\-line}=0)}
\label{classGenomicRegionGFF_59dee8d6f9e73b1da46386497030beb2}

\begin{CompactList}\small\item\em Class constructor. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_3885d42a347283a5fed52791988076ad}{
\hyperlink{classGenomicRegionGFF_3885d42a347283a5fed52791988076ad}{$\sim$GenomicRegionGFF} ()}
\label{classGenomicRegionGFF_3885d42a347283a5fed52791988076ad}

\begin{CompactList}\small\item\em Class destructor. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_eb1390d42b388e32d98f58f57a5c169c}{
void \hyperlink{classGenomicRegionGFF_eb1390d42b388e32d98f58f57a5c169c}{Read} (char $\ast$inp, long int \hyperlink{classGenomicRegion_efe2255aeed5338060190ded05cb9c0c}{n\_\-line}=0)}
\label{classGenomicRegionGFF_eb1390d42b388e32d98f58f57a5c169c}

\begin{CompactList}\small\item\em Reads GFF format from input string {\bf inp}. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_3ffa71d729e4ff060a15a6bab1e924c8}{
void \hyperlink{classGenomicRegionGFF_3ffa71d729e4ff060a15a6bab1e924c8}{Print} (FILE $\ast$file\_\-ptr=stdout)}
\label{classGenomicRegionGFF_3ffa71d729e4ff060a15a6bab1e924c8}

\begin{CompactList}\small\item\em Prints label and genomic region intervals with a newline character at the end. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_96539f4372048c9f0d87826f5856a8ef}{
void \hyperlink{classGenomicRegionGFF_96539f4372048c9f0d87826f5856a8ef}{PrintConstrained} (\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$r, bool merge\_\-labels=false)}
\label{classGenomicRegionGFF_96539f4372048c9f0d87826f5856a8ef}

\begin{CompactList}\small\item\em Same as \hyperlink{classGenomicRegionGFF_3ffa71d729e4ff060a15a6bab1e924c8}{Print()}, but the interval is constrained between {\bf start} and {\bf stop}. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_4b331cd12506aa4793a2cf6ea7620936}{
void \hyperlink{classGenomicRegionGFF_4b331cd12506aa4793a2cf6ea7620936}{PrintModified} (char $\ast$label, long int start, long int stop)}
\label{classGenomicRegionGFF_4b331cd12506aa4793a2cf6ea7620936}

\begin{CompactList}\small\item\em Same as \hyperlink{classGenomicRegionGFF_3ffa71d729e4ff060a15a6bab1e924c8}{Print()}, but the label and interval's coordinates are modified. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_ee70fcc81c9a6629bf0aa1ec37cb1d64}{
void \hyperlink{classGenomicRegionGFF_ee70fcc81c9a6629bf0aa1ec37cb1d64}{Diff} ()}
\label{classGenomicRegionGFF_ee70fcc81c9a6629bf0aa1ec37cb1d64}

\begin{CompactList}\small\item\em Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries). \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_0ff4768578a03011fabad309072fade4}{
void \hyperlink{classGenomicRegionGFF_0ff4768578a03011fabad309072fade4}{RunCalcDistances} (char $\ast$op1, char $\ast$op2)}
\label{classGenomicRegionGFF_0ff4768578a03011fabad309072fade4}

\begin{CompactList}\small\item\em Print distance between successive intervals. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_77975c03dd39b404098eb519e81f67d4}{
void \hyperlink{classGenomicRegionGFF_77975c03dd39b404098eb519e81f67d4}{Divide} ()}
\label{classGenomicRegionGFF_77975c03dd39b404098eb519e81f67d4}

\begin{CompactList}\small\item\em Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region). \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_d4acc1231677efd876084ef36db82de1}{
void \hyperlink{classGenomicRegionGFF_d4acc1231677efd876084ef36db82de1}{RunDivide} ()}
\label{classGenomicRegionGFF_d4acc1231677efd876084ef36db82de1}

\begin{CompactList}\small\item\em Print intervals divided in half (i.e. two intervals for each original interval in the genomic region). \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_d35d9b6e9c57270c0d5648ae9f702959}{
void \hyperlink{classGenomicRegionGFF_d35d9b6e9c57270c0d5648ae9f702959}{Intersect} ()}
\label{classGenomicRegionGFF_d35d9b6e9c57270c0d5648ae9f702959}

\begin{CompactList}\small\item\em Print intersection among all intervals, i.e. maximum start to minimum stop position. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_aa9b1271561e2d62ef1be6464e589cf4}{
void \hyperlink{classGenomicRegionGFF_aa9b1271561e2d62ef1be6464e589cf4}{Select} (bool first, bool last, bool from5p, bool from3p)}
\label{classGenomicRegionGFF_aa9b1271561e2d62ef1be6464e589cf4}

\begin{CompactList}\small\item\em Selects a subset of intervals according to their relative positions. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_bc08552b0ad4ca512f79064b8cd64e54}{
void \hyperlink{classGenomicRegionGFF_bc08552b0ad4ca512f79064b8cd64e54}{RunSplit} ()}
\label{classGenomicRegionGFF_bc08552b0ad4ca512f79064b8cd64e54}

\begin{CompactList}\small\item\em Print region intervals on separate lines. \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_3815f1b27aef231a674c527e11f25880}{
void \hyperlink{classGenomicRegionGFF_3815f1b27aef231a674c527e11f25880}{PrintWindows} (long int win\_\-step, long int win\_\-size)}
\label{classGenomicRegionGFF_3815f1b27aef231a674c527e11f25880}

\begin{CompactList}\small\item\em Print sliding windows (implemented only for single-interval regions). \item\end{CompactList}\end{CompactItemize}
\subsection*{Public Attributes}
\begin{CompactItemize}
\item 
\hypertarget{classGenomicRegionGFF_aec963f555e7e807fb5046566622f5ee}{
long int \hyperlink{classGenomicRegionGFF_aec963f555e7e807fb5046566622f5ee}{n\_\-tokens}}
\label{classGenomicRegionGFF_aec963f555e7e807fb5046566622f5ee}

\begin{CompactList}\small\item\em number of TAB-separated fields in GFF entry (= 11) \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_2f0d4a5f15d8a756233436af03341734}{
char $\ast$ \hyperlink{classGenomicRegionGFF_2f0d4a5f15d8a756233436af03341734}{SOURCE}}
\label{classGenomicRegionGFF_2f0d4a5f15d8a756233436af03341734}

\begin{CompactList}\small\item\em feature source \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_0de447fbfad3da5b7a9cf07a5a6980f0}{
char $\ast$ \hyperlink{classGenomicRegionGFF_0de447fbfad3da5b7a9cf07a5a6980f0}{FEATURE}}
\label{classGenomicRegionGFF_0de447fbfad3da5b7a9cf07a5a6980f0}

\begin{CompactList}\small\item\em feture type \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_0e92da1f0fd153504e0630f27229ba4d}{
char $\ast$ \hyperlink{classGenomicRegionGFF_0e92da1f0fd153504e0630f27229ba4d}{SCORE}}
\label{classGenomicRegionGFF_0e92da1f0fd153504e0630f27229ba4d}

\begin{CompactList}\small\item\em score \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_42122a69eec385e0106e3861cb3b98f3}{
char \hyperlink{classGenomicRegionGFF_42122a69eec385e0106e3861cb3b98f3}{FRAME}}
\label{classGenomicRegionGFF_42122a69eec385e0106e3861cb3b98f3}

\begin{CompactList}\small\item\em coding frame \item\end{CompactList}\item 
\hypertarget{classGenomicRegionGFF_625a023f1bb24fc597e723919a036073}{
char $\ast$ \hyperlink{classGenomicRegionGFF_625a023f1bb24fc597e723919a036073}{COMMENT}}
\label{classGenomicRegionGFF_625a023f1bb24fc597e723919a036073}

\begin{CompactList}\small\item\em comment \item\end{CompactList}\end{CompactItemize}


\subsection{Detailed Description}
This class processes genomic regions in SAM format. 

The documentation for this class was generated from the following files:\begin{CompactItemize}
\item 
genomic\_\-intervals.h\item 
genomic\_\-intervals.cpp\end{CompactItemize}
