\hypertarget{classUnsortedGenomicRegionSetOverlaps}{
\section{UnsortedGenomicRegionSetOverlaps Class Reference}
\label{classUnsortedGenomicRegionSetOverlaps}\index{UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}}
}


Class for computing overlaps between unsorted regions.  




{\ttfamily \#include $<$genomic\_\-intervals.h$>$}

Inheritance diagram for UnsortedGenomicRegionSetOverlaps:\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[height=2.000000cm]{classUnsortedGenomicRegionSetOverlaps}
\end{center}
\end{figure}
\subsection*{Public Types}
\begin{DoxyCompactItemize}
\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a0f0857dd93ad3c047d1f5056a57edc14}{
typedef pair$<$ long int $\ast$, long int $\ast$$\ast$ $>$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_a0f0857dd93ad3c047d1f5056a57edc14}{BinSet}}
\label{classUnsortedGenomicRegionSetOverlaps_a0f0857dd93ad3c047d1f5056a57edc14}

\begin{DoxyCompactList}\small\item\em a two-\/dimensional grid of bins \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Public Member Functions}
\begin{DoxyCompactItemize}
\item 
\hyperlink{classUnsortedGenomicRegionSetOverlaps_aaa858364bf0a3d7986c9d0838a73ccab}{UnsortedGenomicRegionSetOverlaps} (\hyperlink{classGenomicRegionSet}{GenomicRegionSet} $\ast$\hyperlink{classGenomicRegionSetOverlaps_ae513304379055f6d379bc5907733dbe2}{QuerySet}, \hyperlink{classGenomicRegionSet}{GenomicRegionSet} $\ast$\hyperlink{classGenomicRegionSetOverlaps_ac587bf854c827381493735b473622e03}{IndexSet}, char $\ast$bin\_\-bits=NULL)
\begin{DoxyCompactList}\small\item\em Class constructor. \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a7682171bb52c651af44e4e8f3bf33f52}{
\hyperlink{classUnsortedGenomicRegionSetOverlaps_a7682171bb52c651af44e4e8f3bf33f52}{$\sim$UnsortedGenomicRegionSetOverlaps} ()}
\label{classUnsortedGenomicRegionSetOverlaps_a7682171bb52c651af44e4e8f3bf33f52}

\begin{DoxyCompactList}\small\item\em Class destructor. \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_abf7e2f8d35f7a9cf8441caa9d765de44}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_abf7e2f8d35f7a9cf8441caa9d765de44}{GetQuery} ()}
\label{classUnsortedGenomicRegionSetOverlaps_abf7e2f8d35f7a9cf8441caa9d765de44}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the current query region ('NULL' if no more query regions are available). \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a96e68ef99a958665f5da74572edcc854}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_a96e68ef99a958665f5da74572edcc854}{NextQuery} ()}
\label{classUnsortedGenomicRegionSetOverlaps_a96e68ef99a958665f5da74572edcc854}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the next query region ('NULL' if no more query regions are available). \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a61c2c6e9203acb0e0a9bfa4f0176308f}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_a61c2c6e9203acb0e0a9bfa4f0176308f}{GetMatch} ()}
\label{classUnsortedGenomicRegionSetOverlaps_a61c2c6e9203acb0e0a9bfa4f0176308f}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a0bcd12697611caa5216db7a4c4fca4e2}{
\hyperlink{classGenomicRegion}{GenomicRegion} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_a0bcd12697611caa5216db7a4c4fca4e2}{NextMatch} ()}
\label{classUnsortedGenomicRegionSetOverlaps_a0bcd12697611caa5216db7a4c4fca4e2}

\begin{DoxyCompactList}\small\item\em Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_afc18fc203debad6a2d48200ea8ccb42f}{
bool \hyperlink{classUnsortedGenomicRegionSetOverlaps_afc18fc203debad6a2d48200ea8ccb42f}{Done} ()}
\label{classUnsortedGenomicRegionSetOverlaps_afc18fc203debad6a2d48200ea8ccb42f}

\begin{DoxyCompactList}\small\item\em Returns true, if no more query or index regions are available and the \hyperlink{classSortedGenomicRegionSetOverlaps_a19fa18e6abd9f045786698fff48a445f}{SortedGenomicRegionSetOverlaps::IRegBuffer} buffer is empty. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Public Attributes}
\begin{DoxyCompactItemize}
\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a251e09b7f0a13a56b902c30791228ce4}{
long int $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_a251e09b7f0a13a56b902c30791228ce4}{r\_\-next}}
\label{classUnsortedGenomicRegionSetOverlaps_a251e09b7f0a13a56b902c30791228ce4}

\begin{DoxyCompactList}\small\item\em for each region store, store a pointer to the \char`\"{}next\char`\"{} one in the same bin \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a1cc721de471ed40dcc7d8eae258e5833}{
int \hyperlink{classUnsortedGenomicRegionSetOverlaps_a1cc721de471ed40dcc7d8eae258e5833}{n\_\-levels}}
\label{classUnsortedGenomicRegionSetOverlaps_a1cc721de471ed40dcc7d8eae258e5833}

\begin{DoxyCompactList}\small\item\em number of levels in the two-\/dimensional grid of bins \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a76ecebfcf839a940bdae26069ceef748}{
int $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_a76ecebfcf839a940bdae26069ceef748}{n\_\-bits}}
\label{classUnsortedGenomicRegionSetOverlaps_a76ecebfcf839a940bdae26069ceef748}

\begin{DoxyCompactList}\small\item\em number of bits (per level) by which the start position is right-\/shifted to compute its bin \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_abeeb94562abdf4aebc36684959a5bc78}{
map$<$ string, \hyperlink{classUnsortedGenomicRegionSetOverlaps_a0f0857dd93ad3c047d1f5056a57edc14}{BinSet} $\ast$ $>$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_abeeb94562abdf4aebc36684959a5bc78}{index}}
\label{classUnsortedGenomicRegionSetOverlaps_abeeb94562abdf4aebc36684959a5bc78}

\begin{DoxyCompactList}\small\item\em the two-\/dimensional grid of bins per chromosome \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_a6bbb87f3c66d04527f06a5761df1ab56}{
\hyperlink{classUnsortedGenomicRegionSetOverlaps_a0f0857dd93ad3c047d1f5056a57edc14}{BinSet} $\ast$ \hyperlink{classUnsortedGenomicRegionSetOverlaps_a6bbb87f3c66d04527f06a5761df1ab56}{current\_\-binset}}
\label{classUnsortedGenomicRegionSetOverlaps_a6bbb87f3c66d04527f06a5761df1ab56}

\begin{DoxyCompactList}\small\item\em pointer to the current chromosome set of bins \end{DoxyCompactList}\item 
\hypertarget{classUnsortedGenomicRegionSetOverlaps_abafc544ce2c677ce071fa4e30ec2ecaf}{
bool \hyperlink{classUnsortedGenomicRegionSetOverlaps_abafc544ce2c677ce071fa4e30ec2ecaf}{new\_\-query}}
\label{classUnsortedGenomicRegionSetOverlaps_abafc544ce2c677ce071fa4e30ec2ecaf}

\begin{DoxyCompactList}\small\item\em 'true', if a new query is about to be processed \end{DoxyCompactList}\end{DoxyCompactItemize}


\subsection{Detailed Description}
Class for computing overlaps between unsorted regions. 

A simple example for computing RNAseq read densities in known exons (this is actually implemented in the {\bfseries genomic\_\-overlaps} command-\/line tool as 'density' operation): 
\begin{DoxyCode}
    #include "core.h"
    #include "genomic_intervals.h"
    
    // open region sets
    char *REF_REG_FILE = argv[next_arg];
    char *TEST_REG_FILE = next_arg+1==argc ? NULL : argv[next_arg+1];
    GenomicRegionSet RefRegSet(REF_REG_FILE,BUFFER_SIZE,VERBOSE,true);
    GenomicRegionSet TestRegSet(TEST_REG_FILE,BUFFER_SIZE,VERBOSE,false);

    // process overlaps
    GenomicRegionSetOverlaps *overlaps = new UnsortedGenomicRegionSetOverlaps(&Te
      stRegSet,&RefRegSet);
    unsigned long int *coverage = overlaps->CalcIndexCoverage(MATCH_GAPS,IGNORE_S
      TRAND,USE_VALUES); 
    Progress PRG("Printing densities...",RefRegSet.n_regions);
    for (long int k=0; k<RefRegSet.n_regions; k++) {
      GenomicRegion *qreg = RefRegSet.R[k];
      long int qreg_size = MATCH_GAPS ? (qreg->I.back()->STOP-qreg->I.front()->ST
      ART+1) : qreg->GetSize();
      double density = (double)coverage[k]/qreg_size;
      if (density>=MIN_DENSITY) printf("%s\t%.4e\n", qreg->LABEL, density);
      PRG.Check();
    }
    PRG.Done();
    delete coverage;
    delete overlaps;
\end{DoxyCode}
 

\subsection{Constructor \& Destructor Documentation}
\hypertarget{classUnsortedGenomicRegionSetOverlaps_aaa858364bf0a3d7986c9d0838a73ccab}{
\index{UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}!UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}}
\index{UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}!UnsortedGenomicRegionSetOverlaps@{UnsortedGenomicRegionSetOverlaps}}
\subsubsection[{UnsortedGenomicRegionSetOverlaps}]{\setlength{\rightskip}{0pt plus 5cm}UnsortedGenomicRegionSetOverlaps::UnsortedGenomicRegionSetOverlaps (
\begin{DoxyParamCaption}
\item[{{\bf GenomicRegionSet} $\ast$}]{QuerySet, }
\item[{{\bf GenomicRegionSet} $\ast$}]{IndexSet, }
\item[{char $\ast$}]{bin\_\-bits = {\ttfamily NULL}}
\end{DoxyParamCaption}
)}}
\label{classUnsortedGenomicRegionSetOverlaps_aaa858364bf0a3d7986c9d0838a73ccab}


Class constructor. 


\begin{DoxyParams}{Parameters}
{\em QuerySet} & pointer to query region set \\
\hline
{\em IndexSet} & pointer to index region set \\
\hline
{\em bin\_\-bits} & number of shift-\/bits per bin level, e.g. \char`\"{}10,15,18\char`\"{} \\
\hline
\end{DoxyParams}


The documentation for this class was generated from the following files:\begin{DoxyCompactItemize}
\item 
genomic\_\-intervals.h\item 
genomic\_\-intervals.cpp\end{DoxyCompactItemize}
