<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>UnsortedGenomicRegionSetOverlaps Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a>  </div>
  <div class="headertitle">
<div class="title">UnsortedGenomicRegionSetOverlaps Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="UnsortedGenomicRegionSetOverlaps" --><!-- doxytag: inherits="GenomicRegionSetOverlaps" -->
<p>Class for computing overlaps between unsorted regions.  
 <a href="classUnsortedGenomicRegionSetOverlaps.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for UnsortedGenomicRegionSetOverlaps:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classUnsortedGenomicRegionSetOverlaps.png" usemap="#UnsortedGenomicRegionSetOverlaps_map" alt=""/>
  <map id="UnsortedGenomicRegionSetOverlaps_map" name="UnsortedGenomicRegionSetOverlaps_map">
<area href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions." alt="GenomicRegionSetOverlaps" shape="rect" coords="0,0,220,24"/>
</map>
 </div></div>

<p><a href="classUnsortedGenomicRegionSetOverlaps-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-types"></a>
Public Types</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0f0857dd93ad3c047d1f5056a57edc14"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::BinSet" ref="a0f0857dd93ad3c047d1f5056a57edc14" args="" -->
typedef pair&lt; long int *, long <br class="typebreak"/>
int ** &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14">BinSet</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">a two-dimensional grid of bins <br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a8b9f403a94c51065e1c6af7f6180b3ba">UnsortedGenomicRegionSetOverlaps</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2">QuerySet</a>, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03">IndexSet</a>, const char *bin_bits=NULL)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor.  <a href="#a8b9f403a94c51065e1c6af7f6180b3ba"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7682171bb52c651af44e4e8f3bf33f52"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::~UnsortedGenomicRegionSetOverlaps" ref="a7682171bb52c651af44e4e8f3bf33f52" args="()" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a7682171bb52c651af44e4e8f3bf33f52">~UnsortedGenomicRegionSetOverlaps</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class destructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abf7e2f8d35f7a9cf8441caa9d765de44"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::GetQuery" ref="abf7e2f8d35f7a9cf8441caa9d765de44" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#abf7e2f8d35f7a9cf8441caa9d765de44">GetQuery</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the current query region ('NULL' if no more query regions are available). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a96e68ef99a958665f5da74572edcc854"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::NextQuery" ref="a96e68ef99a958665f5da74572edcc854" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a96e68ef99a958665f5da74572edcc854">NextQuery</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the next query region ('NULL' if no more query regions are available). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a61c2c6e9203acb0e0a9bfa4f0176308f"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::GetMatch" ref="a61c2c6e9203acb0e0a9bfa4f0176308f" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a61c2c6e9203acb0e0a9bfa4f0176308f">GetMatch</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0bcd12697611caa5216db7a4c4fca4e2"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::NextMatch" ref="a0bcd12697611caa5216db7a4c4fca4e2" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a0bcd12697611caa5216db7a4c4fca4e2">NextMatch</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afc18fc203debad6a2d48200ea8ccb42f"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::Done" ref="afc18fc203debad6a2d48200ea8ccb42f" args="()" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#afc18fc203debad6a2d48200ea8ccb42f">Done</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns true, if no more query or index regions are available and the <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> buffer is empty. <br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a251e09b7f0a13a56b902c30791228ce4"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::r_next" ref="a251e09b7f0a13a56b902c30791228ce4" args="" -->
long int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a251e09b7f0a13a56b902c30791228ce4">r_next</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">for each region, store a pointer to the "next" one in the same bin <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1cc721de471ed40dcc7d8eae258e5833"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::n_levels" ref="a1cc721de471ed40dcc7d8eae258e5833" args="" -->
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a1cc721de471ed40dcc7d8eae258e5833">n_levels</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">number of levels in the two-dimensional grid of bins <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a76ecebfcf839a940bdae26069ceef748"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::n_bits" ref="a76ecebfcf839a940bdae26069ceef748" args="" -->
int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a76ecebfcf839a940bdae26069ceef748">n_bits</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">number of bits (per level) by which the start position is right-shifted to compute its bin <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abeeb94562abdf4aebc36684959a5bc78"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::index" ref="abeeb94562abdf4aebc36684959a5bc78" args="" -->
map&lt; string, <a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14">BinSet</a> * &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#abeeb94562abdf4aebc36684959a5bc78">index</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">the two-dimensional grid of bins per chromosome <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6bbb87f3c66d04527f06a5761df1ab56"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::current_binset" ref="a6bbb87f3c66d04527f06a5761df1ab56" args="" -->
<a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a0f0857dd93ad3c047d1f5056a57edc14">BinSet</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#a6bbb87f3c66d04527f06a5761df1ab56">current_binset</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">pointer to the current chromosome set of bins <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abafc544ce2c677ce071fa4e30ec2ecaf"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::new_query" ref="abafc544ce2c677ce071fa4e30ec2ecaf" args="" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#abafc544ce2c677ce071fa4e30ec2ecaf">new_query</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">'true', if a new query is about to be processed <br/></td></tr>
</table>
<hr/><a name="details" id="details"></a><h2>Detailed Description</h2>
<div class="textblock"><p>Class for computing overlaps between unsorted regions. </p>
<p>A simple example for computing RNAseq read densities in known exons (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'density' operation): </p>
<div class="fragment"><pre class="fragment"><span class="preprocessor">    #include &quot;core.h&quot;</span>
<span class="preprocessor">    #include &quot;genomic_intervals.h&quot;</span>
    
    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *REF_REG_FILE = argv[next_arg];
    <span class="keywordtype">char</span> *TEST_REG_FILE = next_arg+1==argc ? NULL : argv[next_arg+1];
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> RefRegSet(REF_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">true</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> TestRegSet(TEST_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">false</span>);

    <span class="comment">// process overlaps</span>
    <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a> *overlaps = <span class="keyword">new</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#a8b9f403a94c51065e1c6af7f6180b3ba" title="Class constructor.">UnsortedGenomicRegionSetOverlaps</a>(&amp;TestRegSet,&amp;RefRegSet);
    <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *coverage = overlaps-&gt;<a class="code" href="classGenomicRegionSetOverlaps.html#a5c31081b154c624ba35487d988082ffa" title="Calculates the total overlap for each index region. The index region set must be loaded in memory...">CalcIndexCoverage</a>(MATCH_GAPS,IGNORE_STRAND,USE_VALUES); 
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">&quot;Printing densities...&quot;</span>,RefRegSet.n_regions);
    <span class="keywordflow">for</span> (<span class="keywordtype">long</span> <span class="keywordtype">int</span> k=0; k&lt;RefRegSet.n_regions; k++) {
      <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *qreg = RefRegSet.R[k];
      <span class="keywordtype">long</span> <span class="keywordtype">int</span> qreg_size = MATCH_GAPS ? (qreg-&gt;<a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.back()-&gt;STOP-qreg-&gt;<a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.front()-&gt;START+1) : qreg-&gt;<a class="code" href="classGenomicRegion.html#a953613d099a41f0a9e23dcff0bafa85b" title="Returns region size.">GetSize</a>();
      <span class="keywordtype">double</span> density = (double)coverage[k]/qreg_size;
      <span class="keywordflow">if</span> (density&gt;=MIN_DENSITY) printf(<span class="stringliteral">&quot;%s\t%.4e\n&quot;</span>, qreg-&gt;<a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>, density);
      PRG.Check();
    }
    PRG.Done();
    <span class="keyword">delete</span> coverage;
    <span class="keyword">delete</span> overlaps;
</pre></div> </div><hr/><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a8b9f403a94c51065e1c6af7f6180b3ba"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::UnsortedGenomicRegionSetOverlaps" ref="a8b9f403a94c51065e1c6af7f6180b3ba" args="(GenomicRegionSet *QuerySet, GenomicRegionSet *IndexSet, const char *bin_bits=NULL)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">UnsortedGenomicRegionSetOverlaps::UnsortedGenomicRegionSetOverlaps </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&#160;</td>
          <td class="paramname"><em>QuerySet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&#160;</td>
          <td class="paramname"><em>IndexSet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>bin_bits</em> = <code>NULL</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Class constructor. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">QuerySet</td><td>pointer to query region set </td></tr>
    <tr><td class="paramname">IndexSet</td><td>pointer to index region set </td></tr>
    <tr><td class="paramname">bin_bits</td><td>number of shift-bits per bin level, e.g. "10,15,18" </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a></li>
<li>genomic_intervals.cpp</li>
</ul>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Mon Jan 7 2013 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
