<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>SortedGenomicRegionSetOverlaps Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="#pri-attribs">Private Attributes</a>  </div>
  <div class="headertitle">
<div class="title">SortedGenomicRegionSetOverlaps Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="SortedGenomicRegionSetOverlaps" --><!-- doxytag: inherits="GenomicRegionSetOverlaps" -->
<p>This class is used to find overlaps between two genomic regions sets. See detailed description below for an example.  
 <a href="classSortedGenomicRegionSetOverlaps.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for SortedGenomicRegionSetOverlaps:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classSortedGenomicRegionSetOverlaps.png" usemap="#SortedGenomicRegionSetOverlaps_map" alt=""/>
  <map id="SortedGenomicRegionSetOverlaps_map" name="SortedGenomicRegionSetOverlaps_map">
<area href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions." alt="GenomicRegionSetOverlaps" shape="rect" coords="0,0,207,24"/>
</map>
 </div></div>

<p><a href="classSortedGenomicRegionSetOverlaps-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a0213f6ea908db759f63cea80db4909aa">SortedGenomicRegionSetOverlaps</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#ae513304379055f6d379bc5907733dbe2">QuerySet</a>, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#ac587bf854c827381493735b473622e03">IndexSet</a>, bool <a class="el" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3">sorted_by_strand</a>)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor.  <a href="#a0213f6ea908db759f63cea80db4909aa"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1ac0a70899fce3684da0b7e915a13d1e"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::~SortedGenomicRegionSetOverlaps" ref="a1ac0a70899fce3684da0b7e915a13d1e" args="()" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a1ac0a70899fce3684da0b7e915a13d1e">~SortedGenomicRegionSetOverlaps</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class destructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3b187db0a12f876a9347b4a1641670a0"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::GetQuery" ref="a3b187db0a12f876a9347b4a1641670a0" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a3b187db0a12f876a9347b4a1641670a0">GetQuery</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the current query region ('NULL' if no more query regions are available); also, <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> is automatically updated via <a class="el" href="classSortedGenomicRegionSetOverlaps.html#ad894d118c61c3cd11be8b37b8f185120">SortedGenomicRegionSetOverlaps::LoadIndexBuffer</a>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a127edf3e9d8fb2ebf189af00e6893e96"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::NextQuery" ref="a127edf3e9d8fb2ebf189af00e6893e96" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a127edf3e9d8fb2ebf189af00e6893e96">NextQuery</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the next query region ('NULL' if no more query regions are available); also, <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> is automatically updated via <a class="el" href="classSortedGenomicRegionSetOverlaps.html#ad894d118c61c3cd11be8b37b8f185120">SortedGenomicRegionSetOverlaps::LoadIndexBuffer</a>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad0f7b48e1eb685f0aa62aa799088dbc7"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::GetMatch" ref="ad0f7b48e1eb685f0aa62aa799088dbc7" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#ad0f7b48e1eb685f0aa62aa799088dbc7">GetMatch</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a756bad7ba9a862ae27cd9ad52acbeef2"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::NextMatch" ref="a756bad7ba9a862ae27cd9ad52acbeef2" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a756bad7ba9a862ae27cd9ad52acbeef2">NextMatch</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a726105d1ee63db6271a90e8092e243c9"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::Done" ref="a726105d1ee63db6271a90e8092e243c9" args="()" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a726105d1ee63db6271a90e8092e243c9">Done</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns true, if no more query or index regions are available and the <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> buffer is empty. <br/></td></tr>
<tr><td colspan="2"><h2><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad894d118c61c3cd11be8b37b8f185120"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::LoadIndexBuffer" ref="ad894d118c61c3cd11be8b37b8f185120" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#ad894d118c61c3cd11be8b37b8f185120">LoadIndexBuffer</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">manages <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac5fe4552253a9d9d802d76b0f4356947"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::ClearIndexBuffer" ref="ac5fe4552253a9d9d802d76b0f4356947" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#ac5fe4552253a9d9d802d76b0f4356947">ClearIndexBuffer</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">clears <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br/></td></tr>
<tr><td colspan="2"><h2><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af08c449faa6ba7be7c6ed9c07f4226e3"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::sorted_by_strand" ref="af08c449faa6ba7be7c6ed9c07f4226e3" args="" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#af08c449faa6ba7be7c6ed9c07f4226e3">sorted_by_strand</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">if true, query and index regions sorted by strand <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a19fa18e6abd9f045786698fff48a445f"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::IRegBuffer" ref="a19fa18e6abd9f045786698fff48a445f" args="" -->
GenomicRegionList&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">IRegBuffer</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">temporary buffer containing index regions that overlap with current query; the buffer is cleared only if the next query has no overlap <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a71d6eea0c4bf9f19be6cf0c0c562f297"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::IRegBufferIterator" ref="a71d6eea0c4bf9f19be6cf0c0c562f297" args="" -->
GenomicRegionList::iterator&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a71d6eea0c4bf9f19be6cf0c0c562f297">IRegBufferIterator</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">iterator on <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a962f8e6ce8c15703cfefa574a07ce42c"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::ireg_buffer_interval" ref="a962f8e6ce8c15703cfefa574a07ce42c" args="" -->
<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a962f8e6ce8c15703cfefa574a07ce42c">ireg_buffer_interval</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">keeps tracks of the (linked) interval stored in <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a349d589ff006f742d635b16987fe7b6b"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::max_ireg_buffer_size" ref="a349d589ff006f742d635b16987fe7b6b" args="" -->
unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classSortedGenomicRegionSetOverlaps.html#a349d589ff006f742d635b16987fe7b6b">max_ireg_buffer_size</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">keeps track of the maximum size of <a class="el" href="classSortedGenomicRegionSetOverlaps.html#a19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> <br/></td></tr>
</table>
<hr/><a name="details" id="details"></a><h2>Detailed Description</h2>
<div class="textblock"><p>This class is used to find overlaps between two genomic regions sets. See detailed description below for an example. </p>
<p>A simple example for computing RNAseq read densities in known exons (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'density' operation): </p>
<div class="fragment"><pre class="fragment"><span class="preprocessor">    #include &quot;core.h&quot;</span>
<span class="preprocessor">    #include &quot;genomic_intervals.h&quot;</span>
 
    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *REF_REG_FILE = argv[next_arg];
    <span class="keywordtype">char</span> *TEST_REG_FILE = next_arg+1==argc ? NULL : argv[next_arg+1];
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> RefRegSet(REF_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">true</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> TestRegSet(TEST_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">false</span>);

    <span class="comment">// process overlaps</span>
    <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a> *overlaps = <span class="keyword">new</span> <a class="code" href="classSortedGenomicRegionSetOverlaps.html#a0213f6ea908db759f63cea80db4909aa" title="Class constructor.">SortedGenomicRegionSetOverlaps</a>(&amp;TestRegSet,&amp;RefRegSet,SORTED_BY_STRAND);
    <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *coverage = overlaps-&gt;<a class="code" href="classGenomicRegionSetOverlaps.html#a5c31081b154c624ba35487d988082ffa" title="Calculates the total overlap for each index region. The index region set must be loaded in memory...">CalcIndexCoverage</a>(MATCH_GAPS,IGNORE_STRAND,USE_VALUES); 
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">&quot;Printing densities...&quot;</span>,RefRegSet.n_regions);
    <span class="keywordflow">for</span> (<span class="keywordtype">long</span> <span class="keywordtype">int</span> k=0; k&lt;RefRegSet.n_regions; k++) {
      <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *qreg = RefRegSet.R[k];
      <span class="keywordtype">long</span> <span class="keywordtype">int</span> qreg_size = MATCH_GAPS ? (qreg-&gt;<a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.back()-&gt;STOP-qreg-&gt;<a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.front()-&gt;START+1) : qreg-&gt;<a class="code" href="classGenomicRegion.html#a5bff7a3e83b0bc9bfb5441270d0ba1ae" title="Returns sum of interval&#39;s sizes.">GetSize</a>();
      <span class="keywordtype">double</span> density = (double)coverage[k]/qreg_size;
      <span class="keywordflow">if</span> (density&gt;=MIN_DENSITY) printf(<span class="stringliteral">&quot;%s\t%.4e\n&quot;</span>, qreg-&gt;<a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>, density);
      PRG.Check();
    }
    PRG.Done();
    <span class="keyword">delete</span> coverage;
    <span class="keyword">delete</span> overlaps;
</pre></div><p> A more complex example for creating ChIPseq read profiles in TSS regions (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'offset' operation): </p>
<div class="fragment"><pre class="fragment"><span class="preprocessor">    #include &quot;core.h&quot;</span>
<span class="preprocessor">    #include &quot;genomic_intervals.h&quot;</span>
 
    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *QUERY_REG_FILE = <span class="stringliteral">&quot;chipseq.reads.reg&quot;</span>;
    <span class="keywordtype">char</span> *INDEX_REG_FILE = <span class="stringliteral">&quot;TSS.flank10kb.reg&quot;</span>;
    <span class="keywordtype">char</span> *OFFSET_OP = <span class="stringliteral">&quot;5p&quot;</span>;
    <span class="keywordtype">bool</span> SORTED_BY_STRAND = <span class="keyword">false</span>; 
    <span class="keywordtype">bool</span> MATCH_GAPS = <span class="keyword">true</span>;
    <span class="keywordtype">bool</span> IGNORE_STRAND = <span class="keyword">true</span>;
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> QueryRegSet(QUERY_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or from standard input...">GenomicRegionSet</a> IndexRegSet(INDEX_REG_FILE,10000,<span class="keyword">true</span>,<span class="keyword">false</span>);

    <span class="comment">// process overlaps</span>
    <a class="code" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed description below ...">SortedGenomicRegionSetOverlaps</a> Overlaps(&amp;QueryRegSet,&amp;IndexRegSet,SORTED_BY_STRAND);
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">&quot;Processing queries...&quot;</span>,1);
    <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *qreg=Overlaps.GetQuery(); Overlaps.Done()==<span class="keyword">false</span>; qreg=Overlaps.NextQuery()) {
      <span class="keywordtype">size_t</span> Qsize = qreg-&gt;<a class="code" href="classGenomicRegion.html#a5bff7a3e83b0bc9bfb5441270d0ba1ae" title="Returns sum of interval&#39;s sizes.">GetSize</a>();
      <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals).">GenomicRegion</a> *ireg=Overlaps.GetOverlap(MATCH_GAPS,IGNORE_STRAND); ireg!=NULL; ireg=Overlaps.NextOverlap(MATCH_GAPS,IGNORE_STRAND)) {
        cout &lt;&lt; qreg-&gt;<a class="code" href="classGenomicRegion.html#a7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a> &lt;&lt; <span class="charliteral">&#39;\t&#39;</span>;
        <span class="keywordtype">long</span> <span class="keywordtype">int</span> start_offset, stop_offset;
        ireg-&gt;I.front()-&gt;GetOffsetFrom(qreg-&gt;<a class="code" href="classGenomicRegion.html#ab10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.front(),OFFSET_OP,IGNORE_STRAND,&amp;start_offset,&amp;stop_offset);
        printf(<span class="stringliteral">&quot;%f&quot;</span>, ((<span class="keywordtype">float</span>)start_offset/Qsize+(<span class="keywordtype">float</span>)stop_offset/Qsize)/2); 
        cout &lt;&lt; <span class="charliteral">&#39;\n&#39;</span>;
      }
      PRG.Check();
    }
    PRG.Done();
</pre></div> </div><hr/><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a0213f6ea908db759f63cea80db4909aa"></a><!-- doxytag: member="SortedGenomicRegionSetOverlaps::SortedGenomicRegionSetOverlaps" ref="a0213f6ea908db759f63cea80db4909aa" args="(GenomicRegionSet *QuerySet, GenomicRegionSet *IndexSet, bool sorted_by_strand)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">SortedGenomicRegionSetOverlaps::SortedGenomicRegionSetOverlaps </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&#160;</td>
          <td class="paramname"><em>QuerySet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&#160;</td>
          <td class="paramname"><em>IndexSet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>sorted_by_strand</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Class constructor. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">QuerySet</td><td>pointer to query region set </td></tr>
    <tr><td class="paramname">IndexSet</td><td>pointer to index region set </td></tr>
    <tr><td class="paramname">sorted_by_strand</td><td>if true, query and index regions are sorted by strand </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a></li>
<li>genomic_intervals.cpp</li>
</ul>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Thu Apr 5 2012 09:14:41 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
