<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>UnsortedGenomicRegionSetOverlaps Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>UnsortedGenomicRegionSetOverlaps Class Reference</h1><!-- doxytag: class="UnsortedGenomicRegionSetOverlaps" --><!-- doxytag: inherits="GenomicRegionSetOverlaps" -->Class for computing overlaps between unsorted regions.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>
<div class="dynheader">
Inheritance diagram for UnsortedGenomicRegionSetOverlaps:</div>
<div class="dynsection">

<p><center><img src="classUnsortedGenomicRegionSetOverlaps.png" usemap="#UnsortedGenomicRegionSetOverlaps_map" border="0" alt=""></center>
<map name="UnsortedGenomicRegionSetOverlaps_map">
<area href="classGenomicRegionSetOverlaps.html" alt="GenomicRegionSetOverlaps" shape="rect" coords="0,0,220,24">
</map>
</div>

<p>
<a href="classUnsortedGenomicRegionSetOverlaps-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Types</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0f0857dd93ad3c047d1f5056a57edc14"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::BinSet" ref="0f0857dd93ad3c047d1f5056a57edc14" args="" -->
typedef pair&lt; long int *, long <br>
int ** &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#0f0857dd93ad3c047d1f5056a57edc14">BinSet</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">a two-dimensional grid of bins <br></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#aa858364bf0a3d7986c9d0838a73ccab">UnsortedGenomicRegionSetOverlaps</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#e513304379055f6d379bc5907733dbe2">QuerySet</a>, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *<a class="el" href="classGenomicRegionSetOverlaps.html#c587bf854c827381493735b473622e03">IndexSet</a>, char *bin_bits=NULL)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor.  <a href="#aa858364bf0a3d7986c9d0838a73ccab"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7682171bb52c651af44e4e8f3bf33f52"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::~UnsortedGenomicRegionSetOverlaps" ref="7682171bb52c651af44e4e8f3bf33f52" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#7682171bb52c651af44e4e8f3bf33f52">~UnsortedGenomicRegionSetOverlaps</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bf7e2f8d35f7a9cf8441caa9d765de44"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::GetQuery" ref="bf7e2f8d35f7a9cf8441caa9d765de44" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#bf7e2f8d35f7a9cf8441caa9d765de44">GetQuery</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the current query region ('NULL' if no more query regions are available). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="96e68ef99a958665f5da74572edcc854"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::NextQuery" ref="96e68ef99a958665f5da74572edcc854" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#96e68ef99a958665f5da74572edcc854">NextQuery</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the next query region ('NULL' if no more query regions are available). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="61c2c6e9203acb0e0a9bfa4f0176308f"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::GetMatch" ref="61c2c6e9203acb0e0a9bfa4f0176308f" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#61c2c6e9203acb0e0a9bfa4f0176308f">GetMatch</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the current index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0bcd12697611caa5216db7a4c4fca4e2"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::NextMatch" ref="0bcd12697611caa5216db7a4c4fca4e2" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#0bcd12697611caa5216db7a4c4fca4e2">NextMatch</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns a pointer to the next index region that overlaps the current query region (even if the overlap is only in the gaps between intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fc18fc203debad6a2d48200ea8ccb42f"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::Done" ref="fc18fc203debad6a2d48200ea8ccb42f" args="()" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#fc18fc203debad6a2d48200ea8ccb42f">Done</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns true, if no more query or index regions are available and the <a class="el" href="classSortedGenomicRegionSetOverlaps.html#19fa18e6abd9f045786698fff48a445f">SortedGenomicRegionSetOverlaps::IRegBuffer</a> buffer is empty. <br></td></tr>
<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="251e09b7f0a13a56b902c30791228ce4"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::r_next" ref="251e09b7f0a13a56b902c30791228ce4" args="" -->
long int *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#251e09b7f0a13a56b902c30791228ce4">r_next</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">for each region store, store a pointer to the "next" one in the same bin <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1cc721de471ed40dcc7d8eae258e5833"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::n_levels" ref="1cc721de471ed40dcc7d8eae258e5833" args="" -->
int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#1cc721de471ed40dcc7d8eae258e5833">n_levels</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">number of levels in the two-dimensional grid of bins <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="76ecebfcf839a940bdae26069ceef748"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::n_bits" ref="76ecebfcf839a940bdae26069ceef748" args="" -->
int *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#76ecebfcf839a940bdae26069ceef748">n_bits</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">number of bits (per level) by which the start position is right-shifted to compute its bin <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="beeb94562abdf4aebc36684959a5bc78"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::index" ref="beeb94562abdf4aebc36684959a5bc78" args="" -->
map&lt; string, <a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#0f0857dd93ad3c047d1f5056a57edc14">BinSet</a> * &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#beeb94562abdf4aebc36684959a5bc78">index</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">the two-dimensional grid of bins per chromosome <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6bbb87f3c66d04527f06a5761df1ab56"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::current_binset" ref="6bbb87f3c66d04527f06a5761df1ab56" args="" -->
<a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#0f0857dd93ad3c047d1f5056a57edc14">BinSet</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#6bbb87f3c66d04527f06a5761df1ab56">current_binset</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">pointer to the current chromosome set of bins <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bafc544ce2c677ce071fa4e30ec2ecaf"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::new_query" ref="bafc544ce2c677ce071fa4e30ec2ecaf" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classUnsortedGenomicRegionSetOverlaps.html#bafc544ce2c677ce071fa4e30ec2ecaf">new_query</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">'true', if a new query is about to be processed <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
Class for computing overlaps between unsorted regions. 
<p>
A simple example for computing RNAseq read densities in known exons (this is actually implemented in the <b>genomic_overlaps</b> command-line tool as 'density' operation): <div class="fragment"><pre class="fragment"><span class="preprocessor">    #include "core.h"</span>
<span class="preprocessor">    #include "genomic_intervals.h"</span>
    
    <span class="comment">// open region sets</span>
    <span class="keywordtype">char</span> *REF_REG_FILE = argv[next_arg];
    <span class="keywordtype">char</span> *TEST_REG_FILE = next_arg+1==argc ? NULL : argv[next_arg+1];
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> RefRegSet(REF_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">true</span>);
    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> TestRegSet(TEST_REG_FILE,BUFFER_SIZE,VERBOSE,<span class="keyword">false</span>);

    <span class="comment">// process overlaps</span>
    <a class="code" href="classGenomicRegionSetOverlaps.html" title="Abstract class for computing overlaps between genomic regions.">GenomicRegionSetOverlaps</a> *overlaps = <span class="keyword">new</span> <a class="code" href="classUnsortedGenomicRegionSetOverlaps.html#aa858364bf0a3d7986c9d0838a73ccab" title="Class constructor.">UnsortedGenomicRegionSetOverlaps</a>(&amp;TestRegSet,&amp;RefRegSet);
    <span class="keywordtype">unsigned</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> *coverage = overlaps-&gt;<a class="code" href="classGenomicRegionSetOverlaps.html#5c31081b154c624ba35487d988082ffa" title="Calculates the total overlap for each index region. The index region set must be...">CalcIndexCoverage</a>(MATCH_GAPS,IGNORE_STRAND,USE_VALUES); 
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">"Printing densities..."</span>,RefRegSet.n_regions);
    <span class="keywordflow">for</span> (<span class="keywordtype">long</span> <span class="keywordtype">int</span> k=0; k&lt;RefRegSet.n_regions; k++) {
      <a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *qreg = RefRegSet.R[k];
      <span class="keywordtype">long</span> <span class="keywordtype">int</span> qreg_size = MATCH_GAPS ? (qreg-&gt;<a class="code" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.back()-&gt;STOP-qreg-&gt;<a class="code" href="classGenomicRegion.html#b10b86b03c258958818b00c21e3672a9" title="set of intervals">I</a>.front()-&gt;START+1) : qreg-&gt;<a class="code" href="classGenomicRegion.html#5bff7a3e83b0bc9bfb5441270d0ba1ae" title="Returns sum of interval&amp;#39;s sizes.">GetSize</a>();
      <span class="keywordtype">double</span> density = (double)coverage[k]/qreg_size;
      <span class="keywordflow">if</span> (density&gt;=MIN_DENSITY) printf(<span class="stringliteral">"%s\t%.4e\n"</span>, qreg-&gt;<a class="code" href="classGenomicRegion.html#7eeba95c1e87e100346688681e30ff24" title="genomic region label">LABEL</a>, density);
      PRG.Check();
    }
    PRG.Done();
    <span class="keyword">delete</span> coverage;
    <span class="keyword">delete</span> overlaps;
</pre></div> <hr><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" name="aa858364bf0a3d7986c9d0838a73ccab"></a><!-- doxytag: member="UnsortedGenomicRegionSetOverlaps::UnsortedGenomicRegionSetOverlaps" ref="aa858364bf0a3d7986c9d0838a73ccab" args="(GenomicRegionSet *QuerySet, GenomicRegionSet *IndexSet, char *bin_bits=NULL)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">UnsortedGenomicRegionSetOverlaps::UnsortedGenomicRegionSetOverlaps           </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td>
          <td class="paramname"> <em>QuerySet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td>
          <td class="paramname"> <em>IndexSet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>bin_bits</em> = <code>NULL</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Class constructor. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>QuerySet</em>&nbsp;</td><td>pointer to query region set </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>IndexSet</em>&nbsp;</td><td>pointer to index region set </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>bin_bits</em>&nbsp;</td><td>number of shift-bits per bin level, e.g. "10,15,18" </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Fri Sep 23 12:35:25 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
