<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>GenomicRegionSet Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>GenomicRegionSet Class Reference</h1><!-- doxytag: class="GenomicRegionSet" -->This class is used to read and manipulate a set of genomic regions from a file or from standard input.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>

<p>
<a href="classGenomicRegionSet-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="afc417b2a9a27f3ad833828be5720e92"></a><!-- doxytag: member="GenomicRegionSet::GenomicRegionSet" ref="afc417b2a9a27f3ad833828be5720e92" args="(char *file, unsigned long int buffer_size, bool verbose, bool load_in_memory=true)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#afc417b2a9a27f3ad833828be5720e92">GenomicRegionSet</a> (char *<a class="el" href="classGenomicRegionSet.html#7c593d1beab775590d5b810097644583">file</a>, unsigned long int <a class="el" href="classGenomicRegionSet.html#4d60b5403d2d6aaeda541ed45625b245">buffer_size</a>, bool <a class="el" href="classGenomicRegionSet.html#2646e9e6d5da1693b4bea36a5acc082e">verbose</a>, bool <a class="el" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687">load_in_memory</a>=true)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor, see example below. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8d18f09bd536bae9f2a2494d4cc19f65"></a><!-- doxytag: member="GenomicRegionSet::GenomicRegionSet" ref="8d18f09bd536bae9f2a2494d4cc19f65" args="(FILE *file_ptr, unsigned long int buffer_size, bool verbose, bool load_in_memory=true)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#8d18f09bd536bae9f2a2494d4cc19f65">GenomicRegionSet</a> (FILE *<a class="el" href="classGenomicRegionSet.html#4ca6c6268e39fdea2add6e6381c66962">file_ptr</a>, unsigned long int <a class="el" href="classGenomicRegionSet.html#4d60b5403d2d6aaeda541ed45625b245">buffer_size</a>, bool <a class="el" href="classGenomicRegionSet.html#2646e9e6d5da1693b4bea36a5acc082e">verbose</a>, bool <a class="el" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687">load_in_memory</a>=true)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor, see example below. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3fc7c6dc31cffcedf97a7b69a7f3c2c8"></a><!-- doxytag: member="GenomicRegionSet::~GenomicRegionSet" ref="3fc7c6dc31cffcedf97a7b69a7f3c2c8" args="()" -->
virtual&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#3fc7c6dc31cffcedf97a7b69a7f3c2c8">~GenomicRegionSet</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6992ccc0c4dd8be0ad4be9b991ab9f36"></a><!-- doxytag: member="GenomicRegionSet::ProcessFileHeader" ref="6992ccc0c4dd8be0ad4be9b991ab9f36" args="(bool hide)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#6992ccc0c4dd8be0ad4be9b991ab9f36">ProcessFileHeader</a> (bool hide)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Processes file header. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5721a1d9ca4ca18ddbcad1e563d42019"></a><!-- doxytag: member="GenomicRegionSet::DetectFileFormat" ref="5721a1d9ca4ca18ddbcad1e563d42019" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#5721a1d9ca4ca18ddbcad1e563d42019">DetectFileFormat</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Detects input format. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4e8e101aebcedd6a5892d463fe830450"></a><!-- doxytag: member="GenomicRegionSet::CreateGenomicRegion" ref="4e8e101aebcedd6a5892d463fe830450" args="(FileBuffer *buffer)" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#4e8e101aebcedd6a5892d463fe830450">CreateGenomicRegion</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *<a class="el" href="classGenomicRegionSet.html#df3a5598e31ddbb5b381125925362085">buffer</a>)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Creates new genomic region based on input format. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="94b1955e7a4127efaceba25429af7568"></a><!-- doxytag: member="GenomicRegionSet::Init" ref="94b1955e7a4127efaceba25429af7568" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#94b1955e7a4127efaceba25429af7568">Init</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Initializes data structures. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ae7295913424148013df6a154c5458aa"></a><!-- doxytag: member="GenomicRegionSet::Reset" ref="ae7295913424148013df6a154c5458aa" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#ae7295913424148013df6a154c5458aa">Reset</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Resets file; note, that this is not possible if the regions are being read from the standard input. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="06e2e8f91623c01e6e00500888b9ba3d"></a><!-- doxytag: member="GenomicRegionSet::Begin" ref="06e2e8f91623c01e6e00500888b9ba3d" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#06e2e8f91623c01e6e00500888b9ba3d">Begin</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns pointer to the first region in the set; note, that this is not possible if the regions are being read from the standard input. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="77031066d2648c8935c291ef227b37be"></a><!-- doxytag: member="GenomicRegionSet::Get" ref="77031066d2648c8935c291ef227b37be" args="()" -->
virtual <a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#77031066d2648c8935c291ef227b37be">Get</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns pointer to current region in the set. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4f7501a27d13cae65247686182b78d51"></a><!-- doxytag: member="GenomicRegionSet::Next" ref="4f7501a27d13cae65247686182b78d51" args="()" -->
virtual <a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#4f7501a27d13cae65247686182b78d51">Next</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Moves pointer to next region in the set; if <a class="el" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687">GenomicRegionSet::load_in_memory</a> is 'false', it deletes the current region. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#c11134b4ccc2ad002ef20ac12d3b70b8">Next</a> (bool sorted_by_strand, bool retain_current=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionSet.html#4f7501a27d13cae65247686182b78d51">Next()</a> but allows user to check sort order and/or retain the current region into memory.  <a href="#c11134b4ccc2ad002ef20ac12d3b70b8"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8993c741ef2513ba5c5b88eb390f1f99"></a><!-- doxytag: member="GenomicRegionSet::Release" ref="8993c741ef2513ba5c5b88eb390f1f99" args="(GenomicRegion *r)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#8993c741ef2513ba5c5b88eb390f1f99">Release</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Deletes genomic region from memory. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d7dfea15eb8968d376771d74e042f3a1"></a><!-- doxytag: member="GenomicRegionSet::GetRetain" ref="d7dfea15eb8968d376771d74e042f3a1" args="()" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#d7dfea15eb8968d376771d74e042f3a1">GetRetain</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">[DEPRECATED] Returns current region, and protects its from deletion by <a class="el" href="classGenomicRegionSet.html#4f7501a27d13cae65247686182b78d51">GenomicRegionSet::Next()</a>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="39c746333affb32acffbf0f31b7f6fcf"></a><!-- doxytag: member="GenomicRegionSet::PrintError" ref="39c746333affb32acffbf0f31b7f6fcf" args="(string error_msg)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#39c746333affb32acffbf0f31b7f6fcf">PrintError</a> (string error_msg)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints error message. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="260acb32612d31325099c9e53366c9d2"></a><!-- doxytag: member="GenomicRegionSet::CountRegions" ref="260acb32612d31325099c9e53366c9d2" args="(bool use_label_counts)" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#260acb32612d31325099c9e53366c9d2">CountRegions</a> (bool use_label_counts)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Counts the number of regions in the set. If <b>use_label_counts</b>, the region labels are assumed to be numbers, and the sum of these numbers is reported. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="78c6f4879ac131e6db2094f21b466a98"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalAnnotate" ref="78c6f4879ac131e6db2094f21b466a98" args="(StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#78c6f4879ac131e6db2094f21b466a98">RunGlobalAnnotate</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Annotate input region neighborhoods as upstream, downstream, etc. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="619b51265f7bf5d6c92144de689ca8a2"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalCluster" ref="619b51265f7bf5d6c92144de689ca8a2" args="(bool merge=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#619b51265f7bf5d6c92144de689ca8a2">RunGlobalCluster</a> (bool merge=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Cluster input regions based on pair-wise overlaps (note that this is computationally intensive). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3d550572975732566378797ac1d8119b"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalCalcDistances" ref="3d550572975732566378797ac1d8119b" args="(char *op1, char *op2)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#3d550572975732566378797ac1d8119b">RunGlobalCalcDistances</a> (char *op1, char *op2)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print distance between successive regions. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2d6dc99608a1d4938b421fa5a85f5477"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalSort" ref="2d6dc99608a1d4938b421fa5a85f5477" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#2d6dc99608a1d4938b421fa5a85f5477">RunGlobalSort</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints sorted region set. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1056659ccda2e39febb6b87d858a88c4"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalInvert" ref="1056659ccda2e39febb6b87d858a88c4" args="(StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#1056659ccda2e39febb6b87d858a88c4">RunGlobalInvert</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints the difference between chromosome intervals and input region set (input region set must be sorted). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b9783fb4fead289c7fe107f9bec1865b"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalLink" ref="b9783fb4fead289c7fe107f9bec1865b" args="(bool sorted_by_strand, long int max_difference)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#b9783fb4fead289c7fe107f9bec1865b">RunGlobalLink</a> (bool sorted_by_strand, long int max_difference)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Links successive regions if they overlap (input region set must be sorted). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="40dd79548f9d0333f1169bd17a6e9465"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalPartition" ref="40dd79548f9d0333f1169bd17a6e9465" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#40dd79548f9d0333f1169bd17a6e9465">RunGlobalPartition</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Partitions overlapping regions into a non-overlapping region set (input region set must be sorted). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="83853c6af4b779a37de284751d09348d"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalTest" ref="83853c6af4b779a37de284751d09348d" args="(bool sorted_by_strand)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#83853c6af4b779a37de284751d09348d">RunGlobalTest</a> (bool sorted_by_strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Tests if input region set is sorted. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="53885e10c5dc9aa51ebece77dae63313"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalMerge" ref="53885e10c5dc9aa51ebece77dae63313" args="(long int merge_win)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#53885e10c5dc9aa51ebece77dae63313">RunGlobalMerge</a> (long int merge_win)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Merges successive regions if they are separated by a distance of <b>merge_win</b> or less (input region set must be sorted). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="962bb1b26543df14b324d7ab4dd8e404"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalReverseOrder" ref="962bb1b26543df14b324d7ab4dd8e404" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#962bb1b26543df14b324d7ab4dd8e404">RunGlobalReverseOrder</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reverses the order of reverse-strand regions (for this operation, the entire region set must be loaded in memory). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b8abdbc3c1f4530dc1992fb0766ba5a4"></a><!-- doxytag: member="GenomicRegionSet::PrintDensities" ref="b8abdbc3c1f4530dc1992fb0766ba5a4" args="(GenomicRegionSet &amp;Index, bool report_exact_boundaries)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#b8abdbc3c1f4530dc1992fb0766ba5a4">PrintDensities</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> &amp;Index, bool report_exact_boundaries)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Computes densities of input regions in index regions in <b>Index</b>. This is OBSOLETE, use class <a class="el" href="classSortedGenomicRegionSetOverlaps.html" title="This class is used to find overlaps between two genomic regions sets. See detailed...">SortedGenomicRegionSetOverlaps</a> instead. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="14610521cc47631b1459c2e3488b0eec"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalScan" ref="14610521cc47631b1459c2e3488b0eec" args="(StringLIntMap *bounds, long int win_step, long int win_size)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#14610521cc47631b1459c2e3488b0eec">RunGlobalScan</a> (StringLIntMap *bounds, long int win_step, long int win_size)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Scans input regions in sliding windows and reports count distribution (input regions must be sorted). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#45cac26bedec4d983127b93d2096f735">RunGlobalScanCount</a> (StringLIntMap *bounds, char *ref_reg_file, long int win_step, long int win_size, bool ignore_reverse_strand, char preprocess, bool use_labels_as_values, long int min_reads)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Scans input regions in sliding windows (input regions must be sorted). See also class <a class="el" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description...">GenomicRegionSetScanner</a>.  <a href="#45cac26bedec4d983127b93d2096f735"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6377c10b1e9903ff5ae3b0ff9fe20e4d"></a><!-- doxytag: member="GenomicRegionSet::RunConvertToWIG" ref="6377c10b1e9903ff5ae3b0ff9fe20e4d" args="(char *title, char *color, char *position, char *options, long int span, bool convert_chromosome)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#6377c10b1e9903ff5ae3b0ff9fe20e4d">RunConvertToWIG</a> (char *title, char *color, char *position, char *options, long int span, bool convert_chromosome)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints in Wiggle format. If <b>convert_chromosome</b> is 'true', then convert from ENSEMBL to UCSC names (by adding 'chr' prefix to each chromosome name, and by converting 'MT' to 'chrM'). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="da3da7ef918c09867341cda32881d63e"></a><!-- doxytag: member="GenomicRegionSet::RunAlign" ref="da3da7ef918c09867341cda32881d63e" args="(Chromosomes *C)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#da3da7ef918c09867341cda32881d63e">RunAlign</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints alignments of input sequences to reference genome. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="18669fb4dfdf1287aaf9cfceb8329cb5"></a><!-- doxytag: member="GenomicRegionSet::RunConvertToBED" ref="18669fb4dfdf1287aaf9cfceb8329cb5" args="(char *title, char *color, char *position, bool convert_chromosome)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#18669fb4dfdf1287aaf9cfceb8329cb5">RunConvertToBED</a> (char *title, char *color, char *position, bool convert_chromosome)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Converts to BED format. If <b>convert_chromosome</b> is 'true', then convert from ENSEMBL to UCSC names (by adding 'chr' prefix to each chromosome name, and by converting 'MT' to 'chrM'). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d7ad039dd702837f37c2cdfebd9b2845"></a><!-- doxytag: member="GenomicRegionSet::RunBounds" ref="d7ad039dd702837f37c2cdfebd9b2845" args="(StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#d7ad039dd702837f37c2cdfebd9b2845">RunBounds</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Corrects interval start/stop positions so as to comply with chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1aa8f41f0978e9b66c6f11c0ccb95a0b"></a><!-- doxytag: member="GenomicRegionSet::RunCenter" ref="1aa8f41f0978e9b66c6f11c0ccb95a0b" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#1aa8f41f0978e9b66c6f11c0ccb95a0b">RunCenter</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals' center. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="752b7d4d2ac3568e7bdb13edbfc00af3"></a><!-- doxytag: member="GenomicRegionSet::RunConnect" ref="752b7d4d2ac3568e7bdb13edbfc00af3" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#752b7d4d2ac3568e7bdb13edbfc00af3">RunConnect</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print connected intervals as a single interval from minimum start to maximum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="90256738bd85501460874ab8e69b894b"></a><!-- doxytag: member="GenomicRegionSet::RunDiff" ref="90256738bd85501460874ab8e69b894b" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#90256738bd85501460874ab8e69b894b">RunDiff</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print gap intervals between successive intervals (e.g. for computing intron boundaries). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="22f3fce9421a0f80c96ec0a9346fec19"></a><!-- doxytag: member="GenomicRegionSet::RunCalcDistances" ref="22f3fce9421a0f80c96ec0a9346fec19" args="(char *op1, char *op2)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#22f3fce9421a0f80c96ec0a9346fec19">RunCalcDistances</a> (char *op1, char *op2)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print distances between successive intervals of each region (i.e. local operation). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9acb3424739ed62c0efddad01536ad18"></a><!-- doxytag: member="GenomicRegionSet::RunDivide" ref="9acb3424739ed62c0efddad01536ad18" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#9acb3424739ed62c0efddad01536ad18">RunDivide</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="efa09dc62c768a06e4ca1f56fd683be5"></a><!-- doxytag: member="GenomicRegionSet::RunFix" ref="efa09dc62c768a06e4ca1f56fd683be5" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#efa09dc62c768a06e4ca1f56fd683be5">RunFix</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Removes rogue intervals (e.g. start&gt;stop). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="e598cd9c5f94555a04e43f9fe6334aec"></a><!-- doxytag: member="GenomicRegionSet::RunIntersection" ref="e598cd9c5f94555a04e43f9fe6334aec" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#e598cd9c5f94555a04e43f9fe6334aec">RunIntersection</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b1e96bd7781dac63a1745a8cf5ac5538"></a><!-- doxytag: member="GenomicRegionSet::RunSize" ref="b1e96bd7781dac63a1745a8cf5ac5538" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#b1e96bd7781dac63a1745a8cf5ac5538">RunSize</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints region label and sum of intervals' sizes. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#11f1e58c92e9b0b2e3feb5821cff1a55">RunModifyPos</a> (char *position_op, long int position_shift)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position.  <a href="#11f1e58c92e9b0b2e3feb5821cff1a55"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8f71dd5fa48a0878751cd78a3b485daf"></a><!-- doxytag: member="GenomicRegionSet::RunConvertToREG" ref="8f71dd5fa48a0878751cd78a3b485daf" args="(bool compact=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#8f71dd5fa48a0878751cd78a3b485daf">RunConvertToREG</a> (bool compact=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end for all regions in the set. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="1b2b9edad211ca961f319921d38bfbf6"></a><!-- doxytag: member="GenomicRegionSet::RunRandomize" ref="1b2b9edad211ca961f319921d38bfbf6" args="(gsl_rng *random_generator, StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#1b2b9edad211ca961f319921d38bfbf6">RunRandomize</a> (gsl_rng *random_generator, StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Randomizes interval position within chromosomal bounds. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7b522944fe4f7f489c99fb76ce7f576c"></a><!-- doxytag: member="GenomicRegionSet::RunSelect" ref="7b522944fe4f7f489c99fb76ce7f576c" args="(bool first, bool last, bool from5p, bool from3p)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#7b522944fe4f7f489c99fb76ce7f576c">RunSelect</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#d4b42becffb21de5e89129476c79c851">RunShiftPos</a> (long int start_shift, long int stop_shift, bool shift_5prime)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately.  <a href="#d4b42becffb21de5e89129476c79c851"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="53143cb3271c65fb289faece622cd8bd"></a><!-- doxytag: member="GenomicRegionSet::RunSort" ref="53143cb3271c65fb289faece622cd8bd" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#53143cb3271c65fb289faece622cd8bd">RunSort</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print sorted intervals according to start position (only for compatible intervals). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="570e56362af1b30e7f5cb149ea42e54e"></a><!-- doxytag: member="GenomicRegionSet::RunSplit" ref="570e56362af1b30e7f5cb149ea42e54e" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#570e56362af1b30e7f5cb149ea42e54e">RunSplit</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Split regions into intervals which are printed on separate lines. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f1146729843b82d7637e0adc1b07e51a"></a><!-- doxytag: member="GenomicRegionSet::RunModifyStrand" ref="f1146729843b82d7637e0adc1b07e51a" args="(char *strand_op, bool sorted)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#f1146729843b82d7637e0adc1b07e51a">RunModifyStrand</a> (char *strand_op, bool sorted)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print region with modified strands: '+'=only forward, '-'=only reverse, 'r'=reverse strands from '+' to '-' and vice versa; if <b>sorted</b> is 'true', it calls <a class="el" href="classGenomicRegionSet.html#7c0a328896b88a0c837f60e89fd85584">GenomicRegionSet::RunModifyStrandSorted</a> instead. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7c0a328896b88a0c837f60e89fd85584"></a><!-- doxytag: member="GenomicRegionSet::RunModifyStrandSorted" ref="7c0a328896b88a0c837f60e89fd85584" args="(char *strand_op)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#7c0a328896b88a0c837f60e89fd85584">RunModifyStrandSorted</a> (char *strand_op)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionSet.html#f1146729843b82d7637e0adc1b07e51a">GenomicRegionSet::RunModifyStrand</a> but returns a sorted region set (note: it only works if the original region set is sorted). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="015049fd0bd08587cbdbb49332ee352d"></a><!-- doxytag: member="GenomicRegionSet::RunShuffle" ref="015049fd0bd08587cbdbb49332ee352d" args="(gsl_rng *random_generator, char *ref_reg_file)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#015049fd0bd08587cbdbb49332ee352d">RunShuffle</a> (gsl_rng *random_generator, char *ref_reg_file)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Shuffles input regions within specified reference regions loaded from file <b>ref_reg_file</b>. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b281c8ab57c4181a06161182031c9df4"></a><!-- doxytag: member="GenomicRegionSet::RunUnion" ref="b281c8ab57c4181a06161182031c9df4" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#b281c8ab57c4181a06161182031c9df4">RunUnion</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print intervals' union. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a326d5adcc97f47ed2de35353b60b393"></a><!-- doxytag: member="GenomicRegionSet::RunSlidingWindows" ref="a326d5adcc97f47ed2de35353b60b393" args="(long int win_step, long int win_size)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#a326d5adcc97f47ed2de35353b60b393">RunSlidingWindows</a> (long int win_step, long int win_size)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b219b834c232c091982b066a2163f76a"></a><!-- doxytag: member="GenomicRegionSet::RunExtractSeq" ref="b219b834c232c091982b066a2163f76a" args="(Chromosomes *C, bool replace=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#b219b834c232c091982b066a2163f76a">RunExtractSeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool replace=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints region label, intervals and extracted sequence in SEQ format. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="de00553eb76b71f68e540739bdc776f4"></a><!-- doxytag: member="GenomicRegionSet::PrintBEDGraphFormat" ref="de00553eb76b71f68e540739bdc776f4" args="(char *title, char *color, char *position, bool convert_chromosome)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#de00553eb76b71f68e540739bdc776f4">PrintBEDGraphFormat</a> (char *title, char *color, char *position, bool convert_chromosome)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints in BEDGraph format. If <b>convert_chromosome</b> is 'true', then convert from ENSEMBL to UCSC names (by adding 'chr' prefix to each chromosome name, and by converting 'MT' to 'chrM'). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="3da972129b9b69577bbe0bb9837bbea3"></a><!-- doxytag: member="GenomicRegionSet::PrintSeqLength" ref="3da972129b9b69577bbe0bb9837bbea3" args="(Chromosomes *C)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#3da972129b9b69577bbe0bb9837bbea3">PrintSeqLength</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints region label and sequence length excluding 'N' characters. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#6dc7150acdd7a0614f8a8497da687d0e">PrintRemoveN</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Print sequence with no alignment gaps (requires SEQ input).  <a href="#6dc7150acdd7a0614f8a8497da687d0e"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#e27a5c9f17f08afae277603e33d4fb91">PrintOffsetFormat</a> (char *op, bool fraction)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates the offset distances of interval start and stop position with respect to the first interval in the set.  <a href="#e27a5c9f17f08afae277603e33d4fb91"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="c8de83151ba0cc30ca1b72e83ecea82f"></a><!-- doxytag: member="GenomicRegionSet::PrintReversePos" ref="c8de83151ba0cc30ca1b72e83ecea82f" args="(StringLIntMap *bounds)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#c8de83151ba0cc30ca1b72e83ecea82f">PrintReversePos</a> (StringLIntMap *bounds)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reverses interval start/stop positions with respect to chromosomal bounds, as if the reference point were in the end of the chromosome as opposed to the beginning. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="dfbad7f0acf42c2179bd33103e5d183c"></a><!-- doxytag: member="GenomicRegionSet::PrintSearch" ref="dfbad7f0acf42c2179bd33103e5d183c" args="(char *pattern, bool header, bool summary)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#dfbad7f0acf42c2179bd33103e5d183c">PrintSearch</a> (char *pattern, bool header, bool summary)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Search sequence for short pattern (requires SEQ input). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4c197a09ffcace2ec2736aa089eec448"></a><!-- doxytag: member="GenomicRegionSet::PrintVerifySeq" ref="4c197a09ffcace2ec2736aa089eec448" args="(Chromosomes *C, bool ignore)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#4c197a09ffcace2ec2736aa089eec448">PrintVerifySeq</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C, bool ignore)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Verifies extracted sequence against region label. <br></td></tr>
<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7c593d1beab775590d5b810097644583"></a><!-- doxytag: member="GenomicRegionSet::file" ref="7c593d1beab775590d5b810097644583" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#7c593d1beab775590d5b810097644583">file</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">file name ('NULL' if reading from standard input) <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4ca6c6268e39fdea2add6e6381c66962"></a><!-- doxytag: member="GenomicRegionSet::file_ptr" ref="4ca6c6268e39fdea2add6e6381c66962" args="" -->
FILE *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#4ca6c6268e39fdea2add6e6381c66962">file_ptr</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">file pointer ('NULL' if reading from standard input) <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4d60b5403d2d6aaeda541ed45625b245"></a><!-- doxytag: member="GenomicRegionSet::buffer_size" ref="4d60b5403d2d6aaeda541ed45625b245" args="" -->
unsigned long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#4d60b5403d2d6aaeda541ed45625b245">buffer_size</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">line buffer size (automatically adjusted during execution to ensure the entire line is read) <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2646e9e6d5da1693b4bea36a5acc082e"></a><!-- doxytag: member="GenomicRegionSet::verbose" ref="2646e9e6d5da1693b4bea36a5acc082e" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#2646e9e6d5da1693b4bea36a5acc082e">verbose</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">verbose mode <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bd0fc2542e717e5855e03d3b3892f027"></a><!-- doxytag: member="GenomicRegionSet::from_stdin" ref="bd0fc2542e717e5855e03d3b3892f027" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#bd0fc2542e717e5855e03d3b3892f027">from_stdin</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">'true' if reading from standard input <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="de43c6cd72b95da75395129b00a52687"></a><!-- doxytag: member="GenomicRegionSet::load_in_memory" ref="de43c6cd72b95da75395129b00a52687" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#de43c6cd72b95da75395129b00a52687">load_in_memory</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">'true' if the entire region set is loaded in memory <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ea8b5c9bae036da50d4e8720d2e13231"></a><!-- doxytag: member="GenomicRegionSet::use_interval_array" ref="ea8b5c9bae036da50d4e8720d2e13231" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#ea8b5c9bae036da50d4e8720d2e13231">use_interval_array</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">if 'true' use an array for storing genomic regions' intervals (for time efficiency) <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a4494ed2ae3f7594c29b18c39fed6748"></a><!-- doxytag: member="GenomicRegionSet::format" ref="a4494ed2ae3f7594c29b18c39fed6748" args="" -->
string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#a4494ed2ae3f7594c29b18c39fed6748">format</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">input format <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="08d9a35a7688ac2c193b53c9a8693c6e"></a><!-- doxytag: member="GenomicRegionSet::progress" ref="08d9a35a7688ac2c193b53c9a8693c6e" args="" -->
<a class="el" href="classProgress.html">Progress</a>&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#08d9a35a7688ac2c193b53c9a8693c6e">progress</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">keeps track of progress of computation <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="df3a5598e31ddbb5b381125925362085"></a><!-- doxytag: member="GenomicRegionSet::buffer" ref="df3a5598e31ddbb5b381125925362085" args="" -->
<a class="el" href="classFileBuffer.html">FileBuffer</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#df3a5598e31ddbb5b381125925362085">buffer</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">pointer to the buffer class used to read the region sets <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="5293292428aa06145b38be4b96ab05e4"></a><!-- doxytag: member="GenomicRegionSet::n_regions" ref="5293292428aa06145b38be4b96ab05e4" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#5293292428aa06145b38be4b96ab05e4">n_regions</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">number of regions ('1' if the regions are not loaded in memory) <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="206d91b443832800afc04f786d15971d"></a><!-- doxytag: member="GenomicRegionSet::R" ref="206d91b443832800afc04f786d15971d" args="" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> **&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#206d91b443832800afc04f786d15971d">R</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">array where the regions are stored; if the region set is not entirely loaded in memory, only <b>R[0]</b> is used <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="d89185b88d5b37161cc65bc84b421c0e"></a><!-- doxytag: member="GenomicRegionSet::r_index" ref="d89185b88d5b37161cc65bc84b421c0e" args="" -->
long int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#d89185b88d5b37161cc65bc84b421c0e">r_index</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">index pointing to the current region in <b>R</b> <br></td></tr>
<tr><td colspan="2"><br><h2>Private Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="cd823956ec9f510ee3d52b08eec1b2f6"></a><!-- doxytag: member="GenomicRegionSet::StoreInTempFile" ref="cd823956ec9f510ee3d52b08eec1b2f6" args="()" -->
<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#cd823956ec9f510ee3d52b08eec1b2f6">StoreInTempFile</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Creates a temporary file from this instance of <a class="el" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> which contains the current chromosome's positive strand regions (used in <a class="el" href="classGenomicRegionSet.html#7c0a328896b88a0c837f60e89fd85584">GenomicRegionSet::RunModifyStrandSorted</a>). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b118bb8a6b63ae47a37ef89bff25ab72"></a><!-- doxytag: member="GenomicRegionSet::PrintMergeSort" ref="b118bb8a6b63ae47a37ef89bff25ab72" args="(GenomicRegionSet *rtempSet, char strand)" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#b118bb8a6b63ae47a37ef89bff25ab72">PrintMergeSort</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *rtempSet, char strand)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Uses mergesort between <b>tempSet</b> and this instance of <a class="el" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> (used in <a class="el" href="classGenomicRegionSet.html#7c0a328896b88a0c837f60e89fd85584">GenomicRegionSet::RunModifyStrandSorted</a>). <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="e601878a7754a7e5bc8b4ed340d012f4"></a><!-- doxytag: member="GenomicRegionSet::PrintReverse" ref="e601878a7754a7e5bc8b4ed340d012f4" args="(GenomicRegionSet *rtempSet)" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSet.html#e601878a7754a7e5bc8b4ed340d012f4">PrintReverse</a> (<a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *rtempSet)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reverses the order between <b>tempSet</b> and this instance of <a class="el" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> (used in <a class="el" href="classGenomicRegionSet.html#7c0a328896b88a0c837f60e89fd85584">GenomicRegionSet::RunModifyStrandSorted</a>). <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class is used to read and manipulate a set of genomic regions from a file or from standard input. 
<p>
Example: <div class="fragment"><pre class="fragment">    <a class="code" href="classGenomicRegionSet.html" title="This class is used to read and manipulate a set of genomic regions from a file or...">GenomicRegionSet</a> RegSet(<span class="stringliteral">"test.reg"</span>,10000,<span class="keyword">true</span>);
    <a class="code" href="classProgress.html" title="This class is used to report progress of loop computations.">Progress</a> PRG(<span class="stringliteral">"Printing center of intervals..."</span>,1);
    <span class="keywordflow">for</span> (<a class="code" href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)...">GenomicRegion</a> *r=RegSet.Get(); r!=NULL; r=RegSet.Next()) {
      r-&gt;Center();
      r-&gt;Print();
      PRG.Check();
    }
    PRG.Done();
</pre></div> <hr><h2>Member Function Documentation</h2>
<a class="anchor" name="c11134b4ccc2ad002ef20ac12d3b70b8"></a><!-- doxytag: member="GenomicRegionSet::Next" ref="c11134b4ccc2ad002ef20ac12d3b70b8" args="(bool sorted_by_strand, bool retain_current=false)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classGenomicRegion.html">GenomicRegion</a> * GenomicRegionSet::Next           </td>
          <td>(</td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>sorted_by_strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>retain_current</em> = <code>false</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Same as <a class="el" href="classGenomicRegionSet.html#4f7501a27d13cae65247686182b78d51">Next()</a> but allows user to check sort order and/or retain the current region into memory. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>sorted_by_strand</em>&nbsp;</td><td>if 'true', it takes into account strand information while checking the sort order </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>retain_current</em>&nbsp;</td><td>if 'true', the current region is not deleted from memory; this must be done 'manually' using the <a class="el" href="classGenomicRegionSet.html#8993c741ef2513ba5c5b88eb390f1f99">Release()</a> method (do not use 'delete'). </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="45cac26bedec4d983127b93d2096f735"></a><!-- doxytag: member="GenomicRegionSet::RunGlobalScanCount" ref="45cac26bedec4d983127b93d2096f735" args="(StringLIntMap *bounds, char *ref_reg_file, long int win_step, long int win_size, bool ignore_reverse_strand, char preprocess, bool use_labels_as_values, long int min_reads)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSet::RunGlobalScanCount           </td>
          <td>(</td>
          <td class="paramtype">StringLIntMap *&nbsp;</td>
          <td class="paramname"> <em>bounds</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>ref_reg_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>win_step</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>win_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>ignore_reverse_strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&nbsp;</td>
          <td class="paramname"> <em>preprocess</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>use_labels_as_values</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>min_reads</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Scans input regions in sliding windows (input regions must be sorted). See also class <a class="el" href="classGenomicRegionSetScanner.html" title="This class is used to scan a set of regions by sliding windows. See detailed description...">GenomicRegionSetScanner</a>. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>bounds</em>&nbsp;</td><td>chromosome sizes </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ref_reg_file</em>&nbsp;</td><td>only windows that overlap with regions in this file will be reported </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>win_step</em>&nbsp;</td><td>sliding window step </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>win_size</em>&nbsp;</td><td>sliding window size (must be a multiple of window step) </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ignore_reverse_strand</em>&nbsp;</td><td>if true, no sliding windows on the negative strand are reported </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>preprocess</em>&nbsp;</td><td>if '1', only start position is counted; if 'p', all positions are counted; if 'c', center of interval is counted </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>use_labels_as_values</em>&nbsp;</td><td>if true, genomic region labels are assumed to be integers and are included in the counting </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>min_reads</em>&nbsp;</td><td>report windows only if value is greater that this parameter </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="11f1e58c92e9b0b2e3feb5821cff1a55"></a><!-- doxytag: member="GenomicRegionSet::RunModifyPos" ref="11f1e58c92e9b0b2e3feb5821cff1a55" args="(char *position_op, long int position_shift)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSet::RunModifyPos           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>position_op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>position_shift</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>position_op</em>&nbsp;</td><td>selects fixed position as follows: '1'=start position, 'c'=center position, '5p'=5-prime, '3p'=3-prime </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>position_shift</em>&nbsp;</td><td>determines the shift of the non-fixed position (in the 'c' case, both positions are shifted symmetrically around the center) </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="d4b42becffb21de5e89129476c79c851"></a><!-- doxytag: member="GenomicRegionSet::RunShiftPos" ref="d4b42becffb21de5e89129476c79c851" args="(long int start_shift, long int stop_shift, bool shift_5prime)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSet::RunShiftPos           </td>
          <td>(</td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>start_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&nbsp;</td>
          <td class="paramname"> <em>stop_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>shift_5prime</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>start_shift</em>&nbsp;</td><td>determines the shift of the start position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>stop_shift</em>&nbsp;</td><td>determines the shift of the stop position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>shift_5prime</em>&nbsp;</td><td>if 'true', then the new start position is the old 5-prime position shifted by <b>start_shift</b> and the new stop position is the old 3-prime position shifted by <b>stop_shift</b> </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="6dc7150acdd7a0614f8a8497da687d0e"></a><!-- doxytag: member="GenomicRegionSet::PrintRemoveN" ref="6dc7150acdd7a0614f8a8497da687d0e" args="(Chromosomes *C)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSet::PrintRemoveN           </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classChromosomes.html">Chromosomes</a> *&nbsp;</td>
          <td class="paramname"> <em>C</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Print sequence with no alignment gaps (requires SEQ input). 
<p>
Remove sub-intervals that correspond to sequences of 'N' characters 
</div>
</div><p>
<a class="anchor" name="e27a5c9f17f08afae277603e33d4fb91"></a><!-- doxytag: member="GenomicRegionSet::PrintOffsetFormat" ref="e27a5c9f17f08afae277603e33d4fb91" args="(char *op, bool fraction)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSet::PrintOffsetFormat           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>fraction</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Calculates the offset distances of interval start and stop position with respect to the first interval in the set. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>op</em>&nbsp;</td><td>determines reference point as follows: '1'=start position, '2'=stop position, '5p'=5-prime position, '3p'=3-prime position </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>fraction</em>&nbsp;</td><td>if 'true', offsets are reported as a fraction of the total region length </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Fri Sep 16 19:13:00 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
