<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>GenomicRegionSAM Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a>  </div>
  <div class="headertitle">
<div class="title">GenomicRegionSAM Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="GenomicRegionSAM" --><!-- doxytag: inherits="GenomicRegion" -->
<p>This class processes genomic regions in SAM format.  
 <a href="classGenomicRegionSAM.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for GenomicRegionSAM:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classGenomicRegionSAM.png" usemap="#GenomicRegionSAM_map" alt=""/>
  <map id="GenomicRegionSAM_map" name="GenomicRegionSAM_map">
<area href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)." alt="GenomicRegion" shape="rect" coords="0,0,128,24"/>
</map>
 </div></div>

<p><a href="classGenomicRegionSAM-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-types"></a>
Public Types</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aadf8580c5362304ab4866fa76a59df96"></a><!-- doxytag: member="GenomicRegionSAM::CIGARTokens" ref="aadf8580c5362304ab4866fa76a59df96" args="" -->
typedef vector&lt; pair&lt; long int, <br class="typebreak"/>
char &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96">CIGARTokens</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">This type is used for storing CIGAR tokens, i.e. (token_type, token_length) <br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae6b0828926a573743e7292567346c1a1"></a><!-- doxytag: member="GenomicRegionSAM::GenomicRegionSAM" ref="ae6b0828926a573743e7292567346c1a1" args="(FileBuffer *B)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ae6b0828926a573743e7292567346c1a1">GenomicRegionSAM</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a25247d97947a54583fb17d493df4a0cf"></a><!-- doxytag: member="GenomicRegionSAM::GenomicRegionSAM" ref="a25247d97947a54583fb17d493df4a0cf" args="(char *inp, long int n_line=0)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a25247d97947a54583fb17d493df4a0cf">GenomicRegionSAM</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1d11cc22134d13dbf67ffe4c7220888d"></a><!-- doxytag: member="GenomicRegionSAM::~GenomicRegionSAM" ref="a1d11cc22134d13dbf67ffe4c7220888d" args="()" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a1d11cc22134d13dbf67ffe4c7220888d">~GenomicRegionSAM</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class destructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a096d90b6743400daf1acb50bcf17086a"></a><!-- doxytag: member="GenomicRegionSAM::Read" ref="a096d90b6743400daf1acb50bcf17086a" args="(char *inp, long int n_line=0)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a096d90b6743400daf1acb50bcf17086a">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Reads SAM format from input string <b>inp</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4381c6486f767b6e4055cb729beee181"></a><!-- doxytag: member="GenomicRegionSAM::Print" ref="a4381c6486f767b6e4055cb729beee181" args="(FILE *file_ptr=stdout)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a4381c6486f767b6e4055cb729beee181">Print</a> (FILE *file_ptr=stdout)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adf8beab78f541121e7b56957afffa4ac"></a><!-- doxytag: member="GenomicRegionSAM::PrintConstrained" ref="adf8beab78f541121e7b56957afffa4ac" args="(GenomicRegion *r, bool merge_labels=false)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#adf8beab78f541121e7b56957afffa4ac">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionSAM.html#a4381c6486f767b6e4055cb729beee181" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>start</b> and <b>stop</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1f30589e831c311974aaa5be772e972a"></a><!-- doxytag: member="GenomicRegionSAM::PrintModified" ref="a1f30589e831c311974aaa5be772e972a" args="(const char *label, long int start, long int stop)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a1f30589e831c311974aaa5be772e972a">PrintModified</a> (const char *label, long int start, long int stop)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionSAM.html#a4381c6486f767b6e4055cb729beee181" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6d0949a836520073de57e0f968a6b819"></a><!-- doxytag: member="GenomicRegionSAM::CalcStrandFromFlag" ref="a6d0949a836520073de57e0f968a6b819" args="(unsigned long int flag)" -->
char&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a6d0949a836520073de57e0f968a6b819">CalcStrandFromFlag</a> (unsigned long int flag)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Extracts the strand from the FLAG field. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af583a963f7fb0eb071c1b640c660813a"></a><!-- doxytag: member="GenomicRegionSAM::UpdateFlagFromStrand" ref="af583a963f7fb0eb071c1b640c660813a" args="(char strand)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#af583a963f7fb0eb071c1b640c660813a">UpdateFlagFromStrand</a> (char strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Updates the FLAG field given a new strand. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aea563a4a484784ad231399a3b56c7056"></a><!-- doxytag: member="GenomicRegionSAM::PrintGappedSequence" ref="aea563a4a484784ad231399a3b56c7056" args="(char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#aea563a4a484784ad231399a3b56c7056">PrintGappedSequence</a> (char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints gapped sequence. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9446540150b384ca0d3dbba630c9c6a6"></a><!-- doxytag: member="GenomicRegionSAM::GetGappedSequence" ref="a9446540150b384ca0d3dbba630c9c6a6" args="(char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a9446540150b384ca0d3dbba630c9c6a6">GetGappedSequence</a> (char *seq, char *gap_token_types, bool lowercase_for_sort_clipping=true)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns gapped sequence. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a621b3dd6f4d03a5223bd26a57af14f7e"></a><!-- doxytag: member="GenomicRegionSAM::InvalidateSeqData" ref="a621b3dd6f4d03a5223bd26a57af14f7e" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a621b3dd6f4d03a5223bd26a57af14f7e">InvalidateSeqData</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets CIGAR, SEQ and QUAL fields to "*" and removes OPTIONAL field; also <a class="el" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f">GenomicRegionSAM::n_tokens</a> is set to 11. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a34b0387b9656e48d0c4046d62fe085e1"></a><!-- doxytag: member="GenomicRegionSAM::TokenizeCIGAR" ref="a34b0387b9656e48d0c4046d62fe085e1" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a34b0387b9656e48d0c4046d62fe085e1">TokenizeCIGAR</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes and stores CIGAR tokens in <a class="el" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92">GenomicRegionSAM::T</a>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a50198fead1c40caadf76f40b61a3c14a"></a><!-- doxytag: member="GenomicRegionSAM::GetNextTokenOfCIGAR" ref="a50198fead1c40caadf76f40b61a3c14a" args="(char **cigar, char *type)" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a50198fead1c40caadf76f40b61a3c14a">GetNextTokenOfCIGAR</a> (char **cigar, char *type)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns next token in CIGAR string. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a965bb3cc1217e6c859a07dd23090c2a0"></a><!-- doxytag: member="GenomicRegionSAM::UpdateCigarFromIntervals" ref="a965bb3cc1217e6c859a07dd23090c2a0" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a965bb3cc1217e6c859a07dd23090c2a0">UpdateCigarFromIntervals</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Update CIGAR string based only on interval information (chooses simplest possible CIGAR). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5de5d05dfccd33149153a299c1cb2f6b"></a><!-- doxytag: member="GenomicRegionSAM::CalcFragmentLengthFromCIGAR" ref="a5de5d05dfccd33149153a299c1cb2f6b" args="()" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a5de5d05dfccd33149153a299c1cb2f6b">CalcFragmentLengthFromCIGAR</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate interval length from CIGAR string. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aacff4cd74a3b8c450f9f90c6b23412c4"></a><!-- doxytag: member="GenomicRegionSAM::CalcReferenceLengthFromCIGAR" ref="aacff4cd74a3b8c450f9f90c6b23412c4" args="()" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#aacff4cd74a3b8c450f9f90c6b23412c4">CalcReferenceLengthFromCIGAR</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate interval length from CIGAR string. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a17e8e7054ffe7205d7c34926b60526d8"></a><!-- doxytag: member="GenomicRegionSAM::TrimCIGAR" ref="a17e8e7054ffe7205d7c34926b60526d8" args="(long int d_ref_start, long int d_ref_stop)" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a17e8e7054ffe7205d7c34926b60526d8">TrimCIGAR</a> (long int d_ref_start, long int d_ref_stop)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Trim CIGAR string from left and right by a given distance (measured on the reference sequence). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0ff4472118291a8034efc929c784680c"></a><!-- doxytag: member="GenomicRegionSAM::TrimSequence" ref="a0ff4472118291a8034efc929c784680c" args="(char *seq, long int d_ref_start, long int d_ref_stop)" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a0ff4472118291a8034efc929c784680c">TrimSequence</a> (char *seq, long int d_ref_start, long int d_ref_stop)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Trim input <b>seq</b> from left and right (used for trimming both SEQ and QUAL fields). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a97b9fec94e5f21490d839b3f99235a96"></a><!-- doxytag: member="GenomicRegionSAM::RunAlign" ref="a97b9fec94e5f21490d839b3f99235a96" args="(Chromosomes *C)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a97b9fec94e5f21490d839b3f99235a96">RunAlign</a> (<a class="el" href="classChromosomes.html">Chromosomes</a> *C)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints alignments of input sequences to reference genome. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7315b920fcc6ef84f1ecaee5f842fa82"></a><!-- doxytag: member="GenomicRegionSAM::ApplyBounds" ref="a7315b920fcc6ef84f1ecaee5f842fa82" args="(StringLIntMap *bounds)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a7315b920fcc6ef84f1ecaee5f842fa82">ApplyBounds</a> (StringLIntMap *bounds)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Corrects interval start/stop positions so as to comply with chromosomal bounds. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a52fa8c0420b16fe92f54427365f72899"></a><!-- doxytag: member="GenomicRegionSAM::Center" ref="a52fa8c0420b16fe92f54427365f72899" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a52fa8c0420b16fe92f54427365f72899">Center</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace each interval in the region with the corresponding center interval. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad765c84814151be0e710fefb37c7949a"></a><!-- doxytag: member="GenomicRegionSAM::Connect" ref="ad765c84814151be0e710fefb37c7949a" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ad765c84814151be0e710fefb37c7949a">Connect</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace region with a single interval from minimum start to maximum stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aca38b04fa916ae4c9c28c36f537136ba"></a><!-- doxytag: member="GenomicRegionSAM::Diff" ref="aca38b04fa916ae4c9c28c36f537136ba" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#aca38b04fa916ae4c9c28c36f537136ba">Diff</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afc9c7ec17be517dd5b1ad7cb62f58826"></a><!-- doxytag: member="GenomicRegionSAM::RunCalcDistances" ref="afc9c7ec17be517dd5b1ad7cb62f58826" args="(char *op1, char *op2)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#afc9c7ec17be517dd5b1ad7cb62f58826">RunCalcDistances</a> (char *op1, char *op2)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print distance between successive intervals. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7c16ba87be562950021d3f19eb14ac5e"></a><!-- doxytag: member="GenomicRegionSAM::Divide" ref="a7c16ba87be562950021d3f19eb14ac5e" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a7c16ba87be562950021d3f19eb14ac5e">Divide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="add2ada73e241f593dcdcdfca97df1eb5"></a><!-- doxytag: member="GenomicRegionSAM::RunDivide" ref="add2ada73e241f593dcdcdfca97df1eb5" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#add2ada73e241f593dcdcdfca97df1eb5">RunDivide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad955a6a512678613b4a42d2db0d3f549"></a><!-- doxytag: member="GenomicRegionSAM::Fix" ref="ad955a6a512678613b4a42d2db0d3f549" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ad955a6a512678613b4a42d2db0d3f549">Fix</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes rogue intervals (e.g. start&gt;stop) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4f26b01325ebd4e20c9bf3247d5fa733"></a><!-- doxytag: member="GenomicRegionSAM::Intersect" ref="a4f26b01325ebd4e20c9bf3247d5fa733" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a4f26b01325ebd4e20c9bf3247d5fa733">Intersect</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a538b8fc9969f24da6911ec39c8d70a3b"></a><!-- doxytag: member="GenomicRegionSAM::Randomize" ref="a538b8fc9969f24da6911ec39c8d70a3b" args="(gsl_rng *random_generator, StringLIntMap *bounds)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a538b8fc9969f24da6911ec39c8d70a3b">Randomize</a> (gsl_rng *random_generator, StringLIntMap *bounds)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Randomizes interval position within chromosomal bounds. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a8a5d2a1d1980dcfc238fca5f63765e1f">ModifyPos</a> (const char *position_op, long int position_shift)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position.  <a href="#a8a5d2a1d1980dcfc238fca5f63765e1f"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4dce925fdb55dcc7ef3dad890bc25408"></a><!-- doxytag: member="GenomicRegionSAM::Select" ref="a4dce925fdb55dcc7ef3dad890bc25408" args="(bool first, bool last, bool from5p, bool from3p)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a4dce925fdb55dcc7ef3dad890bc25408">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#afb2701ba1a521ae2c07ea0ace2f9ee77">ShiftPos</a> (long int start_shift, long int stop_shift, bool strand_aware)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately.  <a href="#afb2701ba1a521ae2c07ea0ace2f9ee77"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab31fed1760d543fa3a38bb16a064dcf9"></a><!-- doxytag: member="GenomicRegionSAM::RunShuffle" ref="ab31fed1760d543fa3a38bb16a064dcf9" args="(gsl_rng *random_generator, GenomicRegionSet *refReg, StringLIntMap *index, StringVecLIntMap *loc)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ab31fed1760d543fa3a38bb16a064dcf9">RunShuffle</a> (gsl_rng *random_generator, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints shuffled region within specified reference regions in <b>refReg</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8a27afdcdf11895d7b3f32bd8eed68ae"></a><!-- doxytag: member="GenomicRegionSAM::Sort" ref="a8a27afdcdf11895d7b3f32bd8eed68ae" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a8a27afdcdf11895d7b3f32bd8eed68ae">Sort</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Sort intervals according to start position (only for compatible intervals) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae499e532c831fdfe299d1f7ef877c730"></a><!-- doxytag: member="GenomicRegionSAM::RunSplit" ref="ae499e532c831fdfe299d1f7ef877c730" args="(bool by_chrom_and_strand)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ae499e532c831fdfe299d1f7ef877c730">RunSplit</a> (bool by_chrom_and_strand)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print region intervals on separate lines. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abeb76df49bb89ecf0f3074805d9fe7b2"></a><!-- doxytag: member="GenomicRegionSAM::ModifyStrand" ref="abeb76df49bb89ecf0f3074805d9fe7b2" args="(char *strand_op)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#abeb76df49bb89ecf0f3074805d9fe7b2">ModifyStrand</a> (char *strand_op)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Modified strand information: '+'=only forward, '-'=only reverse, 'r'=reverse strands from '+' to '-' and vice versa, 'b'=both strands. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6161ba10d2a4871cc43aa5bc597a787b"></a><!-- doxytag: member="GenomicRegionSAM::Union" ref="a6161ba10d2a4871cc43aa5bc597a787b" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a6161ba10d2a4871cc43aa5bc597a787b">Union</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute intervals' union. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a27027bc80a3b5a864ebafd415a6ef5a6"></a><!-- doxytag: member="GenomicRegionSAM::PrintWindows" ref="a27027bc80a3b5a864ebafd415a6ef5a6" args="(long int win_step, long int win_size)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a27027bc80a3b5a864ebafd415a6ef5a6">PrintWindows</a> (long int win_step, long int win_size)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions) <br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a281e35ef1bd0b7da811d9b8957f3d08f"></a><!-- doxytag: member="GenomicRegionSAM::n_tokens" ref="a281e35ef1bd0b7da811d9b8957f3d08f" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a281e35ef1bd0b7da811d9b8957f3d08f">n_tokens</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">number of TAB-separated fields in SAM entry (&gt;= 12) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a200d7f7044dd69311eceb851a3c315e5"></a><!-- doxytag: member="GenomicRegionSAM::FLAG" ref="a200d7f7044dd69311eceb851a3c315e5" args="" -->
unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a200d7f7044dd69311eceb851a3c315e5">FLAG</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">bitwise FLAG <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aedacdd8319f4b614229a41b80d4c6f67"></a><!-- doxytag: member="GenomicRegionSAM::MAPQ" ref="aedacdd8319f4b614229a41b80d4c6f67" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#aedacdd8319f4b614229a41b80d4c6f67">MAPQ</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">MAPping Quality. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa5acff481ae565db4976e66933212726"></a><!-- doxytag: member="GenomicRegionSAM::CIGAR" ref="aa5acff481ae565db4976e66933212726" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#aa5acff481ae565db4976e66933212726">CIGAR</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">CIGAR string. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af723505bfafff1a1fa616479023d0c3c"></a><!-- doxytag: member="GenomicRegionSAM::RNEXT" ref="af723505bfafff1a1fa616479023d0c3c" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#af723505bfafff1a1fa616479023d0c3c">RNEXT</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Reference name of the mate/next fragment. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad6f88b4324ba508b86102641d819661f"></a><!-- doxytag: member="GenomicRegionSAM::PNEXT" ref="ad6f88b4324ba508b86102641d819661f" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ad6f88b4324ba508b86102641d819661f">PNEXT</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Position of the mate/next fragment. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af180854a186be37eab49520edaa1262f"></a><!-- doxytag: member="GenomicRegionSAM::TLEN" ref="af180854a186be37eab49520edaa1262f" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#af180854a186be37eab49520edaa1262f">TLEN</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">observed Template LENgth <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6c2887ed1fa3a25d801641ea1b139032"></a><!-- doxytag: member="GenomicRegionSAM::SEQ" ref="a6c2887ed1fa3a25d801641ea1b139032" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a6c2887ed1fa3a25d801641ea1b139032">SEQ</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">fragment SEQuence <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab43ca5f2fbfdee520f1424c3a9fd0706"></a><!-- doxytag: member="GenomicRegionSAM::QUAL" ref="ab43ca5f2fbfdee520f1424c3a9fd0706" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#ab43ca5f2fbfdee520f1424c3a9fd0706">QUAL</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">ASCII of Phred-scaled base QUALity+33. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1622a8804462bad83b1c49d25eeacb49"></a><!-- doxytag: member="GenomicRegionSAM::OPTIONAL" ref="a1622a8804462bad83b1c49d25eeacb49" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a1622a8804462bad83b1c49d25eeacb49">OPTIONAL</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">all SAM fields after the 12th mandatory field <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a47afdde68b89cf4ea199bc91d4954a92"></a><!-- doxytag: member="GenomicRegionSAM::T" ref="a47afdde68b89cf4ea199bc91d4954a92" args="" -->
<a class="el" href="classGenomicRegionSAM.html#aadf8580c5362304ab4866fa76a59df96">CIGARTokens</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionSAM.html#a47afdde68b89cf4ea199bc91d4954a92">T</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">CIGAR tokens. <br/></td></tr>
</table>
<hr/><a name="details" id="details"></a><h2>Detailed Description</h2>
<div class="textblock"><p>This class processes genomic regions in SAM format. </p>
</div><hr/><h2>Member Function Documentation</h2>
<a class="anchor" id="a8a5d2a1d1980dcfc238fca5f63765e1f"></a><!-- doxytag: member="GenomicRegionSAM::ModifyPos" ref="a8a5d2a1d1980dcfc238fca5f63765e1f" args="(const char *position_op, long int position_shift)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSAM::ModifyPos </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>position_op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>position_shift</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">position_op</td><td>selects fixed position as follows: '1'=start position, 'c'=center position, '5p'=5-prime, '3p'=3-prime </td></tr>
    <tr><td class="paramname">position_shift</td><td>determines the shift of the non-fixed position (in the 'c' case, both positions are shifted symmetrically around the center) </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#a1277874633035d4f6fc9dabfb38998a2">GenomicRegion</a>.</p>

</div>
</div>
<a class="anchor" id="afb2701ba1a521ae2c07ea0ace2f9ee77"></a><!-- doxytag: member="GenomicRegionSAM::ShiftPos" ref="afb2701ba1a521ae2c07ea0ace2f9ee77" args="(long int start_shift, long int stop_shift, bool strand_aware)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionSAM::ShiftPos </td>
          <td>(</td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>start_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>stop_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>strand_aware</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">start_shift</td><td>determines the shift of the start position </td></tr>
    <tr><td class="paramname">stop_shift</td><td>determines the shift of the stop position </td></tr>
    <tr><td class="paramname">strand_aware</td><td>if 'true', then start=5-prime and stop=3-prime </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#a0ee8c165839c79afdc586f8b5e07788e">GenomicRegion</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a></li>
<li>genomic_intervals.cpp</li>
</ul>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Mon Jan 7 2013 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
