<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>GenomicRegionBED Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div id="top">
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a>  </div>
  <div class="headertitle">
<div class="title">GenomicRegionBED Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="GenomicRegionBED" --><!-- doxytag: inherits="GenomicRegion" -->
<p>This class processes genomic regions in BED format.  
 <a href="classGenomicRegionBED.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for GenomicRegionBED:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classGenomicRegionBED.png" usemap="#GenomicRegionBED_map" alt=""/>
  <map id="GenomicRegionBED_map" name="GenomicRegionBED_map">
<area href="classGenomicRegion.html" title="This class implements the genomic regions (i.e. ordered sets of genomic intervals)." alt="GenomicRegion" shape="rect" coords="0,0,125,24"/>
</map>
 </div></div>

<p><a href="classGenomicRegionBED-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a89d54edc5809d113202c9f09f82e1057"></a><!-- doxytag: member="GenomicRegionBED::GenomicRegionBED" ref="a89d54edc5809d113202c9f09f82e1057" args="(FileBuffer *B)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a89d54edc5809d113202c9f09f82e1057">GenomicRegionBED</a> (<a class="el" href="classFileBuffer.html">FileBuffer</a> *B)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor, see example below. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4d7c87327c1074258ac1e6b5ac87c9dd"></a><!-- doxytag: member="GenomicRegionBED::GenomicRegionBED" ref="a4d7c87327c1074258ac1e6b5ac87c9dd" args="(char *inp, long int n_line=0)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a4d7c87327c1074258ac1e6b5ac87c9dd">GenomicRegionBED</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor, see example below. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa6e68da10b635bcd84bf320b94a4d414"></a><!-- doxytag: member="GenomicRegionBED::GenomicRegionBED" ref="aa6e68da10b635bcd84bf320b94a4d414" args="(char *label, GenomicInterval *i)" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#aa6e68da10b635bcd84bf320b94a4d414">GenomicRegionBED</a> (char *label, <a class="el" href="classGenomicInterval.html">GenomicInterval</a> *i)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa27241b7fbdb90e0b4b1e411045d080f"></a><!-- doxytag: member="GenomicRegionBED::~GenomicRegionBED" ref="aa27241b7fbdb90e0b4b1e411045d080f" args="()" -->
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#aa27241b7fbdb90e0b4b1e411045d080f">~GenomicRegionBED</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Class destructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a707a1744e040214d78d37b928f1f0252"></a><!-- doxytag: member="GenomicRegionBED::Read" ref="a707a1744e040214d78d37b928f1f0252" args="(char *inp, long int n_line=0)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a707a1744e040214d78d37b928f1f0252">Read</a> (char *inp, long int <a class="el" href="classGenomicRegion.html#aefe2255aeed5338060190ded05cb9c0c">n_line</a>=0)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Reads BED format from input string <b>inp</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="affbd8eb279f10f9b45189e2388fe7e89"></a><!-- doxytag: member="GenomicRegionBED::Print" ref="affbd8eb279f10f9b45189e2388fe7e89" args="(FILE *file_ptr=stdout)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#affbd8eb279f10f9b45189e2388fe7e89">Print</a> (FILE *file_ptr=stdout)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints label and genomic region intervals with a newline character at the end. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae194ce2d67eafe222f10172ee376de94"></a><!-- doxytag: member="GenomicRegionBED::Print" ref="ae194ce2d67eafe222f10172ee376de94" args="(FILE *file_ptr, char *color, bool convert_chromosome)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#ae194ce2d67eafe222f10172ee376de94">Print</a> (FILE *file_ptr, char *color, bool convert_chromosome)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as generic <a class="el" href="classGenomicRegionBED.html#affbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a> except that it modifies the color and the chromosome names. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a04eceeb473bee498747bdd68d8ba8806"></a><!-- doxytag: member="GenomicRegionBED::PrintIntersection" ref="a04eceeb473bee498747bdd68d8ba8806" args="(GenomicRegion *r, bool ignore_strand, bool merge_labels)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a04eceeb473bee498747bdd68d8ba8806">PrintIntersection</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool ignore_strand, bool merge_labels)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all pairwise intersection with the intervals in region <b>r</b> (compatible, sorted and non-overlapping). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a25a8584259b6a8d47aa8c46795de1880"></a><!-- doxytag: member="GenomicRegionBED::PrintConstrained" ref="a25a8584259b6a8d47aa8c46795de1880" args="(GenomicRegion *r, bool merge_labels=false)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a25a8584259b6a8d47aa8c46795de1880">PrintConstrained</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, bool merge_labels=false)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionBED.html#affbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the interval is constrained between <b>start</b> and <b>stop</b> <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aaebba3c31c0fa88e594706f490030549"></a><!-- doxytag: member="GenomicRegionBED::PrintModified" ref="aaebba3c31c0fa88e594706f490030549" args="(char *label, long int start, long int stop)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#aaebba3c31c0fa88e594706f490030549">PrintModified</a> (char *label, long int start, long int stop)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Same as <a class="el" href="classGenomicRegionBED.html#affbd8eb279f10f9b45189e2388fe7e89" title="Prints label and genomic region intervals with a newline character at the end.">Print()</a>, but the label and interval's coordinates are modified. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a01bd886cfa67a3adae10c67dee941924"></a><!-- doxytag: member="GenomicRegionBED::UpdateThick" ref="a01bd886cfa67a3adae10c67dee941924" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a01bd886cfa67a3adae10c67dee941924">UpdateThick</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Auxiliary function: updates thickStart/thickEnd variables to honor changes in coordinates. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a513fb071e2626be58837052daf86bab5"></a><!-- doxytag: member="GenomicRegionBED::PrintBEDFormat" ref="a513fb071e2626be58837052daf86bab5" args="(char *color, bool convert_chromosome)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a513fb071e2626be58837052daf86bab5">PrintBEDFormat</a> (char *color, bool convert_chromosome)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints in BED format. If <b>convert_chromosome</b> is 'true', then convert from ENSEMBL to UCSC names (by adding 'chr' prefix to each chromosome name, and by converting 'MT' to 'chrM'). <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad9ec9c8dd69ee078fde940463ad46e1e"></a><!-- doxytag: member="GenomicRegionBED::ApplyBounds" ref="ad9ec9c8dd69ee078fde940463ad46e1e" args="(StringLIntMap *bounds)" -->
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#ad9ec9c8dd69ee078fde940463ad46e1e">ApplyBounds</a> (StringLIntMap *bounds)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Corrects interval start/stop positions so as to comply with chromosomal bounds. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a35c692c1343327d9cb450cfa1276096f"></a><!-- doxytag: member="GenomicRegionBED::Center" ref="a35c692c1343327d9cb450cfa1276096f" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a35c692c1343327d9cb450cfa1276096f">Center</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace each interval in the region with the corresponding center interval. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad55a9ffea8b0924bac3d11a5722cdfd2"></a><!-- doxytag: member="GenomicRegionBED::Connect" ref="ad55a9ffea8b0924bac3d11a5722cdfd2" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#ad55a9ffea8b0924bac3d11a5722cdfd2">Connect</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace region with a single interval from minimum start to maximum stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9580179a120834e71ac90e56e5f35aa9"></a><!-- doxytag: member="GenomicRegionBED::Diff" ref="a9580179a120834e71ac90e56e5f35aa9" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a9580179a120834e71ac90e56e5f35aa9">Diff</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace regions with the corresponding gap intervals between successive intervals (e.g. used to compute intron boundaries) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a833d50369e61f64603c63b6a05f3d80b"></a><!-- doxytag: member="GenomicRegionBED::RunCalcDistances" ref="a833d50369e61f64603c63b6a05f3d80b" args="(char *op1, char *op2)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a833d50369e61f64603c63b6a05f3d80b">RunCalcDistances</a> (char *op1, char *op2)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print distance between successive intervals. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abe056ec84d2d6890479d47559839bab4"></a><!-- doxytag: member="GenomicRegionBED::Divide" ref="abe056ec84d2d6890479d47559839bab4" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#abe056ec84d2d6890479d47559839bab4">Divide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5f456b4af154c98968370568828d6e3d"></a><!-- doxytag: member="GenomicRegionBED::RunDivide" ref="a5f456b4af154c98968370568828d6e3d" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a5f456b4af154c98968370568828d6e3d">RunDivide</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intervals divided in half (i.e. two intervals for each original interval in the genomic region) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aac069dcf513265f7756b72c106f2765c"></a><!-- doxytag: member="GenomicRegionBED::Fix" ref="aac069dcf513265f7756b72c106f2765c" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#aac069dcf513265f7756b72c106f2765c">Fix</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes rogue intervals (e.g. start&gt;stop) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="acf815a65712e82bf3f8c596c795f55c4"></a><!-- doxytag: member="GenomicRegionBED::Intersect" ref="acf815a65712e82bf3f8c596c795f55c4" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#acf815a65712e82bf3f8c596c795f55c4">Intersect</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print intersection among all intervals, i.e. maximum start to minimum stop position. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#ac515c70f443db400f911452ed359433b">ModifyPos</a> (char *position_op, long int position_shift)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position.  <a href="#ac515c70f443db400f911452ed359433b"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a920d47deb095506c311da2b2324e7a14"></a><!-- doxytag: member="GenomicRegionBED::Randomize" ref="a920d47deb095506c311da2b2324e7a14" args="(gsl_rng *random_generator, StringLIntMap *bounds)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a920d47deb095506c311da2b2324e7a14">Randomize</a> (gsl_rng *random_generator, StringLIntMap *bounds)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Randomizes interval position within chromosomal bounds. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adec7408315765bbdb21465a26400422b"></a><!-- doxytag: member="GenomicRegionBED::Select" ref="adec7408315765bbdb21465a26400422b" args="(bool first, bool last, bool from5p, bool from3p)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#adec7408315765bbdb21465a26400422b">Select</a> (bool first, bool last, bool from5p, bool from3p)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Selects a subset of intervals according to their relative positions. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#afd2f05cec2af5186794f40f221e040b1">ShiftPos</a> (long int start_shift, long int stop_shift, bool strand_aware)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately.  <a href="#afd2f05cec2af5186794f40f221e040b1"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af3793387aa335bf68042f9245bc5992d"></a><!-- doxytag: member="GenomicRegionBED::RunShuffle" ref="af3793387aa335bf68042f9245bc5992d" args="(gsl_rng *random_generator, GenomicRegionSet *refReg, StringLIntMap *index, StringVecLIntMap *loc)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#af3793387aa335bf68042f9245bc5992d">RunShuffle</a> (gsl_rng *random_generator, <a class="el" href="classGenomicRegionSet.html">GenomicRegionSet</a> *refReg, StringLIntMap *index, StringVecLIntMap *loc)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints shuffled region within specified reference regions in <b>refReg</b>. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afd1a61c04f76782abdc83074ae0f7bd3"></a><!-- doxytag: member="GenomicRegionBED::Sort" ref="afd1a61c04f76782abdc83074ae0f7bd3" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#afd1a61c04f76782abdc83074ae0f7bd3">Sort</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Sort intervals according to start position (only for compatible intervals) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abd3b8e4969c2ff2639eb344de7b224f9"></a><!-- doxytag: member="GenomicRegionBED::RunSplit" ref="abd3b8e4969c2ff2639eb344de7b224f9" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#abd3b8e4969c2ff2639eb344de7b224f9">RunSplit</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print region intervals on separate lines. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a605f5bf86254547d117ed0c3f5c64587"></a><!-- doxytag: member="GenomicRegionBED::Union" ref="a605f5bf86254547d117ed0c3f5c64587" args="()" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a605f5bf86254547d117ed0c3f5c64587">Union</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute intervals' union. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3ebdc24dd97c7b106205d932970de535"></a><!-- doxytag: member="GenomicRegionBED::PrintWindows" ref="a3ebdc24dd97c7b106205d932970de535" args="(long int win_step, long int win_size)" -->
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a3ebdc24dd97c7b106205d932970de535">PrintWindows</a> (long int win_step, long int win_size)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Print sliding windows (implemented only for single-interval regions) <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0c4b2a8a590ce1dfbef142d03d53fe5d"></a><!-- doxytag: member="GenomicRegionBED::Constrain" ref="a0c4b2a8a590ce1dfbef142d03d53fe5d" args="(GenomicRegion *r, char *label=NULL)" -->
<a class="el" href="classGenomicRegionBED.html">GenomicRegionBED</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a0c4b2a8a590ce1dfbef142d03d53fe5d">Constrain</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r, char *label=NULL)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns a version of this genomic region constrained by the start/stop coordinates. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6db4aa9e6d1655bf116095d295b27269"></a><!-- doxytag: member="GenomicRegionBED::Diff" ref="a6db4aa9e6d1655bf116095d295b27269" args="(GenomicRegion *r)" -->
<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a6db4aa9e6d1655bf116095d295b27269">Diff</a> (<a class="el" href="classGenomicRegion.html">GenomicRegion</a> *r)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the difference of this genomic region and region <b>r</b>. <br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a60878391f6fb1e0ad0ac9bc88a7625d3"></a><!-- doxytag: member="GenomicRegionBED::n_tokens" ref="a60878391f6fb1e0ad0ac9bc88a7625d3" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a60878391f6fb1e0ad0ac9bc88a7625d3">n_tokens</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">number of tokens <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afa7ba810a0d1673f5dbabcf85c7bbb95"></a><!-- doxytag: member="GenomicRegionBED::score" ref="afa7ba810a0d1673f5dbabcf85c7bbb95" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#afa7ba810a0d1673f5dbabcf85c7bbb95">score</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">score field <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a52fc75b347dfe6696e7ad2beeeb623a9"></a><!-- doxytag: member="GenomicRegionBED::itemRgb" ref="a52fc75b347dfe6696e7ad2beeeb623a9" args="" -->
char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a52fc75b347dfe6696e7ad2beeeb623a9">itemRgb</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">color field <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a019dfaebedc17e8859b36c3359287faa"></a><!-- doxytag: member="GenomicRegionBED::thickStart" ref="a019dfaebedc17e8859b36c3359287faa" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a019dfaebedc17e8859b36c3359287faa">thickStart</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">thickStart field <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5333a7fdc1f21db508aa7af4f50553ba"></a><!-- doxytag: member="GenomicRegionBED::thickEnd" ref="a5333a7fdc1f21db508aa7af4f50553ba" args="" -->
long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionBED.html#a5333a7fdc1f21db508aa7af4f50553ba">thickEnd</a></td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">thickEnd field <br/></td></tr>
</table>
<hr/><a name="details" id="details"></a><h2>Detailed Description</h2>
<div class="textblock"><p>This class processes genomic regions in BED format. </p>
<p>Example: </p>
<div class="fragment"><pre class="fragment">    <a class="code" href="classGenomicRegionBED.html" title="This class processes genomic regions in BED format.">GenomicRegionBED</a> x(<span class="stringliteral">&quot;chr1 132033 140102 Region#1 1000 + 132033 140102 0 2 3899,4104 0,3965&quot;</span>);
    cout &lt;&lt; <span class="stringliteral">&quot;x = \n&quot;</span>; x.Print();
</pre></div> </div><hr/><h2>Member Function Documentation</h2>
<a class="anchor" id="ac515c70f443db400f911452ed359433b"></a><!-- doxytag: member="GenomicRegionBED::ModifyPos" ref="ac515c70f443db400f911452ed359433b" args="(char *position_op, long int position_shift)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionBED::ModifyPos </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>position_op</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>position_shift</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Modifies interval start/stop positions. First it chooses which position to keep fixed, and then it shifts the non-fixed position. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">position_op</td><td>selects fixed position as follows: '1'=start position, 'c'=center position, '5p'=5-prime, '3p'=3-prime </td></tr>
    <tr><td class="paramname">position_shift</td><td>determines the shift of the non-fixed position (in the 'c' case, both positions are shifted symmetrically around the center) </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#a0721b07af0850057e4ab9cd416ecac2f">GenomicRegion</a>.</p>

</div>
</div>
<a class="anchor" id="afd2f05cec2af5186794f40f221e040b1"></a><!-- doxytag: member="GenomicRegionBED::ShiftPos" ref="afd2f05cec2af5186794f40f221e040b1" args="(long int start_shift, long int stop_shift, bool strand_aware)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionBED::ShiftPos </td>
          <td>(</td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>start_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long int&#160;</td>
          <td class="paramname"><em>stop_shift</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>strand_aware</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td><code> [virtual]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Shifts interval start/stop positions. First it chooses reference position, and then shifts start and stop positions separately. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">start_shift</td><td>determines the shift of the start position </td></tr>
    <tr><td class="paramname">stop_shift</td><td>determines the shift of the stop position </td></tr>
    <tr><td class="paramname">strand_aware</td><td>if 'true', then start=5-prime and stop=3-prime </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented from <a class="el" href="classGenomicRegion.html#a0ee8c165839c79afdc586f8b5e07788e">GenomicRegion</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h_source.html">genomic_intervals.h</a></li>
<li>genomic_intervals.cpp</li>
</ul>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr class="footer"/><address class="footer"><small>Generated on Thu Apr 5 2012 09:34:48 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </small></address>
</body>
</html>
