<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>Chromosomes Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.6 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    <li>
      <form action="search.php" method="get">
        <table cellspacing="0" cellpadding="0" border="0">
          <tr>
            <td><label>&nbsp;<u>S</u>earch&nbsp;for&nbsp;</label></td>
            <td><input type="text" name="query" value="" size="20" accesskey="s"/></td>
          </tr>
        </table>
      </form>
    </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>Chromosomes Class Reference</h1><!-- doxytag: class="Chromosomes" -->This class is used to store chromosome DNA sequences.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;<a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a>&gt;</code>
<p>

<p>
<a href="classChromosomes-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Types</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4a233d2b7fabc1a589cbbafd5d0ea243"></a><!-- doxytag: member="Chromosomes::ChromosomeSeqData" ref="4a233d2b7fabc1a589cbbafd5d0ea243" args="" -->
typedef <a class="el" href="classChromosomes.html#8e40e7c99c6c0a40713c7785ba5b5ac4">map</a>&lt; string, pair<br>
&lt; char *, size_t &gt; &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#4a233d2b7fabc1a589cbbafd5d0ea243">ChromosomeSeqData</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">for every chromosome, store a pointer to its sequence, and its size. <br></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#a596430d4a05ce0dc4d4f43c650db83e">Chromosomes</a> (char *<a class="el" href="classChromosomes.html#8df459fcb5ad98a2524fd51ac2b5a60c">map_dir</a>, char *<a class="el" href="classChromosomes.html#9c690f19e04e80b056dac2072476f24c">map_name</a>)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor.  <a href="#a596430d4a05ce0dc4d4f43c650db83e"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#bb3466c6c94f42f22a872b043c415b67">Chromosomes</a> (char *fasta_file, bool <a class="el" href="classChromosomes.html#954bae240564f6f589a30d253291eb75">verbose</a>)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class constructor.  <a href="#bb3466c6c94f42f22a872b043c415b67"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="0b04a0781b2ff7088dd8046493385003"></a><!-- doxytag: member="Chromosomes::~Chromosomes" ref="0b04a0781b2ff7088dd8046493385003" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#0b04a0781b2ff7088dd8046493385003">~Chromosomes</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Class destructor. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="2378b5cab810d50699d97883fe1c04bf"></a><!-- doxytag: member="Chromosomes::FindChromosome" ref="2378b5cab810d50699d97883fe1c04bf" args="(char *chromosome_name)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#2378b5cab810d50699d97883fe1c04bf">FindChromosome</a> (char *chromosome_name)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns 'true' if chromosome name is found in the data. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="819d0fa5f41c034a45949156307d02e4"></a><!-- doxytag: member="Chromosomes::GetSeq" ref="819d0fa5f41c034a45949156307d02e4" args="(GenomicInterval *I, bool replace=false)" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#819d0fa5f41c034a45949156307d02e4">GetSeq</a> (<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *I, bool replace=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns interval sequence. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="c732efe7b6dcc645861507257bf66347"></a><!-- doxytag: member="Chromosomes::PrintSeq" ref="c732efe7b6dcc645861507257bf66347" args="(GenomicInterval *I, bool replace=false)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#c732efe7b6dcc645861507257bf66347">PrintSeq</a> (<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *I, bool replace=false)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints interval sequence. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="044da4339ffcc9a892ea35198bbb9f36"></a><!-- doxytag: member="Chromosomes::GetSeqLength" ref="044da4339ffcc9a892ea35198bbb9f36" args="(GenomicInterval *I)" -->
size_t&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#044da4339ffcc9a892ea35198bbb9f36">GetSeqLength</a> (<a class="el" href="classGenomicInterval.html">GenomicInterval</a> *I)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates interval sequence length excluding 'N' characters. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7f4e8938bd27e1fdb81f7ab33a44d62a"></a><!-- doxytag: member="Chromosomes::LoadChromosome" ref="7f4e8938bd27e1fdb81f7ab33a44d62a" args="(string chromosome_name)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#7f4e8938bd27e1fdb81f7ab33a44d62a">LoadChromosome</a> (string chromosome_name)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Loads chromosome sequence in memory. <br></td></tr>
<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="954bae240564f6f589a30d253291eb75"></a><!-- doxytag: member="Chromosomes::verbose" ref="954bae240564f6f589a30d253291eb75" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#954bae240564f6f589a30d253291eb75">verbose</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">if 'true', print messages while loading chromosome sequences <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="408a06b9ae635ceaf0b75452bcb62763"></a><!-- doxytag: member="Chromosomes::load_in_memory" ref="408a06b9ae635ceaf0b75452bcb62763" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#408a06b9ae635ceaf0b75452bcb62763">load_in_memory</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">'true' if reading from FASTA file <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8e40e7c99c6c0a40713c7785ba5b5ac4"></a><!-- doxytag: member="Chromosomes::map" ref="8e40e7c99c6c0a40713c7785ba5b5ac4" args="" -->
StringMap&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#8e40e7c99c6c0a40713c7785ba5b5ac4">map</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">maps chromosome names to chromosome sequence file names <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="8df459fcb5ad98a2524fd51ac2b5a60c"></a><!-- doxytag: member="Chromosomes::map_dir" ref="8df459fcb5ad98a2524fd51ac2b5a60c" args="" -->
string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#8df459fcb5ad98a2524fd51ac2b5a60c">map_dir</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">the directory where the chromosome map file is located <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9c690f19e04e80b056dac2072476f24c"></a><!-- doxytag: member="Chromosomes::map_name" ref="9c690f19e04e80b056dac2072476f24c" args="" -->
string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#9c690f19e04e80b056dac2072476f24c">map_name</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">the name of the chromosome map file <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="b3b502eec7bd2028297300aeb47b5844"></a><!-- doxytag: member="Chromosomes::current_chromosome_name" ref="b3b502eec7bd2028297300aeb47b5844" args="" -->
string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#b3b502eec7bd2028297300aeb47b5844">current_chromosome_name</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">name of the currently used chromosome <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="fb59cb80755187f4546573f10110e669"></a><!-- doxytag: member="Chromosomes::current_chromosome_size" ref="fb59cb80755187f4546573f10110e669" args="" -->
size_t&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#fb59cb80755187f4546573f10110e669">current_chromosome_size</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">size of the currently used chromosome <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f640cf0f09b109b3d46d70aba9207a2e"></a><!-- doxytag: member="Chromosomes::current_chromosome_seq" ref="f640cf0f09b109b3d46d70aba9207a2e" args="" -->
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#f640cf0f09b109b3d46d70aba9207a2e">current_chromosome_seq</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">pointer to the currently used chromosome sequence <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="673169d55aa0b94815e17b41566d4d8c"></a><!-- doxytag: member="Chromosomes::chromosome_seq_data" ref="673169d55aa0b94815e17b41566d4d8c" args="" -->
<a class="el" href="classChromosomes.html#4a233d2b7fabc1a589cbbafd5d0ea243">ChromosomeSeqData</a>&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classChromosomes.html#673169d55aa0b94815e17b41566d4d8c">chromosome_seq_data</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">all chromosomes sequences and their sizes are stored here <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
This class is used to store chromosome DNA sequences. <hr><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" name="a596430d4a05ce0dc4d4f43c650db83e"></a><!-- doxytag: member="Chromosomes::Chromosomes" ref="a596430d4a05ce0dc4d4f43c650db83e" args="(char *map_dir, char *map_name)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Chromosomes::Chromosomes           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>map_dir</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>map_name</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Class constructor. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>map_dir</em>&nbsp;</td><td>the directory where the chromosome map file is located </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>map_name</em>&nbsp;</td><td>the name of the chromosome map file</td></tr>
  </table>
</dl>
Chromosome sequences are loaded on the fly from the *.dna files. For efficient sequence extraction, genomic intervals must be grouped by chromosome. If not, use the alternative constructor described below.<p>
Example of map file (stored in <a class="el" href="classChromosomes.html#9c690f19e04e80b056dac2072476f24c">map_name</a>):<p>
1 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.1.fa.dna <br>
 10 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.10.fa.dna <br>
 11 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.11.fa.dna <br>
 12 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.12.fa.dna <br>
 13 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.13.fa.dna <br>
 14 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.14.fa.dna <br>
 15 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.15.fa.dna <br>
 16 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.16.fa.dna <br>
 17 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.17.fa.dna <br>
 18 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.18.fa.dna <br>
 19 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.19.fa.dna <br>
 2 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.2.fa.dna <br>
 3 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.3.fa.dna <br>
 4 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.4.fa.dna <br>
 5 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.5.fa.dna <br>
 6 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.6.fa.dna <br>
 7 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.7.fa.dna <br>
 8 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.8.fa.dna <br>
 9 <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.9.fa.dna <br>
 MT <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.MT.fa.dna <br>
 X <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.X.fa.dna <br>
 Y <em>&lt;TAB&gt;</em> Mus_musculus.NCBIM37.52.dna.chromosome.Y.fa.dna <br>
<p>
Each *.dna file contains the chromosome sequence in one line. 
</div>
</div><p>
<a class="anchor" name="bb3466c6c94f42f22a872b043c415b67"></a><!-- doxytag: member="Chromosomes::Chromosomes" ref="bb3466c6c94f42f22a872b043c415b67" args="(char *fasta_file, bool verbose)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Chromosomes::Chromosomes           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>fasta_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&nbsp;</td>
          <td class="paramname"> <em>verbose</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Class constructor. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>fasta_file</em>&nbsp;</td><td>FASTA file containing chromosome sequences </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>verbose</em>&nbsp;</td><td>if 'true', print messages while loading chromosome sequences</td></tr>
  </table>
</dl>
Chromosome sequences are loaded in memory from the FASTA file. Example FASTA file:<p>
&gt;chromosome_1 <br>
 TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTA <br>
 &gt;chromosome_2 <br>
 CGTATCCCACACACCACACCCACACACCACACCCACACACACCCACACCC <br>
 &gt;chromosome_3 <br>
 TCCTCCCAGAATCTGGAGAGGTCAACCTGTTCTTCAAAGCAGTGGTGGAT <br>
 &gt;chromosome_4 <br>
 ACCCTAACCCTAACCCTAACCCTAACCCTAACCCTACCCTAACCCTAACC <br>
 
</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="genomic__intervals_8h-source.html">genomic_intervals.h</a><li>genomic_intervals.cpp</ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Fri Sep 16 19:12:57 2011 by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.6 </small></address>
</body>
</html>
